<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Dan McGuire">
<meta name="dcterms.date" content="2025-10-17">
<meta name="description" content="Efficient cell typing using multiple analytes">

<title>Integrated RNA + protein multiomic cell typing using HieraType – Blog</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/logo_Bruker_black.png" rel="icon" type="image/png">
<script src="../../site_libs/cookie-consent/cookie-consent.js"></script>
<link href="../../site_libs/cookie-consent/cookie-consent.css" rel="stylesheet">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-3a01e2046221230fdceeea94b1ec5d67.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-a4a11d514c7d463668e07712114998e6.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-8dd817d487e8e26707729065bc6b7e6d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-29W4MW0Y2W"></script>

<script type="text/plain" cookie-consent="tracking">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-29W4MW0Y2W', { 'anonymize_ip': true});
</script>

<script type="text/javascript" charset="UTF-8">
document.addEventListener('DOMContentLoaded', function () {
cookieconsent.run({
  "notice_banner_type":"simple",
  "consent_type":"implied",
  "palette":"light",
  "language":"en",
  "page_load_consent_levels":["strictly-necessary","functionality","tracking","targeting"],
  "notice_banner_reject_button_hide":false,
  "preferences_center_close_button_hide":false,
  "website_name":""
  ,
"language":"en"
  });
});
</script> 
  

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../.././assets/logo_Bruker_white.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../license.html"> 
<span class="menu-text">License</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../link-to-code.html"> 
<span class="menu-text">Code</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">Browse Posts by Topic</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/company/bruker-spatial"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Integrated RNA + protein multiomic cell typing using HieraType</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
                  <div>
        <div class="description">
          Efficient cell typing using multiple analytes
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">cell typing</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliations</div>
    
      <div class="quarto-title-meta-contents">
      <p class="author">Dan McGuire <a href="https://orcid.org/0009-0006-0286-9625" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              Bruker Spatial Biology
            </p>
          <p class="affiliation">
              Github: <a href="https://github.com/dan11mcguire" target="_blank">dan11mcguire</a>
            </p>
        </div>
    </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">October 17, 2025</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2">
    <h2 id="toc-title">Contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#a-quick-look-at-the-tissue" id="toc-a-quick-look-at-the-tissue" class="nav-link" data-scroll-target="#a-quick-look-at-the-tissue"><span class="header-section-number">2</span> A quick look at the tissue</a></li>
  <li><a href="#evidence-multiomics-can-improve-cell-typing" id="toc-evidence-multiomics-can-improve-cell-typing" class="nav-link" data-scroll-target="#evidence-multiomics-can-improve-cell-typing"><span class="header-section-number">3</span> Evidence multiomics can improve cell typing</a>
  <ul class="collapse">
  <li><a href="#t-cell-improvement-demonstration-via-area-under-the-roc-curves" id="toc-t-cell-improvement-demonstration-via-area-under-the-roc-curves" class="nav-link" data-scroll-target="#t-cell-improvement-demonstration-via-area-under-the-roc-curves"><span class="header-section-number">3.0.1</span> T cell improvement demonstration via area under the ROC curves</a></li>
  <li><a href="#improved-auc-across-immune-cell-types" id="toc-improved-auc-across-immune-cell-types" class="nav-link" data-scroll-target="#improved-auc-across-immune-cell-types"><span class="header-section-number">3.0.2</span> Improved AUC across immune cell types</a></li>
  </ul></li>
  <li><a href="#multiomic" id="toc-multiomic" class="nav-link" data-scroll-target="#multiomic"><span class="header-section-number">4</span> Running HieraType using Protein + RNA</a></li>
  <li><a href="#integrating-unsupervised-clusters-with-immune-cell-type-labels" id="toc-integrating-unsupervised-clusters-with-immune-cell-type-labels" class="nav-link" data-scroll-target="#integrating-unsupervised-clusters-with-immune-cell-type-labels"><span class="header-section-number">5</span> Integrating unsupervised clusters with immune cell type labels</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion"><span class="header-section-number">6</span> Conclusion</a></li>
  <li><a href="#endtoend" id="toc-endtoend" class="nav-link" data-scroll-target="#endtoend"><span class="header-section-number">7</span> End-to-end detailed analysis code</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Introduction</h1>
<p>This post showcases how the <a href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/HieraType">HieraType cell typing package</a> can be used for integrated same-slide multiomic (protein + rna) cell typing.</p>
<p>We use a breast cancer tissue sample with both 64-plex protein and WTx 18,935 plex RNA assayed, and demonstrate how cell typing can be performed using</p>
<ul>
<li>both RNA + protein</li>
<li>protein-only</li>
<li>RNA-only</li>
</ul>
<p>We compare cell typing performance and show that, overall, protein-only and RNA-only cell typing calls show more disagreement with each other than they do with with integrated multiomic cell type calls.</p>
<p>We also demonstrate evidence of improved coherence of cell type labels with the defining marker genes and proteins when using an integrated approach that makes use of both analytes.</p>
<p>Finally, we show a simple yet effective way to integrate unsupervised clusters with supervised immune cell type labels generated with HieraType. This can enable granular detection of novel or tissue-specific cell types along with well known immune cell types.</p>
<p>Code for this analysis end-to-end can be found by skipping ahead <a href="#endtoend">here</a>.</p>
</section>
<section id="a-quick-look-at-the-tissue" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> A quick look at the tissue</h1>
<p>Here’s a quick look at this breast cancer tissue, where unsupervised clustering was performed using both the pearson residual normalized RNA + protein. Code for creating these unsupervised clusters is <a href="#multiunsup">here</a>. Next, we’ll compare multiomic immune cell typing performance with RNA-only or protein-only cell typing.</p>
<div class="wider-image">
<p><img src="./figures/xyp_multi_unsup.png" alt="Wide Image"></p>
</div>
</section>
<section id="evidence-multiomics-can-improve-cell-typing" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Evidence multiomics can improve cell typing</h1>
<p>We used the Adjusted Rand Index (ARI) metric to quantify the agreement between protein-only, RNA-only, and multiomic cell type labels (<a href="#ari">code</a>). ARI is a numerical measure of similarity between two data clusterings <span class="citation" data-cites="ariref">(<a href="#ref-ariref" role="doc-biblioref">Rand 1971</a>)</span>. We found protein-only and RNA-only cell typing calls show more disagreement with each other than they do with with integrated multiomic cell type calls, with:</p>
<ul>
<li>ARI(multiomic,RNA-only) = 0.679</li>
<li>ARI(multiomic, protein-only) = 0.589</li>
<li>ARI(protein-only, RNA-only) = 0.483</li>
</ul>
<p>We compared the agreement between celltype predictions and canonical protein markers and RNA markers, and found that the multiomic cell type labels often show stronger alignment than RNA-only or protein-only cell typing. ( Code for <a href="#rnaonly">RNA-only</a>, <a href="#proteinonly">protein-only</a>, <a href="#multiomic">multiomic</a> HieraType calls.)</p>
<section id="t-cell-improvement-demonstration-via-area-under-the-roc-curves" class="level3" data-number="3.0.1">
<h3 data-number="3.0.1" class="anchored" data-anchor-id="t-cell-improvement-demonstration-via-area-under-the-roc-curves"><span class="header-section-number">3.0.1</span> T cell improvement demonstration via area under the ROC curves</h3>
<p>For example, the figure below shows ROC curves for T cell predictions along with three canonical cell type markers for T cells (one protein, and two RNA). In this context, for a given level of marker expression, ‘<em>sensitivity</em>’ refers to the frequency a cell is labeled a T-cell (for all cells above that threshold of marker expression). ‘<em>Specificity</em>’ refers to the proportion of times a cell is labeled anything other than a T-cell (for all cells below that threshold).<br>
If we treat our normalized marker expression as a gold standard for cell type calling, then we would want to see high sensitivity and specificity of the T cell calls depending on whether marker expression is high or low. In general, larger area under the ROC curve indicates better agreement between cell type calls and marker expression.</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/rocplot_tcell_lbl.png" class="img-fluid figure-img" width="6000"></p>
</figure>
</div>
</div>
</div>
<p>Above, as we might expect, we see that using <span style="color:#377EB8;">RNA-only</span> for calling T cells does not optimize agreement between CD3 protein (AUC=0.871). Unsurprisingly, agreement improves when using CD3 protein directly for celltyping (<span style="color:#4DAF4A;">protein-only</span> AUC=0.927). In this case, we actually observed the best AUC when both RNA and protein were used (<span style="color:#E41A1C;">multiomics</span> AUC=0.937).</p>
<p>A perhaps unsurprising theme that emerged from these cell typing comparisons was that we often see the worst agreement for protein markers when using RNA-only for celltyping, and the worst agreement for RNA markers when using protein-only for celltyping. Meanwhile, using both RNA and protein analytes often yields the best or nearly-best agreement for both marker types. This is further shown below in the middle-and-right panels when looking at CD3D and CD2 RNA markers. <span style="color:#E41A1C;">Multiomic</span> celltyping outperforms protein-only by the widest margin, and also slightly outperforms <span style="color:#377EB8;">RNA-only</span>.</p>
<p>One interesting point of note when comparing the shapes of the ROC curves above is the smoothness of the curve for the CD3 protein marker, compared to the CD2 and CD3D RNA markers (which show a sharper bump closer to the lower-left corner where sensitivity=0 and specificity=1). Due to the sparsity of RNA counts, we see a sharp spike in sensitivity close to an expression threshold at which cells have a single count of the marker gene. Because RNA-only cell typing uses only the RNA markers, the AUC is larger and more influenced around this local threshold of a single marker gene count, while protein and multiomic cell type labels are impacted directly by marker protein expression.</p>
</section>
<section id="improved-auc-across-immune-cell-types" class="level3" data-number="3.0.2">
<h3 data-number="3.0.2" class="anchored" data-anchor-id="improved-auc-across-immune-cell-types"><span class="header-section-number">3.0.2</span> Improved AUC across immune cell types</h3>
<p>We saw a similar trend as that demonstrated above across immune cell types.<br>
Here, for the most important ‘index markers’ of each immune cell type, we show the area under the ROC curve for RNA-only, protein-only, and multiomic celltyping results (code for this analysis can be found here in the <a href="#endtoend">end-to-end</a> code section). In general, there appears to be an improvement in AUC for both RNA and protein markers compared to if one of the analytes were removed.</p>
<p>Average AUC was highest for multiomic celltyping across all markers and celltypes, with</p>
<ul>
<li><span class="math inline">\(\overline{AUC}(\text{multiomic}) = 0.711\)</span></li>
<li><span class="math inline">\(\overline{AUC}(\text{RNA-only}) = 0.709\)</span></li>
<li><span class="math inline">\(\overline{AUC}(\text{protein-only}) = 0.701\)</span></li>
</ul>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/aucbarplot_wtext.png" class="img-fluid figure-img" width="6000"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="multiomic" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Running HieraType using Protein + RNA</h1>
<p>Below, we’ll step through the process of generating cell type labels with multiomic data. The code for using HieraType with protein data uses very similar syntax as a regular HieraType call, with a few exceptions worth calling out.</p>
<section id="protein-naming-convention" class="level5">
<h5 class="anchored" data-anchor-id="protein-naming-convention">Protein naming convention</h5>
<p>HieraType works from a set of pre-defined ‘markers’ which correspond to each cell type we hope to classify. For multiomic or protein-only data, protein marker names must end with the suffix “_protein” in our <code>markerslist</code> objects.<br>
For example, here is the HieraType package multiomic markerslist for CD8 T cells (<code>cd8t</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(HieraType<span class="sc">::</span>markerslist_multiomic_tcellmajor<span class="sc">$</span>cd8t)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>$index_marker
[1] "CD8A"        "CD8B"        "CD8_protein"

$predictors
 [1] "CD8_protein"    "CD3_protein"    "CCR7_protein"   "CD45RA_protein" "CD27_protein"   "GZMA_protein"  
 [7] "GZMB_protein"   "LAG3_protein"   "TCF7_protein"   "CD3D"           "CD3E"           "CD3G"          
[13] "CD8A"           "CD8B"           "IL7R"           "CD27"           "SELL"           "CCR7"          
[19] "TCF7"           "LEF1"           "FOXP1"          "BACH2"          "KLF2"           "CD28"          
[25] "ICOS"           "STAT3"          "STAT5"          "RUNX3"          "BATF"           "EOMES"         
[31] "TBX21"          "PRDM1"          "PDCD1"          "LAG3"           "TIGIT"          "HAVCR2"        
[37] "GZMK"           "GZMA"           "GZMB"           "PRF1"           "GNLY"           "KLRG1"         
[43] "KLRD1"          "IFNG"           "CCL5"           "CX3CR1"         "NKG7"           "FAS"           
[49] "FASLG"          "IL2"            "CD244"          "CXCR3"          "KLRB1"          "TOX"           
[55] "NR4A2"          "BCL6"           "CXCL13"        

$use_offclass_markers_as_negative_predictors
[1] TRUE</code></pre>
</div>
</div>
</section>
<section id="protein-data-normalization-and-constructing-a-multiomic-adjacency_matrix" class="level5">
<h5 class="anchored" data-anchor-id="protein-data-normalization-and-constructing-a-multiomic-adjacency_matrix">Protein data normalization and constructing a multiomic adjacency_matrix</h5>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Below, we make use of the <a href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/scPearsonPCA">scPearsonPCA package</a>. This R package provides convenience functions for getting the principal component embeddings for pearson-residual normalized data directly from a raw counts matrix. When normalized protein is also available, we can get the PC decomposition from both RNA and normalized protein using the functions below. For more detail on the package, please check out the <a href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/scPearsonPCA">README</a>.</p>
</div>
</div>
<p>Protein data should be normalized, then combined with a sparse un-normalized RNA counts matrix before passing to HieraType. Here we’ll use pearson residuals from a gamma regression model as a normalization method, although other reasonable methods might be considered and left to the user’s disgression. We then add the “_protein” suffix to the protein names, so they’ll align with the HieraType markerlists.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## normalize each protein taking pearson residuals from gamma generalized linear model.</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>pgamm <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">pearson_normalize_gamma_glm</span>(<span class="fu">as.matrix</span>(sem_protein[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>                                                   ,<span class="at">upper_q_thresh =</span> <span class="fl">0.99</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>                                                   ,<span class="at">lower_q_thresh =</span> <span class="fl">0.01</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>                                                   )</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pgamm) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rownames</span>(pgamm), <span class="st">"_protein"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here we conduct a principal component decomposition from combined RNA and proteins across all cells; this can be used to give us the cells x cells similarity graph we’ll pass to HieraType.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## select genes from RNA data to use for PCA analysis.</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Take the top 2k features as well as any pre-defined celltyping markers from HieraType</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindVariableFeatures</span>(sem_rna, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>markerg <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">c</span>(</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  HieraType<span class="sc">::</span>markerslist_l1</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_immune</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd8tminor</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd4tminor</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_tcellmajor</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>), <span class="st">"[["</span>, <span class="st">"predictors"</span>))</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>use_genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(sem_rna<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>var.features, <span class="fu">intersect</span>(markerg, <span class="fu">rownames</span>(sem_rna))))</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="do">### total counts and gene frequency from RNA data computed across full dataset</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>tc <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>genefreq <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">gene_frequency</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Multiomic RNA and Protein cluster analysis</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>pcaobj_multi <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">sparse_quasipoisson_pca_seurat_multiomic</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,]</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">totalcounts =</span> tc</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">grate =</span> genefreq[use_genes]</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">xother =</span> pgamm <span class="do">## normalized protein</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">do.scale =</span> <span class="cn">TRUE</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">do.center =</span> <span class="cn">TRUE</span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">scale.max =</span> <span class="dv">10</span></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">ncores =</span> <span class="dv">1</span></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a><span class="do">### Make a cell-cell adjacency matrix for HieraType </span></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>multi_simgrph <span class="ot">&lt;-</span> uwot<span class="sc">::</span><span class="fu">similarity_graph</span>(pcaobj_multi<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">n_neighbors =</span> <span class="dv">30</span></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">nn_method =</span> <span class="st">"annoy"</span></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">metric =</span> <span class="st">'cosine'</span></span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>                                  )</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(multi_simgrph) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(multi_simgrph) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(pcaobj_multi<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="hieratype-call" class="level5">
<h5 class="anchored" data-anchor-id="hieratype-call">HieraType call</h5>
<p>Here we construct our ‘multiomic’ expression matrix which includes our sparse RNA counts with the normalized protein expression, then we call <code>HieraType::run_pipeline</code> .</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### combining sparse RNA counts and normalized proteins into a single matrix 'multimat'</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>overlapping_cells <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(sem_protein), <span class="fu">colnames</span>(sem_rna))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>multimat <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">t</span>(pgamm[,overlapping_cells])</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>multimat <span class="ot">&lt;-</span> <span class="fu">cbind</span>(multimat, Matrix<span class="sc">::</span><span class="fu">t</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,overlapping_cells]))</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="do">### Define pipeline, using multiomic markerslists</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>pipeline_io_multiomics <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">make_pipeline</span>(</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">markerslists =</span> <span class="fu">list</span>(</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l1"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_l1</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"l2"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_immune</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_tcellmajor</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd4tminor</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd8tminor</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors =</span> <span class="fu">list</span>(</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"l1"</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"l2"</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors_category =</span> <span class="fu">list</span>(</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"immune"</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"tcell"</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"cd4t"</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"cd8t"</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="do">### Run the celltyping pipeline</span></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>multictobj <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>   <span class="at">pipeline =</span> pipeline_io_multiomics</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>   ,<span class="at">totalcounts=</span> tc[<span class="fu">rownames</span>(multimat)]</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>   ,<span class="at">gene_wise_frequency =</span> genefreq</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>   ,<span class="at">counts_matrix =</span> multimat </span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>   ,<span class="at">adjacency_matrix =</span> multi_simgrph[<span class="fu">rownames</span>(multimat), <span class="fu">rownames</span>(multimat)]</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="outputs" class="level5">
<h5 class="anchored" data-anchor-id="outputs">Outputs</h5>
<p>For extensive description of the outputs we get from HieraType, we refer the reader to the original post. Here, we’ll demonstrate how to make a few marker heatmaps from the RNA and protein data to sanity check our cell typing results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>multictobj<span class="sc">$</span>post_probs[[<span class="dv">1</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>] <span class="do">## posterior probabilities for each cell type</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>        cell_ID celltype_thresh celltype_granular best_score_granular
1     c_1_100_1          immune          monocyte           0.2832607
2    c_1_100_10          plasma            plasma           0.9505284
3   c_1_100_100      epithelial        epithelial           0.9894404
4  c_1_100_1000              l1         cd4_naive           0.1029328
5  c_1_100_1001              l1       endothelial           0.3917291
6  c_1_100_1002          immune        macrophage           0.2146396
7  c_1_100_1003          plasma            plasma           0.9697994
8  c_1_100_1004          immune         dendritic           0.3165272
9  c_1_100_1005              l1        fibroblast           0.2769950
10 c_1_100_1006          plasma            plasma           0.9283903
   best_score_thresh
1          0.9979747
2          0.9505284
3          0.9894404
4          0.4220678
5          0.3917291
6          0.5871081
7          0.9697994
8          0.9999999
9          0.2769950
10         0.9283903</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Use the major cell type categories for a heatmap</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>ppmulti <span class="ot">&lt;-</span> <span class="fu">copy</span>(multictobj<span class="sc">$</span>post_probs<span class="sc">$</span>l1)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>ppmulti[,celltype_multi_major<span class="sc">:</span><span class="er">=</span>celltype_granular]</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd4"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd4_tcell"</span>]</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd8"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd8_tcell"</span>]</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"mono|macro"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"macrophage_monocyte"</span>]</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="do">### fold change metrics to show in RNA heatmap</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>fctbl_rna <span class="ot">&lt;-</span> </span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="fu">clusterwise_foldchange_metrics</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,]</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">totalcounts =</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">metadata =</span> ppmulti</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">cluster_column =</span> <span class="st">"celltype_multi_major"</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">cellid_column =</span> <span class="st">"cell_ID"</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>                               )</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>hm_rna <span class="ot">&lt;-</span> <span class="fu">add_ncells</span>(<span class="fu">marker_heatmap</span>(fctbl_rna))</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hm_rna)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="wider-image">
<p><img src="./figures/hm_rna_multi.png" alt="Wide Image"></p>
</div>
<p>For the protein data, almost all proteins have some positive measurement in each cell. We can use the pearson residuals to set a cutoff for computing the proportion of positive cells by marker and by cell type. Below, we’ll use cutoff of ‘pearson residual &gt; 1’ to determine the size of the bubble in the heatmap.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>fctbl_protein <span class="ot">&lt;-</span> </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">clusterwise_foldchange_metrics_protein</span>(sem_protein[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[,overlapping_cells]</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">metadata =</span> ppmulti</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">cluster_column =</span> <span class="st">"celltype_multi_major"</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">propd =</span> (pgamm[,overlapping_cells] <span class="sc">&gt;</span> <span class="dv">1</span>))</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>hm_protein <span class="ot">&lt;-</span> <span class="fu">add_ncells</span>(<span class="fu">marker_heatmap</span>(fctbl_protein))</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hm_protein)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="wider-image">
<p><img src="./figures/hm_protein_multi.png" alt="Wide Image"></p>
</div>
</section>
</section>
<section id="integrating-unsupervised-clusters-with-immune-cell-type-labels" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Integrating unsupervised clusters with immune cell type labels</h1>
HieraType was designed to find celltypes with known, pre-specified markers. Often, unsupervised clustering is sufficiently granular to identify very distinct tissue-specific clusters, but may not identify commune immune cell types of interest due to relative sparsity of the immune markers. We demonstrate an easy way (<a href="#integrateunsup">code</a>) to reconcile the labels between HieraType and unsupervised clustering; keeping the HieraType label when a cell is classified as immune, and taking the unsupervised cluster label otherwise. This can be a useful technique regardless of whether one is using RNA, protein, or both to cell type their single cell dataset.
<div class="wider-image">
<p><img src="./figures/xyp_multi_semisup.png" alt="Wide Image"></p>
</div>
</section>
<section id="conclusion" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Conclusion</h1>
<p>The HieraType method constructs continuous celltype metagene scores, which are then probabilistically clustered to classify cell types. Because these metagene scores are continuous and the clustering uses flexible gaussian mixture models, the inputted expression can be fluidly specified in terms of RNA markers, protein markers, or some combination of both. This allows for an elegant and efficient solution for fully integrated celltyping when both protein and RNA are available.<br>
Overall, we found improved cell typing performance in the breast cancer sample analyzed in this post when both analytes were used compared to either analyte on its own.</p>
<p>While the HieraType method is fully supervised and cannot identify novel cell types not specified in a ‘<code>markerslist</code>’, we demonstrate an easy approach for integrating unsupervised clusters with HieraType immune cell typing calls, enabling high resolution cell typing for both immune cell types as well as epithelial or tissue context specific cell types which may not have a known marker profile a priori. This can help reap benefits of both unsupervised and supervised approaches - well controlled pre-specified immune cell typing combined with granular unsupervised epithelial clusters.</p>
</section>
<section id="endtoend" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> End-to-end detailed analysis code</h1>
<p>For completeness, here are a few end-to-end code blocks for each analysis step.</p>
<section id="reading-in-the-data" class="level5">
<h5 class="anchored" data-anchor-id="reading-in-the-data">Reading in the data</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scPearsonPCA)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(HieraType)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pROC)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mclust)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="do">## Note, using a v3-style Seurat Object below, i.e., `sem[["RNA"]]@counts`</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="do">## For v5-style, the syntax would be like `sem[["RNA"]]$counts`</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">Seurat.object.assay.version =</span> <span class="st">"v3"</span>)  </span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>sem_protein <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Breast/sem_protein.rds"</span>)</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Breast/sem_rna.rds"</span>)</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> <span class="fu">subset</span>(sem_rna, nCount_RNA <span class="sc">&gt;=</span><span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="normalizing-protein-data" class="level5">
<h5 class="anchored" data-anchor-id="normalizing-protein-data">Normalizing protein data</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="do">## normalize each protein taking pearson residuals from gamma generalized linear model.</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>pgamm <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">pearson_normalize_gamma_glm</span>(<span class="fu">as.matrix</span>(sem_protein[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                                                   ,<span class="at">upper_q_thresh =</span> <span class="fl">0.99</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>                                                   ,<span class="at">lower_q_thresh =</span> <span class="fl">0.01</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>                                                   )</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pgamm) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rownames</span>(pgamm), <span class="st">"_protein"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="multiunsup" class="level5">
<h5 class="anchored" data-anchor-id="multiunsup">Multiomic pearson residual PCA analysis using the protein and RNA</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="do">## select genes from RNA data to use for PCA analysis.</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Will take the top 2k features as well as any pre-defined immune celltyping markers from HieraType</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindVariableFeatures</span>(sem_rna, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>markerg <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">c</span>(</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  HieraType<span class="sc">::</span>markerslist_l1</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_immune</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd8tminor</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd4tminor</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_tcellmajor</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>), <span class="st">"[["</span>, <span class="st">"predictors"</span>))</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>use_genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(sem_rna<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>var.features, <span class="fu">intersect</span>(markerg, <span class="fu">rownames</span>(sem_rna))))</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="do">### total counts and gene frequency from RNA data computed across full dataset</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>tc <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>genefreq <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">gene_frequency</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a><span class="do">### add a '_protein' suffix to the protein names - helps differentiate proteins and RNA's with the same name. </span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pgamm) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rownames</span>(pgamm), <span class="st">"_protein"</span>)</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a><span class="do">### Principal components derived from the combined covariance matrix of pearson residual normalized RNA and protein.</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a><span class="do">### </span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>pcaobj_multi <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">sparse_quasipoisson_pca_seurat_multiomic</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,]</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">totalcounts =</span> tc</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">grate =</span> genefreq[use_genes]</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">xother =</span> pgamm <span class="do">## normalized protein</span></span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">do.scale =</span> <span class="cn">TRUE</span></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">do.center =</span> <span class="cn">TRUE</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">scale.max =</span> <span class="dv">10</span></span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">ncores =</span> <span class="dv">1</span></span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>PC_ 1 
Positive:  CD45_protein, CD4_protein, CD39_protein, CD16_protein, Channel-CD45_protein, Fibronectin_protein, CD3_protein, Vimentin_protein, CD40_protein, CD14_protein 
       HLA-DR_protein, STING_protein, SMA_protein, ICAM1_protein, CD68_protein, CD163_protein, CD11c_protein, ICOS_protein, CD20_protein, CD8_protein 
Negative:  AZGP1, KRT7, KRT18, NAMPT, KRT8, KRT19, FASN, S100A11, DBI, SERHL2 
       MYL6, EPCAM, ASPH, PRDX2, ALDOA, WFDC2, CYB5A, GSTP1, TMBIM6, NFIX 
PC_ 2 
Positive:  VIM, SPARC, COL3A1, COL1A1, COL1A2, CD39_protein, IGFBP7, HLA-DRB1, CST3, TIMP2 
       HLA-DRA, AEBP1, CD74, LUM, COL6A2, CD40_protein, CD16_protein, BGN, CYBA, COL6A1 
Negative:  pan-RAS_protein, p53_protein, Channel-PanCK_protein, CD127_protein, EGFR_protein, NF-kB p65_protein, EpCAM_protein, CTLA4_protein, TCF7_protein, LAMP1_protein 
       Beta-catenin_protein, Ki-67_protein, CD138_protein, VISTA_protein, FOXP3_protein, Her2_protein, LAG3_protein, CD45RA_protein, DBI, SERHL2 
PC_ 3 
Positive:  H3C15, H3C2, H3-7, H4C14, H4C11, H3C10, H3C13, H4C12, H3C7, H2AC16 
       H2BC4, H1-5, H2BC9, H1-2, H2AC18, H2AC11, H2BC18, H3C8, H3C11, H1-3 
Negative:  KRT19, p53_protein, FOXP3_protein, LAG3_protein, NDRG1, TCF7_protein, Her2_protein, CTLA4_protein, IgD_protein, KRT8 
       Channel-PanCK_protein, CD56_protein, VEGFA, LDHA, CD123_protein, AFMID, GZMB_protein, PGK1, ALDOA, iNOS_protein 
PC_ 4 
Positive:  HLA-DRB1, HLA-DRA, CD74, HLA-DR_protein, APOE, Channel-CD68_protein, CD11c_protein, CD45_protein, LYZ, HLA-DQB1 
       CD68_protein, CTSD, HLA-DPA1, C1QC, IFI30, CD16_protein, APOC1, HLA-DQA1, C1QB, CD4_protein 
Negative:  CD123_protein, CD34_protein, GZMB_protein, IGFBP7, CD31_protein, COL4A1, SPARC, CD56_protein, LAG3_protein, FOXP3_protein 
       COL4A2, IgD_protein, PLVAP, Her2_protein, GITR_protein, HSPG2, 4-1BB_protein, LAMA4, p53_protein, SMA_protein 
PC_ 5 
Positive:  PLVAP, COL4A1, CD34_protein, KDR, Beta-catenin_protein, CD31_protein, EGFL7, EpCAM_protein, COL4A2, AQP1 
       FLT1, HSPG2, SHANK3, PODXL, A2M, LAMP1_protein, CD34, ECSCR, INSR, STC1 
Negative:  PD-L2_protein, FOXP3_protein, LAG3_protein, 4-1BB_protein, GITR_protein, CD56_protein, GZMB_protein, iNOS_protein, Her2_protein, IgD_protein 
       GZMA_protein, PD-1_protein, CD11b_protein, CD19_protein, TCF7_protein, CD163_protein, CD123_protein, CTLA4_protein, IL-18_protein, Bcl-2_protein </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Make a cell-cell adjacency matrix for HieraType </span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>multi_simgrph <span class="ot">&lt;-</span> uwot<span class="sc">::</span><span class="fu">similarity_graph</span>(pcaobj_multi<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">n_neighbors =</span> <span class="dv">30</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">nn_method =</span> <span class="st">"annoy"</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">metric =</span> <span class="st">'cosine'</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>                                  )</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(multi_simgrph) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(multi_simgrph) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(pcaobj_multi<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="do">###  Unsupervised clusters using multiomic cell-cell adjacency graph</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>sem_rna[[<span class="st">"multigrph"</span>]] <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">as.Graph</span>(multi_simgrph[<span class="fu">colnames</span>(sem_rna),<span class="fu">colnames</span>(sem_rna)])</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindClusters</span>(sem_rna, <span class="at">graph.name=</span><span class="st">"multigrph"</span>)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>sem_rna<span class="sc">@</span>meta.data<span class="sc">$</span>clusters_multi_unsup <span class="ot">&lt;-</span> sem_rna<span class="sc">@</span>meta.data<span class="sc">$</span>seurat_clusters</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>xyp_multi_unsup <span class="ot">&lt;-</span> <span class="fu">xyplot</span>(<span class="st">"clusters_multi_unsup"</span>, <span class="at">metadata =</span> sem_rna<span class="sc">@</span>meta.data, <span class="at">ptsize =</span> <span class="fl">0.05</span>) <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"figures/xyp_multi_unsup.png"</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">plot=</span>xyp_multi_unsup</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">device =</span> <span class="st">"png"</span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">width =</span> <span class="dv">1181</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>)</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">height=</span><span class="dv">689</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>)</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">units =</span> <span class="st">'px'</span></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="multiomic-cell-typing-code-using-hieratype" class="level5">
<h5 class="anchored" data-anchor-id="multiomic-cell-typing-code-using-hieratype">Multiomic cell typing code using HieraType</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Define the pipeline using multiomic markerlists in HieraType package</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>overlapping_cells <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(sem_protein), <span class="fu">colnames</span>(sem_rna))</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>multimat <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">t</span>(pgamm[,overlapping_cells])</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>multimat <span class="ot">&lt;-</span> <span class="fu">cbind</span>(multimat, Matrix<span class="sc">::</span><span class="fu">t</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,overlapping_cells]))</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>pipeline_io_multiomics <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">make_pipeline</span>(</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">markerslists =</span> <span class="fu">list</span>(</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l1"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_l1</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"l2"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_immune</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_tcellmajor</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd4tminor</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd8tminor</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors =</span> <span class="fu">list</span>(</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"l1"</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"l2"</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors_category =</span> <span class="fu">list</span>(</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"immune"</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"tcell"</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"cd4t"</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"cd8t"</span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a>multictobj <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(</span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">pipeline =</span> pipeline_io_multiomics</span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">totalcounts=</span> tc[<span class="fu">rownames</span>(multimat)]</span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">gene_wise_frequency =</span> genefreq</span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">counts_matrix =</span> multimat</span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">adjacency_matrix =</span> multi_simgrph[<span class="fu">rownames</span>(multimat), <span class="fu">rownames</span>(multimat)]</span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="rnaonly" class="level5">
<h5 class="anchored" data-anchor-id="rnaonly">RNA-only cell typing</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="do">### RNA-only PCA and similarity graph</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>pcaobj_rna <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">sparse_quasipoisson_pca_seurat</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,]</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">totalcounts =</span> tc</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">grate =</span> genefreq[use_genes]</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">do.scale =</span> <span class="cn">TRUE</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">do.center =</span> <span class="cn">TRUE</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">scale.max =</span> <span class="dv">10</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">ncores =</span> <span class="dv">1</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="do">### Make a cell-cell adjacency matrix for HieraType </span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>rna_simgrph <span class="ot">&lt;-</span> uwot<span class="sc">::</span><span class="fu">similarity_graph</span>(pcaobj_rna<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">n_neighbors =</span> <span class="dv">30</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">nn_method =</span> <span class="st">"annoy"</span></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">metric =</span> <span class="st">'cosine'</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>                                  )</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(rna_simgrph) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(rna_simgrph) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(pcaobj_rna<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings)</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a><span class="do">## RNA-only hieratype</span></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>rnactobj <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">pipeline =</span> pipeline_io_multiomics <span class="do">## the _protein markers are automatically excluded, </span></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>                                    <span class="do">## because they won't be found in the counts_matrix below</span></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">totalcounts=</span> tc[<span class="fu">rownames</span>(multimat)]</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">gene_wise_frequency =</span> genefreq[use_genes]</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">counts_matrix =</span> multimat[,use_genes] <span class="do">## exclude proteins, only use the 'use_genes'.  </span></span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">adjacency_matrix =</span> rna_simgrph[<span class="fu">rownames</span>(multimat), <span class="fu">rownames</span>(multimat)]</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="proteinonly" class="level5">
<h5 class="anchored" data-anchor-id="proteinonly">Protein-only cell typing</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>pgamm_prot <span class="ot">&lt;-</span> pgamm</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pgamm_prot) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_protein$"</span>, <span class="st">""</span>, <span class="fu">rownames</span>(pgamm_prot))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>sem_protein <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">SetAssayData</span>(sem_protein,<span class="at">layer =</span> <span class="st">"data"</span>, <span class="at">new.data =</span> pgamm_prot)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>sem_protein <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">ScaleData</span>(sem_protein)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>sem_protein <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">RunPCA</span>(sem_protein, <span class="at">features =</span> <span class="fu">rownames</span>(sem_protein))</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="do">### Make a cell-cell adjacency matrix for HieraType </span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>protein_simgrph <span class="ot">&lt;-</span> uwot<span class="sc">::</span><span class="fu">similarity_graph</span>(sem_protein<span class="sc">@</span>reductions<span class="sc">$</span>pca<span class="sc">@</span>cell.embeddings</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">n_neighbors =</span> <span class="dv">30</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">nn_method =</span> <span class="st">"annoy"</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">metric =</span> <span class="st">'cosine'</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>                                  )</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(protein_simgrph) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(protein_simgrph) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sem_protein<span class="sc">@</span>reductions<span class="sc">$</span>pca<span class="sc">@</span>cell.embeddings)</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="do">#### Make a copy of the cell typing pipeline to use for protein</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="do">#### -- Remove 'mast' cells, which dont have a specific protein marker.</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="do">#### -- remove the minor T8 and T4 subtypes</span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein <span class="ot">&lt;-</span> pipeline_io_multiomics</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>markerslists<span class="sc">$</span>l2<span class="sc">$</span>mast <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>markerslists<span class="sc">$</span>lt4minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>markerslists<span class="sc">$</span>lt8minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>priors<span class="sc">$</span>lt4minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>priors<span class="sc">$</span>lt8minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>priors_category<span class="sc">$</span>lt4minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>priors_category<span class="sc">$</span>lt8minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>proteinctobj <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(</span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">pipeline =</span> pipeline_io_protein</span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">counts_matrix =</span> multimat[,<span class="fu">grep</span>(<span class="st">"_protein$"</span>,<span class="fu">colnames</span>(multimat),<span class="at">value=</span><span class="cn">TRUE</span>)]</span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">adjacency_matrix =</span> protein_simgrph[<span class="fu">rownames</span>(multimat), <span class="fu">rownames</span>(multimat)]</span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="combining-metadata" class="level5">
<h5 class="anchored" data-anchor-id="combining-metadata">Combining metadata</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Combining metadata</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Adding on multiomic celltype annotations</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>ppmulti <span class="ot">&lt;-</span> <span class="fu">copy</span>(multictobj<span class="sc">$</span>post_probs<span class="sc">$</span>l1)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>ppmulti[,celltype_multi_major<span class="sc">:</span><span class="er">=</span>celltype_granular]</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd4"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd4_tcell"</span>]</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd8"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd8_tcell"</span>]</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"mono|macro"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"macrophage_monocyte"</span>]</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>ppmulti[,.N,by<span class="ot">=</span>.(celltype_multi_major, celltype_granular)]</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Adding on RNA-only based  celltype annotations</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>ppmulti <span class="ot">&lt;-</span> <span class="fu">merge</span>(ppmulti, rnactobj<span class="sc">$</span>post_probs<span class="sc">$</span>l1[,.(cell_ID, celltype_granular)]</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>                 ,<span class="at">by=</span><span class="st">"cell_ID"</span>, <span class="at">suffixes=</span><span class="fu">c</span>(<span class="st">"_multi"</span>, <span class="st">"_rna"</span>))</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>ppmulti[,celltype_rna_major<span class="sc">:</span><span class="er">=</span>celltype_granular_rna]</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd4"</span>,celltype_rna_major),celltype_rna_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd4_tcell"</span>]</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd8"</span>,celltype_rna_major),celltype_rna_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd8_tcell"</span>]</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"mono|macro"</span>,celltype_rna_major),celltype_rna_major<span class="sc">:</span><span class="er">=</span><span class="st">"macrophage_monocyte"</span>]</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>ppmulti[,.N,by<span class="ot">=</span>.(celltype_rna_major, celltype_granular_rna)]</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a><span class="do">## Adding protein-only based celltype annotations</span></span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>ppmulti <span class="ot">&lt;-</span> <span class="fu">merge</span>(ppmulti, proteinctobj<span class="sc">$</span>post_probs<span class="sc">$</span>l1[,.(cell_ID, celltype_granular)]</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>                 ,<span class="at">by=</span><span class="st">"cell_ID"</span>, <span class="at">suffixes=</span><span class="fu">c</span>(<span class="st">"_multi"</span>, <span class="st">"_protein"</span>))</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a><span class="fu">setnames</span>(ppmulti, <span class="fu">c</span>(<span class="st">"celltype_granular"</span>), <span class="fu">c</span>(<span class="st">"celltype_granular_protein"</span>))</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>ppmulti[,celltype_protein_major<span class="sc">:</span><span class="er">=</span>celltype_granular_protein]</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd4"</span>,celltype_protein_major),celltype_protein_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd4_tcell"</span>]</span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd8"</span>,celltype_protein_major),celltype_protein_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd8_tcell"</span>]</span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"mono|macro"</span>,celltype_protein_major),celltype_protein_major<span class="sc">:</span><span class="er">=</span><span class="st">"macrophage_monocyte"</span>]</span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a>ppmulti[,.N,by<span class="ot">=</span>.(celltype_protein_major, celltype_granular_protein)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="integrateunsup" class="level5">
<h5 class="anchored" data-anchor-id="integrateunsup">Integrating unsupervised clusters with immune cell type labels</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Annotating non-immune clusters with unsupervised labels, creating more granular annotations</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="do">### </span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>ppmulti <span class="ot">&lt;-</span> <span class="fu">merge</span>(ppmulti, <span class="fu">data.table</span>(sem_rna<span class="sc">@</span>meta.data)[,.(cell_ID, x_slide_mm, y_slide_mm, clusters_multi_unsup)], <span class="at">by =</span> <span class="st">"cell_ID"</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>ppmulti[,celltype_multi_semi<span class="sc">:</span><span class="er">=</span>celltype_multi_major]</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>ppmulti[celltype_multi_semi <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"plasma"</span>, <span class="st">"endothelial"</span>, <span class="st">"epithelial"</span>, <span class="st">"fibroblast"</span>, <span class="st">"smooth_muscle"</span>),celltype_multi_semi<span class="sc">:</span><span class="er">=</span>clusters_multi_unsup]</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>ppmulti[,.N,by<span class="ot">=</span>.(clusters_multi_unsup, celltype_multi_major)]</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>class_other <span class="ot">&lt;-</span> ppmulti[,.N,by<span class="ot">=</span>.(clusters_multi_unsup, celltype_multi_major)][,prop<span class="sc">:</span><span class="er">=</span>N<span class="sc">/</span><span class="fu">sum</span>(N),by<span class="ot">=</span>.(clusters_multi_unsup)][<span class="fu">order</span>(<span class="sc">-</span>prop)][</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>  ,<span class="fu">head</span>(.SD,<span class="dv">1</span>),by<span class="ot">=</span>.(clusters_multi_unsup)]</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>others <span class="ot">&lt;-</span> class_other[<span class="sc">!</span>celltype_multi_major <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"plasma"</span>, <span class="st">"endothelial"</span>, <span class="st">"epithelial"</span>, <span class="st">"fibroblast"</span>, <span class="st">"smooth_muscle"</span>)][[<span class="st">"clusters_multi_unsup"</span>]]</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>ppmulti[celltype_multi_semi <span class="sc">%in%</span> others, celltype_multi_semi<span class="sc">:</span><span class="er">=</span><span class="st">"other"</span>]</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>xyp_multi_semisup <span class="ot">&lt;-</span> <span class="fu">xyplot</span>(<span class="st">"celltype_multi_semi"</span>, <span class="at">metadata =</span> ppmulti, <span class="at">ptsize =</span> <span class="fl">0.05</span>) <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"figures/xyp_multi_semisup.png"</span></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">plot=</span>xyp_multi_semisup</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">device =</span> <span class="st">"png"</span></span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">width =</span> <span class="dv">1181</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>)</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">height=</span><span class="dv">689</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>)</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">units =</span> <span class="st">'px'</span></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ari" class="level5">
<h5 class="anchored" data-anchor-id="ari">ARI quantifying agreement between cell type labels</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="do">## ARI</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(ppmulti<span class="sc">$</span>celltype_multi_major, ppmulti<span class="sc">$</span>celltype_protein_major)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(ppmulti<span class="sc">$</span>celltype_multi_major, ppmulti<span class="sc">$</span>celltype_rna_major)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(ppmulti<span class="sc">$</span>celltype_protein_major, ppmulti<span class="sc">$</span>celltype_rna_major)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="making-roc-curves-computing-and-plotting-area-under-the-curve" class="level5">
<h5 class="anchored" data-anchor-id="making-roc-curves-computing-and-plotting-area-under-the-curve">Making ROC curves, computing and plotting area under the curve</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Get the most important 'index markers' for each celltype (including proteins and RNAs)</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>index_markers <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(HieraType<span class="sc">::</span>markerslist_multiomic_immune, <span class="st">"[["</span>, <span class="st">"index_marker"</span>))</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>index_markers <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(<span class="fu">lapply</span>(<span class="fu">names</span>(HieraType<span class="sc">::</span>markerslist_multiomic_immune), <span class="cf">function</span>(ct){</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  markerdt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">celltype =</span> ct</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>                         ,<span class="at">marker =</span> HieraType<span class="sc">::</span>markerslist_multiomic_immune[[ct]][[<span class="st">"index_marker"</span>]]</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>                         ,<span class="at">typ =</span> <span class="st">"rna"</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  markerdt[<span class="fu">grepl</span>(<span class="st">"_protein$"</span>, marker),typ<span class="sc">:</span><span class="er">=</span><span class="st">"protein"</span>]</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>  markerdt[,marker<span class="sc">:</span><span class="er">=</span><span class="fu">gsub</span>(<span class="st">"_protein$"</span>,<span class="st">""</span>, marker)]</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>  markerdt <span class="ot">&lt;-</span> markerdt[(typ<span class="sc">==</span><span class="st">"rna"</span> <span class="sc">&amp;</span> marker <span class="sc">%in%</span> <span class="fu">rownames</span>(sem_rna)) <span class="sc">|</span> (typ<span class="sc">==</span><span class="st">"protein"</span> <span class="sc">&amp;</span> marker <span class="sc">%in%</span> <span class="fu">rownames</span>(sem_protein))]</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(markerdt)</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="do">### Combine macrophage and monocyte due to similarity</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>index_markers[celltype <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"macrophage"</span>, <span class="st">"monocyte"</span>),celltype<span class="sc">:</span><span class="er">=</span><span class="st">"macrophage_monocyte"</span>]</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>aucrocl <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>plotdata_markerl <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pgamm) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_protein"</span>, <span class="st">""</span>, <span class="fu">rownames</span>(pgamm))</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(ii <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(index_markers)){</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(ii)</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>  res <span class="ot">&lt;-</span> index_markers[ii]</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>  typ <span class="ot">&lt;-</span> res[[<span class="st">"typ"</span>]]</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>  marker <span class="ot">&lt;-</span> res[[<span class="st">"marker"</span>]]</span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>  celltype <span class="ot">&lt;-</span> res[[<span class="st">"celltype"</span>]] </span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(typ<span class="sc">==</span><span class="st">"rna"</span>){</span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[marker,ppmulti[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a>    tc <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[,ppmulti[[<span class="st">"cell_ID"</span>]]])</span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a>    muhat <span class="ot">&lt;-</span> genefreq[marker] <span class="sc">*</span> tc </span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> (x <span class="sc">-</span> muhat) <span class="sc">/</span> <span class="fu">sqrt</span>(muhat) <span class="do">## normalized expression for marker (pearson residuals)</span></span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a>  }  <span class="cf">else</span> {</span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> pgamm[marker,ppmulti[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb21-34"><a href="#cb21-34" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-35"><a href="#cb21-35" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-36"><a href="#cb21-36" aria-hidden="true" tabindex="-1"></a>  pdlbl <span class="ot">&lt;-</span> <span class="fu">list</span>() </span>
<span id="cb21-37"><a href="#cb21-37" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(labelname <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"celltype_multi_major"</span>, <span class="st">"celltype_rna_major"</span>, <span class="st">"celltype_protein_major"</span>)){</span>
<span id="cb21-38"><a href="#cb21-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(labelname)</span>
<span id="cb21-39"><a href="#cb21-39" aria-hidden="true" tabindex="-1"></a>    labels <span class="ot">&lt;-</span> ppmulti[[labelname]]</span>
<span id="cb21-40"><a href="#cb21-40" aria-hidden="true" tabindex="-1"></a>    labels[<span class="fu">grep</span>(<span class="st">"tcell"</span>, labels)] <span class="ot">&lt;-</span> <span class="st">"tcell"</span> </span>
<span id="cb21-41"><a href="#cb21-41" aria-hidden="true" tabindex="-1"></a>    y <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(labels<span class="sc">==</span>celltype)</span>
<span id="cb21-42"><a href="#cb21-42" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">sum</span>(y) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb21-43"><a href="#cb21-43" aria-hidden="true" tabindex="-1"></a>      roc_obj <span class="ot">&lt;-</span> <span class="fu">roc</span>(y, x, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-44"><a href="#cb21-44" aria-hidden="true" tabindex="-1"></a>      res[,(labelname)<span class="sc">:</span><span class="er">=</span><span class="fu">as.numeric</span>(<span class="fu">auc</span>(roc_obj))]</span>
<span id="cb21-45"><a href="#cb21-45" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb21-46"><a href="#cb21-46" aria-hidden="true" tabindex="-1"></a>    pdlbl[[labelname]] <span class="ot">&lt;-</span> </span>
<span id="cb21-47"><a href="#cb21-47" aria-hidden="true" tabindex="-1"></a>      <span class="fu">data.table</span>(<span class="at">spec=</span>roc_obj<span class="sc">$</span>specificities, <span class="at">sens=</span>roc_obj<span class="sc">$</span>sensitivities,<span class="at">marker=</span>marker)[,label<span class="sc">:</span><span class="er">=</span>labelname][<span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">1</span>,.N,<span class="dv">10</span>), .N))]</span>
<span id="cb21-48"><a href="#cb21-48" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-49"><a href="#cb21-49" aria-hidden="true" tabindex="-1"></a>  plotdata_markerl[[ii]] <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(pdlbl)[,<span class="st">`</span><span class="at">:=</span><span class="st">`</span>(<span class="at">celltype =</span> celltype, <span class="at">marker=</span>marker, <span class="at">typ =</span> typ)]</span>
<span id="cb21-50"><a href="#cb21-50" aria-hidden="true" tabindex="-1"></a>  <span class="co">#ggplot(rbindlist(pdlbl)[seq(1,.N,by=10)],aes(x = (1-spec), y = sens, color=label)) + geom_line()</span></span>
<span id="cb21-51"><a href="#cb21-51" aria-hidden="true" tabindex="-1"></a>  aucrocl[[ii]] <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">copy</span>(res) <span class="co">#return(res)</span></span>
<span id="cb21-52"><a href="#cb21-52" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="average-auc-across-markers-and-celltypes" class="level4" data-number="7.0.0.1">
<h4 data-number="7.0.0.1" class="anchored" data-anchor-id="average-auc-across-markers-and-celltypes"><span class="header-section-number">7.0.0.1</span> Average AUC across markers and celltypes</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>aucroc[celltype<span class="sc">!=</span><span class="st">"mast"</span>,<span class="fu">lapply</span>(.SD, mean),.SDcols<span class="ot">=</span><span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="t-cell-roc-curves" class="level4" data-number="7.0.0.2">
<h4 data-number="7.0.0.2" class="anchored" data-anchor-id="t-cell-roc-curves"><span class="header-section-number">7.0.0.2</span> T cell ROC curves</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>  shw <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(plotdata_markerl)[marker <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"CD3"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD2"</span>)]</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  shw[,method<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(label, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"celltype_multi_major"</span>, <span class="st">"celltype_rna_major"</span>, <span class="st">"celltype_protein_major"</span>), <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"multiomic typing"</span>, <span class="st">"rna-only cell typing"</span>, <span class="st">"protein-only cell typing"</span>))]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">merge</span>(shw, aucrocl)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  data.table<span class="sc">::</span><span class="fu">setnames</span>(shw, <span class="fu">c</span>(<span class="st">"typ"</span>), <span class="fu">c</span>(<span class="st">"analyte"</span>))</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  shw[,marker<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(marker, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"CD3"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD2"</span>))]</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  shw <span class="ot">&lt;-</span> shw[<span class="fu">order</span>(marker)]</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  aucroc[marker <span class="sc">%in%</span> shw<span class="sc">$</span>marker]</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  rocplot_tcell <span class="ot">&lt;-</span> </span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(shw, <span class="fu">aes</span>(<span class="at">x =</span> (<span class="dv">1</span><span class="sc">-</span>spec), <span class="at">y =</span> sens,<span class="at">group =</span> method,<span class="at">color=</span>method)) <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span> </span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>marker <span class="sc">+</span> analyte, <span class="at">labeller=</span>label_both) <span class="sc">+</span> </span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,.<span class="dv">2</span>), <span class="at">labels =</span> <span class="dv">1</span><span class="sc">-</span><span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,.<span class="dv">2</span>), <span class="at">name =</span> <span class="st">"specificity"</span>) <span class="sc">+</span> </span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,.<span class="dv">2</span>), <span class="at">name =</span> <span class="st">"sensitivity"</span>) <span class="sc">+</span> </span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Set1"</span>)[<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>,<span class="dv">3</span>)]</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>    <span class="co">#scale_color_manual(values = brewer.pal(8,"Dark2")[c(1:3)]</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>                       , <span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">lwd=</span><span class="dv">2</span>)))  <span class="sc">+</span> </span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">16</span>)) <span class="sc">+</span> </span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"bottom"</span>) <span class="sc">+</span>  </span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"ROC curve with T cell RNA and protein markers:</span><span class="sc">\n</span><span class="st">Multimomic integrated celltyping outperforms</span><span class="sc">\n</span><span class="st">either RNA or protein alone when considering cell type markers measured on both analytes"</span>)</span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>  lbld <span class="ot">&lt;-</span> </span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>    aucroc[marker <span class="sc">%in%</span> shw<span class="sc">$</span>marker]</span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>  lbld <span class="ot">&lt;-</span> <span class="fu">melt</span>(lbld, <span class="at">id.vars=</span><span class="fu">c</span>(<span class="st">"celltype"</span>, <span class="st">"marker"</span>, <span class="st">"typ"</span>),<span class="at">value.name=</span><span class="st">"auc"</span>, <span class="at">variable.name=</span><span class="st">"label"</span>)</span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a>  lbld[,method<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(label</span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a>                       ,<span class="at">levels=</span><span class="fu">c</span>(<span class="st">"celltype_multi_major"</span>, <span class="st">"celltype_rna_major"</span>, <span class="st">"celltype_protein_major"</span>)</span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>                       ,<span class="at">labels=</span><span class="fu">c</span>(<span class="st">"multiomic typing"</span>, <span class="st">"rna-only cell typing"</span>, <span class="st">"protein-only cell typing"</span>))]</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a>  lbld[,marker<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(marker, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"CD3"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD2"</span>))]</span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a>  lbld[,analyte<span class="sc">:</span><span class="er">=</span>typ]</span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a>  lbld[,lbl<span class="sc">:</span><span class="er">=</span><span class="fu">paste0</span>(method, <span class="st">" auc="</span>, <span class="fu">formatC</span>(auc, <span class="at">digits=</span><span class="dv">3</span>))]</span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a>  lbld[,spec<span class="sc">:</span><span class="er">=</span><span class="fl">0.7</span>]</span>
<span id="cb23-31"><a href="#cb23-31" aria-hidden="true" tabindex="-1"></a>  lbld[,sens<span class="sc">:</span><span class="er">=</span><span class="fl">0.1</span><span class="sc">-</span><span class="fu">as.numeric</span>(lbld<span class="sc">$</span>method)<span class="sc">*</span><span class="fl">0.05</span>]</span>
<span id="cb23-32"><a href="#cb23-32" aria-hidden="true" tabindex="-1"></a>  rocplot_tcell_lbl <span class="ot">&lt;-</span> rocplot_tcell <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="at">data=</span>lbld, <span class="fu">aes</span>(<span class="at">label =</span> lbl),<span class="at">show.legend=</span><span class="cn">FALSE</span>)</span>
<span id="cb23-33"><a href="#cb23-33" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-34"><a href="#cb23-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="at">plot=</span>rocplot_tcell_lbl</span>
<span id="cb23-35"><a href="#cb23-35" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">filename =</span> <span class="st">"figures/rocplot_tcell_lbl.png"</span></span>
<span id="cb23-36"><a href="#cb23-36" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">width=</span><span class="dv">1372</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb23-37"><a href="#cb23-37" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">height=</span><span class="dv">742</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb23-38"><a href="#cb23-38" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">units =</span> <span class="st">'px'</span></span>
<span id="cb23-39"><a href="#cb23-39" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="auc-barplots" class="level4" data-number="7.0.0.3">
<h4 data-number="7.0.0.3" class="anchored" data-anchor-id="auc-barplots"><span class="header-section-number">7.0.0.3</span> AUC barplots</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>  mroc <span class="ot">&lt;-</span> <span class="fu">melt</span>(aucroc</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">id.vars=</span><span class="fu">c</span>(<span class="st">"celltype"</span>, <span class="st">"marker"</span>, <span class="st">"typ"</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">value.name=</span><span class="st">"auc"</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">variable.name=</span><span class="st">"method"</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">variable.factor =</span> <span class="cn">FALSE</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  mroc[,celltype<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(celltype, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"tcell"</span>, <span class="st">"bcell"</span>, <span class="st">"macrophage_monocyte"</span>, <span class="st">"dendritic"</span>, <span class="st">"neutrophil"</span>, <span class="st">"nk"</span>, <span class="st">"mast"</span>))]</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  mroc[,marker_label<span class="sc">:</span><span class="er">=</span><span class="fu">paste0</span>(marker, <span class="st">" ("</span>, typ, <span class="st">")"</span>)]</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  mroc[,marker_label<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(marker_label,<span class="at">levels=</span>mroc[<span class="fu">order</span>(celltype,typ, <span class="sc">-</span>auc),<span class="fu">unique</span>(marker_label)])]</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>  aucbar <span class="ot">&lt;-</span> </span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(mroc[celltype<span class="sc">!=</span><span class="st">"mast"</span>]</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>           ,<span class="fu">aes</span>(<span class="at">y =</span> marker_label, <span class="at">x =</span> auc, <span class="at">fill=</span>method)) <span class="sc">+</span> </span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">'identity'</span>,<span class="at">position=</span><span class="fu">position_dodge</span>()) <span class="sc">+</span> </span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">brewer.pal</span>(<span class="dv">3</span>,<span class="st">"Set1"</span>)<span class="co">#unname(pals::alphabet())</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>                      , <span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">reverse=</span><span class="cn">TRUE</span>)) <span class="sc">+</span> </span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>celltype, <span class="at">labeller=</span>label_both,<span class="at">scales=</span><span class="st">"free"</span>) <span class="sc">+</span> </span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Agreement between celltype label and marker gene/protein as measured by area under the ROC curve"</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>        ,<span class="at">x =</span> <span class="st">"area under the ROC curve"</span>)</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="at">plot=</span>aucbar</span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">filename =</span> <span class="st">"figures/aucbarplot.png"</span></span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">width=</span><span class="dv">1430</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">height=</span><span class="dv">874</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">units =</span> <span class="st">'px'</span></span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a>  aucbar_with_text <span class="ot">&lt;-</span>  </span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a>  aucbar <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> <span class="fu">formatC</span>(auc, <span class="at">digits =</span> <span class="dv">3</span>)),<span class="at">position=</span><span class="fu">position_dodge</span>(<span class="at">width=</span><span class="fl">0.8</span>))  <span class="sc">+</span> </span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">coord_cartesian</span>(<span class="at">clip =</span> <span class="st">"off"</span>)</span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="at">plot=</span>aucbar_with_text </span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">filename =</span> <span class="st">"figures/aucbarplot_wtext.png"</span></span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">width=</span><span class="dv">1430</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">height=</span><span class="dv">874</span><span class="sc">*</span><span class="fl">1.3</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">units =</span> <span class="st">'px'</span></span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


<!-- -->


</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-ariref" class="csl-entry" role="listitem">
Rand, William M. 1971. <span>“Objective Criteria for the Evaluation of Clustering Methods.”</span> <em>Journal of the American Statistical Association</em> 66 (336): 846–50. <a href="https://doi.org/10.1080/01621459.1971.10482356">https://doi.org/10.1080/01621459.1971.10482356</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/nanostring-biostats\.github\.io\/CosMx-Analysis-Scratch-Space");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb25" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Integrated RNA + protein multiomic cell typing using HieraType"</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co">  - name: Dan McGuire</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co">    orcid: 0009-0006-0286-9625</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="co">    affiliations:</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="co">      - ref: nstg</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="co">      - ref: dan11mcguire</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="an">toc:</span><span class="co"> true</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="an">toc-title:</span><span class="co"> Contents</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a><span class="an">toc-depth:</span><span class="co"> 3</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="an">toc-expand:</span><span class="co"> 2</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a><span class="an">toc-location:</span><span class="co"> left</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="an">number-sections:</span><span class="co"> true</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="an">number-depth:</span><span class="co"> 4</span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "2025-10-17"</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span></span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a><span class="co">  - cell typing</span></span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a><span class="an">image:</span><span class="co"> figures/rocplot_tcell_lbl.png</span></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a><span class="an">description:</span><span class="co"> "Efficient cell typing using multiple analytes"</span></span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a><span class="an">draft:</span><span class="co"> false</span></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a><span class="co">  html: </span></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a><span class="fu"># Introduction</span></span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>This post showcases how the <span class="co">[</span><span class="ot">HieraType cell typing package</span><span class="co">](https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/HieraType)</span> can be used for integrated same-slide multiomic (protein + rna) cell typing.  </span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a>We use a breast cancer tissue sample with both 64-plex protein and WTx 18,935 plex RNA assayed, and demonstrate how cell typing can be performed using </span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>both RNA + protein </span>
<span id="cb25-34"><a href="#cb25-34" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>protein-only</span>
<span id="cb25-35"><a href="#cb25-35" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>RNA-only</span>
<span id="cb25-36"><a href="#cb25-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-37"><a href="#cb25-37" aria-hidden="true" tabindex="-1"></a>We compare cell typing performance and show that, overall, protein-only and RNA-only cell typing calls show more disagreement with each other than they do with with integrated multiomic cell type calls.</span>
<span id="cb25-38"><a href="#cb25-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-39"><a href="#cb25-39" aria-hidden="true" tabindex="-1"></a>We also demonstrate evidence of improved coherence of cell type labels with the defining marker genes and proteins when using an integrated approach that makes use of both analytes.</span>
<span id="cb25-40"><a href="#cb25-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-41"><a href="#cb25-41" aria-hidden="true" tabindex="-1"></a>Finally, we show a simple yet effective way to integrate unsupervised clusters with supervised immune cell type labels generated with HieraType. This can enable granular detection of novel or tissue-specific cell types along with well known immune cell types.</span>
<span id="cb25-42"><a href="#cb25-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-43"><a href="#cb25-43" aria-hidden="true" tabindex="-1"></a>Code for this analysis end-to-end can be found by skipping ahead <span class="co">[</span><span class="ot">here</span><span class="co">](#endtoend)</span>.</span>
<span id="cb25-44"><a href="#cb25-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-45"><a href="#cb25-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-46"><a href="#cb25-46" aria-hidden="true" tabindex="-1"></a><span class="fu">#  A quick look at the tissue</span></span>
<span id="cb25-47"><a href="#cb25-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-48"><a href="#cb25-48" aria-hidden="true" tabindex="-1"></a>Here's a quick look at this breast cancer tissue, where unsupervised clustering was performed using both the pearson residual normalized RNA + protein.  Code for creating these unsupervised clusters is <span class="co">[</span><span class="ot">here</span><span class="co">](#multiunsup)</span>.  Next, we'll compare multiomic immune cell typing performance with RNA-only or protein-only cell typing.</span>
<span id="cb25-49"><a href="#cb25-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-50"><a href="#cb25-50" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb25-51"><a href="#cb25-51" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/xyp_multi_unsup.png" alt="Wide Image"&gt;</span>
<span id="cb25-52"><a href="#cb25-52" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb25-55"><a href="#cb25-55" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-56"><a href="#cb25-56" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-57"><a href="#cb25-57" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-58"><a href="#cb25-58" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.show: "hold"</span></span>
<span id="cb25-59"><a href="#cb25-59" aria-hidden="true" tabindex="-1"></a><span class="co">#| out-width: "6000px"</span></span>
<span id="cb25-60"><a href="#cb25-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-61"><a href="#cb25-61" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/xyp_multi_unsup.png"</span>)</span>
<span id="cb25-62"><a href="#cb25-62" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-63"><a href="#cb25-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-64"><a href="#cb25-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-65"><a href="#cb25-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-66"><a href="#cb25-66" aria-hidden="true" tabindex="-1"></a><span class="fu"># Evidence multiomics can improve cell typing</span></span>
<span id="cb25-67"><a href="#cb25-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-68"><a href="#cb25-68" aria-hidden="true" tabindex="-1"></a>We used the Adjusted Rand Index (ARI) metric to quantify the agreement between protein-only, RNA-only, and multiomic cell type labels (<span class="co">[</span><span class="ot">code</span><span class="co">](#ari)</span>).  ARI is a numerical measure of similarity between two data clusterings <span class="co">[</span><span class="ot">@ariref</span><span class="co">]</span>.  We found protein-only and RNA-only cell typing calls show more disagreement with each other than they do with with integrated multiomic cell type calls, with:</span>
<span id="cb25-69"><a href="#cb25-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-70"><a href="#cb25-70" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>ARI(multiomic,RNA-only) = 0.679</span>
<span id="cb25-71"><a href="#cb25-71" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>ARI(multiomic, protein-only) = 0.589</span>
<span id="cb25-72"><a href="#cb25-72" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>ARI(protein-only, RNA-only) = 0.483 </span>
<span id="cb25-73"><a href="#cb25-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-74"><a href="#cb25-74" aria-hidden="true" tabindex="-1"></a>We compared the agreement between celltype predictions and canonical protein markers and RNA markers, and found that the multiomic cell type labels often show stronger alignment than RNA-only or protein-only cell typing. ( Code for <span class="co">[</span><span class="ot">RNA-only</span><span class="co">](#rnaonly)</span>, <span class="co">[</span><span class="ot">protein-only</span><span class="co">](#proteinonly)</span>, <span class="co">[</span><span class="ot">multiomic</span><span class="co">](#multiomic)</span> HieraType calls.)</span>
<span id="cb25-75"><a href="#cb25-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-76"><a href="#cb25-76" aria-hidden="true" tabindex="-1"></a><span class="fu">### T cell improvement demonstration via area under the ROC curves</span></span>
<span id="cb25-77"><a href="#cb25-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-78"><a href="#cb25-78" aria-hidden="true" tabindex="-1"></a>For example, the figure below shows ROC curves for T cell predictions along with three canonical cell type markers for T cells (one protein, and two RNA).  In this context, for a given level of marker expression, '*sensitivity*' refers to the frequency a cell is labeled a T-cell (for all cells above that threshold of marker expression). '*Specificity*' refers to the proportion of times a cell is labeled anything other than a T-cell (for all cells below that threshold).  </span>
<span id="cb25-79"><a href="#cb25-79" aria-hidden="true" tabindex="-1"></a>If we treat our normalized marker expression as a gold standard for cell type calling, then we would want to see high sensitivity and specificity of the T cell calls depending on whether marker expression is high or low.  In general, larger area under the ROC curve indicates better agreement between cell type calls and marker expression.</span>
<span id="cb25-80"><a href="#cb25-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-81"><a href="#cb25-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-84"><a href="#cb25-84" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-85"><a href="#cb25-85" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb25-86"><a href="#cb25-86" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-87"><a href="#cb25-87" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.show: "hold"</span></span>
<span id="cb25-88"><a href="#cb25-88" aria-hidden="true" tabindex="-1"></a><span class="co">#| out-width: "6000px"</span></span>
<span id="cb25-89"><a href="#cb25-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-90"><a href="#cb25-90" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/rocplot_tcell_lbl.png"</span>)</span>
<span id="cb25-91"><a href="#cb25-91" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-92"><a href="#cb25-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-93"><a href="#cb25-93" aria-hidden="true" tabindex="-1"></a>Above, as we might expect, we see that using  &lt;span style="color:#377EB8;"&gt;RNA-only&lt;/span&gt; for calling T cells does not optimize agreement between CD3 protein (AUC=0.871).  Unsurprisingly, agreement improves when using CD3 protein directly for celltyping (&lt;span style="color:#4DAF4A;"&gt;protein-only&lt;/span&gt; AUC=0.927).  In this case, we actually observed the best AUC when both RNA and protein were used  (&lt;span style="color:#E41A1C;"&gt;multiomics&lt;/span&gt; AUC=0.937).</span>
<span id="cb25-94"><a href="#cb25-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-95"><a href="#cb25-95" aria-hidden="true" tabindex="-1"></a>A perhaps unsurprising theme that emerged from these cell typing comparisons was that we often see the worst agreement for protein markers when using RNA-only for celltyping, and the worst agreement for RNA markers when using protein-only for celltyping.  Meanwhile, using both RNA and protein analytes often yields the best or nearly-best agreement for both marker types.</span>
<span id="cb25-96"><a href="#cb25-96" aria-hidden="true" tabindex="-1"></a>This is further shown below in the middle-and-right panels when looking at CD3D and CD2 RNA markers.  &lt;span style="color:#E41A1C;"&gt;Multiomic&lt;/span&gt; celltyping outperforms protein-only by the widest margin, and also slightly outperforms &lt;span style="color:#377EB8;"&gt;RNA-only&lt;/span&gt;.</span>
<span id="cb25-97"><a href="#cb25-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-98"><a href="#cb25-98" aria-hidden="true" tabindex="-1"></a>One interesting point of note when comparing the shapes of the ROC curves above is the smoothness of the curve for the CD3 protein marker, compared to the CD2 and CD3D RNA markers (which show a sharper bump closer to the lower-left corner where sensitivity=0 and specificity=1).  Due to the sparsity of RNA counts, we see a sharp spike in sensitivity close to an expression threshold at which cells have a single count of the marker gene. Because RNA-only cell typing uses only the RNA markers, the AUC is larger and more influenced around this local threshold of a single marker gene count, while protein and multiomic cell type labels are impacted directly by marker protein expression. </span>
<span id="cb25-99"><a href="#cb25-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-100"><a href="#cb25-100" aria-hidden="true" tabindex="-1"></a><span class="fu">###  Improved AUC across immune cell types</span></span>
<span id="cb25-101"><a href="#cb25-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-102"><a href="#cb25-102" aria-hidden="true" tabindex="-1"></a>We saw a similar trend as that demonstrated above across immune cell types.  </span>
<span id="cb25-103"><a href="#cb25-103" aria-hidden="true" tabindex="-1"></a>Here, for the most important 'index markers' of each immune cell type, we show the area under the ROC curve for RNA-only, protein-only, and multiomic celltyping results (code for this analysis can be found here in the <span class="co">[</span><span class="ot">end-to-end</span><span class="co">](#endtoend)</span> code section).  In general, there appears to be an improvement in AUC for both RNA and protein markers compared to if one of the analytes were removed. </span>
<span id="cb25-104"><a href="#cb25-104" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb25-105"><a href="#cb25-105" aria-hidden="true" tabindex="-1"></a>Average AUC was highest for multiomic celltyping across all markers and celltypes, with</span>
<span id="cb25-106"><a href="#cb25-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-107"><a href="#cb25-107" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>$\overline{AUC}(\text{multiomic}) = 0.711$</span>
<span id="cb25-108"><a href="#cb25-108" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>$\overline{AUC}(\text{RNA-only}) = 0.709$</span>
<span id="cb25-109"><a href="#cb25-109" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>$\overline{AUC}(\text{protein-only}) = 0.701$</span>
<span id="cb25-110"><a href="#cb25-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-113"><a href="#cb25-113" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-114"><a href="#cb25-114" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb25-115"><a href="#cb25-115" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-116"><a href="#cb25-116" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.show: "hold"</span></span>
<span id="cb25-117"><a href="#cb25-117" aria-hidden="true" tabindex="-1"></a><span class="co">#| out-width: "6000px"</span></span>
<span id="cb25-118"><a href="#cb25-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-119"><a href="#cb25-119" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/aucbarplot_wtext.png"</span>)</span>
<span id="cb25-120"><a href="#cb25-120" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-121"><a href="#cb25-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-122"><a href="#cb25-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-123"><a href="#cb25-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-124"><a href="#cb25-124" aria-hidden="true" tabindex="-1"></a><span class="fu"># Running HieraType using Protein + RNA {#multiomic}</span></span>
<span id="cb25-125"><a href="#cb25-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-126"><a href="#cb25-126" aria-hidden="true" tabindex="-1"></a>Below, we'll step through the process of generating cell type labels with multiomic data.</span>
<span id="cb25-127"><a href="#cb25-127" aria-hidden="true" tabindex="-1"></a>The code for using HieraType with protein data uses very similar syntax as a regular HieraType call, with a few exceptions worth calling out.</span>
<span id="cb25-128"><a href="#cb25-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-129"><a href="#cb25-129" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Protein naming convention</span></span>
<span id="cb25-130"><a href="#cb25-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-131"><a href="#cb25-131" aria-hidden="true" tabindex="-1"></a>HieraType works from a set of pre-defined 'markers' which correspond to each cell type we hope to classify.  For multiomic or protein-only data, protein marker names must end with the suffix "_protein" in our <span class="in">`markerslist`</span> objects.  </span>
<span id="cb25-132"><a href="#cb25-132" aria-hidden="true" tabindex="-1"></a>For example, here is the HieraType package multiomic markerslist for CD8 T cells (<span class="in">`cd8t`</span>). </span>
<span id="cb25-133"><a href="#cb25-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-134"><a href="#cb25-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-137"><a href="#cb25-137" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-138"><a href="#cb25-138" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-139"><a href="#cb25-139" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-140"><a href="#cb25-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-141"><a href="#cb25-141" aria-hidden="true" tabindex="-1"></a>example_multi_markerslist <span class="ot">&lt;-</span> utils<span class="sc">::</span><span class="fu">capture.output</span>(<span class="fu">print</span>(</span>
<span id="cb25-142"><a href="#cb25-142" aria-hidden="true" tabindex="-1"></a>  HieraType<span class="sc">::</span>markerslist_multiomic_tcellmajor<span class="sc">$</span>cd8t</span>
<span id="cb25-143"><a href="#cb25-143" aria-hidden="true" tabindex="-1"></a> ))</span>
<span id="cb25-144"><a href="#cb25-144" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(example_multi_markerslist, <span class="st">"figures/example_multi_markerslist.rds"</span>)</span>
<span id="cb25-145"><a href="#cb25-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-146"><a href="#cb25-146" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-149"><a href="#cb25-149" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-150"><a href="#cb25-150" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-151"><a href="#cb25-151" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-152"><a href="#cb25-152" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(HieraType<span class="sc">::</span>markerslist_multiomic_tcellmajor<span class="sc">$</span>cd8t)</span>
<span id="cb25-153"><a href="#cb25-153" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-154"><a href="#cb25-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-157"><a href="#cb25-157" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-158"><a href="#cb25-158" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb25-159"><a href="#cb25-159" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-160"><a href="#cb25-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-161"><a href="#cb25-161" aria-hidden="true" tabindex="-1"></a>example_multi_markerslist <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"figures/example_multi_markerslist.rds"</span>)</span>
<span id="cb25-162"><a href="#cb25-162" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">paste</span>(example_multi_markerslist, <span class="at">collapse =</span> <span class="st">'</span><span class="sc">\n</span><span class="st">'</span>))</span>
<span id="cb25-163"><a href="#cb25-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-164"><a href="#cb25-164" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-165"><a href="#cb25-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-166"><a href="#cb25-166" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Protein data normalization and constructing a multiomic adjacency_matrix</span></span>
<span id="cb25-167"><a href="#cb25-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-168"><a href="#cb25-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-169"><a href="#cb25-169" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb25-170"><a href="#cb25-170" aria-hidden="true" tabindex="-1"></a>Below, we make use of the <span class="co">[</span><span class="ot">scPearsonPCA package</span><span class="co">](https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/scPearsonPCA)</span>.  This R package provides convenience functions for getting the principal component embeddings for pearson-residual normalized data directly from a raw counts matrix.  When normalized protein is also available, we can get the PC decomposition from both RNA and normalized protein using the functions below.   For more detail on the package, please check out the <span class="co">[</span><span class="ot">README</span><span class="co">](https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/scPearsonPCA)</span>. </span>
<span id="cb25-171"><a href="#cb25-171" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-172"><a href="#cb25-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-173"><a href="#cb25-173" aria-hidden="true" tabindex="-1"></a>Protein data should be normalized, then combined with a sparse un-normalized RNA counts matrix before passing to HieraType.</span>
<span id="cb25-174"><a href="#cb25-174" aria-hidden="true" tabindex="-1"></a>Here we'll use pearson residuals from a gamma regression model as a normalization method, although other reasonable methods might be considered and left to the user's disgression.  We then add the "_protein" suffix to the protein names, so they'll align with the HieraType markerlists.</span>
<span id="cb25-175"><a href="#cb25-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-178"><a href="#cb25-178" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-179"><a href="#cb25-179" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-180"><a href="#cb25-180" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-181"><a href="#cb25-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-182"><a href="#cb25-182" aria-hidden="true" tabindex="-1"></a><span class="do">## normalize each protein taking pearson residuals from gamma generalized linear model.</span></span>
<span id="cb25-183"><a href="#cb25-183" aria-hidden="true" tabindex="-1"></a>pgamm <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">pearson_normalize_gamma_glm</span>(<span class="fu">as.matrix</span>(sem_protein[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb25-184"><a href="#cb25-184" aria-hidden="true" tabindex="-1"></a>                                                   ,<span class="at">upper_q_thresh =</span> <span class="fl">0.99</span></span>
<span id="cb25-185"><a href="#cb25-185" aria-hidden="true" tabindex="-1"></a>                                                   ,<span class="at">lower_q_thresh =</span> <span class="fl">0.01</span></span>
<span id="cb25-186"><a href="#cb25-186" aria-hidden="true" tabindex="-1"></a>                                                   )</span>
<span id="cb25-187"><a href="#cb25-187" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pgamm) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rownames</span>(pgamm), <span class="st">"_protein"</span>)</span>
<span id="cb25-188"><a href="#cb25-188" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-189"><a href="#cb25-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-190"><a href="#cb25-190" aria-hidden="true" tabindex="-1"></a>Here we conduct a principal component decomposition from combined RNA and proteins across all cells; this can be used to give us the cells x cells similarity graph we'll pass to HieraType.</span>
<span id="cb25-193"><a href="#cb25-193" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-194"><a href="#cb25-194" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-195"><a href="#cb25-195" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-196"><a href="#cb25-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-197"><a href="#cb25-197" aria-hidden="true" tabindex="-1"></a><span class="do">## select genes from RNA data to use for PCA analysis.</span></span>
<span id="cb25-198"><a href="#cb25-198" aria-hidden="true" tabindex="-1"></a><span class="do">## Take the top 2k features as well as any pre-defined celltyping markers from HieraType</span></span>
<span id="cb25-199"><a href="#cb25-199" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindVariableFeatures</span>(sem_rna, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb25-200"><a href="#cb25-200" aria-hidden="true" tabindex="-1"></a>markerg <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">c</span>(</span>
<span id="cb25-201"><a href="#cb25-201" aria-hidden="true" tabindex="-1"></a>  HieraType<span class="sc">::</span>markerslist_l1</span>
<span id="cb25-202"><a href="#cb25-202" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_immune</span>
<span id="cb25-203"><a href="#cb25-203" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd8tminor</span>
<span id="cb25-204"><a href="#cb25-204" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd4tminor</span>
<span id="cb25-205"><a href="#cb25-205" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_tcellmajor</span>
<span id="cb25-206"><a href="#cb25-206" aria-hidden="true" tabindex="-1"></a>), <span class="st">"[["</span>, <span class="st">"predictors"</span>))</span>
<span id="cb25-207"><a href="#cb25-207" aria-hidden="true" tabindex="-1"></a>use_genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(sem_rna<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>var.features, <span class="fu">intersect</span>(markerg, <span class="fu">rownames</span>(sem_rna))))</span>
<span id="cb25-208"><a href="#cb25-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-209"><a href="#cb25-209" aria-hidden="true" tabindex="-1"></a><span class="do">### total counts and gene frequency from RNA data computed across full dataset</span></span>
<span id="cb25-210"><a href="#cb25-210" aria-hidden="true" tabindex="-1"></a>tc <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb25-211"><a href="#cb25-211" aria-hidden="true" tabindex="-1"></a>genefreq <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">gene_frequency</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb25-212"><a href="#cb25-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-213"><a href="#cb25-213" aria-hidden="true" tabindex="-1"></a><span class="co"># Multiomic RNA and Protein cluster analysis</span></span>
<span id="cb25-214"><a href="#cb25-214" aria-hidden="true" tabindex="-1"></a>pcaobj_multi <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">sparse_quasipoisson_pca_seurat_multiomic</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,]</span>
<span id="cb25-215"><a href="#cb25-215" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">totalcounts =</span> tc</span>
<span id="cb25-216"><a href="#cb25-216" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">grate =</span> genefreq[use_genes]</span>
<span id="cb25-217"><a href="#cb25-217" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">xother =</span> pgamm <span class="do">## normalized protein</span></span>
<span id="cb25-218"><a href="#cb25-218" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">do.scale =</span> <span class="cn">TRUE</span></span>
<span id="cb25-219"><a href="#cb25-219" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">do.center =</span> <span class="cn">TRUE</span></span>
<span id="cb25-220"><a href="#cb25-220" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">scale.max =</span> <span class="dv">10</span></span>
<span id="cb25-221"><a href="#cb25-221" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">ncores =</span> <span class="dv">1</span></span>
<span id="cb25-222"><a href="#cb25-222" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-223"><a href="#cb25-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-224"><a href="#cb25-224" aria-hidden="true" tabindex="-1"></a><span class="do">### Make a cell-cell adjacency matrix for HieraType </span></span>
<span id="cb25-225"><a href="#cb25-225" aria-hidden="true" tabindex="-1"></a>multi_simgrph <span class="ot">&lt;-</span> uwot<span class="sc">::</span><span class="fu">similarity_graph</span>(pcaobj_multi<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings</span>
<span id="cb25-226"><a href="#cb25-226" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">n_neighbors =</span> <span class="dv">30</span></span>
<span id="cb25-227"><a href="#cb25-227" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">nn_method =</span> <span class="st">"annoy"</span></span>
<span id="cb25-228"><a href="#cb25-228" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">metric =</span> <span class="st">'cosine'</span></span>
<span id="cb25-229"><a href="#cb25-229" aria-hidden="true" tabindex="-1"></a>                                  )</span>
<span id="cb25-230"><a href="#cb25-230" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(multi_simgrph) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(multi_simgrph) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(pcaobj_multi<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings)</span>
<span id="cb25-231"><a href="#cb25-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-232"><a href="#cb25-232" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-233"><a href="#cb25-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-234"><a href="#cb25-234" aria-hidden="true" tabindex="-1"></a><span class="fu">##### HieraType call</span></span>
<span id="cb25-235"><a href="#cb25-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-236"><a href="#cb25-236" aria-hidden="true" tabindex="-1"></a>Here we construct our 'multiomic' expression matrix which includes our sparse RNA counts with the normalized protein expression, then we call <span class="in">`HieraType::run_pipeline`</span> .</span>
<span id="cb25-239"><a href="#cb25-239" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-240"><a href="#cb25-240" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-241"><a href="#cb25-241" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-242"><a href="#cb25-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-243"><a href="#cb25-243" aria-hidden="true" tabindex="-1"></a><span class="do">#### combining sparse RNA counts and normalized proteins into a single matrix 'multimat'</span></span>
<span id="cb25-244"><a href="#cb25-244" aria-hidden="true" tabindex="-1"></a>overlapping_cells <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(sem_protein), <span class="fu">colnames</span>(sem_rna))</span>
<span id="cb25-245"><a href="#cb25-245" aria-hidden="true" tabindex="-1"></a>multimat <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">t</span>(pgamm[,overlapping_cells])</span>
<span id="cb25-246"><a href="#cb25-246" aria-hidden="true" tabindex="-1"></a>multimat <span class="ot">&lt;-</span> <span class="fu">cbind</span>(multimat, Matrix<span class="sc">::</span><span class="fu">t</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,overlapping_cells]))</span>
<span id="cb25-247"><a href="#cb25-247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-248"><a href="#cb25-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-249"><a href="#cb25-249" aria-hidden="true" tabindex="-1"></a><span class="do">### Define pipeline, using multiomic markerslists</span></span>
<span id="cb25-250"><a href="#cb25-250" aria-hidden="true" tabindex="-1"></a>pipeline_io_multiomics <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">make_pipeline</span>(</span>
<span id="cb25-251"><a href="#cb25-251" aria-hidden="true" tabindex="-1"></a>  <span class="at">markerslists =</span> <span class="fu">list</span>(</span>
<span id="cb25-252"><a href="#cb25-252" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l1"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_l1</span>
<span id="cb25-253"><a href="#cb25-253" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"l2"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_immune</span>
<span id="cb25-254"><a href="#cb25-254" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_tcellmajor</span>
<span id="cb25-255"><a href="#cb25-255" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd4tminor</span>
<span id="cb25-256"><a href="#cb25-256" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd8tminor</span>
<span id="cb25-257"><a href="#cb25-257" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-258"><a href="#cb25-258" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors =</span> <span class="fu">list</span>(</span>
<span id="cb25-259"><a href="#cb25-259" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"l1"</span></span>
<span id="cb25-260"><a href="#cb25-260" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"l2"</span></span>
<span id="cb25-261"><a href="#cb25-261" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb25-262"><a href="#cb25-262" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb25-263"><a href="#cb25-263" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-264"><a href="#cb25-264" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors_category =</span> <span class="fu">list</span>(</span>
<span id="cb25-265"><a href="#cb25-265" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"immune"</span></span>
<span id="cb25-266"><a href="#cb25-266" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"tcell"</span></span>
<span id="cb25-267"><a href="#cb25-267" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"cd4t"</span></span>
<span id="cb25-268"><a href="#cb25-268" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"cd8t"</span></span>
<span id="cb25-269"><a href="#cb25-269" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-270"><a href="#cb25-270" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-271"><a href="#cb25-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-272"><a href="#cb25-272" aria-hidden="true" tabindex="-1"></a><span class="do">### Run the celltyping pipeline</span></span>
<span id="cb25-273"><a href="#cb25-273" aria-hidden="true" tabindex="-1"></a>multictobj <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(</span>
<span id="cb25-274"><a href="#cb25-274" aria-hidden="true" tabindex="-1"></a>   <span class="at">pipeline =</span> pipeline_io_multiomics</span>
<span id="cb25-275"><a href="#cb25-275" aria-hidden="true" tabindex="-1"></a>   ,<span class="at">totalcounts=</span> tc[<span class="fu">rownames</span>(multimat)]</span>
<span id="cb25-276"><a href="#cb25-276" aria-hidden="true" tabindex="-1"></a>   ,<span class="at">gene_wise_frequency =</span> genefreq</span>
<span id="cb25-277"><a href="#cb25-277" aria-hidden="true" tabindex="-1"></a>   ,<span class="at">counts_matrix =</span> multimat </span>
<span id="cb25-278"><a href="#cb25-278" aria-hidden="true" tabindex="-1"></a>   ,<span class="at">adjacency_matrix =</span> multi_simgrph[<span class="fu">rownames</span>(multimat), <span class="fu">rownames</span>(multimat)]</span>
<span id="cb25-279"><a href="#cb25-279" aria-hidden="true" tabindex="-1"></a> )</span>
<span id="cb25-280"><a href="#cb25-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-281"><a href="#cb25-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-282"><a href="#cb25-282" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-283"><a href="#cb25-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-284"><a href="#cb25-284" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Outputs</span></span>
<span id="cb25-285"><a href="#cb25-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-286"><a href="#cb25-286" aria-hidden="true" tabindex="-1"></a>For extensive description of the outputs we get from HieraType, we refer the reader to the original post.</span>
<span id="cb25-287"><a href="#cb25-287" aria-hidden="true" tabindex="-1"></a>Here, we'll demonstrate how to make a few marker heatmaps from the RNA and protein data to sanity check our cell typing results.</span>
<span id="cb25-290"><a href="#cb25-290" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-291"><a href="#cb25-291" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-292"><a href="#cb25-292" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-293"><a href="#cb25-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-294"><a href="#cb25-294" aria-hidden="true" tabindex="-1"></a>multictobj<span class="sc">$</span>post_probs[[<span class="dv">1</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>] <span class="do">## posterior probabilities for each cell type</span></span>
<span id="cb25-295"><a href="#cb25-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-296"><a href="#cb25-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-297"><a href="#cb25-297" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-298"><a href="#cb25-298" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-301"><a href="#cb25-301" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-302"><a href="#cb25-302" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb25-303"><a href="#cb25-303" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-304"><a href="#cb25-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-305"><a href="#cb25-305" aria-hidden="true" tabindex="-1"></a><span class="co">#saveRDS(multictobj$post_probs[[1]][1:10], file="figures/postprobs_head.rds")</span></span>
<span id="cb25-306"><a href="#cb25-306" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">readRDS</span>(<span class="st">"figures/postprobs_head.rds"</span>))</span>
<span id="cb25-307"><a href="#cb25-307" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-310"><a href="#cb25-310" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-311"><a href="#cb25-311" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-312"><a href="#cb25-312" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-313"><a href="#cb25-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-314"><a href="#cb25-314" aria-hidden="true" tabindex="-1"></a><span class="do">### Use the major cell type categories for a heatmap</span></span>
<span id="cb25-315"><a href="#cb25-315" aria-hidden="true" tabindex="-1"></a>ppmulti <span class="ot">&lt;-</span> <span class="fu">copy</span>(multictobj<span class="sc">$</span>post_probs<span class="sc">$</span>l1)</span>
<span id="cb25-316"><a href="#cb25-316" aria-hidden="true" tabindex="-1"></a>ppmulti[,celltype_multi_major<span class="sc">:</span><span class="er">=</span>celltype_granular]</span>
<span id="cb25-317"><a href="#cb25-317" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd4"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd4_tcell"</span>]</span>
<span id="cb25-318"><a href="#cb25-318" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd8"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd8_tcell"</span>]</span>
<span id="cb25-319"><a href="#cb25-319" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"mono|macro"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"macrophage_monocyte"</span>]</span>
<span id="cb25-320"><a href="#cb25-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-321"><a href="#cb25-321" aria-hidden="true" tabindex="-1"></a><span class="do">### fold change metrics to show in RNA heatmap</span></span>
<span id="cb25-322"><a href="#cb25-322" aria-hidden="true" tabindex="-1"></a>fctbl_rna <span class="ot">&lt;-</span> </span>
<span id="cb25-323"><a href="#cb25-323" aria-hidden="true" tabindex="-1"></a><span class="fu">clusterwise_foldchange_metrics</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,]</span>
<span id="cb25-324"><a href="#cb25-324" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">totalcounts =</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb25-325"><a href="#cb25-325" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">metadata =</span> ppmulti</span>
<span id="cb25-326"><a href="#cb25-326" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">cluster_column =</span> <span class="st">"celltype_multi_major"</span></span>
<span id="cb25-327"><a href="#cb25-327" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">cellid_column =</span> <span class="st">"cell_ID"</span></span>
<span id="cb25-328"><a href="#cb25-328" aria-hidden="true" tabindex="-1"></a>                               )</span>
<span id="cb25-329"><a href="#cb25-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-330"><a href="#cb25-330" aria-hidden="true" tabindex="-1"></a>hm_rna <span class="ot">&lt;-</span> <span class="fu">add_ncells</span>(<span class="fu">marker_heatmap</span>(fctbl_rna))</span>
<span id="cb25-331"><a href="#cb25-331" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hm_rna)</span>
<span id="cb25-332"><a href="#cb25-332" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-333"><a href="#cb25-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-336"><a href="#cb25-336" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-337"><a href="#cb25-337" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-338"><a href="#cb25-338" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-339"><a href="#cb25-339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-340"><a href="#cb25-340" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">plot=</span>hm_rna, <span class="at">filename =</span> <span class="st">"figures/hm_rna_multi.png"</span>, <span class="at">width =</span> <span class="dv">1522</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">height =</span> <span class="dv">430</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">dpi=</span><span class="dv">400</span>,<span class="at">units=</span><span class="st">'px'</span>)</span>
<span id="cb25-341"><a href="#cb25-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-342"><a href="#cb25-342" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-343"><a href="#cb25-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-344"><a href="#cb25-344" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb25-345"><a href="#cb25-345" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/hm_rna_multi.png" alt="Wide Image"&gt;</span>
<span id="cb25-346"><a href="#cb25-346" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb25-347"><a href="#cb25-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-348"><a href="#cb25-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-349"><a href="#cb25-349" aria-hidden="true" tabindex="-1"></a>For the protein data, almost all proteins have some positive measurement in each cell.  We can use the pearson residuals to set a cutoff for computing the proportion of positive cells by marker and by cell type.  Below, we'll use cutoff of 'pearson residual &gt; 1' to determine the size of the bubble in the heatmap.</span>
<span id="cb25-352"><a href="#cb25-352" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-353"><a href="#cb25-353" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-354"><a href="#cb25-354" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-355"><a href="#cb25-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-356"><a href="#cb25-356" aria-hidden="true" tabindex="-1"></a>fctbl_protein <span class="ot">&lt;-</span> </span>
<span id="cb25-357"><a href="#cb25-357" aria-hidden="true" tabindex="-1"></a><span class="fu">clusterwise_foldchange_metrics_protein</span>(sem_protein[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[,overlapping_cells]</span>
<span id="cb25-358"><a href="#cb25-358" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">metadata =</span> ppmulti</span>
<span id="cb25-359"><a href="#cb25-359" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">cluster_column =</span> <span class="st">"celltype_multi_major"</span></span>
<span id="cb25-360"><a href="#cb25-360" aria-hidden="true" tabindex="-1"></a>                               ,<span class="at">propd =</span> (pgamm[,overlapping_cells] <span class="sc">&gt;</span> <span class="dv">1</span>))</span>
<span id="cb25-361"><a href="#cb25-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-362"><a href="#cb25-362" aria-hidden="true" tabindex="-1"></a>hm_protein <span class="ot">&lt;-</span> <span class="fu">add_ncells</span>(<span class="fu">marker_heatmap</span>(fctbl_protein))</span>
<span id="cb25-363"><a href="#cb25-363" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hm_protein)</span>
<span id="cb25-364"><a href="#cb25-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-365"><a href="#cb25-365" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-368"><a href="#cb25-368" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-369"><a href="#cb25-369" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-370"><a href="#cb25-370" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-371"><a href="#cb25-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-372"><a href="#cb25-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-373"><a href="#cb25-373" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">plot=</span>hm_protein, <span class="at">filename =</span> <span class="st">"figures/hm_protein_multi.png"</span>, <span class="at">width =</span> <span class="dv">1522</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">height =</span> <span class="dv">430</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">dpi=</span><span class="dv">400</span>,<span class="at">units=</span><span class="st">'px'</span>)</span>
<span id="cb25-374"><a href="#cb25-374" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-375"><a href="#cb25-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-376"><a href="#cb25-376" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-377"><a href="#cb25-377" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb25-378"><a href="#cb25-378" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/hm_protein_multi.png" alt="Wide Image"&gt;</span>
<span id="cb25-379"><a href="#cb25-379" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb25-380"><a href="#cb25-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-381"><a href="#cb25-381" aria-hidden="true" tabindex="-1"></a><span class="fu"># Integrating unsupervised clusters with immune cell type labels</span></span>
<span id="cb25-382"><a href="#cb25-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-383"><a href="#cb25-383" aria-hidden="true" tabindex="-1"></a>HieraType was designed to find celltypes with known, pre-specified markers.  Often, unsupervised clustering is sufficiently granular to identify very distinct tissue-specific clusters, but may not identify commune immune cell types of interest due to relative sparsity of the immune markers.  We demonstrate an easy way (<span class="co">[</span><span class="ot">code</span><span class="co">](#integrateunsup)</span>) to  reconcile the labels between HieraType and unsupervised clustering; keeping the HieraType label when a cell is classified as immune, and taking the unsupervised cluster label otherwise.  This can be a useful technique regardless of whether one is using RNA, protein, or both to cell type their single cell dataset.</span>
<span id="cb25-384"><a href="#cb25-384" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb25-385"><a href="#cb25-385" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/xyp_multi_semisup.png" alt="Wide Image"&gt;</span>
<span id="cb25-386"><a href="#cb25-386" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb25-389"><a href="#cb25-389" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-390"><a href="#cb25-390" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-391"><a href="#cb25-391" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-392"><a href="#cb25-392" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.show: "hold"</span></span>
<span id="cb25-393"><a href="#cb25-393" aria-hidden="true" tabindex="-1"></a><span class="co">#| out-width: "12000px"</span></span>
<span id="cb25-394"><a href="#cb25-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-395"><a href="#cb25-395" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/xyp_multi_semisup.png"</span>)</span>
<span id="cb25-396"><a href="#cb25-396" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-397"><a href="#cb25-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-398"><a href="#cb25-398" aria-hidden="true" tabindex="-1"></a><span class="fu"># Conclusion</span></span>
<span id="cb25-399"><a href="#cb25-399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-400"><a href="#cb25-400" aria-hidden="true" tabindex="-1"></a>The HieraType method constructs continuous celltype metagene scores, which are then probabilistically clustered to classify cell types. </span>
<span id="cb25-401"><a href="#cb25-401" aria-hidden="true" tabindex="-1"></a>Because these metagene scores are continuous and the clustering uses flexible gaussian mixture models, the inputted expression can be fluidly specified in terms of RNA markers, protein markers, or some combination of both.  This allows for an elegant and efficient solution for fully integrated celltyping when both protein and RNA are available.  </span>
<span id="cb25-402"><a href="#cb25-402" aria-hidden="true" tabindex="-1"></a>Overall, we found improved cell typing performance in the breast cancer sample analyzed in this post when both analytes were used compared to either analyte on its own.</span>
<span id="cb25-403"><a href="#cb25-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-404"><a href="#cb25-404" aria-hidden="true" tabindex="-1"></a>While the HieraType method is fully supervised and cannot identify novel cell types not specified in a '<span class="in">`markerslist`</span>',  we demonstrate an easy approach for integrating unsupervised clusters with HieraType immune cell typing calls, enabling high resolution cell typing for both immune cell types as well as epithelial or tissue context specific cell types which may not have a known marker profile a priori.</span>
<span id="cb25-405"><a href="#cb25-405" aria-hidden="true" tabindex="-1"></a>This can help reap benefits of both unsupervised and supervised approaches - well controlled pre-specified immune cell typing combined with granular unsupervised epithelial clusters.</span>
<span id="cb25-406"><a href="#cb25-406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-407"><a href="#cb25-407" aria-hidden="true" tabindex="-1"></a><span class="fu"># End-to-end detailed analysis code {#endtoend}</span></span>
<span id="cb25-408"><a href="#cb25-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-409"><a href="#cb25-409" aria-hidden="true" tabindex="-1"></a>For completeness, here are a few end-to-end code blocks for each analysis step.</span>
<span id="cb25-410"><a href="#cb25-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-411"><a href="#cb25-411" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Reading in the data  </span></span>
<span id="cb25-414"><a href="#cb25-414" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-415"><a href="#cb25-415" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-416"><a href="#cb25-416" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-417"><a href="#cb25-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-418"><a href="#cb25-418" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb25-419"><a href="#cb25-419" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb25-420"><a href="#cb25-420" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb25-421"><a href="#cb25-421" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scPearsonPCA)</span>
<span id="cb25-422"><a href="#cb25-422" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(HieraType)</span>
<span id="cb25-423"><a href="#cb25-423" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb25-424"><a href="#cb25-424" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pROC)</span>
<span id="cb25-425"><a href="#cb25-425" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mclust)</span>
<span id="cb25-426"><a href="#cb25-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-427"><a href="#cb25-427" aria-hidden="true" tabindex="-1"></a><span class="do">## Note, using a v3-style Seurat Object below, i.e., `sem[["RNA"]]@counts`</span></span>
<span id="cb25-428"><a href="#cb25-428" aria-hidden="true" tabindex="-1"></a><span class="do">## For v5-style, the syntax would be like `sem[["RNA"]]$counts`</span></span>
<span id="cb25-429"><a href="#cb25-429" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">Seurat.object.assay.version =</span> <span class="st">"v3"</span>)  </span>
<span id="cb25-430"><a href="#cb25-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-431"><a href="#cb25-431" aria-hidden="true" tabindex="-1"></a>sem_protein <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Breast/sem_protein.rds"</span>)</span>
<span id="cb25-432"><a href="#cb25-432" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Breast/sem_rna.rds"</span>)</span>
<span id="cb25-433"><a href="#cb25-433" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> <span class="fu">subset</span>(sem_rna, nCount_RNA <span class="sc">&gt;=</span><span class="dv">100</span>)</span>
<span id="cb25-434"><a href="#cb25-434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-435"><a href="#cb25-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-436"><a href="#cb25-436" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-437"><a href="#cb25-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-438"><a href="#cb25-438" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-441"><a href="#cb25-441" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-442"><a href="#cb25-442" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-443"><a href="#cb25-443" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-444"><a href="#cb25-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-445"><a href="#cb25-445" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">Seurat.object.assay.version =</span> <span class="st">"v3"</span>)  </span>
<span id="cb25-446"><a href="#cb25-446" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">load_all</span>(<span class="st">"~/data/dan/imputation_smoothing/HieraType_dev"</span>, <span class="at">export_all =</span> <span class="cn">FALSE</span>)</span>
<span id="cb25-447"><a href="#cb25-447" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">load_all</span>(<span class="st">"~/data/dan/scPearsonPCA"</span>, <span class="at">export_all =</span> <span class="cn">FALSE</span>)</span>
<span id="cb25-448"><a href="#cb25-448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-449"><a href="#cb25-449" aria-hidden="true" tabindex="-1"></a><span class="do">#### Unsupervised clusters using PCA on combined Protein + RNA data </span></span>
<span id="cb25-450"><a href="#cb25-450" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb25-451"><a href="#cb25-451" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb25-452"><a href="#cb25-452" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb25-453"><a href="#cb25-453" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb25-454"><a href="#cb25-454" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-455"><a href="#cb25-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-456"><a href="#cb25-456" aria-hidden="true" tabindex="-1"></a>sem_protein <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"~/data/multiomics_hackathon/Breast/sem_protein.rds"</span>)</span>
<span id="cb25-457"><a href="#cb25-457" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"~/data/multiomics_hackathon/Breast/sem_rna.rds"</span>)</span>
<span id="cb25-458"><a href="#cb25-458" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> <span class="fu">subset</span>(sem_rna, nCount_RNA <span class="sc">&gt;=</span><span class="dv">100</span>)</span>
<span id="cb25-459"><a href="#cb25-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-460"><a href="#cb25-460" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-461"><a href="#cb25-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-462"><a href="#cb25-462" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Normalizing protein data</span></span>
<span id="cb25-463"><a href="#cb25-463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-466"><a href="#cb25-466" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-467"><a href="#cb25-467" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-468"><a href="#cb25-468" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-469"><a href="#cb25-469" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-470"><a href="#cb25-470" aria-hidden="true" tabindex="-1"></a><span class="do">## normalize each protein taking pearson residuals from gamma generalized linear model.</span></span>
<span id="cb25-471"><a href="#cb25-471" aria-hidden="true" tabindex="-1"></a>pgamm <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">pearson_normalize_gamma_glm</span>(<span class="fu">as.matrix</span>(sem_protein[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb25-472"><a href="#cb25-472" aria-hidden="true" tabindex="-1"></a>                                                   ,<span class="at">upper_q_thresh =</span> <span class="fl">0.99</span></span>
<span id="cb25-473"><a href="#cb25-473" aria-hidden="true" tabindex="-1"></a>                                                   ,<span class="at">lower_q_thresh =</span> <span class="fl">0.01</span></span>
<span id="cb25-474"><a href="#cb25-474" aria-hidden="true" tabindex="-1"></a>                                                   )</span>
<span id="cb25-475"><a href="#cb25-475" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pgamm) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rownames</span>(pgamm), <span class="st">"_protein"</span>)</span>
<span id="cb25-476"><a href="#cb25-476" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-477"><a href="#cb25-477" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-478"><a href="#cb25-478" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-479"><a href="#cb25-479" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Multiomic pearson residual PCA analysis using the protein and RNA {#multiunsup}</span></span>
<span id="cb25-480"><a href="#cb25-480" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-483"><a href="#cb25-483" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-484"><a href="#cb25-484" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-485"><a href="#cb25-485" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-486"><a href="#cb25-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-487"><a href="#cb25-487" aria-hidden="true" tabindex="-1"></a><span class="do">## select genes from RNA data to use for PCA analysis.</span></span>
<span id="cb25-488"><a href="#cb25-488" aria-hidden="true" tabindex="-1"></a><span class="do">## Will take the top 2k features as well as any pre-defined immune celltyping markers from HieraType</span></span>
<span id="cb25-489"><a href="#cb25-489" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindVariableFeatures</span>(sem_rna, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb25-490"><a href="#cb25-490" aria-hidden="true" tabindex="-1"></a>markerg <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">c</span>(</span>
<span id="cb25-491"><a href="#cb25-491" aria-hidden="true" tabindex="-1"></a>  HieraType<span class="sc">::</span>markerslist_l1</span>
<span id="cb25-492"><a href="#cb25-492" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_immune</span>
<span id="cb25-493"><a href="#cb25-493" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd8tminor</span>
<span id="cb25-494"><a href="#cb25-494" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd4tminor</span>
<span id="cb25-495"><a href="#cb25-495" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_tcellmajor</span>
<span id="cb25-496"><a href="#cb25-496" aria-hidden="true" tabindex="-1"></a>), <span class="st">"[["</span>, <span class="st">"predictors"</span>))</span>
<span id="cb25-497"><a href="#cb25-497" aria-hidden="true" tabindex="-1"></a>use_genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(sem_rna<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>var.features, <span class="fu">intersect</span>(markerg, <span class="fu">rownames</span>(sem_rna))))</span>
<span id="cb25-498"><a href="#cb25-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-499"><a href="#cb25-499" aria-hidden="true" tabindex="-1"></a><span class="do">### total counts and gene frequency from RNA data computed across full dataset</span></span>
<span id="cb25-500"><a href="#cb25-500" aria-hidden="true" tabindex="-1"></a>tc <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb25-501"><a href="#cb25-501" aria-hidden="true" tabindex="-1"></a>genefreq <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">gene_frequency</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb25-502"><a href="#cb25-502" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-503"><a href="#cb25-503" aria-hidden="true" tabindex="-1"></a><span class="do">### add a '_protein' suffix to the protein names - helps differentiate proteins and RNA's with the same name. </span></span>
<span id="cb25-504"><a href="#cb25-504" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pgamm) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rownames</span>(pgamm), <span class="st">"_protein"</span>)</span>
<span id="cb25-505"><a href="#cb25-505" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-506"><a href="#cb25-506" aria-hidden="true" tabindex="-1"></a><span class="do">### Principal components derived from the combined covariance matrix of pearson residual normalized RNA and protein.</span></span>
<span id="cb25-507"><a href="#cb25-507" aria-hidden="true" tabindex="-1"></a><span class="do">### </span></span>
<span id="cb25-508"><a href="#cb25-508" aria-hidden="true" tabindex="-1"></a>pcaobj_multi <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">sparse_quasipoisson_pca_seurat_multiomic</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,]</span>
<span id="cb25-509"><a href="#cb25-509" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">totalcounts =</span> tc</span>
<span id="cb25-510"><a href="#cb25-510" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">grate =</span> genefreq[use_genes]</span>
<span id="cb25-511"><a href="#cb25-511" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">xother =</span> pgamm <span class="do">## normalized protein</span></span>
<span id="cb25-512"><a href="#cb25-512" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">do.scale =</span> <span class="cn">TRUE</span></span>
<span id="cb25-513"><a href="#cb25-513" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">do.center =</span> <span class="cn">TRUE</span></span>
<span id="cb25-514"><a href="#cb25-514" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">scale.max =</span> <span class="dv">10</span></span>
<span id="cb25-515"><a href="#cb25-515" aria-hidden="true" tabindex="-1"></a>                                                                       ,<span class="at">ncores =</span> <span class="dv">1</span></span>
<span id="cb25-516"><a href="#cb25-516" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-517"><a href="#cb25-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-518"><a href="#cb25-518" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-519"><a href="#cb25-519" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-522"><a href="#cb25-522" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-523"><a href="#cb25-523" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-524"><a href="#cb25-524" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-525"><a href="#cb25-525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-526"><a href="#cb25-526" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(pcaobj_multi, <span class="st">"data2/pcaobj_multi.rds"</span>)</span>
<span id="cb25-527"><a href="#cb25-527" aria-hidden="true" tabindex="-1"></a>pcaobj_multi <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data2/pcaobj_multi.rds"</span>)</span>
<span id="cb25-528"><a href="#cb25-528" aria-hidden="true" tabindex="-1"></a>use_genes <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"_protein"</span>, <span class="fu">rownames</span>(pcaobj_multi<span class="sc">$</span>reduction.data<span class="sc">@</span>feature.loadings) , <span class="at">value=</span><span class="cn">TRUE</span>, <span class="at">invert=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-529"><a href="#cb25-529" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-530"><a href="#cb25-530" aria-hidden="true" tabindex="-1"></a>msg <span class="ot">&lt;-</span> utils<span class="sc">::</span><span class="fu">capture.output</span>(<span class="fu">print</span>(</span>
<span id="cb25-531"><a href="#cb25-531" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> pcaobj_multi<span class="sc">$</span>reduction.data,</span>
<span id="cb25-532"><a href="#cb25-532" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,</span>
<span id="cb25-533"><a href="#cb25-533" aria-hidden="true" tabindex="-1"></a>  <span class="at">nfeatures =</span> <span class="dv">20</span></span>
<span id="cb25-534"><a href="#cb25-534" aria-hidden="true" tabindex="-1"></a>))</span>
<span id="cb25-535"><a href="#cb25-535" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">paste</span>(msg, <span class="at">collapse =</span> <span class="st">'</span><span class="sc">\n</span><span class="st">'</span>))</span>
<span id="cb25-536"><a href="#cb25-536" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(msg, <span class="st">"figures/pca_multiomics_toppcs_msg.rds"</span>)</span>
<span id="cb25-537"><a href="#cb25-537" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-540"><a href="#cb25-540" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-541"><a href="#cb25-541" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-542"><a href="#cb25-542" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-545"><a href="#cb25-545" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-546"><a href="#cb25-546" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb25-547"><a href="#cb25-547" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-548"><a href="#cb25-548" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-549"><a href="#cb25-549" aria-hidden="true" tabindex="-1"></a>msg <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"figures/pca_multiomics_toppcs_msg.rds"</span>)</span>
<span id="cb25-550"><a href="#cb25-550" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">paste</span>(msg, <span class="at">collapse =</span> <span class="st">'</span><span class="sc">\n</span><span class="st">'</span>))</span>
<span id="cb25-551"><a href="#cb25-551" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-552"><a href="#cb25-552" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-553"><a href="#cb25-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-556"><a href="#cb25-556" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-557"><a href="#cb25-557" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-558"><a href="#cb25-558" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-559"><a href="#cb25-559" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-560"><a href="#cb25-560" aria-hidden="true" tabindex="-1"></a><span class="do">### Make a cell-cell adjacency matrix for HieraType </span></span>
<span id="cb25-561"><a href="#cb25-561" aria-hidden="true" tabindex="-1"></a>multi_simgrph <span class="ot">&lt;-</span> uwot<span class="sc">::</span><span class="fu">similarity_graph</span>(pcaobj_multi<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings</span>
<span id="cb25-562"><a href="#cb25-562" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">n_neighbors =</span> <span class="dv">30</span></span>
<span id="cb25-563"><a href="#cb25-563" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">nn_method =</span> <span class="st">"annoy"</span></span>
<span id="cb25-564"><a href="#cb25-564" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">metric =</span> <span class="st">'cosine'</span></span>
<span id="cb25-565"><a href="#cb25-565" aria-hidden="true" tabindex="-1"></a>                                  )</span>
<span id="cb25-566"><a href="#cb25-566" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(multi_simgrph) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(multi_simgrph) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(pcaobj_multi<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings)</span>
<span id="cb25-567"><a href="#cb25-567" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-568"><a href="#cb25-568" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-569"><a href="#cb25-569" aria-hidden="true" tabindex="-1"></a><span class="do">###  Unsupervised clusters using multiomic cell-cell adjacency graph</span></span>
<span id="cb25-570"><a href="#cb25-570" aria-hidden="true" tabindex="-1"></a>sem_rna[[<span class="st">"multigrph"</span>]] <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">as.Graph</span>(multi_simgrph[<span class="fu">colnames</span>(sem_rna),<span class="fu">colnames</span>(sem_rna)])</span>
<span id="cb25-571"><a href="#cb25-571" aria-hidden="true" tabindex="-1"></a>sem_rna <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindClusters</span>(sem_rna, <span class="at">graph.name=</span><span class="st">"multigrph"</span>)</span>
<span id="cb25-572"><a href="#cb25-572" aria-hidden="true" tabindex="-1"></a>sem_rna<span class="sc">@</span>meta.data<span class="sc">$</span>clusters_multi_unsup <span class="ot">&lt;-</span> sem_rna<span class="sc">@</span>meta.data<span class="sc">$</span>seurat_clusters</span>
<span id="cb25-573"><a href="#cb25-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-574"><a href="#cb25-574" aria-hidden="true" tabindex="-1"></a>xyp_multi_unsup <span class="ot">&lt;-</span> <span class="fu">xyplot</span>(<span class="st">"clusters_multi_unsup"</span>, <span class="at">metadata =</span> sem_rna<span class="sc">@</span>meta.data, <span class="at">ptsize =</span> <span class="fl">0.05</span>) <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb25-575"><a href="#cb25-575" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-576"><a href="#cb25-576" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"figures/xyp_multi_unsup.png"</span></span>
<span id="cb25-577"><a href="#cb25-577" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">plot=</span>xyp_multi_unsup</span>
<span id="cb25-578"><a href="#cb25-578" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">device =</span> <span class="st">"png"</span></span>
<span id="cb25-579"><a href="#cb25-579" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">width =</span> <span class="dv">1181</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>)</span>
<span id="cb25-580"><a href="#cb25-580" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">height=</span><span class="dv">689</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>)</span>
<span id="cb25-581"><a href="#cb25-581" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">units =</span> <span class="st">'px'</span></span>
<span id="cb25-582"><a href="#cb25-582" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-583"><a href="#cb25-583" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-584"><a href="#cb25-584" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-585"><a href="#cb25-585" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-586"><a href="#cb25-586" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-587"><a href="#cb25-587" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-588"><a href="#cb25-588" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Multiomic cell typing code using HieraType</span></span>
<span id="cb25-591"><a href="#cb25-591" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-592"><a href="#cb25-592" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-593"><a href="#cb25-593" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-594"><a href="#cb25-594" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-595"><a href="#cb25-595" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-596"><a href="#cb25-596" aria-hidden="true" tabindex="-1"></a><span class="do">## Define the pipeline using multiomic markerlists in HieraType package</span></span>
<span id="cb25-597"><a href="#cb25-597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-598"><a href="#cb25-598" aria-hidden="true" tabindex="-1"></a>overlapping_cells <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(sem_protein), <span class="fu">colnames</span>(sem_rna))</span>
<span id="cb25-599"><a href="#cb25-599" aria-hidden="true" tabindex="-1"></a>multimat <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">t</span>(pgamm[,overlapping_cells])</span>
<span id="cb25-600"><a href="#cb25-600" aria-hidden="true" tabindex="-1"></a>multimat <span class="ot">&lt;-</span> <span class="fu">cbind</span>(multimat, Matrix<span class="sc">::</span><span class="fu">t</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,overlapping_cells]))</span>
<span id="cb25-601"><a href="#cb25-601" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-602"><a href="#cb25-602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-603"><a href="#cb25-603" aria-hidden="true" tabindex="-1"></a>pipeline_io_multiomics <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">make_pipeline</span>(</span>
<span id="cb25-604"><a href="#cb25-604" aria-hidden="true" tabindex="-1"></a>  <span class="at">markerslists =</span> <span class="fu">list</span>(</span>
<span id="cb25-605"><a href="#cb25-605" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l1"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_l1</span>
<span id="cb25-606"><a href="#cb25-606" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"l2"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_immune</span>
<span id="cb25-607"><a href="#cb25-607" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_tcellmajor</span>
<span id="cb25-608"><a href="#cb25-608" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd4tminor</span>
<span id="cb25-609"><a href="#cb25-609" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd8tminor</span>
<span id="cb25-610"><a href="#cb25-610" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-611"><a href="#cb25-611" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors =</span> <span class="fu">list</span>(</span>
<span id="cb25-612"><a href="#cb25-612" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"l1"</span></span>
<span id="cb25-613"><a href="#cb25-613" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"l2"</span></span>
<span id="cb25-614"><a href="#cb25-614" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb25-615"><a href="#cb25-615" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb25-616"><a href="#cb25-616" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-617"><a href="#cb25-617" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors_category =</span> <span class="fu">list</span>(</span>
<span id="cb25-618"><a href="#cb25-618" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"immune"</span></span>
<span id="cb25-619"><a href="#cb25-619" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"tcell"</span></span>
<span id="cb25-620"><a href="#cb25-620" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"cd4t"</span></span>
<span id="cb25-621"><a href="#cb25-621" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"cd8t"</span></span>
<span id="cb25-622"><a href="#cb25-622" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-623"><a href="#cb25-623" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-624"><a href="#cb25-624" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-625"><a href="#cb25-625" aria-hidden="true" tabindex="-1"></a>multictobj <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(</span>
<span id="cb25-626"><a href="#cb25-626" aria-hidden="true" tabindex="-1"></a>  <span class="at">pipeline =</span> pipeline_io_multiomics</span>
<span id="cb25-627"><a href="#cb25-627" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">totalcounts=</span> tc[<span class="fu">rownames</span>(multimat)]</span>
<span id="cb25-628"><a href="#cb25-628" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">gene_wise_frequency =</span> genefreq</span>
<span id="cb25-629"><a href="#cb25-629" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">counts_matrix =</span> multimat</span>
<span id="cb25-630"><a href="#cb25-630" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">adjacency_matrix =</span> multi_simgrph[<span class="fu">rownames</span>(multimat), <span class="fu">rownames</span>(multimat)]</span>
<span id="cb25-631"><a href="#cb25-631" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-632"><a href="#cb25-632" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-633"><a href="#cb25-633" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-636"><a href="#cb25-636" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-637"><a href="#cb25-637" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-638"><a href="#cb25-638" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-639"><a href="#cb25-639" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-640"><a href="#cb25-640" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(multictobj, <span class="st">"data2/multictobj_multigrph.rds"</span>)</span>
<span id="cb25-641"><a href="#cb25-641" aria-hidden="true" tabindex="-1"></a>multictobj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data2/multictobj_multigrph.rds"</span>)</span>
<span id="cb25-642"><a href="#cb25-642" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-643"><a href="#cb25-643" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-644"><a href="#cb25-644" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-645"><a href="#cb25-645" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-646"><a href="#cb25-646" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-647"><a href="#cb25-647" aria-hidden="true" tabindex="-1"></a><span class="fu">##### RNA-only cell typing  {#rnaonly}</span></span>
<span id="cb25-648"><a href="#cb25-648" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-651"><a href="#cb25-651" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-652"><a href="#cb25-652" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-653"><a href="#cb25-653" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-654"><a href="#cb25-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-655"><a href="#cb25-655" aria-hidden="true" tabindex="-1"></a><span class="do">### RNA-only PCA and similarity graph</span></span>
<span id="cb25-656"><a href="#cb25-656" aria-hidden="true" tabindex="-1"></a>pcaobj_rna <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">sparse_quasipoisson_pca_seurat</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[use_genes,]</span>
<span id="cb25-657"><a href="#cb25-657" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">totalcounts =</span> tc</span>
<span id="cb25-658"><a href="#cb25-658" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">grate =</span> genefreq[use_genes]</span>
<span id="cb25-659"><a href="#cb25-659" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">do.scale =</span> <span class="cn">TRUE</span></span>
<span id="cb25-660"><a href="#cb25-660" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">do.center =</span> <span class="cn">TRUE</span></span>
<span id="cb25-661"><a href="#cb25-661" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">scale.max =</span> <span class="dv">10</span></span>
<span id="cb25-662"><a href="#cb25-662" aria-hidden="true" tabindex="-1"></a>                                                       ,<span class="at">ncores =</span> <span class="dv">1</span></span>
<span id="cb25-663"><a href="#cb25-663" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-664"><a href="#cb25-664" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-665"><a href="#cb25-665" aria-hidden="true" tabindex="-1"></a><span class="do">### Make a cell-cell adjacency matrix for HieraType </span></span>
<span id="cb25-666"><a href="#cb25-666" aria-hidden="true" tabindex="-1"></a>rna_simgrph <span class="ot">&lt;-</span> uwot<span class="sc">::</span><span class="fu">similarity_graph</span>(pcaobj_rna<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings</span>
<span id="cb25-667"><a href="#cb25-667" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">n_neighbors =</span> <span class="dv">30</span></span>
<span id="cb25-668"><a href="#cb25-668" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">nn_method =</span> <span class="st">"annoy"</span></span>
<span id="cb25-669"><a href="#cb25-669" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">metric =</span> <span class="st">'cosine'</span></span>
<span id="cb25-670"><a href="#cb25-670" aria-hidden="true" tabindex="-1"></a>                                  )</span>
<span id="cb25-671"><a href="#cb25-671" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(rna_simgrph) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(rna_simgrph) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(pcaobj_rna<span class="sc">$</span>reduction.data<span class="sc">@</span>cell.embeddings)</span>
<span id="cb25-672"><a href="#cb25-672" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-673"><a href="#cb25-673" aria-hidden="true" tabindex="-1"></a><span class="do">## RNA-only hieratype</span></span>
<span id="cb25-674"><a href="#cb25-674" aria-hidden="true" tabindex="-1"></a>rnactobj <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(</span>
<span id="cb25-675"><a href="#cb25-675" aria-hidden="true" tabindex="-1"></a>  <span class="at">pipeline =</span> pipeline_io_multiomics <span class="do">## the _protein markers are automatically excluded, </span></span>
<span id="cb25-676"><a href="#cb25-676" aria-hidden="true" tabindex="-1"></a>                                    <span class="do">## because they won't be found in the counts_matrix below</span></span>
<span id="cb25-677"><a href="#cb25-677" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">totalcounts=</span> tc[<span class="fu">rownames</span>(multimat)]</span>
<span id="cb25-678"><a href="#cb25-678" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">gene_wise_frequency =</span> genefreq[use_genes]</span>
<span id="cb25-679"><a href="#cb25-679" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">counts_matrix =</span> multimat[,use_genes] <span class="do">## exclude proteins, only use the 'use_genes'.  </span></span>
<span id="cb25-680"><a href="#cb25-680" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">adjacency_matrix =</span> rna_simgrph[<span class="fu">rownames</span>(multimat), <span class="fu">rownames</span>(multimat)]</span>
<span id="cb25-681"><a href="#cb25-681" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-682"><a href="#cb25-682" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-683"><a href="#cb25-683" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-686"><a href="#cb25-686" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-687"><a href="#cb25-687" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-688"><a href="#cb25-688" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-689"><a href="#cb25-689" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-690"><a href="#cb25-690" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(pcaobj_rna, <span class="st">"data2/pcaobj_rna.rds"</span>)</span>
<span id="cb25-691"><a href="#cb25-691" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(rnactobj, <span class="st">"data2/rnactobj_rnagrph.rds"</span>)</span>
<span id="cb25-692"><a href="#cb25-692" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-693"><a href="#cb25-693" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-694"><a href="#cb25-694" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-695"><a href="#cb25-695" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-696"><a href="#cb25-696" aria-hidden="true" tabindex="-1"></a>pcaobj_rna <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data2/pcaobj_rna.rds"</span>) </span>
<span id="cb25-697"><a href="#cb25-697" aria-hidden="true" tabindex="-1"></a>rnactobj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data2/rnactobj_rnagrph.rds"</span>) </span>
<span id="cb25-698"><a href="#cb25-698" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb25-699"><a href="#cb25-699" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-700"><a href="#cb25-700" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-701"><a href="#cb25-701" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-702"><a href="#cb25-702" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Protein-only cell typing {#proteinonly}</span></span>
<span id="cb25-703"><a href="#cb25-703" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-706"><a href="#cb25-706" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-707"><a href="#cb25-707" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-708"><a href="#cb25-708" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-709"><a href="#cb25-709" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-710"><a href="#cb25-710" aria-hidden="true" tabindex="-1"></a>pgamm_prot <span class="ot">&lt;-</span> pgamm</span>
<span id="cb25-711"><a href="#cb25-711" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pgamm_prot) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_protein$"</span>, <span class="st">""</span>, <span class="fu">rownames</span>(pgamm_prot))</span>
<span id="cb25-712"><a href="#cb25-712" aria-hidden="true" tabindex="-1"></a>sem_protein <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">SetAssayData</span>(sem_protein,<span class="at">layer =</span> <span class="st">"data"</span>, <span class="at">new.data =</span> pgamm_prot)</span>
<span id="cb25-713"><a href="#cb25-713" aria-hidden="true" tabindex="-1"></a>sem_protein <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">ScaleData</span>(sem_protein)</span>
<span id="cb25-714"><a href="#cb25-714" aria-hidden="true" tabindex="-1"></a>sem_protein <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">RunPCA</span>(sem_protein, <span class="at">features =</span> <span class="fu">rownames</span>(sem_protein))</span>
<span id="cb25-715"><a href="#cb25-715" aria-hidden="true" tabindex="-1"></a><span class="do">### Make a cell-cell adjacency matrix for HieraType </span></span>
<span id="cb25-716"><a href="#cb25-716" aria-hidden="true" tabindex="-1"></a>protein_simgrph <span class="ot">&lt;-</span> uwot<span class="sc">::</span><span class="fu">similarity_graph</span>(sem_protein<span class="sc">@</span>reductions<span class="sc">$</span>pca<span class="sc">@</span>cell.embeddings</span>
<span id="cb25-717"><a href="#cb25-717" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">n_neighbors =</span> <span class="dv">30</span></span>
<span id="cb25-718"><a href="#cb25-718" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">nn_method =</span> <span class="st">"annoy"</span></span>
<span id="cb25-719"><a href="#cb25-719" aria-hidden="true" tabindex="-1"></a>                                  ,<span class="at">metric =</span> <span class="st">'cosine'</span></span>
<span id="cb25-720"><a href="#cb25-720" aria-hidden="true" tabindex="-1"></a>                                  )</span>
<span id="cb25-721"><a href="#cb25-721" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(protein_simgrph) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(protein_simgrph) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sem_protein<span class="sc">@</span>reductions<span class="sc">$</span>pca<span class="sc">@</span>cell.embeddings)</span>
<span id="cb25-722"><a href="#cb25-722" aria-hidden="true" tabindex="-1"></a><span class="do">#### Make a copy of the cell typing pipeline to use for protein</span></span>
<span id="cb25-723"><a href="#cb25-723" aria-hidden="true" tabindex="-1"></a><span class="do">#### -- Remove 'mast' cells, which dont have a specific protein marker.</span></span>
<span id="cb25-724"><a href="#cb25-724" aria-hidden="true" tabindex="-1"></a><span class="do">#### -- remove the minor T8 and T4 subtypes</span></span>
<span id="cb25-725"><a href="#cb25-725" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein <span class="ot">&lt;-</span> pipeline_io_multiomics</span>
<span id="cb25-726"><a href="#cb25-726" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>markerslists<span class="sc">$</span>l2<span class="sc">$</span>mast <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb25-727"><a href="#cb25-727" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>markerslists<span class="sc">$</span>lt4minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb25-728"><a href="#cb25-728" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>markerslists<span class="sc">$</span>lt8minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb25-729"><a href="#cb25-729" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>priors<span class="sc">$</span>lt4minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb25-730"><a href="#cb25-730" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>priors<span class="sc">$</span>lt8minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb25-731"><a href="#cb25-731" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>priors_category<span class="sc">$</span>lt4minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb25-732"><a href="#cb25-732" aria-hidden="true" tabindex="-1"></a>pipeline_io_protein<span class="sc">$</span>priors_category<span class="sc">$</span>lt8minor <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb25-733"><a href="#cb25-733" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-734"><a href="#cb25-734" aria-hidden="true" tabindex="-1"></a>proteinctobj <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(</span>
<span id="cb25-735"><a href="#cb25-735" aria-hidden="true" tabindex="-1"></a>  <span class="at">pipeline =</span> pipeline_io_protein</span>
<span id="cb25-736"><a href="#cb25-736" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">counts_matrix =</span> multimat[,<span class="fu">grep</span>(<span class="st">"_protein$"</span>,<span class="fu">colnames</span>(multimat),<span class="at">value=</span><span class="cn">TRUE</span>)]</span>
<span id="cb25-737"><a href="#cb25-737" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">adjacency_matrix =</span> protein_simgrph[<span class="fu">rownames</span>(multimat), <span class="fu">rownames</span>(multimat)]</span>
<span id="cb25-738"><a href="#cb25-738" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-739"><a href="#cb25-739" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-740"><a href="#cb25-740" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-741"><a href="#cb25-741" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-742"><a href="#cb25-742" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-745"><a href="#cb25-745" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-746"><a href="#cb25-746" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-747"><a href="#cb25-747" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-748"><a href="#cb25-748" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(proteinctobj, <span class="st">"data2/proteinctobj_proteingrph.rds"</span>)</span>
<span id="cb25-749"><a href="#cb25-749" aria-hidden="true" tabindex="-1"></a>proteinctobj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data2/proteinctobj_proteingrph.rds"</span>) </span>
<span id="cb25-750"><a href="#cb25-750" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-751"><a href="#cb25-751" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-752"><a href="#cb25-752" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-753"><a href="#cb25-753" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Combining metadata</span></span>
<span id="cb25-754"><a href="#cb25-754" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-757"><a href="#cb25-757" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-758"><a href="#cb25-758" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-759"><a href="#cb25-759" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-760"><a href="#cb25-760" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-761"><a href="#cb25-761" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-762"><a href="#cb25-762" aria-hidden="true" tabindex="-1"></a><span class="do">### Combining metadata</span></span>
<span id="cb25-763"><a href="#cb25-763" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-764"><a href="#cb25-764" aria-hidden="true" tabindex="-1"></a><span class="do">## Adding on multiomic celltype annotations</span></span>
<span id="cb25-765"><a href="#cb25-765" aria-hidden="true" tabindex="-1"></a>ppmulti <span class="ot">&lt;-</span> <span class="fu">copy</span>(multictobj<span class="sc">$</span>post_probs<span class="sc">$</span>l1)</span>
<span id="cb25-766"><a href="#cb25-766" aria-hidden="true" tabindex="-1"></a>ppmulti[,celltype_multi_major<span class="sc">:</span><span class="er">=</span>celltype_granular]</span>
<span id="cb25-767"><a href="#cb25-767" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd4"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd4_tcell"</span>]</span>
<span id="cb25-768"><a href="#cb25-768" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd8"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd8_tcell"</span>]</span>
<span id="cb25-769"><a href="#cb25-769" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"mono|macro"</span>,celltype_multi_major),celltype_multi_major<span class="sc">:</span><span class="er">=</span><span class="st">"macrophage_monocyte"</span>]</span>
<span id="cb25-770"><a href="#cb25-770" aria-hidden="true" tabindex="-1"></a>ppmulti[,.N,by<span class="ot">=</span>.(celltype_multi_major, celltype_granular)]</span>
<span id="cb25-771"><a href="#cb25-771" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-772"><a href="#cb25-772" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-773"><a href="#cb25-773" aria-hidden="true" tabindex="-1"></a><span class="do">## Adding on RNA-only based  celltype annotations</span></span>
<span id="cb25-774"><a href="#cb25-774" aria-hidden="true" tabindex="-1"></a>ppmulti <span class="ot">&lt;-</span> <span class="fu">merge</span>(ppmulti, rnactobj<span class="sc">$</span>post_probs<span class="sc">$</span>l1[,.(cell_ID, celltype_granular)]</span>
<span id="cb25-775"><a href="#cb25-775" aria-hidden="true" tabindex="-1"></a>                 ,<span class="at">by=</span><span class="st">"cell_ID"</span>, <span class="at">suffixes=</span><span class="fu">c</span>(<span class="st">"_multi"</span>, <span class="st">"_rna"</span>))</span>
<span id="cb25-776"><a href="#cb25-776" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-777"><a href="#cb25-777" aria-hidden="true" tabindex="-1"></a>ppmulti[,celltype_rna_major<span class="sc">:</span><span class="er">=</span>celltype_granular_rna]</span>
<span id="cb25-778"><a href="#cb25-778" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd4"</span>,celltype_rna_major),celltype_rna_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd4_tcell"</span>]</span>
<span id="cb25-779"><a href="#cb25-779" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd8"</span>,celltype_rna_major),celltype_rna_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd8_tcell"</span>]</span>
<span id="cb25-780"><a href="#cb25-780" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"mono|macro"</span>,celltype_rna_major),celltype_rna_major<span class="sc">:</span><span class="er">=</span><span class="st">"macrophage_monocyte"</span>]</span>
<span id="cb25-781"><a href="#cb25-781" aria-hidden="true" tabindex="-1"></a>ppmulti[,.N,by<span class="ot">=</span>.(celltype_rna_major, celltype_granular_rna)]</span>
<span id="cb25-782"><a href="#cb25-782" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-783"><a href="#cb25-783" aria-hidden="true" tabindex="-1"></a><span class="do">## Adding protein-only based celltype annotations</span></span>
<span id="cb25-784"><a href="#cb25-784" aria-hidden="true" tabindex="-1"></a>ppmulti <span class="ot">&lt;-</span> <span class="fu">merge</span>(ppmulti, proteinctobj<span class="sc">$</span>post_probs<span class="sc">$</span>l1[,.(cell_ID, celltype_granular)]</span>
<span id="cb25-785"><a href="#cb25-785" aria-hidden="true" tabindex="-1"></a>                 ,<span class="at">by=</span><span class="st">"cell_ID"</span>, <span class="at">suffixes=</span><span class="fu">c</span>(<span class="st">"_multi"</span>, <span class="st">"_protein"</span>))</span>
<span id="cb25-786"><a href="#cb25-786" aria-hidden="true" tabindex="-1"></a><span class="fu">setnames</span>(ppmulti, <span class="fu">c</span>(<span class="st">"celltype_granular"</span>), <span class="fu">c</span>(<span class="st">"celltype_granular_protein"</span>))</span>
<span id="cb25-787"><a href="#cb25-787" aria-hidden="true" tabindex="-1"></a>ppmulti[,celltype_protein_major<span class="sc">:</span><span class="er">=</span>celltype_granular_protein]</span>
<span id="cb25-788"><a href="#cb25-788" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd4"</span>,celltype_protein_major),celltype_protein_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd4_tcell"</span>]</span>
<span id="cb25-789"><a href="#cb25-789" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"cd8"</span>,celltype_protein_major),celltype_protein_major<span class="sc">:</span><span class="er">=</span><span class="st">"cd8_tcell"</span>]</span>
<span id="cb25-790"><a href="#cb25-790" aria-hidden="true" tabindex="-1"></a>ppmulti[<span class="fu">grepl</span>(<span class="st">"mono|macro"</span>,celltype_protein_major),celltype_protein_major<span class="sc">:</span><span class="er">=</span><span class="st">"macrophage_monocyte"</span>]</span>
<span id="cb25-791"><a href="#cb25-791" aria-hidden="true" tabindex="-1"></a>ppmulti[,.N,by<span class="ot">=</span>.(celltype_protein_major, celltype_granular_protein)]</span>
<span id="cb25-792"><a href="#cb25-792" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb25-793"><a href="#cb25-793" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-794"><a href="#cb25-794" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-795"><a href="#cb25-795" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-796"><a href="#cb25-796" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Integrating unsupervised clusters with immune cell type labels {#integrateunsup}</span></span>
<span id="cb25-799"><a href="#cb25-799" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-800"><a href="#cb25-800" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-801"><a href="#cb25-801" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-802"><a href="#cb25-802" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-803"><a href="#cb25-803" aria-hidden="true" tabindex="-1"></a><span class="do">### Annotating non-immune clusters with unsupervised labels, creating more granular annotations</span></span>
<span id="cb25-804"><a href="#cb25-804" aria-hidden="true" tabindex="-1"></a><span class="do">### </span></span>
<span id="cb25-805"><a href="#cb25-805" aria-hidden="true" tabindex="-1"></a>ppmulti <span class="ot">&lt;-</span> <span class="fu">merge</span>(ppmulti, <span class="fu">data.table</span>(sem_rna<span class="sc">@</span>meta.data)[,.(cell_ID, x_slide_mm, y_slide_mm, clusters_multi_unsup)], <span class="at">by =</span> <span class="st">"cell_ID"</span>)</span>
<span id="cb25-806"><a href="#cb25-806" aria-hidden="true" tabindex="-1"></a>ppmulti[,celltype_multi_semi<span class="sc">:</span><span class="er">=</span>celltype_multi_major]</span>
<span id="cb25-807"><a href="#cb25-807" aria-hidden="true" tabindex="-1"></a>ppmulti[celltype_multi_semi <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"plasma"</span>, <span class="st">"endothelial"</span>, <span class="st">"epithelial"</span>, <span class="st">"fibroblast"</span>, <span class="st">"smooth_muscle"</span>),celltype_multi_semi<span class="sc">:</span><span class="er">=</span>clusters_multi_unsup]</span>
<span id="cb25-808"><a href="#cb25-808" aria-hidden="true" tabindex="-1"></a>ppmulti[,.N,by<span class="ot">=</span>.(clusters_multi_unsup, celltype_multi_major)]</span>
<span id="cb25-809"><a href="#cb25-809" aria-hidden="true" tabindex="-1"></a>class_other <span class="ot">&lt;-</span> ppmulti[,.N,by<span class="ot">=</span>.(clusters_multi_unsup, celltype_multi_major)][,prop<span class="sc">:</span><span class="er">=</span>N<span class="sc">/</span><span class="fu">sum</span>(N),by<span class="ot">=</span>.(clusters_multi_unsup)][<span class="fu">order</span>(<span class="sc">-</span>prop)][</span>
<span id="cb25-810"><a href="#cb25-810" aria-hidden="true" tabindex="-1"></a>  ,<span class="fu">head</span>(.SD,<span class="dv">1</span>),by<span class="ot">=</span>.(clusters_multi_unsup)]</span>
<span id="cb25-811"><a href="#cb25-811" aria-hidden="true" tabindex="-1"></a>others <span class="ot">&lt;-</span> class_other[<span class="sc">!</span>celltype_multi_major <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"plasma"</span>, <span class="st">"endothelial"</span>, <span class="st">"epithelial"</span>, <span class="st">"fibroblast"</span>, <span class="st">"smooth_muscle"</span>)][[<span class="st">"clusters_multi_unsup"</span>]]</span>
<span id="cb25-812"><a href="#cb25-812" aria-hidden="true" tabindex="-1"></a>ppmulti[celltype_multi_semi <span class="sc">%in%</span> others, celltype_multi_semi<span class="sc">:</span><span class="er">=</span><span class="st">"other"</span>]</span>
<span id="cb25-813"><a href="#cb25-813" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-814"><a href="#cb25-814" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb25-815"><a href="#cb25-815" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-816"><a href="#cb25-816" aria-hidden="true" tabindex="-1"></a>xyp_multi_semisup <span class="ot">&lt;-</span> <span class="fu">xyplot</span>(<span class="st">"celltype_multi_semi"</span>, <span class="at">metadata =</span> ppmulti, <span class="at">ptsize =</span> <span class="fl">0.05</span>) <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb25-817"><a href="#cb25-817" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-818"><a href="#cb25-818" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"figures/xyp_multi_semisup.png"</span></span>
<span id="cb25-819"><a href="#cb25-819" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">plot=</span>xyp_multi_semisup</span>
<span id="cb25-820"><a href="#cb25-820" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">device =</span> <span class="st">"png"</span></span>
<span id="cb25-821"><a href="#cb25-821" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">width =</span> <span class="dv">1181</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>)</span>
<span id="cb25-822"><a href="#cb25-822" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">height=</span><span class="dv">689</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>)</span>
<span id="cb25-823"><a href="#cb25-823" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">units =</span> <span class="st">'px'</span></span>
<span id="cb25-824"><a href="#cb25-824" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-825"><a href="#cb25-825" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-826"><a href="#cb25-826" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-827"><a href="#cb25-827" aria-hidden="true" tabindex="-1"></a><span class="fu">##### ARI quantifying agreement between cell type labels {#ari}</span></span>
<span id="cb25-828"><a href="#cb25-828" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-831"><a href="#cb25-831" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-832"><a href="#cb25-832" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-833"><a href="#cb25-833" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-834"><a href="#cb25-834" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-835"><a href="#cb25-835" aria-hidden="true" tabindex="-1"></a><span class="do">## ARI</span></span>
<span id="cb25-836"><a href="#cb25-836" aria-hidden="true" tabindex="-1"></a>mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(ppmulti<span class="sc">$</span>celltype_multi_major, ppmulti<span class="sc">$</span>celltype_protein_major)</span>
<span id="cb25-837"><a href="#cb25-837" aria-hidden="true" tabindex="-1"></a>mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(ppmulti<span class="sc">$</span>celltype_multi_major, ppmulti<span class="sc">$</span>celltype_rna_major)</span>
<span id="cb25-838"><a href="#cb25-838" aria-hidden="true" tabindex="-1"></a>mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(ppmulti<span class="sc">$</span>celltype_protein_major, ppmulti<span class="sc">$</span>celltype_rna_major)</span>
<span id="cb25-839"><a href="#cb25-839" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-840"><a href="#cb25-840" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-841"><a href="#cb25-841" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-842"><a href="#cb25-842" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Making ROC curves, computing and plotting area under the curve</span></span>
<span id="cb25-843"><a href="#cb25-843" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-846"><a href="#cb25-846" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-847"><a href="#cb25-847" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-848"><a href="#cb25-848" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-849"><a href="#cb25-849" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-850"><a href="#cb25-850" aria-hidden="true" tabindex="-1"></a><span class="do">### Get the most important 'index markers' for each celltype (including proteins and RNAs)</span></span>
<span id="cb25-851"><a href="#cb25-851" aria-hidden="true" tabindex="-1"></a>index_markers <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(HieraType<span class="sc">::</span>markerslist_multiomic_immune, <span class="st">"[["</span>, <span class="st">"index_marker"</span>))</span>
<span id="cb25-852"><a href="#cb25-852" aria-hidden="true" tabindex="-1"></a>index_markers <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(<span class="fu">lapply</span>(<span class="fu">names</span>(HieraType<span class="sc">::</span>markerslist_multiomic_immune), <span class="cf">function</span>(ct){</span>
<span id="cb25-853"><a href="#cb25-853" aria-hidden="true" tabindex="-1"></a>  markerdt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">celltype =</span> ct</span>
<span id="cb25-854"><a href="#cb25-854" aria-hidden="true" tabindex="-1"></a>                         ,<span class="at">marker =</span> HieraType<span class="sc">::</span>markerslist_multiomic_immune[[ct]][[<span class="st">"index_marker"</span>]]</span>
<span id="cb25-855"><a href="#cb25-855" aria-hidden="true" tabindex="-1"></a>                         ,<span class="at">typ =</span> <span class="st">"rna"</span></span>
<span id="cb25-856"><a href="#cb25-856" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-857"><a href="#cb25-857" aria-hidden="true" tabindex="-1"></a>  markerdt[<span class="fu">grepl</span>(<span class="st">"_protein$"</span>, marker),typ<span class="sc">:</span><span class="er">=</span><span class="st">"protein"</span>]</span>
<span id="cb25-858"><a href="#cb25-858" aria-hidden="true" tabindex="-1"></a>  markerdt[,marker<span class="sc">:</span><span class="er">=</span><span class="fu">gsub</span>(<span class="st">"_protein$"</span>,<span class="st">""</span>, marker)]</span>
<span id="cb25-859"><a href="#cb25-859" aria-hidden="true" tabindex="-1"></a>  markerdt <span class="ot">&lt;-</span> markerdt[(typ<span class="sc">==</span><span class="st">"rna"</span> <span class="sc">&amp;</span> marker <span class="sc">%in%</span> <span class="fu">rownames</span>(sem_rna)) <span class="sc">|</span> (typ<span class="sc">==</span><span class="st">"protein"</span> <span class="sc">&amp;</span> marker <span class="sc">%in%</span> <span class="fu">rownames</span>(sem_protein))]</span>
<span id="cb25-860"><a href="#cb25-860" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(markerdt)</span>
<span id="cb25-861"><a href="#cb25-861" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb25-862"><a href="#cb25-862" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-863"><a href="#cb25-863" aria-hidden="true" tabindex="-1"></a><span class="do">### Combine macrophage and monocyte due to similarity</span></span>
<span id="cb25-864"><a href="#cb25-864" aria-hidden="true" tabindex="-1"></a>index_markers[celltype <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"macrophage"</span>, <span class="st">"monocyte"</span>),celltype<span class="sc">:</span><span class="er">=</span><span class="st">"macrophage_monocyte"</span>]</span>
<span id="cb25-865"><a href="#cb25-865" aria-hidden="true" tabindex="-1"></a>aucrocl <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb25-866"><a href="#cb25-866" aria-hidden="true" tabindex="-1"></a>plotdata_markerl <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb25-867"><a href="#cb25-867" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pgamm) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_protein"</span>, <span class="st">""</span>, <span class="fu">rownames</span>(pgamm))</span>
<span id="cb25-868"><a href="#cb25-868" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(ii <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(index_markers)){</span>
<span id="cb25-869"><a href="#cb25-869" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(ii)</span>
<span id="cb25-870"><a href="#cb25-870" aria-hidden="true" tabindex="-1"></a>  res <span class="ot">&lt;-</span> index_markers[ii]</span>
<span id="cb25-871"><a href="#cb25-871" aria-hidden="true" tabindex="-1"></a>  typ <span class="ot">&lt;-</span> res[[<span class="st">"typ"</span>]]</span>
<span id="cb25-872"><a href="#cb25-872" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-873"><a href="#cb25-873" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-874"><a href="#cb25-874" aria-hidden="true" tabindex="-1"></a>  marker <span class="ot">&lt;-</span> res[[<span class="st">"marker"</span>]]</span>
<span id="cb25-875"><a href="#cb25-875" aria-hidden="true" tabindex="-1"></a>  celltype <span class="ot">&lt;-</span> res[[<span class="st">"celltype"</span>]] </span>
<span id="cb25-876"><a href="#cb25-876" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(typ<span class="sc">==</span><span class="st">"rna"</span>){</span>
<span id="cb25-877"><a href="#cb25-877" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[marker,ppmulti[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb25-878"><a href="#cb25-878" aria-hidden="true" tabindex="-1"></a>    tc <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(sem_rna[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[,ppmulti[[<span class="st">"cell_ID"</span>]]])</span>
<span id="cb25-879"><a href="#cb25-879" aria-hidden="true" tabindex="-1"></a>    muhat <span class="ot">&lt;-</span> genefreq[marker] <span class="sc">*</span> tc </span>
<span id="cb25-880"><a href="#cb25-880" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> (x <span class="sc">-</span> muhat) <span class="sc">/</span> <span class="fu">sqrt</span>(muhat) <span class="do">## normalized expression for marker (pearson residuals)</span></span>
<span id="cb25-881"><a href="#cb25-881" aria-hidden="true" tabindex="-1"></a>  }  <span class="cf">else</span> {</span>
<span id="cb25-882"><a href="#cb25-882" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> pgamm[marker,ppmulti[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb25-883"><a href="#cb25-883" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb25-884"><a href="#cb25-884" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb25-885"><a href="#cb25-885" aria-hidden="true" tabindex="-1"></a>  pdlbl <span class="ot">&lt;-</span> <span class="fu">list</span>() </span>
<span id="cb25-886"><a href="#cb25-886" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(labelname <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"celltype_multi_major"</span>, <span class="st">"celltype_rna_major"</span>, <span class="st">"celltype_protein_major"</span>)){</span>
<span id="cb25-887"><a href="#cb25-887" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(labelname)</span>
<span id="cb25-888"><a href="#cb25-888" aria-hidden="true" tabindex="-1"></a>    labels <span class="ot">&lt;-</span> ppmulti[[labelname]]</span>
<span id="cb25-889"><a href="#cb25-889" aria-hidden="true" tabindex="-1"></a>    labels[<span class="fu">grep</span>(<span class="st">"tcell"</span>, labels)] <span class="ot">&lt;-</span> <span class="st">"tcell"</span> </span>
<span id="cb25-890"><a href="#cb25-890" aria-hidden="true" tabindex="-1"></a>    y <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(labels<span class="sc">==</span>celltype)</span>
<span id="cb25-891"><a href="#cb25-891" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">sum</span>(y) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb25-892"><a href="#cb25-892" aria-hidden="true" tabindex="-1"></a>      roc_obj <span class="ot">&lt;-</span> <span class="fu">roc</span>(y, x, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-893"><a href="#cb25-893" aria-hidden="true" tabindex="-1"></a>      res[,(labelname)<span class="sc">:</span><span class="er">=</span><span class="fu">as.numeric</span>(<span class="fu">auc</span>(roc_obj))]</span>
<span id="cb25-894"><a href="#cb25-894" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb25-895"><a href="#cb25-895" aria-hidden="true" tabindex="-1"></a>    pdlbl[[labelname]] <span class="ot">&lt;-</span> </span>
<span id="cb25-896"><a href="#cb25-896" aria-hidden="true" tabindex="-1"></a>      <span class="fu">data.table</span>(<span class="at">spec=</span>roc_obj<span class="sc">$</span>specificities, <span class="at">sens=</span>roc_obj<span class="sc">$</span>sensitivities,<span class="at">marker=</span>marker)[,label<span class="sc">:</span><span class="er">=</span>labelname][<span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">1</span>,.N,<span class="dv">10</span>), .N))]</span>
<span id="cb25-897"><a href="#cb25-897" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb25-898"><a href="#cb25-898" aria-hidden="true" tabindex="-1"></a>  plotdata_markerl[[ii]] <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(pdlbl)[,<span class="st">`</span><span class="at">:=</span><span class="st">`</span>(<span class="at">celltype =</span> celltype, <span class="at">marker=</span>marker, <span class="at">typ =</span> typ)]</span>
<span id="cb25-899"><a href="#cb25-899" aria-hidden="true" tabindex="-1"></a>  <span class="co">#ggplot(rbindlist(pdlbl)[seq(1,.N,by=10)],aes(x = (1-spec), y = sens, color=label)) + geom_line()</span></span>
<span id="cb25-900"><a href="#cb25-900" aria-hidden="true" tabindex="-1"></a>  aucrocl[[ii]] <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">copy</span>(res) <span class="co">#return(res)</span></span>
<span id="cb25-901"><a href="#cb25-901" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb25-902"><a href="#cb25-902" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb25-903"><a href="#cb25-903" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-906"><a href="#cb25-906" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-907"><a href="#cb25-907" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-908"><a href="#cb25-908" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-909"><a href="#cb25-909" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-910"><a href="#cb25-910" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-911"><a href="#cb25-911" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(aucrocl, <span class="st">"data2/aucrocl.rds"</span>)</span>
<span id="cb25-912"><a href="#cb25-912" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(plotdata_markerl, <span class="st">"data2/plotdata_markerl.rds"</span>)</span>
<span id="cb25-913"><a href="#cb25-913" aria-hidden="true" tabindex="-1"></a>  <span class="co">#pdlbl</span></span>
<span id="cb25-914"><a href="#cb25-914" aria-hidden="true" tabindex="-1"></a>  aucrocl <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data2/aucrocl.rds"</span>)</span>
<span id="cb25-915"><a href="#cb25-915" aria-hidden="true" tabindex="-1"></a>  plotdata_markerl <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data2/plotdata_markerl.rds"</span>)</span>
<span id="cb25-916"><a href="#cb25-916" aria-hidden="true" tabindex="-1"></a> <span class="in">```</span></span>
<span id="cb25-917"><a href="#cb25-917" aria-hidden="true" tabindex="-1"></a> <span class="in">```{r}</span></span>
<span id="cb25-918"><a href="#cb25-918" aria-hidden="true" tabindex="-1"></a><span class="in">#| eval: false</span></span>
<span id="cb25-919"><a href="#cb25-919" aria-hidden="true" tabindex="-1"></a><span class="in">#| echo: true</span></span>
<span id="cb25-920"><a href="#cb25-920" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-921"><a href="#cb25-921" aria-hidden="true" tabindex="-1"></a><span class="in">  aucroc &lt;- rbindlist(aucrocl, use.names=TRUE,fill=TRUE)</span></span>
<span id="cb25-922"><a href="#cb25-922" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb25-923"><a href="#cb25-923" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb25-924"><a href="#cb25-924" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-925"><a href="#cb25-925" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-926"><a href="#cb25-926" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Average AUC across markers and celltypes</span></span>
<span id="cb25-929"><a href="#cb25-929" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-930"><a href="#cb25-930" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-931"><a href="#cb25-931" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-932"><a href="#cb25-932" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-933"><a href="#cb25-933" aria-hidden="true" tabindex="-1"></a>aucroc[celltype<span class="sc">!=</span><span class="st">"mast"</span>,<span class="fu">lapply</span>(.SD, mean),.SDcols<span class="ot">=</span><span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>]</span>
<span id="cb25-934"><a href="#cb25-934" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-935"><a href="#cb25-935" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-936"><a href="#cb25-936" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-937"><a href="#cb25-937" aria-hidden="true" tabindex="-1"></a><span class="fu">#### T cell ROC curves</span></span>
<span id="cb25-940"><a href="#cb25-940" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-941"><a href="#cb25-941" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-942"><a href="#cb25-942" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb25-943"><a href="#cb25-943" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-944"><a href="#cb25-944" aria-hidden="true" tabindex="-1"></a>  shw <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(plotdata_markerl)[marker <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"CD3"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD2"</span>)]</span>
<span id="cb25-945"><a href="#cb25-945" aria-hidden="true" tabindex="-1"></a>  shw[,method<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(label, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"celltype_multi_major"</span>, <span class="st">"celltype_rna_major"</span>, <span class="st">"celltype_protein_major"</span>), <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"multiomic typing"</span>, <span class="st">"rna-only cell typing"</span>, <span class="st">"protein-only cell typing"</span>))]</span>
<span id="cb25-946"><a href="#cb25-946" aria-hidden="true" tabindex="-1"></a>  <span class="fu">merge</span>(shw, aucrocl)</span>
<span id="cb25-947"><a href="#cb25-947" aria-hidden="true" tabindex="-1"></a>  data.table<span class="sc">::</span><span class="fu">setnames</span>(shw, <span class="fu">c</span>(<span class="st">"typ"</span>), <span class="fu">c</span>(<span class="st">"analyte"</span>))</span>
<span id="cb25-948"><a href="#cb25-948" aria-hidden="true" tabindex="-1"></a>  shw[,marker<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(marker, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"CD3"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD2"</span>))]</span>
<span id="cb25-949"><a href="#cb25-949" aria-hidden="true" tabindex="-1"></a>  shw <span class="ot">&lt;-</span> shw[<span class="fu">order</span>(marker)]</span>
<span id="cb25-950"><a href="#cb25-950" aria-hidden="true" tabindex="-1"></a>  aucroc[marker <span class="sc">%in%</span> shw<span class="sc">$</span>marker]</span>
<span id="cb25-951"><a href="#cb25-951" aria-hidden="true" tabindex="-1"></a>  rocplot_tcell <span class="ot">&lt;-</span> </span>
<span id="cb25-952"><a href="#cb25-952" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(shw, <span class="fu">aes</span>(<span class="at">x =</span> (<span class="dv">1</span><span class="sc">-</span>spec), <span class="at">y =</span> sens,<span class="at">group =</span> method,<span class="at">color=</span>method)) <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span> </span>
<span id="cb25-953"><a href="#cb25-953" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>marker <span class="sc">+</span> analyte, <span class="at">labeller=</span>label_both) <span class="sc">+</span> </span>
<span id="cb25-954"><a href="#cb25-954" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb25-955"><a href="#cb25-955" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,.<span class="dv">2</span>), <span class="at">labels =</span> <span class="dv">1</span><span class="sc">-</span><span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,.<span class="dv">2</span>), <span class="at">name =</span> <span class="st">"specificity"</span>) <span class="sc">+</span> </span>
<span id="cb25-956"><a href="#cb25-956" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,.<span class="dv">2</span>), <span class="at">name =</span> <span class="st">"sensitivity"</span>) <span class="sc">+</span> </span>
<span id="cb25-957"><a href="#cb25-957" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">brewer.pal</span>(<span class="dv">8</span>,<span class="st">"Set1"</span>)[<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>,<span class="dv">3</span>)]</span>
<span id="cb25-958"><a href="#cb25-958" aria-hidden="true" tabindex="-1"></a>    <span class="co">#scale_color_manual(values = brewer.pal(8,"Dark2")[c(1:3)]</span></span>
<span id="cb25-959"><a href="#cb25-959" aria-hidden="true" tabindex="-1"></a>                       , <span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">lwd=</span><span class="dv">2</span>)))  <span class="sc">+</span> </span>
<span id="cb25-960"><a href="#cb25-960" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">16</span>)) <span class="sc">+</span> </span>
<span id="cb25-961"><a href="#cb25-961" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"bottom"</span>) <span class="sc">+</span>  </span>
<span id="cb25-962"><a href="#cb25-962" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"ROC curve with T cell RNA and protein markers:</span><span class="sc">\n</span><span class="st">Multimomic integrated celltyping outperforms</span><span class="sc">\n</span><span class="st">either RNA or protein alone when considering cell type markers measured on both analytes"</span>)</span>
<span id="cb25-963"><a href="#cb25-963" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb25-964"><a href="#cb25-964" aria-hidden="true" tabindex="-1"></a>  lbld <span class="ot">&lt;-</span> </span>
<span id="cb25-965"><a href="#cb25-965" aria-hidden="true" tabindex="-1"></a>    aucroc[marker <span class="sc">%in%</span> shw<span class="sc">$</span>marker]</span>
<span id="cb25-966"><a href="#cb25-966" aria-hidden="true" tabindex="-1"></a>  lbld <span class="ot">&lt;-</span> <span class="fu">melt</span>(lbld, <span class="at">id.vars=</span><span class="fu">c</span>(<span class="st">"celltype"</span>, <span class="st">"marker"</span>, <span class="st">"typ"</span>),<span class="at">value.name=</span><span class="st">"auc"</span>, <span class="at">variable.name=</span><span class="st">"label"</span>)</span>
<span id="cb25-967"><a href="#cb25-967" aria-hidden="true" tabindex="-1"></a>  lbld[,method<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(label</span>
<span id="cb25-968"><a href="#cb25-968" aria-hidden="true" tabindex="-1"></a>                       ,<span class="at">levels=</span><span class="fu">c</span>(<span class="st">"celltype_multi_major"</span>, <span class="st">"celltype_rna_major"</span>, <span class="st">"celltype_protein_major"</span>)</span>
<span id="cb25-969"><a href="#cb25-969" aria-hidden="true" tabindex="-1"></a>                       ,<span class="at">labels=</span><span class="fu">c</span>(<span class="st">"multiomic typing"</span>, <span class="st">"rna-only cell typing"</span>, <span class="st">"protein-only cell typing"</span>))]</span>
<span id="cb25-970"><a href="#cb25-970" aria-hidden="true" tabindex="-1"></a>  lbld[,marker<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(marker, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"CD3"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD2"</span>))]</span>
<span id="cb25-971"><a href="#cb25-971" aria-hidden="true" tabindex="-1"></a>  lbld[,analyte<span class="sc">:</span><span class="er">=</span>typ]</span>
<span id="cb25-972"><a href="#cb25-972" aria-hidden="true" tabindex="-1"></a>  lbld[,lbl<span class="sc">:</span><span class="er">=</span><span class="fu">paste0</span>(method, <span class="st">" auc="</span>, <span class="fu">formatC</span>(auc, <span class="at">digits=</span><span class="dv">3</span>))]</span>
<span id="cb25-973"><a href="#cb25-973" aria-hidden="true" tabindex="-1"></a>  lbld[,spec<span class="sc">:</span><span class="er">=</span><span class="fl">0.7</span>]</span>
<span id="cb25-974"><a href="#cb25-974" aria-hidden="true" tabindex="-1"></a>  lbld[,sens<span class="sc">:</span><span class="er">=</span><span class="fl">0.1</span><span class="sc">-</span><span class="fu">as.numeric</span>(lbld<span class="sc">$</span>method)<span class="sc">*</span><span class="fl">0.05</span>]</span>
<span id="cb25-975"><a href="#cb25-975" aria-hidden="true" tabindex="-1"></a>  rocplot_tcell_lbl <span class="ot">&lt;-</span> rocplot_tcell <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="at">data=</span>lbld, <span class="fu">aes</span>(<span class="at">label =</span> lbl),<span class="at">show.legend=</span><span class="cn">FALSE</span>)</span>
<span id="cb25-976"><a href="#cb25-976" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb25-977"><a href="#cb25-977" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="at">plot=</span>rocplot_tcell_lbl</span>
<span id="cb25-978"><a href="#cb25-978" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">filename =</span> <span class="st">"figures/rocplot_tcell_lbl.png"</span></span>
<span id="cb25-979"><a href="#cb25-979" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">width=</span><span class="dv">1372</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb25-980"><a href="#cb25-980" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">height=</span><span class="dv">742</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb25-981"><a href="#cb25-981" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">units =</span> <span class="st">'px'</span></span>
<span id="cb25-982"><a href="#cb25-982" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-983"><a href="#cb25-983" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-984"><a href="#cb25-984" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-985"><a href="#cb25-985" aria-hidden="true" tabindex="-1"></a><span class="fu">#### AUC barplots</span></span>
<span id="cb25-988"><a href="#cb25-988" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-989"><a href="#cb25-989" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-990"><a href="#cb25-990" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true </span></span>
<span id="cb25-991"><a href="#cb25-991" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-992"><a href="#cb25-992" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-993"><a href="#cb25-993" aria-hidden="true" tabindex="-1"></a>  mroc <span class="ot">&lt;-</span> <span class="fu">melt</span>(aucroc</span>
<span id="cb25-994"><a href="#cb25-994" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">id.vars=</span><span class="fu">c</span>(<span class="st">"celltype"</span>, <span class="st">"marker"</span>, <span class="st">"typ"</span>)</span>
<span id="cb25-995"><a href="#cb25-995" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">value.name=</span><span class="st">"auc"</span></span>
<span id="cb25-996"><a href="#cb25-996" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">variable.name=</span><span class="st">"method"</span></span>
<span id="cb25-997"><a href="#cb25-997" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">variable.factor =</span> <span class="cn">FALSE</span></span>
<span id="cb25-998"><a href="#cb25-998" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-999"><a href="#cb25-999" aria-hidden="true" tabindex="-1"></a>  mroc[,celltype<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(celltype, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"tcell"</span>, <span class="st">"bcell"</span>, <span class="st">"macrophage_monocyte"</span>, <span class="st">"dendritic"</span>, <span class="st">"neutrophil"</span>, <span class="st">"nk"</span>, <span class="st">"mast"</span>))]</span>
<span id="cb25-1000"><a href="#cb25-1000" aria-hidden="true" tabindex="-1"></a>  mroc[,marker_label<span class="sc">:</span><span class="er">=</span><span class="fu">paste0</span>(marker, <span class="st">" ("</span>, typ, <span class="st">")"</span>)]</span>
<span id="cb25-1001"><a href="#cb25-1001" aria-hidden="true" tabindex="-1"></a>  mroc[,marker_label<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(marker_label,<span class="at">levels=</span>mroc[<span class="fu">order</span>(celltype,typ, <span class="sc">-</span>auc),<span class="fu">unique</span>(marker_label)])]</span>
<span id="cb25-1002"><a href="#cb25-1002" aria-hidden="true" tabindex="-1"></a>  aucbar <span class="ot">&lt;-</span> </span>
<span id="cb25-1003"><a href="#cb25-1003" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(mroc[celltype<span class="sc">!=</span><span class="st">"mast"</span>]</span>
<span id="cb25-1004"><a href="#cb25-1004" aria-hidden="true" tabindex="-1"></a>           ,<span class="fu">aes</span>(<span class="at">y =</span> marker_label, <span class="at">x =</span> auc, <span class="at">fill=</span>method)) <span class="sc">+</span> </span>
<span id="cb25-1005"><a href="#cb25-1005" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">'identity'</span>,<span class="at">position=</span><span class="fu">position_dodge</span>()) <span class="sc">+</span> </span>
<span id="cb25-1006"><a href="#cb25-1006" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb25-1007"><a href="#cb25-1007" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">brewer.pal</span>(<span class="dv">3</span>,<span class="st">"Set1"</span>)<span class="co">#unname(pals::alphabet())</span></span>
<span id="cb25-1008"><a href="#cb25-1008" aria-hidden="true" tabindex="-1"></a>                      , <span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">reverse=</span><span class="cn">TRUE</span>)) <span class="sc">+</span> </span>
<span id="cb25-1009"><a href="#cb25-1009" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>celltype, <span class="at">labeller=</span>label_both,<span class="at">scales=</span><span class="st">"free"</span>) <span class="sc">+</span> </span>
<span id="cb25-1010"><a href="#cb25-1010" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Agreement between celltype label and marker gene/protein as measured by area under the ROC curve"</span></span>
<span id="cb25-1011"><a href="#cb25-1011" aria-hidden="true" tabindex="-1"></a>        ,<span class="at">x =</span> <span class="st">"area under the ROC curve"</span>)</span>
<span id="cb25-1012"><a href="#cb25-1012" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb25-1013"><a href="#cb25-1013" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb25-1014"><a href="#cb25-1014" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="at">plot=</span>aucbar</span>
<span id="cb25-1015"><a href="#cb25-1015" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">filename =</span> <span class="st">"figures/aucbarplot.png"</span></span>
<span id="cb25-1016"><a href="#cb25-1016" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">width=</span><span class="dv">1430</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb25-1017"><a href="#cb25-1017" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">height=</span><span class="dv">874</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb25-1018"><a href="#cb25-1018" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">units =</span> <span class="st">'px'</span></span>
<span id="cb25-1019"><a href="#cb25-1019" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-1020"><a href="#cb25-1020" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-1021"><a href="#cb25-1021" aria-hidden="true" tabindex="-1"></a>  aucbar_with_text <span class="ot">&lt;-</span>  </span>
<span id="cb25-1022"><a href="#cb25-1022" aria-hidden="true" tabindex="-1"></a>  aucbar <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> <span class="fu">formatC</span>(auc, <span class="at">digits =</span> <span class="dv">3</span>)),<span class="at">position=</span><span class="fu">position_dodge</span>(<span class="at">width=</span><span class="fl">0.8</span>))  <span class="sc">+</span> </span>
<span id="cb25-1023"><a href="#cb25-1023" aria-hidden="true" tabindex="-1"></a>    <span class="fu">coord_cartesian</span>(<span class="at">clip =</span> <span class="st">"off"</span>)</span>
<span id="cb25-1024"><a href="#cb25-1024" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb25-1025"><a href="#cb25-1025" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="at">plot=</span>aucbar_with_text </span>
<span id="cb25-1026"><a href="#cb25-1026" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">filename =</span> <span class="st">"figures/aucbarplot_wtext.png"</span></span>
<span id="cb25-1027"><a href="#cb25-1027" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">width=</span><span class="dv">1430</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb25-1028"><a href="#cb25-1028" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">height=</span><span class="dv">874</span><span class="sc">*</span><span class="fl">1.3</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">9</span>)</span>
<span id="cb25-1029"><a href="#cb25-1029" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">units =</span> <span class="st">'px'</span></span>
<span id="cb25-1030"><a href="#cb25-1030" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-1031"><a href="#cb25-1031" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb25-1032"><a href="#cb25-1032" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-1033"><a href="#cb25-1033" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="../../license.html">
<p>License</p>
</a>
  </li>  
</ul>
    <div class="cookie-consent-footer"><a href="#" id="open_preferences_center">Cookie Preferences</a></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>