<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Dan McGuire">
<meta name="dcterms.date" content="2025-06-20">
<meta name="description" content="Cell typing using metagenes and smoothing.">

<title>Using marker gene driven metagene scores for granular hierarchical cell typing. – Blog</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/logo_Bruker_black.png" rel="icon" type="image/png">
<script src="../../site_libs/cookie-consent/cookie-consent.js"></script>
<link href="../../site_libs/cookie-consent/cookie-consent.css" rel="stylesheet">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-3a01e2046221230fdceeea94b1ec5d67.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-a4a11d514c7d463668e07712114998e6.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-1ea908db87654478666ac2bdcc4260f8.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-29W4MW0Y2W"></script>

<script type="text/plain" cookie-consent="tracking">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-29W4MW0Y2W', { 'anonymize_ip': true});
</script>

<script type="text/javascript" charset="UTF-8">
document.addEventListener('DOMContentLoaded', function () {
cookieconsent.run({
  "notice_banner_type":"simple",
  "consent_type":"implied",
  "palette":"light",
  "language":"en",
  "page_load_consent_levels":["strictly-necessary","functionality","tracking","targeting"],
  "notice_banner_reject_button_hide":false,
  "preferences_center_close_button_hide":false,
  "website_name":""
  ,
"language":"en"
  });
});
</script> 
  

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../.././assets/logo_Bruker_white.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">Topics</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../guides/index.html"> 
<span class="menu-text">Guides (new!)</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../bbytes/index.html"> 
<span class="menu-text">Browser Bytes (new!)</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../link-to-code.html"> 
<span class="menu-text">Source Code</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/company/bruker-spatial"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Using marker gene driven metagene scores for granular hierarchical cell typing.</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
                  <div>
        <div class="description">
          Cell typing using metagenes and smoothing.
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">cell typing</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliations</div>
    
      <div class="quarto-title-meta-contents">
      <p class="author">Dan McGuire <a href="https://orcid.org/0009-0006-0286-9625" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              Bruker Spatial Biology
            </p>
          <p class="affiliation">
              Github: <a href="https://github.com/dan11mcguire" target="_blank">dan11mcguire</a>
            </p>
        </div>
    </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">June 20, 2025</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2">
    <h2 id="toc-title">Contents</h2>
   
  <ul>
  <li><a href="#abstract" id="toc-abstract" class="nav-link active" data-scroll-target="#abstract"><span class="header-section-number">1</span> Abstract</a></li>
  <li><a href="#quick-links-to-the-code" id="toc-quick-links-to-the-code" class="nav-link" data-scroll-target="#quick-links-to-the-code"><span class="header-section-number">2</span> Quick links to the code</a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction"><span class="header-section-number">3</span> Introduction</a></li>
  <li><a href="#initial-look-at-our-dataset" id="toc-initial-look-at-our-dataset" class="nav-link" data-scroll-target="#initial-look-at-our-dataset"><span class="header-section-number">4</span> Initial look at our dataset</a></li>
  <li><a href="#ex1" id="toc-ex1" class="nav-link" data-scroll-target="#ex1"><span class="header-section-number">5</span> Example case 1: Subclustering T cells</a>
  <ul class="collapse">
  <li><a href="#tldr" id="toc-tldr" class="nav-link" data-scroll-target="#tldr"><span class="header-section-number">5.1</span> ‘Just The Code’</a></li>
  <li><a href="#detailed-walkthrough" id="toc-detailed-walkthrough" class="nav-link" data-scroll-target="#detailed-walkthrough"><span class="header-section-number">5.2</span> Detailed walkthrough</a>
  <ul class="collapse">
  <li><a href="#part-1-specifying-genes-associated-with-my-cell-types-based-on-prior-biology" id="toc-part-1-specifying-genes-associated-with-my-cell-types-based-on-prior-biology" class="nav-link" data-scroll-target="#part-1-specifying-genes-associated-with-my-cell-types-based-on-prior-biology"><span class="header-section-number">5.2.1</span> Part 1: Specifying genes associated with my cell types based on prior biology</a></li>
  <li><a href="#part-2-estimating-a-metagene-score-for-each-cell-type" id="toc-part-2-estimating-a-metagene-score-for-each-cell-type" class="nav-link" data-scroll-target="#part-2-estimating-a-metagene-score-for-each-cell-type"><span class="header-section-number">5.2.2</span> Part 2: Estimating a metagene score for each cell type</a></li>
  <li><a href="#part-3-assigning-cell-types-based-on-metagene-scores" id="toc-part-3-assigning-cell-types-based-on-metagene-scores" class="nav-link" data-scroll-target="#part-3-assigning-cell-types-based-on-metagene-scores"><span class="header-section-number">5.2.3</span> Part 3: Assigning cell types based on metagene scores</a></li>
  <li><a href="#granular-hierarchical-subtyping-and-using-a-pipeline" id="toc-granular-hierarchical-subtyping-and-using-a-pipeline" class="nav-link" data-scroll-target="#granular-hierarchical-subtyping-and-using-a-pipeline"><span class="header-section-number">5.2.4</span> Granular hierarchical subtyping and using a <code>pipeline</code></a></li>
  <li><a href="#examining-and-validating-clustering-results" id="toc-examining-and-validating-clustering-results" class="nav-link" data-scroll-target="#examining-and-validating-clustering-results"><span class="header-section-number">5.2.5</span> Examining and validating clustering results</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#ex2" id="toc-ex2" class="nav-link" data-scroll-target="#ex2"><span class="header-section-number">6</span> Example 2: Broad immune cell classification</a>
  <ul class="collapse">
  <li><a href="#tldr2" id="toc-tldr2" class="nav-link" data-scroll-target="#tldr2"><span class="header-section-number">6.1</span> ‘Just The Code’</a></li>
  <li><a href="#evaluation" id="toc-evaluation" class="nav-link" data-scroll-target="#evaluation"><span class="header-section-number">6.2</span> Evaluation</a></li>
  </ul></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion"><span class="header-section-number">7</span> Conclusion</a></li>
  <li><a href="#quick-summary-of-package-functions-used-in-this-post" id="toc-quick-summary-of-package-functions-used-in-this-post" class="nav-link" data-scroll-target="#quick-summary-of-package-functions-used-in-this-post"><span class="header-section-number">8</span> Quick summary of package functions used in this post</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="abstract" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Abstract</h1>
<p>This post showcases a hierarchical cell typing method (HieraType) optimized for detection and sub-clustering of immune cells. HieraType:</p>
<ul>
<li>Is organized around canonical marker genes, but harnesses cells’ complete expression profiles.</li>
<li>Can be adapted to any arbitrary cell type hierarchy.</li>
</ul>
<p>Code for the method and utility functions used in this post can be found in the R package <a href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/HieraType">HieraType</a>. The package contains carefully designed pipelines for:</p>
<ul>
<li>Subclassifying T-cells into 12 categories</li>
<li>Classifying a whole dataset into broad categories (epithelial, fibroblast, immune, endothelial), then subclassifying immune cells into 9 categories</li>
<li>Classifying native brain cells</li>
</ul>
<p>It is also possible to assemble pipelines for custom applications, e.g.&nbsp;focused on a specific tissue type.</p>
<p>In this post, we use a 6K-plex colon-cancer dataset to demonstrate two common use cases for the method:</p>
<ol type="1">
<li>Subtyping T cells</li>
<li>Broad immune cell typing.</li>
</ol>
</section>
<section id="quick-links-to-the-code" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Quick links to the code</h1>
<p>For those in a hurry to ‘just see the code’, they can jump ahead to the “Just The Code” sections for <a href="#tldr">Example 1</a>, or <a href="#tldr2">Example 2</a>.</p>
</section>
<section id="introduction" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Introduction</h1>
<p>Often (particularly with immune cells), we may want our cell type calls to be guided by specific sets of marker genes.<br>
In some spatially-resolved transcriptomics (SRT) datasets, directly using marker gene expression for determining cell type may be a challenge due to data sparsity, background, or segmentation ambiguity. These challenges together can make marker genes noisy individual data points.</p>
<p>However, marker genes remain a strong organizing principle for differentiating cell types. To overcome the intrinsic noise of single marker genes, the cell typing method we describe here captures information from both (a) similar cells and (b) biologically relevant genes, in order to construct metagenes that more stably estimate true marker expression. These marker metagene scores are then used for probabilistic cell type assignment. The method takes as input predefined ‘marker lists’ for each cell type, which make it easy to enforce prior biological assumptions about the marker genes and expression profiles we expect to see in our cell types. By not requiring any training data or an explicit reference profile, we expect the method to be robust to platform or batch effects that may otherwise challenge cell type labeling from pre-trained models or external reference data.</p>
<p>The basic framework for the method can be described in a few basic steps (we’ll describe each of these steps in more detail throughout our examples.)</p>
<ol type="1">
<li>First, for each cell type, we define a set of genes we expect may be ‘marker genes’ or highly expressed based on prior data or biology (either the literature, a reference dataset, or some combination of both).</li>
<li>Next, we construct metagene scores from our data for each cell type based on the genes specified in (1).</li>
<li>Finally, we model the metagene scores, where the model estimates a posterior probability that each cell belongs to any of the possible cell types.</li>
</ol>
<p>We’ll use an internally generated colon cancer dataset profiled with the CosMx Human 6K Discovery Panel comprised of about 112k cells, and work through two common cell typing scenarios in which we envision this methodology can be useful:</p>
<ol type="1">
<li>Subclustering an existing set of cell types:</li>
</ol>
<ul>
<li>In this first case, we’ll assume we’ve already done some basic clustering, but don’t quite have the level of cell type granularity we’d like. We’ll take a ‘T cell’ cluster and sub-classify those cells into CD8-T and CD4-T, followed by more granular T cell subcategories.</li>
</ul>
<ol start="2" type="1">
<li>Top down hierarchical cell typing with focus on immune cells:</li>
</ol>
<ul>
<li>Here we’ll start from scratch; first assigning cells into major categories and identifying immune cells, then classifying those immune cells into various sub classifications of myeloid and lymphocyte cell types.</li>
</ul>
</section>
<section id="initial-look-at-our-dataset" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Initial look at our dataset</h1>
<p>Let’s start by taking an initial look at our data, then walk through some cell typing scenarios. First, we can load in our data and take a look at some unsupervised leiden clustering we’ve already performed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(HieraType)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Note, using a v3-style Seurat Object below, i.e., `sem[["RNA"]]@counts`</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="do">## For v5-style, the syntax would be like `sem[["RNA"]]$counts`</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">Seurat.object.assay.version =</span> <span class="st">"v3"</span>)  </span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>sem <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"colon_seurat.rds"</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>normed <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">totalcount_norm</span>(Matrix<span class="sc">::</span><span class="fu">t</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts))</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>sem <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">SetAssayData</span>(sem, <span class="at">layer =</span> <span class="st">"data"</span>, <span class="at">new.data =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(normed))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Spatial plot of initial clusters</strong></p>
<p>Here’s a spatial plot of our leiden clusters. At first glance, we see that some of these unsupervised clusters (i.e., 3, 0, 1, 2, and 7 for example) have distinct spatial patterns; it definitely looks like some meaningful biological structure in these tissue were captured by unsupervised clustering.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>sem<span class="sc">@</span>meta.data<span class="sc">$</span>portion <span class="ot">&lt;-</span> <span class="st">"part_a"</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>sem<span class="sc">@</span>meta.data<span class="sc">$</span>portion[sem<span class="sc">@</span>meta.data<span class="sc">$</span>fov <span class="sc">%in%</span> <span class="dv">60</span><span class="sc">:</span><span class="dv">73</span>] <span class="ot">&lt;-</span> <span class="st">"part_b"</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>xyplist <span class="ot">&lt;-</span> </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(<span class="fu">split</span>(<span class="fu">data.table</span>(sem<span class="sc">@</span>meta.data), <span class="at">by =</span> <span class="st">"portion"</span>), <span class="cf">function</span>(xx){</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  HieraType<span class="sc">::</span><span class="fu">xyplot</span>(<span class="st">"clusters_unsup"</span>, <span class="at">metadata =</span> xx,<span class="at">ptsize =</span> <span class="fl">0.1</span>) <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>xypl <span class="ot">&lt;-</span> </span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist=</span>xyplist, <span class="at">nrow=</span><span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">'white'</span>,<span class="at">color=</span><span class="st">'white'</span>)) <span class="sc">+</span> </span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  cowplot<span class="sc">::</span><span class="fu">panel_border</span>(<span class="at">remove=</span><span class="cn">TRUE</span>) </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(xypl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="wider-image">
<p><img src="./figures/xyp_unsup.png" alt="Wide Image"></p>
</div>
<p><strong>Heatmap of initial clusters </strong></p>
<p>We can make a marker gene heatmap to identify some of the top marker genes for each cluster and see if we recognize any of the celltypes represented by the unsupervised clusters. Here we can use a couple of convenience functions provided in the R package <code>HieraType</code>.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## calculate fold change, average expression, and other summary metrics </span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="do">## for each gene and each cluster compared to other clusters</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="do">## This is the data underlying the marker heatmap.</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>fctbl_unsup <span class="ot">&lt;-</span> </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>HieraType<span class="sc">::</span><span class="fu">clusterwise_foldchange_metrics</span>(<span class="at">normed  =</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>data</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>                                          ,<span class="at">metadata =</span> sem<span class="sc">@</span>meta.data</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>                                          ,<span class="at">cluster_column =</span> <span class="st">"clusters_unsup"</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>                                          )</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="do">## make a marker heatmap</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>hm_unsup <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">marker_heatmap</span>(fctbl_unsup</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>                           ,<span class="at">extras =</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"CD4"</span>, <span class="st">"FOXP3"</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>                                       , <span class="st">"CD8B"</span>, <span class="st">"CD8A"</span>)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>                           )</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hm_unsup)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="wider-image">
<p><img src="./figures/hm_unsup_annotated.PNG" alt="Wide Image"></p>
</div>
<p>Let’s suppose for sake of demonstration that I recognize (and am comfortable enough) to label these clusters with cell type names based on the marker genes shown in the heatmap.</p>
<p>However, there are one or a few of my immune clusters I’ve identified where I desire more granularity. In this case, I’m particularly interested in my T cells; in the bottom-left hand corner of the heatmap plot above, I see that ‘cluster 6’ is capturing T-cells, showing enrichment in some key marker genes like CD3 (CD3D, CD3E) as well as other markers like CD2, IL7R, and ITK.</p>
<p>I’d like to split these cells further into at least a few meaningful T-cell subtypes, at minimum: “CD8+”, “CD4+”, and “Treg”.</p>
</section>
<section id="ex1" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Example case 1: Subclustering T cells</h1>
<section id="tldr" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="tldr"><span class="header-section-number">5.1</span> ‘Just The Code’</h2>
<p>Here’s the code for classifying leiden cluster ‘6’ into <code>cd8t</code> and <code>cd4t</code> cell types , followed by more granular T cell subtyping. We use the R package <code>HieraType</code> with a prespecified <code>tcell_pipeline</code>, and corresponding lists of marker genes included as package datasets. A <code>pipeline</code> specifies the hierarchical order we want to do our cell typing, and which cell types go in each level of the hierarchy. Each cell type has a list of associated marker genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"pipeline_tcell"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(pipeline_tcell, <span class="at">max.level =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 3
 $ markerslists   :List of 3
  ..$ tmajor :List of 2
  .. ..- attr(*, "class")= chr [1:2] "list" "markerslist"
  ..$ t4minor:List of 7
  .. ..- attr(*, "class")= chr [1:2] "list" "markerslist"
  ..$ t8minor:List of 5
  .. ..- attr(*, "class")= chr [1:2] "list" "markerslist"
 $ priors         :List of 2
  ..$ t4minor: chr "tmajor"
  ..$ t8minor: chr "tmajor"
 $ priors_category:List of 2
  ..$ t4minor: chr "cd4t"
  ..$ t8minor: chr "cd8t"
 - attr(*, "class")= chr [1:2] "list" "pipeline"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pipeline_tcell<span class="sc">$</span>markerslists<span class="sc">$</span>tmajor)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "cd8t" "cd4t"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pipeline_tcell<span class="sc">$</span>markerslists<span class="sc">$</span>t4minor)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "cd4_naive" "cd4_tem"   "cd4_tcm"   "cd4_th1"   "cd4_th2"   "cd4_th17" 
[7] "cd4_treg" </code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pipeline_tcell<span class="sc">$</span>markerslists<span class="sc">$</span>t8minor)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "cd8_naive"     "cd8_tem"       "cd8_tcm"       "cd8_cytotoxic"
[5] "cd8_exhausted"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="do">### these are cells I want to subcluster</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>tcell_ids <span class="ot">&lt;-</span> sem<span class="sc">@</span>meta.data[sem<span class="sc">@</span>meta.data<span class="sc">$</span>clusters_unsup<span class="sc">==</span><span class="dv">6</span>,<span class="st">"cell_ID"</span>]</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="do">### Run a 'tcell'-typing pipeline, which </span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="do">### first classifies into T8/T4 categories (using the `tmajor` markerslist), </span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="do">### then runs subclassification for T4 and T8 using the t4minor and t8minor markerslists.</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>tcell_typing <span class="ot">&lt;-</span> </span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="fu">run_pipeline</span>(<span class="at">pipeline =</span> pipeline_tcell</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">counts_matrix =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[,tcell_ids])</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">adjacency_matrix =</span> <span class="fu">as</span>(sem<span class="sc">@</span>graphs<span class="sc">$</span>umapgrph</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>                                    ,<span class="st">"CsparseMatrix"</span>)[tcell_ids,tcell_ids]</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">celltype_call_threshold =</span> <span class="fl">0.5</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>             )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The data table <code>post_probs</code> in our <code>tcell_typing</code> object holds the high-level cell typing results; the <code>celltype_thresh</code> column shows the <em>most granular</em> cell type call where posterior probability is <span class="math inline">\(&gt; 0.5\)</span> (our <code>celltype_call_threshold</code>). We can see the posterior probablity for the <code>celltype_thresh</code> category in <code>best_score_thresh</code>.<br>
The best granular cell type call (regardless of score) is given by the <code>celltype_granular</code> column, with <code>best_score_granular</code> indicating the corresponding posterior probability.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(tcell_typing, <span class="at">max.level =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 3
 $ post_probs     :List of 1
  ..$ tmajor:Classes 'data.table' and 'data.frame': 7356 obs. of  5 variables:
  .. ..- attr(*, ".internal.selfref")=&lt;externalptr&gt; 
  .. ..- attr(*, "sorted")= chr "cell_ID"
 $ models         :List of 3
  ..$ tmajor :List of 3
  ..$ t4minor:List of 3
  ..$ t8minor:List of 3
 $ metagene_scores:List of 3
  ..$ tmajor :List of 2
  ..$ t4minor:List of 7
  ..$ t8minor:List of 5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>tcell_typing<span class="sc">$</span>post_probs<span class="sc">$</span>tmajor[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>][]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Key: &lt;cell_ID&gt;
        cell_ID celltype_thresh celltype_granular best_score_granular
         &lt;char&gt;          &lt;char&gt;            &lt;char&gt;               &lt;num&gt;
 1: c_1_12_1014            cd8t     cd8_exhausted           0.4691963
 2: c_1_12_1066            cd4t           cd4_th1           0.3352265
 3: c_1_12_1091         cd4_th2           cd4_th2           0.9683195
 4: c_1_12_1175   cd8_cytotoxic     cd8_cytotoxic           0.8777777
 5: c_1_12_1270         cd8_tem           cd8_tem           0.5882417
 6: c_1_12_1286        cd4_treg          cd4_treg           0.9909980
 7: c_1_12_1325            cd4t           cd4_tcm           0.4647576
 8: c_1_12_1346        cd4_treg          cd4_treg           0.8298557
 9: c_1_12_1349            cd8t           cd8_tem           0.3939835
10: c_1_12_1394   cd8_cytotoxic     cd8_cytotoxic           0.9587283
    best_score_thresh
                &lt;num&gt;
 1:         0.9999999
 2:         0.6751557
 3:         0.9683195
 4:         0.8777777
 5:         0.5882417
 6:         0.9909980
 7:         0.9984388
 8:         0.8298557
 9:         0.9999814
10:         0.9587283</code></pre>
</div>
</div>
<p>Now lets take a closer look at the details and describe the methodology for each step.</p>
</section>
<section id="detailed-walkthrough" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="detailed-walkthrough"><span class="header-section-number">5.2</span> Detailed walkthrough</h2>
<section id="part-1-specifying-genes-associated-with-my-cell-types-based-on-prior-biology" class="level3" data-number="5.2.1">
<h3 data-number="5.2.1" class="anchored" data-anchor-id="part-1-specifying-genes-associated-with-my-cell-types-based-on-prior-biology"><span class="header-section-number">5.2.1</span> Part 1: Specifying genes associated with my cell types based on prior biology</h3>
<p>We’ll describe the first level of cell type classification from our example above to illustrate the method, where we suppose we want to classify <code>cd4t</code> vs.&nbsp;<code>cd8t</code>. Creating a marker list, by specifying genes I expect <strong>might</strong> be associated with my cell types is the first step of this algorithm. For each of my cell types, <code>cd4t</code> and <code>cd8t</code>, I specify</p>
<ul>
<li>a set of primary defining markers or “index genes” (these are genes that should be fairly exclusive to the cell type if possible)</li>
<li>a set of other marker genes which I expect could be expressed as “predictors”</li>
</ul>
<p>In this case, we’ll use [CD4], and [CD8A, CD8B], as “index genes” for <code>cd4t</code> and <code>cd8t</code> respectively. The other genes we’ll specify as predictors are based on the literature. For example, FOXP3 is often associated with cd4t ‘Treg’ cells, so we include it as a predictor here for <code>cd4t</code>. In general, it’s okay to include a gene as a “predictor” even if we only expect to see it in a subset of cells (like FOXP3 in this example, which we expect to be present in a ‘Treg’ subtype, but not all <code>cd4t</code> cells).</p>
<p>It’s okay to include markers in multiple cell types if that fits our mental model. For example, CD3D, CD3E, and CD3G are general T cell markers, and so are included as predictors for both subtypes, as are other genes like IL7R, CD27, SELL which could be seen in either <code>cd4t</code> or <code>cd8t</code> cells. When manually creating a marker list, consider “more predictors is better” as a general rule of thumb; a larger number of predictor genes can help build more accurate metagene models for celltype index markers.</p>
<p>We can create our marker list manually like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>markerslist_tcellmajor <span class="ot">&lt;-</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>HieraType<span class="sc">::</span><span class="fu">make_markerslist</span>(</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">index_marker =</span> <span class="fu">list</span>(</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"cd8t"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"CD8B"</span>, <span class="st">"CD8A"</span>)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"cd4t"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"CD4"</span>)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">predictors =</span> <span class="fu">list</span>(</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"cd8t"</span> <span class="ot">=</span>  <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD3G"</span>, <span class="st">"CD8A"</span>, <span class="st">"CD8B"</span>, <span class="st">"IL7R"</span>, <span class="st">"CD27"</span>, <span class="st">"SELL"</span>, <span class="st">"CCR7"</span>, <span class="st">"TCF7"</span>, <span class="st">"LEF1"</span>, </span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"FOXP1"</span>, <span class="st">"BACH2"</span>, <span class="st">"KLF2"</span>, <span class="st">"CD28"</span>, <span class="st">"ICOS"</span>, <span class="st">"STAT3"</span>, <span class="st">"STAT5"</span>, <span class="st">"RUNX3"</span>, <span class="st">"BATF"</span>, <span class="st">"EOMES"</span>, </span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"TBX21"</span>, <span class="st">"PRDM1"</span>, <span class="st">"PDCD1"</span>, <span class="st">"LAG3"</span>, <span class="st">"TIGIT"</span>, <span class="st">"HAVCR2"</span>, <span class="st">"GZMK"</span>, <span class="st">"GZMA"</span>, <span class="st">"GZMB"</span>, <span class="st">"PRF1"</span>, </span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"GNLY"</span>, <span class="st">"KLRG1"</span>, <span class="st">"KLRD1"</span>, <span class="st">"IFNG"</span>, <span class="st">"CCL5"</span>, <span class="st">"CX3CR1"</span>, <span class="st">"NKG7"</span>, <span class="st">"FAS"</span>, <span class="st">"FASLG"</span>, <span class="st">"IL2"</span>, <span class="st">"CD244"</span>, </span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"CXCR3"</span>, <span class="st">"KLRB1"</span>, <span class="st">"TOX"</span>, <span class="st">"NR4A2"</span>, <span class="st">"BCL6"</span>, <span class="st">"CXCL13"</span>)</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"cd4t"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD3G"</span>, <span class="st">"CD4"</span>, <span class="st">"IL7R"</span>, <span class="st">"CD27"</span>, <span class="st">"SELL"</span>, <span class="st">"CCR7"</span>, <span class="st">"TCF7"</span>, <span class="st">"LEF1"</span>, <span class="st">"FOXP1"</span>, </span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"BACH2"</span>, <span class="st">"KLF2"</span>, <span class="st">"CD28"</span>, <span class="st">"ICOS"</span>, <span class="st">"CD40LG"</span>, <span class="st">"STAT3"</span>, <span class="st">"STAT5"</span>, <span class="st">"RUNX1"</span>, <span class="st">"RUNX3"</span>, <span class="st">"BATF"</span>, </span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"GATA3"</span>, <span class="st">"TBX21"</span>, <span class="st">"RORC"</span>, <span class="st">"FOXP3"</span>, <span class="st">"IL2RA"</span>, <span class="st">"CTLA4"</span>, <span class="st">"PDCD1"</span>, <span class="st">"LAG3"</span>, <span class="st">"TIGIT"</span>, <span class="st">"HAVCR2"</span>, </span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"IL2"</span>, <span class="st">"IFNG"</span>, <span class="st">"IL4"</span>, <span class="st">"IL5"</span>, <span class="st">"IL13"</span>, <span class="st">"IL17A"</span>, <span class="st">"IL17F"</span>, <span class="st">"IL10"</span>, <span class="st">"CCL5"</span>, <span class="st">"CXCR3"</span>, <span class="st">"CCR6"</span>, </span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"CXCR5"</span>, <span class="st">"PRDM1"</span>, <span class="st">"BCL6"</span>, <span class="st">"EOMES"</span>, <span class="st">"CXCL13"</span>, <span class="st">"TOX"</span>)</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The object created looks like this below</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(markerslist_tcellmajor)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 2
 $ cd8t:List of 3
  ..$ index_marker                               : chr [1:2] "CD8B" "CD8A"
  ..$ predictors                                 : chr [1:48] "CD3D" "CD3E" "CD3G" "CD8A" ...
  ..$ use_offclass_markers_as_negative_predictors: logi TRUE
 $ cd4t:List of 3
  ..$ index_marker                               : chr "CD4"
  ..$ predictors                                 : chr [1:48] "CD3D" "CD3E" "CD3G" "CD4" ...
  ..$ use_offclass_markers_as_negative_predictors: logi TRUE
 - attr(*, "class")= chr [1:2] "list" "markerslist"</code></pre>
</div>
</div>
<p>In practice, we can save and reuse our marker lists for different cell types. While it should be easy to create/modify a custom list from the literature, this particular list is included as a package dataset <code>markerslist_tcellmajor</code> to facilitate ease of use.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"markerslist_tcellmajor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(markerslist_tcellmajor)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 2
 $ cd8t:List of 3
  ..$ index_marker                               : chr [1:2] "CD8B" "CD8A"
  ..$ predictors                                 : chr [1:48] "CD3D" "CD3E" "CD3G" "CD8A" ...
  ..$ use_offclass_markers_as_negative_predictors: logi TRUE
 $ cd4t:List of 3
  ..$ index_marker                               : chr "CD4"
  ..$ predictors                                 : chr [1:48] "CD3D" "CD3E" "CD3G" "CD4" ...
  ..$ use_offclass_markers_as_negative_predictors: logi TRUE
 - attr(*, "class")= chr [1:2] "list" "markerslist"</code></pre>
</div>
</div>
</section>
<section id="part-2-estimating-a-metagene-score-for-each-cell-type" class="level3" data-number="5.2.2">
<h3 data-number="5.2.2" class="anchored" data-anchor-id="part-2-estimating-a-metagene-score-for-each-cell-type"><span class="header-section-number">5.2.2</span> Part 2: Estimating a metagene score for each cell type</h3>
<p>Because single count values marker genes can be noisy, we seek to derive variables that more stably estimate marker expression. To do so, we’ll exploit biologically relevant genes (e.g.&nbsp;CD8A expression is predictive of CD8B expression), and cells with similar expression patterns. To do so, we fit a model for each cell type, using the primary index marker as the response variable and covariates derived from the predictor genes. The predicted values of the primary index marker for this model make a continuous “metagene score” that will later be used for probabilistic cell type assignment. You can think of each metagene score as estimating the canonical marker gene’s expression for a single cell type; e.g.&nbsp;we will produce a FOXP3-estimating metagene to capture Treg’s, a CD68-estimating metagene to capture macrophages, etc…</p>
<p>We fit our models for each cell type class using the <code>fit_metagene_scores</code> function. This function uses ‘smoothing’ (averaging over a set of similar cells), as well as our prior biology / assumptions about which markers are associated with each celltype to predict the pearson residuals of the index gene.</p>
<p>The default regression model setup looks like this: <span class="math display">\[Y = WY\eta  + WX\beta + X\gamma + \epsilon \]</span> where</p>
<ul>
<li><span class="math inline">\(Y:=\)</span> pearson residuals of the index marker (our response variable)</li>
<li><span class="math inline">\(X:=\)</span> (totalcount-normalized) matrix of predictor genes</li>
<li><span class="math inline">\(W:=\)</span> a cells <span class="math inline">\(\times\)</span> cells neighborhood matrix, denoting cells with similar expression profiles.</li>
<li><span class="math inline">\(\eta\)</span>, <span class="math inline">\(\beta\)</span>, and <span class="math inline">\(\gamma\)</span> are the regression coefficients associated with the ‘neighbor-averaged’ index gene, ‘neighbor-averaged’ predictor genes, and predictor genes within the cell respectively.</li>
</ul>
<p>One way to think about this model specification could be as an implementation of a ‘smoothing model’, which leverages information across multiple dimensions:</p>
<ul>
<li>the index gene’s mean expression in similar cells (<span class="math inline">\(WY\)</span>) (roughly, neighbors in UMAP space)</li>
<li>expression of other expected marker genes (<span class="math inline">\(X\)</span>) (our predictor genes defined from biological knowledge)</li>
<li>expression of other expected marker genes in similar cells (<span class="math inline">\(WX\)</span>)</li>
</ul>
<p>We use a non-negative least squares <span class="citation" data-cites="nnlsref">(<a href="#ref-nnlsref" role="doc-biblioref">Slawski and Hein 2013</a>)</span> (NNLS) regression to estimate our models. The NNLS model enforces that all of the covariates based on predictor genes we specified must have positive regression coefficients (in practice we’ll observe that a number of regression coefficients may shrink to zero).</p>
<p>By specifying <code>use_offclass_markers_as_negative_predictors = TRUE</code> in the section above, we further assume that predictors for <em>other celltypes</em> that we haven’t listed can be used as <em>negative</em> predictors of our index gene. For example, genes like like “GZMA” and “GZMB”, which are markers associated with <code>cd8t</code>, will be assumed to be <em>negative</em> predictors of the other classes (<code>cd4t</code>) where they are not listed. This argument should in general be set to TRUE, as it’s useful to specify which genes we <em>don’t</em> expect to see in a celltype. If your markers list has extensive genes expected to be enriched in multiple cell types, and those genes are for some reason omitted from many of the other cell types, then a user might consider specifying this argument as FALSE; e.g.&nbsp;if many genes like CD3E are specified in markers lists for some T cell subtypes but not all (despite being ubiquitously expressed).</p>
<p>Estimating this model from the dataset at hand rather than using a reference model could make this framework more robust to platform effects or batch effects compared to pre-trained models or reference profiles which used different technology or data with slightly different biology.</p>
<p>Genes not included in the marker gene lists will not be considered in assigning cell types. Because the model is only using genes we’ve specified in our ‘markers list’, we have complete control over which genes influence the metadata scores, reducing the chance that noisy or irrelevant genes can have an impact on our cell type classifications. Furthermore, by estimating the metagenes directly from the dataset at hand, we are not reliant on reference datasets or pre-trained models influenced by different technology (example: scRNA-seq) or underlying biology (example: different tissue type or disease status, etc).</p>
<p>Here I call the <code>fit_metagene_scores</code> function to estimate these models for cells corresponding to the unsupervised cluster I’ve identified to be T cells.<br>
For the cells <span class="math inline">\(\times\)</span> cells <code>adjacency_matrix</code>, I’ve used the nearest neighbor matrix in my Seurat object. This is the same graph that was used to generate my UMAP and my unsupervised leiden clusters. After fitting our regression models for each cell type, we’ll use the predicted scores as ‘metagenes’ that can be clustered in the next step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>tcell_ids <span class="ot">&lt;-</span> sem<span class="sc">@</span>meta.data[sem<span class="sc">@</span>meta.data<span class="sc">$</span>clusters_unsup<span class="sc">==</span><span class="dv">6</span>,<span class="st">"cell_ID"</span>]</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>metagenes_t <span class="ot">&lt;-</span> </span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fit_metagene_scores</span>(</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">markerslist =</span> markerslist_tcellmajor</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">counts_matrix =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[,tcell_ids])</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">adjacency_matrix =</span> <span class="fu">as</span>(sem<span class="sc">@</span>graphs<span class="sc">$</span>umapgrph</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>                           ,<span class="st">"CsparseMatrix"</span>)[tcell_ids,tcell_ids]</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>metagenes_t</code> object we’ve created has an element for each celltype, including the regression coefficients (<code>bhat</code>), and predicted values we’ll use for clustering (<code>yhat</code>) for each index marker gene for cell type.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(metagenes_t,<span class="at">max.level =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 2
 $ cd8t:List of 7
  ..$ CD8B   :List of 5
  ..$ CD8A   :List of 5
  ..$ yhat   : Named num [1:7356] 0.593 1.32 2.959 1.375 2.478 ...
  .. ..- attr(*, "names")= chr [1:7356] "c_1_12_1066" "c_1_12_1175" "c_1_12_1270" "c_1_12_1349" ...
  ..$ y      : Named num [1:7356] -0.208 -0.178 2.047 -0.207 -0.242 ...
  .. ..- attr(*, "names")= chr [1:7356] "c_1_12_1066" "c_1_12_1175" "c_1_12_1270" "c_1_12_1349" ...
  ..$ ypost  : Named num [1:7356] -0.1796 -0.0223 1.5179 -0.0269 0.1805 ...
  .. ..- attr(*, "names")= chr [1:7356] "c_1_12_1066" "c_1_12_1175" "c_1_12_1270" "c_1_12_1349" ...
  ..$ bhat   : num [1:125, 1:2] 0.1904 0.1342 0.1113 0.1057 0.0912 ...
  .. ..- attr(*, "dimnames")=List of 2
  ..$ nnpcfit:List of 7
  .. ..- attr(*, "class")= chr [1:2] "nsprcomp" "prcomp"
 $ cd4t:List of 6
  ..$ CD4    :List of 5
  ..$ yhat   : Named num [1:7356] 0.29 -1.426 -0.968 -0.692 -1.945 ...
  .. ..- attr(*, "names")= chr [1:7356] "c_1_12_1066" "c_1_12_1175" "c_1_12_1270" "c_1_12_1349" ...
  ..$ y      : Named num [1:7356] -0.271 -0.235 -0.418 -0.27 -0.312 ...
  .. ..- attr(*, "names")= chr [1:7356] "c_1_12_1066" "c_1_12_1175" "c_1_12_1270" "c_1_12_1349" ...
  ..$ ypost  : Named num [1:7356] -0.294 -0.366 -0.469 -0.321 -0.507 ...
  .. ..- attr(*, "names")= chr [1:7356] "c_1_12_1066" "c_1_12_1175" "c_1_12_1270" "c_1_12_1349" ...
  ..$ bhat   : num [1:125, 1] 0.3763 0.2525 0.1551 0.1283 0.0749 ...
  .. ..- attr(*, "dimnames")=List of 2
  ..$ nnpcfit:List of 7
  .. ..- attr(*, "class")= chr [1:2] "nsprcomp" "prcomp"</code></pre>
</div>
</div>
<p>For cell types using multiple ‘index marker genes’, like <code>cd8t</code> here (using CD8A and CD8B), metagene scores are combined together using non-negative sparse pca (implemented in the <code>nsprcomp</code> R package) <span class="citation" data-cites="nonnegpca">(<a href="#ref-nonnegpca" role="doc-biblioref">Sigg and Buhmann 2008</a>)</span>. This allows for summarizing an overall cell type metagene score, using a constrained non-negative combination of individual index-gene scores that explains maximal variation across individual index gene scores. Below, we see the PCA rotation representing the individual weights given to CD8A and CD8B metagene scores.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>metagenes_t<span class="sc">$</span>cd8t<span class="sc">$</span>nnpcfit</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Standard deviations (1, .., p=1):
[1] 1.020128

Rotation (n x k) = (2 x 1):
           PC1
CD8B 0.6578375
CD8A 0.7531599</code></pre>
</div>
</div>
<p>We can get a better understanding of the metagene model outputs by looking at a few diagnostics plots. In the below, we plot the regression coefficients for each model, ordering them on the x-axis by their ranking. For each cell type, the positive <span class="math inline">\(\hat{\beta}\)</span> correspond to genes we specified as predictors, and negative <span class="math inline">\(\hat{\beta}\)</span> were predictors for either of the other two cell types.<br>
A lot of the covariates which have the largest magnitudes have “<em>lag.</em>” in their name indicating they correspond to a smoothed (or neighbor-averaged) covariate for that gene.</p>
<p>We also see that many genes end up with weights of 0. These genes failed to be predictive in the direction our biological priors expected.</p>
<p>For <code>cd4t</code>, we see that smoothed CD4 (indicated by ‘<em>lag.y.CD4</em>’ in the plot) is the strongest predictor of observed ‘CD4’. For <code>cd8t</code>, ‘<em>lag.y.CD8A</em>’ and ‘<em>lag.y.CD8B</em>’ (indicating smoothed CD8A and CD8B) are also among the strongest predictors, but not first in rank.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>coefplots <span class="ot">&lt;-</span> <span class="fu">metagene_coefficients_plot</span>(metagenes_t)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>coefplots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(coefplots, <span class="cf">function</span>(x){x <span class="sc">+</span>   <span class="fu">theme</span>(<span class="at">text=</span><span class="fu">element_text</span>(<span class="at">face=</span><span class="st">'bold'</span>,<span class="at">size=</span><span class="dv">18</span>))</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  }) <span class="do">## just make the text bigger and bold</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> coefplots, <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="wider-image">
<p><img src="./figures/coefplots_tcells.png" alt="Wide Image"></p>
</div>
<p>A pairsplot can show us how well the metagene scores we’ve created are distinct from each other. Ideally, if our cell type classes are mutually exclusive, we’d want to not see too many cells which are ‘positive’ for multiple cell types at the same time.</p>
<p>Here, we see that <code>cd4t</code> and <code>cd8t</code> metagene scores are negatively correlated with each other, which is good to see.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">metagene_pairsplot</span>(metagenes_t, <span class="at">var=</span> <span class="st">"yhat"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/yhatpairs_tcells.png" class="img-fluid figure-img" width="1918"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="part-3-assigning-cell-types-based-on-metagene-scores" class="level3" data-number="5.2.3">
<h3 data-number="5.2.3" class="anchored" data-anchor-id="part-3-assigning-cell-types-based-on-metagene-scores"><span class="header-section-number">5.2.3</span> Part 3: Assigning cell types based on metagene scores</h3>
<p>Now we use the metagene scores to classify cell types. To do so, we estimate the multivariate distribution of metagene scores that characterize each cell type. This allows us to compute the likelihoods for a cell belonging to any of these types, and infer a posterior probability of T-cell subtype for each cell. The <code>cluster_metagenes</code> function we run below estimates a mixture of overfitted multivariate gaussian mixture models <span class="citation" data-cites="aragam2020identifiability">(<a href="#ref-aragam2020identifiability" role="doc-biblioref">Aragam et al. 2020</a>)</span> to estimate the distribution of metagene scores.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>model_t <span class="ot">&lt;-</span> </span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>HieraType<span class="sc">::</span><span class="fu">cluster_metagenes</span>(<span class="at">metagenes =</span> metagenes_t)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>At a high level, the <code>cluster_metagenes()</code> function first identifies all cells with positive metagene scores for a particular cell type. We might loosely interpret these cells as ‘anchor cells’. The data may look like this for example, where here I have subset my original data to cells which have positive scores for my celltype <code>cd8t</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>yh <span class="ot">&lt;-</span> <span class="fu">make_scores_matrix</span>(<span class="at">metagenes =</span> metagenes_t, <span class="st">"yhat"</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>yh[yh[,<span class="st">"cd8t"</span>] <span class="sc">&gt;</span> <span class="dv">0</span>,][<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, a mixture model is fit to these (2-dimensional) scores, by default using 6 components. The estimates corresponding to the individual mixture models for each cell type are found in the <code>models</code> list, with the means (<code>mu_hat</code>), covariance matrices (<code>sigma_hat</code>), and prior probability (<code>pihat</code>) corresponding to each mixture component.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(model_t<span class="sc">$</span>models, <span class="at">max.level =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 2
 $ cd8t:List of 6
  ..$ post_probs  :Classes 'data.table' and 'data.frame':   2528 obs. of  9 variables:
  .. ..- attr(*, ".internal.selfref")=&lt;externalptr&gt; 
  ..$ pihat       : Named num [1:6] 0.2785 0.103 0.0437 0.3024 0.0747 ...
  .. ..- attr(*, "names")= chr [1:6] "k1" "k2" "k3" "k4" ...
  ..$ llmat       : num [1:2528, 1:6] -2.97 -2.87 -2.95 -3.42 -2.43 ...
  .. ..- attr(*, "dimnames")=List of 2
  ..$ mu_hat      : num [1:6, 1:2] 2.136 0.136 2.966 1.196 0.645 ...
  .. ..- attr(*, "dimnames")=List of 2
  ..$ sigma_hat   :List of 6
  ..$ loglik_iters: num [1:262] -4920 -4891 -4883 -4879 -4876 ...
 $ cd4t:List of 6
  ..$ post_probs  :Classes 'data.table' and 'data.frame':   2623 obs. of  9 variables:
  .. ..- attr(*, ".internal.selfref")=&lt;externalptr&gt; 
  ..$ pihat       : Named num [1:6] 0.1544 0.1398 0.2798 0.0414 0.0677 ...
  .. ..- attr(*, "names")= chr [1:6] "k1" "k2" "k3" "k4" ...
  ..$ llmat       : num [1:2623, 1:6] -2.03 -56.28 -126.91 -3.61 -5.12 ...
  .. ..- attr(*, "dimnames")=List of 2
  ..$ mu_hat      : num [1:6, 1:2] -1.074 -0.779 -1.13 -1.796 -1.429 ...
  .. ..- attr(*, "dimnames")=List of 2
  ..$ sigma_hat   :List of 6
  ..$ loglik_iters: num [1:173] -5230 -5189 -5167 -5155 -5147 ...</code></pre>
</div>
</div>
<p>Typically we shouldn’t need to interact with this part of the output. After the mixture models for each cell type are fitted, we can score all cells, and get an overall posterior probability that a cell belongs to any of the two major classes using Bayes rule.</p>
<p>Let <span class="math inline">\(y_c\)</span> be the cell type category for cell <span class="math inline">\(c\)</span>, <span class="math inline">\(scores_c\)</span> the corresponding metagene scores, and <span class="math inline">\(t\)</span> denoting a possible cell ‘type’ or category (for <span class="math inline">\(t = 1,...,T\)</span>).</p>
<p>We compute the overall posterior probability that <span class="math inline">\(y_c\)</span> is a particular type <span class="math inline">\(t\)</span>, as</p>
<p><span class="math display">\[P(y_c = t | scores_c) = \frac{P(scores_c | y_c = t)P(y_c = t)}{\sum_{l=1}^T P(scores_c | y_c = l)P(y_c = l)}\]</span></p>
<p>Using the individual mixture models for a particular cell type with <span class="math inline">\(K\)</span> mixture components, we have <span class="math display">\[P(scores_c | y_c = t) = \sum_{j=1}^K P(scores_c | k=j, y_c = t) P(k=j | y_c = t)\]</span></p>
<p>The prior probability for a particular cell type class <span class="math inline">\(t\)</span> can be estimated by counting the number of cells used to estimate the mixture model in each cell type, and computing a proportion amongst all possible cell type classes, i.e., <span class="math display">\[P(y_c = t) = \frac{\sum_{c=1}^C I(scores_{c,t} &gt; 0)}{\sum_{l=1}^T \sum_{c=1}^C I(scores_{c,l} &gt; 0)}\]</span> These probabilities are in the <code>post_probs</code> data.table object of the model. The “best class” column gives each cell’s best-fitting cell type.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>model_t<span class="sc">$</span>post_probs[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>][]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        cell_ID best_class best_score        cd8t         cd4t
         &lt;char&gt;     &lt;char&gt;      &lt;num&gt;       &lt;num&gt;        &lt;num&gt;
 1: c_1_12_1066       cd4t  0.6751557 0.324844253 6.751557e-01
 2: c_1_12_1175       cd8t  0.9999986 0.999998637 1.363123e-06
 3: c_1_12_1270       cd8t  0.9999989 0.999998939 1.060938e-06
 4: c_1_12_1349       cd8t  0.9999814 0.999981384 1.861648e-05
 5: c_1_12_1394       cd8t  1.0000000 0.999999986 1.438074e-08
 6:  c_1_12_172       cd4t  0.9979380 0.002061995 9.979380e-01
 7:  c_1_12_228       cd4t  0.9955272 0.004472831 9.955272e-01
 8:  c_1_12_246       cd4t  0.9933210 0.006679029 9.933210e-01
 9:  c_1_12_285       cd4t  0.9986854 0.001314640 9.986854e-01
10:  c_1_12_511       cd4t  0.9911673 0.008832705 9.911673e-01</code></pre>
</div>
</div>
</section>
<section id="granular-hierarchical-subtyping-and-using-a-pipeline" class="level3" data-number="5.2.4">
<h3 data-number="5.2.4" class="anchored" data-anchor-id="granular-hierarchical-subtyping-and-using-a-pipeline"><span class="header-section-number">5.2.4</span> Granular hierarchical subtyping and using a <code>pipeline</code></h3>
<p>These posterior-probability scores can be passed directly to subtyping models for <code>cd4t</code> and <code>cd8t</code>, via the <code>prior_level_weights</code> arguments. For example, if I wanted to go ahead and cluster the <code>cd4t</code> cells into subtypes, I could follow the same process, passing in the prior probability of the cell being <code>cd4t</code> as a weight.</p>
<p>In the example below, the probability of <code>cd4t</code> is used as a weight in the NNLS regression fitting (akin to weighted least squares) when constructing metagene scores. Cells with high probability of <code>cd4t</code> are emphasized, while cells with lower probability of <code>cd4t</code> have less influence on the regression coefficients used to generate metagene scores.</p>
<p>In the clustering stage, probability of <code>cd4t</code> is again used as a weight (or prior probability), that factors into the expectation step of the EM algorithm. Letting <code>s(t)</code> denote the <em>subtype</em> of <code>cd4t</code>, then we have the below, where <span class="math inline">\(P(y_c = t)\)</span> would be the probability of <code>cd4t</code> from the previous stage of modeling.</p>
<p>The final returned probability for each <code>cd4t</code> subtype is a <em>joint</em> probability of both being the subtype <span class="math inline">\(s(t)\)</span> and the major celltype <span class="math inline">\(t\)</span>.</p>
<p><span class="math display">\[P(y_c = s(t) \cap y_c = t  | scores_c) = P(y_c = s(t) | scores_c, y_c = t)P(y_c = t)\]</span></p>
<p>Below is an example of this syntax using the <code>cd4t</code> subtypes in the package dataset “<code>markerslist_cd4tminor</code>”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(HieraType<span class="sc">::</span>markerslist_cd4tminor, <span class="at">max.level=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>List of 7
 $ cd4_naive:List of 3
  ..$ index_marker                               : chr [1:4] "TCF7" "LEF1" "SELL" "CCR7"
  ..$ predictors                                 : chr [1:30] "CD3D" "CD3E" "CD3G" "CD4" ...
  ..$ use_offclass_markers_as_negative_predictors: logi TRUE
 $ cd4_tem  :List of 3
  ..$ index_marker                               : chr [1:4] "CD44" "CCR6" "CXCR3" "IL7R"
  ..$ predictors                                 : chr [1:35] "CD3D" "CD3E" "CD3G" "CD4" ...
  ..$ use_offclass_markers_as_negative_predictors: logi TRUE
 $ cd4_tcm  :List of 3
  ..$ index_marker                               : chr [1:3] "CCR7" "IL7R" "SELL"
  ..$ predictors                                 : chr [1:27] "CD3D" "CD3E" "CD3G" "CD4" ...
  ..$ use_offclass_markers_as_negative_predictors: logi TRUE
 $ cd4_th1  :List of 3
  ..$ index_marker                               : chr [1:4] "TBX21" "IFNG" "CXCR3" "STAT4"
  ..$ predictors                                 : chr [1:30] "CD3D" "CD3E" "CD3G" "CD4" ...
  ..$ use_offclass_markers_as_negative_predictors: logi TRUE
 $ cd4_th2  :List of 3
  ..$ index_marker                               : chr [1:5] "GATA3" "IL4" "IL5" "IL13" ...
  ..$ predictors                                 : chr [1:30] "CD3D" "CD3E" "CD3G" "CD4" ...
  ..$ use_offclass_markers_as_negative_predictors: logi TRUE
 $ cd4_th17 :List of 3
  ..$ index_marker                               : chr [1:4] "RORC" "IL17A" "IL17F" "IL23R"
  ..$ predictors                                 : chr [1:28] "CD3D" "CD3E" "CD3G" "CD4" ...
  ..$ use_offclass_markers_as_negative_predictors: logi TRUE
 $ cd4_treg :List of 3
  ..$ index_marker                               : chr [1:4] "FOXP3" "IL2RA" "CTLA4" "TNFRSF18"
  ..$ predictors                                 : chr [1:29] "CD3D" "CD3E" "CD3G" "CD4" ...
  ..$ use_offclass_markers_as_negative_predictors: logi TRUE
 - attr(*, "class")= chr [1:2] "list" "markerslist"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Classify cd4t subtypes</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>metagenes_t4 <span class="ot">&lt;-</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  HieraType<span class="sc">::</span><span class="fu">fit_metagene_scores</span>(</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">markerslist =</span> HieraType<span class="sc">::</span>markerslist_cd4tminor</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">counts_matrix =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[,tcell_ids])</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">adjacency_matrix =</span> <span class="fu">as</span>(sem<span class="sc">@</span>graphs<span class="sc">$</span>umapgrph</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>                           ,<span class="st">"CsparseMatrix"</span>)[tcell_ids,tcell_ids]</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">prior_level_weights =</span> model_t<span class="sc">$</span>post_probs<span class="sc">$</span>cd4t <span class="do">## prior probability of cd4t</span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>model_t4 <span class="ot">&lt;-</span> </span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>HieraType<span class="sc">::</span><span class="fu">cluster_metagenes</span>(<span class="at">metagenes =</span> metagenes_t4, <span class="at">prior_prob_level =</span> model_t<span class="sc">$</span>post_probs<span class="sc">$</span>cd4t)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A convenient way to simplify this process with minimal code is to use a <code>pipeline</code> object. A custom hierarchical cell typing pipeline could be created like this below, where</p>
<ul>
<li><code>markerslists</code> is a named list, with a <code>markerslist</code> for each stage of cell typing. “tmajor” will classify <code>cd8t</code> vs.&nbsp;<code>cd4t</code>, while “t8minor” and “t4minor” will subclassify those cell types respectively.</li>
<li><code>priors</code>, a list specifying ‘parent’ celltyping stage. Here , <code>t4minor</code> and <code>t8minor</code> both need to run after “tmajor”, as they will use the estimated probabilities of “<code>cd4t</code>” and “<code>cd8t</code>” as inputs.<br>
“tmajor” runs first, and does not have prior probability inputs.</li>
<li><code>priors_category</code>, a list specifying ‘parent’ / major categories of each celltype. Here , <code>t4minor</code> is a subset of <code>"cd4t"</code>. “tmajor” runs first, and does not have a prior.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>pipeline_tcell <span class="ot">&lt;-</span> </span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">make_pipeline</span>(<span class="at">markerslists =</span> <span class="fu">list</span>(<span class="st">"tmajor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_tcellmajor</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>                    ,<span class="st">"t4minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_cd4tminor</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>                    ,<span class="st">"t8minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_cd8tminor</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>                    )</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>                 ,<span class="at">priors =</span> <span class="fu">list</span>(<span class="st">"t4minor"</span> <span class="ot">=</span> <span class="st">"tmajor"</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>                               ,<span class="st">"t8minor"</span> <span class="ot">=</span> <span class="st">"tmajor"</span>)</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>                 ,<span class="at">priors_category =</span> <span class="fu">list</span>(<span class="st">"t4minor"</span> <span class="ot">=</span> <span class="st">"cd4t"</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>                                         ,<span class="st">"t8minor"</span> <span class="ot">=</span> <span class="st">"cd8t"</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>                                         )</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(pipeline_tcell)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "list"     "pipeline"</code></pre>
</div>
</div>
<p>This particular pipeline is also saved as a package dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(HieraType<span class="sc">::</span>pipeline_tcell, <span class="at">max.level=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>List of 3
 $ markerslists   :List of 3
  ..$ tmajor :List of 2
  .. ..- attr(*, "class")= chr [1:2] "list" "markerslist"
  ..$ t4minor:List of 7
  .. ..- attr(*, "class")= chr [1:2] "list" "markerslist"
  ..$ t8minor:List of 5
  .. ..- attr(*, "class")= chr [1:2] "list" "markerslist"
 $ priors         :List of 2
  ..$ t4minor: chr "tmajor"
  ..$ t8minor: chr "tmajor"
 $ priors_category:List of 2
  ..$ t4minor: chr "cd4t"
  ..$ t8minor: chr "cd8t"
 - attr(*, "class")= chr [1:2] "list" "pipeline"</code></pre>
</div>
</div>
<p>The code to run the pipeline is shown below. Additional arguments to be passed to <code>fit_metagene_scores</code> or <code>cluster_metagenes</code> functions can be passed as additional arguments to <code>run_pipeline()</code> .</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Run a 'tcell'-typing pipeline, which </span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="do">### first classifies into T8/T4 categories (using the `tmajor` markerslist), </span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="do">### then runs subclassification for T4 and T8 using the t4minor and t8minor markerslists.</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>tcell_typing <span class="ot">&lt;-</span> </span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="fu">run_pipeline</span>(<span class="at">pipeline =</span> pipeline_tcell</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">counts_matrix =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[,tcell_ids])</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">adjacency_matrix =</span> <span class="fu">as</span>(sem<span class="sc">@</span>graphs<span class="sc">$</span>umapgrph</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>                           ,<span class="st">"CsparseMatrix"</span>)[tcell_ids,tcell_ids]</span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">celltype_call_threshold =</span> <span class="fl">0.5</span></span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">return_all_columns_postprobs =</span> <span class="cn">FALSE</span></span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We could take a look at individual <code>models</code> and <code>metagenes</code> objects if needed within our <code>tcell_typing</code> result. Below, we show that the outputted objects are the same in the pipeline format, as if we had ran them individually like in the method demonstration.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(model_t, tcell_typing<span class="sc">$</span>models<span class="sc">$</span>tmajor)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(metagenes_t, tcell_typing<span class="sc">$</span>metagene_scores<span class="sc">$</span>tmajor)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>By default, two possible cell type classifications are returned in the overall <code>post_probs</code> data.table object from our pipeline. Below, <code>celltype_granular</code> shows the most granular classification for each cell, carrying the ‘best possible’ classification at each stage of the pipeline. For example, the first cell below was first identified as <code>cd8t</code> being the best category. The best subcategory of <code>cd8t</code> was <code>cd8_exhausted</code>, and the posterior probability score for <code>cd8t_exhausted</code> was ~0.47.</p>
<p>We set the the cutoff <code>celltype_call_threshold=0.5</code> when we ran the pipeline, so that the <code>celltype_thresh</code> classification shows the most granular cell type with probability scores <em>at least</em> this threshold. The first cell below has <code>cd8_exhausted</code> score ~0.47 (less than 0.5), and so takes the higher level category of <code>cd8t</code>, which is above the threshold at ~0.99999.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>tcell_typing<span class="sc">$</span>post_probs[[<span class="st">"tmajor"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Key: &lt;cell_ID&gt;
        cell_ID celltype_thresh celltype_granular best_score_granular
         &lt;char&gt;          &lt;char&gt;            &lt;char&gt;               &lt;num&gt;
 1: c_1_12_1014            cd8t     cd8_exhausted           0.4691963
 2: c_1_12_1066            cd4t           cd4_th1           0.3352265
 3: c_1_12_1091         cd4_th2           cd4_th2           0.9683195
 4: c_1_12_1175   cd8_cytotoxic     cd8_cytotoxic           0.8777777
 5: c_1_12_1270         cd8_tem           cd8_tem           0.5882417
 6: c_1_12_1286        cd4_treg          cd4_treg           0.9909980
 7: c_1_12_1325            cd4t           cd4_tcm           0.4647576
 8: c_1_12_1346        cd4_treg          cd4_treg           0.8298557
 9: c_1_12_1349            cd8t           cd8_tem           0.3939835
10: c_1_12_1394   cd8_cytotoxic     cd8_cytotoxic           0.9587283
    best_score_thresh
                &lt;num&gt;
 1:         0.9999999
 2:         0.6751557
 3:         0.9683195
 4:         0.8777777
 5:         0.5882417
 6:         0.9909980
 7:         0.9984388
 8:         0.8298557
 9:         0.9999814
10:         0.9587283</code></pre>
</div>
</div>
<p>We can remake this table with a different cutoff in mind as needed, depending on how conservative / aggressive we are comfortable being with the cell type call, using the <code>combine_postprob_tables</code> function.</p>
<p>As demonstration, we can look at the output if we were to raise our threshold from 0.5 to 0.8.<br>
We can “<code>return_all_columns</code>” to show a more verbose result, with individual columns showing the joint probability scores for all possible cell type categories in each cell.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>HieraType<span class="sc">::</span><span class="fu">combine_postprob_tables</span>(<span class="at">pipeline =</span> pipeline_tcell</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>                                   ,<span class="at">models =</span> tcell_typing<span class="sc">$</span>models</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>                                   ,<span class="at">celltype_call_threshold =</span> <span class="fl">0.8</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>                                   ,<span class="at">return_all_columns =</span> <span class="cn">TRUE</span>)[[<span class="st">"tmajor"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>        cell_ID best_class_tmajor best_score_tmajor        cd8t         cd4t
         &lt;char&gt;            &lt;char&gt;             &lt;num&gt;       &lt;num&gt;        &lt;num&gt;
 1: c_1_12_1014              cd8t         0.9999999 0.999999904 9.619835e-08
 2: c_1_12_1066              cd4t         0.6751557 0.324844253 6.751557e-01
 3: c_1_12_1091              cd4t         0.9988244 0.001175604 9.988244e-01
 4: c_1_12_1175              cd8t         0.9999986 0.999998637 1.363123e-06
 5: c_1_12_1270              cd8t         0.9999989 0.999998939 1.060938e-06
 6: c_1_12_1286              cd4t         0.9981013 0.001898693 9.981013e-01
 7: c_1_12_1325              cd4t         0.9984388 0.001561222 9.984388e-01
 8: c_1_12_1346              cd4t         0.9982395 0.001760536 9.982395e-01
 9: c_1_12_1349              cd8t         0.9999814 0.999981384 1.861648e-05
10: c_1_12_1394              cd8t         1.0000000 0.999999986 1.438074e-08
    celltype_thresh celltype_granular best_score_granular best_score_thresh
             &lt;char&gt;            &lt;char&gt;               &lt;num&gt;             &lt;num&gt;
 1:            cd8t     cd8_exhausted           0.4691963         0.9999999
 2:          tmajor           cd4_th1           0.3352265         0.6751557
 3:         cd4_th2           cd4_th2           0.9683195         0.9683195
 4:   cd8_cytotoxic     cd8_cytotoxic           0.8777777         0.8777777
 5:            cd8t           cd8_tem           0.5882417         0.9999989
 6:        cd4_treg          cd4_treg           0.9909980         0.9909980
 7:            cd4t           cd4_tcm           0.4647576         0.9984388
 8:        cd4_treg          cd4_treg           0.8298557         0.8298557
 9:            cd8t           cd8_tem           0.3939835         0.9999814
10:   cd8_cytotoxic     cd8_cytotoxic           0.9587283         0.9587283
    best_class_t4minor best_score_t4minor    cd4_naive      cd4_tem
                &lt;char&gt;              &lt;num&gt;        &lt;num&gt;        &lt;num&gt;
 1:            cd4_tem       7.408648e-08 1.251116e-10 7.408648e-08
 2:            cd4_th1       3.352265e-01 7.836058e-05 2.443358e-01
 3:            cd4_th2       9.683195e-01 1.152001e-03 7.841427e-05
 4:           cd4_treg       1.262446e-06 3.588638e-09 1.405697e-11
 5:            cd4_th1       6.094740e-07 1.213471e-09 2.245479e-07
 6:           cd4_treg       9.909980e-01 1.289310e-06 8.622134e-08
 7:            cd4_tcm       4.647576e-01 1.677040e-01 1.226034e-02
 8:           cd4_treg       8.298557e-01 3.002300e-07 1.581774e-01
 9:           cd4_treg       1.092955e-05 4.486025e-09 8.873803e-09
10:           cd4_treg       6.985845e-09 1.273443e-10 5.957512e-09
         cd4_tcm      cd4_th1      cd4_th2     cd4_th17     cd4_treg
           &lt;num&gt;        &lt;num&gt;        &lt;num&gt;        &lt;num&gt;        &lt;num&gt;
 1: 1.195101e-11 2.129543e-08 4.029363e-15 7.197487e-13 6.786477e-10
 2: 1.867806e-04 3.352265e-01 6.040219e-03 1.158038e-02 7.770771e-02
 3: 1.148814e-02 9.695413e-03 9.683195e-01 2.286910e-11 8.090952e-03
 4: 5.892169e-10 6.060104e-08 3.049262e-08 5.391101e-09 1.262446e-06
 5: 9.412377e-10 6.094740e-07 7.486703e-10 1.940035e-07 3.000945e-08
 6: 1.653479e-08 1.018626e-08 7.094111e-03 7.843421e-06 9.909980e-01
 7: 4.647576e-01 7.313584e-04 2.508449e-01 9.533710e-02 6.803462e-03
 8: 1.020607e-02 6.923060e-14 1.491775e-13 6.882974e-12 8.298557e-01
 9: 6.009531e-08 8.601937e-07 6.614330e-06 1.389513e-07 1.092955e-05
10: 1.844918e-10 1.378047e-10 9.658852e-10 2.186144e-11 6.985845e-09
    best_class_t8minor best_score_t8minor    cd8_naive      cd8_tem
                &lt;char&gt;              &lt;num&gt;        &lt;num&gt;        &lt;num&gt;
 1:      cd8_exhausted       0.4691962886 6.011871e-05 1.860347e-01
 2:      cd8_cytotoxic       0.1649358419 3.833792e-06 1.311043e-01
 3:            cd8_tcm       0.0005912820 4.489162e-04 2.336174e-08
 4:      cd8_cytotoxic       0.8777776850 2.319425e-06 3.868444e-07
 5:            cd8_tem       0.5882416885 1.162076e-05 5.882417e-01
 6:      cd8_exhausted       0.0018765126 1.183680e-06 1.782040e-06
 7:          cd8_naive       0.0009770069 9.770069e-04 2.507600e-05
 8:      cd8_exhausted       0.0017514060 1.019562e-06 1.789781e-07
 9:            cd8_tem       0.3939834697 2.279564e-04 3.939835e-01
10:      cd8_cytotoxic       0.9587282579 5.255921e-06 3.696543e-02
         cd8_tcm cd8_cytotoxic cd8_exhausted
           &lt;num&gt;         &lt;num&gt;         &lt;num&gt;
 1: 2.514628e-12  3.447088e-01  4.691963e-01
 2: 1.210775e-11  1.649358e-01  2.880030e-02
 3: 5.912820e-04  1.352937e-04  8.823050e-08
 4: 9.247577e-09  8.777777e-01  1.222182e-01
 5: 1.619476e-14  2.342515e-01  1.774941e-01
 6: 1.656608e-07  1.904865e-05  1.876513e-03
 7: 3.768608e-04  5.270201e-05  1.295761e-04
 8: 1.001439e-07  7.830808e-06  1.751406e-03
 9: 1.169891e-06  2.354990e-01  3.702698e-01
10: 7.792592e-09  9.587283e-01  4.301033e-03</code></pre>
</div>
</div>
</section>
<section id="examining-and-validating-clustering-results" class="level3" data-number="5.2.5">
<h3 data-number="5.2.5" class="anchored" data-anchor-id="examining-and-validating-clustering-results"><span class="header-section-number">5.2.5</span> Examining and validating clustering results</h3>
<p>Now we can evaluate the cell typing performance. In particular, we want to check that the genes we specified as the index marker and predictors are in fact marker genes in the corresponding cell types.</p>
<p>First, we can look at a marker gene heatmap made with just the t-cells, and show the genes we specified as the most important ‘index genes’ for each of the <code>cd8t</code> and <code>cd4t</code> subtypes. Looking closely, we see that CD8A and CD8B are much higher in <code>cd8_</code> compared to <code>cd4_</code> types. Similarly, CD4 is higher expressed in <code>cd4_</code> subtypes compared to <code>cd8_</code> subtypes. Looking at the specific index markers for each subtype, we see good agreement between marker gene specified for each subtype reflected in our heatmap.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>fct <span class="ot">&lt;-</span> <span class="fu">clusterwise_foldchange_metrics</span>(<span class="at">counts =</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[,tcell_ids]</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>                                      ,<span class="at">metadata =</span> tcell_typing<span class="sc">$</span>post_probs[[<span class="st">"tmajor"</span>]]</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>                                      ,<span class="at">cluster_column =</span> <span class="st">"celltype_granular"</span>)</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>idxs <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">unname</span>(<span class="fu">c</span>(<span class="fu">unlist</span>(<span class="fu">lapply</span>(HieraType<span class="sc">::</span>markerslist_cd4tminor, <span class="st">"[["</span>, <span class="st">"index_marker"</span>))</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>               ,<span class="fu">unlist</span>(<span class="fu">lapply</span>(HieraType<span class="sc">::</span>markerslist_cd8tminor, <span class="st">"[["</span>, <span class="st">"index_marker"</span>)))))</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>nms <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"cd4t"</span>, <span class="fu">names</span>(HieraType<span class="sc">::</span>markerslist_cd4tminor), <span class="st">"cd8t"</span>, <span class="fu">names</span>(HieraType<span class="sc">::</span>markerslist_cd8tminor))</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>hmsubtype <span class="ot">&lt;-</span> <span class="fu">marker_heatmap</span>(fct, <span class="at">featsuse =</span> <span class="fu">c</span>(<span class="st">"CD8A"</span>, <span class="st">"CD8B"</span>, <span class="st">"CD4"</span>, idxs), <span class="at">clusterorder =</span> nms, <span class="at">orient_diagonal =</span> <span class="cn">TRUE</span>)</span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hmsubtype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="do">### The key index markers for cd4 subtypes</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(HieraType<span class="sc">::</span>markerslist_cd4tminor, <span class="st">"[["</span>, <span class="st">"index_marker"</span>)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="do">### The key index markers for cd8 subtypes</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(HieraType<span class="sc">::</span>markerslist_cd8tminor, <span class="st">"[["</span>, <span class="st">"index_marker"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$cd4_naive
[1] "TCF7" "LEF1" "SELL" "CCR7"

$cd4_tem
[1] "CD44"  "CCR6"  "CXCR3" "IL7R" 

$cd4_tcm
[1] "CCR7" "IL7R" "SELL"

$cd4_th1
[1] "TBX21" "IFNG"  "CXCR3" "STAT4"

$cd4_th2
[1] "GATA3" "IL4"   "IL5"   "IL13"  "CCR4" 

$cd4_th17
[1] "RORC"  "IL17A" "IL17F" "IL23R"

$cd4_treg
[1] "FOXP3"    "IL2RA"    "CTLA4"    "TNFRSF18"</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$cd8_naive
[1] "TCF7" "LEF1" "CCR7" "SELL" "IL7R"

$cd8_tem
[1] "GZMK"  "GZMA"  "CXCR3" "CD69"  "IL7R" 

$cd8_tcm
[1] "CCR7" "IL7R" "SELL"

$cd8_cytotoxic
[1] "GZMB"  "PRF1"  "GNLY"  "KLRD1" "IFNG" 

$cd8_exhausted
[1] "PDCD1"  "LAG3"   "TIGIT"  "HAVCR2" "TOX"   </code></pre>
</div>
</div>
<div class="wider-image">
<p><img src="./figures/hm_tsubtype.png" alt="Wide Image"></p>
</div>
<p>Here we can attach our T-cell annotations back onto our metadata, and remake our marker heatmap.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>met <span class="ot">&lt;-</span> </span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">merge</span>(<span class="fu">data.table</span>(sem<span class="sc">@</span>meta.data), tcell_typing<span class="sc">$</span>post_probs[[<span class="st">"tmajor"</span>]][,.(cell_ID, <span class="at">celltype =</span> celltype_granular, best_score_granular)]</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>      ,<span class="at">by=</span><span class="st">"cell_ID"</span>, <span class="at">all.x=</span><span class="cn">TRUE</span>)</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>met[<span class="fu">is.na</span>(celltype),celltype<span class="sc">:</span><span class="er">=</span>clusters_unsup]</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>fc_t_all <span class="ot">&lt;-</span> </span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>HieraType<span class="sc">::</span><span class="fu">clusterwise_foldchange_metrics</span>(<span class="at">normed =</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>data</span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>                   ,<span class="at">metadata =</span> met</span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>                   ,<span class="at">cluster_column =</span> <span class="st">"celltype"</span></span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a>                   )</span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a>hm_tcell <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">marker_heatmap</span>(fc_t_all</span>
<span id="cb59-13"><a href="#cb59-13" aria-hidden="true" tabindex="-1"></a>                           ,<span class="at">extras =</span> <span class="fu">c</span>(<span class="st">"CD4"</span>, <span class="st">"CD8A"</span>, <span class="st">"CD8B"</span>)</span>
<span id="cb59-14"><a href="#cb59-14" aria-hidden="true" tabindex="-1"></a>                          )</span>
<span id="cb59-15"><a href="#cb59-15" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hm_tcell)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="wider-image">
<p><img src="./figures/hm_tcell_annotated.png" alt="Wide Image"></p>
</div>
<p>Treating the ‘posterior probability’ as a confidence that the cell belongs to the corresponding cluster, we can show that the proportion of cells that express the expected marker genes generally increases with higher confidence.</p>
<p>In the figure below, we compare the major <code>cd4t</code> and <code>cd8t</code> T cell types in terms of some of their marker genes which are specific to each. We see that CD8A and CD8B increase in <code>cd8t</code> cells as we gain confidence in the cell type call, along with other <code>cd8t</code> specific markers (GZMB, GNLY, PRF1).<br>
Similarly, we see that CD4 increases in <code>cd4t</code> calls with higher posterior probability scores, along with <code>cd4t</code> subtype-specific markers like FOXP3, CTLA4, TIGIT (<code>cd4_treg</code> marker genes), IL17A and IL17F (<code>cd4_th17</code> markers) and GATA3 (<code>cd4_th2</code> marker).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> </span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="fu">marker_threshold_plot</span>(<span class="at">normed =</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>data</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>                      ,<span class="at">metadata =</span> tcell_typing<span class="sc">$</span>models<span class="sc">$</span>tmajor<span class="sc">$</span>post_probs</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>                      ,<span class="at">clusters =</span> <span class="fu">c</span>(<span class="st">"cd4t"</span>, <span class="st">"cd8t"</span>)</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>                      ,<span class="at">markers =</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"CD4"</span>, <span class="st">"IL17A"</span>, <span class="st">"IL17F"</span>, <span class="st">"GATA3"</span>, <span class="st">"FOXP3"</span>,  <span class="st">"IL2RA"</span>, <span class="st">"CTLA4"</span>, <span class="st">"TIGIT"</span>,<span class="st">"CD8A"</span>, <span class="st">"CD8B"</span>, <span class="st">"GNLY"</span>, <span class="st">"GZMB"</span>,<span class="st">"PRF1"</span>)</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>                      ,<span class="at">thresholds =</span> <span class="fu">c</span>(<span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.7</span>, <span class="fl">0.9</span>, <span class="fl">0.99</span>, <span class="fl">0.995</span>)</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>                      ,<span class="at">score_column =</span> <span class="st">"best_score"</span></span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>                      ,<span class="at">cluster_column =</span> <span class="st">"best_class"</span></span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>                      )</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>,<span class="at">face=</span><span class="st">'bold'</span>)) <span class="sc">+</span> </span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Proportion of cells expressing marker gene increases with </span><span class="sc">\n</span><span class="st">posterior probability confidence scores"</span>)</span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="wider-image">
<p><img src="./figures/markerbarplot_tcells_major.png" alt="Wide Image"></p>
</div>
<p>Here’s another look at the pairsplot for our metagene scores,this time coloring the cells based on their classification.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">metagene_pairsplot</span>(metagenes_t, <span class="st">"yhat"</span>, <span class="at">obs =</span> model_t<span class="sc">$</span>post_probs[,.(cell_ID, best_class)], <span class="at">colorvar =</span> <span class="st">"best_class"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="wider-image">
<p><img src="./figures/pairshat_colored.png" alt="Wide Image"></p>
</div>
</section>
</section>
</section>
<section id="ex2" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Example 2: Broad immune cell classification</h1>
<p>For the second example, we wish to ignore the unsupervised clustering and infer all cell types based entirely on the framework motivated here.</p>
<p>Using the full dataset, we can create a pipeline to first infer amongst 5 major cell type categories [<code>immune</code>, <code>epithelial</code>, <code>endothelial</code>, <code>fibroblast</code>, <code>plasma</code>] in a similar fashion as described in our first example.<br>
Then, we’ll infer subtypes of the immune cell labels with categories including [<code>tcell</code>,<code>bcell</code>,<code>macrophage</code>,<code>neutrophil</code>,<code>mast</code>,<code>monocyte</code>,<code>cDC</code>,<code>pDC</code>]. Finally, we’ll do some basic subclassification of the <code>tcells</code> into [<code>cd4t</code>,<code>cd8t</code>].</p>
<section id="tldr2" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="tldr2"><span class="header-section-number">6.1</span> ‘Just The Code’</h2>
<p>Here’s the code to produce the set of cell types. The next section will cover the details and discuss a few choices that were made.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Pipeline, with </span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="do">### level-1 ('l1') being major categorization </span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="do">### level-2 ('l2') being classification of major immune cell categories</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="do">### level-t ('lt') being T8/T4 classification</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>pipeline_io <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">make_pipeline</span>(</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">markerslists =</span> <span class="fu">list</span>(<span class="st">"l1"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_l1</span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>                      ,<span class="st">"l2"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_immunemajor</span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>                      ,<span class="st">"lt"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_tcellmajor)</span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors =</span> <span class="fu">list</span>(<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"l2"</span></span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a>                 ,<span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"l1"</span>) </span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors_category =</span> <span class="fu">list</span>(<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"tcell"</span></span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a>                          ,<span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"immune"</span>)</span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="do">### run my pipeline (using all cells)</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>immune_typing <span class="ot">&lt;-</span> </span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="fu">run_pipeline</span>(<span class="at">pipeline =</span> pipeline_io</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">counts_matrix =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">adjacency_matrix =</span> <span class="fu">as</span>(sem<span class="sc">@</span>graphs<span class="sc">$</span>umapgrph</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>                                    ,<span class="st">"CsparseMatrix"</span>)</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">celltype_call_threshold =</span> <span class="fl">0.5</span></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">return_all_columns_postprobs =</span> <span class="cn">FALSE</span></span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>             )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="evaluation" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="evaluation"><span class="header-section-number">6.2</span> Evaluation</h2>
<p>Like previously, this <code>pipeline</code>, and the corresponding <code>markerslists</code> are saved as package datasets.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(HieraType<span class="sc">::</span>pipeline_io,<span class="at">max.level=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>List of 3
 $ markerslists   :List of 3
  ..$ l1:List of 6
  .. ..$ epithelial   :List of 3
  .. .. ..$ index_marker                               : chr [1:4] "EPCAM" "KRT18" "KRT19" "CDH1"
  .. .. ..$ predictors                                 : chr [1:167] "EPCAM" "KRT18" "KRT8" "KRT19" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ endothelial  :List of 3
  .. .. ..$ index_marker                               : chr [1:3] "PECAM1" "VWF" "CDH5"
  .. .. ..$ predictors                                 : chr [1:129] "PECAM1" "VWF" "CDH5" "CLDN5" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ fibroblast   :List of 3
  .. .. ..$ index_marker                               : chr [1:3] "COL1A1" "DCN" "PDGFRA"
  .. .. ..$ predictors                                 : chr [1:131] "COL1A2" "VIM" "S100A4" "PDGFRB" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ plasma       :List of 3
  .. .. ..$ index_marker                               : chr [1:5] "IGHG1" "IGHG1/2" "IGHA1" "MZB1" ...
  .. .. ..$ predictors                                 : chr [1:29] "IGHG1" "IGHG1/2" "IGHA1" "MZB1" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ immune       :List of 3
  .. .. ..$ index_marker                               : chr [1:6] "PTPRC" "ITGB2" "SPI1" "CD52" ...
  .. .. ..$ predictors                                 : chr [1:142] "PTPRC" "CD3E" "CD3D" "CD4" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ smooth_muscle:List of 3
  .. .. ..$ index_marker                               : chr [1:4] "MYH11" "ACTA2" "TAGLN" "DES"
  .. .. ..$ predictors                                 : chr [1:64] "MYH11" "ACTA2" "TAGLN" "DES" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..- attr(*, "class")= chr [1:2] "list" "markerslist"
  ..$ l2:List of 7
  .. ..$ tcell     :List of 3
  .. .. ..$ index_marker                               : chr [1:6] "CD3D" "CD3G" "CD3E" "IL7R" ...
  .. .. ..$ predictors                                 : chr [1:117] "CD3D" "CD3E" "CD3G" "CD8A" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ bcell     :List of 3
  .. .. ..$ index_marker                               : chr [1:4] "CD19" "MS4A1" "CD79A" "CD79B"
  .. .. ..$ predictors                                 : chr [1:84] "CD19" "MS4A1" "CD79A" "CD79B" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ macrophage:List of 3
  .. .. ..$ index_marker                               : chr [1:4] "CD163" "CD68" "MARCO" "CSF1R"
  .. .. ..$ predictors                                 : chr [1:100] "CD68" "CD163" "MARCO" "CSF1R" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ neutrophil:List of 3
  .. .. ..$ index_marker                               : chr [1:4] "MPO" "FCGR3B" "S100A8" "S100A9"
  .. .. ..$ predictors                                 : chr [1:100] "MPO" "FCGR3B" "S100A8" "S100A9" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ mast      :List of 3
  .. .. ..$ index_marker                               : chr [1:5] "KIT" "TPSAB1" "TPSB2" "TPSAB1/2" ...
  .. .. ..$ predictors                                 : chr [1:92] "KIT" "TPSAB1" "CPA3" "TPSB2" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ monocyte  :List of 3
  .. .. ..$ index_marker                               : chr [1:4] "CD14" "FCGR3A" "CCR2" "LYZ"
  .. .. ..$ predictors                                 : chr [1:130] "CD14" "FCGR3A" "CCR2" "LYZ" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ dendritic :List of 3
  .. .. ..$ index_marker                               : chr [1:5] "ITGAX" "CLEC9A" "FLT3" "XCR1" ...
  .. .. ..$ predictors                                 : chr [1:103] "ITGAX" "CLEC9A" "FLT3" "XCR1" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..- attr(*, "class")= chr [1:2] "list" "markerslist"
  ..$ lt:List of 2
  .. ..$ cd8t:List of 3
  .. .. ..$ index_marker                               : chr [1:2] "CD8B" "CD8A"
  .. .. ..$ predictors                                 : chr [1:48] "CD3D" "CD3E" "CD3G" "CD8A" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..$ cd4t:List of 3
  .. .. ..$ index_marker                               : chr "CD4"
  .. .. ..$ predictors                                 : chr [1:48] "CD3D" "CD3E" "CD3G" "CD4" ...
  .. .. ..$ use_offclass_markers_as_negative_predictors: logi TRUE
  .. ..- attr(*, "class")= chr [1:2] "list" "markerslist"
 $ priors         :List of 2
  ..$ lt: chr "l2"
  ..$ l2: chr "l1"
 $ priors_category:List of 2
  ..$ lt: chr "tcell"
  ..$ l2: chr "immune"
 - attr(*, "class")= chr [1:2] "list" "pipeline"</code></pre>
</div>
</div>
<p>We can look at a heatmap showing the marker genes for our cell types to check whether the algorithm produced sensible results. Because I have a lot of cell type categories, and some of them are not totally exclusive (for example, macrophage/monocyte are very similar and share many markers), I don’t necessarily expect all my cells to have very high confidence scores for every cell type; I’m okay here with taking the most granular cell type calls and running with them.</p>
<p>These clusters show more granularity in a first pass than my original set of unsupervised clusters. The heatmap below also shows very sane-looking results, with marker genes corresponding to many of the genes I specified in <code>markerslists</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>fc_l1 <span class="ot">&lt;-</span> </span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">clusterwise_foldchange_metrics</span>(<span class="at">normed =</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>data</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>                     ,<span class="at">metadata =</span>immune_typing<span class="sc">$</span>post_probs[[<span class="dv">1</span>]]</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>                     ,<span class="at">cluster_column =</span> <span class="st">"celltype_granular"</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>hm_l1 <span class="ot">&lt;-</span> <span class="fu">marker_heatmap</span>(fc_l1, <span class="at">extras =</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8A"</span>, <span class="st">"CD8B"</span>, <span class="st">"FOXP3"</span>))</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hm_l1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="wider-image">
<p><img src="./figures/hm_l1.png" alt="Wide Image"></p>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(HieraType<span class="sc">::</span>markerslist_immunemajor, <span class="st">"[["</span>, <span class="st">"index_marker"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$tcell
[1] "CD3D"  "CD3G"  "CD3E"  "IL7R"  "TRBC1" "CD2"  

$bcell
[1] "CD19"  "MS4A1" "CD79A" "CD79B"

$macrophage
[1] "CD163" "CD68"  "MARCO" "CSF1R"

$neutrophil
[1] "MPO"    "FCGR3B" "S100A8" "S100A9"

$mast
[1] "KIT"      "TPSAB1"   "TPSB2"    "TPSAB1/2" "CPA3"    

$monocyte
[1] "CD14"   "FCGR3A" "CCR2"   "LYZ"   

$dendritic
[1] "ITGAX"  "CLEC9A" "FLT3"   "XCR1"   "CD1C"  </code></pre>
</div>
</div>
<p>Here’s a spatial plot of the cell types we’ve called so far. We could stop here, or continue in the same manner for subclassifying any other cell types.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>met <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="fu">data.table</span>(sem<span class="sc">@</span>meta.data)</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>             ,immune_typing<span class="sc">$</span>post_probs<span class="sc">$</span>l1[,.(cell_ID, <span class="at">celltype_supervised =</span> celltype_granular)]</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">by=</span><span class="st">"cell_ID"</span>)</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>colorpal <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'#F0A0FF'</span>,<span class="st">'#E0FF66'</span>,<span class="st">'#C20088'</span>,<span class="st">'#00998F'</span>,<span class="st">'#426600'</span></span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>              , <span class="st">"#4C005C"</span> </span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>              ,<span class="st">'#191919'</span></span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>              ,<span class="st">'#003380'</span>, <span class="st">'#005C31'</span>,<span class="st">'#94FFB5'</span>,<span class="st">'#990000'</span>,<span class="st">'#0075DC'</span>,<span class="st">'#FFA405'</span>)</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(colorpal) <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(met[[<span class="st">"celltype_supervised"</span>]]))</span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>xyplist_sup <span class="ot">&lt;-</span> </span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(<span class="fu">split</span>(met, <span class="at">by=</span><span class="st">"portion"</span>), <span class="cf">function</span>(xx){</span>
<span id="cb69-13"><a href="#cb69-13" aria-hidden="true" tabindex="-1"></a>    HieraType<span class="sc">:::</span><span class="fu">xyplot</span>(<span class="at">cluster_column =</span> <span class="st">"celltype_supervised"</span>, <span class="at">metadata =</span> xx</span>
<span id="cb69-14"><a href="#cb69-14" aria-hidden="true" tabindex="-1"></a>                       , <span class="at">ptsize =</span> <span class="fl">0.1</span>, <span class="at">cls =</span> colorpal</span>
<span id="cb69-15"><a href="#cb69-15" aria-hidden="true" tabindex="-1"></a>                       <span class="co">#,clusters = c("cd4t", "cd8t", "bcell")</span></span>
<span id="cb69-16"><a href="#cb69-16" aria-hidden="true" tabindex="-1"></a><span class="co">#                       , ptsize = 0.01, cls = colorpal</span></span>
<span id="cb69-17"><a href="#cb69-17" aria-hidden="true" tabindex="-1"></a>                       ,<span class="at">plotfirst =</span> <span class="fu">c</span>(<span class="st">"epithelial"</span>, <span class="st">"endothelial"</span>, <span class="st">"fibroblast"</span>, <span class="st">"smooth_muscle"</span>, <span class="st">"plasma"</span>)</span>
<span id="cb69-18"><a href="#cb69-18" aria-hidden="true" tabindex="-1"></a>                       ) <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb69-19"><a href="#cb69-19" aria-hidden="true" tabindex="-1"></a>  })  </span>
<span id="cb69-20"><a href="#cb69-20" aria-hidden="true" tabindex="-1"></a>xyplot_sup <span class="ot">&lt;-</span> </span>
<span id="cb69-21"><a href="#cb69-21" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> xyplist_sup,<span class="at">nrow =</span> <span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb69-22"><a href="#cb69-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">'white'</span>,<span class="at">color=</span><span class="st">'white'</span>)) <span class="sc">+</span> </span>
<span id="cb69-23"><a href="#cb69-23" aria-hidden="true" tabindex="-1"></a>  cowplot<span class="sc">::</span><span class="fu">panel_border</span>(<span class="at">remove=</span><span class="cn">TRUE</span>) </span>
<span id="cb69-24"><a href="#cb69-24" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(xyplot_sup)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="wider-image">
<p><img src="./figures/xyplot_sup.png" alt="Wide Image"></p>
</div>
<p>Below, we notice several interesting things about the spatial distribution of <code>cd8t</code>, <code>cd4t</code>, and <code>bcell</code> types. The <code>cd4t</code> cells tend to show higher colocalization with <code>bcell</code>s in tertiary lymphoid type structures, while <code>cd8t</code> spatial distribution are more dispersed throughout the tissue. This reflects the role of <code>cd8t</code> cells in seeking out and killing tumor cells, compared to the role of <code>cd4t</code> cells in local antigen presentation.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>xyplist_lymph <span class="ot">&lt;-</span> </span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(<span class="fu">split</span>(met, <span class="at">by=</span><span class="st">"portion"</span>), <span class="cf">function</span>(xx){</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>    HieraType<span class="sc">:::</span><span class="fu">xyplot</span>(<span class="at">cluster_column =</span> <span class="st">"celltype_supervised"</span>, <span class="at">metadata =</span> xx</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>                       , <span class="at">ptsize =</span> <span class="fl">0.05</span>, <span class="at">cls =</span> colorpal</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>                       ,<span class="at">clusters =</span> <span class="fu">c</span>(<span class="st">"cd4t"</span>, <span class="st">"cd8t"</span>, <span class="st">"bcell"</span>)</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>                       ) <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>lymphplot <span class="ot">&lt;-</span> </span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a>xyplist_lymph[[<span class="dv">1</span>]]  <span class="sc">+</span> </span>
<span id="cb70-12"><a href="#cb70-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>celltype_supervised) <span class="sc">+</span> ggdark<span class="sc">::</span><span class="fu">dark_theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>,<span class="at">face=</span><span class="st">'bold'</span>))</span>
<span id="cb70-13"><a href="#cb70-13" aria-hidden="true" tabindex="-1"></a>,xyplist_lymph[[<span class="dv">2</span>]]  <span class="sc">+</span> </span>
<span id="cb70-14"><a href="#cb70-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>celltype_supervised) <span class="sc">+</span> ggdark<span class="sc">::</span><span class="fu">dark_theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>,<span class="at">face=</span><span class="st">'bold'</span>))</span>
<span id="cb70-15"><a href="#cb70-15" aria-hidden="true" tabindex="-1"></a>,<span class="at">nrow=</span><span class="dv">2</span></span>
<span id="cb70-16"><a href="#cb70-16" aria-hidden="true" tabindex="-1"></a>,<span class="at">rel_heights  =</span> <span class="fu">c</span>(<span class="fl">1.3</span>,<span class="dv">1</span>)</span>
<span id="cb70-17"><a href="#cb70-17" aria-hidden="true" tabindex="-1"></a>)  <span class="sc">+</span> </span>
<span id="cb70-18"><a href="#cb70-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">'black'</span>,<span class="at">color=</span><span class="st">'black'</span>)) <span class="sc">+</span> </span>
<span id="cb70-19"><a href="#cb70-19" aria-hidden="true" tabindex="-1"></a>  cowplot<span class="sc">::</span><span class="fu">panel_border</span>(<span class="at">remove=</span><span class="cn">TRUE</span>) </span>
<span id="cb70-20"><a href="#cb70-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-21"><a href="#cb70-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-22"><a href="#cb70-22" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(lymphplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="wider-image">
<p><img src="./figures/lymphxyplot.png" alt="Wide Image"></p>
</div>
</section>
</section>
<section id="conclusion" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Conclusion</h1>
<p>We’ve discussed a method for cell typing that uses predefined sets of marker genes to build cell type specific metagene scores, which are probabilistically clustered into cell types.</p>
<p>Using a predefined set of marker genes has several advantages. For one, this allows users to prespecify the network of genes which are expected to have positive covariance within the cell type, and restricts any undesired influence of non-informative or noisy genes. Because no assumption is made regarding the average expression or level of covariance between any pair of genes, we suspect that in some cases this approach may be more flexible and robust in the presence of batch effects or platform effects compared to using pre-trained models, data transfer, or reference profiles which may have been trained using different sample conditions or technologies.</p>
<p>The NNLS regression framework for estimating celltype metagenes allows for improved prediction power by incorporating information from similar cells through an adjacency network. By restricting the direction of the regression coefficients, we’ve implicitly imposed an assumption of positive covariance between these markers within the corresponding cell type class. Sign constraints have been shown to be as effective in some cases as explicit regularization <span class="citation" data-cites="nnlsref">(<a href="#ref-nnlsref" role="doc-biblioref">Slawski and Hein 2013</a>)</span>.</p>
<p>We use a mixture of overfitted multivariate gaussian mixture models for probabilistic cell type classification. This approach is effectively a non-parametric mixture model, where we assume that the distribution of metagene scores for each cell type can be approximated by a ‘large-enough’ mixture of multivariate gaussians <span class="citation" data-cites="aragam2020identifiability">(<a href="#ref-aragam2020identifiability" role="doc-biblioref">Aragam et al. 2020</a>)</span>. The simple selection of anchor cells based on score positivity, and ability to fine tune cell typing by controlling behavior of double positivity should allow for easy user control and flexibility.</p>
<p>While we envision the method to be useful for many cell typing applications, there are several caveats and limitations users should keep in mind. One primary assumption is that the user has a sense of the tissue type they are working with and the cell types that are likely to be present. If the user were to run a “tcell typing pipeline” on myeloid cells, they may see unexpected or poor results. Similarly, the “io pipeline” we used in Example 2, may not perform well in a very different tissue type, like brain for example. If we wanted to apply this method to a tissue type like ‘brain’, we may need to generate a new pipeline, with <code>markerslists</code> for brain specific cell types like <code>neurons</code> and <code>oligodendrocytes</code>. For cases where a major cell type exists in a dataset but does not have a <code>markerslists</code> to be modeled, that cell type could manifest as cells which have “low posterior probability” for all of the other cell type classes included in the pipeline.</p>
<p>There are several future improvements for this method we plan to work on. Simultaneous estimation of metagene scores and clustering (perhaps through a regression-based mixture model) might enhance the power of metagene scores and classification by optimizing the cell allocations to the correct cluster simultaneously during model-fitting. Exploring other frameworks outside of NNLS regression could potentially capture more flexible model structures with interaction and non-linearity.<br>
The ability or freedom to learn additional marker genes from the data, and an extension to a semi-supervised framework where cell types where marker genes are unknown a priori would also be useful.</p>
<p>We’ve implemented the tools presented here in the R package <a href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/HieraType">HieraType</a>.</p>
</section>
<section id="quick-summary-of-package-functions-used-in-this-post" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> Quick summary of package functions used in this post</h1>
<p>Below we call out a short list of <code>HieraType</code> package functions which were used throughout this post. Please see the examples throughout the rest of this document for more context on their usage.</p>
<p><em>Model Fitting and Clustering</em></p>
<ul>
<li><p><code>make_markerlist()</code> - Used to create a custom list of cell type markers.</p></li>
<li><p><code>make_pipeline()</code> - Used to create a pipeline, or recipe for hierarchical cell typing.</p></li>
<li><p><code>run_pipeline()</code> - Used to run a hierarchical cell typing pipeline.</p></li>
<li><p><code>fit_metagene_scores()</code> - This function fits cell type metagene scores. Called as a standalone function or automatically within <code>run_pipeline</code>,</p></li>
<li><p><code>cluster_metagenes()</code> - This function clusters cell type metagene scores using overfitted-gaussian mixture models. Also called within <code>run_pipeline</code>.</p></li>
</ul>
<p><em>Plotting and Diagnostics</em></p>
<ul>
<li><p><code>xyplot()</code> - A utility function for plotting cell types in x/y space.</p></li>
<li><p><code>metagene_pairsplot()</code> - Plot pairwise distributions of metagene scores.</p></li>
<li><p><code>metagene_coefficients_plot()</code> - Plot for which genes contribute most to cell type metagene scores.</p></li>
<li><p><code>marker_threshold_plot()</code> - Model diagnostics plot, show specific marker genes and the proportion of cells that express them by cluster, for increasing posterior probability thresholds.</p></li>
<li><p><code>clusterwise_foldchange_metrics()</code> - Computes the fold change, average expression, proportion of cells with at least 1 count for each gene and each cluster. This is the data used as input to a marker gene heatmap.</p></li>
<li><p><code>marker_heatmap()</code> - Makes a bubble-style heatmap plot (clusters x marker genes) showing the proportion of cells with at least one count (bubble size) and relative average expression (bubble color). Easy to add and control the genes shown in the heatmap, and can specify a fixed order for comparing multiple heatmaps.</p></li>
</ul>


<!-- -->


</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-aragam2020identifiability" class="csl-entry" role="listitem">
Aragam, Bryon, Chen Dan, Eric P. Xing, and Pradeep Ravikumar. 2020. <span>“<span class="nocase">Identifiability of nonparametric mixture models and Bayes optimal clustering</span>.”</span> <em>The Annals of Statistics</em> 48 (4): 2277–2302. <a href="https://doi.org/10.1214/19-AOS1887">https://doi.org/10.1214/19-AOS1887</a>.
</div>
<div id="ref-nonnegpca" class="csl-entry" role="listitem">
Sigg, Christian D., and Joachim M. Buhmann. 2008. <span>“Expectation-Maximization for Sparse and Non-Negative PCA.”</span> In <em>Proceedings of the 25th International Conference on Machine Learning</em>, 960–67. ICML ’08. New York, NY, USA: Association for Computing Machinery. <a href="https://doi.org/10.1145/1390156.1390277">https://doi.org/10.1145/1390156.1390277</a>.
</div>
<div id="ref-nnlsref" class="csl-entry" role="listitem">
Slawski, Martin, and Matthias Hein. 2013. <span>“<span class="nocase">Non-negative least squares for high-dimensional linear models: Consistency and sparse recovery without regularization</span>.”</span> <em>Electronic Journal of Statistics</em> 7 (none): 3004–56. <a href="https://doi.org/10.1214/13-EJS868">https://doi.org/10.1214/13-EJS868</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/nanostring-biostats\.github\.io\/CosMx-Analysis-Scratch-Space");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb71" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Using marker gene driven metagene scores for granular hierarchical cell typing."</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a><span class="co">  - name: Dan McGuire</span></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a><span class="co">    orcid: 0009-0006-0286-9625</span></span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a><span class="co">    affiliations:</span></span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a><span class="co">      - ref: nstg</span></span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a><span class="co">      - ref: dan11mcguire</span></span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a><span class="an">toc:</span><span class="co"> true</span></span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a><span class="an">toc-title:</span><span class="co"> Contents</span></span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a><span class="an">toc-depth:</span><span class="co"> 3</span></span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a><span class="an">toc-expand:</span><span class="co"> 2</span></span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a><span class="an">toc-location:</span><span class="co"> left</span></span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a><span class="an">number-sections:</span><span class="co"> true</span></span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a><span class="an">number-depth:</span><span class="co"> 4</span></span>
<span id="cb71-16"><a href="#cb71-16" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "2025-06-20"</span></span>
<span id="cb71-17"><a href="#cb71-17" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span></span>
<span id="cb71-18"><a href="#cb71-18" aria-hidden="true" tabindex="-1"></a><span class="co">  - cell typing</span></span>
<span id="cb71-19"><a href="#cb71-19" aria-hidden="true" tabindex="-1"></a><span class="an">image:</span><span class="co"> figures/hm_tsubtype.png</span></span>
<span id="cb71-20"><a href="#cb71-20" aria-hidden="true" tabindex="-1"></a><span class="an">description:</span><span class="co"> "Cell typing using metagenes and smoothing."</span></span>
<span id="cb71-21"><a href="#cb71-21" aria-hidden="true" tabindex="-1"></a><span class="an">draft:</span><span class="co"> false</span></span>
<span id="cb71-22"><a href="#cb71-22" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb71-23"><a href="#cb71-23" aria-hidden="true" tabindex="-1"></a><span class="co">  html: </span></span>
<span id="cb71-24"><a href="#cb71-24" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb71-25"><a href="#cb71-25" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb71-26"><a href="#cb71-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-27"><a href="#cb71-27" aria-hidden="true" tabindex="-1"></a><span class="fu"># Abstract</span></span>
<span id="cb71-28"><a href="#cb71-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-29"><a href="#cb71-29" aria-hidden="true" tabindex="-1"></a>This post showcases a hierarchical cell typing method (HieraType) optimized for detection and sub-clustering of immune cells.</span>
<span id="cb71-30"><a href="#cb71-30" aria-hidden="true" tabindex="-1"></a>HieraType:</span>
<span id="cb71-31"><a href="#cb71-31" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb71-32"><a href="#cb71-32" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Is organized around canonical marker genes, but harnesses cells' complete expression profiles.</span>
<span id="cb71-33"><a href="#cb71-33" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Can be adapted to any arbitrary cell type hierarchy.</span>
<span id="cb71-34"><a href="#cb71-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-35"><a href="#cb71-35" aria-hidden="true" tabindex="-1"></a>Code for the method and utility functions used in this post can be found in the R package <span class="co">[</span><span class="ot">HieraType</span><span class="co">](https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/HieraType)</span>.  The package contains carefully designed pipelines for:</span>
<span id="cb71-36"><a href="#cb71-36" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb71-37"><a href="#cb71-37" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Subclassifying T-cells into 12 categories</span>
<span id="cb71-38"><a href="#cb71-38" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Classifying a whole dataset into broad categories (epithelial, fibroblast, immune, endothelial), then subclassifying immune cells into 9 categories</span>
<span id="cb71-39"><a href="#cb71-39" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Classifying native brain cells</span>
<span id="cb71-40"><a href="#cb71-40" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb71-41"><a href="#cb71-41" aria-hidden="true" tabindex="-1"></a>It is also possible to assemble pipelines for custom applications, e.g. focused on a specific tissue type.</span>
<span id="cb71-42"><a href="#cb71-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-43"><a href="#cb71-43" aria-hidden="true" tabindex="-1"></a>In this post, we use a 6K-plex colon-cancer dataset to demonstrate two common use cases for the method:</span>
<span id="cb71-44"><a href="#cb71-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-45"><a href="#cb71-45" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Subtyping T cells</span>
<span id="cb71-46"><a href="#cb71-46" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Broad immune cell typing.</span>
<span id="cb71-47"><a href="#cb71-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-48"><a href="#cb71-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-49"><a href="#cb71-49" aria-hidden="true" tabindex="-1"></a><span class="fu"># Quick links to the code</span></span>
<span id="cb71-50"><a href="#cb71-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-51"><a href="#cb71-51" aria-hidden="true" tabindex="-1"></a>For those in a hurry to 'just see the code', they can jump ahead to the "Just The Code" sections for <span class="co">[</span><span class="ot">Example 1</span><span class="co">](#tldr)</span>, or <span class="co">[</span><span class="ot">Example 2</span><span class="co">](#tldr2)</span>.</span>
<span id="cb71-52"><a href="#cb71-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-53"><a href="#cb71-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-54"><a href="#cb71-54" aria-hidden="true" tabindex="-1"></a><span class="fu"># Introduction </span></span>
<span id="cb71-55"><a href="#cb71-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-56"><a href="#cb71-56" aria-hidden="true" tabindex="-1"></a>Often (particularly with immune cells), we may want our cell type calls to be guided by specific sets of marker genes.  </span>
<span id="cb71-57"><a href="#cb71-57" aria-hidden="true" tabindex="-1"></a>In some spatially-resolved transcriptomics (SRT) datasets, directly using marker gene expression for determining cell type may be a challenge due to data sparsity, background, or segmentation ambiguity.  These challenges together can make marker genes noisy individual data points. </span>
<span id="cb71-58"><a href="#cb71-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-59"><a href="#cb71-59" aria-hidden="true" tabindex="-1"></a>However, marker genes remain a strong organizing principle for differentiating cell types. To overcome the intrinsic noise of single marker genes, the cell typing method we describe here captures information from both (a) similar cells and (b) biologically relevant genes, in order to construct metagenes that more stably estimate true marker expression. These marker metagene scores are then used for probabilistic cell type assignment.</span>
<span id="cb71-60"><a href="#cb71-60" aria-hidden="true" tabindex="-1"></a>The method takes as input predefined 'marker lists' for each cell type, which make it easy to enforce prior biological assumptions about the marker genes and expression profiles we expect to see in our cell types.  By not requiring any training data or an explicit reference profile, we expect the method to be robust to platform or batch effects that may otherwise challenge cell type labeling from pre-trained models or external reference data.  </span>
<span id="cb71-61"><a href="#cb71-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-62"><a href="#cb71-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-63"><a href="#cb71-63" aria-hidden="true" tabindex="-1"></a>The basic framework for the method can be described in a few basic steps (we'll describe each of these steps in more detail throughout our examples.)</span>
<span id="cb71-64"><a href="#cb71-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-65"><a href="#cb71-65" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>First, for each cell type, we define a set of genes we expect may be 'marker genes' or highly expressed based on prior data or biology (either the literature, a reference dataset, or some combination of both).</span>
<span id="cb71-66"><a href="#cb71-66" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Next, we construct metagene scores from our data for each cell type based on the genes specified in (1).</span>
<span id="cb71-67"><a href="#cb71-67" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Finally, we model the metagene scores, where the model estimates a posterior probability that each cell belongs to any of the possible cell types.</span>
<span id="cb71-68"><a href="#cb71-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-69"><a href="#cb71-69" aria-hidden="true" tabindex="-1"></a>We'll use an internally generated colon cancer dataset profiled with the CosMx Human 6K Discovery Panel comprised of about 112k cells, and work through two common cell typing scenarios in which we envision this methodology can be useful:</span>
<span id="cb71-70"><a href="#cb71-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-71"><a href="#cb71-71" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Subclustering an existing set of cell types: </span>
<span id="cb71-72"><a href="#cb71-72" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>In this first case, we'll assume we've already done some basic clustering, but don't quite have the level of cell type granularity we'd like.  We'll take a 'T cell' cluster and sub-classify those cells into CD8-T and CD4-T, followed by more granular T cell subcategories.</span>
<span id="cb71-73"><a href="#cb71-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-74"><a href="#cb71-74" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Top down hierarchical cell typing with focus on immune cells: </span>
<span id="cb71-75"><a href="#cb71-75" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Here we'll start from scratch; first assigning cells into major categories and identifying immune cells, then classifying those immune cells into various sub classifications of myeloid and lymphocyte cell types.</span>
<span id="cb71-76"><a href="#cb71-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-77"><a href="#cb71-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-78"><a href="#cb71-78" aria-hidden="true" tabindex="-1"></a><span class="fu"># Initial look at our dataset</span></span>
<span id="cb71-79"><a href="#cb71-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-80"><a href="#cb71-80" aria-hidden="true" tabindex="-1"></a>Let's start by taking an initial look at our data, then walk through some cell typing scenarios.  First, we can load in our data and take a look at some unsupervised leiden clustering we've already performed.</span>
<span id="cb71-81"><a href="#cb71-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-82"><a href="#cb71-82" aria-hidden="true" tabindex="-1"></a><span class="in">```{r load_libraries, message=FALSE, eval = FALSE}</span></span>
<span id="cb71-83"><a href="#cb71-83" aria-hidden="true" tabindex="-1"></a><span class="in">library(ggplot2)</span></span>
<span id="cb71-84"><a href="#cb71-84" aria-hidden="true" tabindex="-1"></a><span class="in">library(ggrepel)</span></span>
<span id="cb71-85"><a href="#cb71-85" aria-hidden="true" tabindex="-1"></a><span class="in">library(data.table)</span></span>
<span id="cb71-86"><a href="#cb71-86" aria-hidden="true" tabindex="-1"></a><span class="in">library(HieraType)</span></span>
<span id="cb71-87"><a href="#cb71-87" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-88"><a href="#cb71-88" aria-hidden="true" tabindex="-1"></a><span class="in">```{r,eval=TRUE,message=FALSE, echo=FALSE}</span></span>
<span id="cb71-89"><a href="#cb71-89" aria-hidden="true" tabindex="-1"></a><span class="in">library(data.table)</span></span>
<span id="cb71-90"><a href="#cb71-90" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-91"><a href="#cb71-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-92"><a href="#cb71-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-93"><a href="#cb71-93" aria-hidden="true" tabindex="-1"></a><span class="in">```{r , message=FALSE, eval = FALSE}</span></span>
<span id="cb71-94"><a href="#cb71-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-95"><a href="#cb71-95" aria-hidden="true" tabindex="-1"></a><span class="in">## Note, using a v3-style Seurat Object below, i.e., `sem[["RNA"]]@counts`</span></span>
<span id="cb71-96"><a href="#cb71-96" aria-hidden="true" tabindex="-1"></a><span class="in">## For v5-style, the syntax would be like `sem[["RNA"]]$counts`</span></span>
<span id="cb71-97"><a href="#cb71-97" aria-hidden="true" tabindex="-1"></a><span class="in">options(Seurat.object.assay.version = "v3")  </span></span>
<span id="cb71-98"><a href="#cb71-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-99"><a href="#cb71-99" aria-hidden="true" tabindex="-1"></a><span class="in">sem &lt;- readRDS("colon_seurat.rds")</span></span>
<span id="cb71-100"><a href="#cb71-100" aria-hidden="true" tabindex="-1"></a><span class="in">normed &lt;- HieraType::totalcount_norm(Matrix::t(sem[["RNA"]]@counts))</span></span>
<span id="cb71-101"><a href="#cb71-101" aria-hidden="true" tabindex="-1"></a><span class="in">sem &lt;- Seurat::SetAssayData(sem, layer = "data", new.data = Matrix::t(normed))</span></span>
<span id="cb71-102"><a href="#cb71-102" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-103"><a href="#cb71-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-106"><a href="#cb71-106" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-107"><a href="#cb71-107" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-108"><a href="#cb71-108" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-109"><a href="#cb71-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-110"><a href="#cb71-110" aria-hidden="true" tabindex="-1"></a><span class="co">#setwd("~/data/dan/scratch_space/CosMx-Analysis-Scratch-Space/posts/immune-cell-typing-smooth")</span></span>
<span id="cb71-111"><a href="#cb71-111" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">load_all</span>(<span class="st">"~/data/dan/imputation_smoothing/HieraType"</span>, <span class="at">export_all =</span> <span class="cn">FALSE</span>)</span>
<span id="cb71-112"><a href="#cb71-112" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-113"><a href="#cb71-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-116"><a href="#cb71-116" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-117"><a href="#cb71-117" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-118"><a href="#cb71-118" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-119"><a href="#cb71-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-120"><a href="#cb71-120" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb71-121"><a href="#cb71-121" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb71-122"><a href="#cb71-122" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb71-123"><a href="#cb71-123" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb71-124"><a href="#cb71-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-125"><a href="#cb71-125" aria-hidden="true" tabindex="-1"></a><span class="co">#sem &lt;- readRDS("colon_seurat.rds")</span></span>
<span id="cb71-126"><a href="#cb71-126" aria-hidden="true" tabindex="-1"></a>sem <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"~/data/dan/imputation_smoothing/colondata/semsub.rds"</span>)</span>
<span id="cb71-127"><a href="#cb71-127" aria-hidden="true" tabindex="-1"></a>normed <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">totalcount_norm</span>(Matrix<span class="sc">::</span><span class="fu">t</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts))</span>
<span id="cb71-128"><a href="#cb71-128" aria-hidden="true" tabindex="-1"></a>sem <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">SetAssayData</span>(sem, <span class="at">layer =</span> <span class="st">"data"</span>, <span class="at">new.data =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(normed))</span>
<span id="cb71-129"><a href="#cb71-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-130"><a href="#cb71-130" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-131"><a href="#cb71-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-132"><a href="#cb71-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-133"><a href="#cb71-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-134"><a href="#cb71-134" aria-hidden="true" tabindex="-1"></a>**Spatial plot of initial clusters**</span>
<span id="cb71-135"><a href="#cb71-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-136"><a href="#cb71-136" aria-hidden="true" tabindex="-1"></a>Here's a spatial plot of our leiden clusters.  At first glance, we see that some of these unsupervised clusters (i.e., 3, 0, 1, 2, and 7 for example) have distinct spatial patterns; it definitely looks like some meaningful biological structure in these tissue were captured by unsupervised clustering.  </span>
<span id="cb71-137"><a href="#cb71-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-138"><a href="#cb71-138" aria-hidden="true" tabindex="-1"></a><span class="in">```{r,eval=FALSE}</span></span>
<span id="cb71-139"><a href="#cb71-139" aria-hidden="true" tabindex="-1"></a><span class="in">#| code-fold: true</span></span>
<span id="cb71-140"><a href="#cb71-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-141"><a href="#cb71-141" aria-hidden="true" tabindex="-1"></a><span class="in">sem@meta.data$portion &lt;- "part_a"</span></span>
<span id="cb71-142"><a href="#cb71-142" aria-hidden="true" tabindex="-1"></a><span class="in">sem@meta.data$portion[sem@meta.data$fov %in% 60:73] &lt;- "part_b"</span></span>
<span id="cb71-143"><a href="#cb71-143" aria-hidden="true" tabindex="-1"></a><span class="in">xyplist &lt;- </span></span>
<span id="cb71-144"><a href="#cb71-144" aria-hidden="true" tabindex="-1"></a><span class="in">lapply(split(data.table(sem@meta.data), by = "portion"), function(xx){</span></span>
<span id="cb71-145"><a href="#cb71-145" aria-hidden="true" tabindex="-1"></a><span class="in">  HieraType::xyplot("clusters_unsup", metadata = xx,ptsize = 0.1) + coord_fixed()</span></span>
<span id="cb71-146"><a href="#cb71-146" aria-hidden="true" tabindex="-1"></a><span class="in">})</span></span>
<span id="cb71-147"><a href="#cb71-147" aria-hidden="true" tabindex="-1"></a><span class="in">xypl &lt;- </span></span>
<span id="cb71-148"><a href="#cb71-148" aria-hidden="true" tabindex="-1"></a><span class="in">cowplot::plot_grid(plotlist=xyplist, nrow=1) + </span></span>
<span id="cb71-149"><a href="#cb71-149" aria-hidden="true" tabindex="-1"></a><span class="in">theme(plot.background = element_rect(fill = 'white',color='white')) + </span></span>
<span id="cb71-150"><a href="#cb71-150" aria-hidden="true" tabindex="-1"></a><span class="in">  cowplot::panel_border(remove=TRUE) </span></span>
<span id="cb71-151"><a href="#cb71-151" aria-hidden="true" tabindex="-1"></a><span class="in">print(xypl)</span></span>
<span id="cb71-152"><a href="#cb71-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-153"><a href="#cb71-153" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-154"><a href="#cb71-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-155"><a href="#cb71-155" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb71-156"><a href="#cb71-156" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/xyp_unsup.png" alt="Wide Image"&gt;</span>
<span id="cb71-157"><a href="#cb71-157" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb71-158"><a href="#cb71-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-161"><a href="#cb71-161" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-162"><a href="#cb71-162" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-163"><a href="#cb71-163" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-164"><a href="#cb71-164" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.show: "hold"</span></span>
<span id="cb71-165"><a href="#cb71-165" aria-hidden="true" tabindex="-1"></a><span class="co">#| out-width: "6000px"</span></span>
<span id="cb71-166"><a href="#cb71-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-167"><a href="#cb71-167" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/xyp_unsup.png"</span>)</span>
<span id="cb71-168"><a href="#cb71-168" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-169"><a href="#cb71-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-170"><a href="#cb71-170" aria-hidden="true" tabindex="-1"></a>**Heatmap of initial clusters **</span>
<span id="cb71-171"><a href="#cb71-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-172"><a href="#cb71-172" aria-hidden="true" tabindex="-1"></a>We can make a marker gene heatmap to identify some of the top marker genes for each cluster and see if we recognize any of the celltypes represented by the unsupervised clusters.</span>
<span id="cb71-173"><a href="#cb71-173" aria-hidden="true" tabindex="-1"></a>Here we can use a couple of convenience functions provided in the R package <span class="in">`HieraType`</span>.</span>
<span id="cb71-174"><a href="#cb71-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-175"><a href="#cb71-175" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-176"><a href="#cb71-176" aria-hidden="true" tabindex="-1"></a><span class="in">#| code-fold: true</span></span>
<span id="cb71-177"><a href="#cb71-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-178"><a href="#cb71-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-179"><a href="#cb71-179" aria-hidden="true" tabindex="-1"></a><span class="in">## calculate fold change, average expression, and other summary metrics </span></span>
<span id="cb71-180"><a href="#cb71-180" aria-hidden="true" tabindex="-1"></a><span class="in">## for each gene and each cluster compared to other clusters</span></span>
<span id="cb71-181"><a href="#cb71-181" aria-hidden="true" tabindex="-1"></a><span class="in">## This is the data underlying the marker heatmap.</span></span>
<span id="cb71-182"><a href="#cb71-182" aria-hidden="true" tabindex="-1"></a><span class="in">fctbl_unsup &lt;- </span></span>
<span id="cb71-183"><a href="#cb71-183" aria-hidden="true" tabindex="-1"></a><span class="in">HieraType::clusterwise_foldchange_metrics(normed  = sem[["RNA"]]@data</span></span>
<span id="cb71-184"><a href="#cb71-184" aria-hidden="true" tabindex="-1"></a><span class="in">                                          ,metadata = sem@meta.data</span></span>
<span id="cb71-185"><a href="#cb71-185" aria-hidden="true" tabindex="-1"></a><span class="in">                                          ,cluster_column = "clusters_unsup"</span></span>
<span id="cb71-186"><a href="#cb71-186" aria-hidden="true" tabindex="-1"></a><span class="in">                                          )</span></span>
<span id="cb71-187"><a href="#cb71-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-188"><a href="#cb71-188" aria-hidden="true" tabindex="-1"></a><span class="in">## make a marker heatmap</span></span>
<span id="cb71-189"><a href="#cb71-189" aria-hidden="true" tabindex="-1"></a><span class="in">hm_unsup &lt;- HieraType::marker_heatmap(fctbl_unsup</span></span>
<span id="cb71-190"><a href="#cb71-190" aria-hidden="true" tabindex="-1"></a><span class="in">                           ,extras = c("CD3D", "CD4", "FOXP3"</span></span>
<span id="cb71-191"><a href="#cb71-191" aria-hidden="true" tabindex="-1"></a><span class="in">                                       , "CD8B", "CD8A")</span></span>
<span id="cb71-192"><a href="#cb71-192" aria-hidden="true" tabindex="-1"></a><span class="in">                           )</span></span>
<span id="cb71-193"><a href="#cb71-193" aria-hidden="true" tabindex="-1"></a><span class="in">print(hm_unsup)</span></span>
<span id="cb71-194"><a href="#cb71-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-195"><a href="#cb71-195" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-196"><a href="#cb71-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-197"><a href="#cb71-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-198"><a href="#cb71-198" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb71-199"><a href="#cb71-199" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/hm_unsup_annotated.PNG" alt="Wide Image"&gt;</span>
<span id="cb71-200"><a href="#cb71-200" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb71-201"><a href="#cb71-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-202"><a href="#cb71-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-205"><a href="#cb71-205" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-206"><a href="#cb71-206" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-207"><a href="#cb71-207" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-208"><a href="#cb71-208" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.show: "hold"</span></span>
<span id="cb71-209"><a href="#cb71-209" aria-hidden="true" tabindex="-1"></a><span class="co">#| out-width: "6000px"</span></span>
<span id="cb71-210"><a href="#cb71-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-211"><a href="#cb71-211" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/hm_unsup_annotated.PNG"</span>)</span>
<span id="cb71-212"><a href="#cb71-212" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-213"><a href="#cb71-213" aria-hidden="true" tabindex="-1"></a>Let's suppose for sake of demonstration that I recognize (and am comfortable enough) to label these clusters with cell type names based on the marker genes shown in the heatmap.  </span>
<span id="cb71-214"><a href="#cb71-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-215"><a href="#cb71-215" aria-hidden="true" tabindex="-1"></a>However, there are one or a few of my immune clusters I've identified where I desire more granularity.</span>
<span id="cb71-216"><a href="#cb71-216" aria-hidden="true" tabindex="-1"></a>In this case, I'm particularly interested in my T cells; in the bottom-left hand corner of the heatmap plot above, I see that 'cluster 6' is capturing T-cells, showing enrichment in some key marker genes like CD3 (CD3D, CD3E) as well as other markers like CD2, IL7R, and ITK.</span>
<span id="cb71-217"><a href="#cb71-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-218"><a href="#cb71-218" aria-hidden="true" tabindex="-1"></a>I'd like to split these cells further into at least a few meaningful T-cell subtypes, at minimum: "CD8+", "CD4+", and "Treg". </span>
<span id="cb71-219"><a href="#cb71-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-220"><a href="#cb71-220" aria-hidden="true" tabindex="-1"></a><span class="fu"># Example case 1: Subclustering T cells {#ex1}</span></span>
<span id="cb71-221"><a href="#cb71-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-222"><a href="#cb71-222" aria-hidden="true" tabindex="-1"></a><span class="fu">## 'Just The Code' {#tldr}</span></span>
<span id="cb71-223"><a href="#cb71-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-224"><a href="#cb71-224" aria-hidden="true" tabindex="-1"></a>Here's the code for classifying leiden cluster '6' into <span class="in">`cd8t`</span> and <span class="in">`cd4t`</span> cell types , followed by more granular T cell subtyping.  We use the R package <span class="in">`HieraType`</span> with a prespecified <span class="in">`tcell_pipeline`</span>, and corresponding lists of marker genes included as package datasets.</span>
<span id="cb71-225"><a href="#cb71-225" aria-hidden="true" tabindex="-1"></a>A <span class="in">`pipeline`</span> specifies the hierarchical order we want to do our cell typing, and which cell types go in each level of the hierarchy.</span>
<span id="cb71-226"><a href="#cb71-226" aria-hidden="true" tabindex="-1"></a>Each cell type has a list of associated marker genes.</span>
<span id="cb71-229"><a href="#cb71-229" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-230"><a href="#cb71-230" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-231"><a href="#cb71-231" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-232"><a href="#cb71-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-233"><a href="#cb71-233" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../../_code/HieraType/data/pipeline_tcell.rda"</span>)</span>
<span id="cb71-234"><a href="#cb71-234" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../../_code/HieraType/data/markerslist_tcellmajor.rda"</span>)</span>
<span id="cb71-235"><a href="#cb71-235" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../../_code/HieraType/data/markerslist_cd4tminor.rda"</span>)</span>
<span id="cb71-236"><a href="#cb71-236" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../../_code/HieraType/data/markerslist_cd8tminor.rda"</span>)</span>
<span id="cb71-237"><a href="#cb71-237" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../../_code/HieraType/data/pipeline_tcell.rda"</span>)</span>
<span id="cb71-238"><a href="#cb71-238" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../../_code/HieraType/data/pipeline_io.rda"</span>)</span>
<span id="cb71-239"><a href="#cb71-239" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../../_code/HieraType/data/markerslist_immunemajor.rda"</span>) </span>
<span id="cb71-240"><a href="#cb71-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-241"><a href="#cb71-241" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-242"><a href="#cb71-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-243"><a href="#cb71-243" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-244"><a href="#cb71-244" aria-hidden="true" tabindex="-1"></a><span class="in">data("pipeline_tcell")</span></span>
<span id="cb71-245"><a href="#cb71-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-246"><a href="#cb71-246" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-247"><a href="#cb71-247" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=TRUE}</span></span>
<span id="cb71-248"><a href="#cb71-248" aria-hidden="true" tabindex="-1"></a><span class="in">str(pipeline_tcell, max.level = 2)</span></span>
<span id="cb71-249"><a href="#cb71-249" aria-hidden="true" tabindex="-1"></a><span class="in">names(pipeline_tcell$markerslists$tmajor)</span></span>
<span id="cb71-250"><a href="#cb71-250" aria-hidden="true" tabindex="-1"></a><span class="in">names(pipeline_tcell$markerslists$t4minor)</span></span>
<span id="cb71-251"><a href="#cb71-251" aria-hidden="true" tabindex="-1"></a><span class="in">names(pipeline_tcell$markerslists$t8minor)</span></span>
<span id="cb71-252"><a href="#cb71-252" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-253"><a href="#cb71-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-254"><a href="#cb71-254" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-255"><a href="#cb71-255" aria-hidden="true" tabindex="-1"></a><span class="in">### these are cells I want to subcluster</span></span>
<span id="cb71-256"><a href="#cb71-256" aria-hidden="true" tabindex="-1"></a><span class="in">tcell_ids &lt;- sem@meta.data[sem@meta.data$clusters_unsup==6,"cell_ID"]</span></span>
<span id="cb71-257"><a href="#cb71-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-258"><a href="#cb71-258" aria-hidden="true" tabindex="-1"></a><span class="in">### Run a 'tcell'-typing pipeline, which </span></span>
<span id="cb71-259"><a href="#cb71-259" aria-hidden="true" tabindex="-1"></a><span class="in">### first classifies into T8/T4 categories (using the `tmajor` markerslist), </span></span>
<span id="cb71-260"><a href="#cb71-260" aria-hidden="true" tabindex="-1"></a><span class="in">### then runs subclassification for T4 and T8 using the t4minor and t8minor markerslists.</span></span>
<span id="cb71-261"><a href="#cb71-261" aria-hidden="true" tabindex="-1"></a><span class="in">tcell_typing &lt;- </span></span>
<span id="cb71-262"><a href="#cb71-262" aria-hidden="true" tabindex="-1"></a><span class="in">run_pipeline(pipeline = pipeline_tcell</span></span>
<span id="cb71-263"><a href="#cb71-263" aria-hidden="true" tabindex="-1"></a><span class="in">             ,counts_matrix = Matrix::t(sem[["RNA"]]@counts[,tcell_ids])</span></span>
<span id="cb71-264"><a href="#cb71-264" aria-hidden="true" tabindex="-1"></a><span class="in">             ,adjacency_matrix = as(sem@graphs$umapgrph</span></span>
<span id="cb71-265"><a href="#cb71-265" aria-hidden="true" tabindex="-1"></a><span class="in">                                    ,"CsparseMatrix")[tcell_ids,tcell_ids]</span></span>
<span id="cb71-266"><a href="#cb71-266" aria-hidden="true" tabindex="-1"></a><span class="in">             ,celltype_call_threshold = 0.5</span></span>
<span id="cb71-267"><a href="#cb71-267" aria-hidden="true" tabindex="-1"></a><span class="in">             )</span></span>
<span id="cb71-268"><a href="#cb71-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-269"><a href="#cb71-269" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-270"><a href="#cb71-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-271"><a href="#cb71-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-274"><a href="#cb71-274" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-275"><a href="#cb71-275" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-276"><a href="#cb71-276" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-277"><a href="#cb71-277" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(tcell_typing, <span class="st">"figures/tcell_typing.rds"</span>)</span>
<span id="cb71-278"><a href="#cb71-278" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-279"><a href="#cb71-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-282"><a href="#cb71-282" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-283"><a href="#cb71-283" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-284"><a href="#cb71-284" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-285"><a href="#cb71-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-286"><a href="#cb71-286" aria-hidden="true" tabindex="-1"></a>tcell_typing <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"figures/tcell_typing.rds"</span>)</span>
<span id="cb71-287"><a href="#cb71-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-288"><a href="#cb71-288" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-289"><a href="#cb71-289" aria-hidden="true" tabindex="-1"></a>The data table <span class="in">`post_probs`</span> in our <span class="in">`tcell_typing`</span> object holds the high-level cell typing results; the <span class="in">`celltype_thresh`</span> column shows the *most granular* cell type call where posterior probability is $&gt; 0.5$ (our <span class="in">`celltype_call_threshold`</span>).  We can see the posterior probablity for the <span class="in">`celltype_thresh`</span> category in <span class="in">`best_score_thresh`</span>.  </span>
<span id="cb71-290"><a href="#cb71-290" aria-hidden="true" tabindex="-1"></a>The best granular cell type call (regardless of score) is given by the <span class="in">`celltype_granular`</span> column, with <span class="in">`best_score_granular`</span> indicating the corresponding posterior probability. </span>
<span id="cb71-291"><a href="#cb71-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-292"><a href="#cb71-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-293"><a href="#cb71-293" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=TRUE}</span></span>
<span id="cb71-294"><a href="#cb71-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-295"><a href="#cb71-295" aria-hidden="true" tabindex="-1"></a><span class="in">str(tcell_typing, max.level = 2)</span></span>
<span id="cb71-296"><a href="#cb71-296" aria-hidden="true" tabindex="-1"></a><span class="in">tcell_typing$post_probs$tmajor[1:10][]</span></span>
<span id="cb71-297"><a href="#cb71-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-298"><a href="#cb71-298" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-299"><a href="#cb71-299" aria-hidden="true" tabindex="-1"></a>Now lets take a closer look at the details and describe the methodology for each step.</span>
<span id="cb71-300"><a href="#cb71-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-301"><a href="#cb71-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-302"><a href="#cb71-302" aria-hidden="true" tabindex="-1"></a><span class="fu">## Detailed walkthrough</span></span>
<span id="cb71-303"><a href="#cb71-303" aria-hidden="true" tabindex="-1"></a><span class="fu">###  Part 1: Specifying genes associated with my cell types based on prior biology</span></span>
<span id="cb71-304"><a href="#cb71-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-305"><a href="#cb71-305" aria-hidden="true" tabindex="-1"></a>We'll describe the first level of cell type classification from our example above to illustrate the method, where we suppose we want to classify <span class="in">`cd4t`</span> vs. <span class="in">`cd8t`</span>.</span>
<span id="cb71-306"><a href="#cb71-306" aria-hidden="true" tabindex="-1"></a>Creating a marker list, by specifying genes I expect **might** be associated with my cell types is the first step of this algorithm.</span>
<span id="cb71-307"><a href="#cb71-307" aria-hidden="true" tabindex="-1"></a>For each of my cell types, <span class="in">`cd4t`</span> and <span class="in">`cd8t`</span>, I specify </span>
<span id="cb71-308"><a href="#cb71-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-309"><a href="#cb71-309" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>a set of primary defining markers or "index genes" (these are genes that should be fairly exclusive to the cell type if possible)</span>
<span id="cb71-310"><a href="#cb71-310" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>a set of other marker genes which I expect could be expressed as "predictors"  </span>
<span id="cb71-311"><a href="#cb71-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-312"><a href="#cb71-312" aria-hidden="true" tabindex="-1"></a>In this case, we'll use <span class="co">[</span><span class="ot">CD4</span><span class="co">]</span>, and <span class="co">[</span><span class="ot">CD8A, CD8B</span><span class="co">]</span>, as "index genes" for <span class="in">`cd4t`</span> and <span class="in">`cd8t`</span> respectively.</span>
<span id="cb71-313"><a href="#cb71-313" aria-hidden="true" tabindex="-1"></a>The other genes we'll specify as predictors are based on the literature.  For example, FOXP3 is often associated with cd4t 'Treg' cells, so we include it as a predictor here for <span class="in">`cd4t`</span>. In general, it's okay to include a gene as a "predictor" even if we only expect to see it in a subset of cells (like FOXP3 in this example, which we expect to be present in a 'Treg' subtype, but not all <span class="in">`cd4t`</span> cells).</span>
<span id="cb71-314"><a href="#cb71-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-315"><a href="#cb71-315" aria-hidden="true" tabindex="-1"></a>It's okay to include markers in multiple cell types if that fits our mental model.</span>
<span id="cb71-316"><a href="#cb71-316" aria-hidden="true" tabindex="-1"></a>For example, CD3D, CD3E, and CD3G are general T cell markers, and so are included as predictors for both subtypes, as are other genes like IL7R, CD27, SELL which could be seen in either <span class="in">`cd4t`</span> or <span class="in">`cd8t`</span> cells.</span>
<span id="cb71-317"><a href="#cb71-317" aria-hidden="true" tabindex="-1"></a>When manually creating a marker list, consider "more predictors is better" as a general rule of thumb; a larger number of predictor genes can help build more accurate metagene models for celltype index markers.</span>
<span id="cb71-318"><a href="#cb71-318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-319"><a href="#cb71-319" aria-hidden="true" tabindex="-1"></a>We can create our marker list manually like this:</span>
<span id="cb71-320"><a href="#cb71-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-321"><a href="#cb71-321" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-322"><a href="#cb71-322" aria-hidden="true" tabindex="-1"></a><span class="in">markerslist_tcellmajor &lt;- </span></span>
<span id="cb71-323"><a href="#cb71-323" aria-hidden="true" tabindex="-1"></a><span class="in">HieraType::make_markerslist(</span></span>
<span id="cb71-324"><a href="#cb71-324" aria-hidden="true" tabindex="-1"></a><span class="in">  index_marker = list(</span></span>
<span id="cb71-325"><a href="#cb71-325" aria-hidden="true" tabindex="-1"></a><span class="in">    "cd8t" = c("CD8B", "CD8A")</span></span>
<span id="cb71-326"><a href="#cb71-326" aria-hidden="true" tabindex="-1"></a><span class="in">    ,"cd4t" = c("CD4")</span></span>
<span id="cb71-327"><a href="#cb71-327" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb71-328"><a href="#cb71-328" aria-hidden="true" tabindex="-1"></a><span class="in">  ,predictors = list(</span></span>
<span id="cb71-329"><a href="#cb71-329" aria-hidden="true" tabindex="-1"></a><span class="in">    "cd8t" =  c("CD3D", "CD3E", "CD3G", "CD8A", "CD8B", "IL7R", "CD27", "SELL", "CCR7", "TCF7", "LEF1", </span></span>
<span id="cb71-330"><a href="#cb71-330" aria-hidden="true" tabindex="-1"></a><span class="in">                    "FOXP1", "BACH2", "KLF2", "CD28", "ICOS", "STAT3", "STAT5", "RUNX3", "BATF", "EOMES", </span></span>
<span id="cb71-331"><a href="#cb71-331" aria-hidden="true" tabindex="-1"></a><span class="in">                    "TBX21", "PRDM1", "PDCD1", "LAG3", "TIGIT", "HAVCR2", "GZMK", "GZMA", "GZMB", "PRF1", </span></span>
<span id="cb71-332"><a href="#cb71-332" aria-hidden="true" tabindex="-1"></a><span class="in">                    "GNLY", "KLRG1", "KLRD1", "IFNG", "CCL5", "CX3CR1", "NKG7", "FAS", "FASLG", "IL2", "CD244", </span></span>
<span id="cb71-333"><a href="#cb71-333" aria-hidden="true" tabindex="-1"></a><span class="in">                    "CXCR3", "KLRB1", "TOX", "NR4A2", "BCL6", "CXCL13")</span></span>
<span id="cb71-334"><a href="#cb71-334" aria-hidden="true" tabindex="-1"></a><span class="in">    ,"cd4t" = c("CD3D", "CD3E", "CD3G", "CD4", "IL7R", "CD27", "SELL", "CCR7", "TCF7", "LEF1", "FOXP1", </span></span>
<span id="cb71-335"><a href="#cb71-335" aria-hidden="true" tabindex="-1"></a><span class="in">                     "BACH2", "KLF2", "CD28", "ICOS", "CD40LG", "STAT3", "STAT5", "RUNX1", "RUNX3", "BATF", </span></span>
<span id="cb71-336"><a href="#cb71-336" aria-hidden="true" tabindex="-1"></a><span class="in">                     "GATA3", "TBX21", "RORC", "FOXP3", "IL2RA", "CTLA4", "PDCD1", "LAG3", "TIGIT", "HAVCR2", </span></span>
<span id="cb71-337"><a href="#cb71-337" aria-hidden="true" tabindex="-1"></a><span class="in">                     "IL2", "IFNG", "IL4", "IL5", "IL13", "IL17A", "IL17F", "IL10", "CCL5", "CXCR3", "CCR6", </span></span>
<span id="cb71-338"><a href="#cb71-338" aria-hidden="true" tabindex="-1"></a><span class="in">                     "CXCR5", "PRDM1", "BCL6", "EOMES", "CXCL13", "TOX")</span></span>
<span id="cb71-339"><a href="#cb71-339" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb71-340"><a href="#cb71-340" aria-hidden="true" tabindex="-1"></a><span class="in">)</span></span>
<span id="cb71-341"><a href="#cb71-341" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-342"><a href="#cb71-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-343"><a href="#cb71-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-344"><a href="#cb71-344" aria-hidden="true" tabindex="-1"></a>The object created looks like this below</span>
<span id="cb71-345"><a href="#cb71-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-346"><a href="#cb71-346" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=TRUE}</span></span>
<span id="cb71-347"><a href="#cb71-347" aria-hidden="true" tabindex="-1"></a><span class="in">str(markerslist_tcellmajor)</span></span>
<span id="cb71-348"><a href="#cb71-348" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-349"><a href="#cb71-349" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-350"><a href="#cb71-350" aria-hidden="true" tabindex="-1"></a>In practice, we can save and reuse our marker lists for different cell types.</span>
<span id="cb71-351"><a href="#cb71-351" aria-hidden="true" tabindex="-1"></a>While it should be easy to create/modify a custom list from the literature,  this particular list is included as a package dataset <span class="in">`markerslist_tcellmajor`</span> to facilitate ease of use.  </span>
<span id="cb71-352"><a href="#cb71-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-353"><a href="#cb71-353" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-354"><a href="#cb71-354" aria-hidden="true" tabindex="-1"></a><span class="in">data("markerslist_tcellmajor")</span></span>
<span id="cb71-355"><a href="#cb71-355" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-358"><a href="#cb71-358" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-359"><a href="#cb71-359" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(markerslist_tcellmajor)</span>
<span id="cb71-360"><a href="#cb71-360" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-361"><a href="#cb71-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-362"><a href="#cb71-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-363"><a href="#cb71-363" aria-hidden="true" tabindex="-1"></a><span class="fu">### Part 2: Estimating a metagene score for each cell type</span></span>
<span id="cb71-364"><a href="#cb71-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-365"><a href="#cb71-365" aria-hidden="true" tabindex="-1"></a>Because single count values marker genes can be noisy, we seek to derive variables</span>
<span id="cb71-366"><a href="#cb71-366" aria-hidden="true" tabindex="-1"></a> that more stably estimate marker expression. To do so, we'll exploit biologically relevant genes (e.g. CD8A expression is predictive of CD8B expression), and cells with similar expression patterns. </span>
<span id="cb71-367"><a href="#cb71-367" aria-hidden="true" tabindex="-1"></a>To do so, we fit a model for each cell type, using the primary index marker as the response variable and covariates derived from the predictor genes.  The predicted values of the primary index marker for this model make a continuous "metagene score" that will later be used for probabilistic cell type assignment. </span>
<span id="cb71-368"><a href="#cb71-368" aria-hidden="true" tabindex="-1"></a>You can think of each metagene score as estimating the canonical marker gene's expression for a single cell type;</span>
<span id="cb71-369"><a href="#cb71-369" aria-hidden="true" tabindex="-1"></a> e.g. we will produce a FOXP3-estimating metagene to capture Treg's, a CD68-estimating metagene to capture macrophages, etc...</span>
<span id="cb71-370"><a href="#cb71-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-371"><a href="#cb71-371" aria-hidden="true" tabindex="-1"></a>We fit our models for each cell type class using the <span class="in">`fit_metagene_scores`</span> function.</span>
<span id="cb71-372"><a href="#cb71-372" aria-hidden="true" tabindex="-1"></a>This function uses 'smoothing' (averaging over a set of similar cells), as well as our prior biology / assumptions about which markers are associated with each celltype to predict the pearson residuals of the index gene.</span>
<span id="cb71-373"><a href="#cb71-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-374"><a href="#cb71-374" aria-hidden="true" tabindex="-1"></a>The default regression model setup looks like this:</span>
<span id="cb71-375"><a href="#cb71-375" aria-hidden="true" tabindex="-1"></a>$$Y = WY\eta  + WX\beta + X\gamma + \epsilon $$</span>
<span id="cb71-376"><a href="#cb71-376" aria-hidden="true" tabindex="-1"></a>where</span>
<span id="cb71-377"><a href="#cb71-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-378"><a href="#cb71-378" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>$Y:=$ pearson residuals of the index marker (our response variable)</span>
<span id="cb71-379"><a href="#cb71-379" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>$X:=$ (totalcount-normalized) matrix of predictor genes</span>
<span id="cb71-380"><a href="#cb71-380" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>$W:=$ a cells $\times$ cells neighborhood matrix, denoting cells with similar expression profiles. </span>
<span id="cb71-381"><a href="#cb71-381" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>$\eta$, $\beta$, and $\gamma$ are the regression coefficients associated with the 'neighbor-averaged' index gene, 'neighbor-averaged' predictor genes, and predictor genes within the cell respectively. </span>
<span id="cb71-382"><a href="#cb71-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-383"><a href="#cb71-383" aria-hidden="true" tabindex="-1"></a>One way to think about this model specification could be as an implementation of a 'smoothing model', which leverages information across multiple dimensions:</span>
<span id="cb71-384"><a href="#cb71-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-385"><a href="#cb71-385" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>the index gene's mean expression in similar cells ($WY$) (roughly, neighbors in UMAP space)</span>
<span id="cb71-386"><a href="#cb71-386" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>expression of other expected marker genes ($X$) (our predictor genes defined from biological knowledge)</span>
<span id="cb71-387"><a href="#cb71-387" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>expression of other expected marker genes in similar cells ($WX$) </span>
<span id="cb71-388"><a href="#cb71-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-389"><a href="#cb71-389" aria-hidden="true" tabindex="-1"></a>We use a non-negative least squares <span class="co">[</span><span class="ot">@nnlsref</span><span class="co">]</span> (NNLS) regression to estimate our models.  The NNLS model enforces that all of the covariates based on predictor genes we specified must have positive regression coefficients (in practice we'll observe that a number of regression coefficients may shrink to zero).</span>
<span id="cb71-390"><a href="#cb71-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-391"><a href="#cb71-391" aria-hidden="true" tabindex="-1"></a>By specifying <span class="in">`use_offclass_markers_as_negative_predictors = TRUE`</span> in the section above, we further assume that predictors for *other celltypes* that we haven't listed can be used as *negative* predictors of our index gene.</span>
<span id="cb71-392"><a href="#cb71-392" aria-hidden="true" tabindex="-1"></a>For example, genes like like "GZMA" and "GZMB", which are markers associated with <span class="in">`cd8t`</span>, will be assumed to be *negative* predictors of the other classes (<span class="in">`cd4t`</span>) where they are not listed. </span>
<span id="cb71-393"><a href="#cb71-393" aria-hidden="true" tabindex="-1"></a>This argument should in general be set to TRUE, as it's useful to specify which genes we *don't* expect to see in a celltype. If your markers list has extensive genes expected to be enriched in multiple cell types, and those genes are for some reason omitted from many of the other cell types, then a user might consider specifying this argument as FALSE; e.g. if many genes like CD3E are specified in markers lists for some T cell subtypes but not all (despite being ubiquitously expressed).</span>
<span id="cb71-394"><a href="#cb71-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-395"><a href="#cb71-395" aria-hidden="true" tabindex="-1"></a>Estimating this model from the dataset at hand rather than using a reference model could make this framework more robust to platform effects or batch effects compared to pre-trained models or reference profiles which used different technology or data with slightly different biology.</span>
<span id="cb71-396"><a href="#cb71-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-397"><a href="#cb71-397" aria-hidden="true" tabindex="-1"></a>Genes not included in the marker gene lists will not be considered in assigning cell types. Because the model is only using genes we’ve specified in our ‘markers list’, we have complete control over which genes influence the metadata scores, reducing the chance that noisy or irrelevant genes can have an impact on our cell type classifications. Furthermore, by estimating the metagenes directly from the dataset at hand, we are not reliant on reference datasets or pre-trained models influenced by different technology (example: scRNA-seq) or underlying biology (example: different tissue type or disease status, etc).</span>
<span id="cb71-398"><a href="#cb71-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-399"><a href="#cb71-399" aria-hidden="true" tabindex="-1"></a>Here I call the <span class="in">`fit_metagene_scores`</span> function to estimate these models for cells corresponding to the unsupervised cluster I've identified to be T cells.  </span>
<span id="cb71-400"><a href="#cb71-400" aria-hidden="true" tabindex="-1"></a>For the cells $\times$ cells <span class="in">`adjacency_matrix`</span>, I've used the nearest neighbor matrix in my Seurat object.  This is the same graph that was used to generate my UMAP and my unsupervised leiden clusters.</span>
<span id="cb71-401"><a href="#cb71-401" aria-hidden="true" tabindex="-1"></a>After fitting our regression models for each cell type, we'll use the predicted scores as 'metagenes' that can be clustered in the next step.</span>
<span id="cb71-402"><a href="#cb71-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-403"><a href="#cb71-403" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-404"><a href="#cb71-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-405"><a href="#cb71-405" aria-hidden="true" tabindex="-1"></a><span class="in">tcell_ids &lt;- sem@meta.data[sem@meta.data$clusters_unsup==6,"cell_ID"]</span></span>
<span id="cb71-406"><a href="#cb71-406" aria-hidden="true" tabindex="-1"></a><span class="in">metagenes_t &lt;- </span></span>
<span id="cb71-407"><a href="#cb71-407" aria-hidden="true" tabindex="-1"></a><span class="in">  fit_metagene_scores(</span></span>
<span id="cb71-408"><a href="#cb71-408" aria-hidden="true" tabindex="-1"></a><span class="in">    markerslist = markerslist_tcellmajor</span></span>
<span id="cb71-409"><a href="#cb71-409" aria-hidden="true" tabindex="-1"></a><span class="in">    ,counts_matrix = Matrix::t(sem[["RNA"]]@counts[,tcell_ids])</span></span>
<span id="cb71-410"><a href="#cb71-410" aria-hidden="true" tabindex="-1"></a><span class="in">    ,adjacency_matrix = as(sem@graphs$umapgrph</span></span>
<span id="cb71-411"><a href="#cb71-411" aria-hidden="true" tabindex="-1"></a><span class="in">                           ,"CsparseMatrix")[tcell_ids,tcell_ids]</span></span>
<span id="cb71-412"><a href="#cb71-412" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb71-413"><a href="#cb71-413" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-414"><a href="#cb71-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-415"><a href="#cb71-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-418"><a href="#cb71-418" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-419"><a href="#cb71-419" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-420"><a href="#cb71-420" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-421"><a href="#cb71-421" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(metagenes_t, <span class="st">"figures/metagenes_t.rds"</span>)</span>
<span id="cb71-422"><a href="#cb71-422" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-423"><a href="#cb71-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-426"><a href="#cb71-426" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-427"><a href="#cb71-427" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-428"><a href="#cb71-428" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-429"><a href="#cb71-429" aria-hidden="true" tabindex="-1"></a>metagenes_t <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"figures/metagenes_t.rds"</span>)</span>
<span id="cb71-430"><a href="#cb71-430" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-431"><a href="#cb71-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-432"><a href="#cb71-432" aria-hidden="true" tabindex="-1"></a>The <span class="in">`metagenes_t`</span> object we've created has an element for each celltype, including the regression coefficients (<span class="in">`bhat`</span>), and predicted values we'll use for clustering (<span class="in">`yhat`</span>) for each index marker gene for cell type.</span>
<span id="cb71-433"><a href="#cb71-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-434"><a href="#cb71-434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-437"><a href="#cb71-437" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-438"><a href="#cb71-438" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(metagenes_t,<span class="at">max.level =</span> <span class="dv">2</span>)</span>
<span id="cb71-439"><a href="#cb71-439" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-440"><a href="#cb71-440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-441"><a href="#cb71-441" aria-hidden="true" tabindex="-1"></a>For cell types using multiple 'index marker genes', like <span class="in">`cd8t`</span> here (using CD8A and CD8B), metagene scores are combined together using non-negative sparse pca (implemented in the <span class="in">`nsprcomp`</span> R package) <span class="co">[</span><span class="ot">@nonnegpca</span><span class="co">]</span>.  This allows for summarizing an overall cell type metagene score, using a constrained non-negative combination of individual index-gene scores that explains maximal variation across individual index gene scores. </span>
<span id="cb71-442"><a href="#cb71-442" aria-hidden="true" tabindex="-1"></a>Below, we see the PCA rotation representing the individual weights given to CD8A and CD8B metagene scores.</span>
<span id="cb71-443"><a href="#cb71-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-446"><a href="#cb71-446" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-447"><a href="#cb71-447" aria-hidden="true" tabindex="-1"></a>metagenes_t<span class="sc">$</span>cd8t<span class="sc">$</span>nnpcfit</span>
<span id="cb71-448"><a href="#cb71-448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-449"><a href="#cb71-449" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-450"><a href="#cb71-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-451"><a href="#cb71-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-452"><a href="#cb71-452" aria-hidden="true" tabindex="-1"></a>We can get a better understanding of the metagene model outputs by looking at a few diagnostics plots.  In the below, we plot the regression coefficients for each model, ordering them on the x-axis by their ranking.</span>
<span id="cb71-453"><a href="#cb71-453" aria-hidden="true" tabindex="-1"></a>For each cell type, the positive $\hat{\beta}$ correspond to genes we specified as predictors, and negative $\hat{\beta}$ were predictors for either of the other two cell types.  </span>
<span id="cb71-454"><a href="#cb71-454" aria-hidden="true" tabindex="-1"></a>A lot of the covariates which have the largest magnitudes have "*lag.*" in their name indicating they correspond to a smoothed (or neighbor-averaged) covariate for that gene. </span>
<span id="cb71-455"><a href="#cb71-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-456"><a href="#cb71-456" aria-hidden="true" tabindex="-1"></a>We also see that many genes end up with weights of 0. These genes failed to be predictive in the direction our biological priors expected. </span>
<span id="cb71-457"><a href="#cb71-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-458"><a href="#cb71-458" aria-hidden="true" tabindex="-1"></a>For <span class="in">`cd4t`</span>, we see that smoothed CD4 (indicated by '*lag.y.CD4*' in the plot) is the strongest predictor of observed 'CD4'.  For `cd8t`, '*lag.y.CD8A*' and '*lag.y.CD8B*' (indicating smoothed CD8A and CD8B) are also among the strongest predictors, but not first in rank. </span>
<span id="cb71-459"><a href="#cb71-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-460"><a href="#cb71-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-461"><a href="#cb71-461" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-462"><a href="#cb71-462" aria-hidden="true" tabindex="-1"></a><span class="in">#| code-fold: true</span></span>
<span id="cb71-463"><a href="#cb71-463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-464"><a href="#cb71-464" aria-hidden="true" tabindex="-1"></a><span class="in">coefplots &lt;- metagene_coefficients_plot(metagenes_t)</span></span>
<span id="cb71-465"><a href="#cb71-465" aria-hidden="true" tabindex="-1"></a><span class="in">coefplots &lt;- lapply(coefplots, function(x){x +   theme(text=element_text(face='bold',size=18))</span></span>
<span id="cb71-466"><a href="#cb71-466" aria-hidden="true" tabindex="-1"></a><span class="in">  }) ## just make the text bigger and bold</span></span>
<span id="cb71-467"><a href="#cb71-467" aria-hidden="true" tabindex="-1"></a><span class="in">cowplot::plot_grid(plotlist = coefplots, nrow = 1)</span></span>
<span id="cb71-468"><a href="#cb71-468" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-469"><a href="#cb71-469" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb71-470"><a href="#cb71-470" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/coefplots_tcells.png" alt="Wide Image"&gt;</span>
<span id="cb71-471"><a href="#cb71-471" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb71-472"><a href="#cb71-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-475"><a href="#cb71-475" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-476"><a href="#cb71-476" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-477"><a href="#cb71-477" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-478"><a href="#cb71-478" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"figures/coefplots_tcells.png"</span>, <span class="at">width =</span> <span class="dv">1904</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">50</span>), <span class="at">height =</span> <span class="dv">671</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">50</span>), <span class="at">units =</span> <span class="st">"px"</span>, <span class="at">dpi =</span> <span class="dv">500</span>)</span>
<span id="cb71-479"><a href="#cb71-479" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-480"><a href="#cb71-480" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-483"><a href="#cb71-483" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-484"><a href="#cb71-484" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-485"><a href="#cb71-485" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-486"><a href="#cb71-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-487"><a href="#cb71-487" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/coefplots_tcells.png"</span>)</span>
<span id="cb71-488"><a href="#cb71-488" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-489"><a href="#cb71-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-490"><a href="#cb71-490" aria-hidden="true" tabindex="-1"></a>A pairsplot can show us how well the metagene scores we've created are distinct from each other.  Ideally, if our cell type classes are mutually exclusive, we'd want to not see too many cells which are 'positive' for multiple cell types at the same time.</span>
<span id="cb71-491"><a href="#cb71-491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-492"><a href="#cb71-492" aria-hidden="true" tabindex="-1"></a>Here, we see that <span class="in">`cd4t`</span> and <span class="in">`cd8t`</span> metagene scores are negatively correlated with each other, which is good to see.</span>
<span id="cb71-493"><a href="#cb71-493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-494"><a href="#cb71-494" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-495"><a href="#cb71-495" aria-hidden="true" tabindex="-1"></a><span class="in">#| code-fold: true</span></span>
<span id="cb71-496"><a href="#cb71-496" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-497"><a href="#cb71-497" aria-hidden="true" tabindex="-1"></a><span class="in">metagene_pairsplot(metagenes_t, var= "yhat")</span></span>
<span id="cb71-498"><a href="#cb71-498" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-499"><a href="#cb71-499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-502"><a href="#cb71-502" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-503"><a href="#cb71-503" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-504"><a href="#cb71-504" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-505"><a href="#cb71-505" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"figures/yhatpairs_tcells.png"</span>, <span class="at">width =</span> <span class="dv">858</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">height =</span> <span class="dv">634</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">units =</span> <span class="st">"px"</span>, <span class="at">dpi =</span> <span class="dv">500</span>)</span>
<span id="cb71-506"><a href="#cb71-506" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-507"><a href="#cb71-507" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-508"><a href="#cb71-508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-509"><a href="#cb71-509" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-512"><a href="#cb71-512" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-513"><a href="#cb71-513" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-514"><a href="#cb71-514" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-515"><a href="#cb71-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-516"><a href="#cb71-516" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/yhatpairs_tcells.png"</span>)</span>
<span id="cb71-517"><a href="#cb71-517" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-518"><a href="#cb71-518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-519"><a href="#cb71-519" aria-hidden="true" tabindex="-1"></a><span class="fu">### Part 3: Assigning cell types based on metagene scores</span></span>
<span id="cb71-520"><a href="#cb71-520" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-521"><a href="#cb71-521" aria-hidden="true" tabindex="-1"></a>Now we use the metagene scores to classify cell types. </span>
<span id="cb71-522"><a href="#cb71-522" aria-hidden="true" tabindex="-1"></a>To do so, we estimate the multivariate distribution of metagene scores that characterize each cell type.  This allows us to compute the likelihoods for a cell belonging to any of these types, and infer a posterior probability of T-cell subtype for each cell.</span>
<span id="cb71-523"><a href="#cb71-523" aria-hidden="true" tabindex="-1"></a>The <span class="in">`cluster_metagenes`</span> function we run below estimates a mixture of overfitted multivariate gaussian mixture models <span class="co">[</span><span class="ot">@aragam2020identifiability</span><span class="co">]</span> to estimate the distribution of metagene scores.</span>
<span id="cb71-524"><a href="#cb71-524" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-527"><a href="#cb71-527" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-528"><a href="#cb71-528" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-529"><a href="#cb71-529" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-530"><a href="#cb71-530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-531"><a href="#cb71-531" aria-hidden="true" tabindex="-1"></a>model_t <span class="ot">&lt;-</span> </span>
<span id="cb71-532"><a href="#cb71-532" aria-hidden="true" tabindex="-1"></a>HieraType<span class="sc">::</span><span class="fu">cluster_metagenes</span>(<span class="at">metagenes =</span> metagenes_t)</span>
<span id="cb71-533"><a href="#cb71-533" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-534"><a href="#cb71-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-537"><a href="#cb71-537" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-538"><a href="#cb71-538" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-539"><a href="#cb71-539" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-540"><a href="#cb71-540" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-541"><a href="#cb71-541" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(model_t, <span class="st">"figures/model_t.rds"</span>)</span>
<span id="cb71-542"><a href="#cb71-542" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-543"><a href="#cb71-543" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-546"><a href="#cb71-546" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-547"><a href="#cb71-547" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-548"><a href="#cb71-548" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-549"><a href="#cb71-549" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-550"><a href="#cb71-550" aria-hidden="true" tabindex="-1"></a>model_t <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"figures/model_t.rds"</span>)</span>
<span id="cb71-551"><a href="#cb71-551" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-552"><a href="#cb71-552" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-553"><a href="#cb71-553" aria-hidden="true" tabindex="-1"></a>At a high level, the <span class="in">`cluster_metagenes()`</span> function first identifies all cells with positive metagene scores for a particular cell type. We might loosely interpret these cells as 'anchor cells'.</span>
<span id="cb71-554"><a href="#cb71-554" aria-hidden="true" tabindex="-1"></a>The data may look like this for example, where here I have subset my original data to cells which have positive scores for my celltype <span class="in">`cd8t`</span>.</span>
<span id="cb71-555"><a href="#cb71-555" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-558"><a href="#cb71-558" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-559"><a href="#cb71-559" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-560"><a href="#cb71-560" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-561"><a href="#cb71-561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-562"><a href="#cb71-562" aria-hidden="true" tabindex="-1"></a>yh <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">make_scores_matrix</span>(<span class="at">metagenes =</span> metagenes_t, <span class="st">"yhat"</span>)</span>
<span id="cb71-563"><a href="#cb71-563" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(yh, <span class="st">"figures/yh_example.rds"</span>)</span>
<span id="cb71-564"><a href="#cb71-564" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-565"><a href="#cb71-565" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-568"><a href="#cb71-568" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-569"><a href="#cb71-569" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-570"><a href="#cb71-570" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-571"><a href="#cb71-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-572"><a href="#cb71-572" aria-hidden="true" tabindex="-1"></a>yh <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"figures/yh_example.rds"</span>)</span>
<span id="cb71-573"><a href="#cb71-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-574"><a href="#cb71-574" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-577"><a href="#cb71-577" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-578"><a href="#cb71-578" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-579"><a href="#cb71-579" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-580"><a href="#cb71-580" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-581"><a href="#cb71-581" aria-hidden="true" tabindex="-1"></a>yh <span class="ot">&lt;-</span> <span class="fu">make_scores_matrix</span>(<span class="at">metagenes =</span> metagenes_t, <span class="st">"yhat"</span>)</span>
<span id="cb71-582"><a href="#cb71-582" aria-hidden="true" tabindex="-1"></a>yh[yh[,<span class="st">"cd8t"</span>] <span class="sc">&gt;</span> <span class="dv">0</span>,][<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,]</span>
<span id="cb71-583"><a href="#cb71-583" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-584"><a href="#cb71-584" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-585"><a href="#cb71-585" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-586"><a href="#cb71-586" aria-hidden="true" tabindex="-1"></a>Then, a mixture model is fit to these (2-dimensional) scores, by default using 6 components.</span>
<span id="cb71-587"><a href="#cb71-587" aria-hidden="true" tabindex="-1"></a>The estimates corresponding to the individual mixture models for each cell type are found in the <span class="in">`models`</span> list, with the means (<span class="in">`mu_hat`</span>), covariance matrices (<span class="in">`sigma_hat`</span>), and prior probability (<span class="in">`pihat`</span>) corresponding to each mixture component. </span>
<span id="cb71-590"><a href="#cb71-590" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-591"><a href="#cb71-591" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(model_t<span class="sc">$</span>models, <span class="at">max.level =</span> <span class="dv">2</span>)</span>
<span id="cb71-592"><a href="#cb71-592" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-593"><a href="#cb71-593" aria-hidden="true" tabindex="-1"></a>Typically we shouldn't need to interact with this part of the output.  After the mixture models for each cell type are fitted, we can score all cells, and get an overall posterior probability that a cell belongs to any of the two major classes using Bayes rule.</span>
<span id="cb71-594"><a href="#cb71-594" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-595"><a href="#cb71-595" aria-hidden="true" tabindex="-1"></a>Let $y_c$ be the cell type category for cell $c$, $scores_c$ the corresponding metagene scores, and $t$ denoting a possible cell 'type' or category (for $t = 1,...,T$).</span>
<span id="cb71-596"><a href="#cb71-596" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-597"><a href="#cb71-597" aria-hidden="true" tabindex="-1"></a>We compute the overall posterior probability that $y_c$ is a particular type $t$, as</span>
<span id="cb71-598"><a href="#cb71-598" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-599"><a href="#cb71-599" aria-hidden="true" tabindex="-1"></a>$$P(y_c = t | scores_c) = \frac{P(scores_c | y_c = t)P(y_c = t)}{\sum_{l=1}^T P(scores_c | y_c = l)P(y_c = l)}$$</span>
<span id="cb71-600"><a href="#cb71-600" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-601"><a href="#cb71-601" aria-hidden="true" tabindex="-1"></a>Using the individual mixture models for a particular cell type with $K$ mixture components, we have</span>
<span id="cb71-602"><a href="#cb71-602" aria-hidden="true" tabindex="-1"></a>$$P(scores_c | y_c = t) = \sum_{j=1}^K P(scores_c | k=j, y_c = t) P(k=j | y_c = t)$$</span>
<span id="cb71-603"><a href="#cb71-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-604"><a href="#cb71-604" aria-hidden="true" tabindex="-1"></a>The prior probability for a particular cell type class $t$ can be estimated by counting the number of cells used to estimate the mixture model in each cell type, and computing a proportion amongst all possible cell type classes, i.e.,</span>
<span id="cb71-605"><a href="#cb71-605" aria-hidden="true" tabindex="-1"></a>$$P(y_c = t) = \frac{\sum_{c=1}^C I(scores_{c,t} &gt; 0)}{\sum_{l=1}^T \sum_{c=1}^C I(scores_{c,l} &gt; 0)}$$</span>
<span id="cb71-606"><a href="#cb71-606" aria-hidden="true" tabindex="-1"></a>These probabilities are in the <span class="in">`post_probs`</span> data.table object of the model.</span>
<span id="cb71-607"><a href="#cb71-607" aria-hidden="true" tabindex="-1"></a>The "best class" column gives each cell's best-fitting cell type. </span>
<span id="cb71-608"><a href="#cb71-608" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-611"><a href="#cb71-611" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-612"><a href="#cb71-612" aria-hidden="true" tabindex="-1"></a>model_t<span class="sc">$</span>post_probs[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>][]</span>
<span id="cb71-613"><a href="#cb71-613" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-614"><a href="#cb71-614" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-615"><a href="#cb71-615" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-616"><a href="#cb71-616" aria-hidden="true" tabindex="-1"></a><span class="fu">### Granular hierarchical subtyping and using a `pipeline`</span></span>
<span id="cb71-617"><a href="#cb71-617" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-618"><a href="#cb71-618" aria-hidden="true" tabindex="-1"></a>These posterior-probability scores can be passed directly to subtyping models for <span class="in">`cd4t`</span> and <span class="in">`cd8t`</span>, via the <span class="in">`prior_level_weights`</span> arguments.</span>
<span id="cb71-619"><a href="#cb71-619" aria-hidden="true" tabindex="-1"></a>For example, if I wanted to go ahead and cluster the <span class="in">`cd4t`</span> cells into subtypes, I could follow the same process, passing in the prior probability of the cell being <span class="in">`cd4t`</span> as a weight.</span>
<span id="cb71-620"><a href="#cb71-620" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-621"><a href="#cb71-621" aria-hidden="true" tabindex="-1"></a>In the example below, the probability of <span class="in">`cd4t`</span> is used as a weight in the NNLS regression fitting (akin to weighted least squares) when constructing metagene scores.  Cells with high probability of <span class="in">`cd4t`</span> are emphasized, while cells with lower probability of <span class="in">`cd4t`</span> have less influence on the regression coefficients used to generate metagene scores.</span>
<span id="cb71-622"><a href="#cb71-622" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-623"><a href="#cb71-623" aria-hidden="true" tabindex="-1"></a>In the clustering stage, probability of <span class="in">`cd4t`</span> is again used as a weight (or prior probability), that factors into the expectation step of the EM algorithm.</span>
<span id="cb71-624"><a href="#cb71-624" aria-hidden="true" tabindex="-1"></a>Letting <span class="in">`s(t)`</span> denote the *subtype* of <span class="in">`cd4t`</span>, then we have the below, where $P(y_c = t)$ would be the probability of <span class="in">`cd4t`</span> from the previous stage of modeling.</span>
<span id="cb71-625"><a href="#cb71-625" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-626"><a href="#cb71-626" aria-hidden="true" tabindex="-1"></a>The final returned probability for each <span class="in">`cd4t`</span> subtype is a *joint* probability of both being the subtype $s(t)$ and the major celltype $t$.</span>
<span id="cb71-627"><a href="#cb71-627" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-628"><a href="#cb71-628" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-629"><a href="#cb71-629" aria-hidden="true" tabindex="-1"></a>$$P(y_c = s(t) \cap y_c = t  | scores_c) = P(y_c = s(t) | scores_c, y_c = t)P(y_c = t)$$</span>
<span id="cb71-630"><a href="#cb71-630" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-631"><a href="#cb71-631" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-632"><a href="#cb71-632" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-633"><a href="#cb71-633" aria-hidden="true" tabindex="-1"></a>Below is an example of this syntax using the <span class="in">`cd4t`</span> subtypes in the package dataset "<span class="in">`markerslist_cd4tminor`</span>".</span>
<span id="cb71-634"><a href="#cb71-634" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-635"><a href="#cb71-635" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-636"><a href="#cb71-636" aria-hidden="true" tabindex="-1"></a><span class="in">str(HieraType::markerslist_cd4tminor, max.level=2)</span></span>
<span id="cb71-637"><a href="#cb71-637" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-638"><a href="#cb71-638" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=TRUE}</span></span>
<span id="cb71-639"><a href="#cb71-639" aria-hidden="true" tabindex="-1"></a><span class="in">#| eval: true</span></span>
<span id="cb71-640"><a href="#cb71-640" aria-hidden="true" tabindex="-1"></a><span class="in">#| echo: false</span></span>
<span id="cb71-641"><a href="#cb71-641" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-642"><a href="#cb71-642" aria-hidden="true" tabindex="-1"></a><span class="in">str(markerslist_cd4tminor, max.level=2)</span></span>
<span id="cb71-643"><a href="#cb71-643" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-644"><a href="#cb71-644" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-645"><a href="#cb71-645" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-646"><a href="#cb71-646" aria-hidden="true" tabindex="-1"></a><span class="in">#### Classify cd4t subtypes</span></span>
<span id="cb71-647"><a href="#cb71-647" aria-hidden="true" tabindex="-1"></a><span class="in">metagenes_t4 &lt;-</span></span>
<span id="cb71-648"><a href="#cb71-648" aria-hidden="true" tabindex="-1"></a><span class="in">  HieraType::fit_metagene_scores(</span></span>
<span id="cb71-649"><a href="#cb71-649" aria-hidden="true" tabindex="-1"></a><span class="in">    markerslist = HieraType::markerslist_cd4tminor</span></span>
<span id="cb71-650"><a href="#cb71-650" aria-hidden="true" tabindex="-1"></a><span class="in">    ,counts_matrix = Matrix::t(sem[["RNA"]]@counts[,tcell_ids])</span></span>
<span id="cb71-651"><a href="#cb71-651" aria-hidden="true" tabindex="-1"></a><span class="in">    ,adjacency_matrix = as(sem@graphs$umapgrph</span></span>
<span id="cb71-652"><a href="#cb71-652" aria-hidden="true" tabindex="-1"></a><span class="in">                           ,"CsparseMatrix")[tcell_ids,tcell_ids]</span></span>
<span id="cb71-653"><a href="#cb71-653" aria-hidden="true" tabindex="-1"></a><span class="in">    ,prior_level_weights = model_t$post_probs$cd4t ## prior probability of cd4t</span></span>
<span id="cb71-654"><a href="#cb71-654" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb71-655"><a href="#cb71-655" aria-hidden="true" tabindex="-1"></a><span class="in">model_t4 &lt;- </span></span>
<span id="cb71-656"><a href="#cb71-656" aria-hidden="true" tabindex="-1"></a><span class="in">HieraType::cluster_metagenes(metagenes = metagenes_t4, prior_prob_level = model_t$post_probs$cd4t)</span></span>
<span id="cb71-657"><a href="#cb71-657" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-658"><a href="#cb71-658" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-659"><a href="#cb71-659" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-660"><a href="#cb71-660" aria-hidden="true" tabindex="-1"></a>A convenient way to simplify this process with minimal code is to use a <span class="in">`pipeline`</span> object.</span>
<span id="cb71-661"><a href="#cb71-661" aria-hidden="true" tabindex="-1"></a>A custom hierarchical cell typing pipeline could be created like this below, where</span>
<span id="cb71-662"><a href="#cb71-662" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-663"><a href="#cb71-663" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`markerslists`</span> is a named list, with a <span class="in">`markerslist`</span> for each stage of cell typing. "tmajor" will classify <span class="in">`cd8t`</span> vs. <span class="in">`cd4t`</span>, while "t8minor" and "t4minor" will subclassify those cell types respectively.</span>
<span id="cb71-664"><a href="#cb71-664" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`priors`</span>, a list specifying 'parent' celltyping stage. Here , <span class="in">`t4minor`</span> and <span class="in">`t8minor`</span> both need to run after "tmajor", as they will use the estimated probabilities of "<span class="in">`cd4t`</span>" and "<span class="in">`cd8t`</span>" as inputs.  </span>
<span id="cb71-665"><a href="#cb71-665" aria-hidden="true" tabindex="-1"></a>       "tmajor" runs first, and does not have prior probability inputs.</span>
<span id="cb71-666"><a href="#cb71-666" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`priors_category`</span>, a list specifying 'parent' /  major categories of each celltype. Here , <span class="in">`t4minor`</span> is a subset of <span class="in">`"cd4t"`</span>.  "tmajor" runs first, and does not have a prior.</span>
<span id="cb71-667"><a href="#cb71-667" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-670"><a href="#cb71-670" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-671"><a href="#cb71-671" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-672"><a href="#cb71-672" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-673"><a href="#cb71-673" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-674"><a href="#cb71-674" aria-hidden="true" tabindex="-1"></a>pipeline_tcell <span class="ot">&lt;-</span> </span>
<span id="cb71-675"><a href="#cb71-675" aria-hidden="true" tabindex="-1"></a>  <span class="fu">make_pipeline</span>(<span class="at">markerslists =</span> <span class="fu">list</span>(<span class="st">"tmajor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_tcellmajor</span>
<span id="cb71-676"><a href="#cb71-676" aria-hidden="true" tabindex="-1"></a>                    ,<span class="st">"t4minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_cd4tminor</span>
<span id="cb71-677"><a href="#cb71-677" aria-hidden="true" tabindex="-1"></a>                    ,<span class="st">"t8minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_cd8tminor</span>
<span id="cb71-678"><a href="#cb71-678" aria-hidden="true" tabindex="-1"></a>                    )</span>
<span id="cb71-679"><a href="#cb71-679" aria-hidden="true" tabindex="-1"></a>                 ,<span class="at">priors =</span> <span class="fu">list</span>(<span class="st">"t4minor"</span> <span class="ot">=</span> <span class="st">"tmajor"</span></span>
<span id="cb71-680"><a href="#cb71-680" aria-hidden="true" tabindex="-1"></a>                               ,<span class="st">"t8minor"</span> <span class="ot">=</span> <span class="st">"tmajor"</span>)</span>
<span id="cb71-681"><a href="#cb71-681" aria-hidden="true" tabindex="-1"></a>                 ,<span class="at">priors_category =</span> <span class="fu">list</span>(<span class="st">"t4minor"</span> <span class="ot">=</span> <span class="st">"cd4t"</span></span>
<span id="cb71-682"><a href="#cb71-682" aria-hidden="true" tabindex="-1"></a>                                         ,<span class="st">"t8minor"</span> <span class="ot">=</span> <span class="st">"cd8t"</span></span>
<span id="cb71-683"><a href="#cb71-683" aria-hidden="true" tabindex="-1"></a>                                         )</span>
<span id="cb71-684"><a href="#cb71-684" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb71-685"><a href="#cb71-685" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-688"><a href="#cb71-688" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-689"><a href="#cb71-689" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(pipeline_tcell)</span>
<span id="cb71-690"><a href="#cb71-690" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-691"><a href="#cb71-691" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-692"><a href="#cb71-692" aria-hidden="true" tabindex="-1"></a>This particular pipeline is also saved as a package dataset.</span>
<span id="cb71-695"><a href="#cb71-695" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-696"><a href="#cb71-696" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-697"><a href="#cb71-697" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-698"><a href="#cb71-698" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-699"><a href="#cb71-699" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(HieraType<span class="sc">::</span>pipeline_tcell, <span class="at">max.level=</span><span class="dv">2</span>)</span>
<span id="cb71-700"><a href="#cb71-700" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-701"><a href="#cb71-701" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-704"><a href="#cb71-704" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-705"><a href="#cb71-705" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-706"><a href="#cb71-706" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-707"><a href="#cb71-707" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-708"><a href="#cb71-708" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(pipeline_tcell, <span class="at">max.level=</span><span class="dv">2</span>)</span>
<span id="cb71-709"><a href="#cb71-709" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-710"><a href="#cb71-710" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-711"><a href="#cb71-711" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-712"><a href="#cb71-712" aria-hidden="true" tabindex="-1"></a>The code to run the pipeline is shown below.</span>
<span id="cb71-713"><a href="#cb71-713" aria-hidden="true" tabindex="-1"></a>Additional arguments to be passed to <span class="in">`fit_metagene_scores`</span> or <span class="in">`cluster_metagenes`</span> functions can be passed as additional arguments to <span class="in">`run_pipeline()`</span> .</span>
<span id="cb71-716"><a href="#cb71-716" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-717"><a href="#cb71-717" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-718"><a href="#cb71-718" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-719"><a href="#cb71-719" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-720"><a href="#cb71-720" aria-hidden="true" tabindex="-1"></a><span class="do">### Run a 'tcell'-typing pipeline, which </span></span>
<span id="cb71-721"><a href="#cb71-721" aria-hidden="true" tabindex="-1"></a><span class="do">### first classifies into T8/T4 categories (using the `tmajor` markerslist), </span></span>
<span id="cb71-722"><a href="#cb71-722" aria-hidden="true" tabindex="-1"></a><span class="do">### then runs subclassification for T4 and T8 using the t4minor and t8minor markerslists.</span></span>
<span id="cb71-723"><a href="#cb71-723" aria-hidden="true" tabindex="-1"></a>tcell_typing <span class="ot">&lt;-</span> </span>
<span id="cb71-724"><a href="#cb71-724" aria-hidden="true" tabindex="-1"></a><span class="fu">run_pipeline</span>(<span class="at">pipeline =</span> pipeline_tcell</span>
<span id="cb71-725"><a href="#cb71-725" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">counts_matrix =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[,tcell_ids])</span>
<span id="cb71-726"><a href="#cb71-726" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">adjacency_matrix =</span> <span class="fu">as</span>(sem<span class="sc">@</span>graphs<span class="sc">$</span>umapgrph</span>
<span id="cb71-727"><a href="#cb71-727" aria-hidden="true" tabindex="-1"></a>                           ,<span class="st">"CsparseMatrix"</span>)[tcell_ids,tcell_ids]</span>
<span id="cb71-728"><a href="#cb71-728" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">celltype_call_threshold =</span> <span class="fl">0.5</span></span>
<span id="cb71-729"><a href="#cb71-729" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">return_all_columns_postprobs =</span> <span class="cn">FALSE</span></span>
<span id="cb71-730"><a href="#cb71-730" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb71-731"><a href="#cb71-731" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-732"><a href="#cb71-732" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-733"><a href="#cb71-733" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-734"><a href="#cb71-734" aria-hidden="true" tabindex="-1"></a>We could take a look at individual <span class="in">`models`</span> and <span class="in">`metagenes`</span> objects if needed within our <span class="in">`tcell_typing`</span> result.</span>
<span id="cb71-735"><a href="#cb71-735" aria-hidden="true" tabindex="-1"></a>Below, we show that the outputted objects are the same in the pipeline format, as if we had ran them individually like in the method demonstration.</span>
<span id="cb71-736"><a href="#cb71-736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-739"><a href="#cb71-739" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-740"><a href="#cb71-740" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(model_t, tcell_typing<span class="sc">$</span>models<span class="sc">$</span>tmajor)</span>
<span id="cb71-741"><a href="#cb71-741" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(metagenes_t, tcell_typing<span class="sc">$</span>metagene_scores<span class="sc">$</span>tmajor)</span>
<span id="cb71-742"><a href="#cb71-742" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-743"><a href="#cb71-743" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-744"><a href="#cb71-744" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-745"><a href="#cb71-745" aria-hidden="true" tabindex="-1"></a>By default, two possible cell type classifications are returned in the overall <span class="in">`post_probs`</span> data.table object from our pipeline.</span>
<span id="cb71-746"><a href="#cb71-746" aria-hidden="true" tabindex="-1"></a>Below, <span class="in">`celltype_granular`</span> shows the most granular classification for each cell, carrying the 'best possible' classification at each stage of the pipeline.</span>
<span id="cb71-747"><a href="#cb71-747" aria-hidden="true" tabindex="-1"></a>For example, the first cell below was first identified as <span class="in">`cd8t`</span> being the best category.  The best subcategory of <span class="in">`cd8t`</span> was <span class="in">`cd8_exhausted`</span>, and the posterior probability score for <span class="in">`cd8t_exhausted`</span> was ~0.47.</span>
<span id="cb71-748"><a href="#cb71-748" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-749"><a href="#cb71-749" aria-hidden="true" tabindex="-1"></a>We set the the cutoff <span class="in">`celltype_call_threshold=0.5`</span> when we ran the pipeline, so that the <span class="in">`celltype_thresh`</span> classification shows the most granular cell type with probability scores *at least* this threshold.</span>
<span id="cb71-750"><a href="#cb71-750" aria-hidden="true" tabindex="-1"></a>The first cell below has <span class="in">`cd8_exhausted`</span> score ~0.47 (less than 0.5), and so takes the higher level category of <span class="in">`cd8t`</span>, which is above the threshold at ~0.99999.</span>
<span id="cb71-751"><a href="#cb71-751" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-754"><a href="#cb71-754" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-755"><a href="#cb71-755" aria-hidden="true" tabindex="-1"></a>tcell_typing<span class="sc">$</span>post_probs[[<span class="st">"tmajor"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb71-756"><a href="#cb71-756" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-757"><a href="#cb71-757" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-758"><a href="#cb71-758" aria-hidden="true" tabindex="-1"></a>We can remake this table with a different cutoff in mind as needed, depending on how conservative / aggressive we are comfortable being with the cell type call, using the </span>
<span id="cb71-759"><a href="#cb71-759" aria-hidden="true" tabindex="-1"></a><span class="in">`combine_postprob_tables`</span> function.</span>
<span id="cb71-760"><a href="#cb71-760" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-761"><a href="#cb71-761" aria-hidden="true" tabindex="-1"></a>As demonstration, we can look at the output if we were to raise our threshold from 0.5 to 0.8.  </span>
<span id="cb71-762"><a href="#cb71-762" aria-hidden="true" tabindex="-1"></a>We can "<span class="in">`return_all_columns`</span>" to show a more verbose result, with individual columns showing the joint probability scores for all possible cell type categories in each cell.</span>
<span id="cb71-763"><a href="#cb71-763" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-766"><a href="#cb71-766" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-767"><a href="#cb71-767" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-768"><a href="#cb71-768" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-769"><a href="#cb71-769" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-770"><a href="#cb71-770" aria-hidden="true" tabindex="-1"></a>pptables <span class="ot">&lt;-</span> </span>
<span id="cb71-771"><a href="#cb71-771" aria-hidden="true" tabindex="-1"></a>HieraType<span class="sc">::</span><span class="fu">combine_postprob_tables</span>(<span class="at">pipeline =</span> pipeline_tcell</span>
<span id="cb71-772"><a href="#cb71-772" aria-hidden="true" tabindex="-1"></a>                                   ,<span class="at">models =</span> tcell_typing<span class="sc">$</span>models</span>
<span id="cb71-773"><a href="#cb71-773" aria-hidden="true" tabindex="-1"></a>                                   ,<span class="at">celltype_call_threshold =</span> <span class="fl">0.8</span></span>
<span id="cb71-774"><a href="#cb71-774" aria-hidden="true" tabindex="-1"></a>                                   ,<span class="at">return_all_columns =</span> <span class="cn">TRUE</span>)[[<span class="st">"tmajor"</span>]]</span>
<span id="cb71-775"><a href="#cb71-775" aria-hidden="true" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">fwrite</span>(pptables, <span class="at">file=</span><span class="st">"figures/postprob_tables_demo.csv"</span>)</span>
<span id="cb71-776"><a href="#cb71-776" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-777"><a href="#cb71-777" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-778"><a href="#cb71-778" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-781"><a href="#cb71-781" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-782"><a href="#cb71-782" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-783"><a href="#cb71-783" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-784"><a href="#cb71-784" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-785"><a href="#cb71-785" aria-hidden="true" tabindex="-1"></a>HieraType<span class="sc">::</span><span class="fu">combine_postprob_tables</span>(<span class="at">pipeline =</span> pipeline_tcell</span>
<span id="cb71-786"><a href="#cb71-786" aria-hidden="true" tabindex="-1"></a>                                   ,<span class="at">models =</span> tcell_typing<span class="sc">$</span>models</span>
<span id="cb71-787"><a href="#cb71-787" aria-hidden="true" tabindex="-1"></a>                                   ,<span class="at">celltype_call_threshold =</span> <span class="fl">0.8</span></span>
<span id="cb71-788"><a href="#cb71-788" aria-hidden="true" tabindex="-1"></a>                                   ,<span class="at">return_all_columns =</span> <span class="cn">TRUE</span>)[[<span class="st">"tmajor"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb71-789"><a href="#cb71-789" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-790"><a href="#cb71-790" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-793"><a href="#cb71-793" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-794"><a href="#cb71-794" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-795"><a href="#cb71-795" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-796"><a href="#cb71-796" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-797"><a href="#cb71-797" aria-hidden="true" tabindex="-1"></a>pptables <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"figures/postprob_tables_demo.csv"</span>)</span>
<span id="cb71-798"><a href="#cb71-798" aria-hidden="true" tabindex="-1"></a>pptables[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>][]</span>
<span id="cb71-799"><a href="#cb71-799" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-800"><a href="#cb71-800" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-801"><a href="#cb71-801" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-802"><a href="#cb71-802" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-803"><a href="#cb71-803" aria-hidden="true" tabindex="-1"></a><span class="fu">### Examining and validating clustering results</span></span>
<span id="cb71-804"><a href="#cb71-804" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-805"><a href="#cb71-805" aria-hidden="true" tabindex="-1"></a>Now we can evaluate the cell typing performance.</span>
<span id="cb71-806"><a href="#cb71-806" aria-hidden="true" tabindex="-1"></a>In particular, we want to check that the genes we specified as the index marker and predictors are in fact marker genes in the corresponding cell types.</span>
<span id="cb71-807"><a href="#cb71-807" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-808"><a href="#cb71-808" aria-hidden="true" tabindex="-1"></a>First, we can look at a marker gene heatmap made with just the t-cells, and show the genes we specified as the most important 'index genes' for each of the <span class="in">`cd8t`</span> and <span class="in">`cd4t`</span> subtypes.</span>
<span id="cb71-809"><a href="#cb71-809" aria-hidden="true" tabindex="-1"></a>Looking closely, we see that CD8A and CD8B are much higher in <span class="in">`cd8_`</span> compared to <span class="in">`cd4_`</span> types.</span>
<span id="cb71-810"><a href="#cb71-810" aria-hidden="true" tabindex="-1"></a>Similarly, CD4 is higher expressed in <span class="in">`cd4_`</span> subtypes compared to <span class="in">`cd8_`</span> subtypes.</span>
<span id="cb71-811"><a href="#cb71-811" aria-hidden="true" tabindex="-1"></a>Looking at the specific index markers for each subtype, we see good agreement between marker gene specified for each subtype reflected in our heatmap. </span>
<span id="cb71-812"><a href="#cb71-812" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-813"><a href="#cb71-813" aria-hidden="true" tabindex="-1"></a><span class="in">```{r,eval=FALSE}</span></span>
<span id="cb71-814"><a href="#cb71-814" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-815"><a href="#cb71-815" aria-hidden="true" tabindex="-1"></a><span class="in">fct &lt;- clusterwise_foldchange_metrics(counts = sem[["RNA"]]@counts[,tcell_ids]</span></span>
<span id="cb71-816"><a href="#cb71-816" aria-hidden="true" tabindex="-1"></a><span class="in">                                      ,metadata = tcell_typing$post_probs[["tmajor"]]</span></span>
<span id="cb71-817"><a href="#cb71-817" aria-hidden="true" tabindex="-1"></a><span class="in">                                      ,cluster_column = "celltype_granular")</span></span>
<span id="cb71-818"><a href="#cb71-818" aria-hidden="true" tabindex="-1"></a><span class="in">idxs &lt;- unique(unname(c(unlist(lapply(HieraType::markerslist_cd4tminor, "[[", "index_marker"))</span></span>
<span id="cb71-819"><a href="#cb71-819" aria-hidden="true" tabindex="-1"></a><span class="in">               ,unlist(lapply(HieraType::markerslist_cd8tminor, "[[", "index_marker")))))</span></span>
<span id="cb71-820"><a href="#cb71-820" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-821"><a href="#cb71-821" aria-hidden="true" tabindex="-1"></a><span class="in">nms &lt;- c("cd4t", names(HieraType::markerslist_cd4tminor), "cd8t", names(HieraType::markerslist_cd8tminor))</span></span>
<span id="cb71-822"><a href="#cb71-822" aria-hidden="true" tabindex="-1"></a><span class="in">hmsubtype &lt;- marker_heatmap(fct, featsuse = c("CD8A", "CD8B", "CD4", idxs), clusterorder = nms, orient_diagonal = TRUE)</span></span>
<span id="cb71-823"><a href="#cb71-823" aria-hidden="true" tabindex="-1"></a><span class="in">print(hmsubtype)</span></span>
<span id="cb71-824"><a href="#cb71-824" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-825"><a href="#cb71-825" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-828"><a href="#cb71-828" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-829"><a href="#cb71-829" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-830"><a href="#cb71-830" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-831"><a href="#cb71-831" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-832"><a href="#cb71-832" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">plot=</span>hmsubtype, <span class="at">filename =</span> <span class="st">"figures/hm_tsubtype.png"</span>, <span class="at">width =</span> <span class="dv">1084</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">height =</span> <span class="dv">420</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>) , <span class="at">dpi =</span> <span class="dv">450</span>, <span class="at">units =</span> <span class="st">"px"</span>)</span>
<span id="cb71-833"><a href="#cb71-833" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(fct, <span class="at">file=</span><span class="st">"data/fc_t_subtypesonly.csv"</span>)</span>
<span id="cb71-834"><a href="#cb71-834" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-835"><a href="#cb71-835" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-836"><a href="#cb71-836" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-839"><a href="#cb71-839" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-840"><a href="#cb71-840" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-841"><a href="#cb71-841" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-842"><a href="#cb71-842" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-843"><a href="#cb71-843" aria-hidden="true" tabindex="-1"></a><span class="do">### The key index markers for cd4 subtypes</span></span>
<span id="cb71-844"><a href="#cb71-844" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(HieraType<span class="sc">::</span>markerslist_cd4tminor, <span class="st">"[["</span>, <span class="st">"index_marker"</span>)</span>
<span id="cb71-845"><a href="#cb71-845" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-846"><a href="#cb71-846" aria-hidden="true" tabindex="-1"></a><span class="do">### The key index markers for cd8 subtypes</span></span>
<span id="cb71-847"><a href="#cb71-847" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(HieraType<span class="sc">::</span>markerslist_cd8tminor, <span class="st">"[["</span>, <span class="st">"index_marker"</span>)</span>
<span id="cb71-848"><a href="#cb71-848" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-851"><a href="#cb71-851" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-852"><a href="#cb71-852" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-853"><a href="#cb71-853" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-854"><a href="#cb71-854" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-855"><a href="#cb71-855" aria-hidden="true" tabindex="-1"></a><span class="do">### The key index markers for cd4 subtypes</span></span>
<span id="cb71-856"><a href="#cb71-856" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(markerslist_cd4tminor, <span class="st">"[["</span>, <span class="st">"index_marker"</span>)</span>
<span id="cb71-857"><a href="#cb71-857" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-858"><a href="#cb71-858" aria-hidden="true" tabindex="-1"></a><span class="do">### The key index markers for cd8 subtypes</span></span>
<span id="cb71-859"><a href="#cb71-859" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(markerslist_cd8tminor, <span class="st">"[["</span>, <span class="st">"index_marker"</span>)</span>
<span id="cb71-860"><a href="#cb71-860" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-861"><a href="#cb71-861" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-862"><a href="#cb71-862" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb71-863"><a href="#cb71-863" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/hm_tsubtype.png" alt="Wide Image"&gt;</span>
<span id="cb71-864"><a href="#cb71-864" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb71-865"><a href="#cb71-865" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-866"><a href="#cb71-866" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-869"><a href="#cb71-869" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-870"><a href="#cb71-870" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-871"><a href="#cb71-871" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-872"><a href="#cb71-872" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-873"><a href="#cb71-873" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/hm_tsubtype.png"</span>)</span>
<span id="cb71-874"><a href="#cb71-874" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-875"><a href="#cb71-875" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-876"><a href="#cb71-876" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-877"><a href="#cb71-877" aria-hidden="true" tabindex="-1"></a>Here we can attach our T-cell annotations back onto our metadata, and remake our marker heatmap.</span>
<span id="cb71-878"><a href="#cb71-878" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-879"><a href="#cb71-879" aria-hidden="true" tabindex="-1"></a><span class="in">```{r,eval=FALSE}</span></span>
<span id="cb71-880"><a href="#cb71-880" aria-hidden="true" tabindex="-1"></a><span class="in">met &lt;- </span></span>
<span id="cb71-881"><a href="#cb71-881" aria-hidden="true" tabindex="-1"></a><span class="in">merge(data.table(sem@meta.data), tcell_typing$post_probs[["tmajor"]][,.(cell_ID, celltype = celltype_granular, best_score_granular)]</span></span>
<span id="cb71-882"><a href="#cb71-882" aria-hidden="true" tabindex="-1"></a><span class="in">      ,by="cell_ID", all.x=TRUE)</span></span>
<span id="cb71-883"><a href="#cb71-883" aria-hidden="true" tabindex="-1"></a><span class="in">met[is.na(celltype),celltype:=clusters_unsup]</span></span>
<span id="cb71-884"><a href="#cb71-884" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-885"><a href="#cb71-885" aria-hidden="true" tabindex="-1"></a><span class="in">fc_t_all &lt;- </span></span>
<span id="cb71-886"><a href="#cb71-886" aria-hidden="true" tabindex="-1"></a><span class="in">HieraType::clusterwise_foldchange_metrics(normed = sem[["RNA"]]@data</span></span>
<span id="cb71-887"><a href="#cb71-887" aria-hidden="true" tabindex="-1"></a><span class="in">                   ,metadata = met</span></span>
<span id="cb71-888"><a href="#cb71-888" aria-hidden="true" tabindex="-1"></a><span class="in">                   ,cluster_column = "celltype"</span></span>
<span id="cb71-889"><a href="#cb71-889" aria-hidden="true" tabindex="-1"></a><span class="in">                   )</span></span>
<span id="cb71-890"><a href="#cb71-890" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-891"><a href="#cb71-891" aria-hidden="true" tabindex="-1"></a><span class="in">hm_tcell &lt;- HieraType::marker_heatmap(fc_t_all</span></span>
<span id="cb71-892"><a href="#cb71-892" aria-hidden="true" tabindex="-1"></a><span class="in">                           ,extras = c("CD4", "CD8A", "CD8B")</span></span>
<span id="cb71-893"><a href="#cb71-893" aria-hidden="true" tabindex="-1"></a><span class="in">                          )</span></span>
<span id="cb71-894"><a href="#cb71-894" aria-hidden="true" tabindex="-1"></a><span class="in">print(hm_tcell)</span></span>
<span id="cb71-895"><a href="#cb71-895" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-896"><a href="#cb71-896" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-899"><a href="#cb71-899" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-900"><a href="#cb71-900" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-901"><a href="#cb71-901" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-902"><a href="#cb71-902" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-903"><a href="#cb71-903" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">plot=</span>hm_tcell, <span class="at">filename =</span> <span class="st">"figures/hm_tcell.png"</span>, <span class="at">width =</span> <span class="dv">1716</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">height =</span> <span class="dv">687</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>) , <span class="at">dpi =</span> <span class="dv">450</span>, <span class="at">units =</span> <span class="st">"px"</span>)</span>
<span id="cb71-904"><a href="#cb71-904" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(fc_t_all, <span class="at">file=</span><span class="st">"data/fc_tcells_allcells.csv"</span>)</span>
<span id="cb71-905"><a href="#cb71-905" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-906"><a href="#cb71-906" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-907"><a href="#cb71-907" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-908"><a href="#cb71-908" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb71-909"><a href="#cb71-909" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/hm_tcell_annotated.png" alt="Wide Image"&gt;</span>
<span id="cb71-910"><a href="#cb71-910" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb71-913"><a href="#cb71-913" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-914"><a href="#cb71-914" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-915"><a href="#cb71-915" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-916"><a href="#cb71-916" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-917"><a href="#cb71-917" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/hm_tcell_annotated.png"</span>)</span>
<span id="cb71-918"><a href="#cb71-918" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-919"><a href="#cb71-919" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-920"><a href="#cb71-920" aria-hidden="true" tabindex="-1"></a>Treating the 'posterior probability' as a confidence that the cell belongs to the corresponding cluster, we can show that </span>
<span id="cb71-921"><a href="#cb71-921" aria-hidden="true" tabindex="-1"></a>the proportion of cells that express the expected marker genes generally increases with higher confidence.</span>
<span id="cb71-922"><a href="#cb71-922" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-923"><a href="#cb71-923" aria-hidden="true" tabindex="-1"></a>In the figure below,  we compare the major <span class="in">`cd4t`</span> and <span class="in">`cd8t`</span> T cell types in terms of some of their marker genes which are specific to each. We see that CD8A and CD8B increase in <span class="in">`cd8t`</span> cells as we gain confidence in the cell type call, along with other <span class="in">`cd8t`</span> specific markers (GZMB, GNLY, PRF1).  </span>
<span id="cb71-924"><a href="#cb71-924" aria-hidden="true" tabindex="-1"></a>Similarly, we see that CD4 increases in <span class="in">`cd4t`</span> calls with higher posterior probability scores, along with <span class="in">`cd4t`</span> subtype-specific markers like FOXP3, CTLA4, TIGIT (<span class="in">`cd4_treg`</span> marker genes), IL17A and IL17F (<span class="in">`cd4_th17`</span> markers) and GATA3 (<span class="in">`cd4_th2`</span> marker).</span>
<span id="cb71-925"><a href="#cb71-925" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-926"><a href="#cb71-926" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-927"><a href="#cb71-927" aria-hidden="true" tabindex="-1"></a><span class="in">```{r,eval=FALSE}</span></span>
<span id="cb71-928"><a href="#cb71-928" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-929"><a href="#cb71-929" aria-hidden="true" tabindex="-1"></a><span class="in">p &lt;- </span></span>
<span id="cb71-930"><a href="#cb71-930" aria-hidden="true" tabindex="-1"></a><span class="in">marker_threshold_plot(normed = sem[["RNA"]]@data</span></span>
<span id="cb71-931"><a href="#cb71-931" aria-hidden="true" tabindex="-1"></a><span class="in">                      ,metadata = tcell_typing$models$tmajor$post_probs</span></span>
<span id="cb71-932"><a href="#cb71-932" aria-hidden="true" tabindex="-1"></a><span class="in">                      ,clusters = c("cd4t", "cd8t")</span></span>
<span id="cb71-933"><a href="#cb71-933" aria-hidden="true" tabindex="-1"></a><span class="in">                      ,markers = c("CD3D", "CD4", "IL17A", "IL17F", "GATA3", "FOXP3",  "IL2RA", "CTLA4", "TIGIT","CD8A", "CD8B", "GNLY", "GZMB","PRF1")</span></span>
<span id="cb71-934"><a href="#cb71-934" aria-hidden="true" tabindex="-1"></a><span class="in">                      ,thresholds = c(0.3, 0.5, 0.7, 0.9, 0.99, 0.995)</span></span>
<span id="cb71-935"><a href="#cb71-935" aria-hidden="true" tabindex="-1"></a><span class="in">                      ,score_column = "best_score"</span></span>
<span id="cb71-936"><a href="#cb71-936" aria-hidden="true" tabindex="-1"></a><span class="in">                      ,cluster_column = "best_class"</span></span>
<span id="cb71-937"><a href="#cb71-937" aria-hidden="true" tabindex="-1"></a><span class="in">                      )</span></span>
<span id="cb71-938"><a href="#cb71-938" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-939"><a href="#cb71-939" aria-hidden="true" tabindex="-1"></a><span class="in">p &lt;- p + theme(text=element_text(size=14,face='bold')) + </span></span>
<span id="cb71-940"><a href="#cb71-940" aria-hidden="true" tabindex="-1"></a><span class="in">  labs(title = "Proportion of cells expressing marker gene increases with \nposterior probability confidence scores")</span></span>
<span id="cb71-941"><a href="#cb71-941" aria-hidden="true" tabindex="-1"></a><span class="in">print(p)</span></span>
<span id="cb71-942"><a href="#cb71-942" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-943"><a href="#cb71-943" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-946"><a href="#cb71-946" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-947"><a href="#cb71-947" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-948"><a href="#cb71-948" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-949"><a href="#cb71-949" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">plot=</span>p , <span class="at">filename =</span> <span class="st">"figures/markerbarplot_tcells_major.png"</span>, <span class="at">width =</span> <span class="dv">1470</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">40</span>), <span class="at">height =</span> <span class="dv">967</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">units =</span> <span class="st">"px"</span>, <span class="at">dpi =</span> <span class="dv">450</span>)</span>
<span id="cb71-950"><a href="#cb71-950" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-951"><a href="#cb71-951" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-952"><a href="#cb71-952" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb71-953"><a href="#cb71-953" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/markerbarplot_tcells_major.png" alt="Wide Image"&gt;</span>
<span id="cb71-954"><a href="#cb71-954" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb71-955"><a href="#cb71-955" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-958"><a href="#cb71-958" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-959"><a href="#cb71-959" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-960"><a href="#cb71-960" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-961"><a href="#cb71-961" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-962"><a href="#cb71-962" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/markerbarplot_tcells_major.png"</span>)</span>
<span id="cb71-963"><a href="#cb71-963" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-964"><a href="#cb71-964" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-965"><a href="#cb71-965" aria-hidden="true" tabindex="-1"></a>Here's another look at the pairsplot for our metagene scores,this time coloring the cells based on their classification.  </span>
<span id="cb71-966"><a href="#cb71-966" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-967"><a href="#cb71-967" aria-hidden="true" tabindex="-1"></a><span class="in">metagene_pairsplot(metagenes_t, "yhat", obs = model_t$post_probs[,.(cell_ID, best_class)], colorvar = "best_class")</span></span>
<span id="cb71-968"><a href="#cb71-968" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-971"><a href="#cb71-971" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-972"><a href="#cb71-972" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-973"><a href="#cb71-973" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-974"><a href="#cb71-974" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-975"><a href="#cb71-975" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">"figures/pairshat_colored.png"</span>, <span class="at">width =</span> <span class="dv">1247</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">height =</span> <span class="dv">806</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">units =</span> <span class="st">"px"</span>, <span class="at">dpi =</span> <span class="dv">450</span>)</span>
<span id="cb71-976"><a href="#cb71-976" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-977"><a href="#cb71-977" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-978"><a href="#cb71-978" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb71-979"><a href="#cb71-979" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/pairshat_colored.png" alt="Wide Image"&gt;</span>
<span id="cb71-980"><a href="#cb71-980" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb71-981"><a href="#cb71-981" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-984"><a href="#cb71-984" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-985"><a href="#cb71-985" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-986"><a href="#cb71-986" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-987"><a href="#cb71-987" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-988"><a href="#cb71-988" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/pairshat_colored.png"</span>)</span>
<span id="cb71-989"><a href="#cb71-989" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-990"><a href="#cb71-990" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-991"><a href="#cb71-991" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-992"><a href="#cb71-992" aria-hidden="true" tabindex="-1"></a><span class="fu"># Example 2:  Broad immune cell classification  {#ex2}</span></span>
<span id="cb71-993"><a href="#cb71-993" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-994"><a href="#cb71-994" aria-hidden="true" tabindex="-1"></a>For the second example, we wish to ignore the unsupervised clustering and infer all cell types based entirely on the framework motivated here.  </span>
<span id="cb71-995"><a href="#cb71-995" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-996"><a href="#cb71-996" aria-hidden="true" tabindex="-1"></a>Using the full dataset, we can create a pipeline to first infer amongst 5 major cell type categories <span class="co">[</span><span class="ot">`immune`, `epithelial`, `endothelial`, `fibroblast`, `plasma`</span><span class="co">]</span> in a similar fashion as described in our first example.  </span>
<span id="cb71-997"><a href="#cb71-997" aria-hidden="true" tabindex="-1"></a>Then, we'll infer subtypes of the immune cell labels with categories including <span class="co">[</span><span class="ot">`tcell`,`bcell`,`macrophage`,`neutrophil`,`mast`,`monocyte`,`cDC`,`pDC`</span><span class="co">]</span>.</span>
<span id="cb71-998"><a href="#cb71-998" aria-hidden="true" tabindex="-1"></a>Finally, we'll do some basic subclassification of the <span class="in">`tcells`</span> into <span class="co">[</span><span class="ot">`cd4t`,`cd8t`</span><span class="co">]</span>.</span>
<span id="cb71-999"><a href="#cb71-999" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1000"><a href="#cb71-1000" aria-hidden="true" tabindex="-1"></a><span class="fu">## 'Just The Code' {#tldr2}</span></span>
<span id="cb71-1001"><a href="#cb71-1001" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1002"><a href="#cb71-1002" aria-hidden="true" tabindex="-1"></a>Here's the code to produce the set of cell types.</span>
<span id="cb71-1003"><a href="#cb71-1003" aria-hidden="true" tabindex="-1"></a>The next section will cover the details and discuss a few choices that were made.</span>
<span id="cb71-1004"><a href="#cb71-1004" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1005"><a href="#cb71-1005" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1008"><a href="#cb71-1008" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1009"><a href="#cb71-1009" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1010"><a href="#cb71-1010" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-1011"><a href="#cb71-1011" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1012"><a href="#cb71-1012" aria-hidden="true" tabindex="-1"></a><span class="do">### Pipeline, with </span></span>
<span id="cb71-1013"><a href="#cb71-1013" aria-hidden="true" tabindex="-1"></a><span class="do">### level-1 ('l1') being major categorization </span></span>
<span id="cb71-1014"><a href="#cb71-1014" aria-hidden="true" tabindex="-1"></a><span class="do">### level-2 ('l2') being classification of major immune cell categories</span></span>
<span id="cb71-1015"><a href="#cb71-1015" aria-hidden="true" tabindex="-1"></a><span class="do">### level-t ('lt') being T8/T4 classification</span></span>
<span id="cb71-1016"><a href="#cb71-1016" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1017"><a href="#cb71-1017" aria-hidden="true" tabindex="-1"></a>pipeline_io <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">make_pipeline</span>(</span>
<span id="cb71-1018"><a href="#cb71-1018" aria-hidden="true" tabindex="-1"></a>  <span class="at">markerslists =</span> <span class="fu">list</span>(<span class="st">"l1"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_l1</span>
<span id="cb71-1019"><a href="#cb71-1019" aria-hidden="true" tabindex="-1"></a>                      ,<span class="st">"l2"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_immunemajor</span>
<span id="cb71-1020"><a href="#cb71-1020" aria-hidden="true" tabindex="-1"></a>                      ,<span class="st">"lt"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_tcellmajor)</span>
<span id="cb71-1021"><a href="#cb71-1021" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors =</span> <span class="fu">list</span>(<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"l2"</span></span>
<span id="cb71-1022"><a href="#cb71-1022" aria-hidden="true" tabindex="-1"></a>                 ,<span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"l1"</span>) </span>
<span id="cb71-1023"><a href="#cb71-1023" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors_category =</span> <span class="fu">list</span>(<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"tcell"</span></span>
<span id="cb71-1024"><a href="#cb71-1024" aria-hidden="true" tabindex="-1"></a>                          ,<span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"immune"</span>)</span>
<span id="cb71-1025"><a href="#cb71-1025" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb71-1026"><a href="#cb71-1026" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1027"><a href="#cb71-1027" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1030"><a href="#cb71-1030" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1031"><a href="#cb71-1031" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1032"><a href="#cb71-1032" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-1033"><a href="#cb71-1033" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1034"><a href="#cb71-1034" aria-hidden="true" tabindex="-1"></a><span class="do">### run my pipeline (using all cells)</span></span>
<span id="cb71-1035"><a href="#cb71-1035" aria-hidden="true" tabindex="-1"></a>immune_typing <span class="ot">&lt;-</span> </span>
<span id="cb71-1036"><a href="#cb71-1036" aria-hidden="true" tabindex="-1"></a><span class="fu">run_pipeline</span>(<span class="at">pipeline =</span> pipeline_io</span>
<span id="cb71-1037"><a href="#cb71-1037" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">counts_matrix =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb71-1038"><a href="#cb71-1038" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">adjacency_matrix =</span> <span class="fu">as</span>(sem<span class="sc">@</span>graphs<span class="sc">$</span>umapgrph</span>
<span id="cb71-1039"><a href="#cb71-1039" aria-hidden="true" tabindex="-1"></a>                                    ,<span class="st">"CsparseMatrix"</span>)</span>
<span id="cb71-1040"><a href="#cb71-1040" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">celltype_call_threshold =</span> <span class="fl">0.5</span></span>
<span id="cb71-1041"><a href="#cb71-1041" aria-hidden="true" tabindex="-1"></a>             ,<span class="at">return_all_columns_postprobs =</span> <span class="cn">FALSE</span></span>
<span id="cb71-1042"><a href="#cb71-1042" aria-hidden="true" tabindex="-1"></a>             )</span>
<span id="cb71-1043"><a href="#cb71-1043" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1044"><a href="#cb71-1044" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1047"><a href="#cb71-1047" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1048"><a href="#cb71-1048" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1049"><a href="#cb71-1049" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-1050"><a href="#cb71-1050" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1051"><a href="#cb71-1051" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(immune_typing, <span class="st">"data/immune_typing.rds"</span>)</span>
<span id="cb71-1052"><a href="#cb71-1052" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1053"><a href="#cb71-1053" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1056"><a href="#cb71-1056" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1057"><a href="#cb71-1057" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-1058"><a href="#cb71-1058" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-1059"><a href="#cb71-1059" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1060"><a href="#cb71-1060" aria-hidden="true" tabindex="-1"></a><span class="co">#immune_typing &lt;- readRDS("data/immune_typing.rds") ## not needed</span></span>
<span id="cb71-1061"><a href="#cb71-1061" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1062"><a href="#cb71-1062" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1063"><a href="#cb71-1063" aria-hidden="true" tabindex="-1"></a><span class="fu">## Evaluation</span></span>
<span id="cb71-1064"><a href="#cb71-1064" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1065"><a href="#cb71-1065" aria-hidden="true" tabindex="-1"></a>Like previously, this <span class="in">`pipeline`</span>, and the corresponding <span class="in">`markerslists`</span> are saved as package datasets.</span>
<span id="cb71-1066"><a href="#cb71-1066" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1069"><a href="#cb71-1069" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1070"><a href="#cb71-1070" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1071"><a href="#cb71-1071" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-1072"><a href="#cb71-1072" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1073"><a href="#cb71-1073" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(HieraType<span class="sc">::</span>pipeline_io,<span class="at">max.level=</span><span class="dv">3</span>)</span>
<span id="cb71-1074"><a href="#cb71-1074" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1077"><a href="#cb71-1077" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1078"><a href="#cb71-1078" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-1079"><a href="#cb71-1079" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-1080"><a href="#cb71-1080" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1081"><a href="#cb71-1081" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(pipeline_io)</span>
<span id="cb71-1082"><a href="#cb71-1082" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1083"><a href="#cb71-1083" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1084"><a href="#cb71-1084" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1085"><a href="#cb71-1085" aria-hidden="true" tabindex="-1"></a>We can look at a heatmap showing the marker genes for our cell types to check whether the algorithm produced sensible results.</span>
<span id="cb71-1086"><a href="#cb71-1086" aria-hidden="true" tabindex="-1"></a>Because I have a lot of cell type categories, and some of them are not totally exclusive (for example, macrophage/monocyte are very similar and share many markers), </span>
<span id="cb71-1087"><a href="#cb71-1087" aria-hidden="true" tabindex="-1"></a>I don't necessarily expect all my cells to have very high confidence scores for every cell type; </span>
<span id="cb71-1088"><a href="#cb71-1088" aria-hidden="true" tabindex="-1"></a>I'm okay here with taking the most granular cell type calls and running with them.</span>
<span id="cb71-1089"><a href="#cb71-1089" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1090"><a href="#cb71-1090" aria-hidden="true" tabindex="-1"></a>These clusters show more granularity in a first pass than my original set of unsupervised clusters.</span>
<span id="cb71-1091"><a href="#cb71-1091" aria-hidden="true" tabindex="-1"></a>The heatmap below also shows very sane-looking results, with marker genes corresponding to many of the genes I specified in <span class="in">`markerslists`</span>. </span>
<span id="cb71-1092"><a href="#cb71-1092" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1095"><a href="#cb71-1095" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1096"><a href="#cb71-1096" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1097"><a href="#cb71-1097" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-1098"><a href="#cb71-1098" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1099"><a href="#cb71-1099" aria-hidden="true" tabindex="-1"></a>fc_l1 <span class="ot">&lt;-</span> </span>
<span id="cb71-1100"><a href="#cb71-1100" aria-hidden="true" tabindex="-1"></a>  <span class="fu">clusterwise_foldchange_metrics</span>(<span class="at">normed =</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>data</span>
<span id="cb71-1101"><a href="#cb71-1101" aria-hidden="true" tabindex="-1"></a>                     ,<span class="at">metadata =</span>immune_typing<span class="sc">$</span>post_probs[[<span class="dv">1</span>]]</span>
<span id="cb71-1102"><a href="#cb71-1102" aria-hidden="true" tabindex="-1"></a>                     ,<span class="at">cluster_column =</span> <span class="st">"celltype_granular"</span></span>
<span id="cb71-1103"><a href="#cb71-1103" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb71-1104"><a href="#cb71-1104" aria-hidden="true" tabindex="-1"></a>hm_l1 <span class="ot">&lt;-</span> <span class="fu">marker_heatmap</span>(fc_l1, <span class="at">extras =</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8A"</span>, <span class="st">"CD8B"</span>, <span class="st">"FOXP3"</span>))</span>
<span id="cb71-1105"><a href="#cb71-1105" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hm_l1)</span>
<span id="cb71-1106"><a href="#cb71-1106" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1107"><a href="#cb71-1107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1108"><a href="#cb71-1108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1111"><a href="#cb71-1111" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1112"><a href="#cb71-1112" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1113"><a href="#cb71-1113" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-1114"><a href="#cb71-1114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1115"><a href="#cb71-1115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1116"><a href="#cb71-1116" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">plot=</span>hm_l1, <span class="at">filename =</span> <span class="st">"figures/hm_l1.png"</span>, <span class="at">width =</span> <span class="dv">1280</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">height =</span> <span class="dv">433</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">units =</span> <span class="st">"px"</span>, <span class="at">dpi =</span> <span class="dv">450</span>)</span>
<span id="cb71-1117"><a href="#cb71-1117" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(fc_l1, <span class="st">"data/fc_l1.csv"</span>)</span>
<span id="cb71-1118"><a href="#cb71-1118" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1119"><a href="#cb71-1119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1120"><a href="#cb71-1120" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb71-1121"><a href="#cb71-1121" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/hm_l1.png" alt="Wide Image"&gt;</span>
<span id="cb71-1122"><a href="#cb71-1122" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb71-1123"><a href="#cb71-1123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1126"><a href="#cb71-1126" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1127"><a href="#cb71-1127" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1128"><a href="#cb71-1128" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-1129"><a href="#cb71-1129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1130"><a href="#cb71-1130" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/hm_l1.png"</span>)</span>
<span id="cb71-1131"><a href="#cb71-1131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1132"><a href="#cb71-1132" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1133"><a href="#cb71-1133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1134"><a href="#cb71-1134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1137"><a href="#cb71-1137" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1138"><a href="#cb71-1138" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1139"><a href="#cb71-1139" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb71-1140"><a href="#cb71-1140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1141"><a href="#cb71-1141" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(HieraType<span class="sc">::</span>markerslist_immunemajor, <span class="st">"[["</span>, <span class="st">"index_marker"</span>)</span>
<span id="cb71-1142"><a href="#cb71-1142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1143"><a href="#cb71-1143" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1144"><a href="#cb71-1144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1147"><a href="#cb71-1147" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1148"><a href="#cb71-1148" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb71-1149"><a href="#cb71-1149" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-1150"><a href="#cb71-1150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1151"><a href="#cb71-1151" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(markerslist_immunemajor, <span class="st">"[["</span>, <span class="st">"index_marker"</span>)</span>
<span id="cb71-1152"><a href="#cb71-1152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1153"><a href="#cb71-1153" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1154"><a href="#cb71-1154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1155"><a href="#cb71-1155" aria-hidden="true" tabindex="-1"></a>Here's a spatial plot of the cell types we've called so far.  We could stop here, or continue in the same manner for subclassifying any other cell types.</span>
<span id="cb71-1156"><a href="#cb71-1156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1157"><a href="#cb71-1157" aria-hidden="true" tabindex="-1"></a><span class="in">```{r,eval=FALSE}</span></span>
<span id="cb71-1158"><a href="#cb71-1158" aria-hidden="true" tabindex="-1"></a><span class="in">#| code-fold: true</span></span>
<span id="cb71-1159"><a href="#cb71-1159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1160"><a href="#cb71-1160" aria-hidden="true" tabindex="-1"></a><span class="in">met &lt;- merge(data.table(sem@meta.data)</span></span>
<span id="cb71-1161"><a href="#cb71-1161" aria-hidden="true" tabindex="-1"></a><span class="in">             ,immune_typing$post_probs$l1[,.(cell_ID, celltype_supervised = celltype_granular)]</span></span>
<span id="cb71-1162"><a href="#cb71-1162" aria-hidden="true" tabindex="-1"></a><span class="in">             ,by="cell_ID")</span></span>
<span id="cb71-1163"><a href="#cb71-1163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1164"><a href="#cb71-1164" aria-hidden="true" tabindex="-1"></a><span class="in">colorpal &lt;- c('#F0A0FF','#E0FF66','#C20088','#00998F','#426600'</span></span>
<span id="cb71-1165"><a href="#cb71-1165" aria-hidden="true" tabindex="-1"></a><span class="in">              , "#4C005C" </span></span>
<span id="cb71-1166"><a href="#cb71-1166" aria-hidden="true" tabindex="-1"></a><span class="in">              ,'#191919'</span></span>
<span id="cb71-1167"><a href="#cb71-1167" aria-hidden="true" tabindex="-1"></a><span class="in">              ,'#003380', '#005C31','#94FFB5','#990000','#0075DC','#FFA405')</span></span>
<span id="cb71-1168"><a href="#cb71-1168" aria-hidden="true" tabindex="-1"></a><span class="in">names(colorpal) &lt;- sort(unique(met[["celltype_supervised"]]))</span></span>
<span id="cb71-1169"><a href="#cb71-1169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1170"><a href="#cb71-1170" aria-hidden="true" tabindex="-1"></a><span class="in">xyplist_sup &lt;- </span></span>
<span id="cb71-1171"><a href="#cb71-1171" aria-hidden="true" tabindex="-1"></a><span class="in">  lapply(split(met, by="portion"), function(xx){</span></span>
<span id="cb71-1172"><a href="#cb71-1172" aria-hidden="true" tabindex="-1"></a><span class="in">    HieraType:::xyplot(cluster_column = "celltype_supervised", metadata = xx</span></span>
<span id="cb71-1173"><a href="#cb71-1173" aria-hidden="true" tabindex="-1"></a><span class="in">                       , ptsize = 0.1, cls = colorpal</span></span>
<span id="cb71-1174"><a href="#cb71-1174" aria-hidden="true" tabindex="-1"></a><span class="in">                       #,clusters = c("cd4t", "cd8t", "bcell")</span></span>
<span id="cb71-1175"><a href="#cb71-1175" aria-hidden="true" tabindex="-1"></a><span class="in">#                       , ptsize = 0.01, cls = colorpal</span></span>
<span id="cb71-1176"><a href="#cb71-1176" aria-hidden="true" tabindex="-1"></a><span class="in">                       ,plotfirst = c("epithelial", "endothelial", "fibroblast", "smooth_muscle", "plasma")</span></span>
<span id="cb71-1177"><a href="#cb71-1177" aria-hidden="true" tabindex="-1"></a><span class="in">                       ) + coord_fixed()</span></span>
<span id="cb71-1178"><a href="#cb71-1178" aria-hidden="true" tabindex="-1"></a><span class="in">  })  </span></span>
<span id="cb71-1179"><a href="#cb71-1179" aria-hidden="true" tabindex="-1"></a><span class="in">xyplot_sup &lt;- </span></span>
<span id="cb71-1180"><a href="#cb71-1180" aria-hidden="true" tabindex="-1"></a><span class="in">cowplot::plot_grid(plotlist = xyplist_sup,nrow = 1) + </span></span>
<span id="cb71-1181"><a href="#cb71-1181" aria-hidden="true" tabindex="-1"></a><span class="in">  theme(plot.background = element_rect(fill = 'white',color='white')) + </span></span>
<span id="cb71-1182"><a href="#cb71-1182" aria-hidden="true" tabindex="-1"></a><span class="in">  cowplot::panel_border(remove=TRUE) </span></span>
<span id="cb71-1183"><a href="#cb71-1183" aria-hidden="true" tabindex="-1"></a><span class="in">print(xyplot_sup)</span></span>
<span id="cb71-1184"><a href="#cb71-1184" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1185"><a href="#cb71-1185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1188"><a href="#cb71-1188" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1189"><a href="#cb71-1189" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1190"><a href="#cb71-1190" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-1191"><a href="#cb71-1191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1192"><a href="#cb71-1192" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb71-1193"><a href="#cb71-1193" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">plot=</span>xyplot_sup</span>
<span id="cb71-1194"><a href="#cb71-1194" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">filename =</span> <span class="st">"figures/xyplot_sup.png"</span></span>
<span id="cb71-1195"><a href="#cb71-1195" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">width=</span><span class="dv">1773</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">14</span>)</span>
<span id="cb71-1196"><a href="#cb71-1196" aria-hidden="true" tabindex="-1"></a>       , <span class="at">height =</span> <span class="dv">909</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">10</span>)</span>
<span id="cb71-1197"><a href="#cb71-1197" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">units =</span> <span class="st">"px"</span>)</span>
<span id="cb71-1198"><a href="#cb71-1198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1199"><a href="#cb71-1199" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1200"><a href="#cb71-1200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1201"><a href="#cb71-1201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1202"><a href="#cb71-1202" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb71-1203"><a href="#cb71-1203" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/xyplot_sup.png" alt="Wide Image"&gt;</span>
<span id="cb71-1204"><a href="#cb71-1204" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb71-1205"><a href="#cb71-1205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1206"><a href="#cb71-1206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1209"><a href="#cb71-1209" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1210"><a href="#cb71-1210" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1211"><a href="#cb71-1211" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-1212"><a href="#cb71-1212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1213"><a href="#cb71-1213" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/xyplot_sup.png"</span>)</span>
<span id="cb71-1214"><a href="#cb71-1214" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1215"><a href="#cb71-1215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1216"><a href="#cb71-1216" aria-hidden="true" tabindex="-1"></a>Below, we notice several interesting things about the spatial distribution of <span class="in">`cd8t`</span>, <span class="in">`cd4t`</span>, and <span class="in">`bcell`</span> types. The <span class="in">`cd4t`</span> cells tend to show higher colocalization with <span class="in">`bcell`</span>s in tertiary lymphoid type structures, while <span class="in">`cd8t`</span> spatial distribution are more dispersed throughout the tissue. This reflects the role of <span class="in">`cd8t`</span> cells in seeking out and killing tumor cells, compared to the role of <span class="in">`cd4t`</span> cells in local antigen presentation.</span>
<span id="cb71-1217"><a href="#cb71-1217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1218"><a href="#cb71-1218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1219"><a href="#cb71-1219" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb71-1220"><a href="#cb71-1220" aria-hidden="true" tabindex="-1"></a><span class="in">#| code-fold: true</span></span>
<span id="cb71-1221"><a href="#cb71-1221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1222"><a href="#cb71-1222" aria-hidden="true" tabindex="-1"></a><span class="in">xyplist_lymph &lt;- </span></span>
<span id="cb71-1223"><a href="#cb71-1223" aria-hidden="true" tabindex="-1"></a><span class="in">  lapply(split(met, by="portion"), function(xx){</span></span>
<span id="cb71-1224"><a href="#cb71-1224" aria-hidden="true" tabindex="-1"></a><span class="in">    HieraType:::xyplot(cluster_column = "celltype_supervised", metadata = xx</span></span>
<span id="cb71-1225"><a href="#cb71-1225" aria-hidden="true" tabindex="-1"></a><span class="in">                       , ptsize = 0.05, cls = colorpal</span></span>
<span id="cb71-1226"><a href="#cb71-1226" aria-hidden="true" tabindex="-1"></a><span class="in">                       ,clusters = c("cd4t", "cd8t", "bcell")</span></span>
<span id="cb71-1227"><a href="#cb71-1227" aria-hidden="true" tabindex="-1"></a><span class="in">                       ) + coord_fixed()</span></span>
<span id="cb71-1228"><a href="#cb71-1228" aria-hidden="true" tabindex="-1"></a><span class="in">  })</span></span>
<span id="cb71-1229"><a href="#cb71-1229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1230"><a href="#cb71-1230" aria-hidden="true" tabindex="-1"></a><span class="in">lymphplot &lt;- </span></span>
<span id="cb71-1231"><a href="#cb71-1231" aria-hidden="true" tabindex="-1"></a><span class="in">cowplot::plot_grid(</span></span>
<span id="cb71-1232"><a href="#cb71-1232" aria-hidden="true" tabindex="-1"></a><span class="in">xyplist_lymph[[1]]  + </span></span>
<span id="cb71-1233"><a href="#cb71-1233" aria-hidden="true" tabindex="-1"></a><span class="in">  facet_wrap(~celltype_supervised) + ggdark::dark_theme_bw() + theme(text=element_text(size=12,face='bold'))</span></span>
<span id="cb71-1234"><a href="#cb71-1234" aria-hidden="true" tabindex="-1"></a><span class="in">,xyplist_lymph[[2]]  + </span></span>
<span id="cb71-1235"><a href="#cb71-1235" aria-hidden="true" tabindex="-1"></a><span class="in">  facet_wrap(~celltype_supervised) + ggdark::dark_theme_bw() + theme(text=element_text(size=12,face='bold'))</span></span>
<span id="cb71-1236"><a href="#cb71-1236" aria-hidden="true" tabindex="-1"></a><span class="in">,nrow=2</span></span>
<span id="cb71-1237"><a href="#cb71-1237" aria-hidden="true" tabindex="-1"></a><span class="in">,rel_heights  = c(1.3,1)</span></span>
<span id="cb71-1238"><a href="#cb71-1238" aria-hidden="true" tabindex="-1"></a><span class="in">)  + </span></span>
<span id="cb71-1239"><a href="#cb71-1239" aria-hidden="true" tabindex="-1"></a><span class="in">  theme(plot.background = element_rect(fill = 'black',color='black')) + </span></span>
<span id="cb71-1240"><a href="#cb71-1240" aria-hidden="true" tabindex="-1"></a><span class="in">  cowplot::panel_border(remove=TRUE) </span></span>
<span id="cb71-1241"><a href="#cb71-1241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1242"><a href="#cb71-1242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1243"><a href="#cb71-1243" aria-hidden="true" tabindex="-1"></a><span class="in">print(lymphplot)</span></span>
<span id="cb71-1244"><a href="#cb71-1244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1245"><a href="#cb71-1245" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1246"><a href="#cb71-1246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1249"><a href="#cb71-1249" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1250"><a href="#cb71-1250" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1251"><a href="#cb71-1251" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-1252"><a href="#cb71-1252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1253"><a href="#cb71-1253" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">plot=</span>lymphplot</span>
<span id="cb71-1254"><a href="#cb71-1254" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">filename =</span> <span class="st">"./figures/lymphxyplot.png"</span></span>
<span id="cb71-1255"><a href="#cb71-1255" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">width =</span> <span class="dv">1042</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">24</span>), <span class="at">height =</span> <span class="dv">840</span> <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="dv">20</span>)</span>
<span id="cb71-1256"><a href="#cb71-1256" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">units =</span> <span class="st">'px'</span></span>
<span id="cb71-1257"><a href="#cb71-1257" aria-hidden="true" tabindex="-1"></a>       )</span>
<span id="cb71-1258"><a href="#cb71-1258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1259"><a href="#cb71-1259" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1260"><a href="#cb71-1260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1261"><a href="#cb71-1261" aria-hidden="true" tabindex="-1"></a>&lt;div class="wider-image"&gt;</span>
<span id="cb71-1262"><a href="#cb71-1262" aria-hidden="true" tabindex="-1"></a>  &lt;img src="./figures/lymphxyplot.png" alt="Wide Image"&gt;</span>
<span id="cb71-1263"><a href="#cb71-1263" aria-hidden="true" tabindex="-1"></a>&lt;/div&gt;</span>
<span id="cb71-1264"><a href="#cb71-1264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1265"><a href="#cb71-1265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1268"><a href="#cb71-1268" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb71-1269"><a href="#cb71-1269" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb71-1270"><a href="#cb71-1270" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb71-1271"><a href="#cb71-1271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1272"><a href="#cb71-1272" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/lymphxyplot.png"</span>)</span>
<span id="cb71-1273"><a href="#cb71-1273" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb71-1274"><a href="#cb71-1274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1275"><a href="#cb71-1275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1276"><a href="#cb71-1276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1277"><a href="#cb71-1277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1278"><a href="#cb71-1278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1279"><a href="#cb71-1279" aria-hidden="true" tabindex="-1"></a><span class="fu"># Conclusion</span></span>
<span id="cb71-1280"><a href="#cb71-1280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1281"><a href="#cb71-1281" aria-hidden="true" tabindex="-1"></a>We've discussed a method for cell typing that uses predefined sets of marker genes to build cell type specific metagene scores, which are probabilistically clustered into cell types.</span>
<span id="cb71-1282"><a href="#cb71-1282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1283"><a href="#cb71-1283" aria-hidden="true" tabindex="-1"></a>Using a predefined set of marker genes has several advantages.  For one, this allows users to prespecify the network of genes which are expected to have positive covariance within the cell type, and restricts any undesired influence of non-informative or noisy genes.  Because no assumption is made regarding the average expression or level of covariance between any pair of genes, we suspect that in some cases this approach may be more flexible and robust in the presence of batch effects or platform effects compared to using pre-trained models, data transfer, or reference profiles which may have been trained using different sample conditions or technologies.</span>
<span id="cb71-1284"><a href="#cb71-1284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1285"><a href="#cb71-1285" aria-hidden="true" tabindex="-1"></a>The NNLS regression framework for estimating celltype metagenes allows for improved prediction power by incorporating information from similar cells through an adjacency network. By restricting the direction of the regression coefficients, we've implicitly imposed an assumption of positive covariance between these markers within the corresponding cell type class.  Sign constraints have been shown to be as effective in some cases as explicit regularization <span class="co">[</span><span class="ot">@nnlsref</span><span class="co">]</span>.  </span>
<span id="cb71-1286"><a href="#cb71-1286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1287"><a href="#cb71-1287" aria-hidden="true" tabindex="-1"></a>We use a mixture of overfitted multivariate gaussian mixture models for probabilistic cell type classification.  This approach is effectively a non-parametric mixture model, where we assume that the distribution of metagene scores for each cell type can be approximated by a 'large-enough' mixture of multivariate gaussians <span class="co">[</span><span class="ot">@aragam2020identifiability</span><span class="co">]</span>. The simple selection of anchor cells based on score positivity, and ability to fine tune cell typing by controlling behavior of double positivity should allow for easy user control and flexibility. </span>
<span id="cb71-1288"><a href="#cb71-1288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1289"><a href="#cb71-1289" aria-hidden="true" tabindex="-1"></a>While we envision the method to be useful for many cell typing applications, there are several caveats and limitations users should keep in mind.</span>
<span id="cb71-1290"><a href="#cb71-1290" aria-hidden="true" tabindex="-1"></a>One primary assumption is that the user has a sense of the tissue type they are working with and the cell types that are likely to be present.  If the user were to run a "tcell typing pipeline"  on myeloid cells, they may see unexpected or poor results.  Similarly, the "io pipeline" we used in Example 2, may not perform well in a very different tissue type, like brain for example.  If we wanted to apply this method to a tissue type like 'brain', we may need to generate a new pipeline, with <span class="in">`markerslists`</span> for brain specific cell types like <span class="in">`neurons`</span> and <span class="in">`oligodendrocytes`</span>.  For cases where a major cell type exists in a dataset but does not have a <span class="in">`markerslists`</span> to be modeled, that cell type could manifest as cells which have "low posterior probability" for all of the other cell type classes included in the pipeline.</span>
<span id="cb71-1291"><a href="#cb71-1291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1292"><a href="#cb71-1292" aria-hidden="true" tabindex="-1"></a>There are several future improvements for this method we plan to work on.</span>
<span id="cb71-1293"><a href="#cb71-1293" aria-hidden="true" tabindex="-1"></a>Simultaneous estimation of metagene scores and clustering (perhaps through a regression-based mixture model) might enhance the power of metagene scores and classification by optimizing the cell allocations to the correct cluster simultaneously during model-fitting.  Exploring other frameworks outside of NNLS regression could potentially capture more flexible model structures with interaction and non-linearity.  </span>
<span id="cb71-1294"><a href="#cb71-1294" aria-hidden="true" tabindex="-1"></a>The ability or freedom to learn additional marker genes from the data, and an extension to a semi-supervised framework where cell types where marker genes are unknown a priori would also be useful.</span>
<span id="cb71-1295"><a href="#cb71-1295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1296"><a href="#cb71-1296" aria-hidden="true" tabindex="-1"></a>We've implemented the tools presented here in the R package <span class="co">[</span><span class="ot">HieraType</span><span class="co">](https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/HieraType)</span>.</span>
<span id="cb71-1297"><a href="#cb71-1297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1298"><a href="#cb71-1298" aria-hidden="true" tabindex="-1"></a><span class="fu"># Quick summary of package functions used in this post</span></span>
<span id="cb71-1299"><a href="#cb71-1299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1300"><a href="#cb71-1300" aria-hidden="true" tabindex="-1"></a>Below we call out a short list of <span class="in">`HieraType`</span> package functions which were used throughout this post.</span>
<span id="cb71-1301"><a href="#cb71-1301" aria-hidden="true" tabindex="-1"></a>Please see the examples throughout the rest of this document for more context on their usage.</span>
<span id="cb71-1302"><a href="#cb71-1302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1303"><a href="#cb71-1303" aria-hidden="true" tabindex="-1"></a>*Model Fitting and Clustering*</span>
<span id="cb71-1304"><a href="#cb71-1304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1305"><a href="#cb71-1305" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`make_markerlist()`</span> - Used to create a custom list of cell type markers.</span>
<span id="cb71-1306"><a href="#cb71-1306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1307"><a href="#cb71-1307" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`make_pipeline()`</span> - Used to create a pipeline, or recipe for hierarchical cell typing.  </span>
<span id="cb71-1308"><a href="#cb71-1308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1309"><a href="#cb71-1309" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`run_pipeline()`</span> - Used to run a hierarchical cell typing pipeline.  </span>
<span id="cb71-1310"><a href="#cb71-1310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1311"><a href="#cb71-1311" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`fit_metagene_scores()`</span> - This function fits cell type metagene scores.  Called as a standalone function or automatically within <span class="in">`run_pipeline`</span>, </span>
<span id="cb71-1312"><a href="#cb71-1312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1313"><a href="#cb71-1313" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`cluster_metagenes()`</span> - This function clusters cell type metagene scores using overfitted-gaussian mixture models. Also called within <span class="in">`run_pipeline`</span>.</span>
<span id="cb71-1314"><a href="#cb71-1314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1315"><a href="#cb71-1315" aria-hidden="true" tabindex="-1"></a>*Plotting and Diagnostics*</span>
<span id="cb71-1316"><a href="#cb71-1316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1317"><a href="#cb71-1317" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`xyplot()`</span>  - A utility function for plotting cell types in x/y space.</span>
<span id="cb71-1318"><a href="#cb71-1318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1319"><a href="#cb71-1319" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`metagene_pairsplot()`</span> - Plot pairwise distributions of metagene scores.</span>
<span id="cb71-1320"><a href="#cb71-1320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1321"><a href="#cb71-1321" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`metagene_coefficients_plot()`</span> - Plot for which genes contribute most to cell type metagene scores.</span>
<span id="cb71-1322"><a href="#cb71-1322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1323"><a href="#cb71-1323" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`marker_threshold_plot()`</span> -  Model diagnostics plot, show specific marker genes and the proportion of cells that express them by cluster, for increasing posterior probability thresholds.</span>
<span id="cb71-1324"><a href="#cb71-1324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1325"><a href="#cb71-1325" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`clusterwise_foldchange_metrics()`</span> - Computes the fold change, average expression, proportion of cells with at least 1 count for each gene and each cluster.  This is the data used as input to a marker gene heatmap.</span>
<span id="cb71-1326"><a href="#cb71-1326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1327"><a href="#cb71-1327" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="in">`marker_heatmap()`</span> -  Makes a bubble-style heatmap plot (clusters x marker genes) showing the proportion of cells with at least one count (bubble size) and relative average expression (bubble color).  Easy to add and control the genes shown in the heatmap, and can specify a fixed order for comparing multiple heatmaps.</span>
<span id="cb71-1328"><a href="#cb71-1328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1329"><a href="#cb71-1329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-1330"><a href="#cb71-1330" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="../../license.html">
<p>License</p>
</a>
  </li>  
</ul>
    <div class="cookie-consent-footer"><a href="#" id="open_preferences_center">Cookie Preferences</a></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>