<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Evelyn Metzger">

<title>6&nbsp; Cell Typing – A Guide to Whole Transcriptome CosMx&lt;sup&gt;®&lt;/sup&gt; SMI Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./pathways.html" rel="next">
<link href="./spatial-domains.html" rel="prev">
<link href="./assets/rocket.svg" rel="icon" type="image/svg+xml">
<script src="site_libs/cookie-consent/cookie-consent.js"></script>
<link href="site_libs/cookie-consent/cookie-consent.css" rel="stylesheet">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark-707d8167ce6003fca903bfe2be84ab7f.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" class="quarto-color-scheme-extra" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-00066ec49057a78057bb708cd7ecd022.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark-cb8585788486bab3a0512e4422583b7d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="site_libs/bootstrap/bootstrap-00066ec49057a78057bb708cd7ecd022.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme-extra" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/live-runtime/live-runtime.js" type="module"></script>
<link href="site_libs/quarto-contrib/live-runtime/live-runtime.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-29W4MW0Y2W"></script>

<script type="text/plain" cookie-consent="tracking">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-29W4MW0Y2W', { 'anonymize_ip': true});
</script>

<script type="text/javascript" charset="UTF-8">
document.addEventListener('DOMContentLoaded', function () {
cookieconsent.run({
  "notice_banner_type":"simple",
  "consent_type":"implied",
  "palette":"light",
  "language":"en",
  "page_load_consent_levels":["strictly-necessary","functionality","tracking","targeting"],
  "notice_banner_reject_button_hide":false,
  "preferences_center_close_button_hide":false,
  "website_name":""
  ,
"language":"en"
  });
});
</script> 
  
<script type="module" src="site_libs/quarto-ojs/quarto-ojs-runtime.js"></script>
<link href="site_libs/quarto-ojs/quarto-ojs.css" rel="stylesheet">


</head>

<body class="nav-sidebar floating nav-fixed quarto-light"><script id="quarto-html-before-body" type="application/javascript">
    const toggleBodyColorMode = (bsSheetEl) => {
      const mode = bsSheetEl.getAttribute("data-mode");
      const bodyEl = window.document.querySelector("body");
      if (mode === "dark") {
        bodyEl.classList.add("quarto-dark");
        bodyEl.classList.remove("quarto-light");
      } else {
        bodyEl.classList.add("quarto-light");
        bodyEl.classList.remove("quarto-dark");
      }
    }
    const toggleBodyColorPrimary = () => {
      const bsSheetEl = window.document.querySelector("link#quarto-bootstrap:not([rel=disabled-stylesheet])");
      if (bsSheetEl) {
        toggleBodyColorMode(bsSheetEl);
      }
    }
    const setColorSchemeToggle = (alternate) => {
      const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
      for (let i=0; i < toggles.length; i++) {
        const toggle = toggles[i];
        if (toggle) {
          if (alternate) {
            toggle.classList.add("alternate");
          } else {
            toggle.classList.remove("alternate");
          }
        }
      }
    };
    const toggleColorMode = (alternate) => {
      // Switch the stylesheets
      const primaryStylesheets = window.document.querySelectorAll('link.quarto-color-scheme:not(.quarto-color-alternate)');
      const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
      manageTransitions('#quarto-margin-sidebar .nav-link', false);
      if (alternate) {
        // note: dark is layered on light, we don't disable primary!
        enableStylesheet(alternateStylesheets);
        for (const sheetNode of alternateStylesheets) {
          if (sheetNode.id === "quarto-bootstrap") {
            toggleBodyColorMode(sheetNode);
          }
        }
      } else {
        disableStylesheet(alternateStylesheets);
        enableStylesheet(primaryStylesheets)
        toggleBodyColorPrimary();
      }
      manageTransitions('#quarto-margin-sidebar .nav-link', true);
      // Switch the toggles
      setColorSchemeToggle(alternate)
      // Hack to workaround the fact that safari doesn't
      // properly recolor the scrollbar when toggling (#1455)
      if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
        manageTransitions("body", false);
        window.scrollTo(0, 1);
        setTimeout(() => {
          window.scrollTo(0, 0);
          manageTransitions("body", true);
        }, 40);
      }
    }
    const disableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        stylesheet.rel = 'disabled-stylesheet';
      }
    }
    const enableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        if(stylesheet.rel !== 'stylesheet') { // for Chrome, which will still FOUC without this check
          stylesheet.rel = 'stylesheet';
        }
      }
    }
    const manageTransitions = (selector, allowTransitions) => {
      const els = window.document.querySelectorAll(selector);
      for (let i=0; i < els.length; i++) {
        const el = els[i];
        if (allowTransitions) {
          el.classList.remove('notransition');
        } else {
          el.classList.add('notransition');
        }
      }
    }
    const isFileUrl = () => {
      return window.location.protocol === 'file:';
    }
    const hasAlternateSentinel = () => {
      let styleSentinel = getColorSchemeSentinel();
      if (styleSentinel !== null) {
        return styleSentinel === "alternate";
      } else {
        return false;
      }
    }
    const setStyleSentinel = (alternate) => {
      const value = alternate ? "alternate" : "default";
      if (!isFileUrl()) {
        window.localStorage.setItem("quarto-color-scheme", value);
      } else {
        localAlternateSentinel = value;
      }
    }
    const getColorSchemeSentinel = () => {
      if (!isFileUrl()) {
        const storageValue = window.localStorage.getItem("quarto-color-scheme");
        return storageValue != null ? storageValue : localAlternateSentinel;
      } else {
        return localAlternateSentinel;
      }
    }
    const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
      const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
      const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
      let newTheme = '';
      if(authorPrefersDark) {
        newTheme = isAlternate ? baseTheme : alternateTheme;
      } else {
        newTheme = isAlternate ? alternateTheme : baseTheme;
      }
      const changeGiscusTheme = () => {
        // From: https://github.com/giscus/giscus/issues/336
        const sendMessage = (message) => {
          const iframe = document.querySelector('iframe.giscus-frame');
          if (!iframe) return;
          iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
        }
        sendMessage({
          setConfig: {
            theme: newTheme
          }
        });
      }
      const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
      if (isGiscussLoaded) {
        changeGiscusTheme();
      }
    };
    const authorPrefersDark = false;
    const darkModeDefault = authorPrefersDark;
      document.querySelector('link#quarto-text-highlighting-styles.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
      document.querySelector('link#quarto-bootstrap.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
    let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
    // Dark / light mode switch
    window.quartoToggleColorScheme = () => {
      // Read the current dark / light value
      let toAlternate = !hasAlternateSentinel();
      toggleColorMode(toAlternate);
      setStyleSentinel(toAlternate);
      toggleGiscusIfUsed(toAlternate, darkModeDefault);
      window.dispatchEvent(new Event('resize'));
    };
    // Switch to dark mode if need be
    if (hasAlternateSentinel()) {
      toggleColorMode(true);
    } else {
      toggleColorMode(false);
    }
  </script>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">A Guide to Whole Transcriptome CosMx<sup>®</sup> SMI Analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="./../../index.html"> <i class="bi bi-house-door" role="img" aria-label="Back to main website">
</i> 
<span class="menu-text">Back to Blog</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./quality-processing.html">Processing Data</a></li><li class="breadcrumb-item"><a href="./cell-typing.html"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Cell Typing</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./quality-processing.html">Processing Data</a></li><li class="breadcrumb-item"><a href="./cell-typing.html"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Cell Typing</span></a></li></ol></nav>
      <div class="quarto-title-block"><div><h1 class="title"><span id="sec-cell-typing" class="quarto-section-identifier"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Cell Typing</span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
            <p class="subtitle lead">Classifying cells into types</p>
                      </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliations</div>
    
      <div class="quarto-title-meta-contents">
      <p class="author">Evelyn Metzger <a href="https://orcid.org/0000-0002-4074-9003" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              Bruker Spatial Biology
            </p>
          <p class="affiliation">
              Github: <a href="https://github.com/eveilyeverafter" target="_blank">eveilyeverafter</a>
            </p>
        </div>
    </div>

  <div class="quarto-title-meta">

        
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      </a>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./how-to-use.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">How to Use This Book</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Getting Started</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction.html" class="sidebar-item-text sidebar-link"><i class="bi bi-rocket-takeoff-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-ingestion.html" class="sidebar-item-text sidebar-link"><i class="bi bi-hdd-network-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data Wrangling</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Processing Data</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quality-processing.html" class="sidebar-item-text sidebar-link"><i class="bi bi-clipboard2-data-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Quality Processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./processing.html" class="sidebar-item-text sidebar-link"><i class="bi bi-rulers" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Normalization, Reduction, and Clustering</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spatial-domains.html" class="sidebar-item-text sidebar-link"><i class="bi bi-diagram-3-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Spatial Domains</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cell-typing.html" class="sidebar-item-text sidebar-link active"><i class="bi bi-layout-wtf" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Cell Typing</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Tertiary Analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pathways.html" class="sidebar-item-text sidebar-link"><i class="bi bi-grid-1x2-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Pathways</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Conclusion</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./conclusion.html" class="sidebar-item-text sidebar-link"><i class="bi bi-check-square-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Conclusion</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Additonal Resources</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-ct-approaches" id="toc-sec-ct-approaches" class="nav-link active" data-scroll-target="#sec-ct-approaches"><span class="header-section-number">6.1</span> Cell Typing Approaches</a>
  <ul>
  <li><a href="#non-spatial-cell-typing-approaches" id="toc-non-spatial-cell-typing-approaches" class="nav-link" data-scroll-target="#non-spatial-cell-typing-approaches"><span class="header-section-number">6.1.1</span> Non-spatial Cell Typing Approaches</a></li>
  <li><a href="#spatially-informed-and-hybridization-aware-methods" id="toc-spatially-informed-and-hybridization-aware-methods" class="nav-link" data-scroll-target="#spatially-informed-and-hybridization-aware-methods"><span class="header-section-number">6.1.2</span> Spatially-Informed and Hybridization-Aware Methods</a></li>
  </ul></li>
  <li><a href="#cell-type-determination" id="toc-cell-type-determination" class="nav-link" data-scroll-target="#cell-type-determination"><span class="header-section-number">6.2</span> Cell Type Determination</a>
  <ul>
  <li><a href="#step-1-read-leiden-clusters" id="toc-step-1-read-leiden-clusters" class="nav-link" data-scroll-target="#step-1-read-leiden-clusters"><span class="header-section-number">6.2.1</span> Step 1: Read Leiden clusters</a></li>
  <li><a href="#step-2-run-hieratype" id="toc-step-2-run-hieratype" class="nav-link" data-scroll-target="#step-2-run-hieratype"><span class="header-section-number">6.2.2</span> Step 2: Run HieraType</a></li>
  <li><a href="#step-3-classify-remaining-cells-using-marker-genes" id="toc-step-3-classify-remaining-cells-using-marker-genes" class="nav-link" data-scroll-target="#step-3-classify-remaining-cells-using-marker-genes"><span class="header-section-number">6.2.3</span> Step 3: Classify remaining cells using marker genes</a></li>
  </ul></li>
  <li><a href="#cell-type-visualization" id="toc-cell-type-visualization" class="nav-link" data-scroll-target="#cell-type-visualization"><span class="header-section-number">6.3</span> Cell Type Visualization</a>
  <ul>
  <li><a href="#broad-cell-types" id="toc-broad-cell-types" class="nav-link" data-scroll-target="#broad-cell-types"><span class="header-section-number">6.3.1</span> Broad cell types</a></li>
  <li><a href="#fine-cell-types" id="toc-fine-cell-types" class="nav-link" data-scroll-target="#fine-cell-types"><span class="header-section-number">6.3.2</span> Fine cell types</a></li>
  </ul></li>
  <li><a href="#sec-integration" id="toc-sec-integration" class="nav-link" data-scroll-target="#sec-integration"><span class="header-section-number">6.4</span> Integration with Spatial Domains</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion"><span class="header-section-number">6.5</span> Conclusion</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">






<p>Okay, let’s chart the course for this crucial phase of our analysis expedition: cell typing. This is where we assign identities to the cell clusters discovered in the previous chapter, transforming abstract groups into recognizable biological entities like T cells, macrophages, or tumor cells.</p>
<p>If pre-processing was about plotting our trajectory through high-dimensional space, cell typing is akin to navigating a dense asteroid field. It’s a critical step, but one fraught with potential obstacles. Assigning definitive labels can be challenging, often requiring careful consideration of marker genes and existing biological knowledge. What constitutes a “correct” cell type label can even be subjective, depending entirely on the resolution needed for your specific biological questions. Are you aiming to distinguish broad categories like “immune” versus “tumor,” or do you need to pinpoint specific T cell subtypes? Like adjusting a microscope’s focus, the level of detail required dictates the approach. Navigating this field requires careful maneuvering and the right instruments. Autopilot hasn’t been invented just yet and so the best recommendation need manual input.</p>
<p>One common hazard is attempting to directly apply analysis pipelines designed for spatially-dissociated scRNA-seq. While scRNA-seq vignettes offer valuable starting points, CosMx SMI data, generated by direct <em>in situ</em> hybridization, has fundamentally different characteristics. Simply copy-pasting scRNA-seq methods without adaptation can lead to suboptimal or incomplete cell type assignments. Instead, we must employ or adapt methods that leverage the strengths of high-plex, direct hybridization data to achieve reliable annotations. In doing so it’s possible to achieve the most comprehensive and biologically-rich cell typing and sub-typing possible in spatial transcriptomics.</p>
<p>It’s also crucial to remember that cell typing, while essential, isn’t the final destination of our journey. It’s a powerful tool for organizing our cellular map, allowing us to ask spatially-aware questions. However, other organizational frameworks, such as classifying cells based on their spatial neighborhood or “niche,” are equally important for understanding the tissue’s complex ecosystem. Ultimately, cell typing is foundational: it establishes the structural anatomy of the dataset, paving the way for us to investigate its functional physiology in the next chapters.</p>
<p>In this chapter, we will navigate the cell typing process in the colon dataset. The main goal is to generate cell type labels that are “close” while also being approachable. I’ll begin (<a href="#sec-ct-approaches" class="quarto-xref"><span>Section 6.1</span></a>) by surveying some of the spatially unaware methods and then summarize the state of the direct hybridization-inspired approaches. Then I will guide you through the specific, hybrid workflow used for this colon dataset. We will start with the broad Leiden clusters defined in the previous chapter, refine them based on spatial observations, and assign primary identities using marker gene analysis. Finally, for the more complex immune populations where standard clustering often struggles, we will employ HieraType, a probabilistic method explicitly designed to handle the specific characteristics of CosMx SMI data.</p>
<div class="cell">
<details class="code-fold">
<summary>R code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-1"><pre class="sourceCode r code-annotation-code code-with-copy"><code class="sourceCode r"><span id="annotated-cell-1-1"><a href="#annotated-cell-1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"./preamble.R"</span>)</span>
<span id="annotated-cell-1-2"><a href="#annotated-cell-1-2" aria-hidden="true" tabindex="-1"></a>reticulate<span class="sc">::</span><span class="fu">source_python</span>(<span class="st">"./preamble.py"</span>)</span>
<span id="annotated-cell-1-3"><a href="#annotated-cell-1-3" aria-hidden="true" tabindex="-1"></a>analysis_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">getwd</span>(), <span class="st">"analysis_results"</span>)</span>
<span id="annotated-cell-1-4"><a href="#annotated-cell-1-4" aria-hidden="true" tabindex="-1"></a>input_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"input_files"</span>)</span>
<span id="annotated-cell-1-5"><a href="#annotated-cell-1-5" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"output_files"</span>)</span>
<span id="annotated-cell-1-6"><a href="#annotated-cell-1-6" aria-hidden="true" tabindex="-1"></a>analysis_asset_dir <span class="ot">&lt;-</span> <span class="st">"./assets/analysis_results"</span></span>
<span id="annotated-cell-1-7"><a href="#annotated-cell-1-7" aria-hidden="true" tabindex="-1"></a>ct_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(output_dir, <span class="st">"ct"</span>)</span>
<span id="annotated-cell-1-8"><a href="#annotated-cell-1-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(ct_dir)){</span>
<span id="annotated-cell-1-9"><a href="#annotated-cell-1-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(ct_dir, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="annotated-cell-1-10"><a href="#annotated-cell-1-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="annotated-cell-1-11"><a href="#annotated-cell-1-11" aria-hidden="true" tabindex="-1"></a>results_list_file <span class="ot">=</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"results_list.rds"</span>)</span>
<span id="annotated-cell-1-12"><a href="#annotated-cell-1-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(results_list_file)){</span>
<span id="annotated-cell-1-13"><a href="#annotated-cell-1-13" aria-hidden="true" tabindex="-1"></a>  results_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="annotated-cell-1-14"><a href="#annotated-cell-1-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="annotated-cell-1-15"><a href="#annotated-cell-1-15" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="annotated-cell-1-16"><a href="#annotated-cell-1-16" aria-hidden="true" tabindex="-1"></a>  results_list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(results_list_file)  </span>
<span id="annotated-cell-1-17"><a href="#annotated-cell-1-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="annotated-cell-1-18"><a href="#annotated-cell-1-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-1-19"><a href="#annotated-cell-1-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="annotated-cell-1-20"><a href="#annotated-cell-1-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="annotated-cell-1-21"><a href="#annotated-cell-1-21" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"./helpers.R"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<section id="sec-ct-approaches" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="sec-ct-approaches"><span class="header-section-number">6.1</span> Cell Typing Approaches</h2>
<section id="non-spatial-cell-typing-approaches" class="level3" data-number="6.1.1">
<h3 data-number="6.1.1" class="anchored" data-anchor-id="non-spatial-cell-typing-approaches"><span class="header-section-number">6.1.1</span> Non-spatial Cell Typing Approaches</h3>
<p>Before discussing spatially-aware or technology-specific approaches, let’s briefly cover common methods designed for non-spatial, single-cell RNA sequencing analysis. These techniques primarily leverage the gene expression matrix (<code>adata.X</code> and normalized layers) and the cluster assignments (e.g., <code>adata.obs['leiden']</code>) derived in the previous chapter. Many excellent vignettes exist for these methods in popular frameworks like scanpy (Python) and Seurat (R), providing a rough “10,000-foot view” of the data’s cellular composition.</p>
<p><strong>Cluster Marker Gene Identification (Leiden/Louvain):</strong> This is often a foundational step in scRNA-seq. After obtaining clusters (like the Leiden clusters from <a href="processing.html" class="quarto-xref"><span>Chapter 4</span></a>), differential expression tests are performed between each cluster and all other cells (or between pairs of clusters). This identifies genes significantly upregulated within a specific cluster. Tools like scanpy’s <code>rank_genes_groups</code> or Seurat’s <code>FindMarkers</code> are standard. The resulting lists of marker genes are then manually compared against known canonical markers from literature or cell atlases to assign putative cell type identities (e.g., observing high <em>CD3E</em>, <em>CD8A</em> suggests a T cell cluster; high <em>EPCAM</em>, <em>KRT19</em> suggests epithelial cells). While effective, this relies heavily on prior biological knowledge and the quality of the initial clustering. It also assumes that transcripts found in the cell do not arise from any spatial co-localization or imperfect cell-cell segmentation boundaries that can arise when assigning transcripts to cells.</p>
<p><strong>Nested / Hierarchical Clustering Analysis:</strong> Sometimes, broad initial clusters (like “Immune Cells” or “Fibroblasts”) contain sub-populations that are meaningful for your analysis. A common refinement is then to subset the data to include only cells from a specific Leiden cluster (or set of clusters) and then re-run the clustering process within that subset. Identifying marker genes for these sub-clusters can reveal finer cell types or states (<em>e.g.</em>, distinguishing CD4+ from CD8+ T cells, or identifying different fibroblast subtypes. This iterative approach allows for a hierarchical understanding of cell identity; however, the imperfect segmentation noted above is amplified in this approach. Consider a workflow where one uses HVGs to create an initial set of “Elementary” clusters and uses a marker gene-based approach to label these broad clusters. When subsetting down to, say, fibroblasts and rerunning the above HVG -&gt; Leiden -&gt; marker gene workflow, one may get non-fibroblast marker genes returned. The reason for this is twofold. First the general fibroblast marker genes are no longer as differentially expressed in the subset of Fibroblasts as they were in the full data and 2) if there are cell segmentation errors of fibroblast sub-populations that are spatially-associated with unrelated cell types – like plasma cells – the HVG algorithm may identify spurious gene markers as highly variable. The result would be markers like <em>JCHAIN</em> in fibroblasts that happen to be located near a germinal center, tertiary lymphoid structure, or similar. Thus, while this nested approach can be valuable, it should be used with caution. For more information on this overlap concept, see Dan McGuire’s <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/smiDE/" target="_blank">blog post</a> on smiDE.</p>
<p><strong>Reference-Based Annotation:</strong> Instead of relying solely on <em>de novo</em> marker finding, many methods leverage existing annotated single-cell datasets (atlases) as references. These algorithms typically project the query data onto the reference or use statistical models trained on the reference to directly assign cell type labels to your individual cells or clusters. Popular tools include:</p>
<ul>
<li><p><a href="https://scanpy.readthedocs.io/en/stable/tutorials/basics/integrating-data-using-ingest.html"><code>scanpy.tl.ingest</code></a>: Integrates query data onto an existing annotated AnnData object.</p></li>
<li><p>Seurat’s <a href="https://satijalab.org/seurat/reference/findtransferanchors"><code>FindTransferAnchors</code></a> and <a href="https://satijalab.org/seurat/reference/transferdata"><code>TransferData</code></a>: A widely used method in the R ecosystem.</p></li>
<li><p>InSituType’s fully supervised method: InSituType – while designed for spatial data – can be used as a fully supervised classifier. I’m including it here since many reference datasets and atlases at the time of writing are based on scRNA-seq data.</p></li>
</ul>
</section>
<section id="spatially-informed-and-hybridization-aware-methods" class="level3" data-number="6.1.2">
<h3 data-number="6.1.2" class="anchored" data-anchor-id="spatially-informed-and-hybridization-aware-methods"><span class="header-section-number">6.1.2</span> Spatially-Informed and Hybridization-Aware Methods</h3>
<p>Unlike standard scRNA-seq workflows, these approaches leverage the unique features of <em>in situ</em> data: specifically the spatial coordinates, the negative probe background, and the available protein data.</p>
<p><strong>Background-Aware Modeling</strong>. Methods like <code>InSituType</code> explicitly model the background noise using the negative control probes. This allows for more accurate classification of cells with low transcript counts that might be discarded or misclassified by standard scRNA-seq tools</p>
<p><strong>Spatial-Molecular Joint Inference</strong>. Tools like <a href="https://www.nature.com/articles/s41588-024-01664-3">Banksy</a> utilize the expression of a cell’s physical neighbors to inform its identity. This leverages the biological reality that cells of the same type often cluster spatially, helping to smooth out technical noise in individual cells.</p>
<p><strong>Multi-modal Integration</strong>. CosMx SMI often includes protein markers (e.g., CD45, PanCK) in the form of IF markers or – more recently – with true RNA and protein same-slide multiomics. “Dual-mode” typing strategies use these robust protein signals to define major lineages (e.g., “This cell is definitely CD45+ Immune”) before using the RNA data to determine the specific subtype (e.g., “It is a CD8+ T Cell”). Similarly, <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/multiomics-hieratype/#introduction">HieraType</a> is a new R package for supervised classification that can combine multi-omics information to cell type. HieraType also classifies RNA-only data (as we’ll see below).</p>
<p><strong>Reference-Based Annotation:</strong> Reference profiles based on CosMx SMI can be used to more quickly cell type new datasets. This approach can be much faster than cell typing <em>de novo</em> but requires such annotations to be available.</p>
</section>
</section>
<section id="cell-type-determination" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="cell-type-determination"><span class="header-section-number">6.2</span> Cell Type Determination</h2>
<p>Just like there are many paths through an asteroid field, there are many ways to navigate cell typing. I find the following approach makes a good starting place for any analysis as it’s relatively fast while still allowing for human intervention.</p>
<ol type="1">
<li>Use Leiden clusters identified in <a href="processing.html" class="quarto-xref"><span>Chapter 4</span></a> as the basis. 1.1. <em>Optional</em>: refine Leiden clusters based on discovered biology. For example, there may be some Leiden clusters that makes sense to merge together or some evidence that splitting a Leiden cluster into two or splitting a given Leiden cluster into multiple sub-clusters based on the spatial layout of cells.</li>
<li>Run HieraType to classify cells and use its integration function to merge its supervised classifications with the unsupervised Leiden clusters.</li>
<li>Of the remaining unclassified cells, use marker genes to infer the most likely cell type.</li>
</ol>
<section id="step-1-read-leiden-clusters" class="level3" data-number="6.2.1">
<h3 data-number="6.2.1" class="anchored" data-anchor-id="step-1-read-leiden-clusters"><span class="header-section-number">6.2.1</span> Step 1: Read Leiden clusters</h3>
<p>Load the data and make a copy of the <code>leiden_pr</code> metadata column that we’ll use.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Python Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> ad.read_h5ad(os.path.join(r.analysis_dir, <span class="st">"anndata-2-leiden.h5ad"</span>))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>cluster_renaming <span class="op">=</span> {}</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>initial <span class="op">=</span> <span class="st">"C"</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> name <span class="kw">in</span> adata.obs[<span class="st">'leiden_pr'</span>].unique():</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  cluster_renaming[name] <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>initial<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>name<span class="sc">}</span><span class="ss">"</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'leiden'</span>] <span class="op">=</span> adata.obs[<span class="st">'leiden_pr'</span>].<span class="bu">map</span>(cluster_renaming)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
</section>
<section id="step-2-run-hieratype" class="level3" data-number="6.2.2">
<h3 data-number="6.2.2" class="anchored" data-anchor-id="step-2-run-hieratype"><span class="header-section-number">6.2.2</span> Step 2: Run HieraType</h3>
<p>Install the HieraType package (and remotes), if needed. Then extract the lists of marker genes.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"mvtnorm"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"Nanostring-Biostats/CosMx-Analysis-Scratch-Space"</span>,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                         <span class="at">subdir =</span> <span class="st">"_code/HieraType"</span>, <span class="at">ref =</span> <span class="st">"Main"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(HieraType)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>markerg <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">c</span>(</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  HieraType<span class="sc">::</span>markerslist_l1</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_immune</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd8tminor</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd4tminor</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_tcellmajor</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>), <span class="st">"[["</span>, <span class="st">"predictors"</span>))</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>marker_genes <span class="ot">&lt;-</span> <span class="fu">as.character</span>(markerg)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>Convert data to R.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Python Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-4"><pre class="sourceCode python code-annotation-code code-with-copy code-annotated"><code class="sourceCode python"><span id="annotated-cell-4-1"><a href="#annotated-cell-4-1" aria-hidden="true" tabindex="-1"></a>total_counts_per_cell <span class="op">=</span> adata.layers[<span class="st">'counts'</span>].<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="annotated-cell-4-2"><a href="#annotated-cell-4-2" aria-hidden="true" tabindex="-1"></a>total_counts_per_cell <span class="op">=</span> np.asarray(total_counts_per_cell).flatten()</span>
<span id="annotated-cell-4-3"><a href="#annotated-cell-4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-4-4"><a href="#annotated-cell-4-4" aria-hidden="true" tabindex="-1"></a>total_counts_per_target <span class="op">=</span> adata.layers[<span class="st">'counts'</span>].<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="annotated-cell-4-5"><a href="#annotated-cell-4-5" aria-hidden="true" tabindex="-1"></a>total_counts_per_target <span class="op">=</span> np.asarray(total_counts_per_target).flatten()</span>
<span id="annotated-cell-4-6"><a href="#annotated-cell-4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-4-7"><a href="#annotated-cell-4-7" aria-hidden="true" tabindex="-1"></a>target_frequencies <span class="op">=</span> total_counts_per_target <span class="op">/</span> total_counts_per_target.<span class="bu">sum</span>()</span>
<span id="annotated-cell-4-8"><a href="#annotated-cell-4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-4-9"><a href="#annotated-cell-4-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Include all HVGs and marker genes (even if not in the HVGs)</span></span>
<span id="annotated-cell-4-10"><a href="#annotated-cell-4-10" aria-hidden="true" tabindex="-1"></a>marker_genes <span class="op">=</span> r.marker_genes</span>
<span id="annotated-cell-4-11"><a href="#annotated-cell-4-11" aria-hidden="true" tabindex="-1"></a>marker_mask <span class="op">=</span> adata.var.index.isin(marker_genes)</span>
<span id="annotated-cell-4-12"><a href="#annotated-cell-4-12" aria-hidden="true" tabindex="-1"></a>highly_variable_mask <span class="op">=</span> adata.var[<span class="st">'highly_variable'</span>].to_numpy()</span>
<span id="annotated-cell-4-13"><a href="#annotated-cell-4-13" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">'hieratype_target'</span>] <span class="op">=</span> (highly_variable_mask <span class="op">|</span> marker_mask)</span>
<span id="annotated-cell-4-14"><a href="#annotated-cell-4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-4-15"><a href="#annotated-cell-4-15" aria-hidden="true" tabindex="-1"></a>adata_ht <span class="op">=</span> adata[:, adata.var.hieratype_target].copy()</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="1">1</button><span id="annotated-cell-4-16" class="code-annotation-target"><a href="#annotated-cell-4-16" aria-hidden="true" tabindex="-1"></a>ht_counts <span class="op">=</span> adata_ht.layers[<span class="st">'counts'</span>].copy().astype(np.int64)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-4" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="16" data-code-annotation="1">In order to use this matrix in R, we need to convert the 32 bit integer counts into 64 bit integers.</span>
</dd>
</dl>
</div>
</div>
<p>Follow <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/multiomics-hieratype/#rnaonly">this tutorial</a> to refine Leiden clusters with immune cells.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>ht_counts <span class="ot">&lt;-</span> py<span class="sc">$</span>ht_counts</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ht_counts) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata_ht<span class="sc">$</span>obs_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ht_counts) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata_ht<span class="sc">$</span>var_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>ht_counts <span class="ot">&lt;-</span> <span class="fu">as</span>(ht_counts, <span class="st">"CsparseMatrix"</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>pca_embeddings <span class="ot">&lt;-</span> py<span class="sc">$</span>adata_ht<span class="sc">$</span>obsm[<span class="st">'X_pca_pr'</span>] <span class="co"># n_cells x n_pcs</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>target_frequencies <span class="ot">&lt;-</span> py<span class="sc">$</span>target_frequencies </span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(target_frequencies) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>var_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>target_frequencies <span class="ot">&lt;-</span> target_frequencies[<span class="fu">colnames</span>(ht_counts)] </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>total_counts_per_cell <span class="ot">&lt;-</span> py<span class="sc">$</span>total_counts_per_cell</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Assumes this was based on cosine.</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>sim_graph <span class="ot">&lt;-</span> py<span class="sc">$</span>adata_ht<span class="sc">$</span>obsp[<span class="st">'neighbors_pr_connectivities'</span>]</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>sim_graph <span class="ot">&lt;-</span> <span class="fu">as</span>(sim_graph, <span class="st">"CsparseMatrix"</span>)</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(sim_graph) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(sim_graph) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(ht_counts)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>Run the pipeline.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>pipeline <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">make_pipeline</span>(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">markerslists =</span> <span class="fu">list</span>(</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l1"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_l1</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"l2"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_immune</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_tcellmajor</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd4tminor</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd8tminor</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors =</span> <span class="fu">list</span>(</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"l1"</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"l2"</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors_category =</span> <span class="fu">list</span>(</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"immune"</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"tcell"</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"cd4t"</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"cd8t"</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>rnactobj <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">pipeline =</span> pipeline,</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts_matrix =</span> ht_counts,</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">adjacency_matrix =</span> sim_graph[<span class="fu">rownames</span>(ht_counts), <span class="fu">rownames</span>(ht_counts)],</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">totalcounts =</span> total_counts_per_cell,</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_wise_frequency =</span> target_frequencies</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(rnactobj, <span class="at">file=</span><span class="fu">file.path</span>(ct_dir, <span class="st">"rnactobj.rds"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>Merge the Leiden cluster calls with the HieraType results. There are several ways we could do this but for this tutorial we’ll use HieraType’s built-in <code>celltype_label_integration</code> function using the default parameters. <code>celltype_label_integration</code> integrates the immune celltype annotations from HieraType with unsupervised cluster labels. The goal of this step is to enable granular detection of novel or tissue-specific cell types (unsupervised), along with well known immune cell types (HieraType). The logic for merging these annotations works like this:</p>
<ol type="1">
<li>All cells which are annotated with an immune-celltype HieraType label keep their label.</li>
<li>All cells with a non-immune HieraType label take the unsupervised cluster label.</li>
<li>Unsupervised clusters that had a high rate of cells taking immune labels (say 90%), and make up only a small proportion of all cells (say 1%) after step 1 are ‘dissolved’. These dissoved cells are assigned the most common cluster label (could be immune or unsupervised) amongst most the similar neighboring cells. Here we used the neighbor cells from the adjacency matrix that was used to run HieraType.</li>
</ol>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-7"><pre class="sourceCode r code-annotation-code code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-7-1"><a href="#annotated-cell-7-1" aria-hidden="true" tabindex="-1"></a>cell_type_dt <span class="ot">&lt;-</span> rnactobj<span class="sc">$</span>post_probs<span class="sc">$</span>l1</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="1">1</button><span id="annotated-cell-7-2" class="code-annotation-target"><a href="#annotated-cell-7-2" aria-hidden="true" tabindex="-1"></a>cell_type_dt <span class="ot">&lt;-</span> <span class="fu">rename</span>(cell_type_dt, <span class="at">cell_id =</span> cell_ID,</span>
<span id="annotated-cell-7-3"><a href="#annotated-cell-7-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">Hieratype_celltype_fine =</span> celltype_granular)</span>
<span id="annotated-cell-7-4"><a href="#annotated-cell-7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-7-5"><a href="#annotated-cell-7-5" aria-hidden="true" tabindex="-1"></a>to_merge <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs <span class="sc">%&gt;%</span> <span class="fu">select</span>(leiden)</span>
<span id="annotated-cell-7-6"><a href="#annotated-cell-7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-7-7"><a href="#annotated-cell-7-7" aria-hidden="true" tabindex="-1"></a>cell_type_dt<span class="sc">$</span>Leiden <span class="ot">&lt;-</span> to_merge<span class="sc">$</span>leiden[<span class="fu">match</span>(<span class="fu">rownames</span>(to_merge), cell_type_dt<span class="sc">$</span>cell_id)]</span>
<span id="annotated-cell-7-8"><a href="#annotated-cell-7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-7-9"><a href="#annotated-cell-7-9" aria-hidden="true" tabindex="-1"></a>cell_type_dt <span class="ot">&lt;-</span> <span class="fu">celltype_label_integration</span>(</span>
<span id="annotated-cell-7-10"><a href="#annotated-cell-7-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">metadata =</span> cell_type_dt, </span>
<span id="annotated-cell-7-11"><a href="#annotated-cell-7-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">adjacency_mat =</span> sim_graph[<span class="fu">rownames</span>(ht_counts), <span class="fu">rownames</span>(ht_counts)],</span>
<span id="annotated-cell-7-12"><a href="#annotated-cell-7-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">cellid_colname =</span> <span class="st">'cell_id'</span>,</span>
<span id="annotated-cell-7-13"><a href="#annotated-cell-7-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">supervised_colname =</span> <span class="st">'Hieratype_celltype_fine'</span>,</span>
<span id="annotated-cell-7-14"><a href="#annotated-cell-7-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">unsupervised_colname =</span> <span class="st">'Leiden'</span>)</span>
<span id="annotated-cell-7-15"><a href="#annotated-cell-7-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-7-16"><a href="#annotated-cell-7-16" aria-hidden="true" tabindex="-1"></a>cell_type_dt <span class="ot">&lt;-</span> <span class="fu">rename</span>(cell_type_dt, <span class="at">merged =</span> celltype)</span>
<span id="annotated-cell-7-17"><a href="#annotated-cell-7-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-7-18"><a href="#annotated-cell-7-18" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(cell_type_dt, <span class="at">file=</span><span class="fu">file.path</span>(ct_dir, <span class="st">"cell_type_dt.rds"</span>))</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-7" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="2" data-code-annotation="1">cell_id here is the row name from the anndata object (as opposed to the c_i_j_k cell_ID notation that is sometimes used).</span>
</dd>
</dl>
</div>
</div>
<p>Save the new calumns back to the anndata object.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Python Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>cell_type_df <span class="op">=</span> r.cell_type_dt</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>adata.obs <span class="op">=</span> adata.obs.join(cell_type_df[[<span class="st">'cell_id'</span>, <span class="st">'Hieratype_celltype_fine'</span>, <span class="st">'merged'</span>]].set_index(<span class="st">'cell_id'</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
</section>
<section id="step-3-classify-remaining-cells-using-marker-genes" class="level3" data-number="6.2.3">
<h3 data-number="6.2.3" class="anchored" data-anchor-id="step-3-classify-remaining-cells-using-marker-genes"><span class="header-section-number">6.2.3</span> Step 3: Classify remaining cells using marker genes</h3>
<p>For the cells still assigned a Leiden cluster label, run a marker gene analysis to determine the dominant cell type per cluster. We’ll subset larger clusters down to 10,000 cells to speed up the computation. Since scanpy’s <code>tl.rank_genes_groups</code> function needs normalizaed data, we’ll use the log1p-transformed matrix that we generated in <a href="processing.html" class="quarto-xref"><span>Chapter 4</span></a>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-9"><pre class="sourceCode python code-annotation-code code-with-copy code-annotated"><code class="sourceCode python"><span id="annotated-cell-9-1"><a href="#annotated-cell-9-1" aria-hidden="true" tabindex="-1"></a>bdata <span class="op">=</span> adata[adata.obs[<span class="st">'leiden'</span>]<span class="op">==</span>adata.obs[<span class="st">'merged'</span>], :].copy()</span>
<span id="annotated-cell-9-2"><a href="#annotated-cell-9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-9-3"><a href="#annotated-cell-9-3" aria-hidden="true" tabindex="-1"></a>random_seed <span class="op">=</span> <span class="dv">98103</span></span>
<span id="annotated-cell-9-4"><a href="#annotated-cell-9-4" aria-hidden="true" tabindex="-1"></a>np.random.seed(random_seed)</span>
<span id="annotated-cell-9-5"><a href="#annotated-cell-9-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-9-6"><a href="#annotated-cell-9-6" aria-hidden="true" tabindex="-1"></a>target_cells_per_cluster <span class="op">=</span> <span class="dv">10000</span></span>
<span id="annotated-cell-9-7"><a href="#annotated-cell-9-7" aria-hidden="true" tabindex="-1"></a>cells_to_keep_indices <span class="op">=</span> []</span>
<span id="annotated-cell-9-8"><a href="#annotated-cell-9-8" aria-hidden="true" tabindex="-1"></a>cluster_counts <span class="op">=</span> bdata.obs[<span class="st">'merged'</span>].value_counts()</span>
<span id="annotated-cell-9-9"><a href="#annotated-cell-9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-9-10"><a href="#annotated-cell-9-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> cluster_label <span class="kw">in</span> cluster_counts.index:</span>
<span id="annotated-cell-9-11"><a href="#annotated-cell-9-11" aria-hidden="true" tabindex="-1"></a>    cluster_cells <span class="op">=</span> bdata.obs_names[bdata.obs[<span class="st">'merged'</span>] <span class="op">==</span> cluster_label]</span>
<span id="annotated-cell-9-12"><a href="#annotated-cell-9-12" aria-hidden="true" tabindex="-1"></a>    num_cells_in_cluster <span class="op">=</span> <span class="bu">len</span>(cluster_cells)</span>
<span id="annotated-cell-9-13"><a href="#annotated-cell-9-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-9-14"><a href="#annotated-cell-9-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> num_cells_in_cluster <span class="op">&gt;</span> target_cells_per_cluster:</span>
<span id="annotated-cell-9-15"><a href="#annotated-cell-9-15" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Cluster '</span><span class="sc">{</span>cluster_label<span class="sc">}</span><span class="ss">' has </span><span class="sc">{</span>num_cells_in_cluster<span class="sc">}</span><span class="ss"> cells, subsampling to </span><span class="sc">{</span>target_cells_per_cluster<span class="sc">}</span><span class="ss">."</span>)</span>
<span id="annotated-cell-9-16"><a href="#annotated-cell-9-16" aria-hidden="true" tabindex="-1"></a>        subsampled_cells <span class="op">=</span> np.random.choice(</span>
<span id="annotated-cell-9-17"><a href="#annotated-cell-9-17" aria-hidden="true" tabindex="-1"></a>            cluster_cells,</span>
<span id="annotated-cell-9-18"><a href="#annotated-cell-9-18" aria-hidden="true" tabindex="-1"></a>            size<span class="op">=</span>target_cells_per_cluster,</span>
<span id="annotated-cell-9-19"><a href="#annotated-cell-9-19" aria-hidden="true" tabindex="-1"></a>            replace<span class="op">=</span><span class="va">False</span></span>
<span id="annotated-cell-9-20"><a href="#annotated-cell-9-20" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="annotated-cell-9-21"><a href="#annotated-cell-9-21" aria-hidden="true" tabindex="-1"></a>        cells_to_keep_indices.extend(subsampled_cells)</span>
<span id="annotated-cell-9-22"><a href="#annotated-cell-9-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="annotated-cell-9-23"><a href="#annotated-cell-9-23" aria-hidden="true" tabindex="-1"></a>      cells_to_keep_indices.extend(cluster_cells)</span>
<span id="annotated-cell-9-24"><a href="#annotated-cell-9-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-9-25"><a href="#annotated-cell-9-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-9-26"><a href="#annotated-cell-9-26" aria-hidden="true" tabindex="-1"></a>bdata_subsampled <span class="op">=</span> bdata[cells_to_keep_indices, :].copy()</span>
<span id="annotated-cell-9-27"><a href="#annotated-cell-9-27" aria-hidden="true" tabindex="-1"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="1">1</button><span id="annotated-cell-9-28" class="code-annotation-target"><a href="#annotated-cell-9-28" aria-hidden="true" tabindex="-1"></a>bdata_subsampled.X <span class="op">=</span> bdata_subsampled.layers[<span class="st">'TC'</span>].copy()</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-9" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="28" data-code-annotation="1">If you do not have the log1p-transformed total counts layer ‘TC’ saved, run <code>sc.pp.normalize_total</code> and <code>sc.pp.log1p</code> (see <span class="quarto-unresolved-ref">?sec-tc-workflow</span>).</span>
</dd>
</dl>
</div>
</div>
<p>Run <code>rank_genes_groups</code> and save results.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>ct_dir <span class="op">=</span> r.ct_dir</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>sc.settings.figdir <span class="op">=</span> ct_dir</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(bdata_subsampled, <span class="st">'leiden'</span>, method<span class="op">=</span><span class="st">'wilcoxon'</span>,</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  solver<span class="op">=</span><span class="st">'saga'</span>, max_iter<span class="op">=</span><span class="dv">200</span>, pts<span class="op">=</span><span class="va">True</span>) <span class="co"># ~an hour using logreg</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups(bdata_subsampled, n_genes<span class="op">=</span><span class="dv">8</span>, sharey<span class="op">=</span><span class="va">True</span>, fontsize<span class="op">=</span><span class="dv">10</span>, save<span class="op">=</span><span class="st">"_marker_gene_ranks_global.png"</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups_dotplot(bdata_subsampled, n_genes<span class="op">=</span><span class="dv">8</span>, save<span class="op">=</span><span class="st">"_marker_gene_dotplot_global.png"</span>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-5-subset-marker-genes.h5ad"</span>)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>bdata_subsampled.write_h5ad(</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>  filename,</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  compression<span class="op">=</span>hdf5plugin.FILTERS[<span class="st">"zstd"</span>],</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  compression_opts<span class="op">=</span>hdf5plugin.Zstd(clevel<span class="op">=</span><span class="dv">5</span>).filter_options</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The next task is determine from the marker gene results the dominant cell types that are likely found within each Leiden cluster. While some Leiden clusters will have a high proportion of cells that are of the same type or state, keep in mind that other Leiden clusters might be composites of cell types. The tools we have for this exercise are:</p>
<ul>
<li>visual aids from the marker gene results and</li>
<li>compare with external sources. The literature and subject matter expertise is the gold-standard for this.</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<div id="fig-global-marker-ranks" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-global-marker-ranks-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/ct/rank_genes_groups_leiden_marker_gene_ranks_global.png" class="img-fluid figure-img" width="1030">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-global-marker-ranks-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.1: Marker gene ranks by Leiden cluster (global).
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-global-marker-dot" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-global-marker-dot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/ct/dotplot__marker_gene_dotplot_global.png" class="img-fluid figure-img" width="1608">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-global-marker-dot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.2: Marker gene dot plot by Leiden cluster (global).
</figcaption>
</figure>
</div>
</div>
</div>
<p>Examine the clusters by extracting the top 25 markers and their relative scores and comparing to the literature. Large Language Models can provide supportive information here but it’s always recommended to use LLMs with caution and to verify any of the results. In other words, LLMs can be fast but at the cost of accuracy. In the example below, I created an LLM prompt that includes the top marker gene statistics. Let’s see what it generates.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>marker_dict, marker_df <span class="op">=</span> format_ranked_genes(bdata_subsampled, <span class="dv">25</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>csv_file_path <span class="op">=</span> os.path.join(ct_dir, <span class="st">"colon_cancer_top_marker_genes_global.csv"</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>marker_df.to_csv(csv_file_path, index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"I am analyzing a CosMx SMI dataset of from Colon cancer and have run Leiden clustering.</span><span class="ch">\n</span><span class="ss">I have already classified several of the immune cells within the larger dataset and now I would like your help in predicting the most like cell type given the marker gene profile from this subset of data.</span><span class="ch">\n</span><span class="ss">Below is a dictionary of results where the key represents the Leiden cluster</span><span class="ch">\n</span><span class="ss">name and its corresponding item is a list of ordered tuples where each tuple</span><span class="ch">\n</span><span class="ss">provides the marker name and the relative score for that marker.</span><span class="ch">\n\n</span><span class="ss"> Please provide a table with the columns that represent 1. inferred cell type 2. your confidence in that cell type 3. up to five markers from the list that are key to your result and 4. a brief justification of your result. Please present the results in a markdown table format to make it easier to copy and paste.</span><span class="ch">\n\n</span><span class="sc">{</span>marker_dict<span class="sc">}</span><span class="ss">"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="tbl-leiden-global" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-leiden-global-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;6.1: Estimating most likely cell type from Leiden clusters’ marker gene rankings. Note: predictions were generated with the aid of an LLM.
</figcaption>
<div aria-describedby="tbl-leiden-global-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<colgroup>
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Cluster</th>
<th style="text-align: left;">Inferred Cell Type</th>
<th style="text-align: left;">Confidence</th>
<th style="text-align: left;">Top 5 Key Markers</th>
<th style="text-align: left;">Brief Justification</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>C-0</strong></td>
<td style="text-align: left;"><strong>Epithelial (Goblet/Secretory)</strong></td>
<td style="text-align: left;">High</td>
<td style="text-align: left;"><em>PIGR, MUC2, EPCAM, KRT8, CLDN7</em></td>
<td style="text-align: left;">Expresses broad epithelial markers (<em>EPCAM, KRT8</em>) alongside specific secretory/mucosal markers (<em>PIGR, MUC2</em>) typical of normal or well-differentiated colon epithelium.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>C-3</strong></td>
<td style="text-align: left;"><strong>Plasma Cells (IgG)</strong></td>
<td style="text-align: left;">High</td>
<td style="text-align: left;"><em>IGHG1, IGKC, MZB1, JCHAIN, XBP1</em></td>
<td style="text-align: left;">Defined by very high expression of Immunoglobulin G heavy chains (<em>IGHG1</em>) and plasma cell maturation factors (<em>MZB1, JCHAIN</em>).</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>C-4</strong></td>
<td style="text-align: left;"><strong>CAF</strong></td>
<td style="text-align: left;">High</td>
<td style="text-align: left;"><em>MMP14, MMP2, FN1, COL1A1, SPARC</em></td>
<td style="text-align: left;">Spatially coincident with tumor. The high levels of matrix metalloproteinases (MMP14, MMP2) and Fibronectin (FN1) identify these as the activated myofibroblasts driving tumor desmoplasia.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>C-5</strong></td>
<td style="text-align: left;"><strong>Unclassified</strong></td>
<td style="text-align: left;">Low</td>
<td style="text-align: left;"><em>MPV17L, ZNF91, DUX4, CXCL8, USP17L</em></td>
<td style="text-align: left;">Lacks clear lineage markers (No <em>EPCAM, CD45,</em> or <em>COL1A1</em>).</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>C-6</strong></td>
<td style="text-align: left;"><strong>Immune (Mixed)</strong></td>
<td style="text-align: left;">Moderate</td>
<td style="text-align: left;"><em>KIR3DL1, CD79A, CSF2, CD53, RAG1</em></td>
<td style="text-align: left;">Contains conflicting a mix of immune markers: NK cell (<em>KIR3DL1</em>) and B cell (<em>CD79A</em>) markers appear together. Likely represents lymphoid aggregates.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>C-7</strong></td>
<td style="text-align: left;"><strong>Endothelial Cells</strong></td>
<td style="text-align: left;">High</td>
<td style="text-align: left;"><em>PECAM1, VWF, PLVAP, EGFL7, SHANK3</em></td>
<td style="text-align: left;">Distinct expression of vascular endothelium markers, including <em>PECAM1</em> (CD31) and <em>VWF</em>. <em>PLVAP</em> suggests fenestrated capillaries common in the gut.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>C-8</strong></td>
<td style="text-align: left;"><strong>Plasma Cells (IgA)</strong></td>
<td style="text-align: left;">High</td>
<td style="text-align: left;"><em>IGHA1, JCHAIN, IGKC, GRP, MZB1</em></td>
<td style="text-align: left;">Similar to C-3 but distinguished by the specific expression of IgA (<em>IGHA1</em>), the dominant isotype in mucosal immunity of the colon.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>C-9</strong></td>
<td style="text-align: left;"><strong>Tumor Epithelial (Invasive)</strong></td>
<td style="text-align: left;">High</td>
<td style="text-align: left;"><em>CEACAM5, MMP7, LCN2, KRT8, CLDN4</em></td>
<td style="text-align: left;">Malignant epithelial profile (<em>EPCAM, KRT8</em>) with high expression of tumor markers (<em>CEACAM5/CEA</em>) and invasion/stress markers (<em>MMP7, LCN2</em>).</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>C-10</strong></td>
<td style="text-align: left;"><strong>Structural Fibroblasts </strong></td>
<td style="text-align: left;">High</td>
<td style="text-align: left;"><em>SFRP2, MGP, FBLN1, C3, CXCL12</em></td>
<td style="text-align: left;">Spatially coincident with muscle (SMC). The profile (SFRP2+, MGP+, FBLN1+) matches “Adventitial” or “Universal” fibroblasts that act as the structural scaffold for muscle layers and vessels.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>C-11</strong></td>
<td style="text-align: left;"><strong>Tumor Epithelial (High Cycling)</strong></td>
<td style="text-align: left;">High</td>
<td style="text-align: left;"><em>RPS19, CEACAM5, EPCAM, RPS18, PIGR</em></td>
<td style="text-align: left;">Epithelial tumor cells (<em>CEACAM5</em>) dominated by ribosomal proteins (<em>RPS/RPL</em>), indicating high protein synthesis and rapid cell cycling.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>C-12</strong></td>
<td style="text-align: left;"><strong>Smooth Muscle Cells</strong></td>
<td style="text-align: left;">High</td>
<td style="text-align: left;"><em>MYH11, ACTG2, DES, CNN1, MYLK</em></td>
<td style="text-align: left;">Expression of contractile proteins (<em>MYH11, ACTG2</em>) and Desmin (<em>DES</em>) identifies these as mural smooth muscle cells (muscularis mucosae/propria).</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>C-13</strong></td>
<td style="text-align: left;"><strong>Enteric Glia / Schwann Cells</strong></td>
<td style="text-align: left;">High</td>
<td style="text-align: left;"><em>CDH19, PMP22, SCN7A, LGI4, CLU</em></td>
<td style="text-align: left;">Presence of peripheral nervous system markers (<em>PMP22, CDH19, SCN7A</em>) identifies these as supportive glial cells of the Enteric Nervous System.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>C-14</strong></td>
<td style="text-align: left;"><strong>Pericytes</strong></td>
<td style="text-align: left;">High</td>
<td style="text-align: left;"><em>NOTCH3, PDGFRB, MCAM, COL4A1, RGS5</em></td>
<td style="text-align: left;">Expresses pericyte markers (<em>PDGFRB, MCAM, NOTCH3</em>) and basement membrane collagen (<em>COL4A1</em>), distinguishing them from the broader fibroblast clusters.</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p>Following a little back and forth and helping provide the LLM with additional spatial context, we have the marker gene results in <a href="#tbl-leiden-global" class="quarto-xref">Table&nbsp;<span>6.1</span></a>, we have most of the structural cell types identified from the remaining clusters. Cluster C-5 lacks a clear and dominant lineage signature. This cluster also occupies roughly the center of the UMAP where poorer quality cells sometimes occupy. For the purposes of this tutorial and the downstream questions, we’ll leave those cells unclassified and label them as “undetermined”. Similarly, cluster C-6 showed two distinct lienage markers for NK cells (<em>KIR3DL1</em>) and B cells (<em>CD79A</em>). Since these markers are unlikely to be present in the same cell it is likely that C-6 comprises a composite of NK and lymphoid cell types. Interestingly, HieraType did not classify the cells within as either NK or B cell. Since our primary interest are on tumor and CAFs, we’ll not unravel this cluster any further and instead label it as a mix of cell types.</p>
<p>From the HieraType-derived merged data, map these new cell type labels.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>remap <span class="op">=</span> {</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-0'</span>: <span class="st">'epithelial'</span>,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-3'</span>: <span class="st">'plasma_IgG'</span>,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-4'</span>: <span class="st">'CAF'</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-5'</span>: <span class="st">'undetermined'</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-6'</span>: <span class="st">'mix_NK_Lymphoid'</span>,</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-7'</span>: <span class="st">'endothelial'</span>,</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-8'</span>: <span class="st">'plasma_IgA'</span>,</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-9'</span>: <span class="st">'tumor'</span>,</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-10'</span>: <span class="st">'fibroblast'</span>,</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-11'</span>: <span class="st">'tumor_cyling'</span>,</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-12'</span>: <span class="st">'smc'</span>,</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-13'</span>: <span class="st">'glial'</span>,</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-14'</span>: <span class="st">'pericyte'</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'celltype_fine'</span>] <span class="op">=</span> adata.obs[<span class="st">'merged'</span>].astype(<span class="bu">str</span>).replace(remap)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Often we want to plot or work with broader classifications of cell types. The code below combines, for example, all the various T cell subtypes into a single group.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Python Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>broad_map <span class="op">=</span> {</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd8_tem'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_treg'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_tem'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd8_cytotoxic'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd8_naive'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd8_exhausted'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_tcm'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_th2'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_th17'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_naive'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_th1'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd8_tcm'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>    <span class="st">'bcell'</span>: <span class="st">'bcell'</span>,</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">'plasma_IgA'</span>: <span class="st">'plasma'</span>,</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">'plasma_IgG'</span>: <span class="st">'plasma'</span>,</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">'tumor'</span>: <span class="st">'tumor'</span>,</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>    <span class="st">'tumor_cyling'</span>: <span class="st">'tumor'</span>,</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>    <span class="st">'epithelial'</span>: <span class="st">'epithelial'</span>,</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>    <span class="st">'CAF'</span>: <span class="st">'CAF'</span>,</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>    <span class="st">'fibroblast'</span>: <span class="st">'fibroblast'</span>,</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>    <span class="st">'smc'</span>: <span class="st">'smc'</span>,</span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>    <span class="st">'pericyte'</span>: <span class="st">'pericyte'</span>,</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>    <span class="st">'endothelial'</span>: <span class="st">'endothelial'</span>,</span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>    <span class="st">'glial'</span>: <span class="st">'glial'</span>,</span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>    <span class="st">'macrophage'</span>: <span class="st">'macrophage'</span>,</span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a>    <span class="st">'monocyte'</span>: <span class="st">'monocyte'</span>,</span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>    <span class="st">'dendritic'</span>: <span class="st">'dendritic'</span>,</span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mast'</span>: <span class="st">'mast'</span>,</span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>    <span class="st">'neutrophil'</span>: <span class="st">'neutrophil'</span>,</span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>    <span class="st">'nk'</span>: <span class="st">'nk'</span>,</span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mix_NK_Lymphoid'</span>: <span class="st">'mix_NK_Lymphoid'</span>,</span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a>    <span class="st">'undetermined'</span>: <span class="st">'undetermined'</span></span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb9-40"><a href="#cb9-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-41"><a href="#cb9-41" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'celltype_broad'</span>] <span class="op">=</span> adata.obs[<span class="st">'celltype_fine'</span>].<span class="bu">map</span>(broad_map)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>Which provides 18 broad cell types and 32 fine cell types.</p>
<p>Save the results to disk.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-6-celltypes.h5ad"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  filename,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  compression<span class="op">=</span>hdf5plugin.FILTERS[<span class="st">"zstd"</span>],</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  compression_opts<span class="op">=</span>hdf5plugin.Zstd(clevel<span class="op">=</span><span class="dv">5</span>).filter_options</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
</section>
</section>
<section id="cell-type-visualization" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="cell-type-visualization"><span class="header-section-number">6.3</span> Cell Type Visualization</h2>
<p>It is useful at this point to visualize the distributions of cells in a variety of ways to get an understanding of their composition. We can visualize these cell types in UMAP space and XY space.</p>
<section id="broad-cell-types" class="level3" data-number="6.3.1">
<h3 data-number="6.3.1" class="anchored" data-anchor-id="broad-cell-types"><span class="header-section-number">6.3.1</span> Broad cell types</h3>
<p>Instead of the default colors in <code>plotDots</code>, let’s create manual ones.</p>
<div class="cell">
<details class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define colors semi-manually instead of using the default colors</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>column <span class="ot">&lt;-</span> <span class="st">'celltype_broad'</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>column_levels <span class="ot">&lt;-</span> <span class="fu">levels</span>(py<span class="sc">$</span>adata<span class="sc">$</span>obs[[column]])</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>n_levels <span class="ot">&lt;-</span> <span class="fu">length</span>(column_levels)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(n_levels<span class="sc">&lt;</span><span class="dv">27</span>){</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  colors_use <span class="ot">&lt;-</span> pals<span class="sc">::</span><span class="fu">alphabet</span>(n_levels)</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> <span class="cf">if</span>(n <span class="sc">&lt;</span> <span class="dv">53</span>){</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  colors_use <span class="ot">&lt;-</span> <span class="fu">c</span>(pals<span class="sc">::</span><span class="fu">alphabet</span>(<span class="dv">26</span>), pals<span class="sc">::</span><span class="fu">alphabet2</span>(n_levels<span class="dv">-26</span>))</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"consider fewer groups"</span>)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(colors_use) <span class="ot">&lt;-</span> column_levels</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>colors_use[<span class="st">'mast'</span>] <span class="ot">=</span> <span class="st">"#FFCC99"</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>colors_use[<span class="st">'tumor'</span>] <span class="ot">=</span> <span class="st">"#C20088"</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>colors_use[<span class="st">'undetermined'</span>] <span class="ot">=</span> <span class="st">"#808080"</span></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>colors_use[<span class="st">'mix_NK_Lymphoid'</span>] <span class="ot">=</span> <span class="st">"#8080FF"</span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a><span class="co"># saving colors for later</span></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'celltype_broad_names'</span>]] <span class="ot">&lt;-</span> <span class="fu">names</span>(colors_use)</span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'celltype_broad_colors'</span>]] <span class="ot">&lt;-</span> <span class="fu">as.character</span>(colors_use)</span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>Plot the cells in XY space and UMAP space and color based on <code>celltype_broad</code>.</p>
<div class="cell">
<details class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-16"><pre class="sourceCode r code-annotation-code code-with-copy code-annotated"><code class="sourceCode r"><button class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="1">1</button><span id="annotated-cell-16-1" class="code-annotation-target"><a href="#annotated-cell-16-1" aria-hidden="true" tabindex="-1"></a>group_prefix <span class="ot">&lt;-</span> <span class="st">"broad"</span></span>
<span id="annotated-cell-16-2"><a href="#annotated-cell-16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-16-3"><a href="#annotated-cell-16-3" aria-hidden="true" tabindex="-1"></a>ct_assets_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>)</span>
<span id="annotated-cell-16-4"><a href="#annotated-cell-16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-16-5"><a href="#annotated-cell-16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># XY (global only)</span></span>
<span id="annotated-cell-16-6"><a href="#annotated-cell-16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, <span class="at">color_by=</span><span class="st">'celltype_broad'</span>,</span>
<span id="annotated-cell-16-7"><a href="#annotated-cell-16-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">TRUE</span>,</span>
<span id="annotated-cell-16-8"><a href="#annotated-cell-16-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">FALSE</span>,</span>
<span id="annotated-cell-16-9"><a href="#annotated-cell-16-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-10"><a href="#annotated-cell-16-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-11"><a href="#annotated-cell-16-11" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span></span>
<span id="annotated-cell-16-12"><a href="#annotated-cell-16-12" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-16-13"><a href="#annotated-cell-16-13" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_bar_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-14"><a href="#annotated-cell-16-14" aria-hidden="true" tabindex="-1"></a>                    <span class="at">location =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">0</span>),</span>
<span id="annotated-cell-16-15"><a href="#annotated-cell-16-15" aria-hidden="true" tabindex="-1"></a>                    <span class="at">width =</span> <span class="dv">2</span>,</span>
<span id="annotated-cell-16-16"><a href="#annotated-cell-16-16" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n =</span> <span class="dv">3</span>,</span>
<span id="annotated-cell-16-17"><a href="#annotated-cell-16-17" aria-hidden="true" tabindex="-1"></a>                    <span class="at">height =</span> <span class="fl">0.1</span>,</span>
<span id="annotated-cell-16-18"><a href="#annotated-cell-16-18" aria-hidden="true" tabindex="-1"></a>                    <span class="at">scale_colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"grey30"</span>),</span>
<span id="annotated-cell-16-19"><a href="#annotated-cell-16-19" aria-hidden="true" tabindex="-1"></a>                    <span class="at">label_nudge_y =</span> <span class="sc">-</span><span class="fl">0.3</span></span>
<span id="annotated-cell-16-20"><a href="#annotated-cell-16-20" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-16-21"><a href="#annotated-cell-16-21" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="annotated-cell-16-22"><a href="#annotated-cell-16-22" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="annotated-cell-16-23"><a href="#annotated-cell-16-23" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">200</span>,</span>
<span id="annotated-cell-16-24"><a href="#annotated-cell-16-24" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">8</span>, </span>
<span id="annotated-cell-16-25"><a href="#annotated-cell-16-25" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">8</span>,</span>
<span id="annotated-cell-16-26"><a href="#annotated-cell-16-26" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="annotated-cell-16-27"><a href="#annotated-cell-16-27" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="annotated-cell-16-28"><a href="#annotated-cell-16-28" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-29"><a href="#annotated-cell-16-29" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="annotated-cell-16-30"><a href="#annotated-cell-16-30" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"X (mm)"</span>),</span>
<span id="annotated-cell-16-31"><a href="#annotated-cell-16-31" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"Y (mm)"</span>), </span>
<span id="annotated-cell-16-32"><a href="#annotated-cell-16-32" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="annotated-cell-16-33"><a href="#annotated-cell-16-33" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors_use),</span>
<span id="annotated-cell-16-34"><a href="#annotated-cell-16-34" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">0.4</span>)),</span>
<span id="annotated-cell-16-35"><a href="#annotated-cell-16-35" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="annotated-cell-16-36"><a href="#annotated-cell-16-36" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="annotated-cell-16-37"><a href="#annotated-cell-16-37" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="annotated-cell-16-38"><a href="#annotated-cell-16-38" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-16-39"><a href="#annotated-cell-16-39" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-16-40"><a href="#annotated-cell-16-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-16-41"><a href="#annotated-cell-16-41" aria-hidden="true" tabindex="-1"></a><span class="co"># XY (facets only)</span></span>
<span id="annotated-cell-16-42"><a href="#annotated-cell-16-42" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, <span class="at">color_by=</span><span class="st">'celltype_broad'</span>,</span>
<span id="annotated-cell-16-43"><a href="#annotated-cell-16-43" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">FALSE</span>,</span>
<span id="annotated-cell-16-44"><a href="#annotated-cell-16-44" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">TRUE</span>,</span>
<span id="annotated-cell-16-45"><a href="#annotated-cell-16-45" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-46"><a href="#annotated-cell-16-46" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-47"><a href="#annotated-cell-16-47" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span></span>
<span id="annotated-cell-16-48"><a href="#annotated-cell-16-48" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-16-49"><a href="#annotated-cell-16-49" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_bar_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-50"><a href="#annotated-cell-16-50" aria-hidden="true" tabindex="-1"></a>                    <span class="at">location =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">0</span>),</span>
<span id="annotated-cell-16-51"><a href="#annotated-cell-16-51" aria-hidden="true" tabindex="-1"></a>                    <span class="at">width =</span> <span class="dv">2</span>,</span>
<span id="annotated-cell-16-52"><a href="#annotated-cell-16-52" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n =</span> <span class="dv">3</span>,</span>
<span id="annotated-cell-16-53"><a href="#annotated-cell-16-53" aria-hidden="true" tabindex="-1"></a>                    <span class="at">height =</span> <span class="fl">0.1</span>,</span>
<span id="annotated-cell-16-54"><a href="#annotated-cell-16-54" aria-hidden="true" tabindex="-1"></a>                    <span class="at">scale_colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"grey30"</span>),</span>
<span id="annotated-cell-16-55"><a href="#annotated-cell-16-55" aria-hidden="true" tabindex="-1"></a>                    <span class="at">label_nudge_y =</span> <span class="sc">-</span><span class="fl">0.3</span></span>
<span id="annotated-cell-16-56"><a href="#annotated-cell-16-56" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-16-57"><a href="#annotated-cell-16-57" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="annotated-cell-16-58"><a href="#annotated-cell-16-58" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="annotated-cell-16-59"><a href="#annotated-cell-16-59" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">100</span>,</span>
<span id="annotated-cell-16-60"><a href="#annotated-cell-16-60" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">5</span>,</span>
<span id="annotated-cell-16-61"><a href="#annotated-cell-16-61" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">5</span>,</span>
<span id="annotated-cell-16-62"><a href="#annotated-cell-16-62" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="annotated-cell-16-63"><a href="#annotated-cell-16-63" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="annotated-cell-16-64"><a href="#annotated-cell-16-64" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-65"><a href="#annotated-cell-16-65" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="annotated-cell-16-66"><a href="#annotated-cell-16-66" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"X (mm)"</span>),</span>
<span id="annotated-cell-16-67"><a href="#annotated-cell-16-67" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"Y (mm)"</span>), </span>
<span id="annotated-cell-16-68"><a href="#annotated-cell-16-68" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="annotated-cell-16-69"><a href="#annotated-cell-16-69" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors_use),</span>
<span id="annotated-cell-16-70"><a href="#annotated-cell-16-70" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">0.4</span>)),</span>
<span id="annotated-cell-16-71"><a href="#annotated-cell-16-71" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="annotated-cell-16-72"><a href="#annotated-cell-16-72" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="annotated-cell-16-73"><a href="#annotated-cell-16-73" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="annotated-cell-16-74"><a href="#annotated-cell-16-74" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-16-75"><a href="#annotated-cell-16-75" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-16-76"><a href="#annotated-cell-16-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-16-77"><a href="#annotated-cell-16-77" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP (global only)</span></span>
<span id="annotated-cell-16-78"><a href="#annotated-cell-16-78" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, </span>
<span id="annotated-cell-16-79"><a href="#annotated-cell-16-79" aria-hidden="true" tabindex="-1"></a>              <span class="at">obsm_key =</span> <span class="st">"umap_pr"</span>,</span>
<span id="annotated-cell-16-80"><a href="#annotated-cell-16-80" aria-hidden="true" tabindex="-1"></a>              <span class="at">color_by=</span><span class="st">'celltype_broad'</span>,</span>
<span id="annotated-cell-16-81"><a href="#annotated-cell-16-81" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">TRUE</span>,</span>
<span id="annotated-cell-16-82"><a href="#annotated-cell-16-82" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">FALSE</span>,</span>
<span id="annotated-cell-16-83"><a href="#annotated-cell-16-83" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-84"><a href="#annotated-cell-16-84" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-85"><a href="#annotated-cell-16-85" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span>, <span class="at">alpha=</span><span class="fl">0.1</span></span>
<span id="annotated-cell-16-86"><a href="#annotated-cell-16-86" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-16-87"><a href="#annotated-cell-16-87" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_label_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-88"><a href="#annotated-cell-16-88" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size =</span> <span class="dv">2</span></span>
<span id="annotated-cell-16-89"><a href="#annotated-cell-16-89" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-16-90"><a href="#annotated-cell-16-90" aria-hidden="true" tabindex="-1"></a>                  <span class="at">labels_on_plot =</span> <span class="fu">data.frame</span>(),</span>
<span id="annotated-cell-16-91"><a href="#annotated-cell-16-91" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="annotated-cell-16-92"><a href="#annotated-cell-16-92" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="annotated-cell-16-93"><a href="#annotated-cell-16-93" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">200</span>,</span>
<span id="annotated-cell-16-94"><a href="#annotated-cell-16-94" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">8</span>,</span>
<span id="annotated-cell-16-95"><a href="#annotated-cell-16-95" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">8</span>,</span>
<span id="annotated-cell-16-96"><a href="#annotated-cell-16-96" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="annotated-cell-16-97"><a href="#annotated-cell-16-97" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="annotated-cell-16-98"><a href="#annotated-cell-16-98" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-99"><a href="#annotated-cell-16-99" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="annotated-cell-16-100"><a href="#annotated-cell-16-100" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"UMAP 1"</span>),</span>
<span id="annotated-cell-16-101"><a href="#annotated-cell-16-101" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"UMAP 2"</span>), </span>
<span id="annotated-cell-16-102"><a href="#annotated-cell-16-102" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="annotated-cell-16-103"><a href="#annotated-cell-16-103" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors_use),</span>
<span id="annotated-cell-16-104"><a href="#annotated-cell-16-104" aria-hidden="true" tabindex="-1"></a>                <span class="co"># guides(color = guide_legend(</span></span>
<span id="annotated-cell-16-105"><a href="#annotated-cell-16-105" aria-hidden="true" tabindex="-1"></a>                <span class="co">#   title="Cell Type",</span></span>
<span id="annotated-cell-16-106"><a href="#annotated-cell-16-106" aria-hidden="true" tabindex="-1"></a>                <span class="co">#   override.aes = list(size = 3) ) ), </span></span>
<span id="annotated-cell-16-107"><a href="#annotated-cell-16-107" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="annotated-cell-16-108"><a href="#annotated-cell-16-108" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-16-109"><a href="#annotated-cell-16-109" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="annotated-cell-16-110"><a href="#annotated-cell-16-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-16-111"><a href="#annotated-cell-16-111" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP (facets only)</span></span>
<span id="annotated-cell-16-112"><a href="#annotated-cell-16-112" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, </span>
<span id="annotated-cell-16-113"><a href="#annotated-cell-16-113" aria-hidden="true" tabindex="-1"></a>              <span class="at">obsm_key =</span> <span class="st">"umap_pr"</span>,</span>
<span id="annotated-cell-16-114"><a href="#annotated-cell-16-114" aria-hidden="true" tabindex="-1"></a>              <span class="at">color_by=</span><span class="st">'celltype_broad'</span>,</span>
<span id="annotated-cell-16-115"><a href="#annotated-cell-16-115" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">FALSE</span>,</span>
<span id="annotated-cell-16-116"><a href="#annotated-cell-16-116" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">TRUE</span>,</span>
<span id="annotated-cell-16-117"><a href="#annotated-cell-16-117" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-118"><a href="#annotated-cell-16-118" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-119"><a href="#annotated-cell-16-119" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span>, <span class="at">alpha=</span><span class="fl">0.1</span></span>
<span id="annotated-cell-16-120"><a href="#annotated-cell-16-120" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-16-121"><a href="#annotated-cell-16-121" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_label_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-122"><a href="#annotated-cell-16-122" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size =</span> <span class="dv">2</span></span>
<span id="annotated-cell-16-123"><a href="#annotated-cell-16-123" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-16-124"><a href="#annotated-cell-16-124" aria-hidden="true" tabindex="-1"></a>                  <span class="at">labels_on_plot =</span> <span class="fu">data.frame</span>(),</span>
<span id="annotated-cell-16-125"><a href="#annotated-cell-16-125" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="annotated-cell-16-126"><a href="#annotated-cell-16-126" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="annotated-cell-16-127"><a href="#annotated-cell-16-127" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">100</span>,</span>
<span id="annotated-cell-16-128"><a href="#annotated-cell-16-128" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">5</span>,</span>
<span id="annotated-cell-16-129"><a href="#annotated-cell-16-129" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">5</span>,</span>
<span id="annotated-cell-16-130"><a href="#annotated-cell-16-130" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="annotated-cell-16-131"><a href="#annotated-cell-16-131" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="annotated-cell-16-132"><a href="#annotated-cell-16-132" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-16-133"><a href="#annotated-cell-16-133" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="annotated-cell-16-134"><a href="#annotated-cell-16-134" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"UMAP 1"</span>),</span>
<span id="annotated-cell-16-135"><a href="#annotated-cell-16-135" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"UMAP 2"</span>), </span>
<span id="annotated-cell-16-136"><a href="#annotated-cell-16-136" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="annotated-cell-16-137"><a href="#annotated-cell-16-137" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors_use),</span>
<span id="annotated-cell-16-138"><a href="#annotated-cell-16-138" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="annotated-cell-16-139"><a href="#annotated-cell-16-139" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="annotated-cell-16-140"><a href="#annotated-cell-16-140" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="annotated-cell-16-141"><a href="#annotated-cell-16-141" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-16-142"><a href="#annotated-cell-16-142" aria-hidden="true" tabindex="-1"></a>            )</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-16" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="1" data-code-annotation="1">in the plotting below, we will use this name to parse out files.</span>
</dd>
</dl>
</div>
</div>
<p>Here are the overall (global) plots.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" href="">Cell Type Broad - UMAP</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false" href="">Cell Type Broad - XY</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell">
<div class="cell-output-display">
<div id="fig-umap-celltype-broad" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-umap-celltype-broad-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/ct/broad__umap_pr__S0__plot.png" class="img-fluid figure-img" width="800">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-umap-celltype-broad-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.3: UMAP with broad cell type in the full dataset.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="cell">
<div class="cell-output-display">
<div id="fig-xy-celltype-broad" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-xy-celltype-broad-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/ct/broad__spatial__S0__plot.png" class="img-fluid figure-img" width="800">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-xy-celltype-broad-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.4: XY with broad cell type in the full dataset.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>And here are the facetted plots.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true" href="">CAF</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false" href="">bcell</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-3" role="tab" aria-controls="tabset-2-3" aria-selected="false" href="">dendritic</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-4" role="tab" aria-controls="tabset-2-4" aria-selected="false" href="">endothelial</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-5-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-5" role="tab" aria-controls="tabset-2-5" aria-selected="false" href="">epithelial</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-6-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-6" role="tab" aria-controls="tabset-2-6" aria-selected="false" href="">fibroblast</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-7-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-7" role="tab" aria-controls="tabset-2-7" aria-selected="false" href="">glial</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-8-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-8" role="tab" aria-controls="tabset-2-8" aria-selected="false" href="">macrophage</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-9-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-9" role="tab" aria-controls="tabset-2-9" aria-selected="false" href="">mast</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-10-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-10" role="tab" aria-controls="tabset-2-10" aria-selected="false" href="">mix_NK_Lymphoid</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-11-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-11" role="tab" aria-controls="tabset-2-11" aria-selected="false" href="">monocyte</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-12-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-12" role="tab" aria-controls="tabset-2-12" aria-selected="false" href="">neutrophil</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-13-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-13" role="tab" aria-controls="tabset-2-13" aria-selected="false" href="">nk</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-14-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-14" role="tab" aria-controls="tabset-2-14" aria-selected="false" href="">pericyte</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-15-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-15" role="tab" aria-controls="tabset-2-15" aria-selected="false" href="">plasma</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-16-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-16" role="tab" aria-controls="tabset-2-16" aria-selected="false" href="">smc</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-17-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-17" role="tab" aria-controls="tabset-2-17" aria-selected="false" href="">tcell</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-18-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-18" role="tab" aria-controls="tabset-2-18" aria-selected="false" href="">tumor</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-19-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-19" role="tab" aria-controls="tabset-2-19" aria-selected="false" href="">undetermined</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_CAF.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_CAF.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_bcell.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_bcell.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-3-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_dendritic.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_dendritic.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-4-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_endothelial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_endothelial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-5" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-5-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_epithelial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_epithelial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-6" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-6-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_fibroblast.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_fibroblast.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-7" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-7-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_glial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_glial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-8" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-8-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_macrophage.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_macrophage.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-9" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-9-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_mast.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_mast.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-10" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-10-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_mix_NK_Lymphoid.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_mix_NK_Lymphoid.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-11" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-11-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_monocyte.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_monocyte.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-12" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-12-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_neutrophil.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_neutrophil.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-13" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-13-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_nk.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_nk.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-14" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-14-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_pericyte.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_pericyte.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-15" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-15-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_plasma.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_plasma.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-16" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-16-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_smc.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_smc.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-17" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-17-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_tcell.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_tcell.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-18" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-18-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_tumor.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_tumor.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-19" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-19-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__spatial__S0__facet_undetermined.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/broad__umap_pr__S0__facet_undetermined.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="fine-cell-types" class="level3" data-number="6.3.2">
<h3 data-number="6.3.2" class="anchored" data-anchor-id="fine-cell-types"><span class="header-section-number">6.3.2</span> Fine cell types</h3>
<p>This section performs a similar routine as above but with the fine cell type classifications.</p>
<p>For the fine cell types, we’ll create colors that are various shades of the parent type.</p>
<div class="cell">
<details class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>broad_map <span class="ot">&lt;-</span> <span class="fu">unlist</span>(py<span class="sc">$</span>broad_map)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>generate_fine_colors <span class="ot">&lt;-</span> <span class="cf">function</span>(mapping, broad_colors, <span class="at">mode =</span> <span class="st">"shades"</span>) {</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  groups <span class="ot">&lt;-</span> <span class="fu">split</span>(<span class="fu">names</span>(mapping), mapping)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  fine_colors <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (broad_name <span class="cf">in</span> <span class="fu">names</span>(groups)) {</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    subtypes <span class="ot">&lt;-</span> groups[[broad_name]]</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>    n <span class="ot">&lt;-</span> <span class="fu">length</span>(subtypes)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>    base_col <span class="ot">&lt;-</span> broad_colors[broad_name]</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (n <span class="sc">==</span> <span class="dv">1</span>) {</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>      cols <span class="ot">&lt;-</span> base_col</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>      <span class="fu">names</span>(cols) <span class="ot">&lt;-</span> subtypes</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (mode <span class="sc">==</span> <span class="st">"shades"</span>) {</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>        ramp_func <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>          <span class="fu">adjustcolor</span>(base_col, <span class="at">transform=</span><span class="fu">diag</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))<span class="sc">*</span><span class="fl">1.4</span>), <span class="co"># Lighten</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>          base_col,</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>          <span class="fu">adjustcolor</span>(base_col, <span class="at">transform=</span><span class="fu">diag</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))<span class="sc">*</span><span class="fl">0.6</span>)  <span class="co"># Darken</span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>        ))</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>        cols <span class="ot">&lt;-</span> <span class="fu">ramp_func</span>(n)</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>        ramp_func <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"#FFFFFF"</span>, base_col, <span class="st">"#000000"</span>))</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>        cols <span class="ot">&lt;-</span> <span class="fu">ramp_func</span>(n <span class="sc">+</span> <span class="dv">2</span>)[<span class="dv">2</span><span class="sc">:</span>(n <span class="sc">+</span> <span class="dv">1</span>)]</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">names</span>(cols) <span class="ot">&lt;-</span> subtypes</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>    fine_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(fine_colors, cols)</span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(fine_colors)</span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>fine_colors <span class="ot">&lt;-</span> <span class="fu">generate_fine_colors</span>(broad_map, colors_use)</span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'celltype_fine_names'</span>]] <span class="ot">&lt;-</span> <span class="fu">names</span>(fine_colors)</span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'celltype_fine_colors'</span>]] <span class="ot">&lt;-</span> <span class="fu">as.character</span>(fine_colors)</span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<div class="cell">
<details class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-18"><pre class="sourceCode r code-annotation-code code-with-copy"><code class="sourceCode r"><span id="annotated-cell-18-1"><a href="#annotated-cell-18-1" aria-hidden="true" tabindex="-1"></a>group_prefix <span class="ot">&lt;-</span> <span class="st">"fine"</span></span>
<span id="annotated-cell-18-2"><a href="#annotated-cell-18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-18-3"><a href="#annotated-cell-18-3" aria-hidden="true" tabindex="-1"></a>ct_assets_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>)</span>
<span id="annotated-cell-18-4"><a href="#annotated-cell-18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-18-5"><a href="#annotated-cell-18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># XY (global only)</span></span>
<span id="annotated-cell-18-6"><a href="#annotated-cell-18-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, <span class="at">color_by=</span><span class="st">'celltype_fine'</span>,</span>
<span id="annotated-cell-18-7"><a href="#annotated-cell-18-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">TRUE</span>,</span>
<span id="annotated-cell-18-8"><a href="#annotated-cell-18-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">FALSE</span>,</span>
<span id="annotated-cell-18-9"><a href="#annotated-cell-18-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-10"><a href="#annotated-cell-18-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-11"><a href="#annotated-cell-18-11" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span></span>
<span id="annotated-cell-18-12"><a href="#annotated-cell-18-12" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-18-13"><a href="#annotated-cell-18-13" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_bar_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-14"><a href="#annotated-cell-18-14" aria-hidden="true" tabindex="-1"></a>                    <span class="at">location =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">0</span>),</span>
<span id="annotated-cell-18-15"><a href="#annotated-cell-18-15" aria-hidden="true" tabindex="-1"></a>                    <span class="at">width =</span> <span class="dv">2</span>,</span>
<span id="annotated-cell-18-16"><a href="#annotated-cell-18-16" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n =</span> <span class="dv">3</span>,</span>
<span id="annotated-cell-18-17"><a href="#annotated-cell-18-17" aria-hidden="true" tabindex="-1"></a>                    <span class="at">height =</span> <span class="fl">0.1</span>,</span>
<span id="annotated-cell-18-18"><a href="#annotated-cell-18-18" aria-hidden="true" tabindex="-1"></a>                    <span class="at">scale_colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"grey30"</span>),</span>
<span id="annotated-cell-18-19"><a href="#annotated-cell-18-19" aria-hidden="true" tabindex="-1"></a>                    <span class="at">label_nudge_y =</span> <span class="sc">-</span><span class="fl">0.3</span></span>
<span id="annotated-cell-18-20"><a href="#annotated-cell-18-20" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-18-21"><a href="#annotated-cell-18-21" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="annotated-cell-18-22"><a href="#annotated-cell-18-22" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="annotated-cell-18-23"><a href="#annotated-cell-18-23" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">200</span>,</span>
<span id="annotated-cell-18-24"><a href="#annotated-cell-18-24" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">8</span>, </span>
<span id="annotated-cell-18-25"><a href="#annotated-cell-18-25" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">8</span>,</span>
<span id="annotated-cell-18-26"><a href="#annotated-cell-18-26" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="annotated-cell-18-27"><a href="#annotated-cell-18-27" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="annotated-cell-18-28"><a href="#annotated-cell-18-28" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-29"><a href="#annotated-cell-18-29" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="annotated-cell-18-30"><a href="#annotated-cell-18-30" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"X (mm)"</span>),</span>
<span id="annotated-cell-18-31"><a href="#annotated-cell-18-31" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"Y (mm)"</span>), </span>
<span id="annotated-cell-18-32"><a href="#annotated-cell-18-32" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="annotated-cell-18-33"><a href="#annotated-cell-18-33" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> fine_colors),</span>
<span id="annotated-cell-18-34"><a href="#annotated-cell-18-34" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">0.4</span>)),</span>
<span id="annotated-cell-18-35"><a href="#annotated-cell-18-35" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="annotated-cell-18-36"><a href="#annotated-cell-18-36" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="annotated-cell-18-37"><a href="#annotated-cell-18-37" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="annotated-cell-18-38"><a href="#annotated-cell-18-38" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-18-39"><a href="#annotated-cell-18-39" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-18-40"><a href="#annotated-cell-18-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-18-41"><a href="#annotated-cell-18-41" aria-hidden="true" tabindex="-1"></a><span class="co"># XY (facets only)</span></span>
<span id="annotated-cell-18-42"><a href="#annotated-cell-18-42" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, <span class="at">color_by=</span><span class="st">'celltype_fine'</span>,</span>
<span id="annotated-cell-18-43"><a href="#annotated-cell-18-43" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">FALSE</span>,</span>
<span id="annotated-cell-18-44"><a href="#annotated-cell-18-44" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">TRUE</span>,</span>
<span id="annotated-cell-18-45"><a href="#annotated-cell-18-45" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-46"><a href="#annotated-cell-18-46" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-47"><a href="#annotated-cell-18-47" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span></span>
<span id="annotated-cell-18-48"><a href="#annotated-cell-18-48" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-18-49"><a href="#annotated-cell-18-49" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_bar_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-50"><a href="#annotated-cell-18-50" aria-hidden="true" tabindex="-1"></a>                    <span class="at">location =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">0</span>),</span>
<span id="annotated-cell-18-51"><a href="#annotated-cell-18-51" aria-hidden="true" tabindex="-1"></a>                    <span class="at">width =</span> <span class="dv">2</span>,</span>
<span id="annotated-cell-18-52"><a href="#annotated-cell-18-52" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n =</span> <span class="dv">3</span>,</span>
<span id="annotated-cell-18-53"><a href="#annotated-cell-18-53" aria-hidden="true" tabindex="-1"></a>                    <span class="at">height =</span> <span class="fl">0.1</span>,</span>
<span id="annotated-cell-18-54"><a href="#annotated-cell-18-54" aria-hidden="true" tabindex="-1"></a>                    <span class="at">scale_colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"grey30"</span>),</span>
<span id="annotated-cell-18-55"><a href="#annotated-cell-18-55" aria-hidden="true" tabindex="-1"></a>                    <span class="at">label_nudge_y =</span> <span class="sc">-</span><span class="fl">0.3</span></span>
<span id="annotated-cell-18-56"><a href="#annotated-cell-18-56" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-18-57"><a href="#annotated-cell-18-57" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="annotated-cell-18-58"><a href="#annotated-cell-18-58" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="annotated-cell-18-59"><a href="#annotated-cell-18-59" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">100</span>,</span>
<span id="annotated-cell-18-60"><a href="#annotated-cell-18-60" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">5</span>,</span>
<span id="annotated-cell-18-61"><a href="#annotated-cell-18-61" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">5</span>,</span>
<span id="annotated-cell-18-62"><a href="#annotated-cell-18-62" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="annotated-cell-18-63"><a href="#annotated-cell-18-63" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="annotated-cell-18-64"><a href="#annotated-cell-18-64" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-65"><a href="#annotated-cell-18-65" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="annotated-cell-18-66"><a href="#annotated-cell-18-66" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"X (mm)"</span>),</span>
<span id="annotated-cell-18-67"><a href="#annotated-cell-18-67" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"Y (mm)"</span>), </span>
<span id="annotated-cell-18-68"><a href="#annotated-cell-18-68" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="annotated-cell-18-69"><a href="#annotated-cell-18-69" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> fine_colors),</span>
<span id="annotated-cell-18-70"><a href="#annotated-cell-18-70" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">0.4</span>)),</span>
<span id="annotated-cell-18-71"><a href="#annotated-cell-18-71" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="annotated-cell-18-72"><a href="#annotated-cell-18-72" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="annotated-cell-18-73"><a href="#annotated-cell-18-73" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="annotated-cell-18-74"><a href="#annotated-cell-18-74" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-18-75"><a href="#annotated-cell-18-75" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-18-76"><a href="#annotated-cell-18-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-18-77"><a href="#annotated-cell-18-77" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP (global only)</span></span>
<span id="annotated-cell-18-78"><a href="#annotated-cell-18-78" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, </span>
<span id="annotated-cell-18-79"><a href="#annotated-cell-18-79" aria-hidden="true" tabindex="-1"></a>              <span class="at">obsm_key =</span> <span class="st">"umap_pr"</span>,</span>
<span id="annotated-cell-18-80"><a href="#annotated-cell-18-80" aria-hidden="true" tabindex="-1"></a>              <span class="at">color_by=</span><span class="st">'celltype_fine'</span>,</span>
<span id="annotated-cell-18-81"><a href="#annotated-cell-18-81" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">TRUE</span>,</span>
<span id="annotated-cell-18-82"><a href="#annotated-cell-18-82" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">FALSE</span>,</span>
<span id="annotated-cell-18-83"><a href="#annotated-cell-18-83" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-84"><a href="#annotated-cell-18-84" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-85"><a href="#annotated-cell-18-85" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span>, <span class="at">alpha=</span><span class="fl">0.1</span></span>
<span id="annotated-cell-18-86"><a href="#annotated-cell-18-86" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-18-87"><a href="#annotated-cell-18-87" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_label_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-88"><a href="#annotated-cell-18-88" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size =</span> <span class="dv">2</span></span>
<span id="annotated-cell-18-89"><a href="#annotated-cell-18-89" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-18-90"><a href="#annotated-cell-18-90" aria-hidden="true" tabindex="-1"></a>                  <span class="at">labels_on_plot =</span> <span class="fu">data.frame</span>(),</span>
<span id="annotated-cell-18-91"><a href="#annotated-cell-18-91" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="annotated-cell-18-92"><a href="#annotated-cell-18-92" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="annotated-cell-18-93"><a href="#annotated-cell-18-93" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">200</span>,</span>
<span id="annotated-cell-18-94"><a href="#annotated-cell-18-94" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">8</span>,</span>
<span id="annotated-cell-18-95"><a href="#annotated-cell-18-95" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">8</span>,</span>
<span id="annotated-cell-18-96"><a href="#annotated-cell-18-96" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="annotated-cell-18-97"><a href="#annotated-cell-18-97" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="annotated-cell-18-98"><a href="#annotated-cell-18-98" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-99"><a href="#annotated-cell-18-99" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="annotated-cell-18-100"><a href="#annotated-cell-18-100" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"UMAP 1"</span>),</span>
<span id="annotated-cell-18-101"><a href="#annotated-cell-18-101" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"UMAP 2"</span>), </span>
<span id="annotated-cell-18-102"><a href="#annotated-cell-18-102" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="annotated-cell-18-103"><a href="#annotated-cell-18-103" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> fine_colors),</span>
<span id="annotated-cell-18-104"><a href="#annotated-cell-18-104" aria-hidden="true" tabindex="-1"></a>                <span class="co"># guides(color = guide_legend(</span></span>
<span id="annotated-cell-18-105"><a href="#annotated-cell-18-105" aria-hidden="true" tabindex="-1"></a>                <span class="co">#   title="Cell Type",</span></span>
<span id="annotated-cell-18-106"><a href="#annotated-cell-18-106" aria-hidden="true" tabindex="-1"></a>                <span class="co">#   override.aes = list(size = 3) ) ), </span></span>
<span id="annotated-cell-18-107"><a href="#annotated-cell-18-107" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="annotated-cell-18-108"><a href="#annotated-cell-18-108" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-18-109"><a href="#annotated-cell-18-109" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="annotated-cell-18-110"><a href="#annotated-cell-18-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-18-111"><a href="#annotated-cell-18-111" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP (facets only)</span></span>
<span id="annotated-cell-18-112"><a href="#annotated-cell-18-112" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, </span>
<span id="annotated-cell-18-113"><a href="#annotated-cell-18-113" aria-hidden="true" tabindex="-1"></a>              <span class="at">obsm_key =</span> <span class="st">"umap_pr"</span>,</span>
<span id="annotated-cell-18-114"><a href="#annotated-cell-18-114" aria-hidden="true" tabindex="-1"></a>              <span class="at">color_by=</span><span class="st">'celltype_fine'</span>,</span>
<span id="annotated-cell-18-115"><a href="#annotated-cell-18-115" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">FALSE</span>,</span>
<span id="annotated-cell-18-116"><a href="#annotated-cell-18-116" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">TRUE</span>,</span>
<span id="annotated-cell-18-117"><a href="#annotated-cell-18-117" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-118"><a href="#annotated-cell-18-118" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-119"><a href="#annotated-cell-18-119" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span>, <span class="at">alpha=</span><span class="fl">0.1</span></span>
<span id="annotated-cell-18-120"><a href="#annotated-cell-18-120" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-18-121"><a href="#annotated-cell-18-121" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_label_params =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-122"><a href="#annotated-cell-18-122" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size =</span> <span class="dv">2</span></span>
<span id="annotated-cell-18-123"><a href="#annotated-cell-18-123" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="annotated-cell-18-124"><a href="#annotated-cell-18-124" aria-hidden="true" tabindex="-1"></a>                  <span class="at">labels_on_plot =</span> <span class="fu">data.frame</span>(),</span>
<span id="annotated-cell-18-125"><a href="#annotated-cell-18-125" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="annotated-cell-18-126"><a href="#annotated-cell-18-126" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="annotated-cell-18-127"><a href="#annotated-cell-18-127" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">100</span>,</span>
<span id="annotated-cell-18-128"><a href="#annotated-cell-18-128" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">5</span>,</span>
<span id="annotated-cell-18-129"><a href="#annotated-cell-18-129" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">5</span>,</span>
<span id="annotated-cell-18-130"><a href="#annotated-cell-18-130" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="annotated-cell-18-131"><a href="#annotated-cell-18-131" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="annotated-cell-18-132"><a href="#annotated-cell-18-132" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-18-133"><a href="#annotated-cell-18-133" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="annotated-cell-18-134"><a href="#annotated-cell-18-134" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"UMAP 1"</span>),</span>
<span id="annotated-cell-18-135"><a href="#annotated-cell-18-135" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"UMAP 2"</span>), </span>
<span id="annotated-cell-18-136"><a href="#annotated-cell-18-136" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="annotated-cell-18-137"><a href="#annotated-cell-18-137" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> fine_colors),</span>
<span id="annotated-cell-18-138"><a href="#annotated-cell-18-138" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="annotated-cell-18-139"><a href="#annotated-cell-18-139" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="annotated-cell-18-140"><a href="#annotated-cell-18-140" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="annotated-cell-18-141"><a href="#annotated-cell-18-141" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="annotated-cell-18-142"><a href="#annotated-cell-18-142" aria-hidden="true" tabindex="-1"></a>            )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-3-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-1" role="tab" aria-controls="tabset-3-1" aria-selected="true" href="">Cell Type Fine - UMAP</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-2" role="tab" aria-controls="tabset-3-2" aria-selected="false" href="">Cell Type Fine - XY</a></li></ul>
<div class="tab-content">
<div id="tabset-3-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-3-1-tab">
<div class="cell">
<div class="cell-output-display">
<div id="fig-umap-celltype-fine" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-umap-celltype-fine-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/ct/fine__umap_pr__S0__plot.png" class="img-fluid figure-img" width="800">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-umap-celltype-fine-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.5: UMAP with fine cell type labels.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-3-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-2-tab">
<div class="cell">
<div class="cell-output-display">
<div id="fig-xy-celltype-fine" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-xy-celltype-fine-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/ct/fine__spatial__S0__plot.png" class="img-fluid figure-img" width="800">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-xy-celltype-fine-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.6: XY with fine cell type labels.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>And here are the facetted plots.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-4-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-1" role="tab" aria-controls="tabset-4-1" aria-selected="true" href="">CAF</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-2" role="tab" aria-controls="tabset-4-2" aria-selected="false" href="">bcell</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-3" role="tab" aria-controls="tabset-4-3" aria-selected="false" href="">cd4_naive</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-4" role="tab" aria-controls="tabset-4-4" aria-selected="false" href="">cd4_tcm</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-5-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-5" role="tab" aria-controls="tabset-4-5" aria-selected="false" href="">cd4_tem</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-6-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-6" role="tab" aria-controls="tabset-4-6" aria-selected="false" href="">cd4_th1</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-7-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-7" role="tab" aria-controls="tabset-4-7" aria-selected="false" href="">cd4_th17</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-8-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-8" role="tab" aria-controls="tabset-4-8" aria-selected="false" href="">cd4_th2</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-9-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-9" role="tab" aria-controls="tabset-4-9" aria-selected="false" href="">cd4_treg</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-10-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-10" role="tab" aria-controls="tabset-4-10" aria-selected="false" href="">cd8_cytotoxic</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-11-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-11" role="tab" aria-controls="tabset-4-11" aria-selected="false" href="">cd8_exhausted</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-12-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-12" role="tab" aria-controls="tabset-4-12" aria-selected="false" href="">cd8_naive</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-13-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-13" role="tab" aria-controls="tabset-4-13" aria-selected="false" href="">cd8_tcm</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-14-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-14" role="tab" aria-controls="tabset-4-14" aria-selected="false" href="">cd8_tem</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-15-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-15" role="tab" aria-controls="tabset-4-15" aria-selected="false" href="">dendritic</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-16-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-16" role="tab" aria-controls="tabset-4-16" aria-selected="false" href="">endothelial</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-17-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-17" role="tab" aria-controls="tabset-4-17" aria-selected="false" href="">epithelial</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-18-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-18" role="tab" aria-controls="tabset-4-18" aria-selected="false" href="">fibroblast</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-19-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-19" role="tab" aria-controls="tabset-4-19" aria-selected="false" href="">glial</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-20-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-20" role="tab" aria-controls="tabset-4-20" aria-selected="false" href="">macrophage</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-21-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-21" role="tab" aria-controls="tabset-4-21" aria-selected="false" href="">mast</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-22-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-22" role="tab" aria-controls="tabset-4-22" aria-selected="false" href="">mix_NK_Lymphoid</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-23-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-23" role="tab" aria-controls="tabset-4-23" aria-selected="false" href="">monocyte</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-24-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-24" role="tab" aria-controls="tabset-4-24" aria-selected="false" href="">neutrophil</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-25-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-25" role="tab" aria-controls="tabset-4-25" aria-selected="false" href="">nk</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-26-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-26" role="tab" aria-controls="tabset-4-26" aria-selected="false" href="">pericyte</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-27-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-27" role="tab" aria-controls="tabset-4-27" aria-selected="false" href="">plasma_IgA</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-28-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-28" role="tab" aria-controls="tabset-4-28" aria-selected="false" href="">plasma_IgG</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-29-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-29" role="tab" aria-controls="tabset-4-29" aria-selected="false" href="">smc</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-30-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-30" role="tab" aria-controls="tabset-4-30" aria-selected="false" href="">tumor</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-31-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-31" role="tab" aria-controls="tabset-4-31" aria-selected="false" href="">tumor_cyling</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-32-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-32" role="tab" aria-controls="tabset-4-32" aria-selected="false" href="">undetermined</a></li></ul>
<div class="tab-content">
<div id="tabset-4-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-4-1-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_CAF.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_CAF.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-2-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_bcell.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_bcell.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-3-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_naive.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_naive.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-4-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_tcm.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_tcm.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-5" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-5-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_tem.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_tem.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-6" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-6-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_th1.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_th1.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-7" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-7-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_th17.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_th17.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-8" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-8-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_th2.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_th2.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-9" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-9-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_treg.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_treg.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-10" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-10-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd8_cytotoxic.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd8_cytotoxic.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-11" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-11-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd8_exhausted.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd8_exhausted.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-12" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-12-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd8_naive.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd8_naive.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-13" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-13-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd8_tcm.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd8_tcm.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-14" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-14-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_cd8_tem.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd8_tem.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-15" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-15-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_dendritic.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_dendritic.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-16" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-16-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_endothelial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_endothelial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-17" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-17-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_epithelial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_epithelial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-18" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-18-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_fibroblast.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_fibroblast.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-19" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-19-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_glial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_glial.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-20" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-20-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_macrophage.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_macrophage.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-21" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-21-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_mast.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_mast.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-22" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-22-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_mix_NK_Lymphoid.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_mix_NK_Lymphoid.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-23" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-23-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_monocyte.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_monocyte.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-24" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-24-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_neutrophil.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_neutrophil.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-25" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-25-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_nk.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_nk.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-26" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-26-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_pericyte.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_pericyte.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-27" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-27-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_plasma_IgA.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_plasma_IgA.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-28" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-28-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_plasma_IgG.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_plasma_IgG.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-29" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-29-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_smc.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_smc.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-30" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-30-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_tumor.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_tumor.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-31" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-31-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_tumor_cyling.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_tumor_cyling.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-32" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-32-tab">
<div class="{columns}">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__spatial__S0__facet_undetermined.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="column" style="width:10%;">

</div>
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./assets/analysis_results/ct/fine__umap_pr__S0__facet_undetermined.png" class="img-fluid figure-img" width="250"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Note that within the SMC XY plots we can see an “outline” that appears to be the muscularis mucosa.</p>
</section>
</section>
<section id="sec-integration" class="level2" data-number="6.4">
<h2 data-number="6.4" class="anchored" data-anchor-id="sec-integration"><span class="header-section-number">6.4</span> Integration with Spatial Domains</h2>
<p>Recall in <a href="spatial-domains.html" class="quarto-xref"><span>Chapter 5</span></a> that we used <code>novae</code> to define spatial domains. Now that we have cell type information, let’s look at the composition of cell types within each domain and assign a biological function to them.</p>
<p>Since we only need the metadata associated with the spatial domain results, we’ll only load the obs.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Python Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>focal_domain_resolution <span class="op">=</span> <span class="st">'novae_domains_6'</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'adata_domains'</span> <span class="kw">not</span> <span class="kw">in</span> <span class="bu">dir</span>():</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-3-novae.h5ad"</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  adata_domains <span class="op">=</span> ad.read_h5ad(filename, backed<span class="op">=</span><span class="st">'r'</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>meta_domains <span class="op">=</span> adata_domains.obs.copy()</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>columns_to_copy <span class="op">=</span> [focal_domain_resolution]</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>adata.obs[columns_to_copy] <span class="op">=</span> meta_domains[columns_to_copy]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>Create a contingency table for the domain and cell type assignments.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>focal_domain_resolution <span class="ot">&lt;-</span> py<span class="sc">$</span>focal_domain_resolution</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>focal_celltype_resolution <span class="ot">&lt;-</span> <span class="st">"celltype_fine"</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs <span class="sc">%&gt;%</span> <span class="fu">select</span>(</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">all_of</span>(focal_domain_resolution), </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">all_of</span>(focal_celltype_resolution))</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> </span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(.data[[focal_domain_resolution]], </span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>           .data[[focal_celltype_resolution]]) <span class="sc">%&gt;%</span> </span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">count =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span> </span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">proportion =</span> count <span class="sc">/</span> <span class="fu">sum</span>(count))</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the geom_tile plot</span></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(plot_data, <span class="fu">aes</span>(<span class="at">x =</span> .data[[focal_domain_resolution]],</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>                           <span class="at">y =</span> .data[[focal_celltype_resolution]], </span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>                           <span class="at">fill =</span> proportion)) <span class="sc">+</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="at">color =</span> <span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> <span class="fu">paste0</span>(<span class="fu">round</span>(<span class="dv">100</span><span class="sc">*</span>proportion, <span class="dv">3</span>), <span class="st">"%"</span>)), <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient</span>(<span class="at">low =</span> <span class="st">"white"</span>, <span class="at">high =</span> <span class="st">"dodgerblue"</span>) <span class="sc">+</span>  <span class="co"># Set color scale</span></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Cell Type"</span>, <span class="at">x =</span> <span class="st">"Spatial Domain"</span>, <span class="at">fill =</span> <span class="st">"Proportion"</span>) <span class="sc">+</span></span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) </span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(ct_dir, <span class="st">"contingency_table.png"</span>), </span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">5</span>, <span class="at">height=</span><span class="dv">7</span>)</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(plot_data, <span class="st">"./tmp.csv"</span>, <span class="at">col.names=</span>T, <span class="at">row.names=</span>F, <span class="at">quote=</span>F)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-contingency-table" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-contingency-table-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/ct/contingency_table.png" class="img-fluid figure-img" width="750">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-contingency-table-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.7: The proportion of a given cell type found in each domain. For example, domain D1015 is comprised of 65.3% smooth muscle cells (SMCs).
</figcaption>
</figure>
</div>
</div>
</div>
<p>From this information and the spatial layout of domains we can assign a function names <a href="#tbl-domains-definitions" class="quarto-xref">Table&nbsp;<span>6.2</span></a>. Just looking at the composition, the differences between three of the domains (D1011, D1015, and D1016) appear to have varying degrees of tumor, cycling tumor, and CAFs. D1015 has relatively more CAFs so let’s call that a Desmoplastic Stroma domain and let’s label D1016, which has the highest concentration of tumor cells, the Tumor Core. D1011 has abundant tumor (cycling, specifically) and also 12% B cell and 20% (normal) epithelial cells. Spatially, D1011 occurs throughout lymphoid structures within the epithelium and around the tumor region (invasive margin).</p>
<div id="tbl-domains-definitions" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-domains-definitions-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;6.2: Functional names associated with each <code>novae</code> domain.
</figcaption>
<div aria-describedby="tbl-domains-definitions-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<colgroup>
<col style="width: 33%">
<col style="width: 33%">
<col style="width: 33%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Domain</th>
<th style="text-align: left;">Inferred Biological Identity</th>
<th style="text-align: left;">Top 3 Cell Types (Proportion)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>D1000</strong></td>
<td style="text-align: left;"><strong>Normal Mucosa</strong></td>
<td style="text-align: left;">1. epithelial (53.4%), 2. plasma_IgA (17.2%), 3. mix_NK_Lymphoid (6.8%)</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>D1003</strong></td>
<td style="text-align: left;"><strong>Muscularis Layer</strong></td>
<td style="text-align: left;">1. smc (65.3%), 2. fibroblast (7.7%), 3. macrophage (6.2%)</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>D1011</strong></td>
<td style="text-align: left;"><strong>Invasive Margin (Immune-Rich)</strong></td>
<td style="text-align: left;">1. epithelial (20.4%), 2. tumor_cyling (18.5%), 3. bcell (12.0%)</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>D1015</strong></td>
<td style="text-align: left;"><strong>Desmoplastic Stroma</strong></td>
<td style="text-align: left;">1. CAF (34.7%), 2. tumor_cyling (11.6%), 3. monocyte (8.3%)</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>D1016</strong></td>
<td style="text-align: left;"><strong>Tumor Core</strong></td>
<td style="text-align: left;">1. tumor_cyling (31.8%), 2. CAF (22.2%), 3. tumor (15.4%)</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>D1017</strong></td>
<td style="text-align: left;"><strong>Heterogeneous</strong></td>
<td style="text-align: left;">1. tumor_cyling (16.1%), 2. undetermined (12.2%), 3. epithelial (12.1%)</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p>Let’s add a new column in the metadata with these values.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Python Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>domain_map <span class="op">=</span> {</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1000'</span>: <span class="st">'Normal Mucosa'</span>,</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1003'</span>: <span class="st">'Muscularis Layer'</span>,</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1011'</span>: <span class="st">'Invasive Margin'</span>,</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1015'</span>: <span class="st">'Desmoplastic Stroma'</span>,</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1016'</span>: <span class="st">'Tumor Core'</span>,</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1017'</span>: <span class="st">'Heterogeneous'</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'annotated_domain'</span>] <span class="op">=</span> adata.obs[focal_domain_resolution].<span class="bu">map</span>(domain_map)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-7-domain_annotations.h5ad"</span>)</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>  filename,</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>  compression<span class="op">=</span>hdf5plugin.FILTERS[<span class="st">"zstd"</span>],</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  compression_opts<span class="op">=</span>hdf5plugin.Zstd(clevel<span class="op">=</span><span class="dv">5</span>).filter_options</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>And while were at it, let’s plot the domains in XY space using <code>plotDots</code> with the new labels.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>domain_colors <span class="ot">&lt;-</span> results_list[[<span class="st">'novae_model_1_domain_colors'</span>]]</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(domain_colors) <span class="ot">&lt;-</span> results_list[[<span class="st">'novae_model_1_domain_names'</span>]]</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>domain_map <span class="ot">&lt;-</span> py<span class="sc">$</span>domain_map</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>domain_map <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">'domain_id'</span> <span class="ot">=</span> <span class="fu">names</span>(domain_map), </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>                         <span class="st">'domain_description'</span><span class="ot">=</span><span class="fu">unlist</span>(domain_map))</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>domain_colors <span class="ot">&lt;-</span> domain_colors[<span class="fu">match</span>(domain_map<span class="sc">$</span>domain_id, <span class="fu">names</span>(domain_colors))]</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>domain_map<span class="sc">$</span>domain_color <span class="ot">&lt;-</span> domain_colors</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'domain_colors'</span>]] <span class="ot">&lt;-</span> domain_map</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>colors_use <span class="ot">&lt;-</span> domain_map<span class="sc">$</span>domain_color</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(colors_use) <span class="ot">&lt;-</span> domain_map<span class="sc">$</span>domain_description</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>group_prefix <span class="ot">&lt;-</span> <span class="st">"annotated_domain"</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, <span class="at">color_by=</span><span class="st">'annotated_domain'</span>,</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">TRUE</span>,</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">FALSE</span>,</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span></span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_bar_params =</span> <span class="fu">list</span>(</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>                    <span class="at">location =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">0</span>),</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a>                    <span class="at">width =</span> <span class="dv">2</span>,</span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n =</span> <span class="dv">3</span>,</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>                    <span class="at">height =</span> <span class="fl">0.1</span>,</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>                    <span class="at">scale_colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"grey30"</span>),</span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>                    <span class="at">label_nudge_y =</span> <span class="sc">-</span><span class="fl">0.3</span></span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">200</span>,</span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">8</span>, </span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">8</span>,</span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="cb16-43"><a href="#cb16-43" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"X (mm)"</span>),</span>
<span id="cb16-44"><a href="#cb16-44" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"Y (mm)"</span>), </span>
<span id="cb16-45"><a href="#cb16-45" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="cb16-46"><a href="#cb16-46" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors_use),</span>
<span id="cb16-47"><a href="#cb16-47" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">0.4</span>)),</span>
<span id="cb16-48"><a href="#cb16-48" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="cb16-49"><a href="#cb16-49" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Spatial Domain"</span>,</span>
<span id="cb16-50"><a href="#cb16-50" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="cb16-51"><a href="#cb16-51" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb16-52"><a href="#cb16-52" aria-hidden="true" tabindex="-1"></a>              )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-xy-domains" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-xy-domains-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/ct/annotated_domain__spatial__S0__plot.png" class="img-fluid figure-img" width="800">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-xy-domains-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.8: XY with spatial domains.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="conclusion" class="level2" data-number="6.5">
<h2 data-number="6.5" class="anchored" data-anchor-id="conclusion"><span class="header-section-number">6.5</span> Conclusion</h2>
<p>We have now mapped the cellular landscape of our tissue. By combining broad unsupervised clustering and the specialized HieraType algorithm, we have moved from a matrix of counts to a spatially annotated map. Then we used this information to create functional names to the domains that we found in <a href="spatial-domains.html" class="quarto-xref"><span>Chapter 5</span></a>.</p>
<p>It is important to recognize that this cell typing workflow is a foundational draft. Depending on your specific research questions, such as defining specialized epithelial cell types or mapping complex stromal niches, more advanced methods may be required.</p>
<p>With our cells now labeled with a type and a domain, we are ready to ask “tertiary” questions. At the moment, this includes an analysis on pathway enrichment but more analysis chapters are in the works.</p>


<!-- -->

<script type="ojs-module-contents">
eyJjb250ZW50cyI6W119
</script>
<div id="exercise-loading-indicator" class="exercise-loading-indicator d-none d-flex align-items-center gap-2">
<div id="exercise-loading-status" class="d-flex gap-2">

</div>
<div class="spinner-grow spinner-grow-sm">

</div>
</div>
<script type="vfs-file">
["./assets/analysis_results/ct/broad__umap_pr__S0__facet_bcell.png","./assets/analysis_results/ct/broad__spatial__S0__facet_fibroblast.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_th17.png","./assets/analysis_results/ct/fine__spatial__S0__facet_plasma_IgA.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_tcm.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_th2.png","./assets/analysis_results/ct/fine__spatial__S0__facet_monocyte.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_fibroblast.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_neutrophil.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_CAF.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_tem.png","./assets/analysis_results/ct/fine__spatial__S0__facet_dendritic.png","./assets/analysis_results/ct/fine__spatial__S0__facet_bcell.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_tumor_cyling.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_treg.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_dendritic.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd8_tcm.png","./assets/analysis_results/ct/broad__spatial__S0__facet_undetermined.png","./assets/analysis_results/ct/fine__spatial__S0__facet_neutrophil.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd8_cytotoxic.png","./assets/analysis_results/ct/fine__spatial__S0__facet_CAF.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_monocyte.png","./assets/analysis_results/ct/broad__spatial__S0__facet_tcell.png","./assets/analysis_results/ct/rank_genes_groups_leiden_marker_gene_ranks_global.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_plasma_IgG.png","./assets/analysis_results/ct/dotplot__marker_gene_dotplot_global.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_undetermined.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_plasma.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd8_tem.png","./assets/analysis_results/ct/broad__spatial__S0__facet_neutrophil.png","./assets/analysis_results/ct/broad__spatial__S0__facet_CAF.png","./assets/analysis_results/ct/annotated_domain__spatial__S0__plot.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_CAF.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_neutrophil.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_tcell.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_undetermined.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_tem.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_th2.png","./assets/analysis_results/ct/fine__spatial__S0__facet_fibroblast.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_bcell.png","./assets/analysis_results/ct/broad__spatial__S0__facet_bcell.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd8_tcm.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd8_cytotoxic.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_fibroblast.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_th17.png","./assets/analysis_results/ct/broad__spatial__S0__facet_monocyte.png","./assets/analysis_results/ct/fine__spatial__S0__facet_plasma_IgG.png","./assets/analysis_results/ct/fine__spatial__S0__facet_undetermined.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_plasma_IgA.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_tcm.png","./assets/analysis_results/ct/broad__spatial__S0__facet_plasma.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd8_tem.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_monocyte.png","./assets/analysis_results/ct/broad__spatial__S0__facet_dendritic.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_treg.png","./assets/analysis_results/ct/fine__spatial__S0__facet_tumor_cyling.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_dendritic.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_nk.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd8_exhausted.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_smc.png","./assets/analysis_results/ct/broad__spatial__S0__facet_glial.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_endothelial.png","./assets/analysis_results/ct/broad__spatial__S0__facet_endothelial.png","./assets/analysis_results/ct/broad__spatial__S0__facet_nk.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_glial.png","./assets/analysis_results/ct/broad__spatial__S0__facet_epithelial.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd8_naive.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_th1.png","./assets/analysis_results/ct/fine__spatial__S0__facet_tumor.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_mix_NK_Lymphoid.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_epithelial.png","./assets/analysis_results/ct/fine__spatial__S0__facet_macrophage.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_naive.png","./assets/analysis_results/ct/fine__umap_pr__S0__plot.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_tumor.png","./assets/analysis_results/ct/fine__spatial__S0__facet_smc.png","./assets/analysis_results/ct/broad__spatial__S0__plot.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_macrophage.png","./assets/analysis_results/ct/broad__spatial__S0__facet_mast.png","./assets/analysis_results/ct/fine__spatial__S0__facet_mix_NK_Lymphoid.png","./assets/analysis_results/ct/broad__spatial__S0__facet_pericyte.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_mast.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_pericyte.png","./assets/analysis_results/ct/broad__spatial__S0__facet_mix_NK_Lymphoid.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd8_naive.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_mast.png","./assets/analysis_results/ct/fine__spatial__S0__facet_epithelial.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_mix_NK_Lymphoid.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_pericyte.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_smc.png","./assets/analysis_results/ct/fine__spatial__S0__facet_glial.png","./assets/analysis_results/ct/contingency_table.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd8_exhausted.png","./assets/analysis_results/ct/broad__spatial__S0__facet_smc.png","./assets/analysis_results/ct/broad__umap_pr__S0__plot.png","./assets/analysis_results/ct/fine__spatial__S0__facet_nk.png","./assets/analysis_results/ct/fine__spatial__S0__facet_endothelial.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_glial.png","./assets/analysis_results/ct/fine__spatial__S0__facet_mast.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_nk.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_endothelial.png","./assets/analysis_results/ct/fine__spatial__S0__facet_pericyte.png","./assets/analysis_results/ct/broad__umap_pr__S0__facet_epithelial.png","./assets/analysis_results/ct/broad__spatial__S0__facet_macrophage.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_tumor.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_cd4_th1.png","./assets/analysis_results/ct/broad__spatial__S0__facet_tumor.png","./assets/analysis_results/ct/fine__spatial__S0__facet_cd4_naive.png","./assets/analysis_results/ct/fine__spatial__S0__plot.png","./assets/analysis_results/ct/fine__umap_pr__S0__facet_macrophage.png"]
</script>
</section>

</main> <!-- /main -->
<script type="ojs-module-contents">
eyJjb250ZW50cyI6W119
</script>
<script type="module">
if (window.location.protocol === "file:") { alert("The OJS runtime does not work with file:// URLs. Please use a web server to view this document."); }
window._ojs.paths.runtimeToDoc = "../..";
window._ojs.paths.runtimeToRoot = "../..";
window._ojs.paths.docToRoot = "";
window._ojs.selfContained = false;
window._ojs.runtime.interpretFromScriptTags();
</script>
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    // Ensure there is a toggle, if there isn't float one in the top right
    if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
      const a = window.document.createElement('a');
      a.classList.add('top-right');
      a.classList.add('quarto-color-scheme-toggle');
      a.href = "";
      a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
      const i = window.document.createElement("i");
      i.classList.add('bi');
      a.appendChild(i);
      window.document.body.appendChild(a);
    }
    setColorSchemeToggle(hasAlternateSentinel())
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/nanostring-biostats\.github\.io\/CosMx-Analysis-Scratch-Space\/the-wtx-guide");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        const annoteTargets = window.document.querySelectorAll('.code-annotation-anchor');
        for (let i=0; i<annoteTargets.length; i++) {
          const annoteTarget = annoteTargets[i];
          const targetCell = annoteTarget.getAttribute("data-target-cell");
          const targetAnnotation = annoteTarget.getAttribute("data-target-annotation");
          const contentFn = () => {
            const content = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
            if (content) {
              const tipContent = content.cloneNode(true);
              tipContent.classList.add("code-annotation-tip-content");
              return tipContent.outerHTML;
            }
          }
          const config = {
            allowHTML: true,
            content: contentFn,
            onShow: (instance) => {
              selectCodeLines(instance.reference);
              instance.reference.classList.add('code-annotation-active');
              window.tippy.hideAll();
            },
            onHide: (instance) => {
              unselectCodeLines();
              instance.reference.classList.remove('code-annotation-active');
            },
            maxWidth: 300,
            delay: [50, 0],
            duration: [200, 0],
            offset: [5, 10],
            arrow: true,
            appendTo: function(el) {
              return el.parentElement.parentElement.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'quarto',
            placement: 'right',
            popperOptions: {
              modifiers: [
              {
                name: 'flip',
                options: {
                  flipVariations: false, // true by default
                  allowedAutoPlacements: ['right'],
                  fallbackPlacements: ['right', 'top', 'top-start', 'top-end', 'bottom', 'bottom-start', 'bottom-end', 'left'],
                },
              },
              {
                name: 'preventOverflow',
                options: {
                  mainAxis: false,
                  altAxis: false
                }
              }
              ]        
            }      
          };
          window.tippy(annoteTarget, config); 
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./spatial-domains.html" class="pagination-link" aria-label="Spatial Domains">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Spatial Domains</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./pathways.html" class="pagination-link" aria-label="Pathways">
        <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Pathways</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="an">subtitle:</span><span class="co"> Classifying cells into types</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> </span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co">- name: Evelyn Metzger</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co">  orcid: 0000-0002-4074-9003</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co">  affiliations: </span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co">  - ref: bsb</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="co">  - ref: eveilyeverafter</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span><span class="co"> </span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="co">  eval: false</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="co">  freeze: auto</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="co">  message: true</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="co">  warning: false</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="an">self-contained:</span><span class="co"> false</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="an">code-fold:</span><span class="co"> false</span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a><span class="an">code-tools:</span><span class="co"> true</span></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a><span class="an">code-annotations:</span><span class="co"> hover</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a><span class="an">engine:</span><span class="co"> knitr</span></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a><span class="an">prefer-html:</span><span class="co"> true</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span><span class="co"> live-html</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a><span class="an">resources:</span></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a><span class="co">  - ./assets/analysis_results/ct</span></span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>{{&lt; include ./_extensions/r-wasm/live/_knitr.qmd &gt;}}</span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a><span class="fu"># Cell Typing {#sec-cell-typing}</span></span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a>Okay, let's chart the course for this crucial phase of our analysis expedition:</span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a>cell typing. This is where we assign identities to the cell clusters discovered</span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a>in the previous chapter, transforming abstract groups into recognizable</span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a>biological entities like T cells, macrophages, or tumor cells.</span>
<span id="cb17-33"><a href="#cb17-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-34"><a href="#cb17-34" aria-hidden="true" tabindex="-1"></a>If pre-processing was about plotting our trajectory through high-dimensional</span>
<span id="cb17-35"><a href="#cb17-35" aria-hidden="true" tabindex="-1"></a>space, cell typing is akin to navigating a dense asteroid field. It's a critical</span>
<span id="cb17-36"><a href="#cb17-36" aria-hidden="true" tabindex="-1"></a>step, but one fraught with potential obstacles. Assigning definitive labels can</span>
<span id="cb17-37"><a href="#cb17-37" aria-hidden="true" tabindex="-1"></a>be challenging, often requiring careful consideration of marker genes and</span>
<span id="cb17-38"><a href="#cb17-38" aria-hidden="true" tabindex="-1"></a>existing biological knowledge. What constitutes a "correct" cell type label can</span>
<span id="cb17-39"><a href="#cb17-39" aria-hidden="true" tabindex="-1"></a>even be subjective, depending entirely on the resolution needed for your</span>
<span id="cb17-40"><a href="#cb17-40" aria-hidden="true" tabindex="-1"></a>specific biological questions. Are you aiming to distinguish broad categories</span>
<span id="cb17-41"><a href="#cb17-41" aria-hidden="true" tabindex="-1"></a>like "immune" versus "tumor," or do you need to pinpoint specific T cell</span>
<span id="cb17-42"><a href="#cb17-42" aria-hidden="true" tabindex="-1"></a>subtypes? Like adjusting a microscope's focus, the level of detail required</span>
<span id="cb17-43"><a href="#cb17-43" aria-hidden="true" tabindex="-1"></a>dictates the approach. Navigating this field requires careful maneuvering and</span>
<span id="cb17-44"><a href="#cb17-44" aria-hidden="true" tabindex="-1"></a>the right instruments. Autopilot hasn't been invented just yet and so the best</span>
<span id="cb17-45"><a href="#cb17-45" aria-hidden="true" tabindex="-1"></a>recommendation need manual input. </span>
<span id="cb17-46"><a href="#cb17-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-47"><a href="#cb17-47" aria-hidden="true" tabindex="-1"></a>One common hazard is attempting to directly apply analysis pipelines designed</span>
<span id="cb17-48"><a href="#cb17-48" aria-hidden="true" tabindex="-1"></a>for spatially-dissociated scRNA-seq. While scRNA-seq</span>
<span id="cb17-49"><a href="#cb17-49" aria-hidden="true" tabindex="-1"></a>vignettes offer valuable starting points, CosMx SMI data, generated by direct _in</span>
<span id="cb17-50"><a href="#cb17-50" aria-hidden="true" tabindex="-1"></a>situ_ hybridization, has fundamentally different characteristics. Simply copy-pasting</span>
<span id="cb17-51"><a href="#cb17-51" aria-hidden="true" tabindex="-1"></a>scRNA-seq methods without adaptation can lead to suboptimal or incomplete cell</span>
<span id="cb17-52"><a href="#cb17-52" aria-hidden="true" tabindex="-1"></a>type assignments. Instead, we must employ or adapt methods that leverage the</span>
<span id="cb17-53"><a href="#cb17-53" aria-hidden="true" tabindex="-1"></a>strengths of high-plex, direct hybridization data to achieve reliable</span>
<span id="cb17-54"><a href="#cb17-54" aria-hidden="true" tabindex="-1"></a>annotations. In doing so it's possible to achieve the most comprehensive and biologically-rich cell typing and sub-typing possible in spatial transcriptomics. </span>
<span id="cb17-55"><a href="#cb17-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-56"><a href="#cb17-56" aria-hidden="true" tabindex="-1"></a>It's also crucial to remember that cell typing, while essential, isn't the final</span>
<span id="cb17-57"><a href="#cb17-57" aria-hidden="true" tabindex="-1"></a>destination of our journey. It's a powerful tool for organizing our cellular</span>
<span id="cb17-58"><a href="#cb17-58" aria-hidden="true" tabindex="-1"></a>map, allowing us to ask spatially-aware questions. However, other organizational</span>
<span id="cb17-59"><a href="#cb17-59" aria-hidden="true" tabindex="-1"></a>frameworks, such as classifying cells based on their spatial neighborhood or</span>
<span id="cb17-60"><a href="#cb17-60" aria-hidden="true" tabindex="-1"></a>"niche," are equally important for understanding the tissue's complex ecosystem.</span>
<span id="cb17-61"><a href="#cb17-61" aria-hidden="true" tabindex="-1"></a>Ultimately, cell typing is foundational: it establishes the structural anatomy </span>
<span id="cb17-62"><a href="#cb17-62" aria-hidden="true" tabindex="-1"></a>of the dataset, paving the way for us to investigate its functional physiology </span>
<span id="cb17-63"><a href="#cb17-63" aria-hidden="true" tabindex="-1"></a>in the next chapters.</span>
<span id="cb17-64"><a href="#cb17-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-65"><a href="#cb17-65" aria-hidden="true" tabindex="-1"></a>In this chapter, we will navigate the cell typing process in the colon</span>
<span id="cb17-66"><a href="#cb17-66" aria-hidden="true" tabindex="-1"></a>dataset. The main goal is to generate cell type labels that are "close" while</span>
<span id="cb17-67"><a href="#cb17-67" aria-hidden="true" tabindex="-1"></a>also being approachable. I'll begin (@sec-ct-approaches) by surveying some of</span>
<span id="cb17-68"><a href="#cb17-68" aria-hidden="true" tabindex="-1"></a>the spatially unaware methods and then summarize the state of the direct</span>
<span id="cb17-69"><a href="#cb17-69" aria-hidden="true" tabindex="-1"></a>hybridization-inspired approaches. Then I will guide you through the specific,</span>
<span id="cb17-70"><a href="#cb17-70" aria-hidden="true" tabindex="-1"></a>hybrid workflow used for this colon dataset. We will start with the broad Leiden</span>
<span id="cb17-71"><a href="#cb17-71" aria-hidden="true" tabindex="-1"></a>clusters defined in the previous chapter, refine them based on spatial</span>
<span id="cb17-72"><a href="#cb17-72" aria-hidden="true" tabindex="-1"></a>observations, and assign primary identities using marker gene analysis. Finally,</span>
<span id="cb17-73"><a href="#cb17-73" aria-hidden="true" tabindex="-1"></a>for the more complex immune populations where standard clustering often</span>
<span id="cb17-74"><a href="#cb17-74" aria-hidden="true" tabindex="-1"></a>struggles, we will employ HieraType, a probabilistic method explicitly designed</span>
<span id="cb17-75"><a href="#cb17-75" aria-hidden="true" tabindex="-1"></a>to handle the specific characteristics of CosMx SMI data.</span>
<span id="cb17-76"><a href="#cb17-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-77"><a href="#cb17-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-80"><a href="#cb17-80" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-81"><a href="#cb17-81" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: Preamble</span></span>
<span id="cb17-82"><a href="#cb17-82" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-83"><a href="#cb17-83" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb17-84"><a href="#cb17-84" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-85"><a href="#cb17-85" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-86"><a href="#cb17-86" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R code"</span></span>
<span id="cb17-87"><a href="#cb17-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-88"><a href="#cb17-88" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"./preamble.R"</span>)</span>
<span id="cb17-89"><a href="#cb17-89" aria-hidden="true" tabindex="-1"></a>reticulate<span class="sc">::</span><span class="fu">source_python</span>(<span class="st">"./preamble.py"</span>)</span>
<span id="cb17-90"><a href="#cb17-90" aria-hidden="true" tabindex="-1"></a>analysis_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">getwd</span>(), <span class="st">"analysis_results"</span>)</span>
<span id="cb17-91"><a href="#cb17-91" aria-hidden="true" tabindex="-1"></a>input_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"input_files"</span>)</span>
<span id="cb17-92"><a href="#cb17-92" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"output_files"</span>)</span>
<span id="cb17-93"><a href="#cb17-93" aria-hidden="true" tabindex="-1"></a>analysis_asset_dir <span class="ot">&lt;-</span> <span class="st">"./assets/analysis_results"</span> <span class="co"># &lt;1&gt;</span></span>
<span id="cb17-94"><a href="#cb17-94" aria-hidden="true" tabindex="-1"></a>ct_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(output_dir, <span class="st">"ct"</span>)</span>
<span id="cb17-95"><a href="#cb17-95" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(ct_dir)){</span>
<span id="cb17-96"><a href="#cb17-96" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(ct_dir, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-97"><a href="#cb17-97" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-98"><a href="#cb17-98" aria-hidden="true" tabindex="-1"></a>results_list_file <span class="ot">=</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"results_list.rds"</span>)</span>
<span id="cb17-99"><a href="#cb17-99" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(results_list_file)){</span>
<span id="cb17-100"><a href="#cb17-100" aria-hidden="true" tabindex="-1"></a>  results_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb17-101"><a href="#cb17-101" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb17-102"><a href="#cb17-102" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb17-103"><a href="#cb17-103" aria-hidden="true" tabindex="-1"></a>  results_list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(results_list_file)  </span>
<span id="cb17-104"><a href="#cb17-104" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-105"><a href="#cb17-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-106"><a href="#cb17-106" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb17-107"><a href="#cb17-107" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb17-108"><a href="#cb17-108" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"./helpers.R"</span>)</span>
<span id="cb17-109"><a href="#cb17-109" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-110"><a href="#cb17-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-111"><a href="#cb17-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-112"><a href="#cb17-112" aria-hidden="true" tabindex="-1"></a><span class="fu">## Cell Typing Approaches {#sec-ct-approaches}</span></span>
<span id="cb17-113"><a href="#cb17-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-114"><a href="#cb17-114" aria-hidden="true" tabindex="-1"></a><span class="fu">### Non-spatial Cell Typing Approaches</span></span>
<span id="cb17-115"><a href="#cb17-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-116"><a href="#cb17-116" aria-hidden="true" tabindex="-1"></a>Before discussing spatially-aware or technology-specific approaches, let's</span>
<span id="cb17-117"><a href="#cb17-117" aria-hidden="true" tabindex="-1"></a>briefly cover common methods designed for non-spatial, single-cell RNA sequencing analysis.</span>
<span id="cb17-118"><a href="#cb17-118" aria-hidden="true" tabindex="-1"></a>These techniques primarily leverage the gene expression matrix (<span class="in">`adata.X`</span> and</span>
<span id="cb17-119"><a href="#cb17-119" aria-hidden="true" tabindex="-1"></a>normalized layers) and the cluster assignments (e.g., <span class="in">`adata.obs['leiden']`</span>)</span>
<span id="cb17-120"><a href="#cb17-120" aria-hidden="true" tabindex="-1"></a>derived in the previous chapter. Many excellent vignettes exist for these</span>
<span id="cb17-121"><a href="#cb17-121" aria-hidden="true" tabindex="-1"></a>methods in popular frameworks like scanpy (Python) and Seurat (R), providing a</span>
<span id="cb17-122"><a href="#cb17-122" aria-hidden="true" tabindex="-1"></a>rough "10,000-foot view" of the data's cellular composition.</span>
<span id="cb17-123"><a href="#cb17-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-124"><a href="#cb17-124" aria-hidden="true" tabindex="-1"></a>**Cluster Marker Gene Identification (Leiden/Louvain):** This is often a</span>
<span id="cb17-125"><a href="#cb17-125" aria-hidden="true" tabindex="-1"></a>foundational step in scRNA-seq. After obtaining clusters (like the Leiden clusters from</span>
<span id="cb17-126"><a href="#cb17-126" aria-hidden="true" tabindex="-1"></a>@sec-preprocessing), differential expression tests are performed between each</span>
<span id="cb17-127"><a href="#cb17-127" aria-hidden="true" tabindex="-1"></a>cluster and all other cells (or between pairs of clusters). This identifies</span>
<span id="cb17-128"><a href="#cb17-128" aria-hidden="true" tabindex="-1"></a>genes significantly upregulated within a specific cluster. Tools like</span>
<span id="cb17-129"><a href="#cb17-129" aria-hidden="true" tabindex="-1"></a>scanpy's <span class="in">`rank_genes_groups`</span> or Seurat's <span class="in">`FindMarkers`</span> are standard. The resulting</span>
<span id="cb17-130"><a href="#cb17-130" aria-hidden="true" tabindex="-1"></a>lists of marker genes are then manually compared against known canonical markers</span>
<span id="cb17-131"><a href="#cb17-131" aria-hidden="true" tabindex="-1"></a>from literature or cell atlases to assign putative cell type identities (e.g.,</span>
<span id="cb17-132"><a href="#cb17-132" aria-hidden="true" tabindex="-1"></a>observing high _CD3E_, _CD8A_ suggests a T cell cluster; high _EPCAM_, _KRT19_ suggests</span>
<span id="cb17-133"><a href="#cb17-133" aria-hidden="true" tabindex="-1"></a>epithelial cells). While effective, this relies heavily on prior biological</span>
<span id="cb17-134"><a href="#cb17-134" aria-hidden="true" tabindex="-1"></a>knowledge and the quality of the initial clustering. It also assumes that transcripts</span>
<span id="cb17-135"><a href="#cb17-135" aria-hidden="true" tabindex="-1"></a>found in the cell do not arise from any spatial co-localization or imperfect </span>
<span id="cb17-136"><a href="#cb17-136" aria-hidden="true" tabindex="-1"></a>cell-cell segmentation boundaries that can arise when assigning transcripts to cells.</span>
<span id="cb17-137"><a href="#cb17-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-138"><a href="#cb17-138" aria-hidden="true" tabindex="-1"></a>**Nested / Hierarchical Clustering Analysis:** Sometimes, broad initial clusters</span>
<span id="cb17-139"><a href="#cb17-139" aria-hidden="true" tabindex="-1"></a>(like "Immune Cells" or "Fibroblasts") contain sub-populations that are meaningful</span>
<span id="cb17-140"><a href="#cb17-140" aria-hidden="true" tabindex="-1"></a>for your analysis. A</span>
<span id="cb17-141"><a href="#cb17-141" aria-hidden="true" tabindex="-1"></a>common refinement is then to subset the data to include only cells from a specific</span>
<span id="cb17-142"><a href="#cb17-142" aria-hidden="true" tabindex="-1"></a>Leiden cluster (or set of clusters) and then re-run the clustering</span>
<span id="cb17-143"><a href="#cb17-143" aria-hidden="true" tabindex="-1"></a>process within that subset. Identifying marker</span>
<span id="cb17-144"><a href="#cb17-144" aria-hidden="true" tabindex="-1"></a>genes for these sub-clusters can reveal finer cell types or states (_e.g._,</span>
<span id="cb17-145"><a href="#cb17-145" aria-hidden="true" tabindex="-1"></a>distinguishing CD4+ from CD8+ T cells, or identifying different fibroblast</span>
<span id="cb17-146"><a href="#cb17-146" aria-hidden="true" tabindex="-1"></a>subtypes. This iterative approach allows for a hierarchical understanding of</span>
<span id="cb17-147"><a href="#cb17-147" aria-hidden="true" tabindex="-1"></a>cell identity; however, the imperfect segmentation noted above is amplified in </span>
<span id="cb17-148"><a href="#cb17-148" aria-hidden="true" tabindex="-1"></a>this approach. Consider a workflow where one uses HVGs to create an initial set of</span>
<span id="cb17-149"><a href="#cb17-149" aria-hidden="true" tabindex="-1"></a>"Elementary" clusters and uses a marker gene-based approach to label these</span>
<span id="cb17-150"><a href="#cb17-150" aria-hidden="true" tabindex="-1"></a>broad clusters. When subsetting down to, say, fibroblasts and rerunning the above</span>
<span id="cb17-151"><a href="#cb17-151" aria-hidden="true" tabindex="-1"></a>HVG -&gt; Leiden -&gt; marker gene workflow, one may get non-fibroblast marker genes returned.</span>
<span id="cb17-152"><a href="#cb17-152" aria-hidden="true" tabindex="-1"></a>The reason for this is twofold. First the general fibroblast marker genes are no</span>
<span id="cb17-153"><a href="#cb17-153" aria-hidden="true" tabindex="-1"></a>longer as differentially expressed in the subset of Fibroblasts as they were in the full data</span>
<span id="cb17-154"><a href="#cb17-154" aria-hidden="true" tabindex="-1"></a>and 2) if there are cell segmentation errors of fibroblast sub-populations that are</span>
<span id="cb17-155"><a href="#cb17-155" aria-hidden="true" tabindex="-1"></a>spatially-associated with unrelated cell types -- like plasma cells -- the HVG</span>
<span id="cb17-156"><a href="#cb17-156" aria-hidden="true" tabindex="-1"></a>algorithm may identify spurious gene markers as highly variable. The result would</span>
<span id="cb17-157"><a href="#cb17-157" aria-hidden="true" tabindex="-1"></a>be markers like _JCHAIN_ in fibroblasts that happen to be located near a germinal</span>
<span id="cb17-158"><a href="#cb17-158" aria-hidden="true" tabindex="-1"></a>center, tertiary lymphoid structure, or similar. Thus, while this nested approach</span>
<span id="cb17-159"><a href="#cb17-159" aria-hidden="true" tabindex="-1"></a>can be valuable, it should be used with caution. For more information on this overlap</span>
<span id="cb17-160"><a href="#cb17-160" aria-hidden="true" tabindex="-1"></a>concept, see Dan McGuire's <span class="co">[</span><span class="ot">blog post</span><span class="co">](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/smiDE/)</span>{target="_blank"} on smiDE.</span>
<span id="cb17-161"><a href="#cb17-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-162"><a href="#cb17-162" aria-hidden="true" tabindex="-1"></a>**Reference-Based Annotation:** Instead of relying solely on _de novo_ marker finding,</span>
<span id="cb17-163"><a href="#cb17-163" aria-hidden="true" tabindex="-1"></a>many methods leverage existing annotated single-cell datasets (atlases) as</span>
<span id="cb17-164"><a href="#cb17-164" aria-hidden="true" tabindex="-1"></a>references. These algorithms typically project the query data</span>
<span id="cb17-165"><a href="#cb17-165" aria-hidden="true" tabindex="-1"></a>onto the reference or use statistical models trained on the reference to</span>
<span id="cb17-166"><a href="#cb17-166" aria-hidden="true" tabindex="-1"></a>directly assign cell type labels to your individual cells or clusters. Popular</span>
<span id="cb17-167"><a href="#cb17-167" aria-hidden="true" tabindex="-1"></a>tools include:</span>
<span id="cb17-168"><a href="#cb17-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-169"><a href="#cb17-169" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="co">[</span><span class="ot">`scanpy.tl.ingest`</span><span class="co">](https://scanpy.readthedocs.io/en/stable/tutorials/basics/integrating-data-using-ingest.html)</span>:</span>
<span id="cb17-170"><a href="#cb17-170" aria-hidden="true" tabindex="-1"></a>Integrates query data onto an existing annotated AnnData object.</span>
<span id="cb17-171"><a href="#cb17-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-172"><a href="#cb17-172" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Seurat's <span class="co">[</span><span class="ot">`FindTransferAnchors`</span><span class="co">](https://satijalab.org/seurat/reference/findtransferanchors)</span></span>
<span id="cb17-173"><a href="#cb17-173" aria-hidden="true" tabindex="-1"></a>and <span class="co">[</span><span class="ot">`TransferData`</span><span class="co">](https://satijalab.org/seurat/reference/transferdata)</span>: A widely used method in the R</span>
<span id="cb17-174"><a href="#cb17-174" aria-hidden="true" tabindex="-1"></a>ecosystem.</span>
<span id="cb17-175"><a href="#cb17-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-176"><a href="#cb17-176" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>InSituType's fully supervised method: InSituType -- while designed for spatial</span>
<span id="cb17-177"><a href="#cb17-177" aria-hidden="true" tabindex="-1"></a>data -- can be used as a fully supervised classifier. I'm including it here since</span>
<span id="cb17-178"><a href="#cb17-178" aria-hidden="true" tabindex="-1"></a>many reference datasets and atlases at the time of writing are based on scRNA-seq</span>
<span id="cb17-179"><a href="#cb17-179" aria-hidden="true" tabindex="-1"></a>data.</span>
<span id="cb17-180"><a href="#cb17-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-181"><a href="#cb17-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-182"><a href="#cb17-182" aria-hidden="true" tabindex="-1"></a><span class="fu">### Spatially-Informed and Hybridization-Aware Methods</span></span>
<span id="cb17-183"><a href="#cb17-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-184"><a href="#cb17-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-185"><a href="#cb17-185" aria-hidden="true" tabindex="-1"></a>Unlike standard scRNA-seq workflows, these approaches leverage the unique features of _in situ_ data: specifically the spatial coordinates, the negative probe background, and the available protein data.</span>
<span id="cb17-186"><a href="#cb17-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-187"><a href="#cb17-187" aria-hidden="true" tabindex="-1"></a>**Background-Aware Modeling**. Methods like <span class="in">`InSituType`</span> explicitly model the background noise using the negative control probes. This allows for more accurate classification of cells with low transcript counts that might be discarded or misclassified by standard scRNA-seq tools</span>
<span id="cb17-188"><a href="#cb17-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-189"><a href="#cb17-189" aria-hidden="true" tabindex="-1"></a>**Spatial-Molecular Joint Inference**. Tools like <span class="co">[</span><span class="ot">Banksy</span><span class="co">](https://www.nature.com/articles/s41588-024-01664-3)</span> utilize the expression of a cell's physical neighbors to inform its identity. This leverages the biological reality that cells of the same type often cluster spatially, helping to smooth out technical noise in individual cells.</span>
<span id="cb17-190"><a href="#cb17-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-191"><a href="#cb17-191" aria-hidden="true" tabindex="-1"></a>**Multi-modal Integration**. CosMx SMI often includes protein markers (e.g., CD45, PanCK) in the form of IF markers or -- more recently -- with true RNA and protein same-slide multiomics. "Dual-mode" typing strategies use these robust protein signals to define major lineages (e.g., "This cell is definitely CD45+ Immune") before using the RNA data to determine the specific subtype (e.g., "It is a CD8+ T Cell"). Similarly, <span class="co">[</span><span class="ot">HieraType</span><span class="co">](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/multiomics-hieratype/#introduction)</span></span>
<span id="cb17-192"><a href="#cb17-192" aria-hidden="true" tabindex="-1"></a>is a new R package for supervised classification that can combine multi-omics</span>
<span id="cb17-193"><a href="#cb17-193" aria-hidden="true" tabindex="-1"></a>information to cell type. HieraType also classifies RNA-only data (as we'll see below).</span>
<span id="cb17-194"><a href="#cb17-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-195"><a href="#cb17-195" aria-hidden="true" tabindex="-1"></a>**Reference-Based Annotation:** Reference profiles based on CosMx SMI can be used to </span>
<span id="cb17-196"><a href="#cb17-196" aria-hidden="true" tabindex="-1"></a>more quickly cell type new datasets. This approach can be much faster than cell</span>
<span id="cb17-197"><a href="#cb17-197" aria-hidden="true" tabindex="-1"></a>typing _de novo_ but requires such annotations to be available.</span>
<span id="cb17-198"><a href="#cb17-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-199"><a href="#cb17-199" aria-hidden="true" tabindex="-1"></a><span class="fu">## Cell Type Determination</span></span>
<span id="cb17-200"><a href="#cb17-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-201"><a href="#cb17-201" aria-hidden="true" tabindex="-1"></a>Just like there are many paths through an asteroid field, there are many ways to</span>
<span id="cb17-202"><a href="#cb17-202" aria-hidden="true" tabindex="-1"></a>navigate cell typing. I find the following approach makes a good starting place</span>
<span id="cb17-203"><a href="#cb17-203" aria-hidden="true" tabindex="-1"></a>for any analysis as it's relatively fast while still allowing for human intervention.</span>
<span id="cb17-204"><a href="#cb17-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-205"><a href="#cb17-205" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Use Leiden clusters identified in @sec-preprocessing as the basis.</span>
<span id="cb17-206"><a href="#cb17-206" aria-hidden="true" tabindex="-1"></a> 1.1. _Optional_: refine Leiden clusters based on discovered biology. For example,</span>
<span id="cb17-207"><a href="#cb17-207" aria-hidden="true" tabindex="-1"></a>there may be some Leiden clusters that makes sense to merge together or some evidence</span>
<span id="cb17-208"><a href="#cb17-208" aria-hidden="true" tabindex="-1"></a>that splitting a Leiden cluster into two or splitting a given</span>
<span id="cb17-209"><a href="#cb17-209" aria-hidden="true" tabindex="-1"></a>Leiden cluster into multiple sub-clusters based on the spatial layout of cells.</span>
<span id="cb17-210"><a href="#cb17-210" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Run HieraType to classify cells and use its integration function to merge its</span>
<span id="cb17-211"><a href="#cb17-211" aria-hidden="true" tabindex="-1"></a>supervised classifications with the unsupervised Leiden clusters.</span>
<span id="cb17-212"><a href="#cb17-212" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Of the remaining unclassified cells, use marker genes to infer the most likely</span>
<span id="cb17-213"><a href="#cb17-213" aria-hidden="true" tabindex="-1"></a>cell type.</span>
<span id="cb17-214"><a href="#cb17-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-215"><a href="#cb17-215" aria-hidden="true" tabindex="-1"></a><span class="fu">### Step 1: Read Leiden clusters</span></span>
<span id="cb17-216"><a href="#cb17-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-217"><a href="#cb17-217" aria-hidden="true" tabindex="-1"></a>Load the data and make a copy of the <span class="in">`leiden_pr`</span> metadata column that we'll use.</span>
<span id="cb17-218"><a href="#cb17-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-221"><a href="#cb17-221" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb17-222"><a href="#cb17-222" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-anndata</span></span>
<span id="cb17-223"><a href="#cb17-223" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-224"><a href="#cb17-224" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb17-225"><a href="#cb17-225" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-226"><a href="#cb17-226" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-227"><a href="#cb17-227" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-228"><a href="#cb17-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-229"><a href="#cb17-229" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> ad.read_h5ad(os.path.join(r.analysis_dir, <span class="st">"anndata-2-leiden.h5ad"</span>))</span>
<span id="cb17-230"><a href="#cb17-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-231"><a href="#cb17-231" aria-hidden="true" tabindex="-1"></a>cluster_renaming <span class="op">=</span> {}</span>
<span id="cb17-232"><a href="#cb17-232" aria-hidden="true" tabindex="-1"></a>initial <span class="op">=</span> <span class="st">"C"</span></span>
<span id="cb17-233"><a href="#cb17-233" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> name <span class="kw">in</span> adata.obs[<span class="st">'leiden_pr'</span>].unique():</span>
<span id="cb17-234"><a href="#cb17-234" aria-hidden="true" tabindex="-1"></a>  cluster_renaming[name] <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>initial<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>name<span class="sc">}</span><span class="ss">"</span></span>
<span id="cb17-235"><a href="#cb17-235" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'leiden'</span>] <span class="op">=</span> adata.obs[<span class="st">'leiden_pr'</span>].<span class="bu">map</span>(cluster_renaming)</span>
<span id="cb17-236"><a href="#cb17-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-237"><a href="#cb17-237" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-238"><a href="#cb17-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-239"><a href="#cb17-239" aria-hidden="true" tabindex="-1"></a><span class="fu">### Step 2: Run HieraType</span></span>
<span id="cb17-240"><a href="#cb17-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-241"><a href="#cb17-241" aria-hidden="true" tabindex="-1"></a>Install the HieraType package (and remotes), if needed. Then extract the lists</span>
<span id="cb17-242"><a href="#cb17-242" aria-hidden="true" tabindex="-1"></a>of marker genes.</span>
<span id="cb17-243"><a href="#cb17-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-246"><a href="#cb17-246" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-247"><a href="#cb17-247" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: H-0</span></span>
<span id="cb17-248"><a href="#cb17-248" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-249"><a href="#cb17-249" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb17-250"><a href="#cb17-250" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-251"><a href="#cb17-251" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-252"><a href="#cb17-252" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-253"><a href="#cb17-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-254"><a href="#cb17-254" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"mvtnorm"</span>)</span>
<span id="cb17-255"><a href="#cb17-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-256"><a href="#cb17-256" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"Nanostring-Biostats/CosMx-Analysis-Scratch-Space"</span>,</span>
<span id="cb17-257"><a href="#cb17-257" aria-hidden="true" tabindex="-1"></a>                         <span class="at">subdir =</span> <span class="st">"_code/HieraType"</span>, <span class="at">ref =</span> <span class="st">"Main"</span>)</span>
<span id="cb17-258"><a href="#cb17-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-259"><a href="#cb17-259" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(HieraType)</span>
<span id="cb17-260"><a href="#cb17-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-261"><a href="#cb17-261" aria-hidden="true" tabindex="-1"></a>markerg <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">c</span>(</span>
<span id="cb17-262"><a href="#cb17-262" aria-hidden="true" tabindex="-1"></a>  HieraType<span class="sc">::</span>markerslist_l1</span>
<span id="cb17-263"><a href="#cb17-263" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_immune</span>
<span id="cb17-264"><a href="#cb17-264" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd8tminor</span>
<span id="cb17-265"><a href="#cb17-265" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_cd4tminor</span>
<span id="cb17-266"><a href="#cb17-266" aria-hidden="true" tabindex="-1"></a>  ,HieraType<span class="sc">::</span>markerslist_tcellmajor</span>
<span id="cb17-267"><a href="#cb17-267" aria-hidden="true" tabindex="-1"></a>), <span class="st">"[["</span>, <span class="st">"predictors"</span>))</span>
<span id="cb17-268"><a href="#cb17-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-269"><a href="#cb17-269" aria-hidden="true" tabindex="-1"></a>marker_genes <span class="ot">&lt;-</span> <span class="fu">as.character</span>(markerg)</span>
<span id="cb17-270"><a href="#cb17-270" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-271"><a href="#cb17-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-272"><a href="#cb17-272" aria-hidden="true" tabindex="-1"></a>Convert data to R.</span>
<span id="cb17-273"><a href="#cb17-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-276"><a href="#cb17-276" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb17-277"><a href="#cb17-277" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: H-1</span></span>
<span id="cb17-278"><a href="#cb17-278" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-279"><a href="#cb17-279" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb17-280"><a href="#cb17-280" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-281"><a href="#cb17-281" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-282"><a href="#cb17-282" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-283"><a href="#cb17-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-284"><a href="#cb17-284" aria-hidden="true" tabindex="-1"></a>total_counts_per_cell <span class="op">=</span> adata.layers[<span class="st">'counts'</span>].<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb17-285"><a href="#cb17-285" aria-hidden="true" tabindex="-1"></a>total_counts_per_cell <span class="op">=</span> np.asarray(total_counts_per_cell).flatten()</span>
<span id="cb17-286"><a href="#cb17-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-287"><a href="#cb17-287" aria-hidden="true" tabindex="-1"></a>total_counts_per_target <span class="op">=</span> adata.layers[<span class="st">'counts'</span>].<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb17-288"><a href="#cb17-288" aria-hidden="true" tabindex="-1"></a>total_counts_per_target <span class="op">=</span> np.asarray(total_counts_per_target).flatten()</span>
<span id="cb17-289"><a href="#cb17-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-290"><a href="#cb17-290" aria-hidden="true" tabindex="-1"></a>target_frequencies <span class="op">=</span> total_counts_per_target <span class="op">/</span> total_counts_per_target.<span class="bu">sum</span>()</span>
<span id="cb17-291"><a href="#cb17-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-292"><a href="#cb17-292" aria-hidden="true" tabindex="-1"></a><span class="co"># Include all HVGs and marker genes (even if not in the HVGs)</span></span>
<span id="cb17-293"><a href="#cb17-293" aria-hidden="true" tabindex="-1"></a>marker_genes <span class="op">=</span> r.marker_genes</span>
<span id="cb17-294"><a href="#cb17-294" aria-hidden="true" tabindex="-1"></a>marker_mask <span class="op">=</span> adata.var.index.isin(marker_genes)</span>
<span id="cb17-295"><a href="#cb17-295" aria-hidden="true" tabindex="-1"></a>highly_variable_mask <span class="op">=</span> adata.var[<span class="st">'highly_variable'</span>].to_numpy()</span>
<span id="cb17-296"><a href="#cb17-296" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">'hieratype_target'</span>] <span class="op">=</span> (highly_variable_mask <span class="op">|</span> marker_mask)</span>
<span id="cb17-297"><a href="#cb17-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-298"><a href="#cb17-298" aria-hidden="true" tabindex="-1"></a>adata_ht <span class="op">=</span> adata[:, adata.var.hieratype_target].copy()</span>
<span id="cb17-299"><a href="#cb17-299" aria-hidden="true" tabindex="-1"></a>ht_counts <span class="op">=</span> adata_ht.layers[<span class="st">'counts'</span>].copy().astype(np.int64) <span class="co"># &lt;1&gt;</span></span>
<span id="cb17-300"><a href="#cb17-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-301"><a href="#cb17-301" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-302"><a href="#cb17-302" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>In order to use this matrix in R, we need to convert the 32 bit integer counts into 64 bit integers.</span>
<span id="cb17-303"><a href="#cb17-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-304"><a href="#cb17-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-305"><a href="#cb17-305" aria-hidden="true" tabindex="-1"></a>Follow <span class="co">[</span><span class="ot">this tutorial</span><span class="co">](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/multiomics-hieratype/#rnaonly)</span> to refine Leiden clusters with immune cells.</span>
<span id="cb17-306"><a href="#cb17-306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-309"><a href="#cb17-309" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-310"><a href="#cb17-310" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: H-2</span></span>
<span id="cb17-311"><a href="#cb17-311" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-312"><a href="#cb17-312" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb17-313"><a href="#cb17-313" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-314"><a href="#cb17-314" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-315"><a href="#cb17-315" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-316"><a href="#cb17-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-317"><a href="#cb17-317" aria-hidden="true" tabindex="-1"></a>ht_counts <span class="ot">&lt;-</span> py<span class="sc">$</span>ht_counts</span>
<span id="cb17-318"><a href="#cb17-318" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ht_counts) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata_ht<span class="sc">$</span>obs_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb17-319"><a href="#cb17-319" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ht_counts) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata_ht<span class="sc">$</span>var_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb17-320"><a href="#cb17-320" aria-hidden="true" tabindex="-1"></a>ht_counts <span class="ot">&lt;-</span> <span class="fu">as</span>(ht_counts, <span class="st">"CsparseMatrix"</span>)</span>
<span id="cb17-321"><a href="#cb17-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-322"><a href="#cb17-322" aria-hidden="true" tabindex="-1"></a>pca_embeddings <span class="ot">&lt;-</span> py<span class="sc">$</span>adata_ht<span class="sc">$</span>obsm[<span class="st">'X_pca_pr'</span>] <span class="co"># n_cells x n_pcs</span></span>
<span id="cb17-323"><a href="#cb17-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-324"><a href="#cb17-324" aria-hidden="true" tabindex="-1"></a>target_frequencies <span class="ot">&lt;-</span> py<span class="sc">$</span>target_frequencies </span>
<span id="cb17-325"><a href="#cb17-325" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(target_frequencies) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>var_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb17-326"><a href="#cb17-326" aria-hidden="true" tabindex="-1"></a>target_frequencies <span class="ot">&lt;-</span> target_frequencies[<span class="fu">colnames</span>(ht_counts)] </span>
<span id="cb17-327"><a href="#cb17-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-328"><a href="#cb17-328" aria-hidden="true" tabindex="-1"></a>total_counts_per_cell <span class="ot">&lt;-</span> py<span class="sc">$</span>total_counts_per_cell</span>
<span id="cb17-329"><a href="#cb17-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-330"><a href="#cb17-330" aria-hidden="true" tabindex="-1"></a><span class="co"># Assumes this was based on cosine.</span></span>
<span id="cb17-331"><a href="#cb17-331" aria-hidden="true" tabindex="-1"></a>sim_graph <span class="ot">&lt;-</span> py<span class="sc">$</span>adata_ht<span class="sc">$</span>obsp[<span class="st">'neighbors_pr_connectivities'</span>]</span>
<span id="cb17-332"><a href="#cb17-332" aria-hidden="true" tabindex="-1"></a>sim_graph <span class="ot">&lt;-</span> <span class="fu">as</span>(sim_graph, <span class="st">"CsparseMatrix"</span>)</span>
<span id="cb17-333"><a href="#cb17-333" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(sim_graph) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(sim_graph) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(ht_counts)</span>
<span id="cb17-334"><a href="#cb17-334" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-335"><a href="#cb17-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-336"><a href="#cb17-336" aria-hidden="true" tabindex="-1"></a>Run the pipeline. </span>
<span id="cb17-337"><a href="#cb17-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-340"><a href="#cb17-340" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-341"><a href="#cb17-341" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: H-3</span></span>
<span id="cb17-342"><a href="#cb17-342" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-343"><a href="#cb17-343" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb17-344"><a href="#cb17-344" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-345"><a href="#cb17-345" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-346"><a href="#cb17-346" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-347"><a href="#cb17-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-348"><a href="#cb17-348" aria-hidden="true" tabindex="-1"></a>pipeline <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">make_pipeline</span>(</span>
<span id="cb17-349"><a href="#cb17-349" aria-hidden="true" tabindex="-1"></a>  <span class="at">markerslists =</span> <span class="fu">list</span>(</span>
<span id="cb17-350"><a href="#cb17-350" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l1"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_l1</span>
<span id="cb17-351"><a href="#cb17-351" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"l2"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_immune</span>
<span id="cb17-352"><a href="#cb17-352" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_tcellmajor</span>
<span id="cb17-353"><a href="#cb17-353" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd4tminor</span>
<span id="cb17-354"><a href="#cb17-354" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_multiomic_cd8tminor</span>
<span id="cb17-355"><a href="#cb17-355" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb17-356"><a href="#cb17-356" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors =</span> <span class="fu">list</span>(</span>
<span id="cb17-357"><a href="#cb17-357" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"l1"</span></span>
<span id="cb17-358"><a href="#cb17-358" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"l2"</span></span>
<span id="cb17-359"><a href="#cb17-359" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb17-360"><a href="#cb17-360" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"lt"</span></span>
<span id="cb17-361"><a href="#cb17-361" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb17-362"><a href="#cb17-362" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors_category =</span> <span class="fu">list</span>(</span>
<span id="cb17-363"><a href="#cb17-363" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"immune"</span></span>
<span id="cb17-364"><a href="#cb17-364" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt"</span> <span class="ot">=</span> <span class="st">"tcell"</span></span>
<span id="cb17-365"><a href="#cb17-365" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt4minor"</span> <span class="ot">=</span> <span class="st">"cd4t"</span></span>
<span id="cb17-366"><a href="#cb17-366" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"lt8minor"</span> <span class="ot">=</span> <span class="st">"cd8t"</span></span>
<span id="cb17-367"><a href="#cb17-367" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb17-368"><a href="#cb17-368" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-369"><a href="#cb17-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-370"><a href="#cb17-370" aria-hidden="true" tabindex="-1"></a>rnactobj <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(</span>
<span id="cb17-371"><a href="#cb17-371" aria-hidden="true" tabindex="-1"></a>  <span class="at">pipeline =</span> pipeline,</span>
<span id="cb17-372"><a href="#cb17-372" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts_matrix =</span> ht_counts,</span>
<span id="cb17-373"><a href="#cb17-373" aria-hidden="true" tabindex="-1"></a>  <span class="at">adjacency_matrix =</span> sim_graph[<span class="fu">rownames</span>(ht_counts), <span class="fu">rownames</span>(ht_counts)],</span>
<span id="cb17-374"><a href="#cb17-374" aria-hidden="true" tabindex="-1"></a>  <span class="at">totalcounts =</span> total_counts_per_cell,</span>
<span id="cb17-375"><a href="#cb17-375" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_wise_frequency =</span> target_frequencies</span>
<span id="cb17-376"><a href="#cb17-376" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-377"><a href="#cb17-377" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(rnactobj, <span class="at">file=</span><span class="fu">file.path</span>(ct_dir, <span class="st">"rnactobj.rds"</span>))</span>
<span id="cb17-378"><a href="#cb17-378" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-379"><a href="#cb17-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-380"><a href="#cb17-380" aria-hidden="true" tabindex="-1"></a>Merge the Leiden cluster calls with the HieraType</span>
<span id="cb17-381"><a href="#cb17-381" aria-hidden="true" tabindex="-1"></a>results. There are several ways we could do this but for this tutorial we'll</span>
<span id="cb17-382"><a href="#cb17-382" aria-hidden="true" tabindex="-1"></a>use HieraType's built-in <span class="in">`celltype_label_integration`</span> function using the default</span>
<span id="cb17-383"><a href="#cb17-383" aria-hidden="true" tabindex="-1"></a>parameters. <span class="in">`celltype_label_integration`</span> integrates the immune celltype annotations</span>
<span id="cb17-384"><a href="#cb17-384" aria-hidden="true" tabindex="-1"></a>from HieraType with unsupervised cluster labels. The goal of this step is to enable</span>
<span id="cb17-385"><a href="#cb17-385" aria-hidden="true" tabindex="-1"></a>granular detection of novel or tissue-specific cell types (unsupervised), along </span>
<span id="cb17-386"><a href="#cb17-386" aria-hidden="true" tabindex="-1"></a>with well known immune cell types (HieraType). The logic for merging these </span>
<span id="cb17-387"><a href="#cb17-387" aria-hidden="true" tabindex="-1"></a>annotations works like this:</span>
<span id="cb17-388"><a href="#cb17-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-389"><a href="#cb17-389" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>All cells which are annotated with an immune-celltype HieraType label keep </span>
<span id="cb17-390"><a href="#cb17-390" aria-hidden="true" tabindex="-1"></a>their label.</span>
<span id="cb17-391"><a href="#cb17-391" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>All cells with a non-immune HieraType label take the unsupervised cluster label.</span>
<span id="cb17-392"><a href="#cb17-392" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Unsupervised clusters that had a high rate of cells taking immune labels </span>
<span id="cb17-393"><a href="#cb17-393" aria-hidden="true" tabindex="-1"></a>(say 90%), and make up only a small proportion of all cells (say 1%) after step </span>
<span id="cb17-394"><a href="#cb17-394" aria-hidden="true" tabindex="-1"></a>1 are 'dissolved'. These dissoved cells are assigned the most common cluster label</span>
<span id="cb17-395"><a href="#cb17-395" aria-hidden="true" tabindex="-1"></a>(could be immune or unsupervised) amongst most the similar neighboring cells. </span>
<span id="cb17-396"><a href="#cb17-396" aria-hidden="true" tabindex="-1"></a>Here we used the neighbor cells from the adjacency matrix that was used to run HieraType.</span>
<span id="cb17-397"><a href="#cb17-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-400"><a href="#cb17-400" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-401"><a href="#cb17-401" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: H-4</span></span>
<span id="cb17-402"><a href="#cb17-402" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-403"><a href="#cb17-403" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb17-404"><a href="#cb17-404" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-405"><a href="#cb17-405" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-406"><a href="#cb17-406" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-407"><a href="#cb17-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-408"><a href="#cb17-408" aria-hidden="true" tabindex="-1"></a>cell_type_dt <span class="ot">&lt;-</span> rnactobj<span class="sc">$</span>post_probs<span class="sc">$</span>l1</span>
<span id="cb17-409"><a href="#cb17-409" aria-hidden="true" tabindex="-1"></a>cell_type_dt <span class="ot">&lt;-</span> <span class="fu">rename</span>(cell_type_dt, <span class="at">cell_id =</span> cell_ID,  <span class="co"># &lt;1&gt;</span></span>
<span id="cb17-410"><a href="#cb17-410" aria-hidden="true" tabindex="-1"></a>                       <span class="at">Hieratype_celltype_fine =</span> celltype_granular)</span>
<span id="cb17-411"><a href="#cb17-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-412"><a href="#cb17-412" aria-hidden="true" tabindex="-1"></a>to_merge <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs <span class="sc">%&gt;%</span> <span class="fu">select</span>(leiden)</span>
<span id="cb17-413"><a href="#cb17-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-414"><a href="#cb17-414" aria-hidden="true" tabindex="-1"></a>cell_type_dt<span class="sc">$</span>Leiden <span class="ot">&lt;-</span> to_merge<span class="sc">$</span>leiden[<span class="fu">match</span>(<span class="fu">rownames</span>(to_merge), cell_type_dt<span class="sc">$</span>cell_id)]</span>
<span id="cb17-415"><a href="#cb17-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-416"><a href="#cb17-416" aria-hidden="true" tabindex="-1"></a>cell_type_dt <span class="ot">&lt;-</span> <span class="fu">celltype_label_integration</span>(</span>
<span id="cb17-417"><a href="#cb17-417" aria-hidden="true" tabindex="-1"></a>  <span class="at">metadata =</span> cell_type_dt, </span>
<span id="cb17-418"><a href="#cb17-418" aria-hidden="true" tabindex="-1"></a>  <span class="at">adjacency_mat =</span> sim_graph[<span class="fu">rownames</span>(ht_counts), <span class="fu">rownames</span>(ht_counts)],</span>
<span id="cb17-419"><a href="#cb17-419" aria-hidden="true" tabindex="-1"></a>  <span class="at">cellid_colname =</span> <span class="st">'cell_id'</span>,</span>
<span id="cb17-420"><a href="#cb17-420" aria-hidden="true" tabindex="-1"></a>  <span class="at">supervised_colname =</span> <span class="st">'Hieratype_celltype_fine'</span>,</span>
<span id="cb17-421"><a href="#cb17-421" aria-hidden="true" tabindex="-1"></a>  <span class="at">unsupervised_colname =</span> <span class="st">'Leiden'</span>)</span>
<span id="cb17-422"><a href="#cb17-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-423"><a href="#cb17-423" aria-hidden="true" tabindex="-1"></a>cell_type_dt <span class="ot">&lt;-</span> <span class="fu">rename</span>(cell_type_dt, <span class="at">merged =</span> celltype)</span>
<span id="cb17-424"><a href="#cb17-424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-425"><a href="#cb17-425" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(cell_type_dt, <span class="at">file=</span><span class="fu">file.path</span>(ct_dir, <span class="st">"cell_type_dt.rds"</span>))</span>
<span id="cb17-426"><a href="#cb17-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-427"><a href="#cb17-427" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-428"><a href="#cb17-428" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>cell_id here is the row name from the anndata object (as opposed to the c_i_j_k cell_ID notation that is sometimes used).</span>
<span id="cb17-429"><a href="#cb17-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-430"><a href="#cb17-430" aria-hidden="true" tabindex="-1"></a>Save the new calumns back to the anndata object.</span>
<span id="cb17-431"><a href="#cb17-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-434"><a href="#cb17-434" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb17-435"><a href="#cb17-435" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: H-5</span></span>
<span id="cb17-436"><a href="#cb17-436" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-437"><a href="#cb17-437" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb17-438"><a href="#cb17-438" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-439"><a href="#cb17-439" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-440"><a href="#cb17-440" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-441"><a href="#cb17-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-442"><a href="#cb17-442" aria-hidden="true" tabindex="-1"></a>cell_type_df <span class="op">=</span> r.cell_type_dt</span>
<span id="cb17-443"><a href="#cb17-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-444"><a href="#cb17-444" aria-hidden="true" tabindex="-1"></a>adata.obs <span class="op">=</span> adata.obs.join(cell_type_df[[<span class="st">'cell_id'</span>, <span class="st">'Hieratype_celltype_fine'</span>, <span class="st">'merged'</span>]].set_index(<span class="st">'cell_id'</span>))</span>
<span id="cb17-445"><a href="#cb17-445" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-446"><a href="#cb17-446" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-447"><a href="#cb17-447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-448"><a href="#cb17-448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-449"><a href="#cb17-449" aria-hidden="true" tabindex="-1"></a><span class="fu">### Step 3: Classify remaining cells using marker genes</span></span>
<span id="cb17-450"><a href="#cb17-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-451"><a href="#cb17-451" aria-hidden="true" tabindex="-1"></a>For the cells still assigned a Leiden cluster label, run a marker gene analysis</span>
<span id="cb17-452"><a href="#cb17-452" aria-hidden="true" tabindex="-1"></a>to determine the dominant cell type per cluster. We'll subset</span>
<span id="cb17-453"><a href="#cb17-453" aria-hidden="true" tabindex="-1"></a>larger clusters down to 10,000 cells to speed up the computation. Since scanpy's</span>
<span id="cb17-454"><a href="#cb17-454" aria-hidden="true" tabindex="-1"></a><span class="in">`tl.rank_genes_groups`</span> function needs normalizaed data, we'll use the log1p-transformed</span>
<span id="cb17-455"><a href="#cb17-455" aria-hidden="true" tabindex="-1"></a>matrix that we generated in @sec-preprocessing.</span>
<span id="cb17-456"><a href="#cb17-456" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-459"><a href="#cb17-459" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb17-460"><a href="#cb17-460" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: L-4</span></span>
<span id="cb17-461"><a href="#cb17-461" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb17-462"><a href="#cb17-462" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb17-463"><a href="#cb17-463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-464"><a href="#cb17-464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-465"><a href="#cb17-465" aria-hidden="true" tabindex="-1"></a>bdata <span class="op">=</span> adata[adata.obs[<span class="st">'leiden'</span>]<span class="op">==</span>adata.obs[<span class="st">'merged'</span>], :].copy()</span>
<span id="cb17-466"><a href="#cb17-466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-467"><a href="#cb17-467" aria-hidden="true" tabindex="-1"></a>random_seed <span class="op">=</span> <span class="dv">98103</span></span>
<span id="cb17-468"><a href="#cb17-468" aria-hidden="true" tabindex="-1"></a>np.random.seed(random_seed)</span>
<span id="cb17-469"><a href="#cb17-469" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-470"><a href="#cb17-470" aria-hidden="true" tabindex="-1"></a>target_cells_per_cluster <span class="op">=</span> <span class="dv">10000</span></span>
<span id="cb17-471"><a href="#cb17-471" aria-hidden="true" tabindex="-1"></a>cells_to_keep_indices <span class="op">=</span> []</span>
<span id="cb17-472"><a href="#cb17-472" aria-hidden="true" tabindex="-1"></a>cluster_counts <span class="op">=</span> bdata.obs[<span class="st">'merged'</span>].value_counts()</span>
<span id="cb17-473"><a href="#cb17-473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-474"><a href="#cb17-474" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> cluster_label <span class="kw">in</span> cluster_counts.index:</span>
<span id="cb17-475"><a href="#cb17-475" aria-hidden="true" tabindex="-1"></a>    cluster_cells <span class="op">=</span> bdata.obs_names[bdata.obs[<span class="st">'merged'</span>] <span class="op">==</span> cluster_label]</span>
<span id="cb17-476"><a href="#cb17-476" aria-hidden="true" tabindex="-1"></a>    num_cells_in_cluster <span class="op">=</span> <span class="bu">len</span>(cluster_cells)</span>
<span id="cb17-477"><a href="#cb17-477" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-478"><a href="#cb17-478" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> num_cells_in_cluster <span class="op">&gt;</span> target_cells_per_cluster:</span>
<span id="cb17-479"><a href="#cb17-479" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Cluster '</span><span class="sc">{</span>cluster_label<span class="sc">}</span><span class="ss">' has </span><span class="sc">{</span>num_cells_in_cluster<span class="sc">}</span><span class="ss"> cells, subsampling to </span><span class="sc">{</span>target_cells_per_cluster<span class="sc">}</span><span class="ss">."</span>)</span>
<span id="cb17-480"><a href="#cb17-480" aria-hidden="true" tabindex="-1"></a>        subsampled_cells <span class="op">=</span> np.random.choice(</span>
<span id="cb17-481"><a href="#cb17-481" aria-hidden="true" tabindex="-1"></a>            cluster_cells,</span>
<span id="cb17-482"><a href="#cb17-482" aria-hidden="true" tabindex="-1"></a>            size<span class="op">=</span>target_cells_per_cluster,</span>
<span id="cb17-483"><a href="#cb17-483" aria-hidden="true" tabindex="-1"></a>            replace<span class="op">=</span><span class="va">False</span></span>
<span id="cb17-484"><a href="#cb17-484" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb17-485"><a href="#cb17-485" aria-hidden="true" tabindex="-1"></a>        cells_to_keep_indices.extend(subsampled_cells)</span>
<span id="cb17-486"><a href="#cb17-486" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb17-487"><a href="#cb17-487" aria-hidden="true" tabindex="-1"></a>      cells_to_keep_indices.extend(cluster_cells)</span>
<span id="cb17-488"><a href="#cb17-488" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-489"><a href="#cb17-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-490"><a href="#cb17-490" aria-hidden="true" tabindex="-1"></a>bdata_subsampled <span class="op">=</span> bdata[cells_to_keep_indices, :].copy()</span>
<span id="cb17-491"><a href="#cb17-491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-492"><a href="#cb17-492" aria-hidden="true" tabindex="-1"></a>bdata_subsampled.X <span class="op">=</span> bdata_subsampled.layers[<span class="st">'TC'</span>].copy() <span class="co"># &lt;1&gt;</span></span>
<span id="cb17-493"><a href="#cb17-493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-494"><a href="#cb17-494" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-495"><a href="#cb17-495" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>If you do not have the log1p-transformed total counts layer 'TC' saved, run <span class="in">`sc.pp.normalize_total`</span></span>
<span id="cb17-496"><a href="#cb17-496" aria-hidden="true" tabindex="-1"></a>and <span class="in">`sc.pp.log1p`</span> (see @sec-tc-workflow).</span>
<span id="cb17-497"><a href="#cb17-497" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-498"><a href="#cb17-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-499"><a href="#cb17-499" aria-hidden="true" tabindex="-1"></a>Run <span class="in">`rank_genes_groups`</span> and save results.</span>
<span id="cb17-500"><a href="#cb17-500" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-503"><a href="#cb17-503" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb17-504"><a href="#cb17-504" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: run-marker-gene-leiden-2</span></span>
<span id="cb17-505"><a href="#cb17-505" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb17-506"><a href="#cb17-506" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb17-507"><a href="#cb17-507" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-508"><a href="#cb17-508" aria-hidden="true" tabindex="-1"></a>ct_dir <span class="op">=</span> r.ct_dir</span>
<span id="cb17-509"><a href="#cb17-509" aria-hidden="true" tabindex="-1"></a>sc.settings.figdir <span class="op">=</span> ct_dir</span>
<span id="cb17-510"><a href="#cb17-510" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-511"><a href="#cb17-511" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(bdata_subsampled, <span class="st">'leiden'</span>, method<span class="op">=</span><span class="st">'wilcoxon'</span>,</span>
<span id="cb17-512"><a href="#cb17-512" aria-hidden="true" tabindex="-1"></a>  solver<span class="op">=</span><span class="st">'saga'</span>, max_iter<span class="op">=</span><span class="dv">200</span>, pts<span class="op">=</span><span class="va">True</span>) <span class="co"># ~an hour using logreg</span></span>
<span id="cb17-513"><a href="#cb17-513" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups(bdata_subsampled, n_genes<span class="op">=</span><span class="dv">8</span>, sharey<span class="op">=</span><span class="va">True</span>, fontsize<span class="op">=</span><span class="dv">10</span>, save<span class="op">=</span><span class="st">"_marker_gene_ranks_global.png"</span>)</span>
<span id="cb17-514"><a href="#cb17-514" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups_dotplot(bdata_subsampled, n_genes<span class="op">=</span><span class="dv">8</span>, save<span class="op">=</span><span class="st">"_marker_gene_dotplot_global.png"</span>)</span>
<span id="cb17-515"><a href="#cb17-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-516"><a href="#cb17-516" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-5-subset-marker-genes.h5ad"</span>)</span>
<span id="cb17-517"><a href="#cb17-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-518"><a href="#cb17-518" aria-hidden="true" tabindex="-1"></a>bdata_subsampled.write_h5ad(</span>
<span id="cb17-519"><a href="#cb17-519" aria-hidden="true" tabindex="-1"></a>  filename,</span>
<span id="cb17-520"><a href="#cb17-520" aria-hidden="true" tabindex="-1"></a>  compression<span class="op">=</span>hdf5plugin.FILTERS[<span class="st">"zstd"</span>],</span>
<span id="cb17-521"><a href="#cb17-521" aria-hidden="true" tabindex="-1"></a>  compression_opts<span class="op">=</span>hdf5plugin.Zstd(clevel<span class="op">=</span><span class="dv">5</span>).filter_options</span>
<span id="cb17-522"><a href="#cb17-522" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-523"><a href="#cb17-523" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-524"><a href="#cb17-524" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-525"><a href="#cb17-525" aria-hidden="true" tabindex="-1"></a>The next task is determine from the marker gene results the dominant cell types</span>
<span id="cb17-526"><a href="#cb17-526" aria-hidden="true" tabindex="-1"></a>that are likely found within each Leiden cluster. While some Leiden clusters </span>
<span id="cb17-527"><a href="#cb17-527" aria-hidden="true" tabindex="-1"></a>will have a high proportion of cells that are of the same type or state, keep</span>
<span id="cb17-528"><a href="#cb17-528" aria-hidden="true" tabindex="-1"></a>in mind that other Leiden clusters might be composites of cell types. The tools</span>
<span id="cb17-529"><a href="#cb17-529" aria-hidden="true" tabindex="-1"></a>we have for this exercise are: </span>
<span id="cb17-530"><a href="#cb17-530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-531"><a href="#cb17-531" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>visual aids from the marker gene results and </span>
<span id="cb17-532"><a href="#cb17-532" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>compare with external sources. The literature and subject matter expertise is the gold-standard for this.</span>
<span id="cb17-533"><a href="#cb17-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-536"><a href="#cb17-536" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-537"><a href="#cb17-537" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-global-marker-ranks</span></span>
<span id="cb17-538"><a href="#cb17-538" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-539"><a href="#cb17-539" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-540"><a href="#cb17-540" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-541"><a href="#cb17-541" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 8</span></span>
<span id="cb17-542"><a href="#cb17-542" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 16</span></span>
<span id="cb17-543"><a href="#cb17-543" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Marker gene ranks by Leiden cluster (global)."</span></span>
<span id="cb17-544"><a href="#cb17-544" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-545"><a href="#cb17-545" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-546"><a href="#cb17-546" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>), <span class="st">"rank_genes_groups_leiden_marker_gene_ranks_global.png"</span>, ct_dir, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb17-547"><a href="#cb17-547" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-548"><a href="#cb17-548" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-551"><a href="#cb17-551" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-552"><a href="#cb17-552" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-global-marker-dot</span></span>
<span id="cb17-553"><a href="#cb17-553" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-554"><a href="#cb17-554" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-555"><a href="#cb17-555" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-556"><a href="#cb17-556" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 12</span></span>
<span id="cb17-557"><a href="#cb17-557" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 5</span></span>
<span id="cb17-558"><a href="#cb17-558" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Marker gene dot plot by Leiden cluster (global)."</span></span>
<span id="cb17-559"><a href="#cb17-559" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-560"><a href="#cb17-560" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-561"><a href="#cb17-561" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>), <span class="st">"dotplot__marker_gene_dotplot_global.png"</span>, ct_dir, <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-562"><a href="#cb17-562" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-563"><a href="#cb17-563" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-564"><a href="#cb17-564" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-565"><a href="#cb17-565" aria-hidden="true" tabindex="-1"></a>Examine the clusters by extracting the top 25 markers and their relative</span>
<span id="cb17-566"><a href="#cb17-566" aria-hidden="true" tabindex="-1"></a>scores and comparing to the literature. Large Language Models can provide supportive</span>
<span id="cb17-567"><a href="#cb17-567" aria-hidden="true" tabindex="-1"></a>information here but it's always recommended to use LLMs with caution and to verify</span>
<span id="cb17-568"><a href="#cb17-568" aria-hidden="true" tabindex="-1"></a>any of the results. In other words, LLMs can be fast but at the cost of accuracy.</span>
<span id="cb17-569"><a href="#cb17-569" aria-hidden="true" tabindex="-1"></a>In the example below, I created an LLM prompt that includes</span>
<span id="cb17-570"><a href="#cb17-570" aria-hidden="true" tabindex="-1"></a>the top marker gene statistics. Let's see what it generates.</span>
<span id="cb17-571"><a href="#cb17-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-574"><a href="#cb17-574" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb17-575"><a href="#cb17-575" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: eval-marker-gene-leiden</span></span>
<span id="cb17-576"><a href="#cb17-576" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb17-577"><a href="#cb17-577" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb17-578"><a href="#cb17-578" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-579"><a href="#cb17-579" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-580"><a href="#cb17-580" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-581"><a href="#cb17-581" aria-hidden="true" tabindex="-1"></a>marker_dict, marker_df <span class="op">=</span> format_ranked_genes(bdata_subsampled, <span class="dv">25</span>)</span>
<span id="cb17-582"><a href="#cb17-582" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-583"><a href="#cb17-583" aria-hidden="true" tabindex="-1"></a>csv_file_path <span class="op">=</span> os.path.join(ct_dir, <span class="st">"colon_cancer_top_marker_genes_global.csv"</span>)</span>
<span id="cb17-584"><a href="#cb17-584" aria-hidden="true" tabindex="-1"></a>marker_df.to_csv(csv_file_path, index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb17-585"><a href="#cb17-585" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-586"><a href="#cb17-586" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"I am analyzing a CosMx SMI dataset of from Colon cancer and have run Leiden clustering.</span><span class="ch">\n</span><span class="ss">I have already classified several of the immune cells within the larger dataset and now I would like your help in predicting the most like cell type given the marker gene profile from this subset of data.</span><span class="ch">\n</span><span class="ss">Below is a dictionary of results where the key represents the Leiden cluster</span><span class="ch">\n</span><span class="ss">name and its corresponding item is a list of ordered tuples where each tuple</span><span class="ch">\n</span><span class="ss">provides the marker name and the relative score for that marker.</span><span class="ch">\n\n</span><span class="ss"> Please provide a table with the columns that represent 1. inferred cell type 2. your confidence in that cell type 3. up to five markers from the list that are key to your result and 4. a brief justification of your result. Please present the results in a markdown table format to make it easier to copy and paste.</span><span class="ch">\n\n</span><span class="sc">{</span>marker_dict<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb17-587"><a href="#cb17-587" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-588"><a href="#cb17-588" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-589"><a href="#cb17-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-590"><a href="#cb17-590" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Cluster <span class="pp">|</span> Inferred Cell Type <span class="pp">|</span> Confidence <span class="pp">|</span> Top 5 Key Markers <span class="pp">|</span> Brief Justification <span class="pp">|</span></span>
<span id="cb17-591"><a href="#cb17-591" aria-hidden="true" tabindex="-1"></a><span class="pp">| :---</span> <span class="pp">| :---</span> <span class="pp">| :---</span> <span class="pp">| :---</span> <span class="pp">| :---</span> <span class="pp">|</span></span>
<span id="cb17-592"><a href="#cb17-592" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-0** | **Epithelial (Goblet/Secretory)** <span class="pp">|</span> High <span class="pp">|</span> *PIGR, MUC2, EPCAM, KRT8, CLDN7* | Expresses broad epithelial markers (*EPCAM, KRT8*) alongside specific secretory/mucosal markers (*PIGR, MUC2*) typical of normal or well-differentiated colon epithelium. <span class="pp">|</span></span>
<span id="cb17-593"><a href="#cb17-593" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-3** | **Plasma Cells (IgG)** <span class="pp">|</span> High <span class="pp">|</span> *IGHG1, IGKC, MZB1, JCHAIN, XBP1* | Defined by very high expression of Immunoglobulin G heavy chains (*IGHG1*) and plasma cell maturation factors (*MZB1, JCHAIN*). <span class="pp">|</span></span>
<span id="cb17-594"><a href="#cb17-594" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-4** | **CAF** <span class="pp">|</span> High <span class="pp">|</span> *MMP14, MMP2, FN1, COL1A1, SPARC* <span class="pp">|</span> Spatially coincident with tumor. The high levels of matrix metalloproteinases (MMP14, MMP2) and Fibronectin (FN1) identify these as the activated myofibroblasts driving tumor desmoplasia. <span class="pp">|</span></span>
<span id="cb17-595"><a href="#cb17-595" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-5** | **Unclassified** <span class="pp">|</span> Low <span class="pp">|</span> *MPV17L, ZNF91, DUX4, CXCL8, USP17L* | Lacks clear lineage markers (No *EPCAM, CD45,* or *COL1A1*). <span class="pp">|</span></span>
<span id="cb17-596"><a href="#cb17-596" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-6** | **Immune (Mixed)** <span class="pp">|</span> Moderate <span class="pp">|</span> *KIR3DL1, CD79A, CSF2, CD53, RAG1* | Contains conflicting a mix of immune markers: NK cell (*KIR3DL1*) and B cell (*CD79A*) markers appear together. Likely represents lymphoid aggregates. <span class="pp">|</span></span>
<span id="cb17-597"><a href="#cb17-597" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-7** | **Endothelial Cells** <span class="pp">|</span> High <span class="pp">|</span> *PECAM1, VWF, PLVAP, EGFL7, SHANK3* | Distinct expression of vascular endothelium markers, including *PECAM1* (CD31) and *VWF*. *PLVAP* suggests fenestrated capillaries common in the gut. <span class="pp">|</span></span>
<span id="cb17-598"><a href="#cb17-598" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-8** | **Plasma Cells (IgA)** <span class="pp">|</span> High <span class="pp">|</span> *IGHA1, JCHAIN, IGKC, GRP, MZB1* | Similar to C-3 but distinguished by the specific expression of IgA (*IGHA1*), the dominant isotype in mucosal immunity of the colon. <span class="pp">|</span></span>
<span id="cb17-599"><a href="#cb17-599" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-9** | **Tumor Epithelial (Invasive)** <span class="pp">|</span> High <span class="pp">|</span> *CEACAM5, MMP7, LCN2, KRT8, CLDN4* | Malignant epithelial profile (*EPCAM, KRT8*) with high expression of tumor markers (*CEACAM5/CEA*) and invasion/stress markers (*MMP7, LCN2*). <span class="pp">|</span></span>
<span id="cb17-600"><a href="#cb17-600" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-10** <span class="pp">|</span> **Structural Fibroblasts ** | High | *SFRP2, MGP, FBLN1, C3, CXCL12* <span class="pp">|</span> Spatially coincident with muscle (SMC). The profile (SFRP2+, MGP+, FBLN1+) matches "Adventitial" or "Universal" fibroblasts that act as the structural scaffold for muscle layers and vessels. <span class="pp">|</span></span>
<span id="cb17-601"><a href="#cb17-601" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-11** | **Tumor Epithelial (High Cycling)** <span class="pp">|</span> High <span class="pp">|</span> *RPS19, CEACAM5, EPCAM, RPS18, PIGR* | Epithelial tumor cells (*CEACAM5*) dominated by ribosomal proteins (*RPS/RPL*), indicating high protein synthesis and rapid cell cycling. <span class="pp">|</span></span>
<span id="cb17-602"><a href="#cb17-602" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-12** | **Smooth Muscle Cells** <span class="pp">|</span> High <span class="pp">|</span> *MYH11, ACTG2, DES, CNN1, MYLK* | Expression of contractile proteins (*MYH11, ACTG2*) and Desmin (*DES*) identifies these as mural smooth muscle cells (muscularis mucosae/propria). <span class="pp">|</span></span>
<span id="cb17-603"><a href="#cb17-603" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-13** | **Enteric Glia / Schwann Cells** <span class="pp">|</span> High <span class="pp">|</span> *CDH19, PMP22, SCN7A, LGI4, CLU* | Presence of peripheral nervous system markers (*PMP22, CDH19, SCN7A*) identifies these as supportive glial cells of the Enteric Nervous System. <span class="pp">|</span></span>
<span id="cb17-604"><a href="#cb17-604" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **C-14** | **Pericytes** <span class="pp">|</span> High <span class="pp">|</span> *NOTCH3, PDGFRB, MCAM, COL4A1, RGS5* | Expresses pericyte markers (*PDGFRB, MCAM, NOTCH3*) and basement membrane collagen (*COL4A1*), distinguishing them from the broader fibroblast clusters. <span class="pp">|</span></span>
<span id="cb17-605"><a href="#cb17-605" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-606"><a href="#cb17-606" aria-hidden="true" tabindex="-1"></a>: Estimating most likely cell type from Leiden clusters' marker gene rankings. Note: predictions were generated with the aid of an LLM. {#tbl-leiden-global}</span>
<span id="cb17-607"><a href="#cb17-607" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-608"><a href="#cb17-608" aria-hidden="true" tabindex="-1"></a>Following a little back and forth and helping provide the LLM with additional</span>
<span id="cb17-609"><a href="#cb17-609" aria-hidden="true" tabindex="-1"></a>spatial context, we have the marker gene results in @tbl-leiden-global, we have </span>
<span id="cb17-610"><a href="#cb17-610" aria-hidden="true" tabindex="-1"></a>most of the structural cell types</span>
<span id="cb17-611"><a href="#cb17-611" aria-hidden="true" tabindex="-1"></a>identified from the remaining clusters. Cluster C-5 lacks a clear and </span>
<span id="cb17-612"><a href="#cb17-612" aria-hidden="true" tabindex="-1"></a>dominant lineage signature. This cluster also occupies roughly the center of the UMAP</span>
<span id="cb17-613"><a href="#cb17-613" aria-hidden="true" tabindex="-1"></a>where poorer quality cells sometimes occupy. For the purposes of this tutorial </span>
<span id="cb17-614"><a href="#cb17-614" aria-hidden="true" tabindex="-1"></a>and the downstream questions, we'll leave those cells unclassified and label them</span>
<span id="cb17-615"><a href="#cb17-615" aria-hidden="true" tabindex="-1"></a>as "undetermined". Similarly, cluster C-6 showed two distinct lienage markers for</span>
<span id="cb17-616"><a href="#cb17-616" aria-hidden="true" tabindex="-1"></a>NK cells (_KIR3DL1_) and B cells (_CD79A_). Since these markers are unlikely to </span>
<span id="cb17-617"><a href="#cb17-617" aria-hidden="true" tabindex="-1"></a>be present in the same cell it is likely that C-6 comprises a composite of NK and</span>
<span id="cb17-618"><a href="#cb17-618" aria-hidden="true" tabindex="-1"></a>lymphoid cell types. Interestingly, HieraType did not classify the cells within</span>
<span id="cb17-619"><a href="#cb17-619" aria-hidden="true" tabindex="-1"></a>as either NK or B cell. Since our primary interest are on tumor and CAFs, we'll</span>
<span id="cb17-620"><a href="#cb17-620" aria-hidden="true" tabindex="-1"></a>not unravel this cluster any further and instead label it as a mix of cell types. </span>
<span id="cb17-621"><a href="#cb17-621" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-622"><a href="#cb17-622" aria-hidden="true" tabindex="-1"></a>From the HieraType-derived merged data, map these new cell type labels.</span>
<span id="cb17-623"><a href="#cb17-623" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-626"><a href="#cb17-626" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb17-627"><a href="#cb17-627" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: eval-marker-gene-leiden1</span></span>
<span id="cb17-628"><a href="#cb17-628" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb17-629"><a href="#cb17-629" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb17-630"><a href="#cb17-630" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-631"><a href="#cb17-631" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-632"><a href="#cb17-632" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-633"><a href="#cb17-633" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-634"><a href="#cb17-634" aria-hidden="true" tabindex="-1"></a>remap <span class="op">=</span> {</span>
<span id="cb17-635"><a href="#cb17-635" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-0'</span>: <span class="st">'epithelial'</span>,</span>
<span id="cb17-636"><a href="#cb17-636" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-3'</span>: <span class="st">'plasma_IgG'</span>,</span>
<span id="cb17-637"><a href="#cb17-637" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-4'</span>: <span class="st">'CAF'</span>,</span>
<span id="cb17-638"><a href="#cb17-638" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-5'</span>: <span class="st">'undetermined'</span>,</span>
<span id="cb17-639"><a href="#cb17-639" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-6'</span>: <span class="st">'mix_NK_Lymphoid'</span>,</span>
<span id="cb17-640"><a href="#cb17-640" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-7'</span>: <span class="st">'endothelial'</span>,</span>
<span id="cb17-641"><a href="#cb17-641" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-8'</span>: <span class="st">'plasma_IgA'</span>,</span>
<span id="cb17-642"><a href="#cb17-642" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-9'</span>: <span class="st">'tumor'</span>,</span>
<span id="cb17-643"><a href="#cb17-643" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-10'</span>: <span class="st">'fibroblast'</span>,</span>
<span id="cb17-644"><a href="#cb17-644" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-11'</span>: <span class="st">'tumor_cyling'</span>,</span>
<span id="cb17-645"><a href="#cb17-645" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-12'</span>: <span class="st">'smc'</span>,</span>
<span id="cb17-646"><a href="#cb17-646" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-13'</span>: <span class="st">'glial'</span>,</span>
<span id="cb17-647"><a href="#cb17-647" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C-14'</span>: <span class="st">'pericyte'</span></span>
<span id="cb17-648"><a href="#cb17-648" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-649"><a href="#cb17-649" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-650"><a href="#cb17-650" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-651"><a href="#cb17-651" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'celltype_fine'</span>] <span class="op">=</span> adata.obs[<span class="st">'merged'</span>].astype(<span class="bu">str</span>).replace(remap)</span>
<span id="cb17-652"><a href="#cb17-652" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-653"><a href="#cb17-653" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-654"><a href="#cb17-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-655"><a href="#cb17-655" aria-hidden="true" tabindex="-1"></a>Often we want to plot or work with broader classifications of cell types. The code</span>
<span id="cb17-656"><a href="#cb17-656" aria-hidden="true" tabindex="-1"></a>below combines, for example, all the various T cell subtypes into a single group.</span>
<span id="cb17-657"><a href="#cb17-657" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-660"><a href="#cb17-660" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb17-661"><a href="#cb17-661" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: H-6</span></span>
<span id="cb17-662"><a href="#cb17-662" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-663"><a href="#cb17-663" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb17-664"><a href="#cb17-664" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-665"><a href="#cb17-665" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-666"><a href="#cb17-666" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-667"><a href="#cb17-667" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-668"><a href="#cb17-668" aria-hidden="true" tabindex="-1"></a>broad_map <span class="op">=</span> {</span>
<span id="cb17-669"><a href="#cb17-669" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd8_tem'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-670"><a href="#cb17-670" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_treg'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-671"><a href="#cb17-671" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_tem'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-672"><a href="#cb17-672" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd8_cytotoxic'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-673"><a href="#cb17-673" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd8_naive'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-674"><a href="#cb17-674" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd8_exhausted'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-675"><a href="#cb17-675" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_tcm'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-676"><a href="#cb17-676" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_th2'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-677"><a href="#cb17-677" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_th17'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-678"><a href="#cb17-678" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_naive'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-679"><a href="#cb17-679" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd4_th1'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-680"><a href="#cb17-680" aria-hidden="true" tabindex="-1"></a>    <span class="st">'cd8_tcm'</span>: <span class="st">'tcell'</span>,</span>
<span id="cb17-681"><a href="#cb17-681" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-682"><a href="#cb17-682" aria-hidden="true" tabindex="-1"></a>    <span class="st">'bcell'</span>: <span class="st">'bcell'</span>,</span>
<span id="cb17-683"><a href="#cb17-683" aria-hidden="true" tabindex="-1"></a>    <span class="st">'plasma_IgA'</span>: <span class="st">'plasma'</span>,</span>
<span id="cb17-684"><a href="#cb17-684" aria-hidden="true" tabindex="-1"></a>    <span class="st">'plasma_IgG'</span>: <span class="st">'plasma'</span>,</span>
<span id="cb17-685"><a href="#cb17-685" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-686"><a href="#cb17-686" aria-hidden="true" tabindex="-1"></a>    <span class="st">'tumor'</span>: <span class="st">'tumor'</span>,</span>
<span id="cb17-687"><a href="#cb17-687" aria-hidden="true" tabindex="-1"></a>    <span class="st">'tumor_cyling'</span>: <span class="st">'tumor'</span>,</span>
<span id="cb17-688"><a href="#cb17-688" aria-hidden="true" tabindex="-1"></a>    <span class="st">'epithelial'</span>: <span class="st">'epithelial'</span>,</span>
<span id="cb17-689"><a href="#cb17-689" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-690"><a href="#cb17-690" aria-hidden="true" tabindex="-1"></a>    <span class="st">'CAF'</span>: <span class="st">'CAF'</span>,</span>
<span id="cb17-691"><a href="#cb17-691" aria-hidden="true" tabindex="-1"></a>    <span class="st">'fibroblast'</span>: <span class="st">'fibroblast'</span>,</span>
<span id="cb17-692"><a href="#cb17-692" aria-hidden="true" tabindex="-1"></a>    <span class="st">'smc'</span>: <span class="st">'smc'</span>,</span>
<span id="cb17-693"><a href="#cb17-693" aria-hidden="true" tabindex="-1"></a>    <span class="st">'pericyte'</span>: <span class="st">'pericyte'</span>,</span>
<span id="cb17-694"><a href="#cb17-694" aria-hidden="true" tabindex="-1"></a>    <span class="st">'endothelial'</span>: <span class="st">'endothelial'</span>,</span>
<span id="cb17-695"><a href="#cb17-695" aria-hidden="true" tabindex="-1"></a>    <span class="st">'glial'</span>: <span class="st">'glial'</span>,</span>
<span id="cb17-696"><a href="#cb17-696" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-697"><a href="#cb17-697" aria-hidden="true" tabindex="-1"></a>    <span class="st">'macrophage'</span>: <span class="st">'macrophage'</span>,</span>
<span id="cb17-698"><a href="#cb17-698" aria-hidden="true" tabindex="-1"></a>    <span class="st">'monocyte'</span>: <span class="st">'monocyte'</span>,</span>
<span id="cb17-699"><a href="#cb17-699" aria-hidden="true" tabindex="-1"></a>    <span class="st">'dendritic'</span>: <span class="st">'dendritic'</span>,</span>
<span id="cb17-700"><a href="#cb17-700" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mast'</span>: <span class="st">'mast'</span>,</span>
<span id="cb17-701"><a href="#cb17-701" aria-hidden="true" tabindex="-1"></a>    <span class="st">'neutrophil'</span>: <span class="st">'neutrophil'</span>,</span>
<span id="cb17-702"><a href="#cb17-702" aria-hidden="true" tabindex="-1"></a>    <span class="st">'nk'</span>: <span class="st">'nk'</span>,</span>
<span id="cb17-703"><a href="#cb17-703" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mix_NK_Lymphoid'</span>: <span class="st">'mix_NK_Lymphoid'</span>,</span>
<span id="cb17-704"><a href="#cb17-704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-705"><a href="#cb17-705" aria-hidden="true" tabindex="-1"></a>    <span class="st">'undetermined'</span>: <span class="st">'undetermined'</span></span>
<span id="cb17-706"><a href="#cb17-706" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-707"><a href="#cb17-707" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-708"><a href="#cb17-708" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'celltype_broad'</span>] <span class="op">=</span> adata.obs[<span class="st">'celltype_fine'</span>].<span class="bu">map</span>(broad_map)</span>
<span id="cb17-709"><a href="#cb17-709" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-710"><a href="#cb17-710" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-711"><a href="#cb17-711" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-712"><a href="#cb17-712" aria-hidden="true" tabindex="-1"></a>Which provides 18 broad cell types and 32 fine cell types.</span>
<span id="cb17-713"><a href="#cb17-713" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-714"><a href="#cb17-714" aria-hidden="true" tabindex="-1"></a>Save the results to disk.</span>
<span id="cb17-715"><a href="#cb17-715" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-718"><a href="#cb17-718" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb17-719"><a href="#cb17-719" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: H-7</span></span>
<span id="cb17-720"><a href="#cb17-720" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-721"><a href="#cb17-721" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb17-722"><a href="#cb17-722" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-723"><a href="#cb17-723" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-724"><a href="#cb17-724" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-725"><a href="#cb17-725" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-726"><a href="#cb17-726" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-6-celltypes.h5ad"</span>)</span>
<span id="cb17-727"><a href="#cb17-727" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(</span>
<span id="cb17-728"><a href="#cb17-728" aria-hidden="true" tabindex="-1"></a>  filename,</span>
<span id="cb17-729"><a href="#cb17-729" aria-hidden="true" tabindex="-1"></a>  compression<span class="op">=</span>hdf5plugin.FILTERS[<span class="st">"zstd"</span>],</span>
<span id="cb17-730"><a href="#cb17-730" aria-hidden="true" tabindex="-1"></a>  compression_opts<span class="op">=</span>hdf5plugin.Zstd(clevel<span class="op">=</span><span class="dv">5</span>).filter_options</span>
<span id="cb17-731"><a href="#cb17-731" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-732"><a href="#cb17-732" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-733"><a href="#cb17-733" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-734"><a href="#cb17-734" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-735"><a href="#cb17-735" aria-hidden="true" tabindex="-1"></a><span class="fu">## Cell Type Visualization</span></span>
<span id="cb17-736"><a href="#cb17-736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-737"><a href="#cb17-737" aria-hidden="true" tabindex="-1"></a>It is useful at this point to visualize the distributions of cells in a variety</span>
<span id="cb17-738"><a href="#cb17-738" aria-hidden="true" tabindex="-1"></a>of ways to get an understanding of their composition. We can visualize these</span>
<span id="cb17-739"><a href="#cb17-739" aria-hidden="true" tabindex="-1"></a>cell types in UMAP space and XY space.</span>
<span id="cb17-740"><a href="#cb17-740" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-741"><a href="#cb17-741" aria-hidden="true" tabindex="-1"></a><span class="fu">### Broad cell types</span></span>
<span id="cb17-742"><a href="#cb17-742" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-743"><a href="#cb17-743" aria-hidden="true" tabindex="-1"></a>Instead of the default colors in <span class="in">`plotDots`</span>, let's create manual ones.</span>
<span id="cb17-744"><a href="#cb17-744" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-745"><a href="#cb17-745" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-748"><a href="#cb17-748" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-749"><a href="#cb17-749" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: V-1</span></span>
<span id="cb17-750"><a href="#cb17-750" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-751"><a href="#cb17-751" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb17-752"><a href="#cb17-752" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-753"><a href="#cb17-753" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-754"><a href="#cb17-754" aria-hidden="true" tabindex="-1"></a><span class="co"># define colors semi-manually instead of using the default colors</span></span>
<span id="cb17-755"><a href="#cb17-755" aria-hidden="true" tabindex="-1"></a>column <span class="ot">&lt;-</span> <span class="st">'celltype_broad'</span></span>
<span id="cb17-756"><a href="#cb17-756" aria-hidden="true" tabindex="-1"></a>column_levels <span class="ot">&lt;-</span> <span class="fu">levels</span>(py<span class="sc">$</span>adata<span class="sc">$</span>obs[[column]])</span>
<span id="cb17-757"><a href="#cb17-757" aria-hidden="true" tabindex="-1"></a>n_levels <span class="ot">&lt;-</span> <span class="fu">length</span>(column_levels)</span>
<span id="cb17-758"><a href="#cb17-758" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-759"><a href="#cb17-759" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(n_levels<span class="sc">&lt;</span><span class="dv">27</span>){</span>
<span id="cb17-760"><a href="#cb17-760" aria-hidden="true" tabindex="-1"></a>  colors_use <span class="ot">&lt;-</span> pals<span class="sc">::</span><span class="fu">alphabet</span>(n_levels)</span>
<span id="cb17-761"><a href="#cb17-761" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> <span class="cf">if</span>(n <span class="sc">&lt;</span> <span class="dv">53</span>){</span>
<span id="cb17-762"><a href="#cb17-762" aria-hidden="true" tabindex="-1"></a>  colors_use <span class="ot">&lt;-</span> <span class="fu">c</span>(pals<span class="sc">::</span><span class="fu">alphabet</span>(<span class="dv">26</span>), pals<span class="sc">::</span><span class="fu">alphabet2</span>(n_levels<span class="dv">-26</span>))</span>
<span id="cb17-763"><a href="#cb17-763" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb17-764"><a href="#cb17-764" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"consider fewer groups"</span>)</span>
<span id="cb17-765"><a href="#cb17-765" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-766"><a href="#cb17-766" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-767"><a href="#cb17-767" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(colors_use) <span class="ot">&lt;-</span> column_levels</span>
<span id="cb17-768"><a href="#cb17-768" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-769"><a href="#cb17-769" aria-hidden="true" tabindex="-1"></a>colors_use[<span class="st">'mast'</span>] <span class="ot">=</span> <span class="st">"#FFCC99"</span></span>
<span id="cb17-770"><a href="#cb17-770" aria-hidden="true" tabindex="-1"></a>colors_use[<span class="st">'tumor'</span>] <span class="ot">=</span> <span class="st">"#C20088"</span></span>
<span id="cb17-771"><a href="#cb17-771" aria-hidden="true" tabindex="-1"></a>colors_use[<span class="st">'undetermined'</span>] <span class="ot">=</span> <span class="st">"#808080"</span></span>
<span id="cb17-772"><a href="#cb17-772" aria-hidden="true" tabindex="-1"></a>colors_use[<span class="st">'mix_NK_Lymphoid'</span>] <span class="ot">=</span> <span class="st">"#8080FF"</span></span>
<span id="cb17-773"><a href="#cb17-773" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-774"><a href="#cb17-774" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-775"><a href="#cb17-775" aria-hidden="true" tabindex="-1"></a><span class="co"># saving colors for later</span></span>
<span id="cb17-776"><a href="#cb17-776" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'celltype_broad_names'</span>]] <span class="ot">&lt;-</span> <span class="fu">names</span>(colors_use)</span>
<span id="cb17-777"><a href="#cb17-777" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'celltype_broad_colors'</span>]] <span class="ot">&lt;-</span> <span class="fu">as.character</span>(colors_use)</span>
<span id="cb17-778"><a href="#cb17-778" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb17-779"><a href="#cb17-779" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-780"><a href="#cb17-780" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-781"><a href="#cb17-781" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-782"><a href="#cb17-782" aria-hidden="true" tabindex="-1"></a>Plot the cells in XY space and UMAP space and color based on <span class="in">`celltype_broad`</span>. </span>
<span id="cb17-783"><a href="#cb17-783" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-786"><a href="#cb17-786" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-787"><a href="#cb17-787" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: V-2</span></span>
<span id="cb17-788"><a href="#cb17-788" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-789"><a href="#cb17-789" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb17-790"><a href="#cb17-790" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-791"><a href="#cb17-791" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-792"><a href="#cb17-792" aria-hidden="true" tabindex="-1"></a>group_prefix <span class="ot">&lt;-</span> <span class="st">"broad"</span> <span class="co"># &lt;1&gt;</span></span>
<span id="cb17-793"><a href="#cb17-793" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-794"><a href="#cb17-794" aria-hidden="true" tabindex="-1"></a>ct_assets_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>)</span>
<span id="cb17-795"><a href="#cb17-795" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-796"><a href="#cb17-796" aria-hidden="true" tabindex="-1"></a><span class="co"># XY (global only)</span></span>
<span id="cb17-797"><a href="#cb17-797" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, <span class="at">color_by=</span><span class="st">'celltype_broad'</span>,</span>
<span id="cb17-798"><a href="#cb17-798" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-799"><a href="#cb17-799" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-800"><a href="#cb17-800" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="cb17-801"><a href="#cb17-801" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-802"><a href="#cb17-802" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span></span>
<span id="cb17-803"><a href="#cb17-803" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-804"><a href="#cb17-804" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_bar_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-805"><a href="#cb17-805" aria-hidden="true" tabindex="-1"></a>                    <span class="at">location =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">0</span>),</span>
<span id="cb17-806"><a href="#cb17-806" aria-hidden="true" tabindex="-1"></a>                    <span class="at">width =</span> <span class="dv">2</span>,</span>
<span id="cb17-807"><a href="#cb17-807" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n =</span> <span class="dv">3</span>,</span>
<span id="cb17-808"><a href="#cb17-808" aria-hidden="true" tabindex="-1"></a>                    <span class="at">height =</span> <span class="fl">0.1</span>,</span>
<span id="cb17-809"><a href="#cb17-809" aria-hidden="true" tabindex="-1"></a>                    <span class="at">scale_colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"grey30"</span>),</span>
<span id="cb17-810"><a href="#cb17-810" aria-hidden="true" tabindex="-1"></a>                    <span class="at">label_nudge_y =</span> <span class="sc">-</span><span class="fl">0.3</span></span>
<span id="cb17-811"><a href="#cb17-811" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-812"><a href="#cb17-812" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="cb17-813"><a href="#cb17-813" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="cb17-814"><a href="#cb17-814" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">200</span>,</span>
<span id="cb17-815"><a href="#cb17-815" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">8</span>, </span>
<span id="cb17-816"><a href="#cb17-816" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">8</span>,</span>
<span id="cb17-817"><a href="#cb17-817" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="cb17-818"><a href="#cb17-818" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb17-819"><a href="#cb17-819" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="cb17-820"><a href="#cb17-820" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="cb17-821"><a href="#cb17-821" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"X (mm)"</span>),</span>
<span id="cb17-822"><a href="#cb17-822" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"Y (mm)"</span>), </span>
<span id="cb17-823"><a href="#cb17-823" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="cb17-824"><a href="#cb17-824" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors_use),</span>
<span id="cb17-825"><a href="#cb17-825" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">0.4</span>)),</span>
<span id="cb17-826"><a href="#cb17-826" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="cb17-827"><a href="#cb17-827" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="cb17-828"><a href="#cb17-828" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="cb17-829"><a href="#cb17-829" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-830"><a href="#cb17-830" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-831"><a href="#cb17-831" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-832"><a href="#cb17-832" aria-hidden="true" tabindex="-1"></a><span class="co"># XY (facets only)</span></span>
<span id="cb17-833"><a href="#cb17-833" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, <span class="at">color_by=</span><span class="st">'celltype_broad'</span>,</span>
<span id="cb17-834"><a href="#cb17-834" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-835"><a href="#cb17-835" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-836"><a href="#cb17-836" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="cb17-837"><a href="#cb17-837" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-838"><a href="#cb17-838" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span></span>
<span id="cb17-839"><a href="#cb17-839" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-840"><a href="#cb17-840" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_bar_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-841"><a href="#cb17-841" aria-hidden="true" tabindex="-1"></a>                    <span class="at">location =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">0</span>),</span>
<span id="cb17-842"><a href="#cb17-842" aria-hidden="true" tabindex="-1"></a>                    <span class="at">width =</span> <span class="dv">2</span>,</span>
<span id="cb17-843"><a href="#cb17-843" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n =</span> <span class="dv">3</span>,</span>
<span id="cb17-844"><a href="#cb17-844" aria-hidden="true" tabindex="-1"></a>                    <span class="at">height =</span> <span class="fl">0.1</span>,</span>
<span id="cb17-845"><a href="#cb17-845" aria-hidden="true" tabindex="-1"></a>                    <span class="at">scale_colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"grey30"</span>),</span>
<span id="cb17-846"><a href="#cb17-846" aria-hidden="true" tabindex="-1"></a>                    <span class="at">label_nudge_y =</span> <span class="sc">-</span><span class="fl">0.3</span></span>
<span id="cb17-847"><a href="#cb17-847" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-848"><a href="#cb17-848" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="cb17-849"><a href="#cb17-849" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="cb17-850"><a href="#cb17-850" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">100</span>,</span>
<span id="cb17-851"><a href="#cb17-851" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">5</span>,</span>
<span id="cb17-852"><a href="#cb17-852" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">5</span>,</span>
<span id="cb17-853"><a href="#cb17-853" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="cb17-854"><a href="#cb17-854" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb17-855"><a href="#cb17-855" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="cb17-856"><a href="#cb17-856" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="cb17-857"><a href="#cb17-857" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"X (mm)"</span>),</span>
<span id="cb17-858"><a href="#cb17-858" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"Y (mm)"</span>), </span>
<span id="cb17-859"><a href="#cb17-859" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="cb17-860"><a href="#cb17-860" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors_use),</span>
<span id="cb17-861"><a href="#cb17-861" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">0.4</span>)),</span>
<span id="cb17-862"><a href="#cb17-862" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="cb17-863"><a href="#cb17-863" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="cb17-864"><a href="#cb17-864" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="cb17-865"><a href="#cb17-865" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-866"><a href="#cb17-866" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-867"><a href="#cb17-867" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-868"><a href="#cb17-868" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP (global only)</span></span>
<span id="cb17-869"><a href="#cb17-869" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, </span>
<span id="cb17-870"><a href="#cb17-870" aria-hidden="true" tabindex="-1"></a>              <span class="at">obsm_key =</span> <span class="st">"umap_pr"</span>,</span>
<span id="cb17-871"><a href="#cb17-871" aria-hidden="true" tabindex="-1"></a>              <span class="at">color_by=</span><span class="st">'celltype_broad'</span>,</span>
<span id="cb17-872"><a href="#cb17-872" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-873"><a href="#cb17-873" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-874"><a href="#cb17-874" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="cb17-875"><a href="#cb17-875" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-876"><a href="#cb17-876" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span>, <span class="at">alpha=</span><span class="fl">0.1</span></span>
<span id="cb17-877"><a href="#cb17-877" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-878"><a href="#cb17-878" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_label_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-879"><a href="#cb17-879" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size =</span> <span class="dv">2</span></span>
<span id="cb17-880"><a href="#cb17-880" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-881"><a href="#cb17-881" aria-hidden="true" tabindex="-1"></a>                  <span class="at">labels_on_plot =</span> <span class="fu">data.frame</span>(),</span>
<span id="cb17-882"><a href="#cb17-882" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="cb17-883"><a href="#cb17-883" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="cb17-884"><a href="#cb17-884" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">200</span>,</span>
<span id="cb17-885"><a href="#cb17-885" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">8</span>,</span>
<span id="cb17-886"><a href="#cb17-886" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">8</span>,</span>
<span id="cb17-887"><a href="#cb17-887" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="cb17-888"><a href="#cb17-888" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb17-889"><a href="#cb17-889" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="cb17-890"><a href="#cb17-890" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="cb17-891"><a href="#cb17-891" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"UMAP 1"</span>),</span>
<span id="cb17-892"><a href="#cb17-892" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"UMAP 2"</span>), </span>
<span id="cb17-893"><a href="#cb17-893" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="cb17-894"><a href="#cb17-894" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors_use),</span>
<span id="cb17-895"><a href="#cb17-895" aria-hidden="true" tabindex="-1"></a>                <span class="co"># guides(color = guide_legend(</span></span>
<span id="cb17-896"><a href="#cb17-896" aria-hidden="true" tabindex="-1"></a>                <span class="co">#   title="Cell Type",</span></span>
<span id="cb17-897"><a href="#cb17-897" aria-hidden="true" tabindex="-1"></a>                <span class="co">#   override.aes = list(size = 3) ) ), </span></span>
<span id="cb17-898"><a href="#cb17-898" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb17-899"><a href="#cb17-899" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-900"><a href="#cb17-900" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb17-901"><a href="#cb17-901" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-902"><a href="#cb17-902" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP (facets only)</span></span>
<span id="cb17-903"><a href="#cb17-903" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, </span>
<span id="cb17-904"><a href="#cb17-904" aria-hidden="true" tabindex="-1"></a>              <span class="at">obsm_key =</span> <span class="st">"umap_pr"</span>,</span>
<span id="cb17-905"><a href="#cb17-905" aria-hidden="true" tabindex="-1"></a>              <span class="at">color_by=</span><span class="st">'celltype_broad'</span>,</span>
<span id="cb17-906"><a href="#cb17-906" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-907"><a href="#cb17-907" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-908"><a href="#cb17-908" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="cb17-909"><a href="#cb17-909" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-910"><a href="#cb17-910" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span>, <span class="at">alpha=</span><span class="fl">0.1</span></span>
<span id="cb17-911"><a href="#cb17-911" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-912"><a href="#cb17-912" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_label_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-913"><a href="#cb17-913" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size =</span> <span class="dv">2</span></span>
<span id="cb17-914"><a href="#cb17-914" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-915"><a href="#cb17-915" aria-hidden="true" tabindex="-1"></a>                  <span class="at">labels_on_plot =</span> <span class="fu">data.frame</span>(),</span>
<span id="cb17-916"><a href="#cb17-916" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="cb17-917"><a href="#cb17-917" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="cb17-918"><a href="#cb17-918" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">100</span>,</span>
<span id="cb17-919"><a href="#cb17-919" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">5</span>,</span>
<span id="cb17-920"><a href="#cb17-920" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">5</span>,</span>
<span id="cb17-921"><a href="#cb17-921" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="cb17-922"><a href="#cb17-922" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb17-923"><a href="#cb17-923" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="cb17-924"><a href="#cb17-924" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="cb17-925"><a href="#cb17-925" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"UMAP 1"</span>),</span>
<span id="cb17-926"><a href="#cb17-926" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"UMAP 2"</span>), </span>
<span id="cb17-927"><a href="#cb17-927" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="cb17-928"><a href="#cb17-928" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors_use),</span>
<span id="cb17-929"><a href="#cb17-929" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="cb17-930"><a href="#cb17-930" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="cb17-931"><a href="#cb17-931" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="cb17-932"><a href="#cb17-932" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-933"><a href="#cb17-933" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb17-934"><a href="#cb17-934" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-935"><a href="#cb17-935" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>in the plotting below, we will use this name to parse out files.</span>
<span id="cb17-936"><a href="#cb17-936" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-937"><a href="#cb17-937" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-938"><a href="#cb17-938" aria-hidden="true" tabindex="-1"></a>Here are the overall (global) plots.</span>
<span id="cb17-939"><a href="#cb17-939" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-940"><a href="#cb17-940" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset}</span>
<span id="cb17-941"><a href="#cb17-941" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-942"><a href="#cb17-942" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Cell Type Broad - UMAP</span></span>
<span id="cb17-943"><a href="#cb17-943" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-946"><a href="#cb17-946" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-947"><a href="#cb17-947" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-umap-celltype-broad</span></span>
<span id="cb17-948"><a href="#cb17-948" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-949"><a href="#cb17-949" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-950"><a href="#cb17-950" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-951"><a href="#cb17-951" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 6</span></span>
<span id="cb17-952"><a href="#cb17-952" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 5</span></span>
<span id="cb17-953"><a href="#cb17-953" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "UMAP with broad cell type in the full dataset."</span></span>
<span id="cb17-954"><a href="#cb17-954" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-955"><a href="#cb17-955" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-956"><a href="#cb17-956" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>, <span class="st">"broad__umap_pr__S0__plot.png"</span>))</span>
<span id="cb17-957"><a href="#cb17-957" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-958"><a href="#cb17-958" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-959"><a href="#cb17-959" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-960"><a href="#cb17-960" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-961"><a href="#cb17-961" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Cell Type Broad - XY</span></span>
<span id="cb17-962"><a href="#cb17-962" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-965"><a href="#cb17-965" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-966"><a href="#cb17-966" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-xy-celltype-broad</span></span>
<span id="cb17-967"><a href="#cb17-967" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-968"><a href="#cb17-968" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-969"><a href="#cb17-969" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-970"><a href="#cb17-970" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 6</span></span>
<span id="cb17-971"><a href="#cb17-971" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 5</span></span>
<span id="cb17-972"><a href="#cb17-972" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "XY with broad cell type in the full dataset."</span></span>
<span id="cb17-973"><a href="#cb17-973" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-974"><a href="#cb17-974" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-975"><a href="#cb17-975" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>, <span class="st">"broad__spatial__S0__plot.png"</span>))</span>
<span id="cb17-976"><a href="#cb17-976" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-977"><a href="#cb17-977" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-978"><a href="#cb17-978" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-979"><a href="#cb17-979" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb17-980"><a href="#cb17-980" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-981"><a href="#cb17-981" aria-hidden="true" tabindex="-1"></a>And here are the facetted plots.</span>
<span id="cb17-982"><a href="#cb17-982" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-983"><a href="#cb17-983" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset}</span>
<span id="cb17-984"><a href="#cb17-984" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-987"><a href="#cb17-987" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-988"><a href="#cb17-988" aria-hidden="true" tabindex="-1"></a><span class="co">#| output: asis</span></span>
<span id="cb17-989"><a href="#cb17-989" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-990"><a href="#cb17-990" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-991"><a href="#cb17-991" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb17-992"><a href="#cb17-992" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-993"><a href="#cb17-993" aria-hidden="true" tabindex="-1"></a>fig_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>)</span>
<span id="cb17-994"><a href="#cb17-994" aria-hidden="true" tabindex="-1"></a>group_prefix <span class="ot">&lt;-</span> <span class="st">"broad"</span></span>
<span id="cb17-995"><a href="#cb17-995" aria-hidden="true" tabindex="-1"></a>slide_id <span class="ot">&lt;-</span> <span class="st">"S0"</span></span>
<span id="cb17-996"><a href="#cb17-996" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-997"><a href="#cb17-997" aria-hidden="true" tabindex="-1"></a>xy_files <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(<span class="fu">file.path</span>(fig_dir, </span>
<span id="cb17-998"><a href="#cb17-998" aria-hidden="true" tabindex="-1"></a>                               <span class="fu">paste0</span>(group_prefix, </span>
<span id="cb17-999"><a href="#cb17-999" aria-hidden="true" tabindex="-1"></a>                                      <span class="st">"__spatial__"</span>, slide_id, <span class="st">"__facet_*png"</span>)))</span>
<span id="cb17-1000"><a href="#cb17-1000" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1001"><a href="#cb17-1001" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1002"><a href="#cb17-1002" aria-hidden="true" tabindex="-1"></a>umap_files <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(<span class="fu">file.path</span>(fig_dir, </span>
<span id="cb17-1003"><a href="#cb17-1003" aria-hidden="true" tabindex="-1"></a>                               <span class="fu">paste0</span>(group_prefix, </span>
<span id="cb17-1004"><a href="#cb17-1004" aria-hidden="true" tabindex="-1"></a>                                      <span class="st">"__umap_pr__"</span>, slide_id, <span class="st">"__facet_*png"</span>)))</span>
<span id="cb17-1005"><a href="#cb17-1005" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1006"><a href="#cb17-1006" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1007"><a href="#cb17-1007" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map2_chr</span>(xy_files, umap_files, \(current_xy_file, current_umap_file){</span>
<span id="cb17-1008"><a href="#cb17-1008" aria-hidden="true" tabindex="-1"></a>    knitr<span class="sc">::</span><span class="fu">knit_child</span>(<span class="at">text =</span> <span class="fu">c</span>(</span>
<span id="cb17-1009"><a href="#cb17-1009" aria-hidden="true" tabindex="-1"></a>      <span class="st">"### `r gsub('.png', '', strsplit(current_umap_file, split='__facet_')[[1]][2])`"</span>,</span>
<span id="cb17-1010"><a href="#cb17-1010" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1011"><a href="#cb17-1011" aria-hidden="true" tabindex="-1"></a>      <span class="st">":::{columns}"</span>,</span>
<span id="cb17-1012"><a href="#cb17-1012" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1013"><a href="#cb17-1013" aria-hidden="true" tabindex="-1"></a>      <span class="st">"::: {.column width='40%'}"</span>,</span>
<span id="cb17-1014"><a href="#cb17-1014" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1015"><a href="#cb17-1015" aria-hidden="true" tabindex="-1"></a>      <span class="st">"```{r eval=TRUE}"</span>,</span>
<span id="cb17-1016"><a href="#cb17-1016" aria-hidden="true" tabindex="-1"></a>      <span class="st">"#| echo: false"</span>,</span>
<span id="cb17-1017"><a href="#cb17-1017" aria-hidden="true" tabindex="-1"></a>      <span class="st">"knitr::include_graphics(current_xy_file)"</span>,</span>
<span id="cb17-1018"><a href="#cb17-1018" aria-hidden="true" tabindex="-1"></a>      <span class="st">"```"</span>,</span>
<span id="cb17-1019"><a href="#cb17-1019" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1020"><a href="#cb17-1020" aria-hidden="true" tabindex="-1"></a>      <span class="st">":::"</span>,</span>
<span id="cb17-1021"><a href="#cb17-1021" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1022"><a href="#cb17-1022" aria-hidden="true" tabindex="-1"></a>      <span class="st">"::: {.column width='10%'}"</span>,</span>
<span id="cb17-1023"><a href="#cb17-1023" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1024"><a href="#cb17-1024" aria-hidden="true" tabindex="-1"></a>      <span class="st">":::"</span>,</span>
<span id="cb17-1025"><a href="#cb17-1025" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1026"><a href="#cb17-1026" aria-hidden="true" tabindex="-1"></a>      <span class="st">"::: {.column width='40%'}"</span>,</span>
<span id="cb17-1027"><a href="#cb17-1027" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1028"><a href="#cb17-1028" aria-hidden="true" tabindex="-1"></a>      <span class="st">"```{r eval=TRUE}"</span>,</span>
<span id="cb17-1029"><a href="#cb17-1029" aria-hidden="true" tabindex="-1"></a>      <span class="st">"#| echo: false"</span>,</span>
<span id="cb17-1030"><a href="#cb17-1030" aria-hidden="true" tabindex="-1"></a>      <span class="st">"knitr::include_graphics(current_umap_file)"</span>,</span>
<span id="cb17-1031"><a href="#cb17-1031" aria-hidden="true" tabindex="-1"></a>      <span class="st">"```"</span>,</span>
<span id="cb17-1032"><a href="#cb17-1032" aria-hidden="true" tabindex="-1"></a>      <span class="st">":::"</span>,</span>
<span id="cb17-1033"><a href="#cb17-1033" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1034"><a href="#cb17-1034" aria-hidden="true" tabindex="-1"></a>      <span class="st">":::"</span>,</span>
<span id="cb17-1035"><a href="#cb17-1035" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1036"><a href="#cb17-1036" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1037"><a href="#cb17-1037" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span></span>
<span id="cb17-1038"><a href="#cb17-1038" aria-hidden="true" tabindex="-1"></a>    ), <span class="at">envir =</span> <span class="fu">environment</span>(), <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-1039"><a href="#cb17-1039" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb17-1040"><a href="#cb17-1040" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1041"><a href="#cb17-1041" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(res, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\n</span><span class="st">'</span>)</span>
<span id="cb17-1042"><a href="#cb17-1042" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1043"><a href="#cb17-1043" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1044"><a href="#cb17-1044" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1045"><a href="#cb17-1045" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb17-1046"><a href="#cb17-1046" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1047"><a href="#cb17-1047" aria-hidden="true" tabindex="-1"></a><span class="fu">### Fine cell types</span></span>
<span id="cb17-1048"><a href="#cb17-1048" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1049"><a href="#cb17-1049" aria-hidden="true" tabindex="-1"></a>This section performs a similar routine as above but with the fine cell type</span>
<span id="cb17-1050"><a href="#cb17-1050" aria-hidden="true" tabindex="-1"></a>classifications.</span>
<span id="cb17-1051"><a href="#cb17-1051" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1052"><a href="#cb17-1052" aria-hidden="true" tabindex="-1"></a>For the fine cell types, we'll create colors that are various shades of the</span>
<span id="cb17-1053"><a href="#cb17-1053" aria-hidden="true" tabindex="-1"></a>parent type.</span>
<span id="cb17-1054"><a href="#cb17-1054" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1057"><a href="#cb17-1057" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-1058"><a href="#cb17-1058" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: V-3</span></span>
<span id="cb17-1059"><a href="#cb17-1059" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-1060"><a href="#cb17-1060" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb17-1061"><a href="#cb17-1061" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-1062"><a href="#cb17-1062" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1063"><a href="#cb17-1063" aria-hidden="true" tabindex="-1"></a>broad_map <span class="ot">&lt;-</span> <span class="fu">unlist</span>(py<span class="sc">$</span>broad_map)</span>
<span id="cb17-1064"><a href="#cb17-1064" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1065"><a href="#cb17-1065" aria-hidden="true" tabindex="-1"></a>generate_fine_colors <span class="ot">&lt;-</span> <span class="cf">function</span>(mapping, broad_colors, <span class="at">mode =</span> <span class="st">"shades"</span>) {</span>
<span id="cb17-1066"><a href="#cb17-1066" aria-hidden="true" tabindex="-1"></a>  groups <span class="ot">&lt;-</span> <span class="fu">split</span>(<span class="fu">names</span>(mapping), mapping)</span>
<span id="cb17-1067"><a href="#cb17-1067" aria-hidden="true" tabindex="-1"></a>  fine_colors <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb17-1068"><a href="#cb17-1068" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (broad_name <span class="cf">in</span> <span class="fu">names</span>(groups)) {</span>
<span id="cb17-1069"><a href="#cb17-1069" aria-hidden="true" tabindex="-1"></a>    subtypes <span class="ot">&lt;-</span> groups[[broad_name]]</span>
<span id="cb17-1070"><a href="#cb17-1070" aria-hidden="true" tabindex="-1"></a>    n <span class="ot">&lt;-</span> <span class="fu">length</span>(subtypes)</span>
<span id="cb17-1071"><a href="#cb17-1071" aria-hidden="true" tabindex="-1"></a>    base_col <span class="ot">&lt;-</span> broad_colors[broad_name]</span>
<span id="cb17-1072"><a href="#cb17-1072" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (n <span class="sc">==</span> <span class="dv">1</span>) {</span>
<span id="cb17-1073"><a href="#cb17-1073" aria-hidden="true" tabindex="-1"></a>      cols <span class="ot">&lt;-</span> base_col</span>
<span id="cb17-1074"><a href="#cb17-1074" aria-hidden="true" tabindex="-1"></a>      <span class="fu">names</span>(cols) <span class="ot">&lt;-</span> subtypes</span>
<span id="cb17-1075"><a href="#cb17-1075" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb17-1076"><a href="#cb17-1076" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (mode <span class="sc">==</span> <span class="st">"shades"</span>) {</span>
<span id="cb17-1077"><a href="#cb17-1077" aria-hidden="true" tabindex="-1"></a>        ramp_func <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(</span>
<span id="cb17-1078"><a href="#cb17-1078" aria-hidden="true" tabindex="-1"></a>          <span class="fu">adjustcolor</span>(base_col, <span class="at">transform=</span><span class="fu">diag</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))<span class="sc">*</span><span class="fl">1.4</span>), <span class="co"># Lighten</span></span>
<span id="cb17-1079"><a href="#cb17-1079" aria-hidden="true" tabindex="-1"></a>          base_col,</span>
<span id="cb17-1080"><a href="#cb17-1080" aria-hidden="true" tabindex="-1"></a>          <span class="fu">adjustcolor</span>(base_col, <span class="at">transform=</span><span class="fu">diag</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))<span class="sc">*</span><span class="fl">0.6</span>)  <span class="co"># Darken</span></span>
<span id="cb17-1081"><a href="#cb17-1081" aria-hidden="true" tabindex="-1"></a>        ))</span>
<span id="cb17-1082"><a href="#cb17-1082" aria-hidden="true" tabindex="-1"></a>        cols <span class="ot">&lt;-</span> <span class="fu">ramp_func</span>(n)</span>
<span id="cb17-1083"><a href="#cb17-1083" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb17-1084"><a href="#cb17-1084" aria-hidden="true" tabindex="-1"></a>        ramp_func <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"#FFFFFF"</span>, base_col, <span class="st">"#000000"</span>))</span>
<span id="cb17-1085"><a href="#cb17-1085" aria-hidden="true" tabindex="-1"></a>        cols <span class="ot">&lt;-</span> <span class="fu">ramp_func</span>(n <span class="sc">+</span> <span class="dv">2</span>)[<span class="dv">2</span><span class="sc">:</span>(n <span class="sc">+</span> <span class="dv">1</span>)]</span>
<span id="cb17-1086"><a href="#cb17-1086" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb17-1087"><a href="#cb17-1087" aria-hidden="true" tabindex="-1"></a>      <span class="fu">names</span>(cols) <span class="ot">&lt;-</span> subtypes</span>
<span id="cb17-1088"><a href="#cb17-1088" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb17-1089"><a href="#cb17-1089" aria-hidden="true" tabindex="-1"></a>    fine_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(fine_colors, cols)</span>
<span id="cb17-1090"><a href="#cb17-1090" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb17-1091"><a href="#cb17-1091" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(fine_colors)</span>
<span id="cb17-1092"><a href="#cb17-1092" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-1093"><a href="#cb17-1093" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1094"><a href="#cb17-1094" aria-hidden="true" tabindex="-1"></a>fine_colors <span class="ot">&lt;-</span> <span class="fu">generate_fine_colors</span>(broad_map, colors_use)</span>
<span id="cb17-1095"><a href="#cb17-1095" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1096"><a href="#cb17-1096" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'celltype_fine_names'</span>]] <span class="ot">&lt;-</span> <span class="fu">names</span>(fine_colors)</span>
<span id="cb17-1097"><a href="#cb17-1097" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'celltype_fine_colors'</span>]] <span class="ot">&lt;-</span> <span class="fu">as.character</span>(fine_colors)</span>
<span id="cb17-1098"><a href="#cb17-1098" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1099"><a href="#cb17-1099" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb17-1100"><a href="#cb17-1100" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1101"><a href="#cb17-1101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1102"><a href="#cb17-1102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1105"><a href="#cb17-1105" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-1106"><a href="#cb17-1106" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-1107"><a href="#cb17-1107" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb17-1108"><a href="#cb17-1108" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-1109"><a href="#cb17-1109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1110"><a href="#cb17-1110" aria-hidden="true" tabindex="-1"></a>group_prefix <span class="ot">&lt;-</span> <span class="st">"fine"</span> <span class="co"># &lt;1&gt;</span></span>
<span id="cb17-1111"><a href="#cb17-1111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1112"><a href="#cb17-1112" aria-hidden="true" tabindex="-1"></a>ct_assets_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>)</span>
<span id="cb17-1113"><a href="#cb17-1113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1114"><a href="#cb17-1114" aria-hidden="true" tabindex="-1"></a><span class="co"># XY (global only)</span></span>
<span id="cb17-1115"><a href="#cb17-1115" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, <span class="at">color_by=</span><span class="st">'celltype_fine'</span>,</span>
<span id="cb17-1116"><a href="#cb17-1116" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-1117"><a href="#cb17-1117" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-1118"><a href="#cb17-1118" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="cb17-1119"><a href="#cb17-1119" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-1120"><a href="#cb17-1120" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span></span>
<span id="cb17-1121"><a href="#cb17-1121" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-1122"><a href="#cb17-1122" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_bar_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-1123"><a href="#cb17-1123" aria-hidden="true" tabindex="-1"></a>                    <span class="at">location =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">0</span>),</span>
<span id="cb17-1124"><a href="#cb17-1124" aria-hidden="true" tabindex="-1"></a>                    <span class="at">width =</span> <span class="dv">2</span>,</span>
<span id="cb17-1125"><a href="#cb17-1125" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n =</span> <span class="dv">3</span>,</span>
<span id="cb17-1126"><a href="#cb17-1126" aria-hidden="true" tabindex="-1"></a>                    <span class="at">height =</span> <span class="fl">0.1</span>,</span>
<span id="cb17-1127"><a href="#cb17-1127" aria-hidden="true" tabindex="-1"></a>                    <span class="at">scale_colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"grey30"</span>),</span>
<span id="cb17-1128"><a href="#cb17-1128" aria-hidden="true" tabindex="-1"></a>                    <span class="at">label_nudge_y =</span> <span class="sc">-</span><span class="fl">0.3</span></span>
<span id="cb17-1129"><a href="#cb17-1129" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-1130"><a href="#cb17-1130" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="cb17-1131"><a href="#cb17-1131" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="cb17-1132"><a href="#cb17-1132" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">200</span>,</span>
<span id="cb17-1133"><a href="#cb17-1133" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">8</span>, </span>
<span id="cb17-1134"><a href="#cb17-1134" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">8</span>,</span>
<span id="cb17-1135"><a href="#cb17-1135" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="cb17-1136"><a href="#cb17-1136" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb17-1137"><a href="#cb17-1137" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="cb17-1138"><a href="#cb17-1138" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="cb17-1139"><a href="#cb17-1139" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"X (mm)"</span>),</span>
<span id="cb17-1140"><a href="#cb17-1140" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"Y (mm)"</span>), </span>
<span id="cb17-1141"><a href="#cb17-1141" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="cb17-1142"><a href="#cb17-1142" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> fine_colors),</span>
<span id="cb17-1143"><a href="#cb17-1143" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">0.4</span>)),</span>
<span id="cb17-1144"><a href="#cb17-1144" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="cb17-1145"><a href="#cb17-1145" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="cb17-1146"><a href="#cb17-1146" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="cb17-1147"><a href="#cb17-1147" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-1148"><a href="#cb17-1148" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-1149"><a href="#cb17-1149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1150"><a href="#cb17-1150" aria-hidden="true" tabindex="-1"></a><span class="co"># XY (facets only)</span></span>
<span id="cb17-1151"><a href="#cb17-1151" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, <span class="at">color_by=</span><span class="st">'celltype_fine'</span>,</span>
<span id="cb17-1152"><a href="#cb17-1152" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-1153"><a href="#cb17-1153" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-1154"><a href="#cb17-1154" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="cb17-1155"><a href="#cb17-1155" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-1156"><a href="#cb17-1156" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span></span>
<span id="cb17-1157"><a href="#cb17-1157" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-1158"><a href="#cb17-1158" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_bar_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-1159"><a href="#cb17-1159" aria-hidden="true" tabindex="-1"></a>                    <span class="at">location =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">0</span>),</span>
<span id="cb17-1160"><a href="#cb17-1160" aria-hidden="true" tabindex="-1"></a>                    <span class="at">width =</span> <span class="dv">2</span>,</span>
<span id="cb17-1161"><a href="#cb17-1161" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n =</span> <span class="dv">3</span>,</span>
<span id="cb17-1162"><a href="#cb17-1162" aria-hidden="true" tabindex="-1"></a>                    <span class="at">height =</span> <span class="fl">0.1</span>,</span>
<span id="cb17-1163"><a href="#cb17-1163" aria-hidden="true" tabindex="-1"></a>                    <span class="at">scale_colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"grey30"</span>),</span>
<span id="cb17-1164"><a href="#cb17-1164" aria-hidden="true" tabindex="-1"></a>                    <span class="at">label_nudge_y =</span> <span class="sc">-</span><span class="fl">0.3</span></span>
<span id="cb17-1165"><a href="#cb17-1165" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-1166"><a href="#cb17-1166" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="cb17-1167"><a href="#cb17-1167" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="cb17-1168"><a href="#cb17-1168" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">100</span>,</span>
<span id="cb17-1169"><a href="#cb17-1169" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">5</span>,</span>
<span id="cb17-1170"><a href="#cb17-1170" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">5</span>,</span>
<span id="cb17-1171"><a href="#cb17-1171" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="cb17-1172"><a href="#cb17-1172" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb17-1173"><a href="#cb17-1173" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="cb17-1174"><a href="#cb17-1174" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="cb17-1175"><a href="#cb17-1175" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"X (mm)"</span>),</span>
<span id="cb17-1176"><a href="#cb17-1176" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"Y (mm)"</span>), </span>
<span id="cb17-1177"><a href="#cb17-1177" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="cb17-1178"><a href="#cb17-1178" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> fine_colors),</span>
<span id="cb17-1179"><a href="#cb17-1179" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">0.4</span>)),</span>
<span id="cb17-1180"><a href="#cb17-1180" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="cb17-1181"><a href="#cb17-1181" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="cb17-1182"><a href="#cb17-1182" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="cb17-1183"><a href="#cb17-1183" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-1184"><a href="#cb17-1184" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-1185"><a href="#cb17-1185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1186"><a href="#cb17-1186" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP (global only)</span></span>
<span id="cb17-1187"><a href="#cb17-1187" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, </span>
<span id="cb17-1188"><a href="#cb17-1188" aria-hidden="true" tabindex="-1"></a>              <span class="at">obsm_key =</span> <span class="st">"umap_pr"</span>,</span>
<span id="cb17-1189"><a href="#cb17-1189" aria-hidden="true" tabindex="-1"></a>              <span class="at">color_by=</span><span class="st">'celltype_fine'</span>,</span>
<span id="cb17-1190"><a href="#cb17-1190" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-1191"><a href="#cb17-1191" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-1192"><a href="#cb17-1192" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="cb17-1193"><a href="#cb17-1193" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-1194"><a href="#cb17-1194" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span>, <span class="at">alpha=</span><span class="fl">0.1</span></span>
<span id="cb17-1195"><a href="#cb17-1195" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-1196"><a href="#cb17-1196" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_label_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-1197"><a href="#cb17-1197" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size =</span> <span class="dv">2</span></span>
<span id="cb17-1198"><a href="#cb17-1198" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-1199"><a href="#cb17-1199" aria-hidden="true" tabindex="-1"></a>                  <span class="at">labels_on_plot =</span> <span class="fu">data.frame</span>(),</span>
<span id="cb17-1200"><a href="#cb17-1200" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="cb17-1201"><a href="#cb17-1201" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="cb17-1202"><a href="#cb17-1202" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">200</span>,</span>
<span id="cb17-1203"><a href="#cb17-1203" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">8</span>,</span>
<span id="cb17-1204"><a href="#cb17-1204" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">8</span>,</span>
<span id="cb17-1205"><a href="#cb17-1205" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="cb17-1206"><a href="#cb17-1206" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb17-1207"><a href="#cb17-1207" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="cb17-1208"><a href="#cb17-1208" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="cb17-1209"><a href="#cb17-1209" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"UMAP 1"</span>),</span>
<span id="cb17-1210"><a href="#cb17-1210" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"UMAP 2"</span>), </span>
<span id="cb17-1211"><a href="#cb17-1211" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="cb17-1212"><a href="#cb17-1212" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> fine_colors),</span>
<span id="cb17-1213"><a href="#cb17-1213" aria-hidden="true" tabindex="-1"></a>                <span class="co"># guides(color = guide_legend(</span></span>
<span id="cb17-1214"><a href="#cb17-1214" aria-hidden="true" tabindex="-1"></a>                <span class="co">#   title="Cell Type",</span></span>
<span id="cb17-1215"><a href="#cb17-1215" aria-hidden="true" tabindex="-1"></a>                <span class="co">#   override.aes = list(size = 3) ) ), </span></span>
<span id="cb17-1216"><a href="#cb17-1216" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb17-1217"><a href="#cb17-1217" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-1218"><a href="#cb17-1218" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb17-1219"><a href="#cb17-1219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1220"><a href="#cb17-1220" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP (facets only)</span></span>
<span id="cb17-1221"><a href="#cb17-1221" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, </span>
<span id="cb17-1222"><a href="#cb17-1222" aria-hidden="true" tabindex="-1"></a>              <span class="at">obsm_key =</span> <span class="st">"umap_pr"</span>,</span>
<span id="cb17-1223"><a href="#cb17-1223" aria-hidden="true" tabindex="-1"></a>              <span class="at">color_by=</span><span class="st">'celltype_fine'</span>,</span>
<span id="cb17-1224"><a href="#cb17-1224" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-1225"><a href="#cb17-1225" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-1226"><a href="#cb17-1226" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="cb17-1227"><a href="#cb17-1227" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-1228"><a href="#cb17-1228" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span>, <span class="at">alpha=</span><span class="fl">0.1</span></span>
<span id="cb17-1229"><a href="#cb17-1229" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-1230"><a href="#cb17-1230" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_label_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-1231"><a href="#cb17-1231" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size =</span> <span class="dv">2</span></span>
<span id="cb17-1232"><a href="#cb17-1232" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-1233"><a href="#cb17-1233" aria-hidden="true" tabindex="-1"></a>                  <span class="at">labels_on_plot =</span> <span class="fu">data.frame</span>(),</span>
<span id="cb17-1234"><a href="#cb17-1234" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="cb17-1235"><a href="#cb17-1235" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="cb17-1236"><a href="#cb17-1236" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">100</span>,</span>
<span id="cb17-1237"><a href="#cb17-1237" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">5</span>,</span>
<span id="cb17-1238"><a href="#cb17-1238" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">5</span>,</span>
<span id="cb17-1239"><a href="#cb17-1239" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="cb17-1240"><a href="#cb17-1240" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb17-1241"><a href="#cb17-1241" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="cb17-1242"><a href="#cb17-1242" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="cb17-1243"><a href="#cb17-1243" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"UMAP 1"</span>),</span>
<span id="cb17-1244"><a href="#cb17-1244" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"UMAP 2"</span>), </span>
<span id="cb17-1245"><a href="#cb17-1245" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="cb17-1246"><a href="#cb17-1246" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> fine_colors),</span>
<span id="cb17-1247"><a href="#cb17-1247" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="cb17-1248"><a href="#cb17-1248" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Cell Type"</span>,</span>
<span id="cb17-1249"><a href="#cb17-1249" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="cb17-1250"><a href="#cb17-1250" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-1251"><a href="#cb17-1251" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb17-1252"><a href="#cb17-1252" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1253"><a href="#cb17-1253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1254"><a href="#cb17-1254" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset}</span>
<span id="cb17-1255"><a href="#cb17-1255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1256"><a href="#cb17-1256" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Cell Type Fine - UMAP</span></span>
<span id="cb17-1257"><a href="#cb17-1257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1260"><a href="#cb17-1260" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-1261"><a href="#cb17-1261" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-umap-celltype-fine</span></span>
<span id="cb17-1262"><a href="#cb17-1262" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-1263"><a href="#cb17-1263" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-1264"><a href="#cb17-1264" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-1265"><a href="#cb17-1265" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 6</span></span>
<span id="cb17-1266"><a href="#cb17-1266" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 5</span></span>
<span id="cb17-1267"><a href="#cb17-1267" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "UMAP with fine cell type labels."</span></span>
<span id="cb17-1268"><a href="#cb17-1268" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-1269"><a href="#cb17-1269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1270"><a href="#cb17-1270" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>, <span class="st">"fine__umap_pr__S0__plot.png"</span>))</span>
<span id="cb17-1271"><a href="#cb17-1271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1272"><a href="#cb17-1272" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1273"><a href="#cb17-1273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1274"><a href="#cb17-1274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1275"><a href="#cb17-1275" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Cell Type Fine - XY</span></span>
<span id="cb17-1276"><a href="#cb17-1276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1279"><a href="#cb17-1279" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-1280"><a href="#cb17-1280" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-xy-celltype-fine</span></span>
<span id="cb17-1281"><a href="#cb17-1281" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-1282"><a href="#cb17-1282" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-1283"><a href="#cb17-1283" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-1284"><a href="#cb17-1284" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 6</span></span>
<span id="cb17-1285"><a href="#cb17-1285" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 5</span></span>
<span id="cb17-1286"><a href="#cb17-1286" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "XY with fine cell type labels."</span></span>
<span id="cb17-1287"><a href="#cb17-1287" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-1288"><a href="#cb17-1288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1289"><a href="#cb17-1289" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>, <span class="st">"fine__spatial__S0__plot.png"</span>))</span>
<span id="cb17-1290"><a href="#cb17-1290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1291"><a href="#cb17-1291" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1292"><a href="#cb17-1292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1293"><a href="#cb17-1293" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb17-1294"><a href="#cb17-1294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1295"><a href="#cb17-1295" aria-hidden="true" tabindex="-1"></a>And here are the facetted plots. </span>
<span id="cb17-1296"><a href="#cb17-1296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1297"><a href="#cb17-1297" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset}</span>
<span id="cb17-1298"><a href="#cb17-1298" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1301"><a href="#cb17-1301" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-1302"><a href="#cb17-1302" aria-hidden="true" tabindex="-1"></a><span class="co">#| output: asis</span></span>
<span id="cb17-1303"><a href="#cb17-1303" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-1304"><a href="#cb17-1304" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-1305"><a href="#cb17-1305" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb17-1306"><a href="#cb17-1306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1307"><a href="#cb17-1307" aria-hidden="true" tabindex="-1"></a>fig_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>)</span>
<span id="cb17-1308"><a href="#cb17-1308" aria-hidden="true" tabindex="-1"></a>group_prefix <span class="ot">&lt;-</span> <span class="st">"fine"</span></span>
<span id="cb17-1309"><a href="#cb17-1309" aria-hidden="true" tabindex="-1"></a>slide_id <span class="ot">&lt;-</span> <span class="st">"S0"</span></span>
<span id="cb17-1310"><a href="#cb17-1310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1311"><a href="#cb17-1311" aria-hidden="true" tabindex="-1"></a>xy_files <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(<span class="fu">file.path</span>(fig_dir, </span>
<span id="cb17-1312"><a href="#cb17-1312" aria-hidden="true" tabindex="-1"></a>                               <span class="fu">paste0</span>(group_prefix, </span>
<span id="cb17-1313"><a href="#cb17-1313" aria-hidden="true" tabindex="-1"></a>                                      <span class="st">"__spatial__"</span>, slide_id, <span class="st">"__facet_*png"</span>)))</span>
<span id="cb17-1314"><a href="#cb17-1314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1315"><a href="#cb17-1315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1316"><a href="#cb17-1316" aria-hidden="true" tabindex="-1"></a>umap_files <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(<span class="fu">file.path</span>(fig_dir, </span>
<span id="cb17-1317"><a href="#cb17-1317" aria-hidden="true" tabindex="-1"></a>                               <span class="fu">paste0</span>(group_prefix, </span>
<span id="cb17-1318"><a href="#cb17-1318" aria-hidden="true" tabindex="-1"></a>                                      <span class="st">"__umap_pr__"</span>, slide_id, <span class="st">"__facet_*png"</span>)))</span>
<span id="cb17-1319"><a href="#cb17-1319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1320"><a href="#cb17-1320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1321"><a href="#cb17-1321" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map2_chr</span>(xy_files, umap_files, \(current_xy_file, current_umap_file){</span>
<span id="cb17-1322"><a href="#cb17-1322" aria-hidden="true" tabindex="-1"></a>    knitr<span class="sc">::</span><span class="fu">knit_child</span>(<span class="at">text =</span> <span class="fu">c</span>(</span>
<span id="cb17-1323"><a href="#cb17-1323" aria-hidden="true" tabindex="-1"></a>      <span class="st">"### `r gsub('.png', '', strsplit(current_umap_file, split='__facet_')[[1]][2])`"</span>,</span>
<span id="cb17-1324"><a href="#cb17-1324" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1325"><a href="#cb17-1325" aria-hidden="true" tabindex="-1"></a>      <span class="st">":::{columns}"</span>,</span>
<span id="cb17-1326"><a href="#cb17-1326" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1327"><a href="#cb17-1327" aria-hidden="true" tabindex="-1"></a>      <span class="st">"::: {.column width='40%'}"</span>,</span>
<span id="cb17-1328"><a href="#cb17-1328" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1329"><a href="#cb17-1329" aria-hidden="true" tabindex="-1"></a>      <span class="st">"```{r eval=TRUE}"</span>,</span>
<span id="cb17-1330"><a href="#cb17-1330" aria-hidden="true" tabindex="-1"></a>      <span class="st">"#| echo: false"</span>,</span>
<span id="cb17-1331"><a href="#cb17-1331" aria-hidden="true" tabindex="-1"></a>      <span class="st">"knitr::include_graphics(current_xy_file)"</span>,</span>
<span id="cb17-1332"><a href="#cb17-1332" aria-hidden="true" tabindex="-1"></a>      <span class="st">"```"</span>,</span>
<span id="cb17-1333"><a href="#cb17-1333" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1334"><a href="#cb17-1334" aria-hidden="true" tabindex="-1"></a>      <span class="st">":::"</span>,</span>
<span id="cb17-1335"><a href="#cb17-1335" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1336"><a href="#cb17-1336" aria-hidden="true" tabindex="-1"></a>      <span class="st">"::: {.column width='10%'}"</span>,</span>
<span id="cb17-1337"><a href="#cb17-1337" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1338"><a href="#cb17-1338" aria-hidden="true" tabindex="-1"></a>      <span class="st">":::"</span>,</span>
<span id="cb17-1339"><a href="#cb17-1339" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1340"><a href="#cb17-1340" aria-hidden="true" tabindex="-1"></a>      <span class="st">"::: {.column width='40%'}"</span>,</span>
<span id="cb17-1341"><a href="#cb17-1341" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1342"><a href="#cb17-1342" aria-hidden="true" tabindex="-1"></a>      <span class="st">"```{r eval=TRUE}"</span>,</span>
<span id="cb17-1343"><a href="#cb17-1343" aria-hidden="true" tabindex="-1"></a>      <span class="st">"#| echo: false"</span>,</span>
<span id="cb17-1344"><a href="#cb17-1344" aria-hidden="true" tabindex="-1"></a>      <span class="st">"knitr::include_graphics(current_umap_file)"</span>,</span>
<span id="cb17-1345"><a href="#cb17-1345" aria-hidden="true" tabindex="-1"></a>      <span class="st">"```"</span>,</span>
<span id="cb17-1346"><a href="#cb17-1346" aria-hidden="true" tabindex="-1"></a>      <span class="st">":::"</span>,</span>
<span id="cb17-1347"><a href="#cb17-1347" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1348"><a href="#cb17-1348" aria-hidden="true" tabindex="-1"></a>      <span class="st">":::"</span>,</span>
<span id="cb17-1349"><a href="#cb17-1349" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1350"><a href="#cb17-1350" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span>,</span>
<span id="cb17-1351"><a href="#cb17-1351" aria-hidden="true" tabindex="-1"></a>      <span class="st">""</span></span>
<span id="cb17-1352"><a href="#cb17-1352" aria-hidden="true" tabindex="-1"></a>    ), <span class="at">envir =</span> <span class="fu">environment</span>(), <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-1353"><a href="#cb17-1353" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb17-1354"><a href="#cb17-1354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1355"><a href="#cb17-1355" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(res, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\n</span><span class="st">'</span>)</span>
<span id="cb17-1356"><a href="#cb17-1356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1357"><a href="#cb17-1357" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1358"><a href="#cb17-1358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1359"><a href="#cb17-1359" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb17-1360"><a href="#cb17-1360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1361"><a href="#cb17-1361" aria-hidden="true" tabindex="-1"></a>Note that within the SMC XY plots we can see an "outline" that appears to be the</span>
<span id="cb17-1362"><a href="#cb17-1362" aria-hidden="true" tabindex="-1"></a>muscularis mucosa.</span>
<span id="cb17-1363"><a href="#cb17-1363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1364"><a href="#cb17-1364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1365"><a href="#cb17-1365" aria-hidden="true" tabindex="-1"></a><span class="fu">## Integration with Spatial Domains {#sec-integration}</span></span>
<span id="cb17-1366"><a href="#cb17-1366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1367"><a href="#cb17-1367" aria-hidden="true" tabindex="-1"></a>Recall in @sec-domains that we used <span class="in">`novae`</span> to define spatial domains. Now that</span>
<span id="cb17-1368"><a href="#cb17-1368" aria-hidden="true" tabindex="-1"></a>we have cell type information, let's look at the composition of cell types</span>
<span id="cb17-1369"><a href="#cb17-1369" aria-hidden="true" tabindex="-1"></a>within each domain and assign a biological function to them.</span>
<span id="cb17-1370"><a href="#cb17-1370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1371"><a href="#cb17-1371" aria-hidden="true" tabindex="-1"></a>Since we only need the metadata associated with the spatial domain results, we'll</span>
<span id="cb17-1372"><a href="#cb17-1372" aria-hidden="true" tabindex="-1"></a>only load the obs.</span>
<span id="cb17-1373"><a href="#cb17-1373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1376"><a href="#cb17-1376" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb17-1377"><a href="#cb17-1377" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: I-1</span></span>
<span id="cb17-1378"><a href="#cb17-1378" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-1379"><a href="#cb17-1379" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb17-1380"><a href="#cb17-1380" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-1381"><a href="#cb17-1381" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-1382"><a href="#cb17-1382" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-1383"><a href="#cb17-1383" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1384"><a href="#cb17-1384" aria-hidden="true" tabindex="-1"></a>focal_domain_resolution <span class="op">=</span> <span class="st">'novae_domains_6'</span></span>
<span id="cb17-1385"><a href="#cb17-1385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1386"><a href="#cb17-1386" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'adata_domains'</span> <span class="kw">not</span> <span class="kw">in</span> <span class="bu">dir</span>():</span>
<span id="cb17-1387"><a href="#cb17-1387" aria-hidden="true" tabindex="-1"></a>  filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-3-novae.h5ad"</span>)</span>
<span id="cb17-1388"><a href="#cb17-1388" aria-hidden="true" tabindex="-1"></a>  adata_domains <span class="op">=</span> ad.read_h5ad(filename, backed<span class="op">=</span><span class="st">'r'</span>)</span>
<span id="cb17-1389"><a href="#cb17-1389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1390"><a href="#cb17-1390" aria-hidden="true" tabindex="-1"></a>meta_domains <span class="op">=</span> adata_domains.obs.copy()</span>
<span id="cb17-1391"><a href="#cb17-1391" aria-hidden="true" tabindex="-1"></a>columns_to_copy <span class="op">=</span> [focal_domain_resolution]</span>
<span id="cb17-1392"><a href="#cb17-1392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1393"><a href="#cb17-1393" aria-hidden="true" tabindex="-1"></a>adata.obs[columns_to_copy] <span class="op">=</span> meta_domains[columns_to_copy]</span>
<span id="cb17-1394"><a href="#cb17-1394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1395"><a href="#cb17-1395" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1396"><a href="#cb17-1396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1397"><a href="#cb17-1397" aria-hidden="true" tabindex="-1"></a>Create a contingency table for the domain and cell type assignments.</span>
<span id="cb17-1398"><a href="#cb17-1398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1401"><a href="#cb17-1401" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-1402"><a href="#cb17-1402" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: I-2</span></span>
<span id="cb17-1403"><a href="#cb17-1403" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-1404"><a href="#cb17-1404" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb17-1405"><a href="#cb17-1405" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-1406"><a href="#cb17-1406" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-1407"><a href="#cb17-1407" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-1408"><a href="#cb17-1408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1409"><a href="#cb17-1409" aria-hidden="true" tabindex="-1"></a>focal_domain_resolution <span class="ot">&lt;-</span> py<span class="sc">$</span>focal_domain_resolution</span>
<span id="cb17-1410"><a href="#cb17-1410" aria-hidden="true" tabindex="-1"></a>focal_celltype_resolution <span class="ot">&lt;-</span> <span class="st">"celltype_fine"</span></span>
<span id="cb17-1411"><a href="#cb17-1411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1412"><a href="#cb17-1412" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs <span class="sc">%&gt;%</span> <span class="fu">select</span>(</span>
<span id="cb17-1413"><a href="#cb17-1413" aria-hidden="true" tabindex="-1"></a>    <span class="fu">all_of</span>(focal_domain_resolution), </span>
<span id="cb17-1414"><a href="#cb17-1414" aria-hidden="true" tabindex="-1"></a>    <span class="fu">all_of</span>(focal_celltype_resolution))</span>
<span id="cb17-1415"><a href="#cb17-1415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1416"><a href="#cb17-1416" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> </span>
<span id="cb17-1417"><a href="#cb17-1417" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(.data[[focal_domain_resolution]], </span>
<span id="cb17-1418"><a href="#cb17-1418" aria-hidden="true" tabindex="-1"></a>           .data[[focal_celltype_resolution]]) <span class="sc">%&gt;%</span> </span>
<span id="cb17-1419"><a href="#cb17-1419" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">count =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span> </span>
<span id="cb17-1420"><a href="#cb17-1420" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">proportion =</span> count <span class="sc">/</span> <span class="fu">sum</span>(count))</span>
<span id="cb17-1421"><a href="#cb17-1421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1422"><a href="#cb17-1422" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the geom_tile plot</span></span>
<span id="cb17-1423"><a href="#cb17-1423" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(plot_data, <span class="fu">aes</span>(<span class="at">x =</span> .data[[focal_domain_resolution]],</span>
<span id="cb17-1424"><a href="#cb17-1424" aria-hidden="true" tabindex="-1"></a>                           <span class="at">y =</span> .data[[focal_celltype_resolution]], </span>
<span id="cb17-1425"><a href="#cb17-1425" aria-hidden="true" tabindex="-1"></a>                           <span class="at">fill =</span> proportion)) <span class="sc">+</span></span>
<span id="cb17-1426"><a href="#cb17-1426" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="at">color =</span> <span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb17-1427"><a href="#cb17-1427" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> <span class="fu">paste0</span>(<span class="fu">round</span>(<span class="dv">100</span><span class="sc">*</span>proportion, <span class="dv">3</span>), <span class="st">"%"</span>)), <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb17-1428"><a href="#cb17-1428" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient</span>(<span class="at">low =</span> <span class="st">"white"</span>, <span class="at">high =</span> <span class="st">"dodgerblue"</span>) <span class="sc">+</span>  <span class="co"># Set color scale</span></span>
<span id="cb17-1429"><a href="#cb17-1429" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Cell Type"</span>, <span class="at">x =</span> <span class="st">"Spatial Domain"</span>, <span class="at">fill =</span> <span class="st">"Proportion"</span>) <span class="sc">+</span></span>
<span id="cb17-1430"><a href="#cb17-1430" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb17-1431"><a href="#cb17-1431" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) </span>
<span id="cb17-1432"><a href="#cb17-1432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1433"><a href="#cb17-1433" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(ct_dir, <span class="st">"contingency_table.png"</span>), </span>
<span id="cb17-1434"><a href="#cb17-1434" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">5</span>, <span class="at">height=</span><span class="dv">7</span>)</span>
<span id="cb17-1435"><a href="#cb17-1435" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(plot_data, <span class="st">"./tmp.csv"</span>, <span class="at">col.names=</span>T, <span class="at">row.names=</span>F, <span class="at">quote=</span>F)</span>
<span id="cb17-1436"><a href="#cb17-1436" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1437"><a href="#cb17-1437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1440"><a href="#cb17-1440" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-1441"><a href="#cb17-1441" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-contingency-table</span></span>
<span id="cb17-1442"><a href="#cb17-1442" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-1443"><a href="#cb17-1443" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-1444"><a href="#cb17-1444" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-1445"><a href="#cb17-1445" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 5</span></span>
<span id="cb17-1446"><a href="#cb17-1446" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 6</span></span>
<span id="cb17-1447"><a href="#cb17-1447" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "The proportion of a given cell type found in each domain. For example, domain D1015 is comprised of 65.3% smooth muscle cells (SMCs)."</span></span>
<span id="cb17-1448"><a href="#cb17-1448" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-1449"><a href="#cb17-1449" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1450"><a href="#cb17-1450" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>), <span class="st">"contingency_table.png"</span>, ct_dir, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb17-1451"><a href="#cb17-1451" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1452"><a href="#cb17-1452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1453"><a href="#cb17-1453" aria-hidden="true" tabindex="-1"></a>From this information and the spatial layout of domains we can assign a function </span>
<span id="cb17-1454"><a href="#cb17-1454" aria-hidden="true" tabindex="-1"></a>names @tbl-domains-definitions. Just looking at the composition, the differences</span>
<span id="cb17-1455"><a href="#cb17-1455" aria-hidden="true" tabindex="-1"></a>between three of the domains (D1011, D1015, and D1016) appear to have varying</span>
<span id="cb17-1456"><a href="#cb17-1456" aria-hidden="true" tabindex="-1"></a>degrees of tumor, cycling tumor, and CAFs. D1015 has relatively more CAFs so let's</span>
<span id="cb17-1457"><a href="#cb17-1457" aria-hidden="true" tabindex="-1"></a>call that a Desmoplastic Stroma domain and let's label D1016, which has the highest</span>
<span id="cb17-1458"><a href="#cb17-1458" aria-hidden="true" tabindex="-1"></a>concentration of tumor cells, the Tumor Core. D1011 has abundant tumor (cycling, </span>
<span id="cb17-1459"><a href="#cb17-1459" aria-hidden="true" tabindex="-1"></a>specifically) and also 12% B cell and 20% (normal) epithelial cells. Spatially, </span>
<span id="cb17-1460"><a href="#cb17-1460" aria-hidden="true" tabindex="-1"></a>D1011 occurs throughout lymphoid structures within the epithelium and around the</span>
<span id="cb17-1461"><a href="#cb17-1461" aria-hidden="true" tabindex="-1"></a>tumor region (invasive margin). </span>
<span id="cb17-1462"><a href="#cb17-1462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1463"><a href="#cb17-1463" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Domain <span class="pp">|</span> Inferred Biological Identity <span class="pp">|</span> Top 3 Cell Types (Proportion) <span class="pp">|</span></span>
<span id="cb17-1464"><a href="#cb17-1464" aria-hidden="true" tabindex="-1"></a><span class="pp">| :---</span> <span class="pp">| :---</span> <span class="pp">| :---</span> <span class="pp">|</span></span>
<span id="cb17-1465"><a href="#cb17-1465" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **D1000** | **Normal Mucosa** <span class="pp">|</span> 1. epithelial (53.4%), 2. plasma_IgA (17.2%), 3. mix_NK_Lymphoid (6.8%) <span class="pp">|</span></span>
<span id="cb17-1466"><a href="#cb17-1466" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **D1003** | **Muscularis Layer** <span class="pp">|</span> 1. smc (65.3%), 2. fibroblast (7.7%), 3. macrophage (6.2%) <span class="pp">|</span></span>
<span id="cb17-1467"><a href="#cb17-1467" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **D1011** | **Invasive Margin (Immune-Rich)** <span class="pp">|</span> 1. epithelial (20.4%), 2. tumor_cyling (18.5%), 3. bcell (12.0%) <span class="pp">|</span></span>
<span id="cb17-1468"><a href="#cb17-1468" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **D1015** | **Desmoplastic Stroma** <span class="pp">|</span> 1. CAF (34.7%), 2. tumor_cyling (11.6%), 3. monocyte (8.3%) <span class="pp">|</span></span>
<span id="cb17-1469"><a href="#cb17-1469" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **D1016** | **Tumor Core** <span class="pp">|</span> 1. tumor_cyling (31.8%), 2. CAF (22.2%), 3. tumor (15.4%) <span class="pp">|</span></span>
<span id="cb17-1470"><a href="#cb17-1470" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **D1017** | **Heterogeneous** <span class="pp">|</span> 1. tumor_cyling (16.1%), 2. undetermined (12.2%), 3. epithelial (12.1%) <span class="pp">|</span></span>
<span id="cb17-1471"><a href="#cb17-1471" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1472"><a href="#cb17-1472" aria-hidden="true" tabindex="-1"></a>: Functional names associated with each <span class="in">`novae`</span> domain. {#tbl-domains-definitions}</span>
<span id="cb17-1473"><a href="#cb17-1473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1474"><a href="#cb17-1474" aria-hidden="true" tabindex="-1"></a>Let's add a new column in the metadata with these values.</span>
<span id="cb17-1475"><a href="#cb17-1475" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1478"><a href="#cb17-1478" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb17-1479"><a href="#cb17-1479" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: I-3</span></span>
<span id="cb17-1480"><a href="#cb17-1480" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-1481"><a href="#cb17-1481" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb17-1482"><a href="#cb17-1482" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-1483"><a href="#cb17-1483" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-1484"><a href="#cb17-1484" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-1485"><a href="#cb17-1485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1486"><a href="#cb17-1486" aria-hidden="true" tabindex="-1"></a>domain_map <span class="op">=</span> {</span>
<span id="cb17-1487"><a href="#cb17-1487" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1000'</span>: <span class="st">'Normal Mucosa'</span>,</span>
<span id="cb17-1488"><a href="#cb17-1488" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1003'</span>: <span class="st">'Muscularis Layer'</span>,</span>
<span id="cb17-1489"><a href="#cb17-1489" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1011'</span>: <span class="st">'Invasive Margin'</span>,</span>
<span id="cb17-1490"><a href="#cb17-1490" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1015'</span>: <span class="st">'Desmoplastic Stroma'</span>,</span>
<span id="cb17-1491"><a href="#cb17-1491" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1016'</span>: <span class="st">'Tumor Core'</span>,</span>
<span id="cb17-1492"><a href="#cb17-1492" aria-hidden="true" tabindex="-1"></a>    <span class="st">'D1017'</span>: <span class="st">'Heterogeneous'</span></span>
<span id="cb17-1493"><a href="#cb17-1493" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-1494"><a href="#cb17-1494" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1495"><a href="#cb17-1495" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'annotated_domain'</span>] <span class="op">=</span> adata.obs[focal_domain_resolution].<span class="bu">map</span>(domain_map)</span>
<span id="cb17-1496"><a href="#cb17-1496" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1497"><a href="#cb17-1497" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-7-domain_annotations.h5ad"</span>)</span>
<span id="cb17-1498"><a href="#cb17-1498" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(</span>
<span id="cb17-1499"><a href="#cb17-1499" aria-hidden="true" tabindex="-1"></a>  filename,</span>
<span id="cb17-1500"><a href="#cb17-1500" aria-hidden="true" tabindex="-1"></a>  compression<span class="op">=</span>hdf5plugin.FILTERS[<span class="st">"zstd"</span>],</span>
<span id="cb17-1501"><a href="#cb17-1501" aria-hidden="true" tabindex="-1"></a>  compression_opts<span class="op">=</span>hdf5plugin.Zstd(clevel<span class="op">=</span><span class="dv">5</span>).filter_options</span>
<span id="cb17-1502"><a href="#cb17-1502" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-1503"><a href="#cb17-1503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1504"><a href="#cb17-1504" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1505"><a href="#cb17-1505" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1506"><a href="#cb17-1506" aria-hidden="true" tabindex="-1"></a>And while were at it, let's plot the domains in XY space using <span class="in">`plotDots`</span> with </span>
<span id="cb17-1507"><a href="#cb17-1507" aria-hidden="true" tabindex="-1"></a>the new labels.</span>
<span id="cb17-1508"><a href="#cb17-1508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1511"><a href="#cb17-1511" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-1512"><a href="#cb17-1512" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: I-4</span></span>
<span id="cb17-1513"><a href="#cb17-1513" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-1514"><a href="#cb17-1514" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb17-1515"><a href="#cb17-1515" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb17-1516"><a href="#cb17-1516" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-1517"><a href="#cb17-1517" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb17-1518"><a href="#cb17-1518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1519"><a href="#cb17-1519" aria-hidden="true" tabindex="-1"></a>domain_colors <span class="ot">&lt;-</span> results_list[[<span class="st">'novae_model_1_domain_colors'</span>]]</span>
<span id="cb17-1520"><a href="#cb17-1520" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(domain_colors) <span class="ot">&lt;-</span> results_list[[<span class="st">'novae_model_1_domain_names'</span>]]</span>
<span id="cb17-1521"><a href="#cb17-1521" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-1522"><a href="#cb17-1522" aria-hidden="true" tabindex="-1"></a>domain_map <span class="ot">&lt;-</span> py<span class="sc">$</span>domain_map</span>
<span id="cb17-1523"><a href="#cb17-1523" aria-hidden="true" tabindex="-1"></a>domain_map <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">'domain_id'</span> <span class="ot">=</span> <span class="fu">names</span>(domain_map), </span>
<span id="cb17-1524"><a href="#cb17-1524" aria-hidden="true" tabindex="-1"></a>                         <span class="st">'domain_description'</span><span class="ot">=</span><span class="fu">unlist</span>(domain_map))</span>
<span id="cb17-1525"><a href="#cb17-1525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1526"><a href="#cb17-1526" aria-hidden="true" tabindex="-1"></a>domain_colors <span class="ot">&lt;-</span> domain_colors[<span class="fu">match</span>(domain_map<span class="sc">$</span>domain_id, <span class="fu">names</span>(domain_colors))]</span>
<span id="cb17-1527"><a href="#cb17-1527" aria-hidden="true" tabindex="-1"></a>domain_map<span class="sc">$</span>domain_color <span class="ot">&lt;-</span> domain_colors</span>
<span id="cb17-1528"><a href="#cb17-1528" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1529"><a href="#cb17-1529" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'domain_colors'</span>]] <span class="ot">&lt;-</span> domain_map</span>
<span id="cb17-1530"><a href="#cb17-1530" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb17-1531"><a href="#cb17-1531" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1532"><a href="#cb17-1532" aria-hidden="true" tabindex="-1"></a>colors_use <span class="ot">&lt;-</span> domain_map<span class="sc">$</span>domain_color</span>
<span id="cb17-1533"><a href="#cb17-1533" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(colors_use) <span class="ot">&lt;-</span> domain_map<span class="sc">$</span>domain_description</span>
<span id="cb17-1534"><a href="#cb17-1534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1535"><a href="#cb17-1535" aria-hidden="true" tabindex="-1"></a>group_prefix <span class="ot">&lt;-</span> <span class="st">"annotated_domain"</span></span>
<span id="cb17-1536"><a href="#cb17-1536" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1537"><a href="#cb17-1537" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(py<span class="sc">$</span>adata, <span class="at">color_by=</span><span class="st">'annotated_domain'</span>,</span>
<span id="cb17-1538"><a href="#cb17-1538" aria-hidden="true" tabindex="-1"></a>              <span class="at">plot_global =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-1539"><a href="#cb17-1539" aria-hidden="true" tabindex="-1"></a>              <span class="at">facet_by_group =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-1540"><a href="#cb17-1540" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_plot_parameters =</span> <span class="fu">list</span>(</span>
<span id="cb17-1541"><a href="#cb17-1541" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom_point_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-1542"><a href="#cb17-1542" aria-hidden="true" tabindex="-1"></a>                    <span class="at">size=</span><span class="fl">0.001</span></span>
<span id="cb17-1543"><a href="#cb17-1543" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-1544"><a href="#cb17-1544" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_bar_params =</span> <span class="fu">list</span>(</span>
<span id="cb17-1545"><a href="#cb17-1545" aria-hidden="true" tabindex="-1"></a>                    <span class="at">location =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">0</span>),</span>
<span id="cb17-1546"><a href="#cb17-1546" aria-hidden="true" tabindex="-1"></a>                    <span class="at">width =</span> <span class="dv">2</span>,</span>
<span id="cb17-1547"><a href="#cb17-1547" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n =</span> <span class="dv">3</span>,</span>
<span id="cb17-1548"><a href="#cb17-1548" aria-hidden="true" tabindex="-1"></a>                    <span class="at">height =</span> <span class="fl">0.1</span>,</span>
<span id="cb17-1549"><a href="#cb17-1549" aria-hidden="true" tabindex="-1"></a>                    <span class="at">scale_colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"grey30"</span>),</span>
<span id="cb17-1550"><a href="#cb17-1550" aria-hidden="true" tabindex="-1"></a>                    <span class="at">label_nudge_y =</span> <span class="sc">-</span><span class="fl">0.3</span></span>
<span id="cb17-1551"><a href="#cb17-1551" aria-hidden="true" tabindex="-1"></a>                  ),</span>
<span id="cb17-1552"><a href="#cb17-1552" aria-hidden="true" tabindex="-1"></a>                  <span class="at">directory =</span> ct_assets_dir,</span>
<span id="cb17-1553"><a href="#cb17-1553" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fileType =</span> <span class="st">"png"</span>,</span>
<span id="cb17-1554"><a href="#cb17-1554" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dpi =</span> <span class="dv">200</span>,</span>
<span id="cb17-1555"><a href="#cb17-1555" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="dv">8</span>, </span>
<span id="cb17-1556"><a href="#cb17-1556" aria-hidden="true" tabindex="-1"></a>                  <span class="at">height =</span> <span class="dv">8</span>,</span>
<span id="cb17-1557"><a href="#cb17-1557" aria-hidden="true" tabindex="-1"></a>                  <span class="at">prefix=</span>group_prefix</span>
<span id="cb17-1558"><a href="#cb17-1558" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb17-1559"><a href="#cb17-1559" aria-hidden="true" tabindex="-1"></a>              <span class="at">additional_ggplot_layers =</span> <span class="fu">list</span>(</span>
<span id="cb17-1560"><a href="#cb17-1560" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>(),</span>
<span id="cb17-1561"><a href="#cb17-1561" aria-hidden="true" tabindex="-1"></a>                <span class="fu">xlab</span>(<span class="st">"X (mm)"</span>),</span>
<span id="cb17-1562"><a href="#cb17-1562" aria-hidden="true" tabindex="-1"></a>                <span class="fu">ylab</span>(<span class="st">"Y (mm)"</span>), </span>
<span id="cb17-1563"><a href="#cb17-1563" aria-hidden="true" tabindex="-1"></a>                <span class="fu">coord_fixed</span>(),</span>
<span id="cb17-1564"><a href="#cb17-1564" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colors_use),</span>
<span id="cb17-1565"><a href="#cb17-1565" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">0.4</span>)),</span>
<span id="cb17-1566"><a href="#cb17-1566" aria-hidden="true" tabindex="-1"></a>                <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="cb17-1567"><a href="#cb17-1567" aria-hidden="true" tabindex="-1"></a>                  <span class="at">title=</span><span class="st">"Spatial Domain"</span>,</span>
<span id="cb17-1568"><a href="#cb17-1568" aria-hidden="true" tabindex="-1"></a>                  <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>) ) )</span>
<span id="cb17-1569"><a href="#cb17-1569" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-1570"><a href="#cb17-1570" aria-hidden="true" tabindex="-1"></a>              )</span>
<span id="cb17-1571"><a href="#cb17-1571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1572"><a href="#cb17-1572" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1573"><a href="#cb17-1573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1576"><a href="#cb17-1576" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-1577"><a href="#cb17-1577" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-xy-domains</span></span>
<span id="cb17-1578"><a href="#cb17-1578" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-1579"><a href="#cb17-1579" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb17-1580"><a href="#cb17-1580" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-1581"><a href="#cb17-1581" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 6</span></span>
<span id="cb17-1582"><a href="#cb17-1582" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 5</span></span>
<span id="cb17-1583"><a href="#cb17-1583" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "XY with spatial domains."</span></span>
<span id="cb17-1584"><a href="#cb17-1584" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-1585"><a href="#cb17-1585" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1586"><a href="#cb17-1586" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"ct"</span>, <span class="st">"annotated_domain__spatial__S0__plot.png"</span>))</span>
<span id="cb17-1587"><a href="#cb17-1587" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1588"><a href="#cb17-1588" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-1589"><a href="#cb17-1589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1590"><a href="#cb17-1590" aria-hidden="true" tabindex="-1"></a><span class="fu">## Conclusion</span></span>
<span id="cb17-1591"><a href="#cb17-1591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1592"><a href="#cb17-1592" aria-hidden="true" tabindex="-1"></a>We have now mapped the cellular landscape of our tissue. By combining broad</span>
<span id="cb17-1593"><a href="#cb17-1593" aria-hidden="true" tabindex="-1"></a>unsupervised clustering  and the specialized</span>
<span id="cb17-1594"><a href="#cb17-1594" aria-hidden="true" tabindex="-1"></a>HieraType algorithm, we have moved from a matrix of counts to a spatially</span>
<span id="cb17-1595"><a href="#cb17-1595" aria-hidden="true" tabindex="-1"></a>annotated map. Then we used this information to create functional names to </span>
<span id="cb17-1596"><a href="#cb17-1596" aria-hidden="true" tabindex="-1"></a>the domains that we found in @sec-domains.</span>
<span id="cb17-1597"><a href="#cb17-1597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1598"><a href="#cb17-1598" aria-hidden="true" tabindex="-1"></a>It is important to recognize that this cell typing workflow is a foundational</span>
<span id="cb17-1599"><a href="#cb17-1599" aria-hidden="true" tabindex="-1"></a>draft. Depending on your specific research questions, such as defining specialized</span>
<span id="cb17-1600"><a href="#cb17-1600" aria-hidden="true" tabindex="-1"></a>epithelial cell types or mapping complex stromal niches, more advanced methods may be</span>
<span id="cb17-1601"><a href="#cb17-1601" aria-hidden="true" tabindex="-1"></a>required.</span>
<span id="cb17-1602"><a href="#cb17-1602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-1603"><a href="#cb17-1603" aria-hidden="true" tabindex="-1"></a>With our cells now labeled with a type and a domain, we are ready to ask</span>
<span id="cb17-1604"><a href="#cb17-1604" aria-hidden="true" tabindex="-1"></a>"tertiary" questions. At the moment, this includes an analysis on pathway </span>
<span id="cb17-1605"><a href="#cb17-1605" aria-hidden="true" tabindex="-1"></a>enrichment but more analysis chapters are in the works.</span>
</code></pre></div><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">

<div class="cookie-consent-footer"><a href="#" id="open_preferences_center">Cookie Preferences</a></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>
<script src="https://unpkg.com/@popperjs/core@2/dist/umd/popper.min.js"></script>
<script src="https://unpkg.com/tippy.js@6/dist/tippy-bundle.umd.min.js"></script>

<script>
window.addEventListener('load', function() {
  // Initialize Tippy.js ONLY on links with the .link-preview class
  tippy('.link-preview', {
    content(reference) {
      const iframe = document.createElement('iframe');
      iframe.src = reference.href;
      iframe.width = "400px";
      iframe.height = "300px";
      iframe.style.border = "1px solid #ddd";
      iframe.style.borderRadius = "4px";
      return iframe;
    },
    allowHTML: true,
    interactive: true,
    placement: 'auto',
    arrow: true,
    animation: 'fade',
    delay: [300, 200],
    maxWidth: 420
  });
});
</script>




</body></html>