<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Felicia New">
<meta name="author" content="Claire Williams">
<meta name="dcterms.date" content="2026-02-13">
<meta name="description" content="Vignette showing an example data analysis of high-plex GeoMx® DSP protein data">

<title>Data Analysis Workflow for GeoMx® NGS-Protein Data – Blog</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/logo_Bruker_black.png" rel="icon" type="image/png">
<script src="../../site_libs/cookie-consent/cookie-consent.js"></script>
<link href="../../site_libs/cookie-consent/cookie-consent.css" rel="stylesheet">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-3a01e2046221230fdceeea94b1ec5d67.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-a4a11d514c7d463668e07712114998e6.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-c1091440317a9281743ded74104a8a94.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-29W4MW0Y2W"></script>

<script type="text/plain" cookie-consent="tracking">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-29W4MW0Y2W', { 'anonymize_ip': true});
</script>

<script type="text/javascript" charset="UTF-8">
document.addEventListener('DOMContentLoaded', function () {
cookieconsent.run({
  "notice_banner_type":"simple",
  "consent_type":"implied",
  "palette":"light",
  "language":"en",
  "page_load_consent_levels":["strictly-necessary","functionality","tracking","targeting"],
  "notice_banner_reject_button_hide":false,
  "preferences_center_close_button_hide":false,
  "website_name":""
  ,
"language":"en"
  });
});
</script> 
  
<script src="../../site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="../../site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="../../site_libs/datatables-binding-0.34.0/datatables.js"></script>
<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="../../site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="../../site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="../../site_libs/dt-ext-buttons-1.13.6/css/buttons.dataTables.min.css" rel="stylesheet">
<script src="../../site_libs/dt-ext-buttons-1.13.6/js/dataTables.buttons.min.js"></script>
<script src="../../site_libs/dt-ext-buttons-1.13.6/js/buttons.html5.min.js"></script>
<script src="../../site_libs/dt-ext-buttons-1.13.6/js/buttons.colVis.min.js"></script>
<script src="../../site_libs/dt-ext-buttons-1.13.6/js/buttons.print.min.js"></script>
<link href="../../site_libs/dt-ext-scroller-1.13.6/css/scroller.dataTables.min.css" rel="stylesheet">
<script src="../../site_libs/dt-ext-scroller-1.13.6/js/dataTables.scroller.min.js"></script>
<link href="../../site_libs/dt-ext-fixedcolumns-1.13.6/css/fixedColumns.dataTables.min.css" rel="stylesheet">
<script src="../../site_libs/dt-ext-fixedcolumns-1.13.6/js/dataTables.fixedColumns.min.js"></script>
<link href="../../site_libs/crosstalk-1.2.2/css/crosstalk.min.css" rel="stylesheet">
<script src="../../site_libs/crosstalk-1.2.2/js/crosstalk.min.js"></script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../.././assets/logo_Bruker_white.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">Topics</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../guides/index.html"> 
<span class="menu-text">Guides (new!)</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../bbytes/index.html"> 
<span class="menu-text">Browser Bytes (new!)</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../link-to-code.html"> 
<span class="menu-text">Source Code</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/company/bruker-spatial"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Data Analysis Workflow for GeoMx® NGS-Protein Data</h1>
                  <div>
        <div class="description">
          Vignette showing an example data analysis of high-plex GeoMx® DSP protein data
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">GeoMx</div>
                <div class="quarto-category">how-tos</div>
                <div class="quarto-category">overview</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-heading">Affiliations</div>
    
      <div class="quarto-title-meta-contents">
      <p class="author">Felicia New <a href="https://orcid.org/0000-0001-6213-4731" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              Bruker Spatial Biology
            </p>
          <p class="affiliation">
              Github: <a href="https://github.com/fnew" target="_blank">fnew</a>
            </p>
        </div>
      <div class="quarto-title-meta-contents">
      <p class="author">Claire Williams <a href="https://orcid.org/0000-0001-5467-149X" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              Bruker Spatial Biology
            </p>
          <p class="affiliation">
              Github: <a href="https://github.com/crwilliams11" target="_blank">crwilliams11</a>
            </p>
        </div>
    </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">February 13, 2026</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">February 13, 2026</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1">
    <h2 id="toc-title">Contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#sec-study-design" id="toc-sec-study-design" class="nav-link" data-scroll-target="#sec-study-design"><span class="header-section-number">2</span> Study Design</a>
  <ul class="collapse">
  <li><a href="#slides-and-profiling-strategy" id="toc-slides-and-profiling-strategy" class="nav-link" data-scroll-target="#slides-and-profiling-strategy"><span class="header-section-number">2.1</span> Slides and Profiling Strategy</a></li>
  </ul></li>
  <li><a href="#sec-quality-control" id="toc-sec-quality-control" class="nav-link" data-scroll-target="#sec-quality-control"><span class="header-section-number">3</span> Quality Control</a>
  <ul class="collapse">
  <li><a href="#segment-qc" id="toc-segment-qc" class="nav-link" data-scroll-target="#segment-qc"><span class="header-section-number">3.1</span> Segment QC</a>
  <ul class="collapse">
  <li><a href="#segment-qc-plots" id="toc-segment-qc-plots" class="nav-link" data-scroll-target="#segment-qc-plots"><span class="header-section-number">3.1.1</span> Segment QC Plots</a></li>
  </ul></li>
  <li><a href="#quality-control-of-housekeeper-proteins-and-background-isotypes" id="toc-quality-control-of-housekeeper-proteins-and-background-isotypes" class="nav-link" data-scroll-target="#quality-control-of-housekeeper-proteins-and-background-isotypes"><span class="header-section-number">3.2</span> Quality control of housekeeper proteins and background (isotypes)</a>
  <ul class="collapse">
  <li><a href="#isotype-controls" id="toc-isotype-controls" class="nav-link" data-scroll-target="#isotype-controls"><span class="header-section-number">3.2.1</span> Isotype Controls</a></li>
  <li><a href="#housekeeping-controls" id="toc-housekeeping-controls" class="nav-link" data-scroll-target="#housekeeping-controls"><span class="header-section-number">3.2.2</span> Housekeeping Controls</a></li>
  <li><a href="#isotype-vs-housekeeper-controls" id="toc-isotype-vs-housekeeper-controls" class="nav-link" data-scroll-target="#isotype-vs-housekeeper-controls"><span class="header-section-number">3.2.3</span> Isotype vs housekeeper controls</a></li>
  </ul></li>
  <li><a href="#signal-to-noise-ratio-snr" id="toc-signal-to-noise-ratio-snr" class="nav-link" data-scroll-target="#signal-to-noise-ratio-snr"><span class="header-section-number">3.3</span> Signal to noise ratio, SNR</a></li>
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization"><span class="header-section-number">3.4</span> Normalization</a></li>
  </ul></li>
  <li><a href="#sec-eda" id="toc-sec-eda" class="nav-link" data-scroll-target="#sec-eda"><span class="header-section-number">4</span> Exploratory Data Analysis</a>
  <ul class="collapse">
  <li><a href="#dimensional-reduction" id="toc-dimensional-reduction" class="nav-link" data-scroll-target="#dimensional-reduction"><span class="header-section-number">4.1</span> Dimensional Reduction</a></li>
  <li><a href="#clustering-features-with-high-coefficient-of-variation" id="toc-clustering-features-with-high-coefficient-of-variation" class="nav-link" data-scroll-target="#clustering-features-with-high-coefficient-of-variation"><span class="header-section-number">4.2</span> Clustering features with high coefficient of variation</a></li>
  </ul></li>
  <li><a href="#sec-de" id="toc-sec-de" class="nav-link" data-scroll-target="#sec-de"><span class="header-section-number">5</span> Differential Expression</a>
  <ul class="collapse">
  <li><a href="#contrasting-groups-with-linear-mixed-models" id="toc-contrasting-groups-with-linear-mixed-models" class="nav-link" data-scroll-target="#contrasting-groups-with-linear-mixed-models"><span class="header-section-number">5.1</span> Contrasting groups with linear mixed models</a></li>
  <li><a href="#ck--responder-vs-non-responder" id="toc-ck--responder-vs-non-responder" class="nav-link" data-scroll-target="#ck--responder-vs-non-responder"><span class="header-section-number">5.2</span> CK- Responder vs Non-responder</a></li>
  </ul></li>
  <li><a href="#sec-conclusion" id="toc-sec-conclusion" class="nav-link" data-scroll-target="#sec-conclusion"><span class="header-section-number">6</span> Conclusions</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Introduction</h1>
<p>In this vignette, we will introduce a data analysis workflow for GeoMx® NGS protein expression data profiled with the <a href="https://nanostring.com/products/geomx-digital-spatial-profiler/geomx-protein-assays/io-proteome-atlas/" target="_blank">GeoMx IO Proteome Atlas (IPA) atlas</a>. The workflow introduced in this vignette can be applied to the new <a href="https://nanostring.com/products/geomx-digital-spatial-profiler/geomx-protein-assays/discovery-proteome-atlas/" target="_blank">Discovery Proteome Atlas (DPA) panel</a>, as well.</p>
<p>The GeoMx Digital Spatial Profiler (DSP) is a platform for capturing spatially resolved high-plex gene or protein expression data from tissue (<span class="citation" data-cites="Merritt2020">Merritt et al. (<a href="#ref-Merritt2020" role="doc-biblioref">2020</a>)</span>). In particular, formalin-fixed paraffin-embedded (FFPE) or fresh-frozen (FF) tissue sections are stained with barcoded in-situ hybridization probes that bind to endogenous mRNA transcripts or barcoded antibodies that recognize endogenous proteins. The user then selects regions of the interest (ROI) to profile; if desired, each ROI segment can be further sub-divided into areas of illumination (AOI) based on tissue morphology. The GeoMx DSP then photo-cleaves and collects expression barcodes for each AOI segment separately for downstream sequencing and data processing.</p>
<p>The final results are spatially resolved unique expression datasets for every protein-coding gene (&gt;18,000 genes) or over 570 proteins from every individual segment profiled from tissue. Here, we walk through an analysis workflow with one example dataset.</p>
<p>This vignette is divided into the following major sections:</p>
<ul>
<li><a href="#sec-study-design" class="quarto-xref">Section&nbsp;2</a> Study Design</li>
<li><a href="#sec-quality-control" class="quarto-xref">Section&nbsp;3</a> Quality Control</li>
<li><a href="#sec-eda" class="quarto-xref">Section&nbsp;4</a> Exploratory Data Analysis</li>
<li><a href="#sec-de" class="quarto-xref">Section&nbsp;5</a> Differential Expression</li>
<li><a href="#sec-conclusion" class="quarto-xref">Section&nbsp;6</a> Conclusions</li>
</ul>
<div class="cell">
<style type="text/css">
.hscroll-plot {
  overflow-x: auto;
  overflow-y: hidden;
  max-width: 100%;
  padding-bottom: 0.25rem;
}
.hscroll-plot img, .hscroll-plot svg, .hscroll-plot canvas {
  display: block;
  max-width: none !important;
}
</style>
</div>
<p>Before getting started, we load in key libraries and set locations of folders we’ll use for intermediate and final analyses and results.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># List of CRAN packages for session</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>.cran_packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"plyr"</span>, <span class="st">"dplyr"</span>, <span class="st">"FactoMineR"</span>, <span class="st">"ggplot2"</span>, <span class="st">"ggforce"</span>, </span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"openxlsx"</span>, <span class="st">"psych"</span>, <span class="st">"kableExtra"</span>, <span class="st">"GGally"</span>, <span class="st">"magick"</span>, </span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"ggrepel"</span>, <span class="st">"circlize"</span>, <span class="st">"pals"</span>, <span class="st">"ggupset"</span>, <span class="st">"tidyr"</span>, </span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"reshape2"</span>, <span class="st">"ggprism"</span>, <span class="st">"gtools"</span>, <span class="st">"parcats"</span>, <span class="st">"svglite"</span>, </span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vdiffr"</span>, <span class="st">"parallel"</span>,</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"easyalluvial"</span>, <span class="st">"ggtext"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co"># List of Bioconductor packages for session</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>.bioc_packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GSEABase"</span>, <span class="st">"GSVA"</span>, <span class="st">"Biobase"</span>, <span class="st">"NanoStringNCTools"</span>, </span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"GeomxTools"</span>, <span class="st">"GeoMxWorkflows"</span>, <span class="st">"SpatialDecon"</span>, </span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"ComplexHeatmap"</span>, <span class="st">"org.Hs.eg.db"</span>, <span class="st">"org.Mm.eg.db"</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="do">## Install packages if any are missing</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Install CRAN packages (if not already installed)</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>.inst <span class="ot">&lt;-</span> .cran_packages <span class="sc">%in%</span> <span class="fu">installed.packages</span>()</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(.cran_packages[<span class="sc">!</span>.inst]) <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="fu">install.packages</span>(.cran_packages[<span class="sc">!</span>.inst])</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Install Bioconductor packages (if not already installed)</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>.bioc_inst <span class="ot">&lt;-</span> .bioc_packages <span class="sc">%in%</span> <span class="fu">installed.packages</span>()</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(.bioc_packages[<span class="sc">!</span>.bioc_inst]) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>  BiocManager<span class="sc">::</span><span class="fu">install</span>(.bioc_packages[<span class="sc">!</span>.bioc_inst])</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="co"># require these packages</span></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>cran_loaded <span class="ot">&lt;-</span> <span class="fu">lapply</span>(.cran_packages, require, <span class="at">character.only=</span><span class="cn">TRUE</span>)</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>bioc_loaded <span class="ot">&lt;-</span> <span class="fu">lapply</span>(.bioc_packages, require, <span class="at">character.only=</span><span class="cn">TRUE</span>)</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Load in useful functions for GeoMx Data Analysis</span></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"https://raw.githubusercontent.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/Main/_code/geomx_vignette/GeoMx_Workflow_Helpers.R"</span>)</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a><span class="do">## Set up directory structure</span></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a><span class="co"># datadir is the location of dcc files, pkc file(s), and images (if applicable).</span></span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a>datadir <span class="ot">&lt;-</span> <span class="st">"/path/to/data/dir/"</span> <span class="co"># Set data directory</span></span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Initiate a date tag &amp; directory naming</span></span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>project_name <span class="ot">&lt;-</span> <span class="st">"Demo"</span> <span class="co"># Set a name for the project</span></span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a>date_tag <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">format</span>(<span class="fu">Sys.Date</span>(), <span class="st">"%Y%m%d"</span>), <span class="st">"_"</span>, project_name)</span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>outdir <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"output_"</span>, date_tag)</span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a>object_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(outdir, <span class="st">"R_objects"</span>)</span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>qc_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(outdir, <span class="st">"qc"</span>)</span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a>de_plot_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(outdir, <span class="st">"de"</span>, <span class="st">"Plots"</span>)</span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a>de_data_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(outdir, <span class="st">"de"</span>, <span class="st">"Data"</span>)</span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize output directory for static images, tables, and serialized R objects.</span></span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(dir_i <span class="cf">in</span> <span class="fu">c</span>(outdir, object_dir, qc_dir, de_plot_dir, de_data_dir)){</span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(dir_i)){</span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(dir_i, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Select your p_adjust method. Note: this value must be one of the following: "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none"</span></span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a>p_adjust_method_selected <span class="ot">&lt;-</span> <span class="st">"fdr"</span></span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a>p_adjust_threshold <span class="ot">&lt;-</span> <span class="fl">0.05</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-study-design" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Study Design</h1>
<p>Here we describe the experimental setup and the raw data that will form the basis of our analysis.</p>
<details>
<summary>
Click to show/hide a description of the study design
</summary>
<section id="slides-and-profiling-strategy" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="slides-and-profiling-strategy"><span class="header-section-number">2.1</span> Slides and Profiling Strategy</h2>
<p>In this vignette, we will analyze a GeoMx DSP breast cancer dataset created with the IPA assay. The dataset of interest is comprised of six breast cancer samples from six patients who received a therapeutic treatment, with 3 responders and 3 non-responders. Regions of interest (ROIs) were further segmented into PanCK+ and PanCK- areas of illumination (AOIs), separating epithelial and tumor compartments from stromal compartments.</p>
<p>To follow along with this vignette, the data set used here can be downloaded from this <a href="https://nanostring.box.com/s/o15x51acki3096y52li2jcxrw3r855wk" target="_blank">Box link</a>. After downloading, unzip the file and place the outputs (key data files) in your <code>datadir</code>, as specified above.</p>
<p>The key data files are:</p>
<ul>
<li>DCCs files - expression count data and sequencing quality metadata</li>
<li>PKCs file(s) - probe assay metadata describing the protein targets present in the data. PKC files can be found <a href="https://nanostring.com/products/geomx-digital-spatial-profiler/geomx-dsp-configuration-files/" target="_blank">here</a>.</li>
<li>Annotation file - This should be the Lab Worksheet from the GeoMx instrument study readout package. Users should add any useful tissue information, including the type of segment profiled (ex: responder vs.&nbsp;non-responder), segment area/nuclei count, and other tissue characteristics when applicable (ex: diseased vs.&nbsp;healthy). Please note, the row order of NTCs is important to ensure proper processing of files. To ensure the proper order, the file should be sorted by DSP file name. The Lab Worksheet needs to be formatted for ingestion with GeoMxTools. First, the lab worksheet should be opened in Excel, and the first ~15 rows should be deleted so that only the header row and sample information remain.</li>
</ul>
<p>We first locate the downloaded files and then we load the data to create a data object using the <code>readNanoStringGeoMxSet</code> function from the <code>GeomxTools</code> package. For more information about this data object, see this <a href="https://www.bioconductor.org/packages/release/workflows/vignettes/GeoMxWorkflows/inst/doc/GeomxTools_RNA-NGS_Analysis.html#91_GeoMxSet_Object_Overview" target="_blank">GeoMxSetObject Overview</a>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./Assets/images/GeomxSetStructure.png" class="img-fluid figure-img"></p>
<figcaption><em>GeoMxSetObject structure</em>.</figcaption>
</figure>
</div>
<div class="cell" data-show="false">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># If the object probe_data already exists in object_dir, </span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># load it from disk. Otherwise, run readNanoStringGeoMxSet</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">file.exists</span>(<span class="fu">file.path</span>(object_dir, <span class="st">"probe_data.RDS"</span>))){</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  probe_data <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(object_dir, <span class="st">"probe_data.RDS"</span>))</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a> DCCFiles <span class="ot">&lt;-</span> <span class="fu">dir</span>(<span class="fu">file.path</span>(datadir, <span class="st">"dccs"</span>), <span class="at">pattern =</span> <span class="st">".dcc$"</span>,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">full.names =</span> <span class="cn">TRUE</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a> PKCFiles <span class="ot">&lt;-</span> <span class="fu">dir</span>(<span class="fu">file.path</span>(datadir, <span class="st">"pkcs"</span>), <span class="at">pattern =</span> <span class="st">".pkc$"</span>,</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">full.names =</span> <span class="cn">TRUE</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a> SampleAnnotationFile <span class="ot">&lt;-</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir</span>(<span class="fu">file.path</span>(datadir), <span class="at">pattern =</span> <span class="st">"demo.xlsx$"</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">full.names =</span> <span class="cn">TRUE</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  probe_data <span class="ot">&lt;-</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">readNanoStringGeoMxSet</span>(<span class="at">dccFiles =</span> DCCFiles,</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>                           <span class="at">pkcFiles =</span> PKCFiles,</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>                           <span class="at">phenoDataFile =</span> SampleAnnotationFile,</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>                           <span class="at">phenoDataSheet =</span> <span class="st">"Sheet 1"</span>,</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>                           <span class="at">phenoDataDccColName =</span> <span class="st">"Sample_ID"</span>,</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>                           <span class="at">protocolDataColNames =</span> <span class="fu">c</span>(<span class="st">"aoi"</span>, <span class="st">"roi"</span>),</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>                           <span class="at">experimentDataColNames =</span> <span class="fu">c</span>(<span class="st">"panel"</span>),</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>                           <span class="at">analyte =</span> <span class="st">'Protein'</span>)</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(<span class="at">object=</span>probe_data, <span class="at">file=</span><span class="fu">file.path</span>(object_dir, <span class="st">"probe_data.RDS"</span>))</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="st">"slide name"</span> <span class="sc">%in%</span> <span class="fu">colnames</span>(<span class="fu">pData</span>(probe_data))){</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"The reserved column </span><span class="sc">\'</span><span class="st">slide name</span><span class="sc">\'</span><span class="st"> was not found in this dataset."</span>)</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Select factors of interest from the data's metadata</span></span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>factors_of_interest <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Responder"</span>,<span class="st">"slide"</span>,<span class="st">"Segment"</span>)</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Set colors for main factors of interest</span></span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>pal_main <span class="ot">&lt;-</span> </span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a> <span class="fu">getColorPalette</span>(<span class="fu">pData</span>(probe_data)[, factors_of_interest[<span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(factors_of_interest),<span class="dv">1</span>)]], <span class="at">method =</span> <span class="st">"Main"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>All of the expression, annotation, and probe information are now linked and stored together into a single data object.</p>
<p>We summarize the study design in a Sankey diagram, which illustrates how samples flow through each step and link to their biological annotations. The width of each connecting band reflects how many segments are shared between the two annotations, making it easy to see where groups diverge or overlap.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>pheno_data <span class="ot">&lt;-</span> <span class="fu">pData</span>(probe_data)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Select factor of interest to color sankey based on this factor</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>sankey_focal_factor <span class="ot">&lt;-</span> factors_of_interest[<span class="dv">1</span>]</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset the desired columns</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>sample_overview <span class="ot">&lt;-</span> pheno_data[factors_of_interest]</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Create alluvial plot</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> easyalluvial<span class="sc">::</span><span class="fu">alluvial_wide</span>(sample_overview, </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">max_variables =</span> <span class="dv">15</span>, </span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">col_vector_value =</span> <span class="fu">unlist</span>(pal_main),</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">col_vector_flow =</span> <span class="fu">as.vector</span>(pal_main[[sankey_focal_factor]]))</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="st">"Main_Alluvial.svg"</span>, <span class="at">path =</span> qc_dir, <span class="at">device =</span> <span class="st">"svg"</span>, <span class="at">width=</span><span class="dv">7</span>, <span class="at">height=</span><span class="dv">7</span>)</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="co"># For an interactive plot, use the code below:</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>pc <span class="ot">&lt;-</span> <span class="fu">parcats</span>(p,</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">marginal_histograms =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">data_input =</span> sample_overview,</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">labelfont =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">12</span>, <span class="at">color =</span> <span class="st">"black"</span>),</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">hoveron=</span><span class="st">"dimension"</span>,</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">hoverinfo =</span> <span class="st">"count"</span>,</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">offset_marginal_histograms =</span> <span class="fl">0.7</span>)</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(pc, <span class="at">file =</span> <span class="fu">paste0</span>(qc_dir, <span class="st">"/pc_interactive.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Here we show the static Sankey diagram:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/Main_Alluvial.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section></details>
</section>

<section id="sec-quality-control" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Quality Control</h1>
<p>Before we dive into deeper analysis, we first run through a few key quality measures to assess the dataset.</p>
<details>
<summary>
Click to show/hide quality control steps
</summary>
<p>The QC steps can be summarized as follows:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./Assets/images/GeoMx_Protein_Workflow.png" class="img-fluid figure-img"></p>
<figcaption><em>A standard GeoMx workflow</em>.</figcaption>
</figure>
</div>
<section id="segment-qc" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="segment-qc"><span class="header-section-number">3.1</span> Segment QC</h2>
<p>Set up QC parameters to use for the study. Here we list parameters that are a good starting point for IPA datasets, but as always we recommend examining the distribution of these metrics in your dataset to choose appropriate thresholds.</p>
<p>Every ROI/AOI segment is tested for:</p>
<ul>
<li><p><strong>Raw sequencing reads</strong>: segments with &lt;1000 raw reads are removed.</p></li>
<li><p><strong>% Aligned, % Trimmed, or % Stitched sequencing reads</strong>: segments below ~80% for one or more of these QC parameters are removed.</p></li>
<li><p><strong>% Sequencing saturation</strong>: Defined as ([1-deduplicated reads/aligned reads]%). Segments below ~50% require additional sequencing to capture full sample diversity and are not typically analyzed until improved.</p></li>
<li><p><strong>Minimum Negative Count</strong>: this is the geometric mean of the several unique negative probes in the GeoMx panel that do not target endogenous proteins and establish the background count level per segment; segments with low negative counts (1-10) are not necessarily removed but may be studied closer for low endogenous signal and/or insufficient tissue sampling.</p></li>
<li><p><strong>No Template Control (NTC) count</strong>: values &gt;1000 could indicate contamination for the segments associated with this NTC; however, in cases where the NTC count is between 1000-10000, the segments may be used if the NTC data is uniformly low (e.g.&nbsp;0-2 counts for all probes).</p></li>
<li><p><strong>Nuclei</strong>: &gt;100 nuclei per segment is generally recommended; however, this cutoff is highly study/tissue dependent and may need to be reduced; what is most important is consistency in the nuclei distribution for segments within the study.</p></li>
<li><p><strong>Area</strong>: generally correlates with nuclei; a strict cutoff is not generally applied based on area.</p></li>
</ul>
<p>More information about these parameters can be found in the <a href="http://www.bioconductor.org/packages/release/bioc/html/GeomxTools.html" target="_blank">GeomxTools Bioconductor page</a> and <a href="http://www.bioconductor.org/packages/release/bioc/manuals/GeomxTools/man/GeomxTools.pdf" target="_blank">reference manual</a>.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Parameters used</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>QC_params <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">minSegmentReads =</span> <span class="dv">1000</span>,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">percentTrimmed =</span> <span class="dv">80</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">percentStitched =</span> <span class="dv">80</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">percentAligned =</span> <span class="dv">75</span>,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">percentSaturation =</span> <span class="dv">50</span>,</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">minNegativeCount =</span> <span class="dv">1</span>,</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">maxNTCCount =</span> <span class="dv">9000</span>,</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">minNuclei =</span> <span class="dv">20</span>,</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">minArea =</span> <span class="dv">1000</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>top_quantile <span class="ot">&lt;-</span> <span class="fl">0.90</span> <span class="co"># quantile of features to select for comparing top CV features</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We first will assess sequencing quality and adequate tissue sampling for every ROI/AOI segment. We do this by comparing segment metrics to the QC thresholds set above. Following this QC check, we show plots visualizing each metric followed by a table summarizing the segments that pass the key QC metrics.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>probe_data_qc <span class="ot">&lt;-</span> probe_data</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Save this as a data frame</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>qc_param_tab <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Parameter=</span><span class="fu">names</span>(QC_params), <span class="st">'Default value'</span><span class="ot">=</span><span class="fu">as.numeric</span>(GeomxTools<span class="sc">:::</span>DEFAULTS[<span class="fu">names</span>(QC_params)]), <span class="st">'Actual value'</span><span class="ot">=</span><span class="fu">as.numeric</span>(QC_params))</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>results_list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/results.rds"</span>))</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>results_list<span class="sc">$</span>qc_param_tab <span class="ot">&lt;-</span> qc_param_tab</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, <span class="fu">paste0</span>(outdir, <span class="st">"/results.rds"</span>))</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Assess QC flags using setSegmentQCFlags</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>probe_data_qc <span class="ot">&lt;-</span> <span class="fu">setSegmentQCFlags</span>(probe_data_qc, <span class="at">qcCutoffs =</span> QC_params)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Collate QC Results</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>QCResults <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">protocolData</span>(probe_data_qc)[[<span class="st">"QCFlags"</span>]])</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>flag_columns <span class="ot">&lt;-</span> <span class="fu">colnames</span>(QCResults)</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>QC_Summary <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Pass =</span> <span class="fu">colSums</span>(<span class="sc">!</span>QCResults[, flag_columns]),</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>                         <span class="at">Warning =</span> <span class="fu">colSums</span>(QCResults[, flag_columns]))</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>QCResults<span class="sc">$</span>QCStatus <span class="ot">&lt;-</span> <span class="fu">apply</span>(QCResults, <span class="dv">1</span>L, <span class="cf">function</span>(x) {</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ifelse</span>(<span class="fu">sum</span>(x) <span class="sc">==</span> <span class="dv">0</span>L, <span class="st">"PASS"</span>, <span class="st">"WARNING"</span>)</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>QC_Summary[<span class="st">"TOTAL FLAGS"</span>, ] <span class="ot">&lt;-</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(<span class="fu">sum</span>(QCResults[, <span class="st">"QCStatus"</span>] <span class="sc">==</span> <span class="st">"PASS"</span>),</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>      <span class="fu">sum</span>(QCResults[, <span class="st">"QCStatus"</span>] <span class="sc">==</span> <span class="st">"WARNING"</span>))</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Add QC status back to probe_data</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a><span class="fu">pData</span>(probe_data)<span class="sc">$</span>QC <span class="ot">&lt;-</span> QCResults<span class="sc">$</span>QCStatus</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Save QC Summary</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>results_list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/results.rds"</span>))</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>results_list<span class="sc">$</span>QC_Summary <span class="ot">&lt;-</span> QC_Summary</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, <span class="fu">paste0</span>(outdir, <span class="st">"/results.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Default.value</th>
<th style="text-align: left;">Actual.value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">minSegmentReads</td>
<td style="text-align: left;">1000</td>
<td style="text-align: left;">1000</td>
</tr>
<tr class="even">
<td style="text-align: left;">percentTrimmed</td>
<td style="text-align: left;">80</td>
<td style="text-align: left;">80</td>
</tr>
<tr class="odd">
<td style="text-align: left;">percentStitched</td>
<td style="text-align: left;">80</td>
<td style="text-align: left;">80</td>
</tr>
<tr class="even">
<td style="text-align: left;">percentAligned</td>
<td style="text-align: left;">80</td>
<td style="text-align: left;">75</td>
</tr>
<tr class="odd">
<td style="text-align: left;">percentSaturation</td>
<td style="text-align: left;">50</td>
<td style="text-align: left;">50</td>
</tr>
<tr class="even">
<td style="text-align: left;">minNegativeCount</td>
<td style="text-align: left;">10</td>
<td style="text-align: left;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">maxNTCCount</td>
<td style="text-align: left;">60</td>
<td style="text-align: left;">9000</td>
</tr>
<tr class="even">
<td style="text-align: left;">minNuclei</td>
<td style="text-align: left;">200</td>
<td style="text-align: left;">20</td>
</tr>
<tr class="odd">
<td style="text-align: left;">minArea</td>
<td style="text-align: left;">16000</td>
<td style="text-align: left;">1000</td>
</tr>
</tbody>
</table>
</div>
</div>
<section id="segment-qc-plots" class="level3" data-number="3.1.1">
<h3 data-number="3.1.1" class="anchored" data-anchor-id="segment-qc-plots"><span class="header-section-number">3.1.1</span> Segment QC Plots</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Processing data for QC plots</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Un-embed data frames from the QC data</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (column <span class="cf">in</span> <span class="fu">colnames</span>(<span class="fu">sData</span>(probe_data_qc))) {</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">inherits</span>(<span class="fu">sData</span>(probe_data_qc)[,column], <span class="st">"data.frame"</span>) <span class="sc">&amp;&amp;</span> <span class="fu">ncol</span>(<span class="fu">sData</span>(probe_data_qc)[,column])<span class="sc">==</span><span class="dv">1</span>){</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>        probe_data_qc<span class="sc">@</span>protocolData<span class="sc">@</span>data[,column] <span class="ot">&lt;-</span> <span class="fu">sData</span>(probe_data_qc)[, column][,<span class="dv">1</span>]</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the negative geometric means for each module</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>negativeGeoMeans <span class="ot">&lt;-</span> </span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">esBy</span>(<span class="fu">negativeControlSubset</span>(probe_data_qc), </span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">GROUP =</span> <span class="st">"Module"</span>, </span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">FUN =</span> <span class="cf">function</span>(x) { </span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>             <span class="fu">assayDataApply</span>(x, <span class="at">MARGIN =</span> <span class="dv">2</span>, <span class="at">FUN =</span> ngeoMean, <span class="at">elt =</span> <span class="st">"exprs"</span>) </span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>         }) </span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert embedded matrix to df then to AnnotatedDataFrame</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a><span class="fu">protocolData</span>(probe_data_qc) <span class="ot">&lt;-</span>  <span class="fu">as</span>(</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cbind</span>(</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as</span>(<span class="fu">protocolData</span>(probe_data_qc), <span class="st">"data.frame"</span>),</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>(negativeGeoMeans)</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>  ), <span class="st">"AnnotatedDataFrame"</span>)</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Here are the module names</span></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>pkcs <span class="ot">&lt;-</span> <span class="fu">annotation</span>(probe_data_qc)</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>modules <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">".pkc"</span>, <span class="st">""</span>, pkcs)</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy the Negative geoMeans from sData to pData</span></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>negCols <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"NegGeoMean_"</span>, modules)</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a><span class="co"># negCols &lt;- colnames(sData(probe_data_qc)[["NegGeoMean"]])</span></span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a><span class="fu">pData</span>(probe_data_qc)[, negCols] <span class="ot">&lt;-</span> <span class="fu">sData</span>(probe_data_qc)[, <span class="fu">which</span>(<span class="fu">colnames</span>(<span class="fu">sData</span>(probe_data_qc)) <span class="sc">%in%</span> modules)]</span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a><span class="co"># detatch neg_geomean columns ahead of aggregateCounts call</span></span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a><span class="fu">pData</span>(probe_data_qc) <span class="ot">&lt;-</span> <span class="fu">pData</span>(probe_data_qc)[, <span class="sc">!</span><span class="fu">colnames</span>(<span class="fu">pData</span>(probe_data_qc)) <span class="sc">%in%</span> negCols]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The tabs below visualize the segment QC distributions for various sequence and segment quality metrics.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">Percent trimmed</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">Percent Aligned</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false">Percent Stitched</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-4" role="tab" aria-controls="tabset-1-4" aria-selected="false">Percent Saturation</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-5-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-5" role="tab" aria-controls="tabset-1-5" aria-selected="false">Segment Areas</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-6-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-6" role="tab" aria-controls="tabset-1-6" aria-selected="false">Segment Nuclei</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-7-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-7" role="tab" aria-controls="tabset-1-7" aria-selected="false">Negative probe means</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<p>The plot below shows the distribution of trimmed reads in the data with the percent threshold as set in the QC parameters code block shown as a vertical dotted line. Typically we want to see very few reads getting trimmed (ie high percentage remaining after trimming) as this indicates a high quality sequencing run.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot trimmed reads with helper function makeQCHistograms</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">makeQCHistogram</span>(probe_data_qc, </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">annotation_col =</span> <span class="st">"Trimmed (%)"</span>, </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">fill_by =</span> <span class="st">"Segment"</span>, </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">bins =</span> <span class="dv">50</span>,</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">xintercept =</span> QC_params<span class="sc">$</span>percentTrimmed)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"trimmed.svg"</span>), <span class="at">width=</span><span class="dv">6</span>, <span class="at">height=</span><span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/trimmed.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p>The plot below shows the distribution of aligned reads in the data with the percent threshold of aligned reads as set in the QC parameters code block shown as a vertical dotted line. Typically we look to see high levels of percent alignment as this indicates a high quality sequencing run without contaminating sequences present.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">makeQCHistogram</span>(probe_data_qc, </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">annotation_col =</span> <span class="st">"Aligned (%)"</span>, </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">fill_by =</span> <span class="st">"Segment"</span>, </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">bins =</span> <span class="dv">50</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">xintercept =</span> QC_params<span class="sc">$</span>percentAligned)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"aligned.svg"</span>), <span class="at">width=</span><span class="dv">6</span>, <span class="at">height=</span><span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/aligned.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<p>The plot below shows the distribution of stitched reads in the data with the percent threshold of stitched reads as set in the QC parameters code block shown as a vertical dotted line. Typically we look to see high levels of percent stitched as this indicates a high quality sequencing run.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">makeQCHistogram</span>(probe_data_qc, </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">annotation_col =</span> <span class="st">"Stitched (%)"</span>, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">fill_by =</span> <span class="st">"Segment"</span>, </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">bins =</span> <span class="dv">50</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">xintercept =</span> QC_params<span class="sc">$</span>percentStitched)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"stitched.svg"</span>), <span class="at">width=</span><span class="dv">6</span>, <span class="at">height=</span><span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/stitched.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-4-tab">
<p>The plot below shows the distribution of aligned reads in the data with the percent saturation threshold as set in the QC parameters code block shown as a vertical dotted line. Saturation is defined as:</p>
<p><span class="math display">\[100* \left( 1 - \frac{deduplicated}{aligned} \right)\]</span></p>
<p>Typically we look to see about 50% saturation or higher, with lower values indicating that the sample may benefit from additional sequencing.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">makeQCHistogram</span>(probe_data_qc, </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">annotation_col =</span> <span class="st">"Saturated (%)"</span>, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">fill_by =</span> <span class="st">"Segment"</span>, </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">bins =</span> <span class="dv">50</span>,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">xintercept =</span> QC_params<span class="sc">$</span>percentSaturation) <span class="sc">+</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Sequencing Saturation (%)"</span>,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Sequencing Saturation (%)"</span>)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"saturated.svg"</span>), <span class="at">width=</span><span class="dv">6</span>, <span class="at">height=</span><span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/saturated.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-5" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-5-tab">
<p>The plot below shows the distribution of area in the data with the area threshold in μm<sup>2</sup> as set in the QC parameters code block shown as a vertical dotted line. AOI area is dependent on experimental design and so thresholds may vary between studies. However, one thing to check for is large outliers in area, which may introduce technical artifacts.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">makeQCHistogram</span>(probe_data_qc, </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">annotation_col =</span> <span class="st">"area"</span>, </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">fill_by =</span> <span class="st">"Segment"</span>, </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">bins =</span> <span class="dv">50</span>,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">xintercept =</span> QC_params<span class="sc">$</span>minArea)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"area.svg"</span>), <span class="at">width=</span><span class="dv">6</span>, <span class="at">height=</span><span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/area.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-6" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-6-tab">
<p>The plot below shows the distribution of nuclei counts in the data with the threshold of nuclei as set in the QC parameters code block shown as a vertical dotted line. Nuclei number per AOI is dependent on experimental design and so thresholds may vary between studies. However, one thing to check for is large outliers, which may introduce technical artifacts.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">makeQCHistogram</span>(probe_data_qc, </span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">annotation_col =</span> <span class="st">"nuclei"</span>, </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">fill_by =</span> <span class="st">"Segment"</span>, </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">bins =</span> <span class="dv">50</span>,</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">xintercept =</span> QC_params<span class="sc">$</span>minNuclei)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"nuclei.svg"</span>), <span class="at">width=</span><span class="dv">6</span>, <span class="at">height=</span><span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/nuclei.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-7" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-7-tab">
<p>Plot the negative probe geometric mean. The vertical bars represent a value of two. Typically geometric mean background values of less than 2 are considered “unstable”.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set core panel name - change this if using a different core panel</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This is the panel that includes the negative probes</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>core_name <span class="ot">=</span> <span class="st">"Hs_P_PRO_NGS_Core_v1.1"</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">makeQCHistogram</span>(probe_data_qc, </span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">annotation_col =</span> core_name, </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>                     <span class="at">fill_by =</span> <span class="st">"Segment"</span>, </span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>                     <span class="at">bins =</span> <span class="dv">50</span>,</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>                     <span class="at">xintercept =</span> <span class="dv">2</span>,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>                     <span class="at">scale_trans =</span> <span class="st">"log10"</span>)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="fu">paste0</span>(core_name, <span class="st">".svg"</span>)), <span class="at">width=</span><span class="dv">6</span>, <span class="at">height=</span><span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/Hs_P_PRO_NGS_Core_v1.1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>As a summary, we show a table of passing and flagged segments.</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Pass</th>
<th style="text-align: right;">Warning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">LowReads</td>
<td style="text-align: right;">190</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">LowTrimmed</td>
<td style="text-align: right;">190</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">LowStitched</td>
<td style="text-align: right;">190</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">LowAligned</td>
<td style="text-align: right;">190</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">LowSaturation</td>
<td style="text-align: right;">176</td>
<td style="text-align: right;">14</td>
</tr>
<tr class="even">
<td style="text-align: left;">HighNTC</td>
<td style="text-align: right;">190</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">LowNuclei</td>
<td style="text-align: right;">189</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">LowArea</td>
<td style="text-align: right;">190</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">TOTAL FLAGS</td>
<td style="text-align: right;">175</td>
<td style="text-align: right;">15</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Additionally, we summarize the No Template Control (NTC) counts on a per plate basis. Values &gt;1,000 could indicate contamination for the segments associated with this NTC; however, in cases where the NTC count is between 1,000- 10,000, the segments may be used if the NTC data is uniformly low (e.g.&nbsp;distributed across the whole panel, this would represent 0-2 counts for all probes).</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create summary table of NTCs</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>ntc_summary_tab <span class="ot">&lt;-</span> <span class="fu">ddply</span>(<span class="fu">sData</span>(probe_data_qc) <span class="sc">%&gt;%</span> </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                          dplyr<span class="sc">::</span><span class="fu">select</span>(Plate_ID, NTC_ID, NTC), .(Plate_ID, NTC_ID), <span class="cf">function</span>(x){</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a> ntc_well <span class="ot">&lt;-</span> x<span class="sc">$</span>NTC_ID[<span class="dv">1</span>]</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a> ntc_count <span class="ot">&lt;-</span> x<span class="sc">$</span>NTC[<span class="dv">1</span>]</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a> n_samples <span class="ot">&lt;-</span> <span class="fu">nrow</span>(x)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a> <span class="fu">return</span>(<span class="fu">data.frame</span>(<span class="st">'NTC Well ID'</span><span class="ot">=</span>ntc_well, <span class="st">'NTC counts'</span> <span class="ot">=</span> ntc_count, <span class="st">'Samples'</span><span class="ot">=</span>n_samples))</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>NTC_ID)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ntc_summary_tab) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Plate"</span>, <span class="st">"NTC Well"</span>, <span class="st">"NTC Counts"</span>, <span class="st">"Samples"</span>)</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Save NTC Summary</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>results_list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/results.rds"</span>))</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>results_list<span class="sc">$</span>NTC_Summary <span class="ot">&lt;-</span> ntc_summary_tab</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, <span class="fu">paste0</span>(outdir, <span class="st">"/results.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Plate</th>
<th style="text-align: left;">NTC Well</th>
<th style="text-align: right;">NTC Counts</th>
<th style="text-align: right;">Samples</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">1001660033000</td>
<td style="text-align: left;">DSP-1001660033000-Y-A01.dcc</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">95</td>
</tr>
<tr class="even">
<td style="text-align: left;">1001660034000</td>
<td style="text-align: left;">DSP-1001660034000-Z-A01.dcc</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">95</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Here, we remove any segments that were flagged in our segment QC.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>seg_qc_samples_to_remove <span class="ot">&lt;-</span> <span class="fu">rownames</span>(QCResults)[QCResults<span class="sc">$</span>QCStatus<span class="sc">!=</span><span class="st">"PASS"</span>]</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(seg_qc_samples_to_remove)<span class="sc">&gt;</span><span class="dv">0</span>){</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  probe_data_qc <span class="ot">&lt;-</span> probe_data_qc[,<span class="sc">-</span><span class="fu">which</span>(<span class="fu">colnames</span>(probe_data_qc) <span class="sc">%in%</span> seg_qc_samples_to_remove)]</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="quality-control-of-housekeeper-proteins-and-background-isotypes" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="quality-control-of-housekeeper-proteins-and-background-isotypes"><span class="header-section-number">3.2</span> Quality control of housekeeper proteins and background (isotypes)</h2>
<p>Before we perform any normalization to control molecules, we need to check if the control molecules are themselves correlated with the predictors of interests. There are two classes of control molecules in the dataset: isotypes and housekeepers. We extract their names here.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>results_list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/results.rds"</span>))</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co"># biological negatives (isotype controls)</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>iggs <span class="ot">=</span> <span class="fu">iggNames</span>(probe_data_qc)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">paste0</span>(<span class="st">"All IgGs: "</span>,<span class="fu">paste0</span>(iggs, <span class="at">collapse=</span><span class="st">", "</span>)))</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="co"># housekeepers:</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>hks <span class="ot">=</span> <span class="fu">hkNames</span>(probe_data_qc)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">paste0</span>(<span class="st">"All HKGs: "</span>,<span class="fu">paste0</span>(hks, <span class="at">collapse=</span><span class="st">", "</span>)))</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Store names</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>results_list<span class="sc">$</span>iggs <span class="ot">&lt;-</span> iggs</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>results_list<span class="sc">$</span>hks <span class="ot">&lt;-</span> hks</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, <span class="fu">paste0</span>(outdir, <span class="st">"/results.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>All IgGs: Hmr IgG, Ms IgG2b, Rb IgG, Rt IgG2a, Ms IgG1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>All HKGs: Calreticulin, Histone H3, RPS6, TOMM20, GAPDH</code></pre>
</div>
</div>
<section id="isotype-controls" class="level3" data-number="3.2.1">
<h3 data-number="3.2.1" class="anchored" data-anchor-id="isotype-controls"><span class="header-section-number">3.2.1</span> Isotype Controls</h3>
<p>Plot the joint expression of each pairwise <code>IgG</code> control molecule to see if:</p>
<ol type="1">
<li>There’s evidence that one <code>IgG</code> is noisy compared to others</li>
<li>There’s clustering by predictors, which could bias the normalized data</li>
</ol>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract counts</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">assayDataElement</span>(probe_data_qc, <span class="at">elt=</span><span class="st">"exprs"</span>))</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>iggs_df <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="fu">as.data.frame</span>(counts[,iggs]))</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Bind to phenotypic data</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>pheno_data <span class="ot">&lt;-</span> <span class="fu">pData</span>(probe_data_qc)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">ncol</span>(iggs_df)</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>iggs_df <span class="ot">&lt;-</span> <span class="fu">cbind</span>(iggs_df, pheno_data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(Responder, slide, Segment))</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Plots</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>p0 <span class="ot">&lt;-</span> GGally<span class="sc">::</span><span class="fu">ggpairs</span>(<span class="at">data=</span>iggs_df, <span class="at">columns=</span><span class="dv">1</span><span class="sc">:</span>n, <span class="at">progress=</span><span class="cn">FALSE</span>,</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>            ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">alpha=</span><span class="fl">0.1</span>)) <span class="sc">+</span> <span class="fu">theme_bw</span>()</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p0, <span class="at">file=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"QC_pairs_background.svg"</span>), <span class="at">width=</span>n<span class="sc">*</span><span class="fl">2.5</span>, <span class="at">height=</span>n<span class="sc">*</span><span class="fl">2.5</span>)</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore by factors of interest</span></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">pairs_custom</span>(iggs_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">eval</span>(iggs), Responder), Responder)</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p2, <span class="at">file=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"QC_pairs_background_Responder.svg"</span>), <span class="at">width=</span>n<span class="sc">*</span><span class="fl">2.8</span>, <span class="at">height=</span>n<span class="sc">*</span><span class="fl">2.8</span>)</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">pairs_custom</span>(iggs_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">eval</span>(iggs), slide), slide) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>, <span class="at">vjust=</span><span class="fl">0.5</span>))</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p3, <span class="at">file=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"QC_pairs_background_slide.svg"</span>), <span class="at">width=</span>n<span class="sc">*</span><span class="fl">2.8</span>, <span class="at">height=</span>n<span class="sc">*</span><span class="fl">2.8</span>)</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">pairs_custom</span>(iggs_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">eval</span>(iggs), Segment), Segment)</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p4, <span class="at">file=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"QC_pairs_background_Segment.svg"</span>), <span class="at">width=</span>n<span class="sc">*</span><span class="fl">2.8</span>, <span class="at">height=</span>n<span class="sc">*</span><span class="fl">2.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true">IgGs Alone</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false">Responder</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-3" role="tab" aria-controls="tabset-2-3" aria-selected="false">Slide</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-4" role="tab" aria-controls="tabset-2-4" aria-selected="false">Segment</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/QC_pairs_background.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/QC_pairs_background_Responder.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-2-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-3-tab">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/QC_pairs_background_slide.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-2-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-4-tab">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/QC_pairs_background_Segment.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>As expected, we see high correlations between isotype controls, indicating that all five isotype controls can be used for normalization. If an isototype control were to show low correlations with all others, we would suggest excluding that isotype control from normalization.</p>
</div>
</div>
</section>
<section id="housekeeping-controls" class="level3" data-number="3.2.2">
<h3 data-number="3.2.2" class="anchored" data-anchor-id="housekeeping-controls"><span class="header-section-number">3.2.2</span> Housekeeping Controls</h3>
<p>Run the same analysis for the housekeepers. This is the same overall process as for the isotypes.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>hks_df <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="fu">as.data.frame</span>(counts[,hks]))</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">ncol</span>(hks_df)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>hks_df <span class="ot">&lt;-</span> <span class="fu">cbind</span>(hks_df, pheno_data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(Responder, slide, Segment))</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>p6 <span class="ot">&lt;-</span> <span class="fu">ggpairs</span>(<span class="at">data=</span>hks_df, <span class="at">columns=</span><span class="dv">1</span><span class="sc">:</span>n, <span class="at">progress=</span><span class="cn">FALSE</span>,</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>            ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">alpha=</span><span class="fl">0.1</span>)) <span class="sc">+</span> <span class="fu">theme_bw</span>()</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p6,<span class="at">file=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"QC_pairs_housekeepers.svg"</span>), <span class="at">width=</span>n<span class="sc">*</span><span class="fl">2.5</span>, <span class="at">height=</span>n<span class="sc">*</span><span class="fl">2.5</span>)</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>p7 <span class="ot">&lt;-</span> <span class="fu">pairs_custom</span>(hks_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">eval</span>(hks), Responder), Responder)</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p7, <span class="at">file=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"QC_pairs_HK_Responder.svg"</span>), <span class="at">width=</span>n<span class="sc">*</span><span class="fl">2.8</span>, <span class="at">height=</span>n<span class="sc">*</span><span class="fl">2.8</span>)</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>p8 <span class="ot">&lt;-</span> <span class="fu">pairs_custom</span>(hks_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">eval</span>(hks), slide), slide) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>, <span class="at">vjust=</span><span class="fl">0.5</span>))</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p8, <span class="at">file=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"QC_pairs_HK_slide.svg"</span>), <span class="at">width=</span>n<span class="sc">*</span><span class="fl">2.8</span>, <span class="at">height=</span>n<span class="sc">*</span><span class="fl">2.8</span>)</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>p9 <span class="ot">&lt;-</span> <span class="fu">pairs_custom</span>(hks_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">eval</span>(hks), Segment), Segment)</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p9, <span class="at">file=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"QC_pairs_HK_Segment.svg"</span>), <span class="at">width=</span>n<span class="sc">*</span><span class="fl">2.8</span>, <span class="at">height=</span>n<span class="sc">*</span><span class="fl">2.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-3-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-1" role="tab" aria-controls="tabset-3-1" aria-selected="true">HKs Alone</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-2" role="tab" aria-controls="tabset-3-2" aria-selected="false">Responder</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-3" role="tab" aria-controls="tabset-3-3" aria-selected="false">Slide</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-4" role="tab" aria-controls="tabset-3-4" aria-selected="false">Segment</a></li></ul>
<div class="tab-content">
<div id="tabset-3-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-3-1-tab">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/QC_pairs_housekeepers.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-3-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-2-tab">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/QC_pairs_HK_Responder.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-3-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-3-tab">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/QC_pairs_HK_slide.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-3-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-4-tab">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/QC_pairs_HK_Segment.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>As expected, we see high correlation between housekeeper proteins. It is possible for one or two of the proteins to show lower correlation with the others, indicating that these should be removed if using the housekeeper proteins for normalization.</p>
</div>
</div>
</section>
<section id="isotype-vs-housekeeper-controls" class="level3" data-number="3.2.3">
<h3 data-number="3.2.3" class="anchored" data-anchor-id="isotype-vs-housekeeper-controls"><span class="header-section-number">3.2.3</span> Isotype vs housekeeper controls</h3>
<p>Next, we look at the correlation between the geometric means of the housekeeper proteins and the isotype (background) controls. We’re looking for concordance between the two sets of controls. In a high quality dataset, we expect to see reasonable correlation between the housekeeper and isotype controls.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>iggs_df <span class="ot">&lt;-</span> (<span class="fu">as.data.frame</span>(counts[,iggs]))</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>hks_df <span class="ot">&lt;-</span> (<span class="fu">as.data.frame</span>(counts[,hks]))</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>igg_geo_means <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">apply</span>(iggs_df, <span class="dv">1</span>, EnvStats<span class="sc">::</span>geoMean))</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>hks_geo_means <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">apply</span>(hks_df, <span class="dv">1</span>, EnvStats<span class="sc">::</span>geoMean))</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>geo_means <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cbind</span>(igg_geo_means, hks_geo_means))</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">ncol</span>(geo_means)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>p10 <span class="ot">&lt;-</span> GGally<span class="sc">::</span><span class="fu">ggpairs</span>(<span class="at">data=</span>geo_means, <span class="at">columns=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="at">progress=</span><span class="cn">FALSE</span>,</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>            ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">alpha=</span><span class="fl">0.1</span>)) <span class="sc">+</span> <span class="fu">theme_bw</span>()</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p10, <span class="at">file=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"QC_pairwise_geomeans.svg"</span>), <span class="at">width=</span>n<span class="sc">*</span><span class="fl">2.8</span>, <span class="at">height=</span>n<span class="sc">*</span><span class="fl">2.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/QC_pairwise_geomeans.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>High correlation between the geometric means of the housekeeper proteins and the isotype (background) controls.</p>
</div>
</div>
</section>
</section>
<section id="signal-to-noise-ratio-snr" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="signal-to-noise-ratio-snr"><span class="header-section-number">3.3</span> Signal to noise ratio, SNR</h2>
<p>We calculate the signal-to-noise ratio (SNR) for every protein in every AOI using the background probes. For this, we use the <code>IgGs</code> to represent background. We then plot the signal-to-noise ratio for all AOIs with all protein probes together. Because we have hundreds of proteins, we make an extra wide plot for exploration such that the protein names are legible.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate SNR</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>backgrounds <span class="ot">&lt;-</span> iggs</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>prData <span class="ot">&lt;-</span> <span class="fu">pData</span>(<span class="fu">protocolData</span>(probe_data_qc))</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>snr <span class="ot">&lt;-</span> <span class="fu">get_snr</span>(<span class="at">exp=</span>counts, <span class="at">backgrounds =</span> backgrounds, <span class="at">n_processors =</span> <span class="dv">5</span>, </span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">type=</span><span class="st">"SBR"</span>)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>snr2 <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(snr, <span class="at">id.var=</span><span class="cn">NA</span>)</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(snr2) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"name"</span>, <span class="st">"protein"</span>, <span class="st">"snr"</span>)</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>snr2 <span class="ot">&lt;-</span> base<span class="sc">::</span><span class="fu">merge</span>(snr2, prData <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">name=</span><span class="fu">rownames</span>(prData)) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(name), <span class="at">by=</span><span class="st">"name"</span>)</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>                            </span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Order proteins based on median SNR, low to high. </span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>medians_snr <span class="ot">&lt;-</span> <span class="fu">ddply</span>(snr2, .(protein), summarize, <span class="at">Median=</span><span class="fu">median</span>(snr), </span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">UQ=</span><span class="fu">as.numeric</span>(<span class="fu">quantile</span>(snr, <span class="at">probs=</span><span class="fl">0.75</span>))) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(UQ)</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>snr2<span class="sc">$</span>protein <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">as.character</span>(snr2<span class="sc">$</span>protein), </span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>              <span class="at">levels=</span><span class="fu">as.character</span>(medians_snr<span class="sc">$</span>protein), <span class="at">order=</span><span class="cn">TRUE</span>)</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>snr2 <span class="ot">&lt;-</span> snr2 <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(protein, name)</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up protein names to color by type</span></span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>levs <span class="ot">&lt;-</span> <span class="fu">levels</span>(<span class="fu">factor</span>(snr2<span class="sc">$</span>protein))                                  </span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>labs <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">paste0</span>(<span class="st">"&lt;span style='color:"</span>,</span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">ifelse</span>(levs <span class="sc">%in%</span> iggs, <span class="st">"red"</span>,</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>                               <span class="fu">ifelse</span>(levs <span class="sc">%in%</span> hks,  <span class="st">"grey"</span>, <span class="st">"black"</span>)),</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"'&gt;"</span>, levs, <span class="st">"&lt;/span&gt;"</span>), levs)                  </span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Add segment metadata</span></span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a>snr2<span class="sc">$</span>Segment <span class="ot">&lt;-</span> pheno_data[snr2<span class="sc">$</span>name, <span class="st">"Segment"</span>]</span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot SNRs</span></span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>snr2, <span class="fu">aes</span>(<span class="at">x=</span>protein, <span class="at">y=</span><span class="fu">log2</span>(snr))) <span class="sc">+</span> </span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span> </span>
<span id="cb22-31"><a href="#cb22-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">size=</span><span class="fl">0.3</span>, <span class="fu">aes</span>(<span class="at">color =</span> Segment)) <span class="sc">+</span></span>
<span id="cb22-32"><a href="#cb22-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> pal_main<span class="sc">$</span>Segment) <span class="sc">+</span></span>
<span id="cb22-33"><a href="#cb22-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">lty=</span><span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb22-34"><a href="#cb22-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"Protein"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"log2(SNR)"</span>) <span class="sc">+</span></span>
<span id="cb22-35"><a href="#cb22-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">labels =</span> labs) <span class="sc">+</span></span>
<span id="cb22-36"><a href="#cb22-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> ggtext<span class="sc">::</span><span class="fu">element_markdown</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="dv">1</span>, </span>
<span id="cb22-37"><a href="#cb22-37" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb22-38"><a href="#cb22-38" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="st">"white"</span>,</span>
<span id="cb22-39"><a href="#cb22-39" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">linewidth =</span> <span class="fl">0.5</span>, <span class="at">linetype=</span><span class="st">"solid"</span>, </span>
<span id="cb22-40"><a href="#cb22-40" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">colour =</span><span class="st">"grey20"</span>),</span>
<span id="cb22-41"><a href="#cb22-41" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.05</span>, <span class="fl">0.75</span>)) <span class="sc">+</span></span>
<span id="cb22-42"><a href="#cb22-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">colour=</span><span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size=</span><span class="dv">2</span>))) </span>
<span id="cb22-43"><a href="#cb22-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-44"><a href="#cb22-44" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">file=</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"snr_plot.png"</span>), <span class="at">height=</span><span class="dv">8</span>, <span class="at">width=</span><span class="dv">80</span>, <span class="at">limitsize=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="hscroll-plot">
<div class="cell" data-layout-align="left" data-fig-responsive="false">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="./Assets/qc/snr_plot.png" class="img-fluid quarto-figure quarto-figure-left figure-img" width="6000"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Most proteins show signal‑to‑noise ratios above background in at least one AOI, though a few have relatively low overall signal. As expected, different antibodies exhibit different background levels, so we look for AOIs with high outlier values as evidence of true expression. For example, compare the <code>Ms IgG1</code> control towards the left, whose distribution is very tight, to <code>CD38</code>, the first protein on the left, which has a much wider spread. The AOIs at the upper end of the <code>CD38</code> distribution are likely true positives because they rise well above the main background range. Another illustrative case is <code>Cytokeratin 19</code> towards the right side of the plot, which shows a long high‑signal tail and a clear separation between CK+ and CK– segments.</p>
<p>We will not be dropping any proteins at this point, but will use these results as context while investigating the data further.</p>
</div>
</div>
</section>
<section id="normalization" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="normalization"><span class="header-section-number">3.4</span> Normalization</h2>
<p>The purpose of normalization is to adjust for technical variables, such as ROI/AOI surface area and tissue quality, and enable meaningful biological and statistical discoveries.</p>
<p>Two common methods for normalization of GeoMx DSP Protein data are i) Background (IgGs) or ii) Housekeeper normalization. Both methods estimate a normalization factor per ROI/AOI segment to bring the segment data distributions together. Background is typically the preferred approach as off-target binding means the negative controls end up measuring the total amount of bindable biological material, which is exactly what we’re trying to normalize out. We demonstrate both approaches here.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>High correlation between the geometric mean of the negative control probe counts and the geometric mean of the housekeeper protein counts suggests that both background and housekeeper-based normalization would work well in this dataset. We generally recommend background normalization as we find it to be robust in the majority of datasets. Therefore, this dataset will undergo background normalization using the isotype controls.</p>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Background or IgG normalization</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>target_data <span class="ot">&lt;-</span> <span class="fu">normalize</span>(probe_data_qc, <span class="at">norm_method =</span> <span class="st">"neg"</span>,</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">fromElt =</span> <span class="st">"exprs"</span>, <span class="at">toElt =</span> <span class="st">"neg_norm"</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="fu">assayDataElement</span>(<span class="at">object =</span> target_data, <span class="at">elt =</span> <span class="st">"log_neg"</span>) <span class="ot">&lt;-</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>     <span class="fu">assayDataApply</span>(target_data, <span class="dv">2</span>, <span class="at">FUN =</span> log, <span class="at">base =</span> <span class="dv">2</span>, <span class="at">elt =</span> <span class="st">"neg_norm"</span>)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="co"># # use norm_method = "hk" for housekeeper norm</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>target_data <span class="ot">&lt;-</span> <span class="fu">normalize</span>(target_data, <span class="at">norm_method =</span> <span class="st">"hk"</span>,</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>                         <span class="at">toElt =</span> <span class="st">"hk_norm"</span>)</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="fu">assayDataElement</span>(<span class="at">object =</span> target_data, <span class="at">elt =</span> <span class="st">"log_hk"</span>) <span class="ot">&lt;-</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">assayDataApply</span>(target_data, <span class="dv">2</span>, <span class="at">FUN =</span> log, <span class="at">base =</span> <span class="dv">2</span>, <span class="at">elt =</span> <span class="st">"hk_norm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>At this point, we have now performed quality control checks on our AOIs (sequencing statistics, area metrics) and on our probes (control probe correlations, SNR). We’ve additionally performed normalization to take technical effects into account and are ready to proceed to exploratory data analyses.</p>
<p>In summary:</p>
<ul>
<li>175 out of 190 AOI segments passed QC</li>
<li>14 AOIs were flagged for low sequencing saturation and 1 AOI was flagged for low nuclei counts</li>
<li>We see high correlation between the IgG proteins, as well as within the housekeeper proteins</li>
<li>Data underwent background (IgG) normalization</li>
</ul>
</section></details>
</section>

<section id="sec-eda" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Exploratory Data Analysis</h1>
<p>The following section is used to understand the general structure of the data and whether the hypotheses that were defined initially are likely to identify significant findings. This qualitative exploration can also help identify any study design considerations for downstream analyses.</p>
<details>
<summary>
Click to show/hide exploratory data analysis steps
</summary>
<section id="dimensional-reduction" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="dimensional-reduction"><span class="header-section-number">4.1</span> Dimensional Reduction</h2>
<p>In this section, we use two dimensional reduction techniques to identify broad patterns in the expression data. For both of these, we start from the normalized data.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-6-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-1" role="tab" aria-controls="tabset-6-1" aria-selected="true">PCA</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-6-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-2" role="tab" aria-controls="tabset-6-2" aria-selected="false">UMAP</a></li></ul>
<div class="tab-content">
<div id="tabset-6-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-6-1-tab">
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>target_data <span class="ot">&lt;-</span> <span class="fu">getPCA</span>(target_data, <span class="at">elt =</span> <span class="st">"log_neg"</span>, <span class="at">the_prefix =</span> <span class="st">"PCA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-4-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-1" role="tab" aria-controls="tabset-4-1" aria-selected="true">Basic</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-2" role="tab" aria-controls="tabset-4-2" aria-selected="false">Faceted</a></li></ul>
<div class="tab-content">
<div id="tabset-4-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-4-1-tab">
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the variables that will determine the color/shape of the points in the dim reduction section:</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>dim_reduction_color_by <span class="ot">&lt;-</span> factors_of_interest[<span class="dv">3</span>]</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>dim_reduction_shape_by <span class="ot">&lt;-</span> factors_of_interest[<span class="dv">1</span>]</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">pData</span>(target_data),</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x =</span> PCA_coords_1, <span class="at">y =</span> PCA_coords_2)) <span class="sc">+</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes_string</span>(<span class="at">color =</span> dim_reduction_color_by,</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>                        <span class="at">shape =</span> dim_reduction_shape_by),</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"PCA 1 ("</span>, <span class="fu">round</span>(<span class="fu">experimentData</span>(target_data)<span class="sc">@</span>other<span class="sc">$</span>PCA[<span class="dv">1</span>], <span class="dv">2</span>), <span class="st">"%)"</span>),</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"PCA 2 ("</span>, <span class="fu">round</span>(<span class="fu">experimentData</span>(target_data)<span class="sc">@</span>other<span class="sc">$</span>PCA[<span class="dv">2</span>], <span class="dv">2</span>), <span class="st">"%)"</span>),</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"PCA"</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>) <span class="sc">+</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(dim_reduction_color_by)) {</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values =</span> pal_main[[dim_reduction_color_by]])</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"PCA_log2_norm.svg"</span>), <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/PCA_log2_norm.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-4-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-2-tab">
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">pData</span>(target_data),</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x =</span> PCA_coords_1, <span class="at">y =</span> PCA_coords_2)) <span class="sc">+</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes_string</span>(<span class="at">color =</span> dim_reduction_color_by,</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">shape =</span> dim_reduction_shape_by),</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> slide, <span class="at">nrow =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"PCA 1 ("</span>, <span class="fu">round</span>(<span class="fu">experimentData</span>(target_data)<span class="sc">@</span>other<span class="sc">$</span>PCA[<span class="dv">1</span>], <span class="dv">2</span>), <span class="st">"%)"</span>),</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"PCA 2 ("</span>, <span class="fu">round</span>(<span class="fu">experimentData</span>(target_data)<span class="sc">@</span>other<span class="sc">$</span>PCA[<span class="dv">2</span>], <span class="dv">2</span>), <span class="st">"%)"</span>)</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>) <span class="sc">+</span></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(dim_reduction_color_by)) {</span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values =</span> pal_main[[dim_reduction_color_by]])</span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"PCA_log2_norm_faceted.svg"</span>), <span class="at">width =</span> <span class="dv">12</span>, <span class="at">height =</span> <span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/PCA_log2_norm_faceted.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-6-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-6-2-tab">
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>target_data <span class="ot">&lt;-</span> <span class="fu">getUMAP</span>(target_data, <span class="at">elt =</span> <span class="st">"log_neg"</span>, <span class="at">the_prefix =</span> <span class="st">"UMAP"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-5-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-1" role="tab" aria-controls="tabset-5-1" aria-selected="true">Basic</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-5-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-2" role="tab" aria-controls="tabset-5-2" aria-selected="false">Faceted</a></li></ul>
<div class="tab-content">
<div id="tabset-5-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-5-1-tab">
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">pData</span>(target_data),</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x =</span> UMAP_coords_1, <span class="at">y =</span> UMAP_coords_2)) <span class="sc">+</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes_string</span>(<span class="at">color =</span> dim_reduction_color_by,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">shape =</span> dim_reduction_shape_by),</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"UMAP 1"</span>, <span class="at">y =</span> <span class="st">"UMAP 2"</span>, <span class="at">title =</span> <span class="st">"UMAP"</span>) <span class="sc">+</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>) <span class="sc">+</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(dim_reduction_color_by)) {</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values =</span> pal_main[[dim_reduction_color_by]])</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"UMAP_log2_norm.svg"</span>), <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/UMAP_log2_norm.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-5-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-5-2-tab">
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">pData</span>(target_data),</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x =</span> UMAP_coords_1, <span class="at">y =</span> UMAP_coords_2)) <span class="sc">+</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes_string</span>(<span class="at">color =</span> dim_reduction_color_by,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">shape =</span> dim_reduction_shape_by),</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> slide, <span class="at">nrow =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"UMAP 1"</span>, <span class="at">y =</span> <span class="st">"UMAP 2"</span>) <span class="sc">+</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>,</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.box =</span> <span class="st">"vertical"</span>,</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.margin =</span> <span class="fu">margin</span>()) <span class="sc">+</span></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(dim_reduction_color_by)) {</span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values =</span> pal_main[[dim_reduction_color_by]])</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"UMAP_log2_norm_faceted.svg"</span>), <span class="at">width =</span> <span class="dv">16</span>, <span class="at">height =</span> <span class="dv">12</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/UMAP_log2_norm_faceted.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Data in the PCA and UMAP show clustering primarily by segment type, with some localized clustering by response status.</p>
</div>
</div>
</section>
<section id="clustering-features-with-high-coefficient-of-variation" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="clustering-features-with-high-coefficient-of-variation"><span class="header-section-number">4.2</span> Clustering features with high coefficient of variation</h2>
<p>Another approach to explore the data is to calculate the coefficient of variation (CV) for each protein (<span class="math inline">\(g\)</span>) using the formula <span class="math inline">\(CV_g = SD_g/mean_g\)</span>. We then identify proteins with high CVs that should have large differences across the various profiled segments. This unbiased approach can reveal highly variable proteins across the study that may be of particular interest.</p>
<p>For this study, we plot the proteins with CVs in the 90th quantile for closer exploration. We use unsupervised hierarchical clustering based on Pearson distance, displayed as a heatmap.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>top_quantile <span class="ot">&lt;-</span> <span class="fl">0.90</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>CV_dat <span class="ot">&lt;-</span> <span class="fu">assayDataApply</span>(target_data, <span class="at">elt =</span> <span class="st">"log_neg"</span>, <span class="at">MARGIN =</span> <span class="dv">1</span>, getCV)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>cv_feats <span class="ot">&lt;-</span> <span class="fu">names</span>(CV_dat)[CV_dat <span class="sc">&gt;</span> <span class="fu">quantile</span>(CV_dat, top_quantile)]</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">file.exists</span>(<span class="fu">file.path</span>(qc_dir, <span class="st">"cv_heatmap.RDS"</span>))){</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a> cv_heatmap <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(qc_dir, <span class="st">"cv_heatmap.RDS"</span>))</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>  cv_heatmap <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(<span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(<span class="fu">assayDataElement</span>(target_data[cv_feats, ], </span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">elt =</span> <span class="st">"log_neg"</span>)))),</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">seq</span>(<span class="sc">-</span><span class="fl">2.5</span>, <span class="fl">2.5</span>, <span class="fl">0.05</span>)), </span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>                         <span class="fu">c</span>(<span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"#0092b5"</span>, <span class="st">"white"</span>, <span class="st">"#a6ce39"</span>))(<span class="dv">101</span>))),</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">name =</span> <span class="st">'z-score'</span>,</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">clustering_distance_rows =</span> <span class="st">"pearson"</span>, <span class="at">clustering_method_rows =</span> <span class="st">"average"</span>,</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">clustering_distance_columns =</span> <span class="st">"pearson"</span>, <span class="at">clustering_method_columns =</span> <span class="st">"average"</span>,</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">column_split =</span> <span class="dv">2</span>, <span class="at">row_split =</span> <span class="dv">2</span>,</span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">border_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> <span class="st">"darkgray"</span>),</span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">show_row_names =</span> <span class="cn">FALSE</span>, <span class="at">show_column_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">use_raster =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">top_annotation =</span> </span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a>         <span class="fu">HeatmapAnnotation</span>(<span class="at">df=</span> <span class="fu">pData</span>(target_data)[, factors_of_interest],</span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a>                           <span class="at">col =</span> pal_main,</span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a>                           <span class="at">gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> <span class="st">"gray"</span>))</span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb30-25"><a href="#cb30-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(cv_heatmap, <span class="at">file=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"cv_heatmap.RDS"</span>))</span>
<span id="cb30-26"><a href="#cb30-26" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb30-27"><a href="#cb30-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-28"><a href="#cb30-28" aria-hidden="true" tabindex="-1"></a>svglite<span class="sc">::</span><span class="fu">svglite</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"cv_heatmap.svg"</span>),</span>
<span id="cb30-29"><a href="#cb30-29" aria-hidden="true" tabindex="-1"></a>                 <span class="at">width=</span><span class="dv">10</span>, <span class="at">height=</span><span class="fl">6.5</span>)</span>
<span id="cb30-30"><a href="#cb30-30" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(cv_heatmap, <span class="at">heatmap_legend_side =</span> <span class="st">"left"</span>)</span>
<span id="cb30-31"><a href="#cb30-31" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/qc/cv_heatmap.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this plot of top CV proteins, we see that the cv data cluster by segment and slide.</p>
</div>
</div>
<p>After these exploratory analyses are complete, we recommend saving the processed object, <code>target_data</code>, for easier reloading in future sessions.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save normalized expression to disk.</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(<span class="at">object=</span>target_data, <span class="at">file=</span><span class="fu">file.path</span>(object_dir, <span class="st">"target_data.RDS"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We’ve now done some basic exploratory analysis to examine which segments and proteins tend to cluster together, and to identify which factors may be the biggest drivers of differences across our dataset. Next, we’ll move on to test explicitly laid out hypotheses in our differential expression section.</p>
</section></details>
</section>

<section id="sec-de" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Differential Expression</h1>
<p>Differential expression (DE) is a common method used to compare protein expression between sample groups and to quantify the significance of these differences. We perform DE on a per-protein basis where the normalized expression is modeled using a linear mixed-effect model (LMM) to account for the sampling of multiple ROI/AOI segments per tissue.</p>
<details>
<summary>
Click to show/hide differential expression steps
</summary>
<p>For a given comparison, we will subset the data down to the relevant segments and perform DE. We provide tables, heatmaps, violin plots, and volcano plots of the top differentially expressed proteins.</p>
<section id="contrasting-groups-with-linear-mixed-models" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="contrasting-groups-with-linear-mixed-models"><span class="header-section-number">5.1</span> Contrasting groups with linear mixed models</h2>
<p>In a given contrast, we are interested in comparing the magnitude of differences between groups and quantifying their significance. In many GeoMx experiments, multiple areas of illumination (AOIs) or regions of interest (ROIs) are sampled from each slide. Because these AOIs are nested within a slide and therefore not independent, we use a linear mixed‑effects model (LMM) with slide as a random effect. This allows us to correctly account for subsampling within each tissue section, rather than treating all AOIs as independent in a traditional statistical test.</p>
<p>There are two common versions of the LMM used for GeoMx data: A) with a random slope, and B) without a random slope.</p>
<p>A random slope is used when the groups being compared co‑exist within a slide (for example, different Regions drawn on the same tissue section). In this case, different slides may show different magnitudes of the Region effect, and the random slope allows each slide to have its own version of that effect. This captures the fact that slides can differ not just in baseline expression (random intercept) but also in how strongly the contrast appears.</p>
<p>When the features being compared are mutually exclusive at the slide level (e.g., each slide is entirely “healthy” or entirely “disease”), the contrast does not vary within any single slide. In such cases, a random slope is unnecessary because there is no within‑slide effect whose size could vary across slides; the model only needs a random intercept to account for per‑slide baseline shifts.</p>
<p>We represent the two variations on the LMM in the schematic below. Note that the schematic says ‘gene’ but the same logic applies to proteins. In this example, there are a number of slides (“Slide”). A slide is classified as either healthy or a disease in the “Disease” category. Within a slide, different regions are present and these are illustrated as pink or green ROIs.</p>
<p><em>Example A</em>: When estimating the effect of “Region”, “Slide” is used as a random effect. Because the slope of “Region” may differ across slides, “Region” is used as a random slope.</p>
<p><em>Example B</em>: In this example study design, Disease is mutually exclusive to a slide. In other words, a slide can only be a “disease” or “healthy”. In this case, there is no need to include the additional term in the model.</p>
<p><img src="./Assets/images/GeomxDETypes.png" class="img-fluid"> In the breast cancer dataset that we’re working with in this vignette, there are not known biological differences (for example tumor vs healthy tissue) within slides and so we will set up our DE test following Example B. We ask: <em>within the stromal compartments, how does protein abundance differ between responders and non-responders?</em></p>
</section>
<section id="ck--responder-vs-non-responder" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="ck--responder-vs-non-responder"><span class="header-section-number">5.2</span> CK- Responder vs Non-responder</h2>
<p>We will compare the stromal compartment between responders vs non-responders in the CK negative compartment.</p>
<p>The following formula was used to model differences for a given protein:</p>
<p><span class="math display">\[ protein \sim Responder + (1|slide) \]</span></p>
<p>We adjust for the multiple sampling of ROI/AOI segments per tissue with the <span class="math inline">\(slide\)</span> variable.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert test variables to factors</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (variable <span class="cf">in</span> factors_of_interest){</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pData</span>(target_data)[[variable]] <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">pData</span>(target_data)[[variable]])</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(variable)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="st">"slide name"</span> <span class="sc">%in%</span> factors_of_interest){</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pData</span>(target_data)[[<span class="st">"slide"</span>]] <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">pData</span>(target_data)[[<span class="st">"slide name"</span>]])</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="fu">pData</span>(target_data)<span class="sc">$</span>Responder <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">pData</span>(target_data)<span class="sc">$</span>Responder, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Yes"</span>, <span class="st">"No"</span>))</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset data: Filter samples for the contrast as needed. </span></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Add and remove parameters as needed. </span></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a><span class="co"># If no filtering is required, use contrast{i}_samples &lt;- rownames(filter(pData(target_data)))</span></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>model_formula1 <span class="ot">&lt;-</span> <span class="er">~</span> Responder <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>slide)</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>group_variable1 <span class="ot">&lt;-</span> <span class="st">"Responder"</span></span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>contrast1_samples <span class="ot">&lt;-</span> </span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a> <span class="fu">rownames</span>(<span class="fu">filter</span>(<span class="fu">pData</span>(target_data),</span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>                 Segment <span class="sc">==</span> <span class="st">"CK neg"</span>))</span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>contrast1 <span class="ot">&lt;-</span> target_data[,contrast1_samples] </span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">file.exists</span>(<span class="fu">file.path</span>(de_data_dir, <span class="st">"de_contrast1.RDS"</span>))){</span>
<span id="cb32-24"><a href="#cb32-24" aria-hidden="true" tabindex="-1"></a> de_contrast1 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(de_data_dir, <span class="st">"de_contrast1.RDS"</span>))</span>
<span id="cb32-25"><a href="#cb32-25" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb32-26"><a href="#cb32-26" aria-hidden="true" tabindex="-1"></a>  de_contrast1 <span class="ot">&lt;-</span> <span class="fu">mixedModelDE</span>(contrast1,</span>
<span id="cb32-27"><a href="#cb32-27" aria-hidden="true" tabindex="-1"></a>                   <span class="at">elt =</span> <span class="st">"log_neg"</span>,</span>
<span id="cb32-28"><a href="#cb32-28" aria-hidden="true" tabindex="-1"></a>                   <span class="at">modelFormula =</span> model_formula1,</span>
<span id="cb32-29"><a href="#cb32-29" aria-hidden="true" tabindex="-1"></a>                   <span class="at">groupVar =</span> group_variable1, </span>
<span id="cb32-30"><a href="#cb32-30" aria-hidden="true" tabindex="-1"></a>                   <span class="at">nCores =</span> parallel<span class="sc">::</span><span class="fu">detectCores</span>()<span class="sc">-</span><span class="dv">1</span>,</span>
<span id="cb32-31"><a href="#cb32-31" aria-hidden="true" tabindex="-1"></a>                   <span class="at">multiCore =</span> <span class="cn">FALSE</span>)</span>
<span id="cb32-32"><a href="#cb32-32" aria-hidden="true" tabindex="-1"></a>  de_contrast1 <span class="ot">&lt;-</span> <span class="fu">formatLMMResults</span>(de_contrast1, <span class="at">p_adjust_method =</span> p_adjust_method_selected)</span>
<span id="cb32-33"><a href="#cb32-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(de_contrast1, <span class="at">file=</span><span class="fu">file.path</span>(de_data_dir, <span class="st">"de_contrast1.RDS"</span>))</span>
<span id="cb32-34"><a href="#cb32-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-35"><a href="#cb32-35" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb32-36"><a href="#cb32-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-37"><a href="#cb32-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify top features to label</span></span>
<span id="cb32-38"><a href="#cb32-38" aria-hidden="true" tabindex="-1"></a>de_padj_threshold_contrast1 <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb32-39"><a href="#cb32-39" aria-hidden="true" tabindex="-1"></a>de_fc_threshold_contrast1 <span class="ot">&lt;-</span> <span class="fl">0.25</span></span>
<span id="cb32-40"><a href="#cb32-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-41"><a href="#cb32-41" aria-hidden="true" tabindex="-1"></a>label_de1 <span class="ot">&lt;-</span> <span class="fu">getTopFeatures</span>(de_contrast1, </span>
<span id="cb32-42"><a href="#cb32-42" aria-hidden="true" tabindex="-1"></a>                            <span class="at">n_features =</span> <span class="dv">11</span>,</span>
<span id="cb32-43"><a href="#cb32-43" aria-hidden="true" tabindex="-1"></a>                            <span class="at">est_thr =</span> de_fc_threshold_contrast1, </span>
<span id="cb32-44"><a href="#cb32-44" aria-hidden="true" tabindex="-1"></a>                            <span class="at">p_adjust_column =</span> <span class="st">"P"</span>,</span>
<span id="cb32-45"><a href="#cb32-45" aria-hidden="true" tabindex="-1"></a>                            <span class="at">p_adjust_thr =</span> de_padj_threshold_contrast1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Figures in the tabs below can be used to visualize differences between our two groups. The <code>Volcano</code> tab plots the log<sub>2</sub> fold change (FC) between groups within the contrast and significance (p-value) for all protein, as well as a correction for multiple comparisons (<code>p_adjust</code>). The <code>Heatmap</code> tab visualizes the top proteins using unsupervised hierarchical clustering on Pearson distances of the normalized data z-scores.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-7-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-1" role="tab" aria-controls="tabset-7-1" aria-selected="true">Volcano</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-7-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-2" role="tab" aria-controls="tabset-7-2" aria-selected="false">Heatmap</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-7-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-3" role="tab" aria-controls="tabset-7-3" aria-selected="false">Violin</a></li></ul>
<div class="tab-content">
<div id="tabset-7-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-7-1-tab">
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>de_contrast1_volcano_list <span class="ot">&lt;-</span> <span class="fu">makeVolcano</span>(<span class="at">df=</span>de_contrast1, </span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">to_label=</span>label_de1<span class="sc">$</span>all,</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">label_color=</span><span class="st">"grey20"</span>, </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">SCALE=</span><span class="fl">1.4</span>, <span class="at">LWD=</span><span class="dv">1</span>,</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">the_name =</span> <span class="fu">paste0</span>(<span class="st">"CK- Responder vs CK- Non-responder"</span>),</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">log_type=</span><span class="st">"log2(FC)"</span>, </span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">fc_cutoff=</span>de_fc_threshold_contrast1,</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">p_adjust_column =</span> <span class="fu">toupper</span>(p_adjust_method_selected),</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">pval_cutoff=</span><span class="fl">0.05</span>)</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(de_contrast1_volcano_list[[<span class="dv">2</span>]], <span class="fu">file.path</span>(de_data_dir, <span class="st">"contrast1_de_volcano.tsv"</span>), <span class="at">sep=</span><span class="st">"   "</span>, <span class="at">quote=</span><span class="cn">FALSE</span>, <span class="at">col.names=</span><span class="cn">TRUE</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>)</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> de_contrast1_volcano_list[[<span class="dv">1</span>]]</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(de_plot_dir, <span class="st">"contrast1_de_volcano.svg"</span>), <span class="at">width=</span><span class="dv">10</span>, <span class="at">height=</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/de/Plots/contrast1_de_volcano.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-7-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-7-2-tab">
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>heatmap1_features <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">subset</span>(de_contrast1, de_contrast1[[<span class="st">"P"</span>]] <span class="sc">&lt;</span> de_padj_threshold_contrast1 <span class="sc">&amp;</span> </span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>                                    <span class="fu">abs</span>(<span class="st">`</span><span class="at">Estimate</span><span class="st">`</span>) <span class="sc">&gt;</span> de_fc_threshold_contrast1)<span class="sc">$</span>Feature))</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>heatmap_dat1 <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a> <span class="fu">assayDataElement</span>(contrast1[heatmap1_features, ], <span class="at">elt =</span> <span class="st">"neg_norm"</span>))))</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>contrast1_de_heatmap <span class="ot">&lt;-</span> </span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a> <span class="fu">Heatmap</span>(heatmap_dat1,</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">col =</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">seq</span>(<span class="sc">-</span><span class="dv">3</span>, <span class="dv">3</span>, <span class="fl">0.05</span>)), </span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>                          <span class="fu">c</span>(<span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"#0092b5"</span>, <span class="st">"white"</span>, <span class="st">"#a6ce39"</span>))(<span class="dv">121</span>))),</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">name =</span> <span class="st">'z-score'</span>, <span class="at">use_raster =</span> <span class="cn">TRUE</span>,</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">clustering_distance_rows =</span> <span class="st">"pearson"</span>, <span class="at">clustering_method_rows =</span> <span class="st">"average"</span>,</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">clustering_distance_columns =</span> <span class="st">"pearson"</span>, <span class="at">clustering_method_columns =</span> <span class="st">"average"</span>,</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">row_split =</span> <span class="dv">2</span>, <span class="at">column_split =</span> <span class="dv">2</span>,</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">border_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> <span class="st">"darkgray"</span>),</span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">show_row_names =</span> <span class="cn">FALSE</span>, <span class="at">show_column_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">right_annotation =</span> </span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>          <span class="fu">rowAnnotation</span>(<span class="at">foo =</span> <span class="fu">anno_mark</span>(<span class="at">at =</span> <span class="fu">match</span>(label_de1<span class="sc">$</span>all, <span class="fu">rownames</span>(heatmap_dat1)),</span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">labels =</span> label_de1<span class="sc">$</span>all)),</span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>         <span class="at">top_annotation =</span> </span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>          <span class="fu">HeatmapAnnotation</span>(<span class="at">df=</span> <span class="fu">pData</span>(contrast1)[,factors_of_interest],</span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a>                            <span class="at">col =</span> pal_main,</span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>                            <span class="at">gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> <span class="st">"gray"</span>)))</span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>svglite<span class="sc">::</span><span class="fu">svglite</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(de_plot_dir, <span class="st">"contrast1_de_heatmap.svg"</span>),</span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>                 <span class="at">width=</span><span class="dv">12</span>, <span class="at">height=</span><span class="dv">10</span>)</span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(contrast1_de_heatmap, <span class="at">merge_legend =</span> <span class="cn">TRUE</span>, <span class="at">heatmap_legend_side =</span> <span class="st">"bottom"</span>, </span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a>      <span class="at">annotation_legend_side =</span> <span class="st">"bottom"</span>, <span class="at">adjust_annotation_extension =</span> <span class="cn">TRUE</span>)</span>
<span id="cb34-28"><a href="#cb34-28" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/de/Plots/contrast1_de_heatmap.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-7-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-7-3-tab">
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">any</span>(label_de1 <span class="sc">%in%</span> <span class="fu">colnames</span>(<span class="fu">pData</span>(contrast1)))){</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"column names in pheno data cannot be label feature names."</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="co"># individual points</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>contrast_factor <span class="ot">&lt;-</span> factors_of_interest[<span class="dv">1</span>]</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>violin_df <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">pData</span>(contrast1) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">eval</span>(contrast_factor)),</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">t</span>(<span class="fu">assayDataElement</span>(contrast1[label_de1<span class="sc">$</span>all, ], <span class="at">elt =</span> <span class="st">"neg_norm"</span>)))</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>violin_df <span class="ot">&lt;-</span> violin_df <span class="sc">%&gt;%</span> tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="sc">-</span><span class="dv">1</span>, <span class="at">names_to =</span> <span class="st">"Feature"</span>, <span class="at">values_to =</span> <span class="st">"Expression"</span>)</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(violin_df)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"contrast_factor"</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Format summary stats</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>violin_p_df <span class="ot">&lt;-</span> <span class="fu">filter</span>(de_contrast1, Feature <span class="sc">%in%</span> label_de1<span class="sc">$</span>all)</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>violin_p_df <span class="ot">&lt;-</span> violin_p_df <span class="sc">%&gt;%</span> tidyr<span class="sc">::</span><span class="fu">separate</span>(<span class="at">col =</span> Comparison, <span class="at">into=</span><span class="fu">c</span>(<span class="st">"group1"</span>, <span class="st">"group2"</span>), <span class="at">sep=</span><span class="st">" vs "</span>)</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>violin_p_df[<span class="fu">toupper</span>(p_adjust_method_selected)] <span class="ot">&lt;-</span> <span class="fu">signif</span>(violin_p_df[<span class="fu">toupper</span>(p_adjust_method_selected)], <span class="dv">3</span>)</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>violin_exp_max <span class="ot">&lt;-</span> <span class="fu">ddply</span>(violin_df, .(Feature), summarize,</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>                        <span class="at">y.position=</span>((<span class="fu">max</span>(Expression)<span class="sc">+</span><span class="dv">1</span>)<span class="sc">*</span><span class="fl">1.1</span>)) <span class="co"># +1 for safe log2</span></span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>violin_p_df <span class="ot">&lt;-</span> base<span class="sc">::</span><span class="fu">merge</span>(violin_p_df, violin_exp_max, <span class="at">by=</span><span class="st">"Feature"</span>)</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(violin_df, </span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x=</span>contrast_factor, <span class="at">y=</span>Expression, <span class="at">fill=</span>contrast_factor)) <span class="sc">+</span> </span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>() <span class="sc">+</span></span>
<span id="cb35-22"><a href="#cb35-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width=</span><span class="fl">0.25</span>, <span class="at">height=</span><span class="dv">0</span>, <span class="at">size =</span> <span class="fl">1.5</span>) <span class="sc">+</span> </span>
<span id="cb35-23"><a href="#cb35-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> pal_main[[contrast_factor]]) <span class="sc">+</span></span>
<span id="cb35-24"><a href="#cb35-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>Feature, <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb35-25"><a href="#cb35-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">eval</span>(contrast_factor), <span class="at">y =</span> <span class="st">"Expression (normalized counts)"</span>) <span class="sc">+</span></span>
<span id="cb35-26"><a href="#cb35-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">trans =</span> <span class="st">"log2"</span>, <span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fl">0.2</span>)) <span class="sc">+</span></span>
<span id="cb35-27"><a href="#cb35-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span> </span>
<span id="cb35-28"><a href="#cb35-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill=</span><span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="fu">eval</span>(contrast_factor)))</span>
<span id="cb35-29"><a href="#cb35-29" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> ggprism<span class="sc">::</span><span class="fu">add_pvalue</span>(</span>
<span id="cb35-30"><a href="#cb35-30" aria-hidden="true" tabindex="-1"></a>  violin_p_df,</span>
<span id="cb35-31"><a href="#cb35-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">label=</span><span class="st">"{classifyPAdjustMethod(p_adjust_method_selected)} = {violin_p_df[[toupper(p_adjust_method_selected)]]}"</span>, <span class="at">label.size =</span> <span class="fl">3.6</span>,</span>
<span id="cb35-32"><a href="#cb35-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">y.position =</span> violin_p_df<span class="sc">$</span>y.position</span>
<span id="cb35-33"><a href="#cb35-33" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"bottom"</span>)</span>
<span id="cb35-34"><a href="#cb35-34" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename =</span> <span class="fu">file.path</span>(de_plot_dir, <span class="st">"contrast1_FOI.svg"</span>), <span class="at">width=</span><span class="dv">12</span>, <span class="at">height=</span><span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./Assets/de/Plots/contrast1_FOI.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>The searchable table below lists log<sub>2</sub> fold change estimates and p-values for each protein used in the above heatmap.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Change the adjusted P-value column name to its classified adjusted P-value method for display.</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(p_adjust_method_selected <span class="sc">==</span> <span class="st">"none"</span>){</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a> dt1 <span class="ot">&lt;-</span> de_contrast1[, <span class="fu">c</span>(<span class="st">"Feature"</span>, <span class="st">"Comparison"</span>, <span class="st">"Estimate"</span>, <span class="st">"P"</span>)]</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a> selected_dt1 <span class="ot">&lt;-</span> <span class="fu">subset</span>(dt1, Feature <span class="sc">%in%</span> heatmap1_features) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(P)</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a> dt1 <span class="ot">&lt;-</span> de_contrast1 <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(P)</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a> <span class="fu">colnames</span>(dt1)[<span class="fu">which</span>(<span class="fu">colnames</span>(de_contrast1) <span class="sc">==</span> p_adjust_method_selected)] <span class="ot">&lt;-</span> <span class="fu">classifyPAdjustMethod</span>(p_adjust_method_selected)</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a> selected_dt1 <span class="ot">&lt;-</span> <span class="fu">subset</span>(dt1, Feature <span class="sc">%in%</span> heatmap1_features) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(p_adjust_method_selected)</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>results_list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/results.rds"</span>))</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>results_list<span class="sc">$</span>dt1 <span class="ot">&lt;-</span> dt1</span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, <span class="fu">paste0</span>(outdir, <span class="st">"/results.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-acd51990154ebcc17ddb" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-acd51990154ebcc17ddb">{"x":{"filter":"none","vertical":false,"extensions":["Buttons","Scroller","FixedColumns"],"data":[["HIF-1 alpha","TLR7","CD68","PRDM1/Blimp1","Syk (phospho Y352) +ZAP70 (phospho Y319)","Histone H3 (tri methyl K9)","Histone H3 (mono methyl K4)","Histone H3 (di methyl K79)","Histone H3","SHP1","S100A8 + S100A9","HLA-DR","Histone H3 (phospho S10)","MRP8","Histone H2A (acetyl K5)","STAT6 (phospho Y641)","Histone H3 (acetyl K14)","LAG-3","p53 (phospho S392)","Bcl-XL","SLC7A5/LAT1","RPS6","p53 (acetyl K373)","Histone H2B (acetyl K5)","Histone H3 (acetyl K18)","Histone H3 (acetyl K9)","Syndecan-1","iNOS","Topoisomerase II alpha (phospho T1343)","Histone H3 (acetyl K27)","CXCR3","GITR","CD31","PI 3 Kinase p85 alpha","CD39","S100A12/CGRP","c-Fos","Cathepsin S","FAK","IFNGR1","Cytokeratin 1","IL-1 alpha","HLA-DPB1","PI 3 Kinase catalytic subunit alpha/PIK3CA","IL22 RA2/IL-22BP","PTGES2/Gbf1","Cdk6","Progesterone Receptor (phospho S190)","NCR1","IL-15RA","PU.1/Spi1","CCR6","IL-5RA","CD16","Myeloperoxidase","LYVE1","NCAM1","Tuberin (phospho T1462)","Oct4","Insulin Receptor beta","TIM 3","CCL18","ErbB4 / HER4+ErbB2 / HER2","Tuberin","Histone H2A.X","Insulin Receptor alpha","MLH1","CD38","Sonic Hedgehog","T-bet / Tbx21","CXCL5 + CXCL6","Somatostatin Receptor 2","GSK3 beta","Histone H3 (tri methyl K36)","S100A9","ST2","Rho A + B + C","Dihydrofolate reductase (DHFR)","Urokinase","HLA E","NKG2D","MEK1 (phospho S298)","CD11c","CDK1","SOCS3","MMP8","IL-6R","IMP3","PF4","Topoisomerase I","p63","TNF alpha","CD133","CD47","Glucocorticoid Receptor (phospho S226)","CD33","Raptor","ARID3A","Rb (phospho S807)","Granzyme A","Catalase","VCAM1","STAT5 (phospho Y694)","Ms IgG2b","FOXO4/AFX","Tuberin (phospho S1254)","GGT1/GGT","CD15","CD40 Ligand","EGFR","ATG7","2B4","Oct-2","Neutrophil Elastase","SHP2","MS4A2","CDK1 + CDK2 + CDK3 + CDK5 (phospho Y15)","TNFAIP3","Indoleamine 2_3-dioxygenase","HLA A","TRAF6","YKL-40/CHI3L1","RPS6 (phospho S235 + S236)","IGF1 Receptor","Liver Arginase","Ras","p73","Cytokeratin 8 (phospho S431)","SDHA","Vimentin","APG5L/ATG5","CD99","STAT6","CD14","Histone H2A (phospho T120)","ALDH1A1","ErbB4 / HER4 (phospho Y1162)","EpCAM","JNK1 + JNK2 + JNK3 (phospho T183+T183+T221)","Glutaminase","HDAC3","TBR2 / Eomes","CD40","MMP7","SIRT1","MMP1","MMP3","Glucose Transporter GLUT1","LILRB3","Histone H3 (di methyl K4)","SIGLEC5","NQO1","TAK1","Rad51","p53","N Cadherin","IRF4","Insulin","TLR9","PI 3 Kinase catalytic subunit gamma/PI3K-gamma","CXCR5","TLR8","IDH1","Nanog","CD79a","p40 - DeltaNp63","GAPDH","HDAC6","Macrophage Inflammatory Protein 3 alpha","Bak","Cyclin E1","ICOS","Bad","HDAC9","Musashi 1 / Msi1","Thymidylate Synthase","KIR3DL1 + KIR3DS1","NFAT2","HMGB2","beta Actin","p57 Kip2","STAT3 (phospho Y705)","CD79b","AKT1 (phospho S473)","UBE2C","IKKi/IKKe","CXCR4","CXCR6","Cytokeratin 17","PLCG 2","HLA F","CD127","Wee1","Rb (phospho S608)","CCR3","Adenosine Receptor A2a","Mannose Receptor","BRCA1","HuR / ELAVL1","ETS1","PI 3 Kinase p85 beta","MelanA","CD134 / OX40L receptor","PDGF B","Osteopontin","LAMP2A","SMAD5","Hsp90","Topoisomerase II alpha (phospho S1106)","NSD3","Galectin 3","CD64","IL-1RA","Rb (phospho T252)","CD34","CD94","Carbonic Anhydrase 9/CA9","CD11b","CD74","SIRT3","Macrophage Inflammatory Protein 1 alpha / CCL3","Wnt3a","PARP1","Occludin","Syk","CD44","p53 (phospho S46)","ATM","Caspase-3","MEKK3","Cdk4","Transferrin Receptor","ErbB2 / HER2 (phospho Y877)","KDM3A / JHDM2A","Bax","Ikaros","STAT3","Poliovirus Receptor/PVR","TAP2","Src","Heme Oxygenase 1","Prostate Specific Antigen","IKK alpha","CD4","Caspase-10/CASP-10","SOX9","CD103","Interferon gamma","Smad2 (phospho S467)","CDKN2A/p16INK4a","NAK/TBK1","AMPK alpha 1","Rb (phospho T356)","Androgen Receptor","PRAS40","HDAC4 + 5 + 9","Bid","RAGE","IL-22RA1","IL3RA/CD123","Ki67","HDAC1","SIRT1 (phospho S47)","Topoisomerase II alpha","p27 KIP 1","Collagen IV","IKK beta","Melanoma gp100","TMS1/ASC","ERG","CD66b","BRAF","VEGF Receptor 1","C Reactive Protein","eNOS","RPS6 (phospho S240 + S244)","IL-12RB1","c-Maf","TRAF1","cIAP2","HLA B7","STING","15-PGDH","Beclin 1","FGFR2","HMGB1","ICAM1","ATM (phospho S1981)","Hsp27 (phospho S78)","NRF1","Lamin A + Lamin B1 + Lamin C","BTK","CD90 / Thy1","IGF2BP1/IMP1","IL-2RG","Wilms Tumor Protein","FANCI","Smac/Diablo","GRP78 BiP","IL-12A","COX2 / Cyclooxygenase 2","FBP2","p38 beta/MAPK11 + p38 alpha/MAPK14","Smad4","CDKN2A/p14ARF","STAT4","Rel B","LAMP1","p27 KIP 1 (phospho S10)","Nectin 2","Met (c-Met) (phospho Y1349)","Cyclin D1","Chk1","Rb IgG","Cathepsin G","CDK1 (phospho T161) + CDK2 / CDK3 (phospho T160)","Cyclin E2","PTCH2","ENO1 + ENO2 + ENO3","HLA G","beta glucuronidase (GUSB)","CD70","CD11a","CD73","CD177","PTEN","ASS1","S100A4","Cleaved PARP1","ERK1","c-Kit","Lactate Dehydrogenase","Actin","Rho","IL-1 beta","CD86","PAK1+PAK2+PAK3 (phospho S141 + S144 + S154)","Axl","CD28","Aldolase B+Aldolase C","CTLA4","Cytokeratin 8","Rb","VEGFD","SCF","S100P","CD27","CD83","IL-18","KIR2DL1 + KIR2DL2","PAK1","KMT6 / EZH2","Bcl6","Pro Caspase-8","NKG2A + NKG2E + NKG2C","Rt IgG2a","RSK1 p90","ERK1 (phospho T202) + ERK2 (phospho T185)","CREB","IRF3","Cyclophilin A","AKT3 + AKT2 + AKT1","Wee1 (phospho S642)","MDM2 (phospho S166)","PRAME","Hmr IgG","AKR1B10","beta 2 Microglobulin","Histone H3 (phospho S28)","VEGF Receptor 2","Dnmt1","CPT1A","Hexokinase 1","CD3D","ERK2","Cytokeratin 19","CD44v6","Granzyme K","Integrin alpha 5","SMAD1 + SMAD5 + SMAD9 (phospho S463 + S465 + S467)","E Cadherin","FADD","FABP4","Cytokeratin 14","ROS1","ATR (phospho S428)","Cyclin A2","ZAP70","Retinoic Acid Receptor beta","SOD2/MnSOD (acetyl K68)","IL-2 Receptor alpha","Calreticulin","Lck","Fibronectin","alpha smooth muscle Actin","STAT1","JAK3","STAT1 (phospho S727)","IRF5","Chk2","GLB1/Beta-galactosidase","TEF1/TEAD-1","ACE2","Ms IgG1","Von Willebrand Factor","GNLY/Granulysin","DUSP6","SFRP1","CD3 epsilon","Cyclin B1","Carbonic Anhydrase 3/CA3","S100 beta","MEK2","C4a","NFkB p105 / p50","TROP2","RSK1 p90 (phospho T359)","FOXP3","Granzyme B","VISTA","uPA Receptor/U-PAR","KMT6/EZH2 (phospho T487)","Nectin-4","Sumo 1","Cytokeratin 5","S6K1","JunD","MyD88","Podoplanin / gp36","Collagen I","CD63","MMP14","CD163","Glucose 6 Phosphate Dehydrogenase","APC","ALK","SOD2/MnSOD","IKK gamma/NEMO","Fatty Acid Synthase","c-Jun (phospho S63)","MEK4/MKK4 (phospho S80)","Fos B","SQSTM1 / p62","AIF","GLP-1R","TOMM20","PCNA","ErbB3 / HER3","c-Myb (phospho S11)","CD18","c-Myc (phospho S62)","Scavenging Receptor SR-BI","Hsp27","p38 (phospho T180)","CD276","GSK3 beta (phospho S9)","PKC beta 1","CD3G","STAT5a","MSH2","Glutathione Peroxidase 4","PDGFR alpha + PDGFR beta","Progesterone Receptor","CD45","Histone H3 (tri methyl K27)","ERK1 (phospho T202 + Y204) + ERK2 (phospho T185 + Y187)","Smad2 (phospho S255)","BRG1","active YAP1","Rb (phospho T780)","Notch1","beta Catenin","PSMA","pan Cytokeratin","Ret","Pan Trk","Cullin 1/CUL-1","DR5","CD20","ROCK2 + ROCK1","NF-kB p65","FOXO1A","CCR7","Integrin beta 1","Integrin alpha 6","Syk (phospho Y323)","Smad2","Cathepsin K","Bim","STAT5b","Ubiquitin","EGFR (phospho Y1068)","IFIT1","CXCL10/IP-10","Brd4","M-CSF","LEF1","IDH2","Argonaute-2","IL-33","15 Lipoxygenase 1","ATF2","YAP1 (phospho S127)","VEGFA","C5 / C5b","Caspase-3 p12","Hsp70","Periostin","CCR5","FGF2","IL4","ATF3","gamma H2A.X (phospho S139)","mTOR","Rb (phospho T373)","MX1","PD-L2","LC3B","mTOR (phospho S2448)","ATG12","ROCK1","GATA3","beta Tubulin","CREB (phospho S133)","AKT1 (phospho T450)","PD-L1","SFRP4","Aurora A","MICA + MICB","CD8 alpha","c-Jun","Tissue Factor","PDGFR alpha","c-Myc","MUC1","CD22","CD36","MEK1","HDAC2","ErbB2 / HER2","Met (c-Met)","EMR1/ADGRE1","Cdk2","SIRT2","BFL-1/GRS","ISG15","GLP-1","Hexokinase II","MSH6","Aldolase","CREBBP","TGF beta 1","PP2A alpha + beta","5 Lipoxygenase/5-LO","beta Arrestin 1","STAT2","Bcl-2","KIR2DL3","TMPRSS2","TRAF2","CD45RA","LDL Receptor","RUNX1 / AML1 + RUNX3 + RUNX2","CD272/BTLA","ARID1A","Smad3 (phospho S423 + S425)","PD1","Smad3","HLA-C","Fas","Caveolin-1","IKB alpha","Survivin","CSF-1-R","Mucin 5AC","CDK1 + Cdk2 + Cdk3 (phospho T14)","Monoamine Oxidase A/MAO-A","MMP9","PTEN (phospho T366)","YAP1"],["No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes","No vs Yes"],[0.6611207754493321,0.4063802434960061,0.801108211617142,0.2503537162683815,0.2981994748665276,1.059430856734876,0.8264992080772733,0.8525453167572088,0.6031747933129367,0.2780353357335477,1.458980088072234,1.529845760608016,1.100751404331,2.090149800480897,0.875016754966534,0.3813821328325142,0.9358025752493909,0.5338855824525087,0.6422059297352777,0.5470630595579358,0.6163051271025128,0.3893147239873748,0.5837248889018072,0.5400349949434959,0.6995746494854306,1.173557188125007,1.52517666795677,0.696267905458094,0.5101531405058584,0.5970839115685861,0.4506172803836964,0.5941731818311772,0.487000642544647,0.414992430549255,0.4862752452629211,0.7096559628882039,0.3261205682762813,0.983611325466066,1.309165216593301,0.3639841819752231,0.320243390792643,1.11814428972961,0.8853777650513588,0.6143173650779766,0.2873647587674098,0.4169274138672563,0.6947405256665959,0.377851862111298,0.3539518437723115,0.5126892101710415,0.3954002229667518,0.7384186453641758,0.3248501703481297,1.076698146909489,0.2583678108136199,0.3746174750958853,0.3090045041951373,0.2483569058734132,0.3420513892359823,0.4807492164161655,0.3896484286046538,0.2988883560267738,0.3934520385455633,1.014440704131324,0.6917974613916209,0.4356512367638897,0.3952156168432663,1.051808939358483,1.070059146268154,0.3223442709046447,0.363344428820141,0.3090924220595924,0.3865162681525688,0.3569372697021697,0.8759956953925342,0.2347859691103093,0.3426731032779705,0.714736245304143,0.3766598085947905,0.6614529644335673,0.2484043609528421,0.3262073512254265,0.3702440300176859,0.5969211171929162,0.3251590631002373,0.2633125236163727,0.2436950845231167,0.2560488381583353,0.2675067544788792,0.3733822314692042,-0.3687286183345732,0.2708975767448453,0.507314175063744,0.3860848526061013,0.372343826164518,0.3322170804465169,0.2752478366786451,0.2971479284427015,0.3306748127744324,0.2620460694877796,0.588721944608888,0.3657814278424805,0.4574445446676403,0.07571573851070877,0.2388630397900357,0.5585619175529816,0.3332718814219012,-0.5146003948777734,0.2071825487507813,0.3910292018642901,0.5251507343943804,0.2335675058599201,0.2705443719101355,0.2347422507347381,0.2596528608891097,0.3188875703640469,-0.2289915924441373,0.3875194637074704,0.4985162945415785,0.9496318711031071,0.2545186469199519,0.2259707940346386,0.5224339202071654,0.2839508799700628,0.1865464138560672,0.2999145720551998,0.25674310932496,-0.5201096045030209,0.3141463191909361,0.7550060972507426,0.1880106466234951,0.2388143609133363,0.462521033971672,0.3933086580155028,0.3760151376754904,0.2798837016472128,0.2641134628057177,-0.7988024950279691,0.2189260959428848,0.6696164034310036,0.2688470112405651,0.2262701428327917,0.5014430246483041,0.2557740256119637,0.2458456734027595,0.1803798770837597,0.2704950231410339,-0.6296353108683508,0.2487633684459481,0.4636313645426101,0.2645226685088564,0.5782810123594742,0.1899113242056886,0.2267403224263966,0.2214911702075115,0.2184798452525123,0.3267471695609942,0.2204561834439543,0.2342200876845875,0.2980726016178446,0.2613380384010473,0.2593875088181912,0.3704126414340834,0.1800511384867812,0.2439074824557407,0.2115830328696549,-0.3905868625430045,0.3619785670728617,0.2002532590505321,0.6637169011547523,0.3208487124822199,0.243072127762589,0.2646977389811755,0.335553979886363,0.1170424492127879,0.284825248256126,0.2145264519035529,0.6000393395805329,0.4329690722210218,0.3517644694746358,0.3009333383810445,0.2645891188845575,0.3395704048544039,0.2353603839326599,0.1927162061995996,0.461460375926366,0.2211092965458355,0.2072775947566081,0.1585995222126484,0.2748875118904003,0.1424646183092159,0.2201864734333262,0.254152661229447,0.273640187990866,0.1878619965059014,0.2800224363489603,0.4997015963938379,0.2369285129532804,0.2964651885720548,0.2152877503403829,0.1762608863465273,0.1881803029246116,0.2444483750319216,0.4840140928339645,-0.3588842950208981,0.34618872376743,0.2711443793095283,0.2043328636798865,0.186122470907622,0.2682415878821493,0.6430817664385824,0.2538446864507237,0.1582584229179302,0.2187447245107896,0.2835169922975136,0.1717189564043499,-0.3118546137552302,0.1559827974315673,0.5549041230427262,0.2087277030667908,0.1642107806255544,0.1088436267702086,0.2557526943112788,0.1383082981985598,0.2248233607326344,0.3633521781226555,0.2138662782065762,0.2213672115891216,0.2447072451225126,0.1786089125350002,0.2198152352395236,0.2346051338472279,0.202373003130577,0.1396715715052371,0.2878542306732227,0.4859149900460588,0.2492432967822046,0.2030326732449208,0.8399030456524708,0.2376413636537645,0.3915364650672782,0.1577299357701743,0.2817757237102555,0.2658284697768623,0.2666382760744213,0.2827685950981214,0.158030939798471,0.1300906995524344,0.2897149477367857,-0.1740527914278348,0.2431100814915534,0.1753906903811254,0.1650321190456085,-0.3411433486577142,0.1651779056979582,0.3099874180331064,0.2338231862358491,0.141153287404925,0.1604355265671726,0.1967523962286649,0.1931049065380444,0.1480777244183566,0.2502480314974256,0.2377855839840341,0.4400162076994395,0.1564753422965113,0.147845514593569,0.1464953294982892,0.5542481693405173,0.2405447250989259,0.2318399994380467,0.2102544599837384,-0.2806523430133966,-0.1533935169259328,0.228606018719903,0.3483832274527814,0.2302151948661577,0.1204864230128621,0.2691607831085494,-0.1473708580044384,-0.3094032303707837,0.3875527438968631,-0.245092470895429,-0.201626227807541,0.1398738666200524,0.3662105235270547,0.1518029767209569,0.2485145212258362,-0.3469674138212699,0.4495029083053995,0.3164864771915233,0.1646156981420853,0.3160365546061102,0.1206235782749046,0.1913566391127973,0.08508915387508972,0.1472679184507549,0.2082341788537648,0.1092868792889259,0.150468542186996,0.07787014135175213,0.09973501687252316,0.09952016750116656,0.1408377600080212,-0.1199621952208774,0.2011060310048744,0.235336827532711,0.2166681594355362,0.1841253628204321,0.1555617551992186,0.2218721721371815,0.1306842427676562,0.1208064845242763,0.09580078907624276,0.3695761020623664,0.3665190329494137,0.1038277085378029,0.1870749576934109,-0.127103444253995,0.2229303416769793,0.5092349010702023,0.1078919143342565,0.1644952144072356,0.1927358343774131,0.1303268378604147,0.1527132898236154,0.1439569508125654,0.2847067249042102,0.1099216302317061,0.2586528697591192,0.1318953580960342,-0.2138175211648528,-0.1713486145424601,-0.1684017614977467,0.2241054530214206,0.1226095789329733,0.2496820465998434,-0.2047736699899232,0.1020716713776804,0.1102623227316995,-0.09923313383483161,-0.2781977766564893,0.09620625951688357,0.1061819122489787,0.1355063306724123,0.08567642321726621,0.2046696667522278,0.1063351680286763,0.1357217767296339,0.1051261064881273,0.1440666906661054,0.1594689627805664,0.09584881607417751,0.1223889134984669,0.1180247620481135,0.06954180984996834,0.1054742228289941,0.08577723983087306,0.1707174506330416,0.2651052976089389,-0.2407564489654498,0.1959409786183495,0.107496872750607,0.1513201475856141,0.1020680147033129,-0.1647652217945831,-0.113849692003514,0.511240639112569,0.2894369292905088,0.1437576169181825,0.09807329340635093,0.1105851690345196,0.174845827742082,0.2507133639610955,0.2845541417766144,-0.3449737738109118,0.138060972321995,0.07747658465902897,-0.2034602049460529,0.1962729262624477,0.3833233674752749,0.1097653239886961,-0.2080081349764023,0.08320464873726659,0.06511137812230688,0.1196820498800921,0.08453699908766796,0.2432701509747909,-0.0741088291036455,0.204846046142154,0.08425121812084606,0.1768632549096263,0.1698969620892861,-0.5803211271804183,-0.3741918696467527,0.3984795635623501,0.06970437897965996,0.3276217791915582,0.07666718486632292,0.1954692023105089,0.2727713900052515,0.08421641584284553,-0.1967989874560089,-0.04766546944256835,0.1523640649499595,0.08779553800320851,0.06934648229523639,0.1045012248635073,0.1859623974148101,-0.08016930803957246,0.1009972766762631,0.1918651854905557,0.1926941870849958,-0.1976958443259795,0.2920223890947448,0.1962279136661107,0.1350692200942802,-0.05917561863657374,0.07506789770089203,0.1718936295846905,0.1235017989451681,0.1505007719465417,0.08152239413743276,0.09862847641804202,-0.2241703506254126,0.1285920420388992,0.1499635771736707,0.09571851838470817,0.1304023224799259,-0.3820390500421316,0.1503519193324136,0.07968264152735642,0.07740987230617398,-0.1186655541374918,0.07404714114807297,0.06257422152999659,0.1804540876663084,-0.07210965633153901,-0.189037672475997,0.05824807992444954,0.1231962302748421,0.06835680002247894,0.239101778528457,0.09231476852394725,0.07278250532284332,0.1383066949730552,-0.163868373994617,-0.09605768249515508,0.1089598231344294,-0.04386459836051232,0.06368206141914635,0.06614291054477428,-0.1638583317116601,0.1081872670685859,-0.2228211694358843,0.05217897152069795,0.06740047440082701,-0.1268456440169886,0.201654489250103,0.08908118122433453,0.09242673839245749,0.1218943891569391,-0.08417231348519881,0.08866626467047162,0.04102235350751094,0.1173743544343772,0.07544035396898757,0.06764562884419094,-0.1329166473351556,0.08466727381708646,0.05095736612403616,0.1600501322484576,0.1203509875985795,0.06887248374055137,0.04083676120542599,0.06926659999394739,0.04969120992067235,-0.05728677430793325,0.2675296983109366,0.06326735435259857,0.09882667405410578,-0.07167141436825399,0.05529101810545854,-0.0540083611952237,-0.06717780280087676,0.05964324495148499,0.06548836179829315,0.04952409687797633,0.04251166346891094,0.1565008628276663,0.05398255420129239,-0.124006524617687,-0.1221563092656626,0.04002762006962609,0.1289280139478512,0.05511951915075902,0.04410025203143405,-0.09511642642531148,-0.05305592702854776,0.04963973250072419,-0.04344965838361698,0.05179354901223202,-0.0636527413490425,-0.06403346597433306,-0.05186610977163927,-0.03605107859971941,0.04686448942910844,-0.1433083557986625,0.0503725827571467,-0.03163612139257563,-0.06108918103918796,-0.04838055317196344,0.05178748386201899,0.03277962892827337,0.03761972968180687,0.1238799357080468,0.02786788244956926,0.03392772808897554,0.04622179026973147,0.04467211279811032,0.05093404543987923,0.03751402534624352,0.05977429172985631,0.08006219257652179,0.03302031613873639,-0.02588419161909808,0.09240823637353454,0.0354230636956353,-0.01251667027941855,0.03863806642117843,-0.03535305983109263,0.02366267169065592,-0.04075153671986891,0.02494396936307729,-0.06719583403933033,-0.02691318408190224,-0.02673304441581857,0.03972598895381791,0.04936863265311226,0.1014367572007782,-0.02593315434928361,-0.02540881170847065,-0.02553015361231586,0.0169175451747052,-0.02315930550884062,-0.03329694487747632,0.011577376423488,0.009381669069421002,0.02418235194317854,-0.0200197410135064,0.0259894027991663,-0.009377476024069379,-0.02367728905517069,0.01671496250764277,-0.01629615539163025,-0.01569015297738431,-0.01964702411119663,-0.0129382093458606,0.007199411755434077,0.01060544335608019,0.01246812728881315,-0.0149034885481886,-0.02221804432439081,0.01323883428352591,0.01403638122723559,0.01259067674587019,-0.01067567804867535,-0.009499665954761392,-0.02934982075053132,-0.004655150774308912,-0.01695720503376364,0.008530164990706966,-0.003007544322027126,0.007345368999063638,0.002371118741552133,0.002752349110520746,-0.004786821753800637,0.001289569604860186,-0.001149018118406728,0.0008069056834506879],[8.253907865882999e-07,4.356055074431498e-06,2.395394249303221e-05,0.0008405693219355069,0.003849063283017301,0.01940777788048553,0.0326598162028711,0.04103069114624449,0.04107731251020114,0.04397006893883364,0.04679354573686929,0.05596371487617147,0.075242877871967,0.07525226274630711,0.07538467718866361,0.07642015023026966,0.07980949537565796,0.08897219560378621,0.08967419146873315,0.09231775022001415,0.09300223456308722,0.09307702408398803,0.09827199890950977,0.09975623906154364,0.1003309462771187,0.1009244118974138,0.1017846801692339,0.1067698917061961,0.1074601082236949,0.1105606138761503,0.1167167877561992,0.1233392335573399,0.124621735984204,0.1271634198090999,0.1318865229210586,0.1324597570307968,0.1333922567348344,0.1344541672694249,0.1368721419225838,0.1369157979626655,0.1411179192647149,0.1457433017951265,0.1461052596345198,0.1465652113989398,0.1487398274002805,0.1492562189933584,0.1547261364838134,0.1550547141050011,0.1590358072930989,0.1624466059853099,0.1653656920463664,0.1671728193407855,0.168227303091967,0.1689658647831241,0.1805817765318972,0.1809054014647609,0.1824535313348958,0.1843188478244256,0.1907528920103198,0.1922117852428037,0.1938652157380856,0.1948141802330094,0.1954854521343173,0.1965080062416971,0.1982101540804746,0.1990560042749212,0.1993233775489432,0.202341613925523,0.2110435235388718,0.2174566447369809,0.2186821259135988,0.2195553158337383,0.220502283248202,0.222139164214146,0.2225420973700383,0.2232446062089606,0.2244169769787878,0.2327774506990082,0.2356963665904668,0.2384811972924115,0.2422524480479346,0.250938413887314,0.2519959439089235,0.2543084485523937,0.2545163827581403,0.2634249129155245,0.2658862314023338,0.266076963295345,0.26671032183841,0.2680363657540854,0.2682014382691137,0.2686661961105934,0.2697774585583923,0.270017109685061,0.2753842761703305,0.279024304004625,0.2803816087149213,0.2807364197953215,0.2826177646767407,0.2832435481598863,0.2836703766738087,0.2843785993465262,0.2847674187475301,0.2890205802631129,0.2897012544134029,0.2922927523622777,0.293998390715266,0.2969931251116102,0.2976519088305853,0.2980555911567648,0.2994255412310627,0.3001540270510539,0.3025944503824214,0.3030489011351628,0.3042591225188526,0.3053107086905596,0.3060793183916632,0.3086165948128725,0.3126620594205187,0.3183501194453107,0.3216591025520797,0.3217153873957991,0.3245874473633672,0.3258509640239542,0.3263069116935605,0.3264886009507298,0.3301734119560312,0.3310718071380773,0.3349374243641121,0.3357237276813601,0.3363193930671623,0.3390990020142181,0.3396910639184018,0.3420357627994181,0.3424533511789892,0.3433511106270437,0.3461779469770102,0.3479869189984884,0.3482890543254262,0.3488407925517154,0.3510670781783521,0.3555560575837872,0.362496292652207,0.368737733044415,0.3695803693722774,0.3723276418273625,0.3734078345278914,0.374133030708202,0.3741659493799555,0.3750220486178661,0.3767469286552986,0.3799250385348377,0.3805399645228412,0.3829940284479599,0.3862940036041933,0.3864317879929425,0.386556360049681,0.386696114562669,0.3874419354380714,0.388246173700407,0.3883267129073854,0.3893580360916399,0.3902761277223328,0.3983148857101633,0.3984248385648325,0.399250670458366,0.3995111080061663,0.4013512199492934,0.4019815750437943,0.4044662769419042,0.4051414349795806,0.4052822981908253,0.4060640636733032,0.4112819917678002,0.4160169460906795,0.4161243185700079,0.4161679409725707,0.4164738962910441,0.4178445539927914,0.4196683564594127,0.4198279953076918,0.4205690760273892,0.4210991515994138,0.4227305552171706,0.422883506676368,0.4231669378581369,0.4236634558181532,0.4237364637876344,0.4249716079952796,0.4267917468583213,0.4280037965708219,0.4291271426576558,0.4293929499575073,0.4304130371606338,0.4304824930508689,0.4314732279533403,0.4334543881217155,0.4341014917626108,0.4342637547036665,0.435498919586149,0.4358182021784635,0.4360676389466543,0.4363489735254126,0.4415109921642733,0.4416577301527359,0.4425862159484288,0.4484724887624797,0.4522038979000675,0.4603099113953978,0.4607233277498751,0.4612218762343154,0.4616576993779254,0.4638261938380622,0.4650982559675215,0.4653629119811786,0.4653909626576818,0.4655143084484215,0.4667428470423064,0.4688820019672272,0.473169930649574,0.4732478092835192,0.4735017834296681,0.4735629312614515,0.4737894778710252,0.4753833989069817,0.4754913845909451,0.4797676438425507,0.4803244554607872,0.480373231721083,0.4813134028531985,0.4813952907132538,0.4830870693085684,0.4850549170541631,0.4856699546111666,0.4861893919217333,0.4868355450614434,0.4882565502535554,0.4908490006647881,0.4930822048492572,0.4938475051212629,0.4948618817922239,0.4949055794272872,0.4956507797742402,0.4957552575708673,0.5011061995439462,0.5036909092404422,0.5041818600751601,0.5063818363512373,0.5069262459541061,0.5070995193129902,0.5099550246054603,0.5103043583710168,0.510617412492629,0.5126120443847652,0.5134527547924698,0.5153770669246746,0.5156804424267729,0.5172095579928161,0.5173440229876576,0.5193953488266553,0.5212485203095414,0.5217204020683475,0.5220321412121569,0.5238707420995321,0.5261625669296027,0.5275011854548235,0.5314832366714193,0.5317143487976977,0.532128067984818,0.532345651040346,0.5360210135641027,0.5371161193708566,0.5383735375771125,0.5393743333300355,0.543065626317862,0.5446718539717357,0.54761314248382,0.5490554664594677,0.5516836735207782,0.552017297332059,0.5530368069113518,0.5575114619018773,0.5589109561479304,0.5622275191571595,0.5653223005330106,0.5686371357394259,0.5748734922310379,0.5822854548406716,0.5867354879805926,0.5874029255666822,0.591265353744113,0.5945844713518009,0.59473801380158,0.5984802554582254,0.5991045551898772,0.5994275362606072,0.6010674463202824,0.6017724300165241,0.6018089271847777,0.6020331428224042,0.6022624608538354,0.6024923287667059,0.6051331378542252,0.6056319782510823,0.607236942609768,0.6077484214608286,0.6082657985745386,0.6089282342098723,0.609209771511417,0.6095733479359886,0.6122259801859511,0.6124432013085628,0.6130010701747708,0.6191899046455011,0.6195212924365986,0.620191532423606,0.6216103973955924,0.6231092370997833,0.6250646729367509,0.6256764525413305,0.6274265759263371,0.630147437151106,0.6314244689923278,0.6315165217588468,0.6320880993772836,0.6323495744087588,0.6353899192884694,0.6363866715482015,0.6366095467675672,0.6380170897040653,0.6381109281516071,0.6401726582304106,0.6412346420533295,0.6448073184968569,0.6494779364787506,0.6511153940917813,0.6512767450551664,0.65170911634831,0.6517964516248553,0.6528181782745978,0.6537924369817556,0.6550433378219251,0.6561007243978706,0.6568447818894479,0.6570339166747079,0.6570341735555041,0.6600651828961758,0.6633883339574629,0.6659491071690806,0.6664500528522209,0.6668373970538665,0.6689806492709882,0.669519192707189,0.6722953658527739,0.6735101014551922,0.6754981305494476,0.6782957518393475,0.6796822805783609,0.6797112597875095,0.6800555832943247,0.6801107330369296,0.6806926726388181,0.6824008189797874,0.6824917085920043,0.6827319101774162,0.684580217865715,0.6847586268312102,0.6857910707080574,0.6864282286470033,0.6882502884146625,0.69119255631583,0.6930476322562856,0.6935127906569589,0.6935855227145007,0.6963595528377342,0.6964876574483897,0.6970383823524197,0.6975593685423352,0.700968418224624,0.7026554677067305,0.7036257591366859,0.7038307686610792,0.7055948911842272,0.7073487821604523,0.7075065842746739,0.710089056225931,0.7121510008943268,0.7141371652697148,0.7145420155819369,0.7153410669507592,0.7160021944970129,0.7173059898619626,0.7188083955828791,0.719386262933341,0.7199320239200582,0.7265315209374166,0.7268709823730045,0.7272739374350811,0.7290872995913833,0.7302123387998192,0.7315686026886129,0.7328830421956996,0.7336757167643172,0.737225498352042,0.7384804779990234,0.7405289302365077,0.7441348291782071,0.7455540915156516,0.7485104844346122,0.7530884517423222,0.75474014202958,0.7551708609756087,0.7582514658540472,0.762530424502877,0.7651817614139227,0.765783294754907,0.7662487582167479,0.7662902738389885,0.7665448462596693,0.7683568641779406,0.7693946800695851,0.7694984247854697,0.7697447135069835,0.7706917859260781,0.7727669118109888,0.7730735479794156,0.7736643815373169,0.7737423838505955,0.7737876988121902,0.7741413960231778,0.7742689484712775,0.7784222210126386,0.7809589708736406,0.7836190765337256,0.7850355096239844,0.7853015369186344,0.7937544326301933,0.794014350066953,0.797741949615286,0.8010724303950993,0.8040948984275134,0.8045614993927517,0.8048623462491397,0.805596430558296,0.8080597700329017,0.8084335742145894,0.8090011051201825,0.8096621662743952,0.8104307243086032,0.8108581072603689,0.8122882324708051,0.815518731489106,0.8162041329559775,0.8183318645594312,0.8189464233529206,0.8197724956308987,0.8221697908801544,0.8238473662188505,0.8247110882490876,0.8268189079750923,0.8271213389967174,0.8277768732973928,0.8307285678462719,0.8314503020174119,0.8419207659607935,0.8419736579075086,0.8424318992159582,0.8432301714735367,0.8441187571249125,0.8442744102119922,0.8448375080142584,0.8457480724319377,0.8478463057921035,0.848975716602909,0.8494067035653894,0.8507680697178284,0.852424559594577,0.8570439618432705,0.8609100085352859,0.8616072765003222,0.8620531532357459,0.8637259252141698,0.864371530408812,0.8644911028561356,0.8649907414915124,0.865712035058574,0.8675658010328244,0.8693615667842323,0.8716750423181312,0.872106559359128,0.8725688593228056,0.8764805736010083,0.8785759968614935,0.8797683879422241,0.8822387410954706,0.885593068759949,0.885792341604842,0.88730511234607,0.8883426515012358,0.889869584828708,0.8899768628709125,0.8905001040863582,0.8907670839891528,0.8908378123534438,0.8943519530198935,0.894954288278067,0.8957642576126225,0.8964083178192064,0.8972586026168966,0.8973083054136686,0.8999791311329383,0.9006301425992134,0.9033371299937933,0.9034326730282806,0.9042549906201188,0.9048035512792262,0.90638577312387,0.9068970892985556,0.9070223796862823,0.9090704748368598,0.9099149025225896,0.9107499250231736,0.9178575628654616,0.9178765521577654,0.9181789547956851,0.9204228581576052,0.9205378882786726,0.9248882905101045,0.9268437135642885,0.9291788605577278,0.9295132474839907,0.9320899140503964,0.9339470502427062,0.9340447693691427,0.9357790449663665,0.9357880639172933,0.9378362937966813,0.9385106113806434,0.9419460264483479,0.9435824964044039,0.944723570942656,0.9462154637121627,0.950767895720494,0.951574965253846,0.9541844615738355,0.9557421237690407,0.9557689263261803,0.9562334006657176,0.9572046614103871,0.9587651164104521,0.9594995574823789,0.9601087142766813,0.9608582338276317,0.9633262110072029,0.966631700155518,0.9682749988613152,0.9683023338899421,0.969593493125578,0.9726004354808423,0.9726372743414865,0.9727940435841254,0.9727965434226473,0.9729153381950107,0.9733875216071542,0.9742688925726883,0.9766046416170575,0.9780709784334094,0.9784383089647872,0.9795280528040801,0.9812971755656376,0.9815797817998061,0.9827343088036202,0.9877478248745364,0.9894266278706182,0.9896688777849489,0.9903412213452722,0.9905317310158708,0.9960557140996535,0.9976727071399817,0.9989937653917627],[0.0004787266562212139,0.001263255971585134,0.004631095548652894,0.1218825516806485,0.446491340830007,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9956818093400434,0.9989937653917627,0.9989937653917627,0.9989937653917627]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>Feature<\/th>\n      <th>Comparison<\/th>\n      <th>Estimate<\/th>\n      <th>P<\/th>\n      <th>FDR<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":3,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":4,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":2,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"className":"dt-right","targets":[2,3,4]},{"name":"Feature","targets":0},{"name":"Comparison","targets":1},{"name":"Estimate","targets":2},{"name":"P","targets":3},{"name":"FDR","targets":4}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render","options.columnDefs.2.render"],"jsHooks":[]}</script>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Overall, we see a handful of proteins are differentially expressed between these two groups. In the heatmap we see that these proteins are clustering primarily by responder.</p>
</div>
</div>
</section></details>
</section>

<section id="sec-conclusion" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Conclusions</h1>
<p>Together, these steps provide a reproducible framework for processing, exploring, and interpreting GeoMx protein expression data. By establishing data quality, normalization, and early biological patterns up front, the workflow sets a strong foundation for downstream differential expression and biological insight.</p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Merritt2020" class="csl-entry" role="listitem">
Merritt, Christopher R., Giang T. Ong, Sarah E. Church, Kristi Barker, Patrick Danaher, Gary Geiss, Margaret Hoang, et al. 2020. <span>“Multiplex Digital Spatial Profiling of Proteins and RNA in Fixed Tissue.”</span> <em>Nature Biotechnology</em> 38 (5): 586–99. <a href="https://doi.org/10.1038/s41587-020-0472-9">https://doi.org/10.1038/s41587-020-0472-9</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/nanostring-biostats\.github\.io\/CosMx-Analysis-Scratch-Space");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="../../license.html">
<p>License</p>
</a>
  </li>  
</ul>
    <div class="cookie-consent-footer"><a href="#" id="open_preferences_center">Cookie Preferences</a></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>