<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Patrick Danaher">
<meta name="dcterms.date" content="2025-12-11">

<title>Basic Analysis Vignette: current best practices for QC, normalization, dimension reduction and cell typing – Blog</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/logo_Bruker_black.png" rel="icon" type="image/png">
<script src="../../site_libs/cookie-consent/cookie-consent.js"></script>
<link href="../../site_libs/cookie-consent/cookie-consent.css" rel="stylesheet">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-3a01e2046221230fdceeea94b1ec5d67.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-a4a11d514c7d463668e07712114998e6.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-8dd817d487e8e26707729065bc6b7e6d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-29W4MW0Y2W"></script>

<script type="text/plain" cookie-consent="tracking">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-29W4MW0Y2W', { 'anonymize_ip': true});
</script>

<script type="text/javascript" charset="UTF-8">
document.addEventListener('DOMContentLoaded', function () {
cookieconsent.run({
  "notice_banner_type":"simple",
  "consent_type":"implied",
  "palette":"light",
  "language":"en",
  "page_load_consent_levels":["strictly-necessary","functionality","tracking","targeting"],
  "notice_banner_reject_button_hide":false,
  "preferences_center_close_button_hide":false,
  "website_name":""
  ,
"language":"en"
  });
});
</script> 
  


</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../.././assets/logo_Bruker_white.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../license.html"> 
<span class="menu-text">License</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../link-to-code.html"> 
<span class="menu-text">Code</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">Browse Posts by Topic</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/company/bruker-spatial"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Basic Analysis Vignette: current best practices for QC, normalization, dimension reduction and cell typing</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">recommended</div>
                <div class="quarto-category">quality control</div>
                <div class="quarto-category">cell typing</div>
                <div class="quarto-category">normalization</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliations</div>
    
      <div class="quarto-title-meta-contents">
      <p class="author">Patrick Danaher <a href="https://orcid.org/0000-0002-2844-5883" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              Bruker Spatial Biology
            </p>
          <p class="affiliation">
              Github: <a href="https://github.com/patrickjdanaher" target="_blank">patrickjdanaher</a>
            </p>
        </div>
    </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">December 11, 2025</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">December 3, 2025</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1">
    <h2 id="toc-title">Contents</h2>
   
  <ul>
  <li><a href="#new-best-practices-for-cosmx-analysis" id="toc-new-best-practices-for-cosmx-analysis" class="nav-link active" data-scroll-target="#new-best-practices-for-cosmx-analysis"><span class="header-section-number">1</span> New best practices for CosMx analysis</a>
  <ul class="collapse">
  <li><a href="#basic-analysis-strategy" id="toc-basic-analysis-strategy" class="nav-link" data-scroll-target="#basic-analysis-strategy"><span class="header-section-number">1.1</span> Basic analysis strategy</a></li>
  <li><a href="#basic-analysis-strategy-large-studies" id="toc-basic-analysis-strategy-large-studies" class="nav-link" data-scroll-target="#basic-analysis-strategy-large-studies"><span class="header-section-number">1.2</span> Basic analysis strategy: large studies</a></li>
  <li><a href="#outline" id="toc-outline" class="nav-link" data-scroll-target="#outline"><span class="header-section-number">1.3</span> Outline</a></li>
  </ul></li>
  <li><a href="#loading" id="toc-loading" class="nav-link" data-scroll-target="#loading"><span class="header-section-number">2</span> Loading flat files</a></li>
  <li><a href="#xy" id="toc-xy" class="nav-link" data-scroll-target="#xy"><span class="header-section-number">3</span> Customizing the tissues’ spatial arrangements</a></li>
  <li><a href="#qc" id="toc-qc" class="nav-link" data-scroll-target="#qc"><span class="header-section-number">4</span> QC</a>
  <ul class="collapse">
  <li><a href="#per-cell-qc" id="toc-per-cell-qc" class="nav-link" data-scroll-target="#per-cell-qc"><span class="header-section-number">4.1</span> Per-cell QC:</a></li>
  <li><a href="#fov-qc" id="toc-fov-qc" class="nav-link" data-scroll-target="#fov-qc"><span class="header-section-number">4.2</span> FOV QC</a></li>
  <li><a href="#descriptive-qc-plots" id="toc-descriptive-qc-plots" class="nav-link" data-scroll-target="#descriptive-qc-plots"><span class="header-section-number">4.3</span> Descriptive QC plots</a></li>
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization"><span class="header-section-number">4.4</span> Normalization</a></li>
  </ul></li>
  <li><a href="#dimred" id="toc-dimred" class="nav-link" data-scroll-target="#dimred"><span class="header-section-number">5</span> Dimension reduction</a></li>
  <li><a href="#clust" id="toc-clust" class="nav-link" data-scroll-target="#clust"><span class="header-section-number">6</span> Cell typing</a>
  <ul class="collapse">
  <li><a href="#naming-clusters" id="toc-naming-clusters" class="nav-link" data-scroll-target="#naming-clusters"><span class="header-section-number">6.1</span> Naming clusters</a></li>
  </ul></li>
  <li><a href="#clustqc" id="toc-clustqc" class="nav-link" data-scroll-target="#clustqc"><span class="header-section-number">7</span> QC of clustering results</a></li>
  <li><a href="#hieratype" id="toc-hieratype" class="nav-link" data-scroll-target="#hieratype"><span class="header-section-number">8</span> Optimizing immune cell calling with the HieraType algorithm</a>
  <ul class="collapse">
  <li><a href="#merging-clustering-and-hieratype-results" id="toc-merging-clustering-and-hieratype-results" class="nav-link" data-scroll-target="#merging-clustering-and-hieratype-results"><span class="header-section-number">8.0.1</span> Merging clustering and HieraType results</a></li>
  </ul></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion"><span class="header-section-number">9</span> Conclusion</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="new-best-practices-for-cosmx-analysis" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> New best practices for CosMx analysis</h1>
<p>Here we propose an updated pipeline for basic CosMx<sup>®</sup> analysis in R, from data loading through cell typing. This post is the first in a series; coming soon will be posts on exploratory analysis, on differential expression analysis, and on pipelines for large studies. This workflow assumes a whole study can be fit into your compute environment’s memory. For very large studies, see our upcoming vignette on the topic, or the summary below.</p>
<p>This post is exclusively in R. Many steps have easy python analogs, but some, like the hieraType immune cell typing algorithm and the FOV QC functions, only have R implementations.</p>
<section id="basic-analysis-strategy" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="basic-analysis-strategy"><span class="header-section-number">1.1</span> Basic analysis strategy</h2>
<p>Analyses can meander and sprawl, but often enough they are reducible to 3 phases:</p>
<ol type="1">
<li><strong>Basic pre-processing</strong>: Prepare your data for deeper analysis: perform QC, normalize your data, perform dimension reduction, and finalize your cell typing. This vignette covers this workflow.</li>
<li><strong>Exploratory analysis</strong>: See what unexpected trends your data holds. Ideally, these trends can be phrased as formal hypotheses to be tested. (Post coming soon.)</li>
<li><strong>Hypothesis testing</strong>: In our experience, most analyses culminate here, mainly taking the form of differential expression (DE) testing. DE is a flexible tool to interrogate both the hypotheses you designed the study around and new hypotheses discovered in exploratory analysis. We see DE as the workhorse of spatial analyses, but there are other useful tools: e.g., you might calculate changes in cell type abundance or co-localization across space, or compute various metrics of ligand-receptor interactions. (Post coming soon.)</li>
</ol>
</section>
<section id="basic-analysis-strategy-large-studies" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="basic-analysis-strategy-large-studies"><span class="header-section-number">1.2</span> Basic analysis strategy: large studies</h2>
<p>When your data is too large to fit into memory – and this can happen with just a few slides of WTX data – you need to modify the above strategy. We propose the below approach, which we will demonstrate in a separate vignette in late 2025:</p>
<ol type="1">
<li>Obtain a subsample of the larger study. This could be FOVs chosen at random from all the tissues in the study, or just the first couple slides to come off the instrument.</li>
<li>Use this subsample to fix a basic pre-processing pipeline. Importantly, this pipeline has to be applicable to new samples: their data has to be projected onto the same dimension-reduced embedding, and their cell typing has to map to the cell typing of the subsample.</li>
<li>Apply the above pre-processing pipeline separately to each tissue/slide in the study.</li>
<li>Define an additional “hypothesis testing” pipeline, also to be applied separately to each tissue. This pipeline will compute whatever summary statistics you’re interested in, e.g.&nbsp;DE results, cell type abundance, LR interaction scores.</li>
<li>Perform meta-analysis of every tissues’ summary statistics together. E.g. contrast DE results or cell type abundances across responders and non-responders.</li>
<li>Custom subset analyses: you may also have to load in custom subsets of the data for analysis, e.g.&nbsp;just the macrophages from every tissue.</li>
</ol>
</section>
<section id="outline" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="outline"><span class="header-section-number">1.3</span> Outline</h2>
<p>Here we’ll walk through a full pipeline for getting your data ready to explore and analyze. We will:</p>
<ol type="1">
<li><a href="#loading">Load in flat files exported by AtoMx<sup>®</sup> SIP</a></li>
<li><a href="#xy">Create a custom spatial layout of our tissues for better visualization (optional)</a></li>
<li><a href="#qc">QC our data, removing cells and FOVs with bad data</a></li>
<li><a href="#dimred">Perform dimension reduction (features a new algorithm for obtaining better PC projections)</a></li>
<li><a href="#clust">Cluster our data and name the clusters we discover</a></li>
<li><a href="#hieratype">Perform immune cell typing using the HieraType algorithm</a></li>
</ol>
</section>
</section>
<section id="loading" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Loading flat files</h1>
<p>First we’ll load the packages we need. Importantly, we’ll load “Matrix”, which handles <em>sparse matrices</em>. When a matrix has mainly zeroes, which is generally the case for single cell data, storing it in sparse matrix format has huge memory savings. Given the size of CosMx data, it is vital to be memory-efficient.</p>
<p>Note: we’ll store our count matrices with cells in rows and genes in columns. Many protocols arrange their count matrix the other way around, but our approach here will be more performant for most operations.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># necessary libraries</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table) <span class="co"># for more memory efficient data frames</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix) <span class="co"># for sparse matrices like our counts matrix</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p><a href="../../posts/flat-file-exports/flat-files-compare.html" target="_blank">AtoMx flat file exports</a> come with each slide packaged in its own directory. Our example here has one slide, titled “BreastCancer”; its contents look like this:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/flatfiles.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
<p>So we have 5 files: an expression matrix, cell metadata, cell polygons, a map of fov positions, and the (very large) transcript locations file from this analysis. This analysis will only use the expression matrix and metadata. As such, you could reasonably export only those files from AtoMx. For examples on how to use the cell polygons in plots, see <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/spatial-plotting/" target="_blank">this post</a>.</p>
<p>Now we’ll load in the flat files and format as we want.</p>
<p>Trying to load an entire matrix into memory at once may cause out-of-memory issues for large files.<br>
In the code below, we load the counts matrices in chunks.<br>
* First, we’ll use <code>data.table::fread</code> function which can directly read gzipped files, and determine how many rows and columns we’re working with.<br>
* Then, we’ll read each chunk of data in, and convert it to a low-memory sparse matrix format before moving onto the next chunk.<br>
* At the end, we’ll combine our sparse matrices to create the full data matrix.</p>
<p>To begin, tell your script where to operate:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">## set up:</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># location of flat files:</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># (point to a directory containing one folder of flat files per slide)</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>myflatfiledir <span class="ot">&lt;-</span> <span class="st">"path_to_your_flat_files_folder/flatFiles"</span>    </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># where to write output:</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>outdir <span class="ot">&lt;-</span> <span class="st">"path_to_your_desired_output_location"</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># set up results directories for later</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/results"</span>))){</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/results"</span>))</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># directory to hold intermediate data files</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/processed_data"</span>))){</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/processed_data"</span>))</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(outdir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The below code performs the loading and merging operations, producing 4 data objects:</p>
<ul>
<li>A cell metadata matrix (cells x variables)</li>
<li>A counts matrix (cells x genes)</li>
<li>A negative control probe (“negprobe”) counts matrix (cells x negprobes)</li>
<li>A SystemControl (“falsecode”) counts matrix (cells x falsecodes)</li>
</ul>
<p>If given a flat files directory with multiple slides, it will seamlessly load and combine their data into a unified dataset.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">### source readFlatFiles:</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"https://raw.githubusercontent.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/Main/_code/vignette2/utils/readFlatFiles.R"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="do">### automatically get slide names:</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>slidenames <span class="ot">&lt;-</span> <span class="fu">dir</span>(myflatfiledir)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="do">### load the flat files and return as distinct data objects:</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>temp <span class="ot">&lt;-</span> <span class="fu">readFlatFiles</span>(myflatfiledir, <span class="at">slidenames =</span> slidenames, <span class="at">return_negprobes =</span> <span class="cn">FALSE</span>, <span class="at">return_falsecodes =</span> <span class="cn">FALSE</span>, <span class="at">offset_slides =</span> <span class="cn">TRUE</span>) </span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>counts   <span class="ot">&lt;-</span> temp<span class="sc">$</span>counts;    temp<span class="sc">$</span>counts <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> temp<span class="sc">$</span>metadata;  temp<span class="sc">$</span>metadata <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>xy       <span class="ot">&lt;-</span> temp<span class="sc">$</span>xy;        temp<span class="sc">$</span>xy <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(temp)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co"># checks</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">identical</span>(<span class="fu">rownames</span>(counts), metadata<span class="sc">$</span>cell_id))</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">max</span>(<span class="fu">abs</span>(Matrix<span class="sc">::</span><span class="fu">rowSums</span>(counts) <span class="sc">-</span> metadata<span class="sc">$</span>nCount_RNA)) <span class="sc">==</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Note that the above performs some well-advised modifications to the metadata:</p>
<ul>
<li>Removes the “fov” column, which has duplicate values over multiple slides (e.g.&nbsp;any two slides will each have a fov 1). Replace it with the “FOV” column, which appends a slide identifier to FOV IDs so that each FOV has a unique id.</li>
<li>Removes the “cell_ID” column, which also can have duplicate values across slides, and retain the duplicate-free column “cell_id”.</li>
<li>For convenience, extracts cells’ xy positions as a separate data object, and recasts them from pixel-scale to mm-scale.</li>
<li>Adds the mean negprobe counts to the metadata (the metadata has the <em>sum</em> of the negprobes already, but since negprobe counts vary across panels it’s worth recording this value now so we won’t have to remember our negprobe number later.)</li>
</ul>
<p>Finally, we save these data objects as .RDS files. These files are smaller on disk (they benefit from more compression than flat files do), and they’re faster to read back into R.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Please note that these files hold intermediate results for convenience during analysis; they are not official file formats supported by BSB.</strong></p>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save data in R-friendly form:</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(counts, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/counts_unfiltered.RDS"</span>))</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(negcounts, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/negcounts_unfiltered.RDS"</span>))</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(falsecounts, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/falsecounts_unfiltered.RDS"</span>))</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(metadata, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/metadata_unfiltered.RDS"</span>))</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(xy, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/xy_unfiltered.RDS"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now we have data in 3 locations:</p>
<ul>
<li>The RDS files we just created</li>
<li>The flat files we began with</li>
<li>The original data on AtoMx SIP</li>
</ul>
<p>This is unnecessary duplication, and since CosMx data is so large, storing extra copies can incur real cost. There are two feasible approaches:</p>
<ol type="1">
<li><p>Once you’ve finished your analysis, delete all the large .RDS files created during it. They can always be regenerated by re-running the analysis scripts. If you are computing on the cloud, re-running does incur some compute costs, but these are minor compared to the cost of indefinitely storing large files.</p></li>
<li><p>Once you’ve confirmed this data loading script has run successfully, you can safely delete the flat files - you won’t be using them again. If you do need to go back to them, they’ll be on AtoMx, ready to be downloaded again: AtoMx acts as a persistent, authoritative source of unmodified data.</p></li>
</ol>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Note on data size:</strong> A reasonably high-performance compute instance can generally handle a few slides of CosMx data using the scripts we use here. But once studies get to millions of cells, they can overwhelm the memory available to most analysts. At this point, you have two options:</p>
<ol type="1">
<li>Use data formats that store data on disk, not in memory, e.g.&nbsp;TileDB, SeuratDisk, hdf5…</li>
<li>Analyze your data one tissue at a time. Put all tissues through a standard analysis pipeline, then, for multi-tissue analyses, judiciously load in just the data you need for a given analysis. This is our recommendation; vignettes demonstrating this workflow are coming in 2026.</li>
</ol>
<p>Neither option is terribly convenient, but having millions or tens of millions of cells to analyze is worth some hassle.</p>
</div>
</div>
</section>
<section id="xy" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Customizing the tissues’ spatial arrangements</h1>
<p>To show multiple slides in the same spatial plot, or just to produce spatial plots without a lot of extraneous white space, we often have to shift around tissues’ and/or FOVs’ xy positions. (We if course will preserve the relative xy positions of any cells we plan to analyze together spatially.) Only a small number of operations are required to do so; below are some tools and code chunks for facilitating these operations, some borrowed from <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/visuals-reduce-whitespace/" target="_blank">this post</a>.</p>
<p>The basic operations of customized xy layouts are as follows:</p>
<p><strong>First</strong>, for multi-slide studies, adjust the slides’ xy positions do they don’t overlap. This function is called within our flat file reader, but you will often want to customize how slides are tiled.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># source the tiling function:</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"https://raw.githubusercontent.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/Main/_code/vignette2/utils/readFlatFiles.R"</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># define new xy positions in which multiple slides no longer overlap each other:</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>xy <span class="ot">&lt;-</span> <span class="fu">condenseTissues</span>(<span class="at">xy =</span> xy, </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">tissue =</span> metadata<span class="sc">$</span>Run_Tissue_name, </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>                      <span class="at">tissueorder =</span> <span class="cn">NULL</span>,  <span class="co"># optional, specify the order tissues are tiled in</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>                      <span class="at">buffer =</span> <span class="dv">1</span>, <span class="co"># space between tissues</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>                      <span class="at">widthheightratio =</span> <span class="dv">4</span><span class="sc">/</span><span class="dv">3</span>) <span class="co"># desired shape of final xy locations </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p><strong>Second</strong>, break your cells into spatially contiguous groups (which may be separate tissues from one slide, or merely discontinuous FOV blocks from the same tissue). This can be done manually, e.g.&nbsp;below where we select cells in a selected xy range:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select a group of cells based on cutoffs in xy:</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>xrange <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">3.4</span>, <span class="fl">5.2</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>yrange <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">6.1</span>, <span class="dv">8</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>thisgroup <span class="ot">&lt;-</span> ((xy[, <span class="dv">1</span>] <span class="sc">&gt;</span> xrange[<span class="dv">1</span>]) <span class="sc">&amp;</span> (xy[, <span class="dv">1</span>] <span class="sc">&lt;</span> xrange[<span class="dv">2</span>])) <span class="sc">&amp;</span> </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  ((xy[, <span class="dv">2</span>] <span class="sc">&gt;</span> yrange[<span class="dv">1</span>]) <span class="sc">&amp;</span> (xy[, <span class="dv">2</span>] <span class="sc">&lt;</span> yrange[<span class="dv">2</span>]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Or, we can automatically group contiguous FOVs using the dbscan algorithm:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cells within eps range of each other will be grouped together. </span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Set eps appropriately to link sufficiently close FOVs and separate sufficiently distant FOVs. </span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># An FOV's width is about 1 mm. </span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>cellgroups <span class="ot">&lt;-</span> dbscan<span class="sc">::</span><span class="fu">dbscan</span>(xy, <span class="at">minPts =</span> <span class="dv">1</span>, <span class="at">eps =</span> <span class="dv">1</span>)<span class="sc">$</span>cluster</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># confirm the grouping is as desired:</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(xy, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.1</span>, <span class="at">asp =</span> <span class="dv">1</span>, </span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">sample</span>(<span class="fu">colors</span>())[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(cellgroups))])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p><strong>Third</strong>, we rearrange groups. We can move groups manually:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># shift a group of cells to the right:</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>xy[thisgroup, <span class="dv">1</span>] <span class="ot">&lt;-</span> xy[thisgroup, <span class="dv">1</span>] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># or up:</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>xy[thisgroup, <span class="dv">2</span>] <span class="ot">&lt;-</span> xy[thisgroup, <span class="dv">2</span>] <span class="sc">+</span> <span class="dv">5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>…or we can move them automatically using the same <code>condenseTissues</code> function we used above:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>xy <span class="ot">&lt;-</span> <span class="fu">condenseTissues</span>(<span class="at">xy =</span> xy, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">tissue =</span> cellgroups, </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">tissueorder =</span> <span class="cn">NULL</span>,  <span class="co"># optional, specify the order tissues are tiled in</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">buffer =</span> <span class="dv">1</span>, <span class="co"># space between tissues</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">widthheightratio =</span> <span class="dv">4</span><span class="sc">/</span><span class="dv">3</span>) <span class="co"># desired shape of final xy locations </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="qc" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> QC</h1>
<p>QC of CosMx data is mostly straightforward. Technical effects can be complex, but they manifest in limited ways:</p>
<ul>
<li>Many individual cells can have low signal</li>
<li>Segmentation errors can create “cells” with bad data</li>
<li>Sporadic FOVs can have low signal</li>
<li>Sporadic FOVs can have distorted / outlier expression profiles</li>
</ul>
<p>Our workflow will create then add to <code>flag</code>, a vector marking cells flagged for one reason or another. When we’re done, we’ll delete everything we’ve flagged. (The full original data will remain on AtoMx, so this step is not as drastic as it seems. For smaller studies it could still be reasonable to hold on to the pre-QC .RDS files.)</p>
<section id="per-cell-qc" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="per-cell-qc"><span class="header-section-number">4.1</span> Per-cell QC:</h2>
<p>First, we’ll apply a threshold on cells’ total counts. Our defaults are fairly permissive; it would be reasonable double them in datasets where you can do so without throwing out too many cells.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># require sufficient counts per cell </span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log2</span>(metadata<span class="sc">$</span>nCount_RNA), <span class="at">breaks =</span> <span class="dv">100</span>, <span class="at">xlab =</span> <span class="st">"Log2 counts per cell"</span>, <span class="at">ylab =</span> <span class="st">"N cells"</span>, <span class="at">main =</span> <span class="st">""</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="do">## cut off the low tail: </span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># recommendations:</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># - 1000plex: 20 (permissive) or 50 (conservative)</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co"># - 6000plex: 50 (permissive) or 100 (conservative)</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># - WTX: 100 (permissive) or 200 (conservative)</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co"># - all panels: no more than 5% of cells (permissive) or 10% of cells (conservative)</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>count_threshold <span class="ot">&lt;-</span> <span class="fu">min</span>(<span class="dv">200</span>, <span class="fu">quantile</span>(metadata<span class="sc">$</span>nCount_RNA, <span class="fl">0.05</span>)) <span class="co"># apply threshold, but don't flag &gt;5% of cells</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="fu">log2</span>(count_threshold), <span class="at">col =</span> <span class="st">"red"</span>)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>flag <span class="ot">&lt;-</span> metadata<span class="sc">$</span>nCount_RNA <span class="sc">&lt;</span> count_threshold</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="at">legend =</span> <span class="fu">paste0</span>(<span class="fu">round</span>(<span class="fu">mean</span>(flag) <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>), <span class="st">"% rejected"</span>))</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(flag)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/QC_counts_histogram.png" class="img-fluid figure-img" style="width:115.0%"></p>
</figure>
</div>
</div>
</div>
<p>Next we’ll throw out cells with excessively large areas. In most studies, these “cells” are the result of segmentation errors. This step typically removes very few cells.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># what's the distribution of areas?</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(metadata<span class="sc">$</span>Area, <span class="at">breaks =</span> <span class="dv">100</span>, <span class="at">xlab =</span> <span class="st">"Cell Area"</span>, <span class="at">main =</span> <span class="st">""</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># based on the above, set a threshold well out into the tail:</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>area_threshold <span class="ot">&lt;-</span> <span class="dv">30000</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> area_threshold, <span class="at">col =</span> <span class="st">"red"</span>)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">paste0</span>(<span class="fu">round</span>(<span class="fu">mean</span>(metadata<span class="sc">$</span>Area <span class="sc">&gt;</span> area_threshold) <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">2</span>), <span class="st">"% rejected"</span>))</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co"># flag cells based on area:</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>flag <span class="ot">&lt;-</span> flag <span class="sc">|</span> (metadata<span class="sc">$</span>Area <span class="sc">&gt;</span> area_threshold)</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(flag)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/QC_area_histogram.png" class="img-fluid figure-img" style="width:115.0%"></p>
</figure>
</div>
</div>
</div>
<p>Note: some tissues do have very large cells, e.g.&nbsp;purkinje cells in the brain or cardiomyocytes in the heart.</p>
<p>To check for segmentation errors, consider our R package <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/segmentation-error-evaluation/" target="_blank">FastReseg</a>, which uses transcript positions to flag cells with likely segmentation errors, then remove or reassign transcripts assigned to the wrong cells. We also advise examining the segmentation results in AtoMx’s tissue viewer; studies will occasionally need to be re-segmented under a different set of parameters.</p>
</section>
<section id="fov-qc" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="fov-qc"><span class="header-section-number">4.2</span> FOV QC</h2>
<p>We check FOVs for two indicators of artifacts: diminished total counts, and biased gene expression profiles. Procedures for both these checks, along with a detailed discussion, can be found in a method described <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/fov-qc/" target="_blank">here</a>.</p>
<p>To summarize briefly: for each “bit” in our barcodes (e.g.&nbsp;red in reporter cycle 8), we look for FOVs where genes using the bit are underexpressed. We also look for FOVs where total expression is suppressed.</p>
<p>Below we’ll run our FOV QC code and examine its results:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="do">## preparing resources:</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co"># source the FOV QC tool:</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"https://raw.githubusercontent.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/Main/_code/FOV%20QC/FOV%20QC%20utils.R"</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># load necessary information for the QC tool: the gene to barcode map:</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>allbarcodes <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">url</span>(<span class="st">"https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/raw/Main/_code/FOV%20QC/barcodes_by_panel.RDS"</span>))</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(allbarcodes)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># get the barcodes for the panel we want:</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>barcodemap <span class="ot">&lt;-</span> allbarcodes<span class="sc">$</span>Hs_WTX    <span class="co"># &lt;------- choose the right map for your panel</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(barcodemap)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="do">## run the method:</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>fovqcresult <span class="ot">&lt;-</span> <span class="fu">runFOVQC</span>(<span class="at">counts =</span> counts, <span class="at">xy =</span> xy, <span class="at">fov =</span> metadata<span class="sc">$</span>FOV, <span class="at">barcodemap =</span> barcodemap) </span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(fovqcresult, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/fovqcresult.RDS"</span>))</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a><span class="co"># map of flagged FOVs:</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a><span class="fu">mapFlaggedFOVs</span>(fovqcresult)</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a><span class="co"># list FOVs flagged for any reason, for loss of signal, for bias:</span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>fovqcresult<span class="sc">$</span>flaggedfovs</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>fovqcresult<span class="sc">$</span>flaggedfovs_fortotalcounts</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>fovqcresult<span class="sc">$</span>flaggedfovs_forbias</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a><span class="co"># flag cells in flagged FOVs:</span></span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>flag[<span class="fu">is.element</span>(metadata<span class="sc">$</span>FOV, fovqcresult<span class="sc">$</span>flaggedfovs)] <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the result:</span></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a><span class="fu">FOVSignalLossSpatialPlot</span>(fovqcresult) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/QC_FOVQC.png" class="img-fluid figure-img" width="264"></p>
</figure>
</div>
</div>
</div>
<p>We don’t have any flagged FOVs in this study. For examples of studies with more flagged FOVs, see <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/fov-qc/" target="_blank">the FOV QC post</a> and <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/vignette-basic-analysis/assets/2.-QC-and-normalization.html" target="_blank">this older vignette</a>.</p>
<p>Finally, we’ll remove flagged cells and save the filtered data for downstream analysis:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save data with flagged cells removed</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> counts[<span class="sc">!</span>flag, ]</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> metadata[<span class="sc">!</span>flag, ]</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>xy <span class="ot">&lt;-</span> xy[<span class="sc">!</span>flag, ]</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="co"># save filtered data objects, to be used in the remainder of the analysis</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(counts, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/counts.RDS"</span>))</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(metadata, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/metadata.RDS"</span>))</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(xy, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/xy.RDS"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="descriptive-qc-plots" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="descriptive-qc-plots"><span class="header-section-number">4.3</span> Descriptive QC plots</h2>
<p>In addition to the above per-cell flags, it can be worth plotting some QC metrics atop the space of your tissues. The purpose of these plots is to become aware of possible technical effects that might explain any artifacts you discover later in analysis. In the vast majority of studies, these plots will not require any action from you.</p>
<p>First, plot total counts over space:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="do">## plot total counts over space:</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(xy, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.1</span>, <span class="at">asp =</span> <span class="dv">1</span>, </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">viridis_pal</span>(<span class="at">option =</span> <span class="st">"B"</span>)(<span class="dv">101</span>)[</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>       <span class="fu">round</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="dv">100</span> <span class="sc">*</span> (<span class="fu">log2</span>(<span class="fu">pmin</span>(<span class="fu">pmax</span>(metadata<span class="sc">$</span>nCount_RNA, <span class="dv">10</span>), <span class="dv">5000</span>)) <span class="sc">-</span> <span class="fu">log2</span>(<span class="dv">10</span>)) <span class="sc">/</span> <span class="fu">log2</span>(<span class="dv">5000</span>))</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>     ])</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>legendvals <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">1000</span>, <span class="dv">5000</span>)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"bottomright"</span>, <span class="at">pch =</span> <span class="dv">16</span>, </span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="fu">viridis_pal</span>(<span class="at">option =</span> <span class="st">"B"</span>)(<span class="dv">101</span>)[<span class="fu">round</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="dv">100</span> <span class="sc">*</span> (<span class="fu">log2</span>(legendvals) <span class="sc">-</span> <span class="fu">log2</span>(<span class="dv">10</span>)) <span class="sc">/</span> <span class="fu">log2</span>(<span class="dv">5000</span>))]),</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"Total counts:"</span>, legendvals))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/QC_counts_over_xy.png" class="img-fluid figure-img" style="width:115.0%"></p>
</figure>
</div>
</div>
</div>
<p>We see ample variability in total counts over space, but it all appears biological, not technical, with the exception of some FOV edge effects in which cells partially captured at some FOV edges have suppressed counts.</p>
<p>We also like to plot background over space. Because negative control counts are so sparse at the level of single cells, we plot <em>spatially smoothed</em> background, i.e.&nbsp;we color each cell by the mean background over its 50 nearest neighbors. To isolate background tendency from overall signal strength, we also normalize our background (negprobe) counts to each cells’ total RNA counts.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install insitucor:</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="cn">FALSE</span>) {</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"https://github.com/Nanostring-Biostats/InSituCor"</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="do">## plot smoothed background over space:</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># get spatial neighbors:</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>neighbors <span class="ot">&lt;-</span> InSituCor<span class="sc">:::</span><span class="fu">nearestNeighborGraph</span>(xy[, <span class="dv">1</span>], xy[, <span class="dv">2</span>], <span class="at">N =</span> <span class="dv">50</span>, <span class="at">subset =</span> metadata<span class="sc">$</span>slide_ID)  </span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate spatially-smoothed and totalcount-normalized background:</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>smoothedgbrate <span class="ot">&lt;-</span> InSituCor<span class="sc">:::</span><span class="fu">neighbor_mean</span>(<span class="at">x =</span> metadata<span class="sc">$</span>nCount_negprobes <span class="sc">/</span> metadata<span class="sc">$</span>nCount_RNA, <span class="at">neighbors =</span> neighbors)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co"># plot:</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(xy[, <span class="dv">2</span><span class="sc">:</span><span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.1</span>, <span class="at">asp =</span> <span class="dv">1</span>, </span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">viridis_pal</span>(<span class="at">option =</span> <span class="st">"B"</span>)(<span class="dv">101</span>)[</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>       <span class="fu">round</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="dv">100</span> <span class="sc">*</span> <span class="fu">pmin</span>(smoothedgbrate <span class="sc">/</span> <span class="fu">quantile</span>(smoothedgbrate, <span class="fl">0.9995</span>), <span class="dv">1</span>))</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>     ])</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>legendvals <span class="ot">&lt;-</span> <span class="fu">signif</span>(<span class="fu">median</span>(smoothedgbrate) <span class="sc">*</span> <span class="fu">c</span>(<span class="fl">0.5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">10</span>), <span class="dv">2</span>)</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"bottomright"</span>, <span class="at">pch =</span> <span class="dv">16</span>, </span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="fu">viridis_pal</span>(<span class="at">option =</span> <span class="st">"B"</span>)(<span class="dv">101</span>)[<span class="fu">signif</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="dv">100</span> <span class="sc">*</span> <span class="fu">pmin</span>(legendvals <span class="sc">/</span> <span class="fu">quantile</span>(smoothedgbrate, <span class="fl">0.9995</span>), <span class="dv">1</span>), <span class="dv">2</span>)]),</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"Spatially smoothed totalcounts-normalized background:"</span>, <span class="fu">paste0</span>(<span class="fu">c</span>(<span class="fl">0.5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">10</span>), <span class="st">"x median"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/spatial_bgoversignal.png" class="img-fluid figure-img" width="464"></p>
</figure>
</div>
</div>
</div>
<p>FOV edge effects are also notable here, presumably downstream from suppressed signal in those locations. Of greater note is a band of higher-background FOVs towards the lower left of the tissue. No immediate action is needed, but by becoming aware of this small technical effect we’ll be armed to interpret any strange results driven by it.</p>
<p>Finally, we just flagged cells in space:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot flagged cells in space:</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(xy[, <span class="dv">2</span><span class="sc">:</span><span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.1</span>, <span class="at">asp =</span> <span class="dv">1</span>, </span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"grey80"</span>, <span class="st">"red"</span>)[<span class="dv">1</span> <span class="sc">+</span> flag])</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>legendvals <span class="ot">&lt;-</span> <span class="fu">signif</span>(<span class="fu">median</span>(smoothedgbrate) <span class="sc">*</span> <span class="fu">c</span>(<span class="fl">0.5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">10</span>), <span class="dv">2</span>)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"bottomright"</span>, <span class="at">pch =</span> <span class="dv">16</span>, </span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="fu">viridis_pal</span>(<span class="at">option =</span> <span class="st">"B"</span>)(<span class="dv">101</span>)[<span class="fu">signif</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="dv">100</span> <span class="sc">*</span> <span class="fu">pmin</span>(legendvals <span class="sc">/</span> <span class="fu">quantile</span>(smoothedgbrate, <span class="fl">0.9995</span>), <span class="dv">1</span>), <span class="dv">2</span>)]),</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"Spatially smoothed totalcounts-normalized background:"</span>, <span class="fu">paste0</span>(<span class="fu">c</span>(<span class="fl">0.5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">10</span>), <span class="st">"x median"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/QC_flaggedcells.png" class="img-fluid figure-img" style="width:115.0%"></p>
</figure>
</div>
</div>
</div>
<p>We can see flagged cells clustered in ducts/vessels/gaps in the tissue. Again, no action is needed, but we should keep this in mind while we interpret results.</p>
</section>
<section id="normalization" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="normalization"><span class="header-section-number">4.4</span> Normalization</h2>
<p>We typically normalize each cell’s expression profile to its total counts. The below code does this efficiently. In fact, given the size of the matrix and the speed of this calculation, there is no need to save a normalized data object; it’s usually cheaper to re-compute it that to store it.</p>
<p>Note that no steps in this workflow actually use the normalized data, although the dimension reduction step performs its own normalization under the hood. Normalized data will be useful in later analyses, e.g.&nbsp;for plotting expression over space or running differential expression.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># normalize counts matrix using efficent sparse matrix calls:</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>scale_row <span class="ot">&lt;-</span> <span class="fu">mean</span>(metadata<span class="sc">$</span>nCount_RNA)<span class="sc">/</span>metadata<span class="sc">$</span>nCount_RNA</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>norm <span class="ot">&lt;-</span> counts</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>norm<span class="sc">@</span>x <span class="ot">&lt;-</span> norm<span class="sc">@</span>x <span class="sc">*</span> scale_row[norm<span class="sc">@</span>i <span class="sc">+</span> <span class="dv">1</span>L]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="dimred" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Dimension reduction</h1>
<p>Here we demonstrate an important improvement in dimension reduction for large studies. The single cell analysis world has long used Pearson residuals to normalize &amp; transform their data prior to running PCA; these residuals scale count data to better balance the information available in each cell / gene. However, Pearson residuals are dense, meaning we cannot encode them with a sparse matrix, making them computationally intractable for larger datasets. To benefit from Pearson residuals without taxing our compute instances, we have developed a method of calculating the principal components of a Pearson residual matrix without ever having to store the entire matrix. We demonstrate this function below. (Scratch Space post on the subject coming soon.)</p>
<p>We’ll start by identifying highly variable genes. In this section we’ll take advantage of some of the functions in the Seurat ecosystem.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a seurat object:</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>seu <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(counts),</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">meta.data =</span> metadata)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># find HVG's:</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="co"># recommend nfeatures = 3000 for WTX, 2000 for 6k panel, and using all features for 1k panels</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>seu <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindVariableFeatures</span>(seu, <span class="at">nfeatures =</span> <span class="dv">3000</span>) </span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>hvgs <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(seu<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.data<span class="sc">$</span>var.features, <span class="cn">NA</span>)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(hvgs, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/hvgs.RDS"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Next we run our Pearson residual-based PCA algorithm. You can install it with this command:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install the sparse pearson PCA package:</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"Nanostring-Biostats/CosMx-Analysis-Scratch-Space"</span>,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">subdir =</span> <span class="st">"_code/scPearsonPCA"</span>, <span class="at">ref =</span> <span class="st">"Main"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The algorithm runs in two versions: standard, and batch-aware. The batch-aware version allows for gene frequencies that vary across some categorical batch variable like tissue or slide.</p>
<p>Note the code saving the output. This lets us re-run the script without waiting for long algorithms to reprocess. However, this particular routine runs fast enough, and generates a large enough data object, that it would also be reasonable to recalculate rather than store its output.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">Standard mode</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">Batch-aware mode</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="do">## get PCs based on Pearson residuals:</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co"># version without batch adjustment:</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="cn">TRUE</span>) {</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  genefreq <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">gene_frequency</span>(Matrix<span class="sc">::</span><span class="fu">t</span>(counts)) <span class="do">## gene frequency (across all cells)</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  pcaobj <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">sparse_quasipoisson_pca_seurat</span>(</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(counts[, hvgs]),</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">totalcounts =</span> metadata<span class="sc">$</span>nCount_RNA,</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">grate =</span> genefreq[hvgs],</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">scale.max =</span> <span class="dv">10</span>, <span class="do">## PC's reflect clipping pearson residuals &gt; 10 SDs above the mean pearson residual</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.scale =</span> <span class="cn">TRUE</span>, <span class="do">## PC's reflect as if pearson residuals for each gene were scaled to have standard deviation=1</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.center =</span> <span class="cn">TRUE</span> <span class="do">## PC's reflect as if pearson residuals for each gene were centered to have mean=0</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(pcaobj, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/pcaobj.RDS"</span>))</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>  pcaobj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/pcaobj.RDS"</span>))</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># In this example code we assume the metadata has a column "patient" that we wish to treat as a possible batch effect</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="cn">TRUE</span>) {</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  genefreq_batch <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">gene_frequency</span>(</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(counts),</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">obs =</span> metadata[,.(cell_id, patient)],</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">cellid_colname =</span> <span class="st">"cell_id"</span>,</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">batch_variable =</span> <span class="st">"patient"</span>)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>  pcaobj <span class="ot">&lt;-</span> </span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sparse_quasipoisson_pca_seurat_batch</span>(</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>      Matrix<span class="sc">::</span><span class="fu">t</span>(counts[, hvgs]),</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>      <span class="at">totalcounts =</span> metadata<span class="sc">$</span>nCount_RNA,</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">grate =</span> genefreq_batch[hvgs,],</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">obs =</span> metadata[,.(cell_id, patient)],</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">batch_variable =</span> <span class="st">"patient"</span>,</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">cellid_colname =</span> <span class="st">"cell_id"</span>,</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale.max =</span> <span class="dv">10</span>,</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">do.scale =</span> <span class="cn">TRUE</span>,</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">do.center =</span> <span class="cn">TRUE</span></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(pcaobj, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/pcaobj.RDS"</span>))</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a>  pcaobj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/pcaobj.RDS"</span>))</span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</div>
</div>
</div>
<p>Next we’ll obtain a UMAP projection from those PCs:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="do">## run UMAP:</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="cn">TRUE</span>) {</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  umapobj <span class="ot">&lt;-</span> scPearsonPCA<span class="sc">::</span><span class="fu">make_umap</span>(pcaobj)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># save the full object: the umap connectivity graph will be useful later on:</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(umapobj, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/umapobj.RDS"</span>))</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>  umapobj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/umapobj.RDS"</span>))</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="do">## plot the UMAP projection, coloring by total counts:</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(umapobj<span class="sc">$</span>ump<span class="sc">@</span>cell.embeddings, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">asp =</span> <span class="dv">1</span>, <span class="at">cex =</span> <span class="fl">0.1</span>,</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> scales<span class="sc">::</span><span class="fu">alpha</span>(<span class="fu">viridis_pal</span>(<span class="at">option =</span> <span class="st">"B"</span>)(<span class="dv">101</span>)[</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>       <span class="fu">round</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="dv">100</span> <span class="sc">*</span> (<span class="fu">log2</span>(<span class="fu">pmin</span>(<span class="fu">pmax</span>(metadata<span class="sc">$</span>nCount_RNA, <span class="dv">10</span>), <span class="dv">5000</span>)) <span class="sc">-</span> <span class="fu">log2</span>(<span class="dv">10</span>)) <span class="sc">/</span> <span class="fu">log2</span>(<span class="dv">5000</span>))</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>     ], <span class="fl">0.05</span>))</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>legendvals <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">1000</span>, <span class="dv">5000</span>)</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="at">pch =</span> <span class="dv">16</span>, </span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="fu">viridis_pal</span>(<span class="at">option =</span> <span class="st">"B"</span>)(<span class="dv">101</span>)[<span class="fu">round</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="dv">100</span> <span class="sc">*</span> (<span class="fu">log2</span>(legendvals) <span class="sc">-</span> <span class="fu">log2</span>(<span class="dv">10</span>)) <span class="sc">/</span> <span class="fu">log2</span>(<span class="dv">5000</span>))]),</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"Total counts:"</span>, legendvals))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/umap_totalcounts.png" class="img-fluid figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="clust" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Cell typing</h1>
<p>Here we demo the most consistently-successful approach to cell typing we have found. First, we perform unsupervised clustering using a set of a few thousand highly variable genes (hvg’s). This approach is robust, and requires minimal configuration and re-running. Naming clusters used to be time-consuming, but we have found that the right prompts to a large language model (LLM), <strong>combined with rigorous oversight</strong>, minimize the burden of cluster naming. We will demonstrate two approaches to unsupervised clustering:</p>
<ul>
<li>Louvain clustering, using PC’s derived from Pearson residuals of hvg’s.</li>
<li>InSituType’s unsupervised mode, run on hvg’s.</li>
</ul>
<p>Of the above two methods, Louvain tends to be faster, while InSituType is easier to refine, and, more importantly, straightforward to project onto future datasets.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true">Louvain clustering</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false">Clustering with InSituType</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<p><strong>Note:</strong> Here we run louvain clustering with a high resolution, intentionally over-clustering with the plan to condense afterwards. It’s often better to run with the default resolution (0.8) at first, then subcluster selected cell types afterwards. To sub-cluster a cell type, first filter out genes for which segmentation errors are likely to bias its profiles. Use <code>smiDE::overlap_ratio_metric</code> to do so. (smiDE is available on the scratch space.)</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use Seurat to run Louvain clustering on the graph from our umap object</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="cn">TRUE</span>) {</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  seu[[<span class="st">"pearsongraph"</span>]] <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">as.Graph</span>(umapobj<span class="sc">$</span>grph) <span class="do">## nearest neighbors / adjacency matrix used for unsupervised clustering</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  seu <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindClusters</span>(seu, <span class="at">graph =</span> <span class="st">"pearsongraph"</span>, <span class="at">resolution =</span> <span class="fl">1.2</span>)  <span class="co"># choosing a high resolution with the plan to condense afterwards</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  clust <span class="ot">&lt;-</span> seu<span class="sc">@</span>meta.data<span class="sc">$</span>seurat_clusters</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(clust, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/clust.RDS"</span>))</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  clust <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/clust.RDS"</span>))</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="co"># save to metadata</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>metadata<span class="sc">$</span>clust <span class="ot">&lt;-</span> clust</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>cellcols <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">c</span>(</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#E41A1C"</span>, <span class="st">"#377EB8"</span>, <span class="st">"#4DAF4A"</span>, <span class="st">"#984EA3"</span>, <span class="st">"#FF7F00"</span>,</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#A65628"</span>, <span class="st">"#F781BF"</span>, <span class="st">"#999999"</span>, <span class="st">"#66C2A5"</span>, <span class="st">"#FC8D62"</span>,</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#8DA0CB"</span>, <span class="st">"#E78AC3"</span>, <span class="st">"#A6D854"</span>, <span class="st">"#FFD92F"</span>, <span class="st">"#E5C494"</span>,</span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#B3B3B3"</span>, <span class="st">"#1B9E77"</span>, <span class="st">"#D95F02"</span>, <span class="st">"#7570B3"</span>, <span class="st">"#E7298A"</span>,</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#66A61E"</span>, <span class="st">"#E6AB02"</span>, <span class="st">"#A6761D"</span>, <span class="st">"#666666"</span>, <span class="st">"#A6CEE3"</span>,</span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#1F78B4"</span>, <span class="st">"#B2DF8A"</span>, <span class="st">"#33A02C"</span>, <span class="st">"#FB9A99"</span>, <span class="st">"#E31A1C"</span>,</span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#FDBF6F"</span>, <span class="st">"#FF7F00"</span>, <span class="st">"#CAB2D6"</span>, <span class="st">"#6A3D9A"</span>, <span class="st">"#FFFF99"</span>,</span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#B15928"</span>, <span class="st">"#8DD3C7"</span>, <span class="st">"#FFFFB3"</span>, <span class="st">"#BEBADA"</span>, <span class="st">"#FB8072"</span></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>), <span class="dv">3</span>)[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(clust))]</span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Quick xy and umap plots of clusters to confirm resolution looks right:</span></span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(umapobj<span class="sc">$</span>ump<span class="sc">@</span>cell.embeddings, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">asp =</span> <span class="dv">1</span>, <span class="at">cex =</span> <span class="fl">0.1</span>, </span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> scales<span class="sc">::</span><span class="fu">alpha</span>(cellcols[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(clust))], <span class="fl">0.05</span>))</span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(xy, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">asp =</span> <span class="dv">1</span>, <span class="at">cex =</span> <span class="fl">0.1</span>, </span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> cellcols[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(clust))])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/umap_louvain.png" class="img-fluid figure-img" width="312"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/cluster_xy_1stlook.png" class="img-fluid figure-img" width="464"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<p>Here we refer you to the <a href="https://github.com/Nanostring-Biostats/InSituType/blob/main/vignettes/NSCLC-RNA_InsituType-vignette.Rmd" target="_blank">InSituType vignette</a>. Running in unsupervised mode is simplest, though supervised and semi-supervised modes can work well provided with good reference profiles.</p>
<p>A call to InSituType that would work with our setup here is:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="cn">TRUE</span>) {</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  unsup <span class="ot">&lt;-</span> <span class="fu">insitutype</span>(</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> counts[, hvgs],  <span class="co"># only use HVG's. InSituType's computation time scales with the number of genes, so analyzing the full WTX panel is very slow. </span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">neg =</span> metadata<span class="sc">$</span>negmean,</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay_type =</span> <span class="st">"RNA"</span>, </span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">reference_profiles =</span> <span class="cn">NULL</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">reference_sds =</span> <span class="cn">NULL</span>,</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">cohort =</span> <span class="cn">NULL</span>,   <span class="co"># see the InSituType docs for how to use this argument to incorporate spatial information into clustering</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_clusts =</span> <span class="dv">20</span>,</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_starts =</span> <span class="dv">5</span>,</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">max_iters =</span> <span class="dv">25</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>  ) </span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(unsup, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/unsup.RDS"</span>))</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>  unsup <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/unsup.RDS"</span>))</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a><span class="co"># save results</span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>metadata<span class="sc">$</span>clust <span class="ot">&lt;-</span> clust <span class="ot">&lt;-</span> unsup<span class="sc">$</span>clust</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</div>
</div>
</div>
<section id="naming-clusters" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="naming-clusters"><span class="header-section-number">6.1</span> Naming clusters</h2>
<p>Determining which cell type a cluster corresponds to is a complex exercise that should take into account:</p>
<ul>
<li>The cluster’s marker genes</li>
<li>Its overall abundance</li>
<li>Its spatial locations</li>
</ul>
<p>This can be onerous, but we can get a big jump start on this work by using large language models (LLM’s). Below, we’ll create a custom prompt asking a LLM to name your cluster. We have found LLM’s to do a very good (<strong>though not perfect</strong>) job of naming cell types given this information. After this step we’ll show how to QC your cell type assignments.</p>
<p>As a first step, we’ll derive marker genes for our clusters:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="cn">TRUE</span>) {</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  markers <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">clusterwise_foldchange_metrics</span>(</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    Matrix<span class="sc">::</span><span class="fu">t</span>(counts),</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">metadata =</span> metadata,</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_column =</span> <span class="st">"clust"</span>,</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">cellid_column =</span> <span class="st">"cell_id"</span>)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(markers, <span class="at">file =</span> <span class="st">"processed_data/markers.RDS"</span>)</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  markers <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"processed_data/markers.RDS"</span>)</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="co"># get a more succinct list of markers: only the top N per cell type:</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a><span class="co"># prioritize by fold change, but penalize low-expressers:</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>markers<span class="sc">$</span>prioritystat <span class="ot">&lt;-</span> (markers<span class="sc">$</span>cluster_expr <span class="sc">+</span> <span class="fl">0.025</span>) <span class="sc">/</span> (markers<span class="sc">$</span>clusterprime_expr <span class="sc">+</span> <span class="fl">0.025</span>)</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>markersshort <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>nperclust <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (name <span class="cf">in</span> <span class="fu">unique</span>(markers<span class="sc">$</span>cluster)) {</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>  inds <span class="ot">&lt;-</span> markers<span class="sc">$</span>cluster <span class="sc">==</span> name</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(inds) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>    markersshort <span class="ot">&lt;-</span> <span class="fu">c</span>(markersshort,</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>                      markers<span class="sc">$</span>gene[inds][<span class="fu">order</span>(markers<span class="sc">$</span>prioritystat[inds], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="fu">min</span>(nperclust, <span class="fu">sum</span>(inds))]])</span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>In addition to data-derived marker genes, we’ll also look at a pre-defined list of well-known markers of broad cell types and lineages.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load general-purpose lineage and marker genes:</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"https://raw.githubusercontent.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/Main/_code/vignette2/lineage_and_marker_genes.R"</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># (loads the list "lineagegenes")</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co"># save both selected markers and lineage genes:</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>allusefulmarkers <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(counts), <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">unlist</span>(lineagegenes), markersshort)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now that we have marker genes identified, we can proceed to crafting an LLM prompt:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="do">## describe your tissue here. </span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>mytissuetype <span class="ot">&lt;-</span> <span class="st">"HER2+ breast cancer"</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="co"># the preamble:</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>prompt_preamble <span class="ot">&lt;-</span> <span class="fu">paste0</span>(</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="st">"Please propose cell type names for the clusters I</span><span class="sc">\'</span><span class="st">ve found in a CosMx study. This is data from a "</span>, mytissuetype, <span class="st">". "</span>,</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="st">"Below are two tables. The first is each cell type</span><span class="sc">\'</span><span class="st">s abundance in the dataset. "</span>,</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a><span class="st">"The second is a table of mean expression levels of various marker genes and lineage-defining genes in each cluster. "</span>,</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="st">"It</span><span class="sc">\'</span><span class="st">s possible that some clusters could be closely-related cell types, or distinct states of the same cell type. Call that out when it</span><span class="sc">\'</span><span class="st">s apparent. "</span>,</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a><span class="st">"For each cluster, give me your best guess at its identity, and give me your justification. Let me know when you</span><span class="sc">\'</span><span class="st">re uncertain. "</span>,</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a><span class="st">"Then, give me R code defining a named vector called </span><span class="sc">\'</span><span class="st">cluster_labels</span><span class="sc">\'</span><span class="st"> in which my cluster IDs are the names and the values are your proposed cell types. "</span>,</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a><span class="st">"Don</span><span class="sc">\'</span><span class="st">t hallucinate, and check your work three times. "</span>)</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a><span class="co"># info on cluster frequencies:</span></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>prompt_frequencies <span class="ot">&lt;-</span> <span class="fu">paste0</span>(</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a><span class="st">"</span></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a><span class="st">The cluster frquencies are: "</span>, </span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="fu">paste0</span>(<span class="st">"cluster "</span>, <span class="fu">names</span>(<span class="fu">table</span>(clust)), <span class="st">":"</span>, <span class="fu">table</span>(clust)), <span class="at">collapse =</span> <span class="st">", "</span>))</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a><span class="do">## give the LLM each cluster's mean expression profile over relevant genes:</span></span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a><span class="co"># obtain a matrix of the clusters' marker gene profiles:</span></span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a>allusefulmarkers <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(counts), <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">unlist</span>(lineagegenes), markersshort)))</span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>meanexpression <span class="ot">&lt;-</span> markers[, <span class="fu">dcast</span>(.SD, gene <span class="sc">~</span> cluster, <span class="at">value.var =</span> <span class="st">"cluster_expr"</span>)]</span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a>meanexpression <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(meanexpression[, <span class="sc">-</span><span class="dv">1</span>])          <span class="co"># drop gene col</span></span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(meanexpression) <span class="ot">&lt;-</span> markers[, <span class="fu">unique</span>(gene)]</span>
<span id="cb27-27"><a href="#cb27-27" aria-hidden="true" tabindex="-1"></a>meanexpression <span class="ot">&lt;-</span> meanexpression[allusefulmarkers, ]</span>
<span id="cb27-28"><a href="#cb27-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-29"><a href="#cb27-29" aria-hidden="true" tabindex="-1"></a><span class="co"># reduce to just lineage genes with decent expression (but keep de novo markers):</span></span>
<span id="cb27-30"><a href="#cb27-30" aria-hidden="true" tabindex="-1"></a>isdenovo <span class="ot">&lt;-</span> <span class="fu">is.element</span>(<span class="fu">rownames</span>(meanexpression), markersshort)</span>
<span id="cb27-31"><a href="#cb27-31" aria-hidden="true" tabindex="-1"></a>meanexpression <span class="ot">&lt;-</span> meanexpression[isdenovo <span class="sc">|</span> <span class="fu">apply</span>(meanexpression, <span class="dv">1</span>, max) <span class="sc">&gt;</span> <span class="fl">0.2</span>, ]</span>
<span id="cb27-32"><a href="#cb27-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-33"><a href="#cb27-33" aria-hidden="true" tabindex="-1"></a>meanexpressionastext <span class="ot">&lt;-</span> <span class="fu">write.table</span>(<span class="fu">round</span>(meanexpression[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>], <span class="dv">1</span>), <span class="at">sep =</span> <span class="st">","</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb27-34"><a href="#cb27-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-35"><a href="#cb27-35" aria-hidden="true" tabindex="-1"></a>prompt_profiles <span class="ot">&lt;-</span> <span class="fu">paste0</span>(</span>
<span id="cb27-36"><a href="#cb27-36" aria-hidden="true" tabindex="-1"></a>  <span class="st">"And here is a table of each cluster's mean expression of selected data- and biology-derived marker genes: "</span>,</span>
<span id="cb27-37"><a href="#cb27-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste</span>(<span class="fu">capture.output</span>(</span>
<span id="cb27-38"><a href="#cb27-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">write.table</span>(<span class="fu">round</span>(meanexpression, <span class="dv">1</span>), <span class="at">sep =</span> <span class="st">","</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb27-39"><a href="#cb27-39" aria-hidden="true" tabindex="-1"></a>  ), <span class="at">collapse =</span> <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb27-40"><a href="#cb27-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-41"><a href="#cb27-41" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">paste0</span>(prompt_preamble, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>, prompt_frequencies, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>, prompt_profiles))</span>
<span id="cb27-42"><a href="#cb27-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-43"><a href="#cb27-43" aria-hidden="true" tabindex="-1"></a><span class="co"># save old results for rewriting later on:</span></span>
<span id="cb27-44"><a href="#cb27-44" aria-hidden="true" tabindex="-1"></a>unnamedclust <span class="ot">&lt;-</span> clust</span>
<span id="cb27-45"><a href="#cb27-45" aria-hidden="true" tabindex="-1"></a>unnamedmeanexpression <span class="ot">&lt;-</span> meanexpression</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>This creates the below prompt, which can be copied into whichever LLM you prefer:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/LLMprompt.png" class="img-fluid figure-img" style="width:120.0%"></p>
</figure>
</div>
</div>
</div>
<p>When this prompt was passed to ChatGPT5, it produced a writeup that looked like this:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/LLMresponse.png" class="img-fluid figure-img" style="width:110.0%"></p>
</figure>
</div>
</div>
</div>
<p>… and more importantly, the below code ready to be copy-pasted into our analysis:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>cluster_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"0"</span>  <span class="ot">=</span> <span class="st">"Luminal epithelial (low-RNA/quiescent)"</span>,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"1"</span>  <span class="ot">=</span> <span class="st">"Neutrophils (state C, degranulating/secretory-high)"</span>,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"2"</span>  <span class="ot">=</span> <span class="st">"T cells (CD4-like, resting)"</span>,</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"3"</span>  <span class="ot">=</span> <span class="st">"IFN-high epithelial (luminal, inflamed)"</span>,</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"4"</span>  <span class="ot">=</span> <span class="st">"Luminal epithelial (apocrine-leaning; state A)"</span>,</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">"5"</span>  <span class="ot">=</span> <span class="st">"CAFs (matrix-rich fibroblasts)"</span>,</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">"6"</span>  <span class="ot">=</span> <span class="st">"Antigen-presenting myeloid (cDC2 / moDC)"</span>,</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">"7"</span>  <span class="ot">=</span> <span class="st">"Luminal epithelial (apocrine-high; state B)"</span>,</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">"8"</span>  <span class="ot">=</span> <span class="st">"Luminal epithelial (apocrine-high; state C)"</span>,</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">"9"</span>  <span class="ot">=</span> <span class="st">"Neutrophils (state B)"</span>,</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"10"</span> <span class="ot">=</span> <span class="st">"Luminal epithelial (apocrine-secretory)"</span>,</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>  <span class="st">"11"</span> <span class="ot">=</span> <span class="st">"Luminal epithelial (apocrine-high; state D)"</span>,</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>  <span class="st">"12"</span> <span class="ot">=</span> <span class="st">"Luminal epithelial (apocrine-very-high; state E)"</span>,</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>  <span class="st">"13"</span> <span class="ot">=</span> <span class="st">"Macrophages (APOE+/C1QA+ TAMs)"</span>,</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>  <span class="st">"14"</span> <span class="ot">=</span> <span class="st">"Neutrophils (state A; S100A8/A9-max)"</span>,</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>  <span class="st">"15"</span> <span class="ot">=</span> <span class="st">"Epithelial–neutrophil admixture (uncertain)"</span>,</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>  <span class="st">"16"</span> <span class="ot">=</span> <span class="st">"CAFs (cycling)"</span>,</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>  <span class="st">"17"</span> <span class="ot">=</span> <span class="st">"Endothelial (blood/lymphatic mix)"</span>,</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>  <span class="st">"18"</span> <span class="ot">=</span> <span class="st">"Inflamed epithelial with myeloid shadow (uncertain)"</span>,</span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>  <span class="st">"19"</span> <span class="ot">=</span> <span class="st">"Pericytes / vascular SMC"</span>,</span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>  <span class="st">"20"</span> <span class="ot">=</span> <span class="st">"Luminal epithelial (mucociliary-like secretory)"</span>,</span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>  <span class="st">"21"</span> <span class="ot">=</span> <span class="st">"Cycling epithelial"</span>,</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>  <span class="st">"22"</span> <span class="ot">=</span> <span class="st">"Luminal epithelial (baseline state)"</span>,</span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a>  <span class="st">"23"</span> <span class="ot">=</span> <span class="st">"IFN-high epithelial (luminal, inflamed)"</span>,</span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>  <span class="st">"24"</span> <span class="ot">=</span> <span class="st">"Luminal epithelial (mucociliary-like secretory)"</span>,</span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>  <span class="st">"25"</span> <span class="ot">=</span> <span class="st">"Plasma cells"</span>,</span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>  <span class="st">"26"</span> <span class="ot">=</span> <span class="st">"Antigen-presenting cells (B/cDC; cycling, uncertain)"</span>,</span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a>  <span class="st">"27"</span> <span class="ot">=</span> <span class="st">"Cytotoxic T / NK cells"</span></span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a><span class="co"># for now, just uncritically take the names. Clustering QC and possible renaming will come later.</span></span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>metadata<span class="sc">$</span>clust <span class="ot">&lt;-</span> clust <span class="ot">&lt;-</span> cluster_labels[<span class="fu">as.character</span>(unnamedclust)]</span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(meanexpression) <span class="ot">&lt;-</span> cluster_labels[<span class="fu">as.character</span>(<span class="fu">gsub</span>(<span class="st">"cluster"</span>, <span class="st">""</span>, <span class="fu">colnames</span>(unnamedmeanexpression)))]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="clustqc" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> QC of clustering results</h1>
<p>The LLM gave us a good guess at cluster names, but we still need to verify them. To assist in this process, we recommend looking at 4 outputs:</p>
<ol type="1">
<li>The LLM’s justifications for its calls</li>
<li>A heatmap of each cluster’s expression of key marker genes</li>
<li>The clusters’ distributions in space.</li>
<li>The clusters’ UMAP positions</li>
</ol>
<p>The code below draws the key plots:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># heatmap of marker genes - drawn tall so we can read all gene names:</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>() <span class="co"># need to silence the current graphics device for a pheatmap pdf to render properly</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/results/marker heatmap tall.pdf"</span>), <span class="at">height =</span> <span class="dv">17</span>, <span class="at">width =</span> <span class="dv">6</span>)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(<span class="fu">sweep</span>(meanexpression, <span class="dv">1</span>, <span class="fu">pmax</span>(<span class="fl">0.1</span>, <span class="fu">apply</span>(meanexpression, <span class="dv">1</span>, max)), <span class="st">"/"</span>),</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">col =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"darkblue"</span>))(<span class="dv">101</span>),</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">fontsize_row =</span> <span class="dv">6</span>)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a><span class="co"># umap positions</span></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="st">"results/umap_by_cluster.png"</span>, <span class="at">width =</span> <span class="dv">16</span>, <span class="at">height =</span> <span class="dv">16</span>, <span class="at">units =</span> <span class="st">"in"</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>()</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a><span class="co"># xy positions:</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (name <span class="cf">in</span> <span class="fu">unique</span>(clust)) {</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">png</span>(<span class="fu">paste0</span>(<span class="st">"results/cluster xy - "</span>, <span class="fu">make.names</span>(name), <span class="st">".png"</span>), </span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">width =</span> <span class="fu">diff</span>(<span class="fu">range</span>(xy[, <span class="dv">1</span>])) <span class="sc">*</span> <span class="fl">0.5</span>, </span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">height =</span> <span class="fu">diff</span>(<span class="fu">range</span>(xy[, <span class="dv">2</span>])) <span class="sc">*</span> <span class="fl">0.5</span>,</span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">unit =</span> <span class="st">"in"</span>, <span class="at">res =</span> <span class="dv">350</span>)</span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(xy, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.1</span>, </span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> scales<span class="sc">::</span><span class="fu">alpha</span>(cellcols, <span class="fl">0.3</span>)[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(clust))], </span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab =</span> <span class="st">""</span>, <span class="at">ylab =</span> <span class="st">""</span>, <span class="at">xaxt =</span> <span class="st">"n"</span>, <span class="at">yaxt =</span> <span class="st">"n"</span>)</span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">points</span>(xy[clust <span class="sc">==</span> name, ], <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.1</span>, </span>
<span id="cb29-25"><a href="#cb29-25" aria-hidden="true" tabindex="-1"></a>         <span class="at">col =</span> <span class="st">"black"</span>)</span>
<span id="cb29-26"><a href="#cb29-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">legend</span>(<span class="st">"top"</span>, <span class="at">legend =</span> name, <span class="at">bty =</span> <span class="st">"n"</span>)</span>
<span id="cb29-27"><a href="#cb29-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dev.off</span>()</span>
<span id="cb29-28"><a href="#cb29-28" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We obtain:</p>
<p>A .pdf heatmap of marker gene expression, stretched out to allow for scrutinizing gene names (not legible zoomed-out; meant to be explored via zooming and scrolling):</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/tallheatmapsnippet.png" class="img-fluid figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<p>A .png showing each cluster in umap space, giving you another look at which clusters are similar to each other:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/umapbyclustersnippet.png" class="img-fluid figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<p>A .png for each cluster showing its positions in xy space (bold black points atop faded points for other clusters):</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/cluster_xy_snippet.png" class="img-fluid figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
<p>With these plots in hand, you should scrutinize the initial cluster names and correct as needed. Note that you can gloss over immune clusters: we’ll be overriding them later with the HieraType algorithm’s results.</p>
<p>To rename clusters, simply edit the cluster names in the definition of <code>cluster_labels</code> above, and re-run from that code block onward. This step can also be used to merge closely-related clusters by assigning multiple clusters the same name. In our case, we renamed three clusters of cancer cells that the LLM labeled as neutrophils (some of the classic neutrophil markers are also expressed by cancer cells).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>For tumor studies, additional techniques can help classify cancer cells:</p>
<ol type="1">
<li>Cancer cells tend to be extremely transcriptionally active. Clusters with high total counts per cell are usually cancer.</li>
<li>The <a href="https://github.com/Jonyyqn/scMalignantFinder" target="_blank">scMalignantFinder algorithm</a> promises to call cells as cancer vs.&nbsp;non-cancer. We have not tried it, but it seems promising.</li>
<li>For WTX studies, it’s possible to estimate cells’ copy number variants. We have used <a href="https://github.com/Moldia/InSituCNV" target="_blank">insituCNV</a> successfully. A newer method, <a href="https://github.com/seasoncloud/Clonalscope" target="_blank">ClonalScope</a>, also looks promising.</li>
</ol>
</div>
</div>
</section>
<section id="hieratype" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> Optimizing immune cell calling with the HieraType algorithm</h1>
<p>We created the <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/immune-cell-typing-smooth/" target="_blank">HieraType algorithm</a> to classify immune cells in a way that is informed by marker genes. We have found it to consistently outperform other algorithms in immune cell typing. Our standard approach now is to overwrite our initial clustering results with HieraType’s results whenever it calls an immune cell.</p>
<p>Below, we run HieraType in two steps. In the first, we call broad cell types (e.g.&nbsp;epithelial, immune, …) then subclassify the immune cells (B, T, macrophage…)</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="do">### set up hieratype pipeline:</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>pipeline_io_rna <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">make_pipeline</span>(</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">markerslists =</span> <span class="fu">list</span>(</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l1"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_l1</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    ,<span class="st">"l2"</span> <span class="ot">=</span> HieraType<span class="sc">::</span>markerslist_immune</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors =</span> <span class="fu">list</span>(</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"l1"</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">priors_category =</span> <span class="fu">list</span>(</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"l2"</span> <span class="ot">=</span> <span class="st">"immune"</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="cn">TRUE</span>) {</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a>  hieratyperes <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(</span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">pipeline =</span> pipeline_io_rna</span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">counts_matrix =</span> counts</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">totalcounts =</span> metadata<span class="sc">$</span>nCount_RNA</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a>    ,<span class="at">adjacency_matrix =</span> umapobj<span class="sc">$</span>grph[<span class="fu">rownames</span>(counts), <span class="fu">rownames</span>(counts)]</span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(hieratyperes, <span class="at">file =</span> <span class="st">"processed_data/hieratyperes.RDS"</span>)</span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a>  hieratyperes <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"processed_data/hieratyperes.RDS"</span>)</span>
<span id="cb30-25"><a href="#cb30-25" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb30-26"><a href="#cb30-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-27"><a href="#cb30-27" aria-hidden="true" tabindex="-1"></a><span class="co"># further subclassify T-cells:</span></span>
<span id="cb30-28"><a href="#cb30-28" aria-hidden="true" tabindex="-1"></a>tcell_ids <span class="ot">&lt;-</span> hieratyperes<span class="sc">$</span>post_probs<span class="sc">$</span>l1[celltype_granular<span class="sc">==</span><span class="st">"tcell"</span>][[<span class="st">"cell_ID"</span>]]</span>
<span id="cb30-29"><a href="#cb30-29" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="cn">TRUE</span>) {</span>
<span id="cb30-30"><a href="#cb30-30" aria-hidden="true" tabindex="-1"></a>  tcell_typing <span class="ot">&lt;-</span> </span>
<span id="cb30-31"><a href="#cb30-31" aria-hidden="true" tabindex="-1"></a>    HieraType<span class="sc">::</span><span class="fu">run_pipeline</span>(<span class="at">pipeline =</span> HieraType<span class="sc">::</span>pipeline_tcell</span>
<span id="cb30-32"><a href="#cb30-32" aria-hidden="true" tabindex="-1"></a>                 ,<span class="at">counts_matrix =</span> counts[tcell_ids, ]</span>
<span id="cb30-33"><a href="#cb30-33" aria-hidden="true" tabindex="-1"></a>                 ,<span class="at">adjacency_matrix =</span> umapobj<span class="sc">$</span>grph[tcell_ids,tcell_ids]</span>
<span id="cb30-34"><a href="#cb30-34" aria-hidden="true" tabindex="-1"></a>                 ,<span class="at">celltype_call_threshold =</span> <span class="fl">0.5</span></span>
<span id="cb30-35"><a href="#cb30-35" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb30-36"><a href="#cb30-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(tcell_typing, <span class="at">file =</span> <span class="st">"processed_data/tcell_typing.RDS"</span>)</span>
<span id="cb30-37"><a href="#cb30-37" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb30-38"><a href="#cb30-38" aria-hidden="true" tabindex="-1"></a>  tcell_typing <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"processed_data/tcell_typing.RDS"</span>)</span>
<span id="cb30-39"><a href="#cb30-39" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb30-40"><a href="#cb30-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-41"><a href="#cb30-41" aria-hidden="true" tabindex="-1"></a><span class="co"># unify the two levels of hieratype results:</span></span>
<span id="cb30-42"><a href="#cb30-42" aria-hidden="true" tabindex="-1"></a>htclust <span class="ot">&lt;-</span> hieratyperes<span class="sc">$</span>post_probs<span class="sc">$</span>l1<span class="sc">$</span>celltype_granular</span>
<span id="cb30-43"><a href="#cb30-43" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(htclust) <span class="ot">&lt;-</span> hieratyperes<span class="sc">$</span>post_probs<span class="sc">$</span>l1<span class="sc">$</span>cell_ID</span>
<span id="cb30-44"><a href="#cb30-44" aria-hidden="true" tabindex="-1"></a>htclust[tcell_typing<span class="sc">$</span>post_probs<span class="sc">$</span>tmajor<span class="sc">$</span>cell_ID] <span class="ot">&lt;-</span> tcell_typing<span class="sc">$</span>post_probs<span class="sc">$</span>tmajor<span class="sc">$</span>celltype_granular</span>
<span id="cb30-45"><a href="#cb30-45" aria-hidden="true" tabindex="-1"></a>htclust <span class="ot">&lt;-</span> htclust[<span class="fu">match</span>(metadata<span class="sc">$</span>cell_id, <span class="fu">names</span>(htclust))]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now we’ll use a simple marker heatmap to confirm HieraType worked as hoped:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pull out canonical ("index") markers for immune cells:</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>indexmarkers <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unlist</span>(pipeline_io_rna<span class="sc">$</span>markerslists)[<span class="fu">grepl</span>(<span class="st">"index_marker"</span>, <span class="fu">names</span>(<span class="fu">unlist</span>(pipeline_io_rna<span class="sc">$</span>markerslists)))],</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unlist</span>(HieraType<span class="sc">::</span>pipeline_tcell)[<span class="fu">grepl</span>(<span class="st">"index_marker"</span>, <span class="fu">names</span>(<span class="fu">unlist</span>(HieraType<span class="sc">::</span>pipeline_tcell)))]</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>))</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>indexmarkers <span class="ot">&lt;-</span> <span class="fu">intersect</span>(indexmarkers, hvgs)</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a><span class="co"># make a marker heatmap for the hieratype immune cell calls:</span></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>fcmetrics <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">clusterwise_foldchange_metrics</span>(</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">normed  =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(norm[, indexmarkers]),</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">metadata =</span> <span class="fu">data.frame</span>(<span class="st">"cell_ID"</span> <span class="ot">=</span> <span class="fu">names</span>(htclust), <span class="st">"hieratype_call"</span> <span class="ot">=</span> htclust),</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_column =</span> <span class="st">"hieratype_call"</span></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>hm_unsup <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">marker_heatmap</span>(fcmetrics, <span class="at">extras =</span> <span class="cn">NULL</span>)</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hm_unsup)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/hieratype_marker_heatmap.png" class="img-fluid figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
<p>The above heatmap doesn’t show the strongest markers; it shows the most biologically-relevant markers. We can gain further confidence in our immune cell typing results looking at data-driven markers:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>fctbl_unsup <span class="ot">&lt;-</span> </span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  HieraType<span class="sc">::</span><span class="fu">clusterwise_foldchange_metrics</span>(<span class="at">normed  =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(norm),</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">metadata =</span> <span class="fu">data.frame</span>(<span class="st">"cell_ID"</span> <span class="ot">=</span> <span class="fu">names</span>(htclust), <span class="st">"hieratype_call"</span> <span class="ot">=</span> htclust),</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">cluster_column =</span> <span class="st">"hieratype_call"</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="do">## make a marker heatmap</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>hm_unsup <span class="ot">&lt;-</span> HieraType<span class="sc">::</span><span class="fu">marker_heatmap</span>(fctbl_unsup)</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hm_unsup)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/hieratype_marker_heatmap2.png" class="img-fluid figure-img" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
<section id="merging-clustering-and-hieratype-results" class="level3" data-number="8.0.1">
<h3 data-number="8.0.1" class="anchored" data-anchor-id="merging-clustering-and-hieratype-results"><span class="header-section-number">8.0.1</span> Merging clustering and HieraType results</h3>
<p>To finish our cell typing exercise, we now merge the Louvain and HieraType results. All cells called immune by HieraType will get HieraType labels. All other cells will retain their old labels, with one important exception: any cells in an immune Louvain cluster but not called immune by HieraType will be reclassified as “unknown”:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># combine hieratype and clustering:</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>immune.ids <span class="ot">&lt;-</span> hieratyperes<span class="sc">$</span>post_probs<span class="sc">$</span>l1<span class="sc">$</span>cell_ID[</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="sc">!</span><span class="fu">is.element</span>(hieratyperes<span class="sc">$</span>post_probs<span class="sc">$</span>l1<span class="sc">$</span>celltype_thresh, <span class="fu">c</span>(<span class="st">"epithelial"</span>, <span class="st">"fibroblast"</span>, <span class="st">"endothelial"</span>, <span class="st">"smooth_muscle"</span>))]</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>celltype <span class="ot">&lt;-</span> clust</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(celltype) <span class="ot">&lt;-</span> metadata<span class="sc">$</span>cell_id</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>celltype[immune.ids] <span class="ot">&lt;-</span> htclust[immune.ids]</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a><span class="co"># which clusters to reassign: &gt;90% taken by hieratype, and &lt; 5% of total cells left.</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>propoverwritten <span class="ot">&lt;-</span> <span class="fu">by</span>(clust <span class="sc">!=</span> celltype, clust, mean)</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>proppercluster <span class="ot">&lt;-</span> <span class="fu">table</span>(celltype) <span class="sc">/</span> <span class="fu">length</span>(celltype)</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>clusterstodrop <span class="ot">&lt;-</span> <span class="fu">union</span>(</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">intersect</span>(</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(<span class="fu">which</span>(propoverwritten <span class="sc">&gt;</span> <span class="fl">0.9</span>)),</span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(<span class="fu">which</span>(proppercluster <span class="sc">&lt;</span> <span class="fl">0.05</span>))),</span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">intersect</span>(</span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(<span class="fu">which</span>(propoverwritten <span class="sc">&gt;</span> <span class="fl">0.5</span>)),</span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(<span class="fu">which</span>(proppercluster <span class="sc">&lt;</span> <span class="fl">0.01</span>)))</span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb33-21"><a href="#cb33-21" aria-hidden="true" tabindex="-1"></a><span class="co"># find nearest neighbors (in valid clusters) for cells in target clusters:</span></span>
<span id="cb33-22"><a href="#cb33-22" aria-hidden="true" tabindex="-1"></a>cellstoreclassify <span class="ot">&lt;-</span> <span class="fu">names</span>(celltype)[<span class="fu">is.element</span>(celltype, clusterstodrop)]</span>
<span id="cb33-23"><a href="#cb33-23" aria-hidden="true" tabindex="-1"></a>neighborclusts <span class="ot">&lt;-</span> InSituCor<span class="sc">:::</span><span class="fu">neighbor_tabulate</span>(celltype, <span class="at">neighbors =</span> umapobj<span class="sc">$</span>grph)</span>
<span id="cb33-24"><a href="#cb33-24" aria-hidden="true" tabindex="-1"></a>neighborclusts <span class="ot">&lt;-</span> neighborclusts[, <span class="sc">!</span><span class="fu">is.element</span>(<span class="fu">colnames</span>(neighborclusts), clusterstodrop)]</span>
<span id="cb33-25"><a href="#cb33-25" aria-hidden="true" tabindex="-1"></a><span class="co"># reclassify by most common nearest neighbor among valid clusters:</span></span>
<span id="cb33-26"><a href="#cb33-26" aria-hidden="true" tabindex="-1"></a>celltype[cellstoreclassify] <span class="ot">&lt;-</span> <span class="fu">colnames</span>(neighborclusts)[<span class="fu">apply</span>(neighborclusts[cellstoreclassify, ], <span class="dv">1</span>, which.max)]</span>
<span id="cb33-27"><a href="#cb33-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb33-28"><a href="#cb33-28" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(celltype)[<span class="fu">order</span>(<span class="fu">table</span>(celltype))]</span>
<span id="cb33-29"><a href="#cb33-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-30"><a href="#cb33-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-31"><a href="#cb33-31" aria-hidden="true" tabindex="-1"></a><span class="co"># # what to do with the leftover immune cells from clustering?</span></span>
<span id="cb33-32"><a href="#cb33-32" aria-hidden="true" tabindex="-1"></a><span class="co"># setdiff(unique(celltype), unique(htclust))</span></span>
<span id="cb33-33"><a href="#cb33-33" aria-hidden="true" tabindex="-1"></a><span class="co"># louvain_immune_names &lt;- c("Macrophages (APOE+/C1QA+ TAMs)",</span></span>
<span id="cb33-34"><a href="#cb33-34" aria-hidden="true" tabindex="-1"></a><span class="co">#                           "Neutrophils (state A; S100A8/A9-max)",</span></span>
<span id="cb33-35"><a href="#cb33-35" aria-hidden="true" tabindex="-1"></a><span class="co">#                           "Neutrophils (state C, degranulating/secretory-high)",</span></span>
<span id="cb33-36"><a href="#cb33-36" aria-hidden="true" tabindex="-1"></a><span class="co">#                           "T cells (CD4-like, resting)",</span></span>
<span id="cb33-37"><a href="#cb33-37" aria-hidden="true" tabindex="-1"></a><span class="co">#                           "Antigen-presenting myeloid (cDC2 / moDC)",</span></span>
<span id="cb33-38"><a href="#cb33-38" aria-hidden="true" tabindex="-1"></a><span class="co">#                           "Neutrophils (state B)",</span></span>
<span id="cb33-39"><a href="#cb33-39" aria-hidden="true" tabindex="-1"></a><span class="co">#                           "Plasma cells",</span></span>
<span id="cb33-40"><a href="#cb33-40" aria-hidden="true" tabindex="-1"></a><span class="co">#                           "Antigen-presenting cells (B/cDC; cycling, uncertain)",</span></span>
<span id="cb33-41"><a href="#cb33-41" aria-hidden="true" tabindex="-1"></a><span class="co">#                           "Cytotoxic T / NK cells")</span></span>
<span id="cb33-42"><a href="#cb33-42" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb33-43"><a href="#cb33-43" aria-hidden="true" tabindex="-1"></a><span class="co"># # check overlaps, making sure there aren't huge numbers of cells called immune by Louvain but not HieraType:</span></span>
<span id="cb33-44"><a href="#cb33-44" aria-hidden="true" tabindex="-1"></a><span class="co"># par(mar = c(22,4,2,1))</span></span>
<span id="cb33-45"><a href="#cb33-45" aria-hidden="true" tabindex="-1"></a><span class="co"># barplot(table(celltype), las = 2, </span></span>
<span id="cb33-46"><a href="#cb33-46" aria-hidden="true" tabindex="-1"></a><span class="co">#         col = c("grey", "orange", "cornflowerblue")[1 + is.element(names(table(celltype)), louvain_immune_names) + 2 * is.element(names(table(celltype)), unique(htclust))])</span></span>
<span id="cb33-47"><a href="#cb33-47" aria-hidden="true" tabindex="-1"></a><span class="co"># legend("topleft", fill = c("grey", "orange", "cornflowerblue"), legend = c("Louvain non-immune", "Louvain immune", "HieraType immune"))</span></span>
<span id="cb33-48"><a href="#cb33-48" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb33-49"><a href="#cb33-49" aria-hidden="true" tabindex="-1"></a><span class="co"># # overwrite left-behind louvain immune clusters:</span></span>
<span id="cb33-50"><a href="#cb33-50" aria-hidden="true" tabindex="-1"></a><span class="co"># celltype[is.element(celltype, louvain_immune_names)] &lt;- "immune_unspecified"</span></span>
<span id="cb33-51"><a href="#cb33-51" aria-hidden="true" tabindex="-1"></a>metadata<span class="sc">$</span>celltype <span class="ot">&lt;-</span> celltype</span>
<span id="cb33-52"><a href="#cb33-52" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(celltype, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/celltype.RDS"</span>))</span>
<span id="cb33-53"><a href="#cb33-53" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(metadata, <span class="at">file =</span> <span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/metadata.RDS"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="figures/celltype_merge_barplot.png" class="img-fluid figure-img" style="width:75.0%"></p>
</figure>
</div>
</div>
</div>
<p>Note that in the above plot, we have 3 tall bars from neutrophil clusters called by Louvain. This result would be concerning for a merge of HieraType and Louvain, implying many immune cells were getting left behind. In our case, these are the 3 clusters that the LLM misclassified; we renamed them to “cancer” for our analysis.</p>
<p>Once our cell types are complete, we re-generate the spatial plots:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># xy positions:</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (name <span class="cf">in</span> <span class="fu">unique</span>(celltype)) {</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">png</span>(<span class="fu">paste0</span>(<span class="st">"results/celltype xy - "</span>, <span class="fu">make.names</span>(name), <span class="st">".png"</span>), </span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">width =</span> <span class="fu">diff</span>(<span class="fu">range</span>(xy[, <span class="dv">1</span>])) <span class="sc">*</span> <span class="fl">0.5</span>, </span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">height =</span> <span class="fu">diff</span>(<span class="fu">range</span>(xy[, <span class="dv">2</span>])) <span class="sc">*</span> <span class="fl">0.5</span>,</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">unit =</span> <span class="st">"in"</span>, <span class="at">res =</span> <span class="dv">350</span>)</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(xy, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.1</span>, </span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> scales<span class="sc">::</span><span class="fu">alpha</span>(cellcols, <span class="fl">0.3</span>)[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(celltype))], </span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab =</span> <span class="st">""</span>, <span class="at">ylab =</span> <span class="st">""</span>, <span class="at">xaxt =</span> <span class="st">"n"</span>, <span class="at">yaxt =</span> <span class="st">"n"</span>)</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">points</span>(xy[celltype <span class="sc">==</span> name, ], <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.1</span>, </span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">col =</span> <span class="st">"black"</span>)</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">legend</span>(<span class="st">"top"</span>, <span class="at">legend =</span> name, <span class="at">bty =</span> <span class="st">"n"</span>)</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dev.off</span>()</span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="conclusion" class="level1" data-number="9">
<h1 data-number="9"><span class="header-section-number">9</span> Conclusion</h1>
<p>The code above provides a direct path through the basic preprocessing of CosMx data. With these results secured, you’re ready to proceed to the rewarding parts of data analysis: exploring your data and testing hypotheses. We will have posts on these subjects soon. For now, here are some reasonable next steps:</p>
<p>Exploratory analysis:</p>
<ul>
<li>If your experiment has baseline / control tissues, try <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/insitudiff/" target="_blank">InSituDiff</a>.</li>
<li>To explore genes’ spatial correlation in studies without control tissues (e.g.&nbsp;tumors), try <a href="https://github.com/Nanostring-Biostats/InSitucor" target="_blank">InSituCor</a>.</li>
<li>Calculate pathway scores using <a href="https://github.com/aertslab/AUCell" target="_blank">AUCell</a>.</li>
</ul>
<p>Hypothesis testing:</p>
<ul>
<li>Use <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/smiDE/" target="_blank">smiDE</a> for differential expression testing.</li>
<li>Try Ligand-Receptor analysis (many packages exist; we have no recommendation yet).</li>
</ul>
<p>We recommend containing those analyses in standalone scripts, each beginning by loading the data we’ve prepared here:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load basic data:</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/counts.RDS"</span>))</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/metadata.RDS"</span>))</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>xy <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(outdir, <span class="st">"/processed_data/xy.RDS"</span>))</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="co"># get normalized data if desired:</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>scale_row <span class="ot">&lt;-</span> <span class="fu">mean</span>(metadata<span class="sc">$</span>nCount_RNA)<span class="sc">/</span>metadata<span class="sc">$</span>nCount_RNA</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>norm <span class="ot">&lt;-</span> counts</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>norm<span class="sc">@</span>x <span class="ot">&lt;-</span> norm<span class="sc">@</span>x <span class="sc">*</span> scale_row[norm<span class="sc">@</span>i <span class="sc">+</span> <span class="dv">1</span>L]</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="co"># likely to be useful: compute nearest neighbors:</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>neighbors <span class="ot">&lt;-</span> InSituCor<span class="sc">:::</span><span class="fu">nearestNeighborGraph</span>(xy[, <span class="dv">1</span>], xy[, <span class="dv">2</span>], <span class="at">N =</span> <span class="dv">50</span>, <span class="at">subset =</span> metadata<span class="sc">$</span>slide_ID)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/nanostring-biostats\.github\.io\/CosMx-Analysis-Scratch-Space");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="../../license.html">
<p>License</p>
</a>
  </li>  
</ul>
    <div class="cookie-consent-footer"><a href="#" id="open_preferences_center">Cookie Preferences</a></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>