<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Evelyn Metzger">

<title>3&nbsp; Quality Processing – A Guide to Whole Transcriptome CosMx&lt;sup&gt;®&lt;/sup&gt; SMI Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./processing.html" rel="next">
<link href="./data-ingestion.html" rel="prev">
<link href="./assets/rocket.svg" rel="icon" type="image/svg+xml">
<script src="site_libs/cookie-consent/cookie-consent.js"></script>
<link href="site_libs/cookie-consent/cookie-consent.css" rel="stylesheet">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark-707d8167ce6003fca903bfe2be84ab7f.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" class="quarto-color-scheme-extra" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-00066ec49057a78057bb708cd7ecd022.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark-cb8585788486bab3a0512e4422583b7d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="site_libs/bootstrap/bootstrap-00066ec49057a78057bb708cd7ecd022.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme-extra" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/live-runtime/live-runtime.js" type="module"></script>
<link href="site_libs/quarto-contrib/live-runtime/live-runtime.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-29W4MW0Y2W"></script>

<script type="text/plain" cookie-consent="tracking">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-29W4MW0Y2W', { 'anonymize_ip': true});
</script>

<script type="text/javascript" charset="UTF-8">
document.addEventListener('DOMContentLoaded', function () {
cookieconsent.run({
  "notice_banner_type":"simple",
  "consent_type":"implied",
  "palette":"light",
  "language":"en",
  "page_load_consent_levels":["strictly-necessary","functionality","tracking","targeting"],
  "notice_banner_reject_button_hide":false,
  "preferences_center_close_button_hide":false,
  "website_name":""
  ,
"language":"en"
  });
});
</script> 
  
<script type="module" src="site_libs/quarto-ojs/quarto-ojs-runtime.js"></script>
<link href="site_libs/quarto-ojs/quarto-ojs.css" rel="stylesheet">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating nav-fixed slimcontent quarto-light"><script id="quarto-html-before-body" type="application/javascript">
    const toggleBodyColorMode = (bsSheetEl) => {
      const mode = bsSheetEl.getAttribute("data-mode");
      const bodyEl = window.document.querySelector("body");
      if (mode === "dark") {
        bodyEl.classList.add("quarto-dark");
        bodyEl.classList.remove("quarto-light");
      } else {
        bodyEl.classList.add("quarto-light");
        bodyEl.classList.remove("quarto-dark");
      }
    }
    const toggleBodyColorPrimary = () => {
      const bsSheetEl = window.document.querySelector("link#quarto-bootstrap:not([rel=disabled-stylesheet])");
      if (bsSheetEl) {
        toggleBodyColorMode(bsSheetEl);
      }
    }
    const setColorSchemeToggle = (alternate) => {
      const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
      for (let i=0; i < toggles.length; i++) {
        const toggle = toggles[i];
        if (toggle) {
          if (alternate) {
            toggle.classList.add("alternate");
          } else {
            toggle.classList.remove("alternate");
          }
        }
      }
    };
    const toggleColorMode = (alternate) => {
      // Switch the stylesheets
      const primaryStylesheets = window.document.querySelectorAll('link.quarto-color-scheme:not(.quarto-color-alternate)');
      const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
      manageTransitions('#quarto-margin-sidebar .nav-link', false);
      if (alternate) {
        // note: dark is layered on light, we don't disable primary!
        enableStylesheet(alternateStylesheets);
        for (const sheetNode of alternateStylesheets) {
          if (sheetNode.id === "quarto-bootstrap") {
            toggleBodyColorMode(sheetNode);
          }
        }
      } else {
        disableStylesheet(alternateStylesheets);
        enableStylesheet(primaryStylesheets)
        toggleBodyColorPrimary();
      }
      manageTransitions('#quarto-margin-sidebar .nav-link', true);
      // Switch the toggles
      setColorSchemeToggle(alternate)
      // Hack to workaround the fact that safari doesn't
      // properly recolor the scrollbar when toggling (#1455)
      if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
        manageTransitions("body", false);
        window.scrollTo(0, 1);
        setTimeout(() => {
          window.scrollTo(0, 0);
          manageTransitions("body", true);
        }, 40);
      }
    }
    const disableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        stylesheet.rel = 'disabled-stylesheet';
      }
    }
    const enableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        if(stylesheet.rel !== 'stylesheet') { // for Chrome, which will still FOUC without this check
          stylesheet.rel = 'stylesheet';
        }
      }
    }
    const manageTransitions = (selector, allowTransitions) => {
      const els = window.document.querySelectorAll(selector);
      for (let i=0; i < els.length; i++) {
        const el = els[i];
        if (allowTransitions) {
          el.classList.remove('notransition');
        } else {
          el.classList.add('notransition');
        }
      }
    }
    const isFileUrl = () => {
      return window.location.protocol === 'file:';
    }
    const hasAlternateSentinel = () => {
      let styleSentinel = getColorSchemeSentinel();
      if (styleSentinel !== null) {
        return styleSentinel === "alternate";
      } else {
        return false;
      }
    }
    const setStyleSentinel = (alternate) => {
      const value = alternate ? "alternate" : "default";
      if (!isFileUrl()) {
        window.localStorage.setItem("quarto-color-scheme", value);
      } else {
        localAlternateSentinel = value;
      }
    }
    const getColorSchemeSentinel = () => {
      if (!isFileUrl()) {
        const storageValue = window.localStorage.getItem("quarto-color-scheme");
        return storageValue != null ? storageValue : localAlternateSentinel;
      } else {
        return localAlternateSentinel;
      }
    }
    const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
      const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
      const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
      let newTheme = '';
      if(authorPrefersDark) {
        newTheme = isAlternate ? baseTheme : alternateTheme;
      } else {
        newTheme = isAlternate ? alternateTheme : baseTheme;
      }
      const changeGiscusTheme = () => {
        // From: https://github.com/giscus/giscus/issues/336
        const sendMessage = (message) => {
          const iframe = document.querySelector('iframe.giscus-frame');
          if (!iframe) return;
          iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
        }
        sendMessage({
          setConfig: {
            theme: newTheme
          }
        });
      }
      const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
      if (isGiscussLoaded) {
        changeGiscusTheme();
      }
    };
    const authorPrefersDark = false;
    const darkModeDefault = authorPrefersDark;
      document.querySelector('link#quarto-text-highlighting-styles.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
      document.querySelector('link#quarto-bootstrap.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
    let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
    // Dark / light mode switch
    window.quartoToggleColorScheme = () => {
      // Read the current dark / light value
      let toAlternate = !hasAlternateSentinel();
      toggleColorMode(toAlternate);
      setStyleSentinel(toAlternate);
      toggleGiscusIfUsed(toAlternate, darkModeDefault);
      window.dispatchEvent(new Event('resize'));
    };
    // Switch to dark mode if need be
    if (hasAlternateSentinel()) {
      toggleColorMode(true);
    } else {
      toggleColorMode(false);
    }
  </script>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">A Guide to Whole Transcriptome CosMx<sup>®</sup> SMI Analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="./../../index.html"> <i class="bi bi-house-door" role="img" aria-label="Back to main website">
</i> 
<span class="menu-text">Back to Blog</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./quality-processing.html">Processing Data</a></li><li class="breadcrumb-item"><a href="./quality-processing.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Quality Processing</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./quality-processing.html">Processing Data</a></li><li class="breadcrumb-item"><a href="./quality-processing.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Quality Processing</span></a></li></ol></nav>
      <div class="quarto-title-block"><div><h1 class="title"><span id="sec-quality-processing" class="quarto-section-identifier"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Quality Processing</span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
            <p class="subtitle lead">Running quality processing steps</p>
                      </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliations</div>
    
      <div class="quarto-title-meta-contents">
      <p class="author">Evelyn Metzger <a href="https://orcid.org/0000-0002-4074-9003" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              Bruker Spatial Biology
            </p>
          <p class="affiliation">
              Github: <a href="https://github.com/eveilyeverafter" target="_blank">eveilyeverafter</a>
            </p>
        </div>
    </div>

  <div class="quarto-title-meta">

        
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      </a>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./how-to-use.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">How to Use This Book</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Getting Started</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction.html" class="sidebar-item-text sidebar-link"><i class="bi bi-rocket-takeoff-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-ingestion.html" class="sidebar-item-text sidebar-link"><i class="bi bi-hdd-network-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data Wrangling</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Processing Data</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quality-processing.html" class="sidebar-item-text sidebar-link active"><i class="bi bi-clipboard2-data-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Quality Processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./processing.html" class="sidebar-item-text sidebar-link"><i class="bi bi-rulers" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Normalization, Reduction, and Clustering</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spatial-domains.html" class="sidebar-item-text sidebar-link"><i class="bi bi-diagram-3-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Spatial Domains</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cell-typing.html" class="sidebar-item-text sidebar-link"><i class="bi bi-layout-wtf" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Cell Typing</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Tertiary Analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pathways.html" class="sidebar-item-text sidebar-link"><i class="bi bi-grid-1x2-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Pathways</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Conclusion</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./conclusion.html" class="sidebar-item-text sidebar-link"><i class="bi bi-check-square-fill" role="img">
</i> 
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Conclusion</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Additonal Resources</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-fov-qc" id="toc-sec-fov-qc" class="nav-link active" data-scroll-target="#sec-fov-qc"><span class="header-section-number">3.1</span> FOV QC</a></li>
  <li><a href="#cell-level-qc" id="toc-cell-level-qc" class="nav-link" data-scroll-target="#cell-level-qc"><span class="header-section-number">3.2</span> Cell-level QC</a>
  <ul>
  <li><a href="#counts-and-features" id="toc-counts-and-features" class="nav-link" data-scroll-target="#counts-and-features"><span class="header-section-number">3.2.1</span> Counts and Features</a></li>
  <li><a href="#optional-cells-near-fov-borders" id="toc-optional-cells-near-fov-borders" class="nav-link" data-scroll-target="#optional-cells-near-fov-borders"><span class="header-section-number">3.2.2</span> (Optional) Cells Near FOV Borders</a></li>
  </ul></li>
  <li><a href="#regional-level-cell-qc" id="toc-regional-level-cell-qc" class="nav-link" data-scroll-target="#regional-level-cell-qc"><span class="header-section-number">3.3</span> Regional-level cell QC</a></li>
  <li><a href="#combining-flags" id="toc-combining-flags" class="nav-link" data-scroll-target="#combining-flags"><span class="header-section-number">3.4</span> Combining Flags</a></li>
  <li><a href="#filtering" id="toc-filtering" class="nav-link" data-scroll-target="#filtering"><span class="header-section-number">3.5</span> Filtering</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block page-columns page-full" id="quarto-document-content">






<p>This section is critical in any data analysis. We’ll pass the data through a series of filters to ensure the downstream processing steps contain high-quality data. We do this by screening and flagging any FOVs that might have lower quality and running a counts-based filter.</p>
<div class="cell">
<details class="code-fold">
<summary>R code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-1"><pre class="sourceCode r code-annotation-code code-with-copy"><code class="sourceCode r"><span id="annotated-cell-1-1"><a href="#annotated-cell-1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"./preamble.R"</span>)</span>
<span id="annotated-cell-1-2"><a href="#annotated-cell-1-2" aria-hidden="true" tabindex="-1"></a>reticulate<span class="sc">::</span><span class="fu">source_python</span>(<span class="st">"./preamble.py"</span>)</span>
<span id="annotated-cell-1-3"><a href="#annotated-cell-1-3" aria-hidden="true" tabindex="-1"></a>analysis_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">getwd</span>(), <span class="st">"analysis_results"</span>)</span>
<span id="annotated-cell-1-4"><a href="#annotated-cell-1-4" aria-hidden="true" tabindex="-1"></a>input_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"input_files"</span>)</span>
<span id="annotated-cell-1-5"><a href="#annotated-cell-1-5" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"output_files"</span>)</span>
<span id="annotated-cell-1-6"><a href="#annotated-cell-1-6" aria-hidden="true" tabindex="-1"></a>analysis_asset_dir <span class="ot">&lt;-</span> <span class="st">"./assets/analysis_results"</span></span>
<span id="annotated-cell-1-7"><a href="#annotated-cell-1-7" aria-hidden="true" tabindex="-1"></a>qc_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(output_dir, <span class="st">"qc"</span>)</span>
<span id="annotated-cell-1-8"><a href="#annotated-cell-1-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(qc_dir)){</span>
<span id="annotated-cell-1-9"><a href="#annotated-cell-1-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(qc_dir, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="annotated-cell-1-10"><a href="#annotated-cell-1-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="annotated-cell-1-11"><a href="#annotated-cell-1-11" aria-hidden="true" tabindex="-1"></a>results_list_file <span class="ot">=</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"results_list.rds"</span>)</span>
<span id="annotated-cell-1-12"><a href="#annotated-cell-1-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(results_list_file)){</span>
<span id="annotated-cell-1-13"><a href="#annotated-cell-1-13" aria-hidden="true" tabindex="-1"></a>  results_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="annotated-cell-1-14"><a href="#annotated-cell-1-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="annotated-cell-1-15"><a href="#annotated-cell-1-15" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="annotated-cell-1-16"><a href="#annotated-cell-1-16" aria-hidden="true" tabindex="-1"></a>  results_list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(results_list_file)  </span>
<span id="annotated-cell-1-17"><a href="#annotated-cell-1-17" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>Load the dataset we generated in <a href="data-ingestion.html" class="quarto-xref"><span>Chapter 2</span></a>.</p>
<div class="cell">
<details class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>py<span class="sc">$</span>adata <span class="ot">&lt;-</span> py<span class="sc">$</span>ad<span class="sc">$</span><span class="fu">read_h5ad</span>(py<span class="sc">$</span>os<span class="sc">$</span>path<span class="sc">$</span><span class="fu">join</span>(analysis_dir, <span class="st">"anndata-0-initial.h5ad"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<section id="sec-fov-qc" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="sec-fov-qc"><span class="header-section-number">3.1</span> FOV QC</h2>
<p>Lower quality FOVs generally result in reduced overall gene expression or reduced signal from select genes. Potential causes for lower FOV quality (e.g.&nbsp;lower relative signal for all/select genes for a given FOV compared to majority of FOVs) include tissue/section quality, high autofluorescence, and inadequate fiducials. We can quantify this signal loss using the FOV QC method as described recently in this <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/fov-qc/index.html#fov-quality">scratch space tool</a>. For a more detailed explanation of the the FOV QC functions, expand the box below.</p>
<p>Since the FOV QC code was written in R, convert relevant annData components to R objects.</p>
<div class="cell">
<details class="code-fold">
<summary>Python Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>meta <span class="op">=</span> adata.obs.copy()</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>expr <span class="op">=</span> adata.X.astype(<span class="st">'float64'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>Run the <code>runFOVQC</code> routine. The <code>source</code> call below will load several functions into memory. Since this dataset is from WTX, we’ll use the barcode patterns from <code>Hs_WTX</code>.</p>
<div class="cell">
<details class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-4"><pre class="sourceCode r code-annotation-code code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-4-1"><a href="#annotated-cell-4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># source the necessary functions:</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="1">1</button><span id="annotated-cell-4-2" class="code-annotation-target"><a href="#annotated-cell-4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"https://raw.githubusercontent.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/Main/_code/FOV%20QC/FOV%20QC%20utils.R"</span>)</span>
<span id="annotated-cell-4-3"><a href="#annotated-cell-4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-4-4"><a href="#annotated-cell-4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-4-5"><a href="#annotated-cell-4-5" aria-hidden="true" tabindex="-1"></a>allbarcodes <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">url</span>(<span class="st">"https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/raw/Main/_code/FOV%20QC/barcodes_by_panel.RDS"</span>))</span>
<span id="annotated-cell-4-6"><a href="#annotated-cell-4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-4-7"><a href="#annotated-cell-4-7" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'barcodes'</span>]] <span class="ot">&lt;-</span> allbarcodes[<span class="st">'Hs_WTX'</span>]</span>
<span id="annotated-cell-4-8"><a href="#annotated-cell-4-8" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-4" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="2" data-code-annotation="1">For historic reasons, the characters in the barcode string are duplicated.</span>
</dd>
</dl>
</div>
</div>
<div class="noteworthybox callout-collapsible" title="Understanding the FOV QC parameters" data-collapse="true">
<div class="noteworthybox-header" aria-expanded="false" data-bs-target="#custom-box-body-1" data-bs-toggle="collapse">
<div class="noteworthybox-icon">

</div>
<p>Understanding the FOV QC parameters</p>
<div class="callout-tools">
<a class="callout-collapse-toggle"></a>
</div>
</div>
<div id="custom-box-body-1" class="noteworthybox-body collapse">
<p>This deep dive shows a how the two tuning parameters of the FOV QC function, <code>max_prop_loss</code> and <code>max_totalcount_loss</code>), affect the results.</p>
<p>The parameter <code>max_prop_loss</code> sets a tolerance for the maximum allowable proportion of signal loss for any single barcode bit before an FOV is flagged. For example, if we set <code>max_prop_loss</code> to 0.3, it means we’d accept a given <code>bit</code> given if it has lost 30% of its original strength. While <code>max_prop_loss</code> focuses on the errors in a single specific barcode <code>bit</code>, <code>max_totalcounts_loss</code> looks at the bigger picture. It asks, “Is this entire FOV significantly dimmer or less populated with transcripts than it should be?” This can help detect broader issues like poor tissue quality, focusing problems, or reagent failures in that specific area.</p>
<p>Let’s see what happens to the number of flagged FOVs when adjust these parameters.</p>
<div class="cell" data-eval="true">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code hidden" id="cb3" data-startfrom="261" data-source-offset="-0"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 260;"><span id="cb3-261"><a href="#cb3-261" aria-hidden="true" tabindex="-1"></a>viewof max_prop_loss <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">range</span>([<span class="fl">0.1</span><span class="op">,</span> <span class="fl">0.6</span>]<span class="op">,</span> {</span>
<span id="cb3-262"><a href="#cb3-262" aria-hidden="true" tabindex="-1"></a>  <span class="dt">step</span><span class="op">:</span> <span class="fl">0.1</span><span class="op">,</span></span>
<span id="cb3-263"><a href="#cb3-263" aria-hidden="true" tabindex="-1"></a>  <span class="dt">label</span><span class="op">:</span> <span class="st">"Max Proportion Loss"</span></span>
<span id="cb3-264"><a href="#cb3-264" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb3-265"><a href="#cb3-265" aria-hidden="true" tabindex="-1"></a>viewof max_totalcounts_loss <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">range</span>([<span class="fl">0.1</span><span class="op">,</span> <span class="fl">0.6</span>]<span class="op">,</span> {</span>
<span id="cb3-266"><a href="#cb3-266" aria-hidden="true" tabindex="-1"></a>  <span class="dt">step</span><span class="op">:</span> <span class="fl">0.1</span><span class="op">,</span></span>
<span id="cb3-267"><a href="#cb3-267" aria-hidden="true" tabindex="-1"></a>  <span class="dt">label</span><span class="op">:</span> <span class="st">"Max Total Counts Loss"</span></span>
<span id="cb3-268"><a href="#cb3-268" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-1-1" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-1-2" data-nodetype="declaration">

</div>
</div>
</div>
</div>
<div class="cell" data-edit="false" data-define="fov_locations">
<div>
<div id="webr-1" class="exercise-cell">

</div>
<script type="webr-1-contents">
eyJhdHRyIjp7ImRlZmluZSI6WyJmb3ZfbG9jYXRpb25zIl0sImVjaG8iOmZhbHNlLCJlZGl0IjpmYWxzZSwiZXZhbCI6dHJ1ZX0sImNvZGUiOiJcbiMgVGhpcyBqdXN0IHJlYWRzIHRoZSBGT1YgbG9jYXRpb25zIHBhcnF1ZXQgYW5kIG1ha2VzIHRoZSBkYXRhXG4jIGF2YWlsYWJsZSB0byB0byB3ZWJyLiBJdCBvbmx5IG5lZWRzIHRvIGJlIGxvYWRlZCBvbmNlLlxuY29uMCA8LSBkYkNvbm5lY3QoZHVja2RiKCkpXG5mb3ZfbG9jYXRpb25zIDwtIGRiR2V0UXVlcnkoXG4gIGNvbjAsIFxuICBcIlNFTEVDVCAqIEZST00gcmVhZF9wYXJxdWV0KCdhc3NldHMvaW50ZXJhY3RpdmVzL2Zvdl9sb2NhdGlvbnMucGFycXVldCcpO1wiXG4gIClcbmRiRGlzY29ubmVjdChjb24wLCBzaHV0ZG93biA9IFRSVUUpIn0=
</script>
</div>
</div>
<div class="cell" data-edit="false" data-input="[&quot;max_prop_loss&quot;,&quot;max_totalcounts_loss&quot;]" data-define="filtered_data">
<div>
<div id="webr-2" class="exercise-cell">

</div>
<script type="webr-2-contents">
eyJhdHRyIjp7ImRlZmluZSI6WyJmaWx0ZXJlZF9kYXRhIl0sImVjaG8iOmZhbHNlLCJlZGl0IjpmYWxzZSwiZXZhbCI6dHJ1ZSwiaW5wdXQiOlsibWF4X3Byb3BfbG9zcyIsIm1heF90b3RhbGNvdW50c19sb3NzIl19LCJjb2RlIjoiXG4jIFRoaXMgd2lsbCByZS1ydW4gZXZlcnkgdGltZSBtYXhfcHJvcF9sb3NzIGFuZCBtYXhfdG90YWxjb3VudHNfbG9zc1xuIyBjaGFuZ2UgKHJlYWN0aXZlIHdpdGggT0pTIHNsaWRlcnMpXG5cbiMgY29ubmVjdGlvbiB0byBhbiBpbi1tZW1vcnkgRHVja0RCIGRhdGFiYXNlXG5jb24gPC0gZGJDb25uZWN0KGR1Y2tkYigpKVxuXG4jIGJ1aWxkIHRoZSBTUUwgcXVlcnlcbnF1ZXJ5IDwtIHBhc3RlMChcbiAgXCJTRUxFQ1QgKiBGUk9NICdhc3NldHMvaW50ZXJhY3RpdmVzL2Zvdl9xY192YXJ5LnBhcnF1ZXQnIFdIRVJFIG1heF9wcm9wX2xvc3MgPSBcIixcbiAgbWF4X3Byb3BfbG9zcyxcbiAgXCIgQU5EIG1heF90b3RhbGNvdW50c19sb3NzID0gXCIsXG4gIG1heF90b3RhbGNvdW50c19sb3NzXG4pXG5cbnF1ZXJ5X2RhdGEgPC0gZGJHZXRRdWVyeShjb24sIHF1ZXJ5KVxuXG5kYkRpc2Nvbm5lY3QoY29uLCBzaHV0ZG93biA9IFRSVUUpXG5cbmZpbHRlcmVkX2RhdGEgPC0gZm92X2xvY2F0aW9uc1xuZmlsdGVyZWRfZGF0YSRmbGFnZ2VkIDwtIGlmZWxzZShcbiAgZmlsdGVyZWRfZGF0YSRGT1YgJWluJSBxdWVyeV9kYXRhJEZPViwgVFJVRSwgRkFMU0UpIn0=
</script>
</div>
</div>
<p>This code below will plot the spatial distribution of failed FOVs. Feel free to adjust the code, if desired.</p>
<div class="cell" data-autorun="true" data-input="filtered_data">
<div>
<div id="webr-3" class="exercise-cell">

</div>
<script type="webr-3-contents">
eyJhdHRyIjp7ImF1dG9ydW4iOnRydWUsImVkaXQiOnRydWUsImV2YWwiOnRydWUsImlucHV0IjpbImZpbHRlcmVkX2RhdGEiXX0sImNvZGUiOiJcbnAgPC0gZ2dwbG90KGZpbHRlcmVkX2RhdGEsIFxuICAgICAgICAgYWVzKHg9eF9nbG9iYWxfbW0sIFxuICAgICAgICAgICAgIHk9eV9nbG9iYWxfbW0sIFxuICAgICAgICAgICAgIGNvbG9yPWZhY3RvcihmbGFnZ2VkKSxcbiAgICAgICAgICAgICBsYWJlbD1GT1YpKSArIFxuICBnZW9tX3BvaW50KCkgKyB0aGVtZV9idygpICsgXG4gIGNvb3JkX2ZpeGVkKCkgKyBcbiAgc2NhbGVfY29sb3JfbWFudWFsKHZhbHVlcz1jKGBUUlVFYD0ncmVkJyxcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIGBGQUxTRWA9J2dyZXknKSlcbnBsb3RseTo6Z2dwbG90bHkocCkifQ==
</script>
</div>
</div>
</div>
</div>
<p>With the functions loaded, execute the <code>runFOVQC</code> function. For this example, I choose a less restrictive parameter set (<code>max_prop_loss</code> = 0.4 and <code>max_totalcounts_loss</code> = 0.3).</p>
<p>Plot the flagged FOVs in R simply with <code>mapFlaggedFOVs(res)</code> function to get the spatial layout of any FOVs that were flagged. In this case (<a href="#fig-qc-fov-1" class="quarto-xref">Figure&nbsp;<span>3.1</span></a>), you can see some FOVs that were flagged along the tissue perimeter.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-qc-fov-1" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qc-fov-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/qc/flagged_fovs.png" class="img-fluid figure-img" width="1750">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qc-fov-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.1: Spatial location of the flagged FOVs.
</figcaption>
</figure>
</div>
</div>
</div>
<p>For the targets that were flagged, plot how many FOVs they were flagged in.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-5"><pre class="sourceCode r code-annotation-code code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-5-1"><a href="#annotated-cell-5-1" aria-hidden="true" tabindex="-1"></a>flagged_genes <span class="ot">&lt;-</span> res<span class="sc">$</span>flagged_fov_x_gene[,<span class="dv">2</span>]</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="1">1</button><span id="annotated-cell-5-2" class="code-annotation-target"><a href="#annotated-cell-5-2" aria-hidden="true" tabindex="-1"></a>flagged_genes <span class="ot">&lt;-</span> flagged_genes[<span class="sc">!</span><span class="fu">is.na</span>(flagged_genes)]</span>
<span id="annotated-cell-5-3"><a href="#annotated-cell-5-3" aria-hidden="true" tabindex="-1"></a>flagged_genes <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">rev</span>(<span class="fu">sort</span>(<span class="fu">table</span>(flagged_genes))))</span>
<span id="annotated-cell-5-4"><a href="#annotated-cell-5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(flagged_genes)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">'target'</span></span>
<span id="annotated-cell-5-5"><a href="#annotated-cell-5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-5-6"><a href="#annotated-cell-5-6" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">target=</span>py<span class="sc">$</span>adata<span class="sc">$</span>var_names<span class="sc">$</span><span class="fu">to_list</span>())</span>
<span id="annotated-cell-5-7"><a href="#annotated-cell-5-7" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">merge</span>(df, flagged_genes, <span class="at">by=</span><span class="st">"target"</span>, <span class="at">all.x=</span><span class="cn">TRUE</span>, <span class="at">all.y=</span><span class="cn">FALSE</span>)</span>
<span id="annotated-cell-5-8"><a href="#annotated-cell-5-8" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>Freq[<span class="fu">is.na</span>(df<span class="sc">$</span>Freq)] <span class="ot">&lt;-</span> <span class="dv">0</span>L</span>
<span id="annotated-cell-5-9"><a href="#annotated-cell-5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-5-10"><a href="#annotated-cell-5-10" aria-hidden="true" tabindex="-1"></a>py<span class="sc">$</span>adata<span class="sc">$</span>var[<span class="st">'fov_qc_flagged'</span>] <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(df<span class="sc">$</span>Freq)</span>
<span id="annotated-cell-5-11"><a href="#annotated-cell-5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-5-12"><a href="#annotated-cell-5-12" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(flagged_genes, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(Freq), <span class="at">fill =</span> <span class="fu">factor</span>(Freq))) <span class="sc">+</span></span>
<span id="annotated-cell-5-13"><a href="#annotated-cell-5-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="annotated-cell-5-14"><a href="#annotated-cell-5-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">stat=</span><span class="st">'count'</span>, <span class="fu">aes</span>(<span class="at">label=</span><span class="fu">after_stat</span>(count)), <span class="at">vjust=</span><span class="sc">-</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="annotated-cell-5-15"><a href="#annotated-cell-5-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Number of FOVs"</span>, <span class="at">y =</span> <span class="st">"Number of Flagged Genes"</span>) <span class="sc">+</span></span>
<span id="annotated-cell-5-16"><a href="#annotated-cell-5-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="annotated-cell-5-17"><a href="#annotated-cell-5-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">breaks =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">13</span>) <span class="sc">+</span> </span>
<span id="annotated-cell-5-18"><a href="#annotated-cell-5-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="annotated-cell-5-19"><a href="#annotated-cell-5-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-5-20"><a href="#annotated-cell-5-20" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"genes_impacted_in_flagged_fovs.png"</span>),</span>
<span id="annotated-cell-5-21"><a href="#annotated-cell-5-21" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">4</span>, <span class="at">height=</span><span class="dv">5</span>, <span class="at">dpi=</span><span class="dv">150</span>, <span class="at">type=</span><span class="st">"cairo"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-5" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="2" data-code-annotation="1">A value of NA here means the gene in a given FOV was <em>not</em> flagged.</span>
</dd>
</dl>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-qc-fov-2" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qc-fov-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/qc/genes_impacted_in_flagged_fovs.png" class="img-fluid figure-img" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qc-fov-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.2: Number of targets impacted in a given number of FOVs.
</figcaption>
</figure>
</div>
</div>
</div>
<p>In <a href="#fig-qc-fov-2" class="quarto-xref">Figure&nbsp;<span>3.2</span></a>, there are only a few genes (149) that were flagged in three FOVs. If we found a set of genes that were systematically flagged in many FOVs, we may consider removing the impacted FOVs. Alternatively we could remove the impacted genes from all FOVs. In this sample, it’s probably fine to ignore this result.</p>
<p>With the <code>FOVSignalLossSpatialPlot</code> function, we can gain a higher resolution perspective by plotting the 7x7 sub-FOV grids across space and coloring these sub-FOV grids by the change in our expected total counts. In <a href="#fig-qc-fov-3" class="quarto-xref">Figure&nbsp;<span>3.3</span></a>, our focus is on FOVs that are solidly blue which would suggestion fewer transcripts than other FOVs. In this example, eight of the FOVs flagged as low total counts occured along the tissue border and were not complete 7x7 grids. The other section in <a href="#fig-qc-fov-3" class="quarto-xref">Figure&nbsp;<span>3.3</span></a> is contiguous and may suggest a domain or type of tissue with lower overall transcripts.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-qc-fov-3" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qc-fov-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/qc/signal_loss_plot.png" class="img-fluid figure-img" width="1750">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qc-fov-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.3: Number of genes impacted in a given number of FOVs.
</figcaption>
</figure>
</div>
</div>
</div>
<p>To see details how these FOVs might have been impacted, we can use the function below to plot the signal lost found for each cycle-by-reporter combination.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>FOVEffectsHeatmap <span class="ot">&lt;-</span> <span class="cf">function</span>(res, <span class="at">fovs=</span><span class="cn">NULL</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                              <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>, </span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">cluster_columns =</span> <span class="cn">TRUE</span>) {</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">is.null</span>(fovs)){</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    fovs <span class="ot">&lt;-</span> <span class="fu">rownames</span>(res<span class="sc">$</span>fovstats<span class="sc">$</span>bias)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  transposed_matrix <span class="ot">&lt;-</span> <span class="fu">t</span>(res<span class="sc">$</span>fovstats<span class="sc">$</span>bias[fovs,])</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  col_fun <span class="ot">=</span> <span class="fu">colorRamp2</span>(</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">breaks =</span> <span class="fu">c</span>(<span class="fu">min</span>(transposed_matrix), <span class="dv">0</span>, <span class="fu">max</span>(transposed_matrix)), </span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"magenta"</span>, <span class="st">"white"</span>, <span class="st">"cyan"</span>) </span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>  ht_obj <span class="ot">&lt;-</span> ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">matrix =</span> transposed_matrix,</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">name =</span> <span class="st">"FOV bias"</span>, </span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">col =</span> col_fun, </span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">cluster_rows =</span> cluster_rows, </span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">cluster_columns =</span> cluster_columns, </span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>      <span class="at">show_row_dend =</span> <span class="cn">TRUE</span>, </span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">show_column_dend =</span> <span class="cn">TRUE</span>, </span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>      <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">7</span>), </span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>      <span class="at">column_names_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">8</span>)</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(ht_obj)</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">file.path</span>(qc_dir, <span class="st">"FOVEffectsHeatmap.png"</span>), </span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">width=</span><span class="dv">10</span>, <span class="at">height=</span><span class="dv">15</span>, <span class="at">res=</span><span class="dv">350</span>, <span class="at">units =</span> <span class="st">"in"</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">draw</span>(<span class="fu">FOVEffectsHeatmap</span>(res, <span class="at">fovs=</span>res<span class="sc">$</span>flaggedfovs))</span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a>results_list<span class="sc">$</span>flaggedfovs <span class="ot">&lt;-</span> res<span class="sc">$</span>flaggedfovs</span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-qc-fov-4" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qc-fov-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/qc/FOVEffectsHeatmap.png" class="img-fluid figure-img" width="1750">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qc-fov-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.4: Reporter-by-FOV combinations that were flagged.
</figcaption>
</figure>
</div>
</div>
</div>
<p>What we are checking for in <a href="#fig-qc-fov-4" class="quarto-xref">Figure&nbsp;<span>3.4</span></a> is:</p>
<ol type="1">
<li>Any report cycles that show systematically reduced signal across many FOVs</li>
<li>Any FOVs that show underperforming multiple reporter cycles.</li>
</ol>
<p>While we do see two reporter cycles (19 and 27) with two flagged FOVs (155 and 239), each, there’s no evidence of a systemic signal lost in any reporter cycle. If we did find strong evidence, one could conservatively filter out the genes that are present in that reporter cycle. This would be roughly 2,000 genes for a given reporter cycle. Looking from the perspective of the FOVs (#2), we see only a handful of FOVs showed one (or two) underperforming reporter cycles. These include FOVs 92, 113, 155, 160, and 239. At this point, one could remove all FOVs that were flagged (<code>length(res$flaggedfovs)</code>; 23 in this case), just the five that we found evidence of signal lost in one or more reporter cycles, or flag the cells within the impacted FOVs and see if any downstream results are impacted. If you wanted to be on the conservative side, you would remove all flagged FOVs unless you have biological justification that the signal difference isn’t due to FOV effects. For this analysis, I’ll simply do the “flag and monitor” approach.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>fovqcflag <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(meta<span class="sc">$</span>fov <span class="sc">%in%</span> <span class="fu">as.numeric</span>(res<span class="sc">$</span>flaggedfovs), <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>py<span class="sc">$</span>adata<span class="sc">$</span>obs<span class="sc">$</span>fovqcflag <span class="ot">&lt;-</span> meta<span class="sc">$</span>fovqcflag</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
</section>
<section id="cell-level-qc" class="level2 page-columns page-full" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="cell-level-qc"><span class="header-section-number">3.2</span> Cell-level QC</h2>
<p>Where the previous section examined signal at the FOV level, this section examines target counts at the cell level. The purpose of this filter is simple: to remove <em>low signal outliers</em> while keeping the <em>biology</em>. The cells in a given dataset comprise of a mix of:</p>
<ul>
<li>viable, intact cells with high signal</li>
<li>damaged or dying cells. These may have leaky membranes and may have reduced cytoplasmic RNA</li>
<li>imperfect segmentation of cells – especially around FOV borders</li>
</ul>
<p>and so our aim is to filter out cells from the latter two scenarios.</p>
<p>Counts-based filtering has a trade-off. While several downstream “tertiary” analyses account for low signal in a variety of ways (discussed later), many of the pre-processing lenient cutoff can introduce low-quality cells in dimensional reductions like PCA and UMAP and can break up clustering patterns (creating spurious clusters or merging otherwise distinct clusters). On the other hand, a stringent cutoff can introduce unexpected signal lost. This is because cell types vary in their transcript counts so removing cells based on counts will distort the composition of cells in your data.</p>
<p>Best practice for counts based filtering of CosMx data includes:</p>
<ul>
<li>examining the distribution of transcript and gene counts</li>
<li>choosing an initial permissive or lenient cutoff</li>
<li>examining signal relative to background</li>
<li>plotting the spatial distribution of flagged cells in space</li>
<li>removing small cell artifacts around the FOV borders</li>
<li>including plots of total transcripts in your downstream pre-processing steps</li>
</ul>

<div class="no-row-height column-margin column-container"><div class="">
<div class="otherapproachesbox" title="Alternative Approach">
<div class="otherapproachesbox-header">
<div class="otherapproachesbox-icon">

</div>
<p>Alternative Approach</p>
</div>
<div class="otherapproachesbox-body">
<p>There are many cell-level attributes that one can use as a bases for filtering and it is recommended to use metrics relevant to your tissue. For example, if you are analyzing brain, you many want to incorporate or evaluate cell area, eccentricity, or other morphological characteristics that are available in the metadata. For a list of cell metadata column definitions, you can see <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/flat-file-exports/flat-files-compare.html#sec-meta">this Scratch Space blog post</a></p>
</div>
</div>
</div></div><section id="counts-and-features" class="level3" data-number="3.2.1">
<h3 data-number="3.2.1" class="anchored" data-anchor-id="counts-and-features"><span class="header-section-number">3.2.1</span> Counts and Features</h3>
<p>For the distribution of transcript and gene counts, I like to look at the joint distribution in addition to the univariate distributions. In the code below I set a permissive, symmetrical threshold of 2.5 standard deviations (SDs) around the mean (in log10 space) but this thresholding is sample-dependent so different values might make more sense on your data.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># meta &lt;- py$adata$obs</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>counts_min_threshold <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">10</span><span class="sc">^</span>(<span class="fu">mean</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nCount_RNA)) <span class="sc">-</span> <span class="fl">2.5</span><span class="sc">*</span><span class="fu">sd</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nCount_RNA))))</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>features_min_threshold <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">10</span><span class="sc">^</span>(<span class="fu">mean</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nFeature_RNA)) <span class="sc">-</span> <span class="fl">2.5</span><span class="sc">*</span><span class="fu">sd</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nFeature_RNA))))</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>counts_max_threshold <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">10</span><span class="sc">^</span>(<span class="fu">mean</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nCount_RNA)) <span class="sc">+</span> <span class="fl">2.5</span><span class="sc">*</span><span class="fu">sd</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nCount_RNA))))</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>features_max_threshold <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">10</span><span class="sc">^</span>(<span class="fu">mean</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nFeature_RNA)) <span class="sc">+</span> <span class="fl">2.5</span><span class="sc">*</span><span class="fu">sd</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nFeature_RNA))))</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'counts_min_threshold'</span>] <span class="ot">=</span> counts_min_threshold</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'features_min_threshold'</span>] <span class="ot">=</span> features_min_threshold</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'counts_max_threshold'</span>] <span class="ot">=</span> counts_max_threshold</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'features_max_threshold'</span>] <span class="ot">=</span> features_max_threshold</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>counts_filter <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  meta<span class="sc">$</span>nFeature_RNA <span class="sc">&gt;=</span> features_min_threshold <span class="sc">&amp;</span> </span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>         meta<span class="sc">$</span>nFeature_RNA <span class="sc">&lt;=</span> features_max_threshold <span class="sc">&amp;</span> </span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>         meta<span class="sc">$</span>nCount_RNA <span class="sc">&gt;=</span> counts_min_threshold <span class="sc">&amp;</span> </span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>         meta<span class="sc">$</span>nCount_RNA <span class="sc">&lt;=</span> counts_max_threshold</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>  , <span class="st">"pass"</span>, <span class="st">"filtered"</span>)</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'prop_counts_based_filter'</span>] <span class="ot">&lt;-</span> <span class="fu">sum</span>(meta<span class="sc">$</span>counts_filter<span class="sc">==</span><span class="st">"pass"</span>) <span class="sc">/</span> <span class="fu">nrow</span>(meta)</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> meta,</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">x=</span>nFeature_RNA, <span class="at">y=</span>nCount_RNA, <span class="at">color=</span>counts_filter)) <span class="sc">+</span> </span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.002</span>, <span class="at">pch=</span><span class="st">"."</span>) <span class="sc">+</span></span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"pass"</span><span class="ot">=</span><span class="st">"dodgerblue"</span>, <span class="st">"filtered"</span><span class="ot">=</span><span class="st">"darkorange"</span>)) <span class="sc">+</span></span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(features_min_threshold, features_max_threshold), <span class="at">lty=</span><span class="st">"dotted"</span>) <span class="sc">+</span> </span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">c</span>(counts_min_threshold, counts_max_threshold), <span class="at">lty=</span><span class="st">"dotted"</span>) <span class="sc">+</span> </span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>(</span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">breaks =</span> <span class="fu">trans_breaks</span>(<span class="st">"log10"</span>, <span class="cf">function</span>(x) <span class="dv">10</span><span class="sc">^</span>x),</span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> <span class="fu">trans_format</span>(<span class="st">"log10"</span>, <span class="fu">math_format</span>(<span class="dv">10</span><span class="sc">^</span>.x))</span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>(</span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">breaks =</span> <span class="fu">trans_breaks</span>(<span class="st">"log10"</span>, <span class="cf">function</span>(x) <span class="dv">10</span><span class="sc">^</span>x),</span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> <span class="fu">trans_format</span>(<span class="st">"log10"</span>, <span class="fu">math_format</span>(<span class="dv">10</span><span class="sc">^</span>.x))</span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> </span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>() <span class="sc">+</span> </span>
<span id="cb6-42"><a href="#cb6-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb6-43"><a href="#cb6-43" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-44"><a href="#cb6-44" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggMarginal</span>(p, <span class="at">type =</span> <span class="st">"density"</span>)</span>
<span id="cb6-45"><a href="#cb6-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-46"><a href="#cb6-46" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_joint_distribution.png"</span>), p,</span>
<span id="cb6-47"><a href="#cb6-47" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">8</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span>
<span id="cb6-48"><a href="#cb6-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-49"><a href="#cb6-49" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb6-50"><a href="#cb6-50" aria-hidden="true" tabindex="-1"></a>      <span class="at">data=</span>meta, </span>
<span id="cb6-51"><a href="#cb6-51" aria-hidden="true" tabindex="-1"></a>      <span class="fu">aes</span>(<span class="at">x=</span>nCount_RNA)) <span class="sc">+</span> </span>
<span id="cb6-52"><a href="#cb6-52" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_density</span>(<span class="at">alpha=</span><span class="fl">0.01</span>) <span class="sc">+</span> </span>
<span id="cb6-53"><a href="#cb6-53" aria-hidden="true" tabindex="-1"></a>      <span class="co"># scale_x_log10() +</span></span>
<span id="cb6-54"><a href="#cb6-54" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb6-55"><a href="#cb6-55" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(counts_min_threshold, counts_max_threshold), <span class="at">lty=</span><span class="st">"dotted"</span>) </span>
<span id="cb6-56"><a href="#cb6-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-57"><a href="#cb6-57" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_dist_counts.png"</span>), p1,</span>
<span id="cb6-58"><a href="#cb6-58" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">8</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span>
<span id="cb6-59"><a href="#cb6-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-60"><a href="#cb6-60" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb6-61"><a href="#cb6-61" aria-hidden="true" tabindex="-1"></a>      <span class="at">data=</span>meta, </span>
<span id="cb6-62"><a href="#cb6-62" aria-hidden="true" tabindex="-1"></a>      <span class="fu">aes</span>(<span class="at">x=</span>nFeature_RNA)) <span class="sc">+</span> </span>
<span id="cb6-63"><a href="#cb6-63" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_density</span>(<span class="at">alpha=</span><span class="fl">0.01</span>) <span class="sc">+</span> </span>
<span id="cb6-64"><a href="#cb6-64" aria-hidden="true" tabindex="-1"></a>      <span class="co"># scale_x_log10() +</span></span>
<span id="cb6-65"><a href="#cb6-65" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb6-66"><a href="#cb6-66" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(features_min_threshold, features_max_threshold), <span class="at">lty=</span><span class="st">"dotted"</span>) </span>
<span id="cb6-67"><a href="#cb6-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-68"><a href="#cb6-68" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_dist_features.png"</span>), p2,</span>
<span id="cb6-69"><a href="#cb6-69" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">8</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span>
<span id="cb6-70"><a href="#cb6-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-71"><a href="#cb6-71" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb6-72"><a href="#cb6-72" aria-hidden="true" tabindex="-1"></a>        <span class="at">data=</span>meta) <span class="sc">+</span></span>
<span id="cb6-73"><a href="#cb6-73" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_point</span>(</span>
<span id="cb6-74"><a href="#cb6-74" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x=</span>x_slide_mm, <span class="at">y=</span>y_slide_mm, <span class="at">color=</span>counts_filter),</span>
<span id="cb6-75"><a href="#cb6-75" aria-hidden="true" tabindex="-1"></a>        <span class="at">size=</span><span class="fl">0.001</span>, <span class="at">alpha=</span><span class="fl">0.02</span>) <span class="sc">+</span> </span>
<span id="cb6-76"><a href="#cb6-76" aria-hidden="true" tabindex="-1"></a>      <span class="fu">coord_fixed</span>() <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb6-77"><a href="#cb6-77" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"pass"</span><span class="ot">=</span><span class="st">"dodgerblue"</span>, <span class="st">"filtered"</span><span class="ot">=</span><span class="st">"darkorange"</span>)) <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>counts_filter) <span class="sc">+</span> </span>
<span id="cb6-78"><a href="#cb6-78" aria-hidden="true" tabindex="-1"></a>      <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">7</span>, <span class="at">alpha =</span> <span class="dv">1</span>)))</span>
<span id="cb6-79"><a href="#cb6-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-80"><a href="#cb6-80" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_xy_flagged.png"</span>), p3,</span>
<span id="cb6-81"><a href="#cb6-81" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">8</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The tabs below show the summary plots. Based on the distributions, the following thresholds we derived from the data: min counts = 148, max counts = 9973, min features = 120, max features = 6016. This removes about 2% of the data. The spatial layout of these flagged cells shows some concentration along the perimeter and epithelial regions. No concentration within FOVs were observed (<a href="#fig-qc-xy-flagged" class="quarto-xref">Figure&nbsp;<span>3.8</span></a>).</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" href="">Transcripts per cell</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false" href="">Features per cell</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false" href="">Joint Distribution</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-4" role="tab" aria-controls="tabset-1-4" aria-selected="false" href="">Spatial Arrangement of flagged cells</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>), <span class="st">"qc_dist_counts.png"</span>, qc_dir)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div id="fig-qc-dist-tx-per-cell" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qc-dist-tx-per-cell-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/qc/qc_dist_counts.png" class="img-fluid figure-img" width="1200">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qc-dist-tx-per-cell-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.5: Distribution of transcripts per cell.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="cell">
<div class="cell-output-display">
<div id="fig-qc-dist-featrues-per-cell" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qc-dist-featrues-per-cell-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/qc/qc_dist_features.png" class="img-fluid figure-img" width="1200">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qc-dist-featrues-per-cell-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.6: Distribution of features per cell.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<div class="cell">
<div class="cell-output-display">
<div id="fig-qc-joint" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qc-joint-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/qc/qc_joint_distribution.png" class="img-fluid figure-img" width="1200">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qc-joint-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.7: Joint distribution of targets (features vs counts).
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-4-tab">
<div class="cell">
<div class="cell-output-display">
<div id="fig-qc-xy-flagged" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qc-xy-flagged-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/qc/qc_xy_flagged.png" class="img-fluid figure-img" width="1200">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qc-xy-flagged-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.8: Spatial arrangement of flagged cells.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="optional-cells-near-fov-borders" class="level3" data-number="3.2.2">
<h3 data-number="3.2.2" class="anchored" data-anchor-id="optional-cells-near-fov-borders"><span class="header-section-number">3.2.2</span> (Optional) Cells Near FOV Borders</h3>
<p>In this particular dataset, 6.2% of the cells were located on the border of FOVs (<em>i.e.</em>, touching the FOV edge). This is a minority of cells that in the analysis as a whole shouldn’t affect downstream analyses. For this reason this step is optional.</p>
<p>Within the metadata there is a column named <code>SplitRatioToLocal</code>. For cells that are adjacent to the FOV boundaries, the <code>SplitRatioToLocal</code> metric measures the cell area relative to the mean area of cells in the FOVs. For <code>0 &lt; SplitRatioToLocal &lt; 1</code>, the cell is smaller than average and for <code>SplitRatioToLocal &gt; 1</code> the cell is larger than average. A value of 0 means the cell is not along the border. Let’s plot the distribution of boarder abutting cells (<em>i.e.</em>, <code>SplitRatioToLocal &gt; 0</code>) and tentatively choose a cutoff of SplitRatioToLocal = 0.5 which translates to removing cells that are 0.5 (50%) the average cell size. Since this is a ratio, we’ll plot in log2 space (<code>log2(SplitRatioToLocal) &lt; -1</code>).</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'SplitRatioToLocalThreshold'</span>] <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># meta &lt;- py$adata$obs</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>p_detail <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> meta <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SplitRatioToLocal <span class="sc">!=</span> <span class="dv">0</span>), </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log2</span>(SplitRatioToLocal))) <span class="sc">+</span> </span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">100</span>, <span class="at">fill =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="fu">expression</span>(log[<span class="dv">2</span>](<span class="st">"SplitRatioToLocal"</span>)),</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Number of FOV border cells"</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">log2</span>(results_list[[<span class="st">'SplitRatioToLocalThreshold'</span>]]), </span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>               <span class="at">color =</span> <span class="st">"darkorange"</span>, <span class="at">lty =</span> <span class="st">"dashed"</span>)</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>pie_data <span class="ot">&lt;-</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Category =</span> <span class="fu">if_else</span>(SplitRatioToLocal <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="st">"Border cell"</span>, <span class="st">"Non-border cell"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(Category) <span class="sc">%&gt;%</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">Percentage =</span> (n <span class="sc">/</span> <span class="fu">sum</span>(n)) <span class="sc">*</span> <span class="dv">100</span>,</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">Label =</span> <span class="fu">paste0</span>(Category, <span class="st">" ("</span>, <span class="fu">round</span>(Percentage, <span class="dv">1</span>), <span class="st">"%)"</span>)</span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>p_pie <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pie_data, <span class="fu">aes</span>(<span class="at">x =</span> <span class="dv">2</span>, <span class="at">y =</span> n, <span class="at">fill =</span> Category)) <span class="sc">+</span></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">width =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_polar</span>(<span class="st">"y"</span>, <span class="at">start =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Border cell"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>, <span class="st">"Non-border cell"</span> <span class="ot">=</span> <span class="st">"gray90"</span>)) <span class="sc">+</span></span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> Label), </span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>            <span class="at">position =</span> <span class="fu">position_stack</span>(<span class="at">vjust =</span> <span class="fl">0.5</span>), </span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>            <span class="at">size =</span> <span class="dv">3</span>, </span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>            <span class="at">fontface =</span> <span class="st">"bold"</span>) <span class="sc">+</span></span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fl">0.5</span>, <span class="fl">2.5</span>) <span class="sc">+</span> </span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>full_plot <span class="ot">&lt;-</span> p_detail <span class="sc">+</span> </span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">inset_element</span>(</span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a>    p_pie, </span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a>    <span class="at">left =</span> <span class="fl">0.6</span>, <span class="at">bottom =</span> <span class="fl">0.6</span>, </span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a>    <span class="at">right =</span> <span class="fl">1.0</span>,</span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a>    <span class="at">top =</span> <span class="fl">1.0</span></span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_splitratiotolocal.png"</span>), full_plot,</span>
<span id="cb8-49"><a href="#cb8-49" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">5</span>, <span class="at">height=</span><span class="dv">5</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-qc-split" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qc-split-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/qc/qc_splitratiotolocal.png" class="img-fluid figure-img" width="750">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qc-split-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.9: Distribution of SplitRatioToLocal for cells abutting an FOV border. Overlay: proportion of border cells relative to all cells.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="regional-level-cell-qc" class="level2 page-columns page-full" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="regional-level-cell-qc"><span class="header-section-number">3.3</span> Regional-level cell QC</h2>
<p>It’s possible that some regions of the tissue might have high background. We can evaluate the total signal relative to the background. For this approach we are interested in the Signal to Background Ratio (SBR) in neighborhoods or tissue regions as opposed to cell-level measures or target-level measures. Since the number of targets is not comparable to the number of negative probes, we’ll consider the average target expression (S) divided by the average background expression.</p>
<p><span class="math display">\[sbr = \frac{S}{\mu_{B}}\]</span></p>
<p>When passing the data through this equation, you’ll likely find that many cells have an undefined SBR due to the sparse nature of background counts. To mitigate this effect, we can spatially smooth the transcript and background counts before applying this equation. To do this we’ll use the <code>liana</code> package that creates a connectivities matrix based on a Gaussian kerneland then multiple this matrix by the negative probe matrix and the expression matrix to get the spatially-weighted matrices for each. For more information on this spatial smoothing see <a href="pathways.html" class="quarto-xref"><span>Chapter 7</span></a>.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Python Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-10"><pre class="sourceCode python code-annotation-code code-with-copy code-annotated"><code class="sourceCode python"><span id="annotated-cell-10-1"><a href="#annotated-cell-10-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> liana <span class="im">as</span> li</span>
<span id="annotated-cell-10-2"><a href="#annotated-cell-10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-10-3"><a href="#annotated-cell-10-3" aria-hidden="true" tabindex="-1"></a>li.ut.spatial_neighbors(</span>
<span id="annotated-cell-10-4"><a href="#annotated-cell-10-4" aria-hidden="true" tabindex="-1"></a>    adata,</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="1">1</button><span id="annotated-cell-10-5" class="code-annotation-target"><a href="#annotated-cell-10-5" aria-hidden="true" tabindex="-1"></a>    bandwidth<span class="op">=</span><span class="fl">0.010</span>,</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="2">2</button><span id="annotated-cell-10-6" class="code-annotation-target"><a href="#annotated-cell-10-6" aria-hidden="true" tabindex="-1"></a>    cutoff<span class="op">=</span><span class="fl">0.080</span>,</span>
<span id="annotated-cell-10-7"><a href="#annotated-cell-10-7" aria-hidden="true" tabindex="-1"></a>    kernel<span class="op">=</span><span class="st">'gaussian'</span>,</span>
<span id="annotated-cell-10-8"><a href="#annotated-cell-10-8" aria-hidden="true" tabindex="-1"></a>    set_diag<span class="op">=</span><span class="va">True</span>,</span>
<span id="annotated-cell-10-9"><a href="#annotated-cell-10-9" aria-hidden="true" tabindex="-1"></a>    key_added <span class="op">=</span> <span class="st">'liana'</span>,</span>
<span id="annotated-cell-10-10"><a href="#annotated-cell-10-10" aria-hidden="true" tabindex="-1"></a>    standardize<span class="op">=</span><span class="va">True</span></span>
<span id="annotated-cell-10-11"><a href="#annotated-cell-10-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="annotated-cell-10-12"><a href="#annotated-cell-10-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-10-13"><a href="#annotated-cell-10-13" aria-hidden="true" tabindex="-1"></a>connectivities <span class="op">=</span> adata.obsp[<span class="st">'liana_connectivities'</span>]</span>
<span id="annotated-cell-10-14"><a href="#annotated-cell-10-14" aria-hidden="true" tabindex="-1"></a>adata.obsm[<span class="st">'counts_neg_spatially_smooth'</span>] <span class="op">=</span> connectivities <span class="op">@</span> adata.obsm[<span class="st">'counts_neg'</span>]</span>
<span id="annotated-cell-10-15"><a href="#annotated-cell-10-15" aria-hidden="true" tabindex="-1"></a>adata.obsm[<span class="st">'counts_spatially_smooth'</span>] <span class="op">=</span> connectivities <span class="op">@</span> adata.X</span>
<span id="annotated-cell-10-16"><a href="#annotated-cell-10-16" aria-hidden="true" tabindex="-1"></a>smoothed_mean_neg <span class="op">=</span> np.ravel(adata.obsm[<span class="st">'counts_neg_spatially_smooth'</span>].mean(axis<span class="op">=</span><span class="dv">1</span>))</span>
<span id="annotated-cell-10-17"><a href="#annotated-cell-10-17" aria-hidden="true" tabindex="-1"></a>smoothed_mean_target <span class="op">=</span> np.ravel(adata.obsm[<span class="st">'counts_spatially_smooth'</span>].mean(axis<span class="op">=</span><span class="dv">1</span>))</span>
<span id="annotated-cell-10-18"><a href="#annotated-cell-10-18" aria-hidden="true" tabindex="-1"></a>epsilon <span class="op">=</span> <span class="fl">1e-9</span></span>
<span id="annotated-cell-10-19"><a href="#annotated-cell-10-19" aria-hidden="true" tabindex="-1"></a>SBR <span class="op">=</span> smoothed_mean_target <span class="op">/</span> (smoothed_mean_neg <span class="op">+</span> epsilon)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-10" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="5" data-code-annotation="1">use 10 µm bandwidth (this tunable)</span>
</dd>
<dt data-target-cell="annotated-cell-10" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="6" data-code-annotation="2">anything below this value is considered 0 (this is tunable)</span>
</dd>
</dl>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<div class="otherapproachesbox" title="Alternative Approach">
<div class="otherapproachesbox-header">
<div class="otherapproachesbox-icon">

</div>
<p>Alternative Approach</p>
</div>
<div class="otherapproachesbox-body">
<p>While not entirely equivalent, if you run into memory issues with creating a spatially-weighted expression matrix, you can instead spatially smooth the total counts from the metadata column <code>nCount_RNA</code>.</p>
</div>
</div>
</div></div><p>What we are looking for in SBR are larger tissue regions with low SBR – not necessarily individual cells with low SBR. Looking at <a href="#fig-qc-sbr" class="quarto-xref">Figure&nbsp;<span>3.10</span></a> we see a low SBR region on the top of the tissue just below the epithelium. This region corresponds to an area of smooth muscle and we’ll see in the next section that this region partially overlaps with an area flagged by the FOV-level QC. Since these low SBR cells (i.e., <code>log2(SBR) &lt; 0</code>) are concentrated spatially, we’ll choose to flag them for removal.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>SBR <span class="ot">&lt;-</span> py<span class="sc">$</span>SBR</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>sbr_breaks <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="sc">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="cn">Inf</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>sbr_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"&lt;= -1"</span>, <span class="st">"-1 to 0"</span>, <span class="st">"0 to 1"</span>, <span class="st">"1 to 2"</span>, </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"2 to 3"</span>, <span class="st">"&gt; 3"</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>sbr_labels <span class="ot">&lt;-</span> <span class="fu">paste0</span>(sbr_labels,</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>                     <span class="fu">paste0</span>(<span class="st">" ("</span>, <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> <span class="fu">table</span>(<span class="fu">cut</span>(<span class="fu">log2</span>(meta<span class="sc">$</span>SBR), </span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>                                <span class="at">breaks =</span> sbr_breaks)) <span class="sc">/</span> <span class="fu">nrow</span>(meta), <span class="dv">1</span>), <span class="st">"%)"</span>))</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>manual_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"&lt;= -1"</span>   <span class="ot">=</span> <span class="st">"#D73027"</span>,  <span class="co"># </span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>  <span class="st">"-1 to 0"</span> <span class="ot">=</span> <span class="st">"#F46D43"</span>,  <span class="co"># </span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>  <span class="st">"0 to 1"</span>  <span class="ot">=</span> <span class="st">"#D9EF8B"</span>,  <span class="co"># </span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  <span class="st">"1 to 2"</span>  <span class="ot">=</span> <span class="st">"#A6D96A"</span>,  <span class="co"># </span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  <span class="st">"2 to 3"</span>  <span class="ot">=</span> <span class="st">"#1A9850"</span>,  <span class="co"># </span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>  <span class="st">"&gt; 3"</span>     <span class="ot">=</span> <span class="st">"#006837"</span>   <span class="co"># </span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(manual_colors) <span class="ot">&lt;-</span> sbr_labels</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> meta) <span class="sc">+</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x_slide_mm,</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> y_slide_mm,</span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>      <span class="at">color =</span> <span class="fu">cut</span>(<span class="fu">log2</span>(SBR), <span class="at">breaks =</span> sbr_breaks, <span class="at">labels =</span> sbr_labels)</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.001</span>, <span class="at">alpha =</span> <span class="dv">1</span></span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> manual_colors) <span class="sc">+</span></span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">7</span>, <span class="at">alpha =</span> <span class="dv">1</span>) <span class="co"># Sets legend points to size 5 and opaque</span></span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">+</span></span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">"log2(SBR)"</span>) <span class="sc">+</span></span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>() <span class="sc">+</span></span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_xy_sbr.png"</span>), p,</span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">10</span>, <span class="at">height=</span><span class="dv">10</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-qc-sbr" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qc-sbr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/qc/qc_xy_sbr.png" class="img-fluid figure-img" width="1500">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qc-sbr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.10: Spatially-smoothed Signal to Background for each cell.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="combining-flags" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="combining-flags"><span class="header-section-number">3.4</span> Combining Flags</h2>
<p>Taken together, what cells were flagged and why? <a href="#fig-qc-upset" class="quarto-xref">Figure&nbsp;<span>3.11</span></a> shows the configuration of the flags.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(UpSetR)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>filter_list <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Low SplitRatio</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(SplitRatioToLocal <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> SplitRatioToLocal <span class="sc">&lt;</span> results_list[[<span class="st">'SplitRatioToLocalThreshold'</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID),</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Low Counts</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(nCount_RNA <span class="sc">&lt;</span> results_list[[<span class="st">'counts_min_threshold'</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID),</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">High Counts</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(nCount_RNA <span class="sc">&gt;</span> results_list[[<span class="st">'counts_max_threshold'</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID),</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Low Features</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(nFeature_RNA <span class="sc">&lt;</span> results_list[[<span class="st">'features_min_threshold'</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID),</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">High Features</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(nFeature_RNA <span class="sc">&gt;</span> results_list[[<span class="st">'features_max_threshold'</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID), </span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Low SBR</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span> </span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="fu">log2</span>(SBR) <span class="sc">&lt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID),</span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">FOV QC</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span> </span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(fovqcflag <span class="sc">==</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID)</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>filter_list <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">keep</span>(filter_list, <span class="sc">~</span> <span class="fu">length</span>(.) <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">file.path</span>(qc_dir, <span class="st">"flagged_cells.png"</span>), </span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">width=</span><span class="dv">10</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">res=</span><span class="dv">350</span>, <span class="at">units =</span> <span class="st">"in"</span>, <span class="at">type =</span> <span class="st">"cairo"</span>)</span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">upset</span>(</span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fromList</span>(filter_list),</span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a>  <span class="at">nintersects =</span> <span class="dv">10</span>,</span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a>  <span class="at">order.by =</span> <span class="st">"freq"</span>,     </span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a>  <span class="at">nsets =</span> <span class="fu">length</span>(filter_list), </span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a>  <span class="at">text.scale =</span> <span class="fl">1.5</span>)</span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-qc-upset" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qc-upset-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./assets/analysis_results/qc/flagged_cells.png" class="img-fluid figure-img" width="1750">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qc-upset-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.11: Combinations of quality filters flagged in this study. Only the 10 most frequent sets are shown for clarity.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="filtering" class="level2" data-number="3.5">
<h2 data-number="3.5" class="anchored" data-anchor-id="filtering"><span class="header-section-number">3.5</span> Filtering</h2>
<p>Now that we have flagged the cells, this section does the actual filtering and saves data to disk so that we can use it in the pre-processing step.</p>
<p>In regular analyses, I simply filter the cells based on the chosen cutoffs and continue with the pre-processing step. In this case, I am interested in showing the downstream effects of leaving poor quality cells in versus filtering them. To that end, I’ll save two annData objects: unfiltered and filtered. In both datasets, I will add a qc flag column to keep track of what QC conditions were flagged in any given cell.</p>
<p>To be efficient I’ll make use of bitwise shifts to create the QC flag column.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>qc_masks <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">LOW_COUNTS =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">0</span>),  <span class="co"># 1</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">HIGH_COUNTS =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">1</span>), <span class="co"># 2</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">LOW_FEAT =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">2</span>),    <span class="co"># 4</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">HIGH_FEAT =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">3</span>),   <span class="co"># 8</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">LOW_SPLIT =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">4</span>),   <span class="co"># 16</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">FOV_QC =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">5</span>),      <span class="co"># 32</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">LOW_SBR =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">6</span>),     <span class="co"># 64 </span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">BLANK7 =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">7</span>)       <span class="co"># 128 (not used)</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">qc_flag =</span> (</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 1</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>      (nCount_RNA <span class="sc">&lt;</span> results_list[[<span class="st">'counts_min_threshold'</span>]]) <span class="sc">*</span> qc_masks<span class="sc">$</span>LOW_COUNTS <span class="sc">+</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 2</span></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>      (nCount_RNA <span class="sc">&gt;</span> results_list[[<span class="st">'counts_max_threshold'</span>]]) <span class="sc">*</span> qc_masks<span class="sc">$</span>HIGH_COUNTS <span class="sc">+</span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 3</span></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>      (nFeature_RNA <span class="sc">&lt;</span> results_list[[<span class="st">'features_min_threshold'</span>]]) <span class="sc">*</span> qc_masks<span class="sc">$</span>LOW_FEAT <span class="sc">+</span></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 4</span></span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a>      (nFeature_RNA <span class="sc">&gt;</span> results_list[[<span class="st">'features_max_threshold'</span>]]) <span class="sc">*</span> qc_masks<span class="sc">$</span>HIGH_FEAT <span class="sc">+</span></span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 5</span></span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a>      (SplitRatioToLocal <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> SplitRatioToLocal <span class="sc">&lt;</span> results_list[[<span class="st">'SplitRatioToLocalThreshold'</span>]]) <span class="sc">*</span> qc_masks<span class="sc">$</span>LOW_SPLIT <span class="sc">+</span> </span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 6</span></span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true" tabindex="-1"></a>      (fovqcflag <span class="sc">==</span> <span class="dv">1</span>) <span class="sc">*</span> qc_masks<span class="sc">$</span>FOV_QC <span class="sc">+</span></span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 7</span></span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true" tabindex="-1"></a>      (<span class="fu">log2</span>(SBR) <span class="sc">&lt;</span> <span class="dv">0</span>) <span class="sc">*</span> qc_masks<span class="sc">$</span>LOW_SBR</span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-39"><a href="#cb11-39" aria-hidden="true" tabindex="-1"></a>py<span class="sc">$</span>adata<span class="sc">$</span>obs<span class="sc">$</span>qcflag <span class="ot">&lt;-</span> meta<span class="sc">$</span>qc_flag</span>
<span id="cb11-40"><a href="#cb11-40" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'n_cells_before_qc'</span>] <span class="ot">=</span> py<span class="sc">$</span>adata<span class="sc">$</span>n_obs</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>Remove some of the intermediate matrices and save the annData to disk.</p>
<div class="cell">
<details class="code-fold">
<summary>Python Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'qcflag'</span>] <span class="op">=</span> adata.obs[<span class="st">'qcflag'</span>].astype(<span class="st">'uint8'</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> adata.obsp[<span class="st">'liana_connectivities'</span>]</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> adata.obsm[<span class="st">'counts_neg_spatially_smooth'</span>]</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> adata.obsm[<span class="st">'counts_spatially_smooth'</span>]</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-1-qc-flagged.h5ad"</span>)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  filename,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  compression<span class="op">=</span>hdf5plugin.FILTERS[<span class="st">"zstd"</span>],</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  compression_opts<span class="op">=</span>hdf5plugin.Zstd(clevel<span class="op">=</span><span class="dv">5</span>).filter_options</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>In addition to saving a copy of flagged data, let’s filter out any cells with a QC flag that is not “32” and save that “filtered” dataset to disk.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Python Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'qcflag'</span>] <span class="op">=</span> adata.obs[<span class="st">'qcflag'</span>].astype(<span class="st">'uint8'</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[adata.obs[<span class="st">'qcflag'</span>].isin([<span class="dv">0</span>, <span class="dv">32</span>]), :].copy()</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-1-qc-filtered.h5ad"</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  filename,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  compression<span class="op">=</span>hdf5plugin.FILTERS[<span class="st">"zstd"</span>],</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  compression_opts<span class="op">=</span>hdf5plugin.Zstd(clevel<span class="op">=</span><span class="dv">5</span>).filter_options</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>Save project-level results.</p>
<div class="cell">
<details class="code-fold">
<summary>R Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'n_cells_after_qc'</span>] <span class="ot">=</span> py<span class="sc">$</span>adata<span class="sc">$</span>n_obs</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<p>After filtering, we have 466820 (94.5 %) that are ready to analyze. Note that we did not do any gene-level filtering. My primary reason for leaving all genes in the data at this point is primarily because – while we do not expect every gene to be highly expressed in every cell – there are rare cells that may express genes that would considered “below background” in the rest of the data. My general approach is to keep all targets in the data and identify which of these are highly variable and use those genes when appropriate (<em>e.g.</em>, PCA, UMAP, Leiden).</p>


<!-- -->

<script type="webr-data">
eyJvcHRpb25zIjp7ImJhc2VVcmwiOiJodHRwczovL3dlYnIuci13YXNtLm9yZy92MC41LjQvIn0sInBhY2thZ2VzIjp7InBrZ3MiOlsiZXZhbHVhdGUiLCJrbml0ciIsImh0bWx0b29scyIsImRwbHlyIiwiZHVja2RiIiwiZ2dwbG90MiIsInBsb3RseSJdLCJyZXBvcyI6WyJodHRwczovL3ItbGliLnItdW5pdmVyc2UuZGV2Il19LCJyZW5kZXJfZGYiOiJkZWZhdWx0In0=
</script>
<script type="ojs-module-contents">
{"contents":[{"cellName":"webr-widget-3","inline":false,"methodName":"interpretQuiet","source":"{\n  // Wait for output to be written to the DOM, then trigger widget rendering\n  await _webr_value_3;\n  if (window.HTMLWidgets) {\n    window.HTMLWidgets.staticRender();\n  }\n  if (window.PagedTableDoc) {\n    window.PagedTableDoc.initAll();\n  }\n}\n"},{"cellName":"webr-3","inline":false,"methodName":"interpret","source":"viewof _webr_editor_3 = {\n  const { WebRExerciseEditor, b64Decode } = window._exercise_ojs_runtime;\n  const scriptContent = document.querySelector(`script[type=\\\"webr-3-contents\\\"]`).textContent;\n  const block = JSON.parse(b64Decode(scriptContent));\n\n  const options = Object.assign({ id: `webr-3-contents` }, block.attr);\n  const editor = new WebRExerciseEditor(webROjs.webRPromise, block.code, options);\n\n  return editor.container;\n}\n_webr_value_3 = webROjs.process(_webr_editor_3, {filtered_data});\n"},{"cellName":"webr-widget-2","inline":false,"methodName":"interpretQuiet","source":"{\n  // Wait for output to be written to the DOM, then trigger widget rendering\n  await _webr_value_2;\n  if (window.HTMLWidgets) {\n    window.HTMLWidgets.staticRender();\n  }\n  if (window.PagedTableDoc) {\n    window.PagedTableDoc.initAll();\n  }\n}\n"},{"cellName":"webr-2","inline":false,"methodName":"interpret","source":"_webr_value_2 = {\n  const { highlightR, b64Decode } = window._exercise_ojs_runtime;\n  const scriptContent = document.querySelector(`script[type=\\\"webr-2-contents\\\"]`).textContent;\n  const block = JSON.parse(b64Decode(scriptContent));\n\n  // Default evaluation configuration\n  const options = Object.assign({\n    id: \"webr-2-contents\",\n    echo: true,\n    output: true\n  }, block.attr);\n\n  // Evaluate the provided R code\n  const result = webROjs.process({code: block.code, options}, {max_prop_loss, max_totalcounts_loss});\n\n  // Early yield while we wait for the first evaluation and render\n  if (options.output && !(\"2\" in webROjs.renderedOjs)) {\n    const container = document.createElement(\"div\");\n    const spinner = document.createElement(\"div\");\n\n    if (options.echo) {\n      // Show output as highlighted source\n      const preElem = document.createElement(\"pre\");\n      container.className = \"sourceCode\";\n      preElem.className = \"sourceCode r\";\n      preElem.appendChild(highlightR(block.code));\n      spinner.className = \"spinner-grow spinner-grow-sm m-2 position-absolute top-0 end-0\";\n      preElem.appendChild(spinner);\n      container.appendChild(preElem);\n    } else {\n      spinner.className = \"spinner-border spinner-border-sm\";\n      container.appendChild(spinner);\n    }\n\n    yield container;\n  }\n\n  webROjs.renderedOjs[\"2\"] = true;\n  yield await result;\n}\n"},{"cellName":"webr-widget-1","inline":false,"methodName":"interpretQuiet","source":"{\n  // Wait for output to be written to the DOM, then trigger widget rendering\n  await _webr_value_1;\n  if (window.HTMLWidgets) {\n    window.HTMLWidgets.staticRender();\n  }\n  if (window.PagedTableDoc) {\n    window.PagedTableDoc.initAll();\n  }\n}\n"},{"cellName":"webr-1","inline":false,"methodName":"interpret","source":"_webr_value_1 = {\n  const { highlightR, b64Decode } = window._exercise_ojs_runtime;\n  const scriptContent = document.querySelector(`script[type=\\\"webr-1-contents\\\"]`).textContent;\n  const block = JSON.parse(b64Decode(scriptContent));\n\n  // Default evaluation configuration\n  const options = Object.assign({\n    id: \"webr-1-contents\",\n    echo: true,\n    output: true\n  }, block.attr);\n\n  // Evaluate the provided R code\n  const result = webROjs.process({code: block.code, options}, {});\n\n  // Early yield while we wait for the first evaluation and render\n  if (options.output && !(\"1\" in webROjs.renderedOjs)) {\n    const container = document.createElement(\"div\");\n    const spinner = document.createElement(\"div\");\n\n    if (options.echo) {\n      // Show output as highlighted source\n      const preElem = document.createElement(\"pre\");\n      container.className = \"sourceCode\";\n      preElem.className = \"sourceCode r\";\n      preElem.appendChild(highlightR(block.code));\n      spinner.className = \"spinner-grow spinner-grow-sm m-2 position-absolute top-0 end-0\";\n      preElem.appendChild(spinner);\n      container.appendChild(preElem);\n    } else {\n      spinner.className = \"spinner-border spinner-border-sm\";\n      container.appendChild(spinner);\n    }\n\n    yield container;\n  }\n\n  webROjs.renderedOjs[\"1\"] = true;\n  yield await result;\n}\n"},{"cellName":"webr-prelude","inline":false,"methodName":"interpretQuiet","source":"webROjs = {\n  const { WebR, ChannelType } = window._exercise_ojs_runtime.WebR;\n  const {\n    WebREvaluator,\n    WebREnvironmentManager,\n    setupR,\n    b64Decode,\n    collapsePath\n  } = window._exercise_ojs_runtime;\n\n  const statusContainer = document.getElementById(\"exercise-loading-status\");\n  const indicatorContainer = document.getElementById(\"exercise-loading-indicator\");\n  indicatorContainer.classList.remove(\"d-none\");\n\n  let statusText = document.createElement(\"div\")\n  statusText.classList = \"exercise-loading-details\";\n  statusText = statusContainer.appendChild(statusText);\n  statusText.textContent = `Initialise`;\n\n  // Hoist indicator out from final slide when running under reveal\n  const revealStatus = document.querySelector(\".reveal .exercise-loading-indicator\");\n  if (revealStatus) {\n    revealStatus.remove();\n    document.querySelector(\".reveal > .slides\").appendChild(revealStatus);\n  }\n\n  // Make any reveal slides with live cells scrollable\n  document.querySelectorAll(\".reveal .exercise-cell\").forEach((el) => {\n    el.closest('section.slide').classList.add(\"scrollable\");\n  })\n\n  // webR supplemental data and options\n  const dataContent = document.querySelector(`script[type=\\\"webr-data\\\"]`).textContent;\n  const data = JSON.parse(b64Decode(dataContent));\n\n  // Grab list of resources to be downloaded\n  const filesContent = document.querySelector(`script[type=\\\"vfs-file\\\"]`).textContent;\n  const files = JSON.parse(b64Decode(filesContent));\n\n  // Use PostMessage channel for async communication\n  // We do not expect to take nested prompt input in Quarto Live cells\n  data.options.channelType = ChannelType.PostMessage;\n\n  // Initialise webR and setup for R code evaluation\n  let webRPromise = (async (webR) => {\n    statusText.textContent = `Downloading webR`;\n    await webR.init();\n\n    // Install provided list of packages\n    // Ensure webR default repo is included\n    data.packages.repos.push(\"https://repo.r-wasm.org\")\n    await data.packages.pkgs.map((pkg) => () => {\n      statusText.textContent = `Downloading package: ${pkg}`;\n      return webR.evalRVoid(`\n        webr::install(pkg, repos = repos)\n        library(pkg, character.only = TRUE)\n      `, { env: {\n        pkg: pkg,\n        repos: data.packages.repos,\n      }});\n    }).reduce((cur, next) => cur.then(next), Promise.resolve());\n\n    // Download and install resources\n    await files.map((file) => async () => {\n      const name = file.substring(file.lastIndexOf('/') + 1);\n      statusText.textContent = `Downloading resource: ${name}`;\n      const response = await fetch(file);\n      if (!response.ok) {\n        throw new Error(`Can't download \\`${file}\\`. Error ${response.status}: \"${response.statusText}\".`);\n      }\n      const data = await response.arrayBuffer();\n\n      // Store URLs in the cwd without any subdirectory structure\n      if (file.includes(\"://\")) {\n        file = name;\n      }\n\n      // Collapse higher directory structure\n      file = collapsePath(file);\n\n      // Create directory tree, ignoring \"directory exists\" VFS errors\n      const parts = file.split('/').slice(0, -1);\n      let path = '';\n      while (parts.length > 0) {\n        path += parts.shift() + '/';\n        const analysis = await webR.FS.analyzePath(path);\n        if (!analysis.exists) {\n          try {\n            await webR.FS.mkdir(path);\n          } catch (e) {\n            throw new Error(`Filesystem Error: \"${e.message}\".`);\n          }\n        }\n      }\n\n      // Write this file to the VFS\n      return await webR.FS.writeFile(file, new Uint8Array(data));\n    }).reduce((cur, next) => cur.then(next), Promise.resolve());\n\n    statusText.textContent = `Installing webR shims`;\n    await webR.evalRVoid(`webr::shim_install()`);\n\n    statusText.textContent = `WebR environment setup`;\n    await setupR(webR, data);\n\n    statusText.remove();\n    if (statusContainer.children.length == 0) {\n      statusContainer.parentNode.remove();\n    }\n    return webR;\n  })(new WebR(data.options));\n\n  // Keep track of initial OJS block render\n  const renderedOjs = {};\n\n  const process = async (context, inputs) => {\n    const webR = await webRPromise;\n    const evaluator = new WebREvaluator(webR, context)\n    await evaluator.process(inputs);\n    return evaluator.container;\n  }\n\n  return {\n    process,\n    webRPromise,\n    renderedOjs,\n  };\n}\n"}]}
</script>
<div id="exercise-loading-indicator" class="exercise-loading-indicator d-none d-flex align-items-center gap-2">
<div id="exercise-loading-status" class="d-flex gap-2">

</div>
<div class="spinner-grow spinner-grow-sm">

</div>
</div>
<script type="vfs-file">
WyJhc3NldHMvaW50ZXJhY3RpdmVzL2Zvdl9xY192YXJ5LnBhcnF1ZXQiLCJhc3NldHMvaW50ZXJhY3RpdmVzL2Zvdl9sb2NhdGlvbnMucGFycXVldCJd
</script>
</section>

</main> <!-- /main -->
<script type="ojs-module-contents">
eyJjb250ZW50cyI6W3sibWV0aG9kTmFtZSI6ImludGVycHJldCIsImNlbGxOYW1lIjoib2pzLWNlbGwtMSIsImlubGluZSI6ZmFsc2UsInNvdXJjZSI6InZpZXdvZiBtYXhfcHJvcF9sb3NzID0gSW5wdXRzLnJhbmdlKFswLjEsIDAuNl0sIHtcbiAgc3RlcDogMC4xLFxuICBsYWJlbDogXCJNYXggUHJvcG9ydGlvbiBMb3NzXCJcbn0pXG52aWV3b2YgbWF4X3RvdGFsY291bnRzX2xvc3MgPSBJbnB1dHMucmFuZ2UoWzAuMSwgMC42XSwge1xuICBzdGVwOiAwLjEsXG4gIGxhYmVsOiBcIk1heCBUb3RhbCBDb3VudHMgTG9zc1wiXG59KVxuIn0seyJtZXRob2ROYW1lIjoiaW50ZXJwcmV0UXVpZXQiLCJzb3VyY2UiOiJzaGlueUlucHV0KCdtYXhfcHJvcF9sb3NzJykifSx7Im1ldGhvZE5hbWUiOiJpbnRlcnByZXRRdWlldCIsInNvdXJjZSI6InNoaW55SW5wdXQoJ21heF90b3RhbGNvdW50c19sb3NzJykifV19
</script>
<script type="module">
if (window.location.protocol === "file:") { alert("The OJS runtime does not work with file:// URLs. Please use a web server to view this document."); }
window._ojs.paths.runtimeToDoc = "../..";
window._ojs.paths.runtimeToRoot = "../..";
window._ojs.paths.docToRoot = "";
window._ojs.selfContained = false;
window._ojs.runtime.interpretFromScriptTags();
</script>
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    // Ensure there is a toggle, if there isn't float one in the top right
    if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
      const a = window.document.createElement('a');
      a.classList.add('top-right');
      a.classList.add('quarto-color-scheme-toggle');
      a.href = "";
      a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
      const i = window.document.createElement("i");
      i.classList.add('bi');
      a.appendChild(i);
      window.document.body.appendChild(a);
    }
    setColorSchemeToggle(hasAlternateSentinel())
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/nanostring-biostats\.github\.io\/CosMx-Analysis-Scratch-Space\/the-wtx-guide");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        const annoteTargets = window.document.querySelectorAll('.code-annotation-anchor');
        for (let i=0; i<annoteTargets.length; i++) {
          const annoteTarget = annoteTargets[i];
          const targetCell = annoteTarget.getAttribute("data-target-cell");
          const targetAnnotation = annoteTarget.getAttribute("data-target-annotation");
          const contentFn = () => {
            const content = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
            if (content) {
              const tipContent = content.cloneNode(true);
              tipContent.classList.add("code-annotation-tip-content");
              return tipContent.outerHTML;
            }
          }
          const config = {
            allowHTML: true,
            content: contentFn,
            onShow: (instance) => {
              selectCodeLines(instance.reference);
              instance.reference.classList.add('code-annotation-active');
              window.tippy.hideAll();
            },
            onHide: (instance) => {
              unselectCodeLines();
              instance.reference.classList.remove('code-annotation-active');
            },
            maxWidth: 300,
            delay: [50, 0],
            duration: [200, 0],
            offset: [5, 10],
            arrow: true,
            appendTo: function(el) {
              return el.parentElement.parentElement.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'quarto',
            placement: 'right',
            popperOptions: {
              modifiers: [
              {
                name: 'flip',
                options: {
                  flipVariations: false, // true by default
                  allowedAutoPlacements: ['right'],
                  fallbackPlacements: ['right', 'top', 'top-start', 'top-end', 'bottom', 'bottom-start', 'bottom-end', 'left'],
                },
              },
              {
                name: 'preventOverflow',
                options: {
                  mainAxis: false,
                  altAxis: false
                }
              }
              ]        
            }      
          };
          window.tippy(annoteTarget, config); 
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./data-ingestion.html" class="pagination-link" aria-label="Data Wrangling">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data Wrangling</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./processing.html" class="pagination-link" aria-label="Normalization, Reduction, and Clustering">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Normalization, Reduction, and Clustering</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb15" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="an">subtitle:</span><span class="co"> Running quality processing steps </span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> </span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co">- name: Evelyn Metzger</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co">  orcid: 0000-0002-4074-9003</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co">  affiliations: </span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co">  - ref: bsb</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co">  - ref: eveilyeverafter</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span><span class="co"> </span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="co">  eval: true</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co">  freeze: auto</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co">  message: true</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="co">  warning: false</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="an">self-contained:</span><span class="co"> false</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a><span class="an">code-fold:</span><span class="co"> false</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="an">code-tools:</span><span class="co"> true</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="an">code-annotations:</span><span class="co"> hover</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="an">engine:</span><span class="co"> knitr</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a><span class="an">prefer-html:</span><span class="co"> true</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span><span class="co"> live-html</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a><span class="an">webr:</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="co">  packages:</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a><span class="co">    - dplyr</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a><span class="co">    - duckdb</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a><span class="co">    - ggplot2</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a><span class="co">    - plotly</span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a><span class="co">  repos:</span></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a><span class="co">    - https://r-lib.r-universe.dev</span></span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a><span class="an">resources:</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a><span class="co">  - assets/interactives/fov_qc_vary.parquet</span></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a><span class="co">  - assets/interactives/fov_locations.parquet</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a>{{&lt; include ./_extensions/r-wasm/live/_knitr.qmd &gt;}}</span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a><span class="fu"># Quality Processing {#sec-quality-processing}</span></span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a>This section is critical in any data analysis. We'll pass the data through</span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a>a series of filters to ensure the downstream processing steps contain high-quality</span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a>data. We do this by screening and flagging any FOVs that might have lower quality</span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a>and running a counts-based filter. </span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: Preamble</span></span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R code"</span></span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-53"><a href="#cb15-53" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"./preamble.R"</span>)</span>
<span id="cb15-54"><a href="#cb15-54" aria-hidden="true" tabindex="-1"></a>reticulate<span class="sc">::</span><span class="fu">source_python</span>(<span class="st">"./preamble.py"</span>)</span>
<span id="cb15-55"><a href="#cb15-55" aria-hidden="true" tabindex="-1"></a>analysis_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">getwd</span>(), <span class="st">"analysis_results"</span>)</span>
<span id="cb15-56"><a href="#cb15-56" aria-hidden="true" tabindex="-1"></a>input_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"input_files"</span>)</span>
<span id="cb15-57"><a href="#cb15-57" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"output_files"</span>)</span>
<span id="cb15-58"><a href="#cb15-58" aria-hidden="true" tabindex="-1"></a>analysis_asset_dir <span class="ot">&lt;-</span> <span class="st">"./assets/analysis_results"</span> <span class="co"># &lt;1&gt;</span></span>
<span id="cb15-59"><a href="#cb15-59" aria-hidden="true" tabindex="-1"></a>qc_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(output_dir, <span class="st">"qc"</span>)</span>
<span id="cb15-60"><a href="#cb15-60" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(qc_dir)){</span>
<span id="cb15-61"><a href="#cb15-61" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(qc_dir, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb15-62"><a href="#cb15-62" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-63"><a href="#cb15-63" aria-hidden="true" tabindex="-1"></a>results_list_file <span class="ot">=</span> <span class="fu">file.path</span>(analysis_dir, <span class="st">"results_list.rds"</span>)</span>
<span id="cb15-64"><a href="#cb15-64" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(results_list_file)){</span>
<span id="cb15-65"><a href="#cb15-65" aria-hidden="true" tabindex="-1"></a>  results_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb15-66"><a href="#cb15-66" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb15-67"><a href="#cb15-67" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb15-68"><a href="#cb15-68" aria-hidden="true" tabindex="-1"></a>  results_list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(results_list_file)  </span>
<span id="cb15-69"><a href="#cb15-69" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-70"><a href="#cb15-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-71"><a href="#cb15-71" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-72"><a href="#cb15-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-73"><a href="#cb15-73" aria-hidden="true" tabindex="-1"></a>Load the dataset we generated in @sec-data-wrangle. </span>
<span id="cb15-74"><a href="#cb15-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-77"><a href="#cb15-77" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-78"><a href="#cb15-78" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-anndata</span></span>
<span id="cb15-79"><a href="#cb15-79" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-80"><a href="#cb15-80" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-81"><a href="#cb15-81" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: true</span></span>
<span id="cb15-82"><a href="#cb15-82" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-83"><a href="#cb15-83" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb15-84"><a href="#cb15-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-85"><a href="#cb15-85" aria-hidden="true" tabindex="-1"></a>py<span class="sc">$</span>adata <span class="ot">&lt;-</span> py<span class="sc">$</span>ad<span class="sc">$</span><span class="fu">read_h5ad</span>(py<span class="sc">$</span>os<span class="sc">$</span>path<span class="sc">$</span><span class="fu">join</span>(analysis_dir, <span class="st">"anndata-0-initial.h5ad"</span>))</span>
<span id="cb15-86"><a href="#cb15-86" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-87"><a href="#cb15-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-88"><a href="#cb15-88" aria-hidden="true" tabindex="-1"></a><span class="fu">## FOV QC {#sec-fov-qc}</span></span>
<span id="cb15-89"><a href="#cb15-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-90"><a href="#cb15-90" aria-hidden="true" tabindex="-1"></a>Lower quality FOVs generally result in reduced overall gene expression or </span>
<span id="cb15-91"><a href="#cb15-91" aria-hidden="true" tabindex="-1"></a>reduced signal from select genes. Potential causes for lower FOV quality (e.g. lower relative signal for all/select genes for a given FOV compared to majority of FOVs) include tissue/section quality, high autofluorescence, and inadequate fiducials. We can </span>
<span id="cb15-92"><a href="#cb15-92" aria-hidden="true" tabindex="-1"></a>quantify this signal loss using the FOV QC method as described recently in </span>
<span id="cb15-93"><a href="#cb15-93" aria-hidden="true" tabindex="-1"></a>this <span class="co">[</span><span class="ot">scratch space tool</span><span class="co">](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/fov-qc/index.html#fov-quality)</span>.</span>
<span id="cb15-94"><a href="#cb15-94" aria-hidden="true" tabindex="-1"></a>For a more detailed explanation of the the FOV</span>
<span id="cb15-95"><a href="#cb15-95" aria-hidden="true" tabindex="-1"></a>QC functions, expand the box below.</span>
<span id="cb15-96"><a href="#cb15-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-97"><a href="#cb15-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-98"><a href="#cb15-98" aria-hidden="true" tabindex="-1"></a>Since the FOV QC code was written in R, convert relevant annData components to R objects.</span>
<span id="cb15-99"><a href="#cb15-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-100"><a href="#cb15-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-103"><a href="#cb15-103" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb15-104"><a href="#cb15-104" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: sparse-to-dense-and-meta</span></span>
<span id="cb15-105"><a href="#cb15-105" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-106"><a href="#cb15-106" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb15-107"><a href="#cb15-107" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-108"><a href="#cb15-108" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-109"><a href="#cb15-109" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb15-110"><a href="#cb15-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-111"><a href="#cb15-111" aria-hidden="true" tabindex="-1"></a>meta <span class="op">=</span> adata.obs.copy()</span>
<span id="cb15-112"><a href="#cb15-112" aria-hidden="true" tabindex="-1"></a>expr <span class="op">=</span> adata.X.astype(<span class="st">'float64'</span>)</span>
<span id="cb15-113"><a href="#cb15-113" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-114"><a href="#cb15-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-115"><a href="#cb15-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-116"><a href="#cb15-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-117"><a href="#cb15-117" aria-hidden="true" tabindex="-1"></a>Run the <span class="in">`runFOVQC`</span> routine. The <span class="in">`source`</span> call below will load several functions</span>
<span id="cb15-118"><a href="#cb15-118" aria-hidden="true" tabindex="-1"></a>into memory. Since this dataset is from WTX, we'll use the barcode patterns</span>
<span id="cb15-119"><a href="#cb15-119" aria-hidden="true" tabindex="-1"></a>from <span class="in">`Hs_WTX`</span>. </span>
<span id="cb15-120"><a href="#cb15-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-121"><a href="#cb15-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-124"><a href="#cb15-124" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-125"><a href="#cb15-125" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fov-qc-1</span></span>
<span id="cb15-126"><a href="#cb15-126" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-127"><a href="#cb15-127" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-128"><a href="#cb15-128" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-129"><a href="#cb15-129" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-130"><a href="#cb15-130" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb15-131"><a href="#cb15-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-132"><a href="#cb15-132" aria-hidden="true" tabindex="-1"></a><span class="co"># source the necessary functions:</span></span>
<span id="cb15-133"><a href="#cb15-133" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"https://raw.githubusercontent.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/Main/_code/FOV%20QC/FOV%20QC%20utils.R"</span>) <span class="co"># &lt;1&gt;</span></span>
<span id="cb15-134"><a href="#cb15-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-135"><a href="#cb15-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-136"><a href="#cb15-136" aria-hidden="true" tabindex="-1"></a>allbarcodes <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">url</span>(<span class="st">"https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/raw/Main/_code/FOV%20QC/barcodes_by_panel.RDS"</span>))</span>
<span id="cb15-137"><a href="#cb15-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-138"><a href="#cb15-138" aria-hidden="true" tabindex="-1"></a>results_list[[<span class="st">'barcodes'</span>]] <span class="ot">&lt;-</span> allbarcodes[<span class="st">'Hs_WTX'</span>]</span>
<span id="cb15-139"><a href="#cb15-139" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb15-140"><a href="#cb15-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-141"><a href="#cb15-141" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-142"><a href="#cb15-142" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>For historic reasons, the characters in the barcode string are duplicated. </span>
<span id="cb15-143"><a href="#cb15-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-144"><a href="#cb15-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-145"><a href="#cb15-145" aria-hidden="true" tabindex="-1"></a>:::{.noteworthybox title="Understanding the FOV QC parameters" collapse="true"}</span>
<span id="cb15-146"><a href="#cb15-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-147"><a href="#cb15-147" aria-hidden="true" tabindex="-1"></a>This deep dive shows a how the two tuning parameters of the FOV QC function, </span>
<span id="cb15-148"><a href="#cb15-148" aria-hidden="true" tabindex="-1"></a><span class="in">`max_prop_loss`</span> and <span class="in">`max_totalcount_loss`</span>), affect the results.</span>
<span id="cb15-149"><a href="#cb15-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-150"><a href="#cb15-150" aria-hidden="true" tabindex="-1"></a>The parameter <span class="in">`max_prop_loss`</span> sets a tolerance for the maximum allowable </span>
<span id="cb15-151"><a href="#cb15-151" aria-hidden="true" tabindex="-1"></a>proportion of signal loss for any single barcode bit before an FOV is flagged. </span>
<span id="cb15-152"><a href="#cb15-152" aria-hidden="true" tabindex="-1"></a>For example, if we set <span class="in">`max_prop_loss`</span> to 0.3, it means we'd accept a given <span class="in">`bit`</span></span>
<span id="cb15-153"><a href="#cb15-153" aria-hidden="true" tabindex="-1"></a>given if it has lost 30% of its original strength. </span>
<span id="cb15-154"><a href="#cb15-154" aria-hidden="true" tabindex="-1"></a>While <span class="in">`max_prop_loss`</span> focuses on the errors in a single specific barcode <span class="in">`bit`</span>,</span>
<span id="cb15-155"><a href="#cb15-155" aria-hidden="true" tabindex="-1"></a><span class="in">`max_totalcounts_loss`</span> looks at the bigger picture. It asks, "Is this entire FOV</span>
<span id="cb15-156"><a href="#cb15-156" aria-hidden="true" tabindex="-1"></a>significantly dimmer or less populated with transcripts than it should be?" This</span>
<span id="cb15-157"><a href="#cb15-157" aria-hidden="true" tabindex="-1"></a>can help detect broader issues like poor tissue quality, focusing problems, or </span>
<span id="cb15-158"><a href="#cb15-158" aria-hidden="true" tabindex="-1"></a>reagent failures in that specific area.</span>
<span id="cb15-159"><a href="#cb15-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-160"><a href="#cb15-160" aria-hidden="true" tabindex="-1"></a>:::: {.content-visible unless-format="pdf"}</span>
<span id="cb15-161"><a href="#cb15-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-162"><a href="#cb15-162" aria-hidden="true" tabindex="-1"></a>Let's see what happens to the number of flagged FOVs when adjust these parameters.</span>
<span id="cb15-163"><a href="#cb15-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-164"><a href="#cb15-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-167"><a href="#cb15-167" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-168"><a href="#cb15-168" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-169"><a href="#cb15-169" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-170"><a href="#cb15-170" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb15-171"><a href="#cb15-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-172"><a href="#cb15-172" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb15-173"><a href="#cb15-173" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr)</span>
<span id="cb15-174"><a href="#cb15-174" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(nanoparquet)</span>
<span id="cb15-175"><a href="#cb15-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-176"><a href="#cb15-176" aria-hidden="true" tabindex="-1"></a>raw_data <span class="ot">&lt;-</span> py<span class="sc">$</span>expr</span>
<span id="cb15-177"><a href="#cb15-177" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(raw_data) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>var_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb15-178"><a href="#cb15-178" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(raw_data) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb15-179"><a href="#cb15-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-180"><a href="#cb15-180" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to sparse</span></span>
<span id="cb15-181"><a href="#cb15-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-182"><a href="#cb15-182" aria-hidden="true" tabindex="-1"></a>xy <span class="ot">&lt;-</span> py<span class="sc">$</span>meta <span class="sc">%&gt;%</span> <span class="fu">select</span>(x_slide_mm, y_slide_mm) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb15-183"><a href="#cb15-183" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(xy) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb15-184"><a href="#cb15-184" aria-hidden="true" tabindex="-1"></a>fov <span class="ot">&lt;-</span> py<span class="sc">$</span>meta <span class="sc">%&gt;%</span> <span class="fu">select</span>(fov) <span class="sc">%&gt;%</span> <span class="fu">pull</span>()</span>
<span id="cb15-185"><a href="#cb15-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-186"><a href="#cb15-186" aria-hidden="true" tabindex="-1"></a>vary_max_prop_loss <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="at">from=</span><span class="fl">0.1</span>, <span class="at">to=</span><span class="fl">0.6</span>, <span class="at">by=</span><span class="fl">0.1</span>) <span class="co"># &lt;1&gt;</span></span>
<span id="cb15-187"><a href="#cb15-187" aria-hidden="true" tabindex="-1"></a>vary_max_totalcounts_loss <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="at">from=</span><span class="fl">0.1</span>, <span class="at">to=</span><span class="fl">0.6</span>, <span class="at">by=</span><span class="fl">0.1</span>)</span>
<span id="cb15-188"><a href="#cb15-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-189"><a href="#cb15-189" aria-hidden="true" tabindex="-1"></a>param_grid <span class="ot">&lt;-</span> <span class="fu">crossing</span>(</span>
<span id="cb15-190"><a href="#cb15-190" aria-hidden="true" tabindex="-1"></a>  <span class="at">i =</span> vary_max_prop_loss,</span>
<span id="cb15-191"><a href="#cb15-191" aria-hidden="true" tabindex="-1"></a>  <span class="at">j =</span> vary_max_totalcounts_loss</span>
<span id="cb15-192"><a href="#cb15-192" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-193"><a href="#cb15-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-194"><a href="#cb15-194" aria-hidden="true" tabindex="-1"></a>fov_qc_vary_file <span class="ot">&lt;-</span> <span class="st">"./assets/interactives/fov_qc_vary.parquet"</span></span>
<span id="cb15-195"><a href="#cb15-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-196"><a href="#cb15-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-197"><a href="#cb15-197" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(fov_qc_vary_file)) {</span>
<span id="cb15-198"><a href="#cb15-198" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Existing results file found. Determining which simulations to resume.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb15-199"><a href="#cb15-199" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-200"><a href="#cb15-200" aria-hidden="true" tabindex="-1"></a>  existing_results <span class="ot">&lt;-</span> <span class="fu">read_parquet</span>(fov_qc_vary_file) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb15-201"><a href="#cb15-201" aria-hidden="true" tabindex="-1"></a>  existing_results<span class="sc">$</span>max_prop_loss <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(existing_results<span class="sc">$</span>max_prop_loss)</span>
<span id="cb15-202"><a href="#cb15-202" aria-hidden="true" tabindex="-1"></a>  existing_results<span class="sc">$</span>max_totalcounts_loss <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(existing_results<span class="sc">$</span>max_totalcounts_loss)</span>
<span id="cb15-203"><a href="#cb15-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-204"><a href="#cb15-204" aria-hidden="true" tabindex="-1"></a>  completed_params <span class="ot">&lt;-</span> existing_results <span class="sc">%&gt;%</span> </span>
<span id="cb15-205"><a href="#cb15-205" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(max_prop_loss, max_totalcounts_loss) <span class="sc">%&gt;%</span></span>
<span id="cb15-206"><a href="#cb15-206" aria-hidden="true" tabindex="-1"></a>    <span class="fu">distinct</span>()</span>
<span id="cb15-207"><a href="#cb15-207" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(completed_params) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'i'</span>, <span class="st">'j'</span>)</span>
<span id="cb15-208"><a href="#cb15-208" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-209"><a href="#cb15-209" aria-hidden="true" tabindex="-1"></a>  params_to_run <span class="ot">&lt;-</span> <span class="fu">anti_join</span>(param_grid, completed_params, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"i"</span>, <span class="st">"j"</span>))</span>
<span id="cb15-210"><a href="#cb15-210" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-211"><a href="#cb15-211" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb15-212"><a href="#cb15-212" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"No existing results file found. Starting simulation from scratch.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb15-213"><a href="#cb15-213" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-214"><a href="#cb15-214" aria-hidden="true" tabindex="-1"></a>  params_to_run <span class="ot">&lt;-</span> param_grid</span>
<span id="cb15-215"><a href="#cb15-215" aria-hidden="true" tabindex="-1"></a>  existing_results <span class="ot">&lt;-</span> <span class="fu">tibble</span>()</span>
<span id="cb15-216"><a href="#cb15-216" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-217"><a href="#cb15-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-218"><a href="#cb15-218" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">nrow</span>(params_to_run) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb15-219"><a href="#cb15-219" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="fu">sprintf</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Running %d new simulation(s)...</span><span class="sc">\n</span><span class="st">"</span>, <span class="fu">nrow</span>(params_to_run)))</span>
<span id="cb15-220"><a href="#cb15-220" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-221"><a href="#cb15-221" aria-hidden="true" tabindex="-1"></a>  results <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb15-222"><a href="#cb15-222" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-223"><a href="#cb15-223" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (row_index <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(params_to_run)) {</span>
<span id="cb15-224"><a href="#cb15-224" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="fu">paste0</span>(<span class="st">"Running row "</span>, row_index))</span>
<span id="cb15-225"><a href="#cb15-225" aria-hidden="true" tabindex="-1"></a>    current_params <span class="ot">&lt;-</span> params_to_run[row_index, ]</span>
<span id="cb15-226"><a href="#cb15-226" aria-hidden="true" tabindex="-1"></a>    i_val <span class="ot">&lt;-</span> current_params<span class="sc">$</span>i</span>
<span id="cb15-227"><a href="#cb15-227" aria-hidden="true" tabindex="-1"></a>    j_val <span class="ot">&lt;-</span> current_params<span class="sc">$</span>j</span>
<span id="cb15-228"><a href="#cb15-228" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-229"><a href="#cb15-229" aria-hidden="true" tabindex="-1"></a>    res <span class="ot">&lt;-</span> <span class="fu">runFOVQC</span>(<span class="at">counts =</span> raw_data, <span class="at">xy =</span> xy, <span class="at">fov =</span> fov,</span>
<span id="cb15-230"><a href="#cb15-230" aria-hidden="true" tabindex="-1"></a>                        <span class="at">barcodemap =</span> results_list[[<span class="st">'barcodes'</span>]][[<span class="dv">1</span>]],</span>
<span id="cb15-231"><a href="#cb15-231" aria-hidden="true" tabindex="-1"></a>                        <span class="at">max_prop_loss =</span> i_val, <span class="at">max_totalcounts_loss =</span> j_val)</span>
<span id="cb15-232"><a href="#cb15-232" aria-hidden="true" tabindex="-1"></a>    flagged_fovs <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(res<span class="sc">$</span>flaggedfovs)</span>
<span id="cb15-233"><a href="#cb15-233" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-234"><a href="#cb15-234" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(flagged_fovs) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb15-235"><a href="#cb15-235" aria-hidden="true" tabindex="-1"></a>      <span class="co"># If FOVs were flagged, create a tibble and add it to our list</span></span>
<span id="cb15-236"><a href="#cb15-236" aria-hidden="true" tabindex="-1"></a>      results[[<span class="fu">length</span>(results) <span class="sc">+</span> <span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">FOV =</span> flagged_fovs) <span class="sc">%&gt;%</span></span>
<span id="cb15-237"><a href="#cb15-237" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(</span>
<span id="cb15-238"><a href="#cb15-238" aria-hidden="true" tabindex="-1"></a>          <span class="at">max_prop_loss =</span> i_val,</span>
<span id="cb15-239"><a href="#cb15-239" aria-hidden="true" tabindex="-1"></a>          <span class="at">max_totalcounts_loss =</span> j_val</span>
<span id="cb15-240"><a href="#cb15-240" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb15-241"><a href="#cb15-241" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb15-242"><a href="#cb15-242" aria-hidden="true" tabindex="-1"></a>      results[[<span class="fu">length</span>(results) <span class="sc">+</span> <span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">FOV=</span><span class="cn">NA</span>, </span>
<span id="cb15-243"><a href="#cb15-243" aria-hidden="true" tabindex="-1"></a>             <span class="at">max_prop_loss =</span> i_val,</span>
<span id="cb15-244"><a href="#cb15-244" aria-hidden="true" tabindex="-1"></a>          <span class="at">max_totalcounts_loss =</span> j_val)</span>
<span id="cb15-245"><a href="#cb15-245" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb15-246"><a href="#cb15-246" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb15-247"><a href="#cb15-247" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-248"><a href="#cb15-248" aria-hidden="true" tabindex="-1"></a>  new_results <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(results)</span>
<span id="cb15-249"><a href="#cb15-249" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-250"><a href="#cb15-250" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Simulation batch complete. Combining and saving results.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb15-251"><a href="#cb15-251" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-252"><a href="#cb15-252" aria-hidden="true" tabindex="-1"></a>  combined_results <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(existing_results, new_results) <span class="sc">%&gt;%</span></span>
<span id="cb15-253"><a href="#cb15-253" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(max_prop_loss, max_totalcounts_loss, FOV)</span>
<span id="cb15-254"><a href="#cb15-254" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-255"><a href="#cb15-255" aria-hidden="true" tabindex="-1"></a>  existing_results<span class="sc">$</span>max_prop_loss <span class="ot">&lt;-</span> <span class="fu">as.character</span>(existing_results<span class="sc">$</span>max_prop_loss)</span>
<span id="cb15-256"><a href="#cb15-256" aria-hidden="true" tabindex="-1"></a>  existing_results<span class="sc">$</span>max_totalcounts_loss <span class="ot">&lt;-</span> <span class="fu">as.character</span>(existing_results<span class="sc">$</span>max_totalcounts_loss)</span>
<span id="cb15-257"><a href="#cb15-257" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-258"><a href="#cb15-258" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write_parquet</span>(combined_results, fov_qc_vary_file)</span>
<span id="cb15-259"><a href="#cb15-259" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-260"><a href="#cb15-260" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Successfully saved updated results to:"</span>, fov_qc_vary_file, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb15-261"><a href="#cb15-261" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-262"><a href="#cb15-262" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb15-263"><a href="#cb15-263" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">All simulations are already complete. Nothing to do.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb15-264"><a href="#cb15-264" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-265"><a href="#cb15-265" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-266"><a href="#cb15-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-267"><a href="#cb15-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-270"><a href="#cb15-270" aria-hidden="true" tabindex="-1"></a><span class="in">```{ojs}</span></span>
<span id="cb15-271"><a href="#cb15-271" aria-hidden="true" tabindex="-1"></a><span class="co">//| eval: true</span></span>
<span id="cb15-272"><a href="#cb15-272" aria-hidden="true" tabindex="-1"></a><span class="co">//| echo: false</span></span>
<span id="cb15-273"><a href="#cb15-273" aria-hidden="true" tabindex="-1"></a>viewof max_prop_loss <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">range</span>([<span class="fl">0.1</span><span class="op">,</span> <span class="fl">0.6</span>]<span class="op">,</span> {</span>
<span id="cb15-274"><a href="#cb15-274" aria-hidden="true" tabindex="-1"></a>  <span class="dt">step</span><span class="op">:</span> <span class="fl">0.1</span><span class="op">,</span></span>
<span id="cb15-275"><a href="#cb15-275" aria-hidden="true" tabindex="-1"></a>  <span class="dt">label</span><span class="op">:</span> <span class="st">"Max Proportion Loss"</span></span>
<span id="cb15-276"><a href="#cb15-276" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb15-277"><a href="#cb15-277" aria-hidden="true" tabindex="-1"></a>viewof max_totalcounts_loss <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">range</span>([<span class="fl">0.1</span><span class="op">,</span> <span class="fl">0.6</span>]<span class="op">,</span> {</span>
<span id="cb15-278"><a href="#cb15-278" aria-hidden="true" tabindex="-1"></a>  <span class="dt">step</span><span class="op">:</span> <span class="fl">0.1</span><span class="op">,</span></span>
<span id="cb15-279"><a href="#cb15-279" aria-hidden="true" tabindex="-1"></a>  <span class="dt">label</span><span class="op">:</span> <span class="st">"Max Total Counts Loss"</span></span>
<span id="cb15-280"><a href="#cb15-280" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb15-281"><a href="#cb15-281" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-282"><a href="#cb15-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-283"><a href="#cb15-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-286"><a href="#cb15-286" aria-hidden="true" tabindex="-1"></a><span class="in">```{webr}</span></span>
<span id="cb15-287"><a href="#cb15-287" aria-hidden="true" tabindex="-1"></a><span class="in">#| echo: false</span></span>
<span id="cb15-288"><a href="#cb15-288" aria-hidden="true" tabindex="-1"></a><span class="in">#| eval: true</span></span>
<span id="cb15-289"><a href="#cb15-289" aria-hidden="true" tabindex="-1"></a><span class="in">#| edit: false</span></span>
<span id="cb15-290"><a href="#cb15-290" aria-hidden="true" tabindex="-1"></a><span class="in">#| define:</span></span>
<span id="cb15-291"><a href="#cb15-291" aria-hidden="true" tabindex="-1"></a><span class="in">#|   - fov_locations</span></span>
<span id="cb15-292"><a href="#cb15-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-293"><a href="#cb15-293" aria-hidden="true" tabindex="-1"></a><span class="in"># This just reads the FOV locations parquet and makes the data</span></span>
<span id="cb15-294"><a href="#cb15-294" aria-hidden="true" tabindex="-1"></a><span class="in"># available to to webr. It only needs to be loaded once.</span></span>
<span id="cb15-295"><a href="#cb15-295" aria-hidden="true" tabindex="-1"></a><span class="in">con0 &lt;- dbConnect(duckdb())</span></span>
<span id="cb15-296"><a href="#cb15-296" aria-hidden="true" tabindex="-1"></a><span class="in">fov_locations &lt;- dbGetQuery(</span></span>
<span id="cb15-297"><a href="#cb15-297" aria-hidden="true" tabindex="-1"></a><span class="in">  con0, </span></span>
<span id="cb15-298"><a href="#cb15-298" aria-hidden="true" tabindex="-1"></a><span class="in">  "SELECT * FROM read_parquet('assets/interactives/fov_locations.parquet');"</span></span>
<span id="cb15-299"><a href="#cb15-299" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb15-300"><a href="#cb15-300" aria-hidden="true" tabindex="-1"></a><span class="in">dbDisconnect(con0, shutdown = TRUE)</span></span>
<span id="cb15-301"><a href="#cb15-301" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-302"><a href="#cb15-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-305"><a href="#cb15-305" aria-hidden="true" tabindex="-1"></a><span class="in">```{webr}</span></span>
<span id="cb15-306"><a href="#cb15-306" aria-hidden="true" tabindex="-1"></a><span class="in">#| echo: false</span></span>
<span id="cb15-307"><a href="#cb15-307" aria-hidden="true" tabindex="-1"></a><span class="in">#| eval: true</span></span>
<span id="cb15-308"><a href="#cb15-308" aria-hidden="true" tabindex="-1"></a><span class="in">#| edit: false</span></span>
<span id="cb15-309"><a href="#cb15-309" aria-hidden="true" tabindex="-1"></a><span class="in">#| input:</span></span>
<span id="cb15-310"><a href="#cb15-310" aria-hidden="true" tabindex="-1"></a><span class="in">#|   - max_prop_loss</span></span>
<span id="cb15-311"><a href="#cb15-311" aria-hidden="true" tabindex="-1"></a><span class="in">#|   - max_totalcounts_loss</span></span>
<span id="cb15-312"><a href="#cb15-312" aria-hidden="true" tabindex="-1"></a><span class="in">#| define:</span></span>
<span id="cb15-313"><a href="#cb15-313" aria-hidden="true" tabindex="-1"></a><span class="in">#|   - filtered_data</span></span>
<span id="cb15-314"><a href="#cb15-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-315"><a href="#cb15-315" aria-hidden="true" tabindex="-1"></a><span class="in"># This will re-run every time max_prop_loss and max_totalcounts_loss</span></span>
<span id="cb15-316"><a href="#cb15-316" aria-hidden="true" tabindex="-1"></a><span class="in"># change (reactive with OJS sliders)</span></span>
<span id="cb15-317"><a href="#cb15-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-318"><a href="#cb15-318" aria-hidden="true" tabindex="-1"></a><span class="in"># connection to an in-memory DuckDB database</span></span>
<span id="cb15-319"><a href="#cb15-319" aria-hidden="true" tabindex="-1"></a><span class="in">con &lt;- dbConnect(duckdb())</span></span>
<span id="cb15-320"><a href="#cb15-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-321"><a href="#cb15-321" aria-hidden="true" tabindex="-1"></a><span class="in"># build the SQL query</span></span>
<span id="cb15-322"><a href="#cb15-322" aria-hidden="true" tabindex="-1"></a><span class="in">query &lt;- paste0(</span></span>
<span id="cb15-323"><a href="#cb15-323" aria-hidden="true" tabindex="-1"></a><span class="in">  "SELECT * FROM 'assets/interactives/fov_qc_vary.parquet' WHERE max_prop_loss = ",</span></span>
<span id="cb15-324"><a href="#cb15-324" aria-hidden="true" tabindex="-1"></a><span class="in">  max_prop_loss,</span></span>
<span id="cb15-325"><a href="#cb15-325" aria-hidden="true" tabindex="-1"></a><span class="in">  " AND max_totalcounts_loss = ",</span></span>
<span id="cb15-326"><a href="#cb15-326" aria-hidden="true" tabindex="-1"></a><span class="in">  max_totalcounts_loss</span></span>
<span id="cb15-327"><a href="#cb15-327" aria-hidden="true" tabindex="-1"></a><span class="in">)</span></span>
<span id="cb15-328"><a href="#cb15-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-329"><a href="#cb15-329" aria-hidden="true" tabindex="-1"></a><span class="in">query_data &lt;- dbGetQuery(con, query)</span></span>
<span id="cb15-330"><a href="#cb15-330" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-331"><a href="#cb15-331" aria-hidden="true" tabindex="-1"></a><span class="in">dbDisconnect(con, shutdown = TRUE)</span></span>
<span id="cb15-332"><a href="#cb15-332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-333"><a href="#cb15-333" aria-hidden="true" tabindex="-1"></a><span class="in">filtered_data &lt;- fov_locations</span></span>
<span id="cb15-334"><a href="#cb15-334" aria-hidden="true" tabindex="-1"></a><span class="in">filtered_data$flagged &lt;- ifelse(</span></span>
<span id="cb15-335"><a href="#cb15-335" aria-hidden="true" tabindex="-1"></a><span class="in">  filtered_data$FOV %in% query_data$FOV, TRUE, FALSE)</span></span>
<span id="cb15-336"><a href="#cb15-336" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-337"><a href="#cb15-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-338"><a href="#cb15-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-339"><a href="#cb15-339" aria-hidden="true" tabindex="-1"></a>This code below will plot the spatial distribution of failed FOVs. Feel free</span>
<span id="cb15-340"><a href="#cb15-340" aria-hidden="true" tabindex="-1"></a>to adjust the code, if desired. </span>
<span id="cb15-343"><a href="#cb15-343" aria-hidden="true" tabindex="-1"></a><span class="in">```{webr}</span></span>
<span id="cb15-344"><a href="#cb15-344" aria-hidden="true" tabindex="-1"></a><span class="in">#| autorun: true</span></span>
<span id="cb15-345"><a href="#cb15-345" aria-hidden="true" tabindex="-1"></a><span class="in">#| input:</span></span>
<span id="cb15-346"><a href="#cb15-346" aria-hidden="true" tabindex="-1"></a><span class="in">#|   - filtered_data</span></span>
<span id="cb15-347"><a href="#cb15-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-348"><a href="#cb15-348" aria-hidden="true" tabindex="-1"></a><span class="in">p &lt;- ggplot(filtered_data, </span></span>
<span id="cb15-349"><a href="#cb15-349" aria-hidden="true" tabindex="-1"></a><span class="in">         aes(x=x_global_mm, </span></span>
<span id="cb15-350"><a href="#cb15-350" aria-hidden="true" tabindex="-1"></a><span class="in">             y=y_global_mm, </span></span>
<span id="cb15-351"><a href="#cb15-351" aria-hidden="true" tabindex="-1"></a><span class="in">             color=factor(flagged),</span></span>
<span id="cb15-352"><a href="#cb15-352" aria-hidden="true" tabindex="-1"></a><span class="in">             label=FOV)) + </span></span>
<span id="cb15-353"><a href="#cb15-353" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_point() + theme_bw() + </span></span>
<span id="cb15-354"><a href="#cb15-354" aria-hidden="true" tabindex="-1"></a><span class="in">  coord_fixed() + </span></span>
<span id="cb15-355"><a href="#cb15-355" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_color_manual(values=c(`TRUE`='red',</span></span>
<span id="cb15-356"><a href="#cb15-356" aria-hidden="true" tabindex="-1"></a><span class="in">                              `FALSE`='grey'))</span></span>
<span id="cb15-357"><a href="#cb15-357" aria-hidden="true" tabindex="-1"></a><span class="in">plotly::ggplotly(p)</span></span>
<span id="cb15-358"><a href="#cb15-358" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-359"><a href="#cb15-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-360"><a href="#cb15-360" aria-hidden="true" tabindex="-1"></a>::::</span>
<span id="cb15-361"><a href="#cb15-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-362"><a href="#cb15-362" aria-hidden="true" tabindex="-1"></a>:::: {.content-visible when-format="pdf"}</span>
<span id="cb15-363"><a href="#cb15-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-364"><a href="#cb15-364" aria-hidden="true" tabindex="-1"></a>For an interactive visual of the effect of these tuning parameters, please</span>
<span id="cb15-365"><a href="#cb15-365" aria-hidden="true" tabindex="-1"></a>see the online version of this book.</span>
<span id="cb15-366"><a href="#cb15-366" aria-hidden="true" tabindex="-1"></a>::::</span>
<span id="cb15-367"><a href="#cb15-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-368"><a href="#cb15-368" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb15-369"><a href="#cb15-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-370"><a href="#cb15-370" aria-hidden="true" tabindex="-1"></a>With the functions loaded, execute the <span class="in">`runFOVQC`</span> function. For this example,</span>
<span id="cb15-371"><a href="#cb15-371" aria-hidden="true" tabindex="-1"></a>I choose a less restrictive parameter set (<span class="in">`max_prop_loss`</span> = 0.4 </span>
<span id="cb15-372"><a href="#cb15-372" aria-hidden="true" tabindex="-1"></a>and <span class="in">`max_totalcounts_loss`</span> = 0.3).</span>
<span id="cb15-373"><a href="#cb15-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-376"><a href="#cb15-376" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-377"><a href="#cb15-377" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: runFOVQC</span></span>
<span id="cb15-378"><a href="#cb15-378" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-379"><a href="#cb15-379" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-380"><a href="#cb15-380" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-381"><a href="#cb15-381" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-382"><a href="#cb15-382" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb15-383"><a href="#cb15-383" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-384"><a href="#cb15-384" aria-hidden="true" tabindex="-1"></a>raw_data <span class="ot">&lt;-</span> py<span class="sc">$</span>expr</span>
<span id="cb15-385"><a href="#cb15-385" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(raw_data) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>var_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb15-386"><a href="#cb15-386" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(raw_data) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb15-387"><a href="#cb15-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-388"><a href="#cb15-388" aria-hidden="true" tabindex="-1"></a>xy <span class="ot">&lt;-</span> py<span class="sc">$</span>meta <span class="sc">%&gt;%</span> <span class="fu">select</span>(x_slide_mm, y_slide_mm) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb15-389"><a href="#cb15-389" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(xy) <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs_names<span class="sc">$</span><span class="fu">to_list</span>()</span>
<span id="cb15-390"><a href="#cb15-390" aria-hidden="true" tabindex="-1"></a>fov <span class="ot">&lt;-</span> py<span class="sc">$</span>meta <span class="sc">%&gt;%</span> <span class="fu">select</span>(fov) <span class="sc">%&gt;%</span> <span class="fu">pull</span>()</span>
<span id="cb15-391"><a href="#cb15-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-392"><a href="#cb15-392" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">runFOVQC</span>(<span class="at">counts =</span> raw_data, <span class="at">xy =</span> xy, <span class="at">fov =</span> fov, </span>
<span id="cb15-393"><a href="#cb15-393" aria-hidden="true" tabindex="-1"></a>                <span class="at">barcodemap =</span> results_list[[<span class="st">'barcodes'</span>]]<span class="sc">$</span>Hs_WTX,</span>
<span id="cb15-394"><a href="#cb15-394" aria-hidden="true" tabindex="-1"></a>                <span class="at">max_prop_loss =</span> <span class="fl">0.4</span>, <span class="at">max_totalcounts_loss =</span> <span class="fl">0.3</span>)</span>
<span id="cb15-395"><a href="#cb15-395" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(res, <span class="fu">file.path</span>(qc_dir, <span class="st">"fovqc.rds"</span>))</span>
<span id="cb15-396"><a href="#cb15-396" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-397"><a href="#cb15-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-398"><a href="#cb15-398" aria-hidden="true" tabindex="-1"></a>Plot the flagged FOVs in R simply with <span class="in">`mapFlaggedFOVs(res)`</span> function to get the </span>
<span id="cb15-399"><a href="#cb15-399" aria-hidden="true" tabindex="-1"></a>spatial layout of any FOVs that were flagged. In this case (@fig-qc-fov-1), you</span>
<span id="cb15-400"><a href="#cb15-400" aria-hidden="true" tabindex="-1"></a>can see some FOVs that were flagged along the tissue perimeter. </span>
<span id="cb15-401"><a href="#cb15-401" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-404"><a href="#cb15-404" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-405"><a href="#cb15-405" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-flagged-fovs</span></span>
<span id="cb15-406"><a href="#cb15-406" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-407"><a href="#cb15-407" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-408"><a href="#cb15-408" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-409"><a href="#cb15-409" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-410"><a href="#cb15-410" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-411"><a href="#cb15-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-412"><a href="#cb15-412" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">file.path</span>(qc_dir, <span class="st">"flagged_fovs.png"</span>), </span>
<span id="cb15-413"><a href="#cb15-413" aria-hidden="true" tabindex="-1"></a>    <span class="at">width=</span><span class="dv">10</span>, <span class="at">height=</span><span class="dv">10</span>, <span class="at">res=</span><span class="dv">350</span>, <span class="at">units =</span> <span class="st">"in"</span>, <span class="at">type =</span> <span class="st">"cairo"</span>)</span>
<span id="cb15-414"><a href="#cb15-414" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mapFlaggedFOVs</span>(res)</span>
<span id="cb15-415"><a href="#cb15-415" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb15-416"><a href="#cb15-416" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-417"><a href="#cb15-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-420"><a href="#cb15-420" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-421"><a href="#cb15-421" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-qc-fov-1</span></span>
<span id="cb15-422"><a href="#cb15-422" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-423"><a href="#cb15-423" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-424"><a href="#cb15-424" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-425"><a href="#cb15-425" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 6</span></span>
<span id="cb15-426"><a href="#cb15-426" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 6</span></span>
<span id="cb15-427"><a href="#cb15-427" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Spatial location of the flagged FOVs."</span></span>
<span id="cb15-428"><a href="#cb15-428" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-429"><a href="#cb15-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-430"><a href="#cb15-430" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>), <span class="st">"flagged_fovs.png"</span>, qc_dir)</span>
<span id="cb15-431"><a href="#cb15-431" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-432"><a href="#cb15-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-433"><a href="#cb15-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-434"><a href="#cb15-434" aria-hidden="true" tabindex="-1"></a>For the targets that were flagged, plot how many FOVs they were flagged in.</span>
<span id="cb15-435"><a href="#cb15-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-438"><a href="#cb15-438" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-439"><a href="#cb15-439" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fov-qc-2</span></span>
<span id="cb15-440"><a href="#cb15-440" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-441"><a href="#cb15-441" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-442"><a href="#cb15-442" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-443"><a href="#cb15-443" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-444"><a href="#cb15-444" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb15-445"><a href="#cb15-445" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-446"><a href="#cb15-446" aria-hidden="true" tabindex="-1"></a>flagged_genes <span class="ot">&lt;-</span> res<span class="sc">$</span>flagged_fov_x_gene[,<span class="dv">2</span>]</span>
<span id="cb15-447"><a href="#cb15-447" aria-hidden="true" tabindex="-1"></a>flagged_genes <span class="ot">&lt;-</span> flagged_genes[<span class="sc">!</span><span class="fu">is.na</span>(flagged_genes)] <span class="co"># &lt;1&gt;</span></span>
<span id="cb15-448"><a href="#cb15-448" aria-hidden="true" tabindex="-1"></a>flagged_genes <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">rev</span>(<span class="fu">sort</span>(<span class="fu">table</span>(flagged_genes))))</span>
<span id="cb15-449"><a href="#cb15-449" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(flagged_genes)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">'target'</span></span>
<span id="cb15-450"><a href="#cb15-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-451"><a href="#cb15-451" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">target=</span>py<span class="sc">$</span>adata<span class="sc">$</span>var_names<span class="sc">$</span><span class="fu">to_list</span>())</span>
<span id="cb15-452"><a href="#cb15-452" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">merge</span>(df, flagged_genes, <span class="at">by=</span><span class="st">"target"</span>, <span class="at">all.x=</span><span class="cn">TRUE</span>, <span class="at">all.y=</span><span class="cn">FALSE</span>)</span>
<span id="cb15-453"><a href="#cb15-453" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>Freq[<span class="fu">is.na</span>(df<span class="sc">$</span>Freq)] <span class="ot">&lt;-</span> <span class="dv">0</span>L</span>
<span id="cb15-454"><a href="#cb15-454" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-455"><a href="#cb15-455" aria-hidden="true" tabindex="-1"></a>py<span class="sc">$</span>adata<span class="sc">$</span>var[<span class="st">'fov_qc_flagged'</span>] <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(df<span class="sc">$</span>Freq) <span class="co"># &lt;2&gt;</span></span>
<span id="cb15-456"><a href="#cb15-456" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-457"><a href="#cb15-457" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(flagged_genes, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(Freq), <span class="at">fill =</span> <span class="fu">factor</span>(Freq))) <span class="sc">+</span></span>
<span id="cb15-458"><a href="#cb15-458" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb15-459"><a href="#cb15-459" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">stat=</span><span class="st">'count'</span>, <span class="fu">aes</span>(<span class="at">label=</span><span class="fu">after_stat</span>(count)), <span class="at">vjust=</span><span class="sc">-</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb15-460"><a href="#cb15-460" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Number of FOVs"</span>, <span class="at">y =</span> <span class="st">"Number of Flagged Genes"</span>) <span class="sc">+</span></span>
<span id="cb15-461"><a href="#cb15-461" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb15-462"><a href="#cb15-462" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">breaks =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">13</span>) <span class="sc">+</span> </span>
<span id="cb15-463"><a href="#cb15-463" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb15-464"><a href="#cb15-464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-465"><a href="#cb15-465" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(p, <span class="at">filename=</span><span class="fu">file.path</span>(qc_dir, <span class="st">"genes_impacted_in_flagged_fovs.png"</span>),</span>
<span id="cb15-466"><a href="#cb15-466" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">4</span>, <span class="at">height=</span><span class="dv">5</span>, <span class="at">dpi=</span><span class="dv">150</span>, <span class="at">type=</span><span class="st">"cairo"</span>)</span>
<span id="cb15-467"><a href="#cb15-467" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-468"><a href="#cb15-468" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-469"><a href="#cb15-469" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>A value of NA here means the gene in a given FOV was _not_ flagged.</span>
<span id="cb15-470"><a href="#cb15-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-473"><a href="#cb15-473" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-474"><a href="#cb15-474" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-qc-fov-2</span></span>
<span id="cb15-475"><a href="#cb15-475" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-476"><a href="#cb15-476" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-477"><a href="#cb15-477" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-478"><a href="#cb15-478" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 6</span></span>
<span id="cb15-479"><a href="#cb15-479" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 6</span></span>
<span id="cb15-480"><a href="#cb15-480" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Number of targets impacted in a given number of FOVs."</span></span>
<span id="cb15-481"><a href="#cb15-481" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-482"><a href="#cb15-482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-483"><a href="#cb15-483" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>), <span class="st">"genes_impacted_in_flagged_fovs.png"</span>, qc_dir)</span>
<span id="cb15-484"><a href="#cb15-484" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-485"><a href="#cb15-485" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-486"><a href="#cb15-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-487"><a href="#cb15-487" aria-hidden="true" tabindex="-1"></a>In @fig-qc-fov-2, there are only a few genes (149) that were flagged in three FOVs. </span>
<span id="cb15-488"><a href="#cb15-488" aria-hidden="true" tabindex="-1"></a>If we found a set of genes that were systematically flagged in many FOVs, we may</span>
<span id="cb15-489"><a href="#cb15-489" aria-hidden="true" tabindex="-1"></a>consider removing the impacted FOVs. Alternatively we could remove the impacted</span>
<span id="cb15-490"><a href="#cb15-490" aria-hidden="true" tabindex="-1"></a>genes from all FOVs. In this sample, it's probably fine to </span>
<span id="cb15-491"><a href="#cb15-491" aria-hidden="true" tabindex="-1"></a>ignore this result.</span>
<span id="cb15-492"><a href="#cb15-492" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-493"><a href="#cb15-493" aria-hidden="true" tabindex="-1"></a>With the <span class="in">`FOVSignalLossSpatialPlot`</span> function, we can gain a higher resolution </span>
<span id="cb15-494"><a href="#cb15-494" aria-hidden="true" tabindex="-1"></a>perspective by plotting the 7x7 sub-FOV grids across space and coloring these</span>
<span id="cb15-495"><a href="#cb15-495" aria-hidden="true" tabindex="-1"></a>sub-FOV grids by the change in our expected total counts. In @fig-qc-fov-3, our</span>
<span id="cb15-496"><a href="#cb15-496" aria-hidden="true" tabindex="-1"></a>focus is on FOVs that are solidly blue which would suggestion fewer transcripts</span>
<span id="cb15-497"><a href="#cb15-497" aria-hidden="true" tabindex="-1"></a>than other FOVs. In this example, eight of the FOVs flagged as low total counts</span>
<span id="cb15-498"><a href="#cb15-498" aria-hidden="true" tabindex="-1"></a>occured along the tissue border and were not complete 7x7 grids. The other section </span>
<span id="cb15-499"><a href="#cb15-499" aria-hidden="true" tabindex="-1"></a>in @fig-qc-fov-3 is contiguous and may</span>
<span id="cb15-500"><a href="#cb15-500" aria-hidden="true" tabindex="-1"></a>suggest a domain or type of tissue with lower overall transcripts.</span>
<span id="cb15-501"><a href="#cb15-501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-502"><a href="#cb15-502" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-505"><a href="#cb15-505" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-506"><a href="#cb15-506" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-signal-loss</span></span>
<span id="cb15-507"><a href="#cb15-507" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-508"><a href="#cb15-508" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-509"><a href="#cb15-509" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-510"><a href="#cb15-510" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-511"><a href="#cb15-511" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-512"><a href="#cb15-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-513"><a href="#cb15-513" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">file.path</span>(qc_dir, <span class="st">"signal_loss_plot.png"</span>), </span>
<span id="cb15-514"><a href="#cb15-514" aria-hidden="true" tabindex="-1"></a>    <span class="at">width=</span><span class="dv">10</span>, <span class="at">height=</span><span class="dv">10</span>, <span class="at">res=</span><span class="dv">350</span>, <span class="at">units =</span> <span class="st">"in"</span>, <span class="at">type=</span><span class="st">"cairo"</span>)</span>
<span id="cb15-515"><a href="#cb15-515" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FOVSignalLossSpatialPlot</span>(res, <span class="at">shownames =</span> <span class="cn">TRUE</span>) </span>
<span id="cb15-516"><a href="#cb15-516" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb15-517"><a href="#cb15-517" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-518"><a href="#cb15-518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-521"><a href="#cb15-521" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-522"><a href="#cb15-522" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-qc-fov-3</span></span>
<span id="cb15-523"><a href="#cb15-523" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-524"><a href="#cb15-524" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-525"><a href="#cb15-525" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-526"><a href="#cb15-526" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 10</span></span>
<span id="cb15-527"><a href="#cb15-527" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 10</span></span>
<span id="cb15-528"><a href="#cb15-528" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Number of genes impacted in a given number of FOVs."</span></span>
<span id="cb15-529"><a href="#cb15-529" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-530"><a href="#cb15-530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-531"><a href="#cb15-531" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>),<span class="st">"signal_loss_plot.png"</span>, qc_dir)</span>
<span id="cb15-532"><a href="#cb15-532" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-533"><a href="#cb15-533" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-534"><a href="#cb15-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-535"><a href="#cb15-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-536"><a href="#cb15-536" aria-hidden="true" tabindex="-1"></a>To see details how these FOVs might have been impacted, we can use the function</span>
<span id="cb15-537"><a href="#cb15-537" aria-hidden="true" tabindex="-1"></a>below to plot the signal lost</span>
<span id="cb15-538"><a href="#cb15-538" aria-hidden="true" tabindex="-1"></a>found for each cycle-by-reporter combination.</span>
<span id="cb15-539"><a href="#cb15-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-540"><a href="#cb15-540" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-543"><a href="#cb15-543" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-544"><a href="#cb15-544" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-fov-effects</span></span>
<span id="cb15-545"><a href="#cb15-545" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-546"><a href="#cb15-546" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-547"><a href="#cb15-547" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-548"><a href="#cb15-548" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-549"><a href="#cb15-549" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb15-550"><a href="#cb15-550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-551"><a href="#cb15-551" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb15-552"><a href="#cb15-552" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb15-553"><a href="#cb15-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-554"><a href="#cb15-554" aria-hidden="true" tabindex="-1"></a>FOVEffectsHeatmap <span class="ot">&lt;-</span> <span class="cf">function</span>(res, <span class="at">fovs=</span><span class="cn">NULL</span>,</span>
<span id="cb15-555"><a href="#cb15-555" aria-hidden="true" tabindex="-1"></a>                              <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>, </span>
<span id="cb15-556"><a href="#cb15-556" aria-hidden="true" tabindex="-1"></a>                              <span class="at">cluster_columns =</span> <span class="cn">TRUE</span>) {</span>
<span id="cb15-557"><a href="#cb15-557" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">is.null</span>(fovs)){</span>
<span id="cb15-558"><a href="#cb15-558" aria-hidden="true" tabindex="-1"></a>    fovs <span class="ot">&lt;-</span> <span class="fu">rownames</span>(res<span class="sc">$</span>fovstats<span class="sc">$</span>bias)</span>
<span id="cb15-559"><a href="#cb15-559" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb15-560"><a href="#cb15-560" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-561"><a href="#cb15-561" aria-hidden="true" tabindex="-1"></a>  transposed_matrix <span class="ot">&lt;-</span> <span class="fu">t</span>(res<span class="sc">$</span>fovstats<span class="sc">$</span>bias[fovs,])</span>
<span id="cb15-562"><a href="#cb15-562" aria-hidden="true" tabindex="-1"></a>  col_fun <span class="ot">=</span> <span class="fu">colorRamp2</span>(</span>
<span id="cb15-563"><a href="#cb15-563" aria-hidden="true" tabindex="-1"></a>    <span class="at">breaks =</span> <span class="fu">c</span>(<span class="fu">min</span>(transposed_matrix), <span class="dv">0</span>, <span class="fu">max</span>(transposed_matrix)), </span>
<span id="cb15-564"><a href="#cb15-564" aria-hidden="true" tabindex="-1"></a>    <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"magenta"</span>, <span class="st">"white"</span>, <span class="st">"cyan"</span>) </span>
<span id="cb15-565"><a href="#cb15-565" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-566"><a href="#cb15-566" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-567"><a href="#cb15-567" aria-hidden="true" tabindex="-1"></a>  ht_obj <span class="ot">&lt;-</span> ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb15-568"><a href="#cb15-568" aria-hidden="true" tabindex="-1"></a>      <span class="at">matrix =</span> transposed_matrix,</span>
<span id="cb15-569"><a href="#cb15-569" aria-hidden="true" tabindex="-1"></a>      <span class="at">name =</span> <span class="st">"FOV bias"</span>, </span>
<span id="cb15-570"><a href="#cb15-570" aria-hidden="true" tabindex="-1"></a>      <span class="at">col =</span> col_fun, </span>
<span id="cb15-571"><a href="#cb15-571" aria-hidden="true" tabindex="-1"></a>      <span class="at">cluster_rows =</span> cluster_rows, </span>
<span id="cb15-572"><a href="#cb15-572" aria-hidden="true" tabindex="-1"></a>      <span class="at">cluster_columns =</span> cluster_columns, </span>
<span id="cb15-573"><a href="#cb15-573" aria-hidden="true" tabindex="-1"></a>      <span class="at">show_row_dend =</span> <span class="cn">TRUE</span>, </span>
<span id="cb15-574"><a href="#cb15-574" aria-hidden="true" tabindex="-1"></a>      <span class="at">show_column_dend =</span> <span class="cn">TRUE</span>, </span>
<span id="cb15-575"><a href="#cb15-575" aria-hidden="true" tabindex="-1"></a>      <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">7</span>), </span>
<span id="cb15-576"><a href="#cb15-576" aria-hidden="true" tabindex="-1"></a>      <span class="at">column_names_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">8</span>)</span>
<span id="cb15-577"><a href="#cb15-577" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-578"><a href="#cb15-578" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-579"><a href="#cb15-579" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(ht_obj)</span>
<span id="cb15-580"><a href="#cb15-580" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-581"><a href="#cb15-581" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-582"><a href="#cb15-582" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">file.path</span>(qc_dir, <span class="st">"FOVEffectsHeatmap.png"</span>), </span>
<span id="cb15-583"><a href="#cb15-583" aria-hidden="true" tabindex="-1"></a>    <span class="at">width=</span><span class="dv">10</span>, <span class="at">height=</span><span class="dv">15</span>, <span class="at">res=</span><span class="dv">350</span>, <span class="at">units =</span> <span class="st">"in"</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span>
<span id="cb15-584"><a href="#cb15-584" aria-hidden="true" tabindex="-1"></a>  <span class="fu">draw</span>(<span class="fu">FOVEffectsHeatmap</span>(res, <span class="at">fovs=</span>res<span class="sc">$</span>flaggedfovs))</span>
<span id="cb15-585"><a href="#cb15-585" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb15-586"><a href="#cb15-586" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-587"><a href="#cb15-587" aria-hidden="true" tabindex="-1"></a>results_list<span class="sc">$</span>flaggedfovs <span class="ot">&lt;-</span> res<span class="sc">$</span>flaggedfovs</span>
<span id="cb15-588"><a href="#cb15-588" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb15-589"><a href="#cb15-589" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-590"><a href="#cb15-590" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-591"><a href="#cb15-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-592"><a href="#cb15-592" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-595"><a href="#cb15-595" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-596"><a href="#cb15-596" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-qc-fov-4</span></span>
<span id="cb15-597"><a href="#cb15-597" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-598"><a href="#cb15-598" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-599"><a href="#cb15-599" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-600"><a href="#cb15-600" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 10</span></span>
<span id="cb15-601"><a href="#cb15-601" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 15</span></span>
<span id="cb15-602"><a href="#cb15-602" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Reporter-by-FOV combinations that were flagged."</span></span>
<span id="cb15-603"><a href="#cb15-603" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-604"><a href="#cb15-604" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-605"><a href="#cb15-605" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>),<span class="st">"FOVEffectsHeatmap.png"</span>, qc_dir)</span>
<span id="cb15-606"><a href="#cb15-606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-607"><a href="#cb15-607" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-608"><a href="#cb15-608" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-609"><a href="#cb15-609" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-610"><a href="#cb15-610" aria-hidden="true" tabindex="-1"></a>What we are checking for in @fig-qc-fov-4 is: </span>
<span id="cb15-611"><a href="#cb15-611" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-612"><a href="#cb15-612" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Any report cycles that show systematically reduced signal across many FOVs</span>
<span id="cb15-613"><a href="#cb15-613" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Any FOVs that show underperforming multiple reporter cycles. </span>
<span id="cb15-614"><a href="#cb15-614" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-615"><a href="#cb15-615" aria-hidden="true" tabindex="-1"></a>While we do see two reporter cycles (19 and 27) with two flagged FOVs (155 and 239),</span>
<span id="cb15-616"><a href="#cb15-616" aria-hidden="true" tabindex="-1"></a>each, there's no evidence of a systemic signal lost in any reporter cycle. If we did find</span>
<span id="cb15-617"><a href="#cb15-617" aria-hidden="true" tabindex="-1"></a>strong evidence, one could conservatively filter out the genes that are present</span>
<span id="cb15-618"><a href="#cb15-618" aria-hidden="true" tabindex="-1"></a>in that reporter cycle. This would be roughly 2,000 genes for a given reporter cycle.</span>
<span id="cb15-619"><a href="#cb15-619" aria-hidden="true" tabindex="-1"></a>Looking from the perspective of the FOVs (#2), we see only a handful </span>
<span id="cb15-620"><a href="#cb15-620" aria-hidden="true" tabindex="-1"></a>of FOVs showed one (or two) underperforming</span>
<span id="cb15-621"><a href="#cb15-621" aria-hidden="true" tabindex="-1"></a>reporter cycles. These include FOVs 92, 113, 155, 160, and 239. At this point,</span>
<span id="cb15-622"><a href="#cb15-622" aria-hidden="true" tabindex="-1"></a>one could remove all FOVs that were flagged (<span class="in">`length(res$flaggedfovs)`</span>; </span>
<span id="cb15-623"><a href="#cb15-623" aria-hidden="true" tabindex="-1"></a><span class="in">`r length(results_list$flaggedfovs)`</span> in this case), just the five that we found</span>
<span id="cb15-624"><a href="#cb15-624" aria-hidden="true" tabindex="-1"></a>evidence of signal lost in one or more reporter cycles, or flag the cells within the impacted</span>
<span id="cb15-625"><a href="#cb15-625" aria-hidden="true" tabindex="-1"></a>FOVs and see if any downstream results are impacted. </span>
<span id="cb15-626"><a href="#cb15-626" aria-hidden="true" tabindex="-1"></a>If you wanted to be on the conservative</span>
<span id="cb15-627"><a href="#cb15-627" aria-hidden="true" tabindex="-1"></a>side, you would remove all flagged FOVs unless you have biological justification</span>
<span id="cb15-628"><a href="#cb15-628" aria-hidden="true" tabindex="-1"></a>that the signal difference isn't due to FOV effects. </span>
<span id="cb15-629"><a href="#cb15-629" aria-hidden="true" tabindex="-1"></a>For this analysis, I'll </span>
<span id="cb15-630"><a href="#cb15-630" aria-hidden="true" tabindex="-1"></a>simply do the "flag and monitor" approach. </span>
<span id="cb15-631"><a href="#cb15-631" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-632"><a href="#cb15-632" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-635"><a href="#cb15-635" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-636"><a href="#cb15-636" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fov-qc-flag-impacted-cells</span></span>
<span id="cb15-637"><a href="#cb15-637" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb15-638"><a href="#cb15-638" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-639"><a href="#cb15-639" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-640"><a href="#cb15-640" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb15-641"><a href="#cb15-641" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-642"><a href="#cb15-642" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs</span>
<span id="cb15-643"><a href="#cb15-643" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>fovqcflag <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(meta<span class="sc">$</span>fov <span class="sc">%in%</span> <span class="fu">as.numeric</span>(res<span class="sc">$</span>flaggedfovs), <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb15-644"><a href="#cb15-644" aria-hidden="true" tabindex="-1"></a>py<span class="sc">$</span>adata<span class="sc">$</span>obs<span class="sc">$</span>fovqcflag <span class="ot">&lt;-</span> meta<span class="sc">$</span>fovqcflag</span>
<span id="cb15-645"><a href="#cb15-645" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-646"><a href="#cb15-646" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-647"><a href="#cb15-647" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-648"><a href="#cb15-648" aria-hidden="true" tabindex="-1"></a><span class="fu">## Cell-level QC</span></span>
<span id="cb15-649"><a href="#cb15-649" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-650"><a href="#cb15-650" aria-hidden="true" tabindex="-1"></a>Where the previous section examined signal at the FOV level, this section examines</span>
<span id="cb15-651"><a href="#cb15-651" aria-hidden="true" tabindex="-1"></a>target counts at the cell level. The purpose of this filter is simple: to remove</span>
<span id="cb15-652"><a href="#cb15-652" aria-hidden="true" tabindex="-1"></a>_low signal outliers_ while keeping the _biology_. The cells in a given dataset comprise of a mix</span>
<span id="cb15-653"><a href="#cb15-653" aria-hidden="true" tabindex="-1"></a>of: </span>
<span id="cb15-654"><a href="#cb15-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-655"><a href="#cb15-655" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>viable, intact cells with high signal</span>
<span id="cb15-656"><a href="#cb15-656" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>damaged or dying cells. These may have leaky membranes and may have reduced cytoplasmic RNA</span>
<span id="cb15-657"><a href="#cb15-657" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>imperfect segmentation of cells -- especially around FOV borders</span>
<span id="cb15-658"><a href="#cb15-658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-659"><a href="#cb15-659" aria-hidden="true" tabindex="-1"></a>and so our aim is to filter out cells from the latter two scenarios. </span>
<span id="cb15-660"><a href="#cb15-660" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-661"><a href="#cb15-661" aria-hidden="true" tabindex="-1"></a>Counts-based filtering has a trade-off. While several downstream "tertiary" analyses</span>
<span id="cb15-662"><a href="#cb15-662" aria-hidden="true" tabindex="-1"></a>account for low signal in a variety of ways (discussed later), many of the pre-processing</span>
<span id="cb15-663"><a href="#cb15-663" aria-hidden="true" tabindex="-1"></a>lenient cutoff can introduce low-quality cells in dimensional reductions like PCA and UMAP </span>
<span id="cb15-664"><a href="#cb15-664" aria-hidden="true" tabindex="-1"></a>and can break up clustering patterns (creating spurious clusters or merging</span>
<span id="cb15-665"><a href="#cb15-665" aria-hidden="true" tabindex="-1"></a>otherwise distinct clusters). On the other hand, a stringent cutoff can introduce</span>
<span id="cb15-666"><a href="#cb15-666" aria-hidden="true" tabindex="-1"></a>unexpected signal lost. This is because cell types vary in their transcript counts so </span>
<span id="cb15-667"><a href="#cb15-667" aria-hidden="true" tabindex="-1"></a>removing cells based on counts will distort the composition of cells in your data. </span>
<span id="cb15-668"><a href="#cb15-668" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-669"><a href="#cb15-669" aria-hidden="true" tabindex="-1"></a>Best practice for counts based filtering of CosMx data includes:</span>
<span id="cb15-670"><a href="#cb15-670" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-671"><a href="#cb15-671" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>examining the distribution of transcript and gene counts</span>
<span id="cb15-672"><a href="#cb15-672" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>choosing an initial permissive or lenient cutoff</span>
<span id="cb15-673"><a href="#cb15-673" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>examining signal relative to background</span>
<span id="cb15-674"><a href="#cb15-674" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>plotting the spatial distribution of flagged cells in space</span>
<span id="cb15-675"><a href="#cb15-675" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>removing small cell artifacts around the FOV borders</span>
<span id="cb15-676"><a href="#cb15-676" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>including plots of total transcripts in your downstream pre-processing steps</span>
<span id="cb15-677"><a href="#cb15-677" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-678"><a href="#cb15-678" aria-hidden="true" tabindex="-1"></a>::: {.column-margin}</span>
<span id="cb15-679"><a href="#cb15-679" aria-hidden="true" tabindex="-1"></a>::: {.otherapproachesbox title="Alternative Approach"}</span>
<span id="cb15-680"><a href="#cb15-680" aria-hidden="true" tabindex="-1"></a>There are many cell-level attributes that one can use as a bases for filtering and</span>
<span id="cb15-681"><a href="#cb15-681" aria-hidden="true" tabindex="-1"></a>it is recommended to use metrics relevant to your tissue. For example, if you are</span>
<span id="cb15-682"><a href="#cb15-682" aria-hidden="true" tabindex="-1"></a>analyzing brain, you many want to incorporate or evaluate cell area, eccentricity, or</span>
<span id="cb15-683"><a href="#cb15-683" aria-hidden="true" tabindex="-1"></a>other morphological characteristics that are available in the metadata. For a list</span>
<span id="cb15-684"><a href="#cb15-684" aria-hidden="true" tabindex="-1"></a>of cell metadata column definitions, you can see </span>
<span id="cb15-685"><a href="#cb15-685" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">this Scratch Space blog post</span><span class="co">](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/flat-file-exports/flat-files-compare.html#sec-meta)</span></span>
<span id="cb15-686"><a href="#cb15-686" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb15-687"><a href="#cb15-687" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb15-688"><a href="#cb15-688" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-689"><a href="#cb15-689" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-690"><a href="#cb15-690" aria-hidden="true" tabindex="-1"></a><span class="fu">### Counts and Features</span></span>
<span id="cb15-691"><a href="#cb15-691" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-692"><a href="#cb15-692" aria-hidden="true" tabindex="-1"></a>For the distribution of transcript and gene counts, I like to look at the joint</span>
<span id="cb15-693"><a href="#cb15-693" aria-hidden="true" tabindex="-1"></a>distribution in addition to the univariate distributions. In the code below</span>
<span id="cb15-694"><a href="#cb15-694" aria-hidden="true" tabindex="-1"></a>I set a permissive, symmetrical threshold of 2.5 standard deviations (SDs) around</span>
<span id="cb15-695"><a href="#cb15-695" aria-hidden="true" tabindex="-1"></a>the mean (in log10 space) but this thresholding is sample-dependent so different</span>
<span id="cb15-696"><a href="#cb15-696" aria-hidden="true" tabindex="-1"></a>values might make more sense on your data.</span>
<span id="cb15-697"><a href="#cb15-697" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-698"><a href="#cb15-698" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-701"><a href="#cb15-701" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-702"><a href="#cb15-702" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: count-based-filtering</span></span>
<span id="cb15-703"><a href="#cb15-703" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-704"><a href="#cb15-704" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-705"><a href="#cb15-705" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-706"><a href="#cb15-706" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-707"><a href="#cb15-707" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-708"><a href="#cb15-708" aria-hidden="true" tabindex="-1"></a><span class="co"># meta &lt;- py$adata$obs</span></span>
<span id="cb15-709"><a href="#cb15-709" aria-hidden="true" tabindex="-1"></a>counts_min_threshold <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">10</span><span class="sc">^</span>(<span class="fu">mean</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nCount_RNA)) <span class="sc">-</span> <span class="fl">2.5</span><span class="sc">*</span><span class="fu">sd</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nCount_RNA))))</span>
<span id="cb15-710"><a href="#cb15-710" aria-hidden="true" tabindex="-1"></a>features_min_threshold <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">10</span><span class="sc">^</span>(<span class="fu">mean</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nFeature_RNA)) <span class="sc">-</span> <span class="fl">2.5</span><span class="sc">*</span><span class="fu">sd</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nFeature_RNA))))</span>
<span id="cb15-711"><a href="#cb15-711" aria-hidden="true" tabindex="-1"></a>counts_max_threshold <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">10</span><span class="sc">^</span>(<span class="fu">mean</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nCount_RNA)) <span class="sc">+</span> <span class="fl">2.5</span><span class="sc">*</span><span class="fu">sd</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nCount_RNA))))</span>
<span id="cb15-712"><a href="#cb15-712" aria-hidden="true" tabindex="-1"></a>features_max_threshold <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">10</span><span class="sc">^</span>(<span class="fu">mean</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nFeature_RNA)) <span class="sc">+</span> <span class="fl">2.5</span><span class="sc">*</span><span class="fu">sd</span>(<span class="fu">log10</span>(meta<span class="sc">$</span>nFeature_RNA))))</span>
<span id="cb15-713"><a href="#cb15-713" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-714"><a href="#cb15-714" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'counts_min_threshold'</span>] <span class="ot">=</span> counts_min_threshold</span>
<span id="cb15-715"><a href="#cb15-715" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'features_min_threshold'</span>] <span class="ot">=</span> features_min_threshold</span>
<span id="cb15-716"><a href="#cb15-716" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'counts_max_threshold'</span>] <span class="ot">=</span> counts_max_threshold</span>
<span id="cb15-717"><a href="#cb15-717" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'features_max_threshold'</span>] <span class="ot">=</span> features_max_threshold</span>
<span id="cb15-718"><a href="#cb15-718" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-719"><a href="#cb15-719" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-720"><a href="#cb15-720" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>counts_filter <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(</span>
<span id="cb15-721"><a href="#cb15-721" aria-hidden="true" tabindex="-1"></a>  meta<span class="sc">$</span>nFeature_RNA <span class="sc">&gt;=</span> features_min_threshold <span class="sc">&amp;</span> </span>
<span id="cb15-722"><a href="#cb15-722" aria-hidden="true" tabindex="-1"></a>         meta<span class="sc">$</span>nFeature_RNA <span class="sc">&lt;=</span> features_max_threshold <span class="sc">&amp;</span> </span>
<span id="cb15-723"><a href="#cb15-723" aria-hidden="true" tabindex="-1"></a>         meta<span class="sc">$</span>nCount_RNA <span class="sc">&gt;=</span> counts_min_threshold <span class="sc">&amp;</span> </span>
<span id="cb15-724"><a href="#cb15-724" aria-hidden="true" tabindex="-1"></a>         meta<span class="sc">$</span>nCount_RNA <span class="sc">&lt;=</span> counts_max_threshold</span>
<span id="cb15-725"><a href="#cb15-725" aria-hidden="true" tabindex="-1"></a>  , <span class="st">"pass"</span>, <span class="st">"filtered"</span>)</span>
<span id="cb15-726"><a href="#cb15-726" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-727"><a href="#cb15-727" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'prop_counts_based_filter'</span>] <span class="ot">&lt;-</span> <span class="fu">sum</span>(meta<span class="sc">$</span>counts_filter<span class="sc">==</span><span class="st">"pass"</span>) <span class="sc">/</span> <span class="fu">nrow</span>(meta)</span>
<span id="cb15-728"><a href="#cb15-728" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb15-729"><a href="#cb15-729" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-730"><a href="#cb15-730" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-731"><a href="#cb15-731" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb15-732"><a href="#cb15-732" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb15-733"><a href="#cb15-733" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> meta,</span>
<span id="cb15-734"><a href="#cb15-734" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">x=</span>nFeature_RNA, <span class="at">y=</span>nCount_RNA, <span class="at">color=</span>counts_filter)) <span class="sc">+</span> </span>
<span id="cb15-735"><a href="#cb15-735" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.002</span>, <span class="at">pch=</span><span class="st">"."</span>) <span class="sc">+</span></span>
<span id="cb15-736"><a href="#cb15-736" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"pass"</span><span class="ot">=</span><span class="st">"dodgerblue"</span>, <span class="st">"filtered"</span><span class="ot">=</span><span class="st">"darkorange"</span>)) <span class="sc">+</span></span>
<span id="cb15-737"><a href="#cb15-737" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(features_min_threshold, features_max_threshold), <span class="at">lty=</span><span class="st">"dotted"</span>) <span class="sc">+</span> </span>
<span id="cb15-738"><a href="#cb15-738" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">c</span>(counts_min_threshold, counts_max_threshold), <span class="at">lty=</span><span class="st">"dotted"</span>) <span class="sc">+</span> </span>
<span id="cb15-739"><a href="#cb15-739" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>(</span>
<span id="cb15-740"><a href="#cb15-740" aria-hidden="true" tabindex="-1"></a>    <span class="at">breaks =</span> <span class="fu">trans_breaks</span>(<span class="st">"log10"</span>, <span class="cf">function</span>(x) <span class="dv">10</span><span class="sc">^</span>x),</span>
<span id="cb15-741"><a href="#cb15-741" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> <span class="fu">trans_format</span>(<span class="st">"log10"</span>, <span class="fu">math_format</span>(<span class="dv">10</span><span class="sc">^</span>.x))</span>
<span id="cb15-742"><a href="#cb15-742" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb15-743"><a href="#cb15-743" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>(</span>
<span id="cb15-744"><a href="#cb15-744" aria-hidden="true" tabindex="-1"></a>    <span class="at">breaks =</span> <span class="fu">trans_breaks</span>(<span class="st">"log10"</span>, <span class="cf">function</span>(x) <span class="dv">10</span><span class="sc">^</span>x),</span>
<span id="cb15-745"><a href="#cb15-745" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> <span class="fu">trans_format</span>(<span class="st">"log10"</span>, <span class="fu">math_format</span>(<span class="dv">10</span><span class="sc">^</span>.x))</span>
<span id="cb15-746"><a href="#cb15-746" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> </span>
<span id="cb15-747"><a href="#cb15-747" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb15-748"><a href="#cb15-748" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>() <span class="sc">+</span> </span>
<span id="cb15-749"><a href="#cb15-749" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb15-750"><a href="#cb15-750" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-751"><a href="#cb15-751" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggMarginal</span>(p, <span class="at">type =</span> <span class="st">"density"</span>)</span>
<span id="cb15-752"><a href="#cb15-752" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-753"><a href="#cb15-753" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_joint_distribution.png"</span>), p,</span>
<span id="cb15-754"><a href="#cb15-754" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">8</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span>
<span id="cb15-755"><a href="#cb15-755" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-756"><a href="#cb15-756" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb15-757"><a href="#cb15-757" aria-hidden="true" tabindex="-1"></a>      <span class="at">data=</span>meta, </span>
<span id="cb15-758"><a href="#cb15-758" aria-hidden="true" tabindex="-1"></a>      <span class="fu">aes</span>(<span class="at">x=</span>nCount_RNA)) <span class="sc">+</span> </span>
<span id="cb15-759"><a href="#cb15-759" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_density</span>(<span class="at">alpha=</span><span class="fl">0.01</span>) <span class="sc">+</span> </span>
<span id="cb15-760"><a href="#cb15-760" aria-hidden="true" tabindex="-1"></a>      <span class="co"># scale_x_log10() +</span></span>
<span id="cb15-761"><a href="#cb15-761" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb15-762"><a href="#cb15-762" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(counts_min_threshold, counts_max_threshold), <span class="at">lty=</span><span class="st">"dotted"</span>) </span>
<span id="cb15-763"><a href="#cb15-763" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-764"><a href="#cb15-764" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_dist_counts.png"</span>), p1,</span>
<span id="cb15-765"><a href="#cb15-765" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">8</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span>
<span id="cb15-766"><a href="#cb15-766" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-767"><a href="#cb15-767" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb15-768"><a href="#cb15-768" aria-hidden="true" tabindex="-1"></a>      <span class="at">data=</span>meta, </span>
<span id="cb15-769"><a href="#cb15-769" aria-hidden="true" tabindex="-1"></a>      <span class="fu">aes</span>(<span class="at">x=</span>nFeature_RNA)) <span class="sc">+</span> </span>
<span id="cb15-770"><a href="#cb15-770" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_density</span>(<span class="at">alpha=</span><span class="fl">0.01</span>) <span class="sc">+</span> </span>
<span id="cb15-771"><a href="#cb15-771" aria-hidden="true" tabindex="-1"></a>      <span class="co"># scale_x_log10() +</span></span>
<span id="cb15-772"><a href="#cb15-772" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb15-773"><a href="#cb15-773" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(features_min_threshold, features_max_threshold), <span class="at">lty=</span><span class="st">"dotted"</span>) </span>
<span id="cb15-774"><a href="#cb15-774" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-775"><a href="#cb15-775" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_dist_features.png"</span>), p2,</span>
<span id="cb15-776"><a href="#cb15-776" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">8</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span>
<span id="cb15-777"><a href="#cb15-777" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-778"><a href="#cb15-778" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb15-779"><a href="#cb15-779" aria-hidden="true" tabindex="-1"></a>        <span class="at">data=</span>meta) <span class="sc">+</span></span>
<span id="cb15-780"><a href="#cb15-780" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_point</span>(</span>
<span id="cb15-781"><a href="#cb15-781" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x=</span>x_slide_mm, <span class="at">y=</span>y_slide_mm, <span class="at">color=</span>counts_filter),</span>
<span id="cb15-782"><a href="#cb15-782" aria-hidden="true" tabindex="-1"></a>        <span class="at">size=</span><span class="fl">0.001</span>, <span class="at">alpha=</span><span class="fl">0.02</span>) <span class="sc">+</span> </span>
<span id="cb15-783"><a href="#cb15-783" aria-hidden="true" tabindex="-1"></a>      <span class="fu">coord_fixed</span>() <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb15-784"><a href="#cb15-784" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"pass"</span><span class="ot">=</span><span class="st">"dodgerblue"</span>, <span class="st">"filtered"</span><span class="ot">=</span><span class="st">"darkorange"</span>)) <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>counts_filter) <span class="sc">+</span> </span>
<span id="cb15-785"><a href="#cb15-785" aria-hidden="true" tabindex="-1"></a>      <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">7</span>, <span class="at">alpha =</span> <span class="dv">1</span>)))</span>
<span id="cb15-786"><a href="#cb15-786" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-787"><a href="#cb15-787" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_xy_flagged.png"</span>), p3,</span>
<span id="cb15-788"><a href="#cb15-788" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">8</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span>
<span id="cb15-789"><a href="#cb15-789" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb15-790"><a href="#cb15-790" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-791"><a href="#cb15-791" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-792"><a href="#cb15-792" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-793"><a href="#cb15-793" aria-hidden="true" tabindex="-1"></a>The tabs below show the summary plots. Based on the distributions, the following thresholds we derived from the data: <span class="in">`r paste0('min counts = ', results_list$counts_min_threshold, ', max counts = ', results_list$counts_max_threshold, ', min features = ', results_list$features_min_threshold, ', max features = ', results_list$features_max_threshold)`</span>. This removes about <span class="in">`r round(100*(1 - results_list$prop_counts_based_filter))`</span>% of the data. The spatial layout of these flagged</span>
<span id="cb15-794"><a href="#cb15-794" aria-hidden="true" tabindex="-1"></a>cells shows some concentration along the perimeter and epithelial regions. </span>
<span id="cb15-795"><a href="#cb15-795" aria-hidden="true" tabindex="-1"></a>No concentration within FOVs were observed (@fig-qc-xy-flagged).</span>
<span id="cb15-796"><a href="#cb15-796" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-797"><a href="#cb15-797" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset}</span>
<span id="cb15-798"><a href="#cb15-798" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-799"><a href="#cb15-799" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Transcripts per cell</span></span>
<span id="cb15-800"><a href="#cb15-800" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-803"><a href="#cb15-803" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-804"><a href="#cb15-804" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-qc-dist-tx-per-cell</span></span>
<span id="cb15-805"><a href="#cb15-805" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-806"><a href="#cb15-806" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-807"><a href="#cb15-807" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Distribution of transcripts per cell."</span></span>
<span id="cb15-808"><a href="#cb15-808" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 4</span></span>
<span id="cb15-809"><a href="#cb15-809" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb15-810"><a href="#cb15-810" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-811"><a href="#cb15-811" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>), <span class="st">"qc_dist_counts.png"</span>, qc_dir)</span>
<span id="cb15-812"><a href="#cb15-812" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-813"><a href="#cb15-813" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-814"><a href="#cb15-814" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-815"><a href="#cb15-815" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Features per cell</span></span>
<span id="cb15-816"><a href="#cb15-816" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-819"><a href="#cb15-819" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-820"><a href="#cb15-820" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-qc-dist-featrues-per-cell</span></span>
<span id="cb15-821"><a href="#cb15-821" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-822"><a href="#cb15-822" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-823"><a href="#cb15-823" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-824"><a href="#cb15-824" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Distribution of features per cell."</span></span>
<span id="cb15-825"><a href="#cb15-825" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 4</span></span>
<span id="cb15-826"><a href="#cb15-826" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb15-827"><a href="#cb15-827" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-828"><a href="#cb15-828" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>), <span class="st">"qc_dist_features.png"</span>, qc_dir)</span>
<span id="cb15-829"><a href="#cb15-829" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-830"><a href="#cb15-830" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-831"><a href="#cb15-831" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-832"><a href="#cb15-832" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Joint Distribution</span></span>
<span id="cb15-833"><a href="#cb15-833" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-836"><a href="#cb15-836" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-837"><a href="#cb15-837" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-qc-joint</span></span>
<span id="cb15-838"><a href="#cb15-838" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-839"><a href="#cb15-839" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-840"><a href="#cb15-840" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-841"><a href="#cb15-841" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Joint distribution of targets (features vs counts)."</span></span>
<span id="cb15-842"><a href="#cb15-842" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 4</span></span>
<span id="cb15-843"><a href="#cb15-843" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb15-844"><a href="#cb15-844" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-845"><a href="#cb15-845" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>), <span class="st">"qc_joint_distribution.png"</span>, qc_dir)</span>
<span id="cb15-846"><a href="#cb15-846" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-847"><a href="#cb15-847" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-848"><a href="#cb15-848" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-849"><a href="#cb15-849" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Spatial Arrangement of flagged cells</span></span>
<span id="cb15-850"><a href="#cb15-850" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-853"><a href="#cb15-853" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-854"><a href="#cb15-854" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-qc-xy-flagged</span></span>
<span id="cb15-855"><a href="#cb15-855" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-856"><a href="#cb15-856" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-857"><a href="#cb15-857" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-858"><a href="#cb15-858" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Spatial arrangement of flagged cells."</span></span>
<span id="cb15-859"><a href="#cb15-859" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-860"><a href="#cb15-860" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>), <span class="st">"qc_xy_flagged.png"</span>, <span class="fu">file.path</span>(qc_dir))</span>
<span id="cb15-861"><a href="#cb15-861" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-862"><a href="#cb15-862" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-863"><a href="#cb15-863" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-864"><a href="#cb15-864" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb15-865"><a href="#cb15-865" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-866"><a href="#cb15-866" aria-hidden="true" tabindex="-1"></a><span class="fu">### (Optional) Cells Near FOV Borders</span></span>
<span id="cb15-867"><a href="#cb15-867" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-868"><a href="#cb15-868" aria-hidden="true" tabindex="-1"></a>In this particular dataset, 6.2% of the cells were located on the border of FOVs (_i.e._, </span>
<span id="cb15-869"><a href="#cb15-869" aria-hidden="true" tabindex="-1"></a>touching the FOV edge). This is a minority of cells that in the analysis as a whole shouldn't</span>
<span id="cb15-870"><a href="#cb15-870" aria-hidden="true" tabindex="-1"></a>affect downstream analyses. For this reason this step is optional. </span>
<span id="cb15-871"><a href="#cb15-871" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-872"><a href="#cb15-872" aria-hidden="true" tabindex="-1"></a>Within the metadata there is a column named <span class="in">`SplitRatioToLocal`</span>. For cells that </span>
<span id="cb15-873"><a href="#cb15-873" aria-hidden="true" tabindex="-1"></a>are adjacent to the FOV boundaries, the <span class="in">`SplitRatioToLocal`</span> metric measures the </span>
<span id="cb15-874"><a href="#cb15-874" aria-hidden="true" tabindex="-1"></a>cell area relative to the mean area of cells in the FOVs. For <span class="in">`0 &lt; SplitRatioToLocal &lt; 1`</span>, </span>
<span id="cb15-875"><a href="#cb15-875" aria-hidden="true" tabindex="-1"></a>the cell is smaller than average and for <span class="in">`SplitRatioToLocal &gt; 1`</span> the cell is </span>
<span id="cb15-876"><a href="#cb15-876" aria-hidden="true" tabindex="-1"></a>larger than average. A value of 0 means the cell is not along the border. Let's</span>
<span id="cb15-877"><a href="#cb15-877" aria-hidden="true" tabindex="-1"></a>plot the distribution of boarder abutting cells (_i.e._, <span class="in">`SplitRatioToLocal &gt; 0`</span>)</span>
<span id="cb15-878"><a href="#cb15-878" aria-hidden="true" tabindex="-1"></a>and tentatively choose a cutoff of SplitRatioToLocal = 0.5 which translates to</span>
<span id="cb15-879"><a href="#cb15-879" aria-hidden="true" tabindex="-1"></a>removing cells that are 0.5 (50%) the average cell size. Since this is a ratio,</span>
<span id="cb15-880"><a href="#cb15-880" aria-hidden="true" tabindex="-1"></a>we'll plot in log2 space (<span class="in">`log2(SplitRatioToLocal) &lt; -1`</span>).</span>
<span id="cb15-881"><a href="#cb15-881" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-884"><a href="#cb15-884" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-885"><a href="#cb15-885" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-886"><a href="#cb15-886" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-887"><a href="#cb15-887" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-888"><a href="#cb15-888" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-889"><a href="#cb15-889" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb15-890"><a href="#cb15-890" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-891"><a href="#cb15-891" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'SplitRatioToLocalThreshold'</span>] <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb15-892"><a href="#cb15-892" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb15-893"><a href="#cb15-893" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-894"><a href="#cb15-894" aria-hidden="true" tabindex="-1"></a><span class="co"># meta &lt;- py$adata$obs</span></span>
<span id="cb15-895"><a href="#cb15-895" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-896"><a href="#cb15-896" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb15-897"><a href="#cb15-897" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb15-898"><a href="#cb15-898" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-899"><a href="#cb15-899" aria-hidden="true" tabindex="-1"></a>p_detail <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb15-900"><a href="#cb15-900" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> meta <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SplitRatioToLocal <span class="sc">!=</span> <span class="dv">0</span>), </span>
<span id="cb15-901"><a href="#cb15-901" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log2</span>(SplitRatioToLocal))) <span class="sc">+</span> </span>
<span id="cb15-902"><a href="#cb15-902" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">100</span>, <span class="at">fill =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span></span>
<span id="cb15-903"><a href="#cb15-903" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb15-904"><a href="#cb15-904" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="fu">expression</span>(log[<span class="dv">2</span>](<span class="st">"SplitRatioToLocal"</span>)),</span>
<span id="cb15-905"><a href="#cb15-905" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Number of FOV border cells"</span></span>
<span id="cb15-906"><a href="#cb15-906" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb15-907"><a href="#cb15-907" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb15-908"><a href="#cb15-908" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">log2</span>(results_list[[<span class="st">'SplitRatioToLocalThreshold'</span>]]), </span>
<span id="cb15-909"><a href="#cb15-909" aria-hidden="true" tabindex="-1"></a>               <span class="at">color =</span> <span class="st">"darkorange"</span>, <span class="at">lty =</span> <span class="st">"dashed"</span>)</span>
<span id="cb15-910"><a href="#cb15-910" aria-hidden="true" tabindex="-1"></a>pie_data <span class="ot">&lt;-</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb15-911"><a href="#cb15-911" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Category =</span> <span class="fu">if_else</span>(SplitRatioToLocal <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="st">"Border cell"</span>, <span class="st">"Non-border cell"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb15-912"><a href="#cb15-912" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(Category) <span class="sc">%&gt;%</span></span>
<span id="cb15-913"><a href="#cb15-913" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb15-914"><a href="#cb15-914" aria-hidden="true" tabindex="-1"></a>    <span class="at">Percentage =</span> (n <span class="sc">/</span> <span class="fu">sum</span>(n)) <span class="sc">*</span> <span class="dv">100</span>,</span>
<span id="cb15-915"><a href="#cb15-915" aria-hidden="true" tabindex="-1"></a>    <span class="at">Label =</span> <span class="fu">paste0</span>(Category, <span class="st">" ("</span>, <span class="fu">round</span>(Percentage, <span class="dv">1</span>), <span class="st">"%)"</span>)</span>
<span id="cb15-916"><a href="#cb15-916" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-917"><a href="#cb15-917" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-918"><a href="#cb15-918" aria-hidden="true" tabindex="-1"></a>p_pie <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pie_data, <span class="fu">aes</span>(<span class="at">x =</span> <span class="dv">2</span>, <span class="at">y =</span> n, <span class="at">fill =</span> Category)) <span class="sc">+</span></span>
<span id="cb15-919"><a href="#cb15-919" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">width =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb15-920"><a href="#cb15-920" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_polar</span>(<span class="st">"y"</span>, <span class="at">start =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb15-921"><a href="#cb15-921" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Border cell"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>, <span class="st">"Non-border cell"</span> <span class="ot">=</span> <span class="st">"gray90"</span>)) <span class="sc">+</span></span>
<span id="cb15-922"><a href="#cb15-922" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> Label), </span>
<span id="cb15-923"><a href="#cb15-923" aria-hidden="true" tabindex="-1"></a>            <span class="at">position =</span> <span class="fu">position_stack</span>(<span class="at">vjust =</span> <span class="fl">0.5</span>), </span>
<span id="cb15-924"><a href="#cb15-924" aria-hidden="true" tabindex="-1"></a>            <span class="at">size =</span> <span class="dv">3</span>, </span>
<span id="cb15-925"><a href="#cb15-925" aria-hidden="true" tabindex="-1"></a>            <span class="at">fontface =</span> <span class="st">"bold"</span>) <span class="sc">+</span></span>
<span id="cb15-926"><a href="#cb15-926" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb15-927"><a href="#cb15-927" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fl">0.5</span>, <span class="fl">2.5</span>) <span class="sc">+</span> </span>
<span id="cb15-928"><a href="#cb15-928" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb15-929"><a href="#cb15-929" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-930"><a href="#cb15-930" aria-hidden="true" tabindex="-1"></a>full_plot <span class="ot">&lt;-</span> p_detail <span class="sc">+</span> </span>
<span id="cb15-931"><a href="#cb15-931" aria-hidden="true" tabindex="-1"></a>  <span class="fu">inset_element</span>(</span>
<span id="cb15-932"><a href="#cb15-932" aria-hidden="true" tabindex="-1"></a>    p_pie, </span>
<span id="cb15-933"><a href="#cb15-933" aria-hidden="true" tabindex="-1"></a>    <span class="at">left =</span> <span class="fl">0.6</span>, <span class="at">bottom =</span> <span class="fl">0.6</span>, </span>
<span id="cb15-934"><a href="#cb15-934" aria-hidden="true" tabindex="-1"></a>    <span class="at">right =</span> <span class="fl">1.0</span>,</span>
<span id="cb15-935"><a href="#cb15-935" aria-hidden="true" tabindex="-1"></a>    <span class="at">top =</span> <span class="fl">1.0</span></span>
<span id="cb15-936"><a href="#cb15-936" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-937"><a href="#cb15-937" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-938"><a href="#cb15-938" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_splitratiotolocal.png"</span>), full_plot,</span>
<span id="cb15-939"><a href="#cb15-939" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">5</span>, <span class="at">height=</span><span class="dv">5</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span>
<span id="cb15-940"><a href="#cb15-940" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-941"><a href="#cb15-941" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-944"><a href="#cb15-944" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-945"><a href="#cb15-945" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-qc-split</span></span>
<span id="cb15-946"><a href="#cb15-946" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-947"><a href="#cb15-947" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-948"><a href="#cb15-948" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Distribution of SplitRatioToLocal for cells abutting an FOV border. Overlay: proportion of border cells relative to all cells."</span></span>
<span id="cb15-949"><a href="#cb15-949" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 5</span></span>
<span id="cb15-950"><a href="#cb15-950" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb15-951"><a href="#cb15-951" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-952"><a href="#cb15-952" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-953"><a href="#cb15-953" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>), <span class="st">"qc_splitratiotolocal.png"</span>, qc_dir, <span class="at">overwrite =</span> T)</span>
<span id="cb15-954"><a href="#cb15-954" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-955"><a href="#cb15-955" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-956"><a href="#cb15-956" aria-hidden="true" tabindex="-1"></a><span class="fu">## Regional-level cell QC</span></span>
<span id="cb15-957"><a href="#cb15-957" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-958"><a href="#cb15-958" aria-hidden="true" tabindex="-1"></a>It's possible that some regions of the tissue might have high background. We can</span>
<span id="cb15-959"><a href="#cb15-959" aria-hidden="true" tabindex="-1"></a>evaluate the total signal relative to the background. For this approach we</span>
<span id="cb15-960"><a href="#cb15-960" aria-hidden="true" tabindex="-1"></a>are interested in the Signal to Background Ratio (SBR) in neighborhoods or tissue</span>
<span id="cb15-961"><a href="#cb15-961" aria-hidden="true" tabindex="-1"></a>regions as opposed to cell-level measures or target-level measures. Since the number of targets is not comparable to the</span>
<span id="cb15-962"><a href="#cb15-962" aria-hidden="true" tabindex="-1"></a>number of negative probes, we'll consider the average target expression (S) </span>
<span id="cb15-963"><a href="#cb15-963" aria-hidden="true" tabindex="-1"></a>divided by the average background expression. </span>
<span id="cb15-964"><a href="#cb15-964" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-965"><a href="#cb15-965" aria-hidden="true" tabindex="-1"></a>$$sbr = \frac{S}{\mu_{B}}$$</span>
<span id="cb15-966"><a href="#cb15-966" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-967"><a href="#cb15-967" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-968"><a href="#cb15-968" aria-hidden="true" tabindex="-1"></a>When passing the data through this equation, you'll likely find that many cells</span>
<span id="cb15-969"><a href="#cb15-969" aria-hidden="true" tabindex="-1"></a>have an undefined SBR due to the sparse nature of background counts. To mitigate</span>
<span id="cb15-970"><a href="#cb15-970" aria-hidden="true" tabindex="-1"></a>this effect, we can spatially smooth the transcript and background counts before applying this </span>
<span id="cb15-971"><a href="#cb15-971" aria-hidden="true" tabindex="-1"></a>equation. To do this we'll use the <span class="in">`liana`</span> package that creates a connectivities</span>
<span id="cb15-972"><a href="#cb15-972" aria-hidden="true" tabindex="-1"></a>matrix based on a Gaussian kerneland then multiple this matrix by the negative</span>
<span id="cb15-973"><a href="#cb15-973" aria-hidden="true" tabindex="-1"></a>probe matrix and the expression matrix to get the spatially-weighted matrices for each. </span>
<span id="cb15-974"><a href="#cb15-974" aria-hidden="true" tabindex="-1"></a>For more information on this spatial smoothing see @sec-pathways.</span>
<span id="cb15-975"><a href="#cb15-975" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-978"><a href="#cb15-978" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb15-979"><a href="#cb15-979" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: sbr-1</span></span>
<span id="cb15-980"><a href="#cb15-980" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-981"><a href="#cb15-981" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb15-982"><a href="#cb15-982" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-983"><a href="#cb15-983" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-984"><a href="#cb15-984" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb15-985"><a href="#cb15-985" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-986"><a href="#cb15-986" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> liana <span class="im">as</span> li</span>
<span id="cb15-987"><a href="#cb15-987" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-988"><a href="#cb15-988" aria-hidden="true" tabindex="-1"></a>li.ut.spatial_neighbors(</span>
<span id="cb15-989"><a href="#cb15-989" aria-hidden="true" tabindex="-1"></a>    adata,</span>
<span id="cb15-990"><a href="#cb15-990" aria-hidden="true" tabindex="-1"></a>    bandwidth<span class="op">=</span><span class="fl">0.010</span>, <span class="co"># &lt;1&gt;</span></span>
<span id="cb15-991"><a href="#cb15-991" aria-hidden="true" tabindex="-1"></a>    cutoff<span class="op">=</span><span class="fl">0.080</span>, <span class="co"># &lt;2&gt;</span></span>
<span id="cb15-992"><a href="#cb15-992" aria-hidden="true" tabindex="-1"></a>    kernel<span class="op">=</span><span class="st">'gaussian'</span>,</span>
<span id="cb15-993"><a href="#cb15-993" aria-hidden="true" tabindex="-1"></a>    set_diag<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb15-994"><a href="#cb15-994" aria-hidden="true" tabindex="-1"></a>    key_added <span class="op">=</span> <span class="st">'liana'</span>,</span>
<span id="cb15-995"><a href="#cb15-995" aria-hidden="true" tabindex="-1"></a>    standardize<span class="op">=</span><span class="va">True</span></span>
<span id="cb15-996"><a href="#cb15-996" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-997"><a href="#cb15-997" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-998"><a href="#cb15-998" aria-hidden="true" tabindex="-1"></a>connectivities <span class="op">=</span> adata.obsp[<span class="st">'liana_connectivities'</span>]</span>
<span id="cb15-999"><a href="#cb15-999" aria-hidden="true" tabindex="-1"></a>adata.obsm[<span class="st">'counts_neg_spatially_smooth'</span>] <span class="op">=</span> connectivities <span class="op">@</span> adata.obsm[<span class="st">'counts_neg'</span>]</span>
<span id="cb15-1000"><a href="#cb15-1000" aria-hidden="true" tabindex="-1"></a>adata.obsm[<span class="st">'counts_spatially_smooth'</span>] <span class="op">=</span> connectivities <span class="op">@</span> adata.X</span>
<span id="cb15-1001"><a href="#cb15-1001" aria-hidden="true" tabindex="-1"></a>smoothed_mean_neg <span class="op">=</span> np.ravel(adata.obsm[<span class="st">'counts_neg_spatially_smooth'</span>].mean(axis<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb15-1002"><a href="#cb15-1002" aria-hidden="true" tabindex="-1"></a>smoothed_mean_target <span class="op">=</span> np.ravel(adata.obsm[<span class="st">'counts_spatially_smooth'</span>].mean(axis<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb15-1003"><a href="#cb15-1003" aria-hidden="true" tabindex="-1"></a>epsilon <span class="op">=</span> <span class="fl">1e-9</span></span>
<span id="cb15-1004"><a href="#cb15-1004" aria-hidden="true" tabindex="-1"></a>SBR <span class="op">=</span> smoothed_mean_target <span class="op">/</span> (smoothed_mean_neg <span class="op">+</span> epsilon)</span>
<span id="cb15-1005"><a href="#cb15-1005" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1006"><a href="#cb15-1006" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-1007"><a href="#cb15-1007" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>use 10 µm bandwidth (this tunable)</span>
<span id="cb15-1008"><a href="#cb15-1008" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>anything below this value is considered 0 (this is tunable)</span>
<span id="cb15-1009"><a href="#cb15-1009" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1010"><a href="#cb15-1010" aria-hidden="true" tabindex="-1"></a>::: {.column-margin}</span>
<span id="cb15-1011"><a href="#cb15-1011" aria-hidden="true" tabindex="-1"></a>::: {.otherapproachesbox title="Alternative Approach"}</span>
<span id="cb15-1012"><a href="#cb15-1012" aria-hidden="true" tabindex="-1"></a>While not entirely equivalent, if you run into memory issues with creating a spatially-weighted</span>
<span id="cb15-1013"><a href="#cb15-1013" aria-hidden="true" tabindex="-1"></a>expression matrix, you can instead spatially smooth the total counts from the </span>
<span id="cb15-1014"><a href="#cb15-1014" aria-hidden="true" tabindex="-1"></a>metadata column <span class="in">`nCount_RNA`</span>.</span>
<span id="cb15-1015"><a href="#cb15-1015" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb15-1016"><a href="#cb15-1016" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb15-1017"><a href="#cb15-1017" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1018"><a href="#cb15-1018" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1019"><a href="#cb15-1019" aria-hidden="true" tabindex="-1"></a>What we are looking for in SBR are larger tissue regions with low SBR -- not necessarily</span>
<span id="cb15-1020"><a href="#cb15-1020" aria-hidden="true" tabindex="-1"></a>individual cells with low SBR. Looking at @fig-qc-sbr we see a low SBR region on the top of the tissue just</span>
<span id="cb15-1021"><a href="#cb15-1021" aria-hidden="true" tabindex="-1"></a>below the epithelium. This region corresponds to an </span>
<span id="cb15-1022"><a href="#cb15-1022" aria-hidden="true" tabindex="-1"></a>area of smooth muscle and we'll see in the next section that this region partially</span>
<span id="cb15-1023"><a href="#cb15-1023" aria-hidden="true" tabindex="-1"></a>overlaps with an area flagged by the FOV-level QC. Since these low SBR cells (i.e., <span class="in">`log2(SBR) &lt; 0`</span>)</span>
<span id="cb15-1024"><a href="#cb15-1024" aria-hidden="true" tabindex="-1"></a>are concentrated spatially, we'll choose to flag them for removal.</span>
<span id="cb15-1025"><a href="#cb15-1025" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1026"><a href="#cb15-1026" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1029"><a href="#cb15-1029" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-1030"><a href="#cb15-1030" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-1031"><a href="#cb15-1031" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-1032"><a href="#cb15-1032" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-1033"><a href="#cb15-1033" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-1034"><a href="#cb15-1034" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb15-1035"><a href="#cb15-1035" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1036"><a href="#cb15-1036" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>SBR <span class="ot">&lt;-</span> py<span class="sc">$</span>SBR</span>
<span id="cb15-1037"><a href="#cb15-1037" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1038"><a href="#cb15-1038" aria-hidden="true" tabindex="-1"></a>sbr_breaks <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="sc">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="cn">Inf</span>)</span>
<span id="cb15-1039"><a href="#cb15-1039" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1040"><a href="#cb15-1040" aria-hidden="true" tabindex="-1"></a>sbr_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"&lt;= -1"</span>, <span class="st">"-1 to 0"</span>, <span class="st">"0 to 1"</span>, <span class="st">"1 to 2"</span>, </span>
<span id="cb15-1041"><a href="#cb15-1041" aria-hidden="true" tabindex="-1"></a>                <span class="st">"2 to 3"</span>, <span class="st">"&gt; 3"</span>)</span>
<span id="cb15-1042"><a href="#cb15-1042" aria-hidden="true" tabindex="-1"></a>sbr_labels <span class="ot">&lt;-</span> <span class="fu">paste0</span>(sbr_labels,</span>
<span id="cb15-1043"><a href="#cb15-1043" aria-hidden="true" tabindex="-1"></a>                     <span class="fu">paste0</span>(<span class="st">" ("</span>, <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> <span class="fu">table</span>(<span class="fu">cut</span>(<span class="fu">log2</span>(meta<span class="sc">$</span>SBR), </span>
<span id="cb15-1044"><a href="#cb15-1044" aria-hidden="true" tabindex="-1"></a>                                <span class="at">breaks =</span> sbr_breaks)) <span class="sc">/</span> <span class="fu">nrow</span>(meta), <span class="dv">1</span>), <span class="st">"%)"</span>))</span>
<span id="cb15-1045"><a href="#cb15-1045" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1046"><a href="#cb15-1046" aria-hidden="true" tabindex="-1"></a>manual_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb15-1047"><a href="#cb15-1047" aria-hidden="true" tabindex="-1"></a>  <span class="st">"&lt;= -1"</span>   <span class="ot">=</span> <span class="st">"#D73027"</span>,  <span class="co"># </span></span>
<span id="cb15-1048"><a href="#cb15-1048" aria-hidden="true" tabindex="-1"></a>  <span class="st">"-1 to 0"</span> <span class="ot">=</span> <span class="st">"#F46D43"</span>,  <span class="co"># </span></span>
<span id="cb15-1049"><a href="#cb15-1049" aria-hidden="true" tabindex="-1"></a>  <span class="st">"0 to 1"</span>  <span class="ot">=</span> <span class="st">"#D9EF8B"</span>,  <span class="co"># </span></span>
<span id="cb15-1050"><a href="#cb15-1050" aria-hidden="true" tabindex="-1"></a>  <span class="st">"1 to 2"</span>  <span class="ot">=</span> <span class="st">"#A6D96A"</span>,  <span class="co"># </span></span>
<span id="cb15-1051"><a href="#cb15-1051" aria-hidden="true" tabindex="-1"></a>  <span class="st">"2 to 3"</span>  <span class="ot">=</span> <span class="st">"#1A9850"</span>,  <span class="co"># </span></span>
<span id="cb15-1052"><a href="#cb15-1052" aria-hidden="true" tabindex="-1"></a>  <span class="st">"&gt; 3"</span>     <span class="ot">=</span> <span class="st">"#006837"</span>   <span class="co"># </span></span>
<span id="cb15-1053"><a href="#cb15-1053" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-1054"><a href="#cb15-1054" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(manual_colors) <span class="ot">&lt;-</span> sbr_labels</span>
<span id="cb15-1055"><a href="#cb15-1055" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1056"><a href="#cb15-1056" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> meta) <span class="sc">+</span></span>
<span id="cb15-1057"><a href="#cb15-1057" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(</span>
<span id="cb15-1058"><a href="#cb15-1058" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb15-1059"><a href="#cb15-1059" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x_slide_mm,</span>
<span id="cb15-1060"><a href="#cb15-1060" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> y_slide_mm,</span>
<span id="cb15-1061"><a href="#cb15-1061" aria-hidden="true" tabindex="-1"></a>      <span class="at">color =</span> <span class="fu">cut</span>(<span class="fu">log2</span>(SBR), <span class="at">breaks =</span> sbr_breaks, <span class="at">labels =</span> sbr_labels)</span>
<span id="cb15-1062"><a href="#cb15-1062" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb15-1063"><a href="#cb15-1063" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.001</span>, <span class="at">alpha =</span> <span class="dv">1</span></span>
<span id="cb15-1064"><a href="#cb15-1064" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb15-1065"><a href="#cb15-1065" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> manual_colors) <span class="sc">+</span></span>
<span id="cb15-1066"><a href="#cb15-1066" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(</span>
<span id="cb15-1067"><a href="#cb15-1067" aria-hidden="true" tabindex="-1"></a>    <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">7</span>, <span class="at">alpha =</span> <span class="dv">1</span>) <span class="co"># Sets legend points to size 5 and opaque</span></span>
<span id="cb15-1068"><a href="#cb15-1068" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">+</span></span>
<span id="cb15-1069"><a href="#cb15-1069" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">"log2(SBR)"</span>) <span class="sc">+</span></span>
<span id="cb15-1070"><a href="#cb15-1070" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>() <span class="sc">+</span></span>
<span id="cb15-1071"><a href="#cb15-1071" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span>
<span id="cb15-1072"><a href="#cb15-1072" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1073"><a href="#cb15-1073" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"qc_xy_sbr.png"</span>), p,</span>
<span id="cb15-1074"><a href="#cb15-1074" aria-hidden="true" tabindex="-1"></a>       <span class="at">width=</span><span class="dv">10</span>, <span class="at">height=</span><span class="dv">10</span>, <span class="at">type =</span> <span class="st">'cairo'</span>)</span>
<span id="cb15-1075"><a href="#cb15-1075" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb15-1076"><a href="#cb15-1076" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-1077"><a href="#cb15-1077" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1078"><a href="#cb15-1078" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1081"><a href="#cb15-1081" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-1082"><a href="#cb15-1082" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-qc-sbr</span></span>
<span id="cb15-1083"><a href="#cb15-1083" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-1084"><a href="#cb15-1084" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-1085"><a href="#cb15-1085" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-1086"><a href="#cb15-1086" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Spatially-smoothed Signal to Background for each cell."</span></span>
<span id="cb15-1087"><a href="#cb15-1087" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb15-1088"><a href="#cb15-1088" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 10</span></span>
<span id="cb15-1089"><a href="#cb15-1089" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1090"><a href="#cb15-1090" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>), <span class="st">"qc_xy_sbr.png"</span>, qc_dir, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb15-1091"><a href="#cb15-1091" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-1092"><a href="#cb15-1092" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1093"><a href="#cb15-1093" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1094"><a href="#cb15-1094" aria-hidden="true" tabindex="-1"></a><span class="fu">## Combining Flags</span></span>
<span id="cb15-1095"><a href="#cb15-1095" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1096"><a href="#cb15-1096" aria-hidden="true" tabindex="-1"></a>Taken together, what cells were flagged and why? @fig-qc-upset shows the</span>
<span id="cb15-1097"><a href="#cb15-1097" aria-hidden="true" tabindex="-1"></a>configuration of the flags. </span>
<span id="cb15-1098"><a href="#cb15-1098" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1101"><a href="#cb15-1101" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-1102"><a href="#cb15-1102" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-1103"><a href="#cb15-1103" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-1104"><a href="#cb15-1104" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-1105"><a href="#cb15-1105" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-1106"><a href="#cb15-1106" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb15-1107"><a href="#cb15-1107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1108"><a href="#cb15-1108" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(UpSetR)</span>
<span id="cb15-1109"><a href="#cb15-1109" aria-hidden="true" tabindex="-1"></a>filter_list <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb15-1110"><a href="#cb15-1110" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Low SplitRatio</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb15-1111"><a href="#cb15-1111" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(SplitRatioToLocal <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> SplitRatioToLocal <span class="sc">&lt;</span> results_list[[<span class="st">'SplitRatioToLocalThreshold'</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb15-1112"><a href="#cb15-1112" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID),</span>
<span id="cb15-1113"><a href="#cb15-1113" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-1114"><a href="#cb15-1114" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Low Counts</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb15-1115"><a href="#cb15-1115" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(nCount_RNA <span class="sc">&lt;</span> results_list[[<span class="st">'counts_min_threshold'</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb15-1116"><a href="#cb15-1116" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID),</span>
<span id="cb15-1117"><a href="#cb15-1117" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-1118"><a href="#cb15-1118" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">High Counts</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb15-1119"><a href="#cb15-1119" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(nCount_RNA <span class="sc">&gt;</span> results_list[[<span class="st">'counts_max_threshold'</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb15-1120"><a href="#cb15-1120" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID),</span>
<span id="cb15-1121"><a href="#cb15-1121" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-1122"><a href="#cb15-1122" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Low Features</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb15-1123"><a href="#cb15-1123" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(nFeature_RNA <span class="sc">&lt;</span> results_list[[<span class="st">'features_min_threshold'</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb15-1124"><a href="#cb15-1124" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID),</span>
<span id="cb15-1125"><a href="#cb15-1125" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-1126"><a href="#cb15-1126" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">High Features</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb15-1127"><a href="#cb15-1127" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(nFeature_RNA <span class="sc">&gt;</span> results_list[[<span class="st">'features_max_threshold'</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb15-1128"><a href="#cb15-1128" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID), </span>
<span id="cb15-1129"><a href="#cb15-1129" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-1130"><a href="#cb15-1130" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Low SBR</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span> </span>
<span id="cb15-1131"><a href="#cb15-1131" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="fu">log2</span>(SBR) <span class="sc">&lt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-1132"><a href="#cb15-1132" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID),</span>
<span id="cb15-1133"><a href="#cb15-1133" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-1134"><a href="#cb15-1134" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">FOV QC</span><span class="st">`</span> <span class="ot">=</span> meta <span class="sc">%&gt;%</span> </span>
<span id="cb15-1135"><a href="#cb15-1135" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(fovqcflag <span class="sc">==</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-1136"><a href="#cb15-1136" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(cell_ID)</span>
<span id="cb15-1137"><a href="#cb15-1137" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-1138"><a href="#cb15-1138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1139"><a href="#cb15-1139" aria-hidden="true" tabindex="-1"></a>filter_list <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">keep</span>(filter_list, <span class="sc">~</span> <span class="fu">length</span>(.) <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb15-1140"><a href="#cb15-1140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1141"><a href="#cb15-1141" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">file.path</span>(qc_dir, <span class="st">"flagged_cells.png"</span>), </span>
<span id="cb15-1142"><a href="#cb15-1142" aria-hidden="true" tabindex="-1"></a>    <span class="at">width=</span><span class="dv">10</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">res=</span><span class="dv">350</span>, <span class="at">units =</span> <span class="st">"in"</span>, <span class="at">type =</span> <span class="st">"cairo"</span>)</span>
<span id="cb15-1143"><a href="#cb15-1143" aria-hidden="true" tabindex="-1"></a>  <span class="fu">upset</span>(</span>
<span id="cb15-1144"><a href="#cb15-1144" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fromList</span>(filter_list),</span>
<span id="cb15-1145"><a href="#cb15-1145" aria-hidden="true" tabindex="-1"></a>  <span class="at">nintersects =</span> <span class="dv">10</span>,</span>
<span id="cb15-1146"><a href="#cb15-1146" aria-hidden="true" tabindex="-1"></a>  <span class="at">order.by =</span> <span class="st">"freq"</span>,     </span>
<span id="cb15-1147"><a href="#cb15-1147" aria-hidden="true" tabindex="-1"></a>  <span class="at">nsets =</span> <span class="fu">length</span>(filter_list), </span>
<span id="cb15-1148"><a href="#cb15-1148" aria-hidden="true" tabindex="-1"></a>  <span class="at">text.scale =</span> <span class="fl">1.5</span>)</span>
<span id="cb15-1149"><a href="#cb15-1149" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb15-1150"><a href="#cb15-1150" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb15-1151"><a href="#cb15-1151" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-1152"><a href="#cb15-1152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1153"><a href="#cb15-1153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1156"><a href="#cb15-1156" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-1157"><a href="#cb15-1157" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-qc-upset</span></span>
<span id="cb15-1158"><a href="#cb15-1158" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb15-1159"><a href="#cb15-1159" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb15-1160"><a href="#cb15-1160" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-1161"><a href="#cb15-1161" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Combinations of quality filters flagged in this study. Only the 10 most frequent sets are shown for clarity."</span></span>
<span id="cb15-1162"><a href="#cb15-1162" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb15-1163"><a href="#cb15-1163" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb15-1164"><a href="#cb15-1164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1165"><a href="#cb15-1165" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(<span class="fu">file.path</span>(analysis_asset_dir, <span class="st">"qc"</span>), <span class="st">"flagged_cells.png"</span>, qc_dir, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span>
<span id="cb15-1166"><a href="#cb15-1166" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-1167"><a href="#cb15-1167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1168"><a href="#cb15-1168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1169"><a href="#cb15-1169" aria-hidden="true" tabindex="-1"></a><span class="fu">## Filtering</span></span>
<span id="cb15-1170"><a href="#cb15-1170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1171"><a href="#cb15-1171" aria-hidden="true" tabindex="-1"></a>Now that we have flagged the cells, this section does the actual filtering and</span>
<span id="cb15-1172"><a href="#cb15-1172" aria-hidden="true" tabindex="-1"></a>saves data to disk so that we can use it in the pre-processing step.</span>
<span id="cb15-1173"><a href="#cb15-1173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1174"><a href="#cb15-1174" aria-hidden="true" tabindex="-1"></a>In regular analyses, I simply filter the cells based on the chosen cutoffs and continue</span>
<span id="cb15-1175"><a href="#cb15-1175" aria-hidden="true" tabindex="-1"></a>with the pre-processing step. In this case, I am interested in showing the downstream</span>
<span id="cb15-1176"><a href="#cb15-1176" aria-hidden="true" tabindex="-1"></a>effects of leaving poor quality cells in versus filtering them. To that end, I'll</span>
<span id="cb15-1177"><a href="#cb15-1177" aria-hidden="true" tabindex="-1"></a>save two annData objects: unfiltered and filtered. In both datasets, I will add</span>
<span id="cb15-1178"><a href="#cb15-1178" aria-hidden="true" tabindex="-1"></a>a qc flag column to keep track of what QC conditions were flagged in any given cell. </span>
<span id="cb15-1179"><a href="#cb15-1179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1180"><a href="#cb15-1180" aria-hidden="true" tabindex="-1"></a>To be efficient I'll make use of bitwise shifts to create the QC flag column. </span>
<span id="cb15-1181"><a href="#cb15-1181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1184"><a href="#cb15-1184" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-1185"><a href="#cb15-1185" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-1186"><a href="#cb15-1186" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-1187"><a href="#cb15-1187" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-1188"><a href="#cb15-1188" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-1189"><a href="#cb15-1189" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb15-1190"><a href="#cb15-1190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1191"><a href="#cb15-1191" aria-hidden="true" tabindex="-1"></a>qc_masks <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb15-1192"><a href="#cb15-1192" aria-hidden="true" tabindex="-1"></a>  <span class="at">LOW_COUNTS =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">0</span>),  <span class="co"># 1</span></span>
<span id="cb15-1193"><a href="#cb15-1193" aria-hidden="true" tabindex="-1"></a>  <span class="at">HIGH_COUNTS =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">1</span>), <span class="co"># 2</span></span>
<span id="cb15-1194"><a href="#cb15-1194" aria-hidden="true" tabindex="-1"></a>  <span class="at">LOW_FEAT =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">2</span>),    <span class="co"># 4</span></span>
<span id="cb15-1195"><a href="#cb15-1195" aria-hidden="true" tabindex="-1"></a>  <span class="at">HIGH_FEAT =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">3</span>),   <span class="co"># 8</span></span>
<span id="cb15-1196"><a href="#cb15-1196" aria-hidden="true" tabindex="-1"></a>  <span class="at">LOW_SPLIT =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">4</span>),   <span class="co"># 16</span></span>
<span id="cb15-1197"><a href="#cb15-1197" aria-hidden="true" tabindex="-1"></a>  <span class="at">FOV_QC =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">5</span>),      <span class="co"># 32</span></span>
<span id="cb15-1198"><a href="#cb15-1198" aria-hidden="true" tabindex="-1"></a>  <span class="at">LOW_SBR =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">6</span>),     <span class="co"># 64 </span></span>
<span id="cb15-1199"><a href="#cb15-1199" aria-hidden="true" tabindex="-1"></a>  <span class="at">BLANK7 =</span> <span class="fu">bitwShiftL</span>(<span class="dv">1</span>, <span class="dv">7</span>)       <span class="co"># 128 (not used)</span></span>
<span id="cb15-1200"><a href="#cb15-1200" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-1201"><a href="#cb15-1201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1202"><a href="#cb15-1202" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb15-1203"><a href="#cb15-1203" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb15-1204"><a href="#cb15-1204" aria-hidden="true" tabindex="-1"></a>    <span class="at">qc_flag =</span> (</span>
<span id="cb15-1205"><a href="#cb15-1205" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 1</span></span>
<span id="cb15-1206"><a href="#cb15-1206" aria-hidden="true" tabindex="-1"></a>      (nCount_RNA <span class="sc">&lt;</span> results_list[[<span class="st">'counts_min_threshold'</span>]]) <span class="sc">*</span> qc_masks<span class="sc">$</span>LOW_COUNTS <span class="sc">+</span></span>
<span id="cb15-1207"><a href="#cb15-1207" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-1208"><a href="#cb15-1208" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 2</span></span>
<span id="cb15-1209"><a href="#cb15-1209" aria-hidden="true" tabindex="-1"></a>      (nCount_RNA <span class="sc">&gt;</span> results_list[[<span class="st">'counts_max_threshold'</span>]]) <span class="sc">*</span> qc_masks<span class="sc">$</span>HIGH_COUNTS <span class="sc">+</span></span>
<span id="cb15-1210"><a href="#cb15-1210" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-1211"><a href="#cb15-1211" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 3</span></span>
<span id="cb15-1212"><a href="#cb15-1212" aria-hidden="true" tabindex="-1"></a>      (nFeature_RNA <span class="sc">&lt;</span> results_list[[<span class="st">'features_min_threshold'</span>]]) <span class="sc">*</span> qc_masks<span class="sc">$</span>LOW_FEAT <span class="sc">+</span></span>
<span id="cb15-1213"><a href="#cb15-1213" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-1214"><a href="#cb15-1214" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 4</span></span>
<span id="cb15-1215"><a href="#cb15-1215" aria-hidden="true" tabindex="-1"></a>      (nFeature_RNA <span class="sc">&gt;</span> results_list[[<span class="st">'features_max_threshold'</span>]]) <span class="sc">*</span> qc_masks<span class="sc">$</span>HIGH_FEAT <span class="sc">+</span></span>
<span id="cb15-1216"><a href="#cb15-1216" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-1217"><a href="#cb15-1217" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 5</span></span>
<span id="cb15-1218"><a href="#cb15-1218" aria-hidden="true" tabindex="-1"></a>      (SplitRatioToLocal <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> SplitRatioToLocal <span class="sc">&lt;</span> results_list[[<span class="st">'SplitRatioToLocalThreshold'</span>]]) <span class="sc">*</span> qc_masks<span class="sc">$</span>LOW_SPLIT <span class="sc">+</span> </span>
<span id="cb15-1219"><a href="#cb15-1219" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-1220"><a href="#cb15-1220" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 6</span></span>
<span id="cb15-1221"><a href="#cb15-1221" aria-hidden="true" tabindex="-1"></a>      (fovqcflag <span class="sc">==</span> <span class="dv">1</span>) <span class="sc">*</span> qc_masks<span class="sc">$</span>FOV_QC <span class="sc">+</span></span>
<span id="cb15-1222"><a href="#cb15-1222" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-1223"><a href="#cb15-1223" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Condition 7</span></span>
<span id="cb15-1224"><a href="#cb15-1224" aria-hidden="true" tabindex="-1"></a>      (<span class="fu">log2</span>(SBR) <span class="sc">&lt;</span> <span class="dv">0</span>) <span class="sc">*</span> qc_masks<span class="sc">$</span>LOW_SBR</span>
<span id="cb15-1225"><a href="#cb15-1225" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-1226"><a href="#cb15-1226" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb15-1227"><a href="#cb15-1227" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-1228"><a href="#cb15-1228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1229"><a href="#cb15-1229" aria-hidden="true" tabindex="-1"></a>py<span class="sc">$</span>adata<span class="sc">$</span>obs<span class="sc">$</span>qcflag <span class="ot">&lt;-</span> meta<span class="sc">$</span>qc_flag</span>
<span id="cb15-1230"><a href="#cb15-1230" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'n_cells_before_qc'</span>] <span class="ot">=</span> py<span class="sc">$</span>adata<span class="sc">$</span>n_obs</span>
<span id="cb15-1231"><a href="#cb15-1231" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-1232"><a href="#cb15-1232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1233"><a href="#cb15-1233" aria-hidden="true" tabindex="-1"></a>Remove some of the intermediate matrices and save the annData to disk.</span>
<span id="cb15-1234"><a href="#cb15-1234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1237"><a href="#cb15-1237" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb15-1238"><a href="#cb15-1238" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-1239"><a href="#cb15-1239" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb15-1240"><a href="#cb15-1240" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-1241"><a href="#cb15-1241" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-1242"><a href="#cb15-1242" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb15-1243"><a href="#cb15-1243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1244"><a href="#cb15-1244" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'qcflag'</span>] <span class="op">=</span> adata.obs[<span class="st">'qcflag'</span>].astype(<span class="st">'uint8'</span>)</span>
<span id="cb15-1245"><a href="#cb15-1245" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> adata.obsp[<span class="st">'liana_connectivities'</span>]</span>
<span id="cb15-1246"><a href="#cb15-1246" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> adata.obsm[<span class="st">'counts_neg_spatially_smooth'</span>]</span>
<span id="cb15-1247"><a href="#cb15-1247" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> adata.obsm[<span class="st">'counts_spatially_smooth'</span>]</span>
<span id="cb15-1248"><a href="#cb15-1248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1249"><a href="#cb15-1249" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-1-qc-flagged.h5ad"</span>)</span>
<span id="cb15-1250"><a href="#cb15-1250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1251"><a href="#cb15-1251" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(</span>
<span id="cb15-1252"><a href="#cb15-1252" aria-hidden="true" tabindex="-1"></a>  filename,</span>
<span id="cb15-1253"><a href="#cb15-1253" aria-hidden="true" tabindex="-1"></a>  compression<span class="op">=</span>hdf5plugin.FILTERS[<span class="st">"zstd"</span>],</span>
<span id="cb15-1254"><a href="#cb15-1254" aria-hidden="true" tabindex="-1"></a>  compression_opts<span class="op">=</span>hdf5plugin.Zstd(clevel<span class="op">=</span><span class="dv">5</span>).filter_options</span>
<span id="cb15-1255"><a href="#cb15-1255" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-1256"><a href="#cb15-1256" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-1257"><a href="#cb15-1257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1258"><a href="#cb15-1258" aria-hidden="true" tabindex="-1"></a>In addition to saving a copy of flagged data, let's filter out any cells with a</span>
<span id="cb15-1259"><a href="#cb15-1259" aria-hidden="true" tabindex="-1"></a>QC flag that is not "32" and save that "filtered" dataset to disk.</span>
<span id="cb15-1260"><a href="#cb15-1260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1263"><a href="#cb15-1263" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb15-1264"><a href="#cb15-1264" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-1265"><a href="#cb15-1265" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Python Code"</span></span>
<span id="cb15-1266"><a href="#cb15-1266" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-1267"><a href="#cb15-1267" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-1268"><a href="#cb15-1268" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: show</span></span>
<span id="cb15-1269"><a href="#cb15-1269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1270"><a href="#cb15-1270" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'qcflag'</span>] <span class="op">=</span> adata.obs[<span class="st">'qcflag'</span>].astype(<span class="st">'uint8'</span>)</span>
<span id="cb15-1271"><a href="#cb15-1271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1272"><a href="#cb15-1272" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[adata.obs[<span class="st">'qcflag'</span>].isin([<span class="dv">0</span>, <span class="dv">32</span>]), :].copy()</span>
<span id="cb15-1273"><a href="#cb15-1273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1274"><a href="#cb15-1274" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> os.path.join(r.analysis_dir, <span class="st">"anndata-1-qc-filtered.h5ad"</span>)</span>
<span id="cb15-1275"><a href="#cb15-1275" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(</span>
<span id="cb15-1276"><a href="#cb15-1276" aria-hidden="true" tabindex="-1"></a>  filename,</span>
<span id="cb15-1277"><a href="#cb15-1277" aria-hidden="true" tabindex="-1"></a>  compression<span class="op">=</span>hdf5plugin.FILTERS[<span class="st">"zstd"</span>],</span>
<span id="cb15-1278"><a href="#cb15-1278" aria-hidden="true" tabindex="-1"></a>  compression_opts<span class="op">=</span>hdf5plugin.Zstd(clevel<span class="op">=</span><span class="dv">5</span>).filter_options</span>
<span id="cb15-1279"><a href="#cb15-1279" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-1280"><a href="#cb15-1280" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-1281"><a href="#cb15-1281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1282"><a href="#cb15-1282" aria-hidden="true" tabindex="-1"></a>Save project-level results. </span>
<span id="cb15-1285"><a href="#cb15-1285" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb15-1286"><a href="#cb15-1286" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb15-1287"><a href="#cb15-1287" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "R Code"</span></span>
<span id="cb15-1288"><a href="#cb15-1288" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb15-1289"><a href="#cb15-1289" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb15-1290"><a href="#cb15-1290" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb15-1291"><a href="#cb15-1291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1292"><a href="#cb15-1292" aria-hidden="true" tabindex="-1"></a>results_list[<span class="st">'n_cells_after_qc'</span>] <span class="ot">=</span> py<span class="sc">$</span>adata<span class="sc">$</span>n_obs</span>
<span id="cb15-1293"><a href="#cb15-1293" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results_list, results_list_file)</span>
<span id="cb15-1294"><a href="#cb15-1294" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb15-1295"><a href="#cb15-1295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-1296"><a href="#cb15-1296" aria-hidden="true" tabindex="-1"></a>After filtering, we have <span class="in">`r results_list$n_cells_after_qc`</span> </span>
<span id="cb15-1297"><a href="#cb15-1297" aria-hidden="true" tabindex="-1"></a>(<span class="in">`r round(100 * results_list$n_cells_after_qc / results_list$n_cells_before_qc, 1)`</span></span>
<span id="cb15-1298"><a href="#cb15-1298" aria-hidden="true" tabindex="-1"></a>%) that are ready to analyze. Note that we did not do any gene-level filtering. </span>
<span id="cb15-1299"><a href="#cb15-1299" aria-hidden="true" tabindex="-1"></a>My primary reason for leaving all genes in the data at this point is primarily</span>
<span id="cb15-1300"><a href="#cb15-1300" aria-hidden="true" tabindex="-1"></a>because -- while we do not expect every gene to be highly expressed in every cell -- there are rare cells that may express genes that would considered</span>
<span id="cb15-1301"><a href="#cb15-1301" aria-hidden="true" tabindex="-1"></a>"below background" in the rest of the data. My general approach is to keep all</span>
<span id="cb15-1302"><a href="#cb15-1302" aria-hidden="true" tabindex="-1"></a>targets in the data and identify which of these are highly variable and use those</span>
<span id="cb15-1303"><a href="#cb15-1303" aria-hidden="true" tabindex="-1"></a>genes when appropriate (_e.g._, PCA, UMAP, Leiden).</span>
</code></pre></div><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">

<div class="cookie-consent-footer"><a href="#" id="open_preferences_center">Cookie Preferences</a></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>
<script src="https://unpkg.com/@popperjs/core@2/dist/umd/popper.min.js"></script>
<script src="https://unpkg.com/tippy.js@6/dist/tippy-bundle.umd.min.js"></script>

<script>
window.addEventListener('load', function() {
  // Initialize Tippy.js ONLY on links with the .link-preview class
  tippy('.link-preview', {
    content(reference) {
      const iframe = document.createElement('iframe');
      iframe.src = reference.href;
      iframe.width = "400px";
      iframe.height = "300px";
      iframe.style.border = "1px solid #ddd";
      iframe.style.borderRadius = "4px";
      return iframe;
    },
    allowHTML: true,
    interactive: true,
    placement: 'auto',
    arrow: true,
    animation: 'fade',
    delay: [300, 200],
    maxWidth: 420
  });
});
</script>




</body></html>